<!DOCTYPE html>
<html class="js" lang="en-gb"><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
        <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1">
<meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=2.5,user-scalable=yes">
    <meta name="citation_publisher" content="Springer International Publishing">
    <meta name="citation_title" content="A review of data mining using big data in health informatics">
    <meta name="citation_doi" content="10.1186/2196-1115-1-2">
    <meta name="citation_language" content="en">
    <meta name="citation_abstract_html_url" content="https://link.springer.com/article/10.1186/2196-1115-1-2">
    <meta name="citation_fulltext_html_url" content="https://link.springer.com/article/10.1186/2196-1115-1-2">
    <meta name="citation_pdf_url" content="https://link.springer.com/content/pdf/10.1186%2F2196-1115-1-2.pdf">
    <meta name="citation_springer_api_url" content="http://api.springer.com/metadata/pam?q=doi:10.1186/2196-1115-1-2&amp;api_key=">
    <meta name="citation_firstpage" content="2">
    <meta name="citation_author" content="Matthew Herland">
    <meta name="citation_author_institution" content="Florida Atlantic University">
    <meta name="citation_author" content="Taghi M Khoshgoftaar">
    <meta name="citation_author_institution" content="Florida Atlantic University">
    <meta name="citation_author" content="Randall Wald">
    <meta name="citation_author_institution" content="Florida Atlantic University">
    <meta name="citation_author_email" content="rwald1@fau.edu">
    <meta name="dc.identifier" content="10.1186/2196-1115-1-2">
    <meta name="format-detection" content="telephone=no">
    <meta name="citation_fulltext_world_readable" content="">
    <meta name="description" content="The amount of data produced within Health Informatics has grown to be quite vast, and analysis of this Big Data grants potentially limitless possibilities for knowledge to be gained. In addition,...">
    <meta name="twitter:card" content="summary">
    <meta name="twitter:title" content="A review of data mining using big data in health informatics">
    <meta name="twitter:image" content="https://static-content.springer.com/cover/journal/40537/1/1.jpg">
    <meta name="twitter:image:alt" content="Content cover image">
    <meta name="twitter:site" content="SpringerLink">
    <meta name="twitter:description" content="The amount of data produced within Health Informatics has grown to be quite vast, and analysis of this Big Data grants potentially limitless possibilities for knowledge to be gained. In addition,...">
    <meta name="citation_journal_title" content="Journal Of Big Data">
    <meta name="citation_journal_abbrev" content="Journal of Big Data">
    <meta name="citation_volume" content="1">
    <meta name="citation_issue" content="1">
    <meta name="citation_issn" content="2196-1115">
    <meta name="citation_online_date" content="2014/06/24">
    <meta name="citation_cover_date" content="2014/12/01">
    <meta name="citation_article_type" content="Survey paper">
    <meta property="og:title" content="A review of data mining using big data in health informatics">
    <meta property="og:type" content="Article">
    <meta property="og:url" content="https://link.springer.com/article/10.1186/2196-1115-1-2">
    <meta property="og:image" content="https://static-content.springer.com/cover/journal/40537/1/1.jpg">
    <meta property="og:site_name" content="SpringerLink">
    <meta property="og:description" content="The amount of data produced within Health Informatics has grown to be quite vast, and analysis of this Big Data grants potentially limitless possibilities for knowledge to be gained. In addition,...">

        <title>A review of data mining using big data in health informatics | SpringerLink</title>
        <link rel="canonical" href="https://link.springer.com/article/10.1186/2196-1115-1-2">
        <link rel="shortcut icon" href="https://link.springer.com/springerlink-static/481091012/images/favicon/favicon.ico">
<link rel="icon" sizes="16x16 32x32 48x48" href="https://link.springer.com/springerlink-static/481091012/images/favicon/favicon.ico">
<link rel="icon" sizes="16x16" type="image/png" href="https://link.springer.com/springerlink-static/481091012/images/favicon/favicon-16x16.png">
<link rel="icon" sizes="32x32" type="image/png" href="https://link.springer.com/springerlink-static/481091012/images/favicon/favicon-32x32.png">
<link rel="icon" sizes="48x48" type="image/png" href="https://link.springer.com/springerlink-static/481091012/images/favicon/favicon-48x48.png">
<link rel="apple-touch-icon" href="https://link.springer.com/springerlink-static/481091012/images/favicon/app-icon-iphone@3x.png">
<link rel="apple-touch-icon" sizes="72x72" href="https://link.springer.com/springerlink-static/481091012/images/favicon/ic_launcher_hdpi.png">
<link rel="apple-touch-icon" sizes="76x76" href="https://link.springer.com/springerlink-static/481091012/images/favicon/app-icon-ipad.png">
<link rel="apple-touch-icon" sizes="114x114" href="https://link.springer.com/springerlink-static/481091012/images/favicon/app-icon-114x114.png">
<link rel="apple-touch-icon" sizes="120x120" href="https://link.springer.com/springerlink-static/481091012/images/favicon/app-icon-iphone@2x.png">
<link rel="apple-touch-icon" sizes="144x144" href="https://link.springer.com/springerlink-static/481091012/images/favicon/ic_launcher_xxhdpi.png">
<link rel="apple-touch-icon" sizes="152x152" href="https://link.springer.com/springerlink-static/481091012/images/favicon/app-icon-ipad@2x.png">
<link rel="apple-touch-icon" sizes="180x180" href="https://link.springer.com/springerlink-static/481091012/images/favicon/app-icon-iphone@3x.png">
<meta name="msapplication-TileColor" content="#ffffff">
<meta name="msapplication-TileImage" content="/springerlink-static/481091012/images/favicon/ic_launcher_xxhdpi.png">
        <link rel="dns-prefetch" href="https://fonts.gstatic.com/">
<link rel="dns-prefetch" href="https://fonts.googleapis.com/">
<link rel="dns-prefetch" href="https://google-analytics.com/">
<link rel="dns-prefetch" href="https://www.google-analytics.com/">
<link rel="dns-prefetch" href="https://www.googletagservices.com/">
<link rel="dns-prefetch" href="https://www.googletagmanager.com/">
<link rel="dns-prefetch" href="https://static-content.springer.com/">
        <link rel="stylesheet" href="basic.css" media="screen">
<link rel="stylesheet" href="styles.css" class="js-ctm" media="only screen and (-webkit-min-device-pixel-ratio:0) and (min-color-index:0), (-ms-high-contrast: none), only all and (min--moz-device-pixel-ratio:0) and (min-resolution: 3e1dpcm)">
<link rel="stylesheet" href="print.css" media="print">


            <script type="text/javascript" src="https://cdn.krxd.net/userdata/get?request_id=7ace42dd-3d25-4062-857b-cace58685326&amp;pub=bd339c69-af54-4a21-b4f1-654bcfcd83ca&amp;callback=Krux.ns._default.kxjsonp_userdata"></script><script type="text/javascript" src="https://beacon.krxd.net/optout_check?request_id=43ca6e30-01e9-4d74-9119-dd57dd0039e6&amp;callback=Krux.ns._default.kxjsonp_optOutCheck"></script><script type="text/javascript" src="https://consumer.krxd.net/consent/set/bd339c69-af54-4a21-b4f1-654bcfcd83ca?request_id=6ffd0939-d6b7-4b1b-8caf-e3ab6a39760c&amp;idt=device&amp;dt=kxcookie&amp;dc=1&amp;al=1&amp;tg=1&amp;cd=0&amp;sh=0&amp;re=0&amp;callback=Krux.ns._default.kxjsonp_consent_set_1"></script><script type="text/javascript" src="https://consumer.krxd.net/consent/get/bd339c69-af54-4a21-b4f1-654bcfcd83ca?idt=device&amp;dt=kxcookie&amp;callback=Krux.ns._default.kxjsonp_consent_get_0"></script><script type="text/javascript" async="" src="//cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><script async="" src="//cdn.krxd.net/ctjs/controltag.js.8f9c5605187855d5a137991abae6f700"></script><script type="text/javascript" async="" src="KDqyaFZ_.js"></script><script type="text/javascript" async="" src="analytics.js"></script><script async="" src="mathJax.js"></script><script async="" src="gtm.js"></script><script>
        window.Krux||((Krux=function(){Krux.q.push(arguments);}).q=[]);
        var dataLayer = [{
                'GA Key':'UA-26408784-1',
                'Features':["leaderboardadverts","abtesting","reprintsandpermissionsabtest"],
                'Event Category':'Article',
                'Open Access':'Y',
                'Labs':'Y',
                'DOI':'10.1186\/2196-1115-1-2',
                'VG Wort Identifier':'vgzm.415900-10.1186-2196-1115-1-2',
                'HasAccess':'Y',
                'Full HTML':'Y',
                'Has Body':'Y',
                'Static Hash':'481091012',
                'Has Preview':'N',
                'user':{'license': {'businessPartnerID': ['3000197460', '3000201946'], 'businessPartnerIDString': '3000197460|3000201946'}},
                'content':{'type': 'article', 'serial': {'eissn': '2196-1115', 'pissn': ''}, 'category': {'pmc': {'primarySubject': 'Computer Science', 'primarySubjectCode': 'I', 'secondarySubjects': {'4': 'Computational Science and Engineering', '5': 'Mathematical Applications in Computer Science', '6': 'Communications Engineering, Networks', '1': 'Database Management', '2': 'Information Storage and Retrieval', '3': 'Data Mining and Knowledge Discovery'}, 'secondarySubjectCodes': {'4': 'M14026', '5': 'M13110', '6': 'T24035', '1': 'I18024', '2': 'I18032', '3': 'I18030'}}, 'sucode': ''}},
                'Access Type':'subscription',
                'Page':'article',
                'Bpids':'3000197460, 3000201946',
                'Bpnames':'CAPES MEC, Universidade Tecnologica Federal do Parana',
                'SubjectCodes':'SCI, SCI18024, SCI18032, SCI18030, SCM14026, SCM13110, SCT24035',
                'session':{'authentication': {'loginStatus': 'N'}, 'attributes': {'edition': 'academic'}},
                'Keywords':'Big data, Health informatics, Bioinformatics, Neuroinformatics, Clinical informatics, Public health informatics, Social media',
                'Country':'BR',
                'Journal Id':'40537',
                'Journal Title':'Journal Of Big Data',

                'doi': "10.1186-2196-1115-1-2",
                'kwrd': ["Big_data","Health_informatics","Bioinformatics","Neuroinformatics","Clinical_informatics","Public_health_informatics","Social_media"],
                'pmc': ["I","I18024","I18032","I18030","M14026","M13110","T24035"],
                'BPID': ["3000197460","3000201946"],
                'ksg': Krux.segments,
                'kuid': Krux.uid,

        }];
    </script>

    <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
                new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
            j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
            'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
    })(window,document,'script','dataLayer','GTM-WCF9Z9');</script>

    <style>@font-face{font-family:'Source Sans Pro';font-weight:400;src:url(data:application/font-woff;base64,d09GRgABAAAAAERcABEAAAAAiPgAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABgAAAADYAAABAA0QDckdQT1MAAAG4AAAHPAAAGXDUveN5R1NVQgAACPQAAACJAAAA4PFn1ldPUy8yAAAJgAAAAFcAAABgWrSUW2NtYXAAAAnYAAABRQAAAebzMPm1Y3Z0IAAACyAAAAAoAAAAKA2ZAPpmcGdtAAALSAAAAQIAAAFzBlmcN2dhc3AAAAxMAAAACAAAAAj//wADZ2x5ZgAADFQAADCcAABd7CMplIhoZWFkAAA88AAAADYAAAA2/hSz4mhoZWEAAD0oAAAAHwAAACQHowOhaG10eAAAPUgAAAHlAAADxKaoJjpsb2NhAAA/MAAAAd4AAAHkOIhQ1G1heHAAAEEQAAAAIAAAACADCwJAbmFtZQAAQTAAAADXAAABxiK6PDBwb3N0AABCCAAAAgUAAAM5bFBIb3ByZXAAAEQQAAAASwAAAEuWBPrreAENwbEBQDAQAMADgH10RmAMHSUyaqbJ36mwAmqj1qayx8rhVLli5faofLHyS2o5VgWBWga3AAB4AdxWVXDkRhB9ZqjyYZiZmfknzJzvMDMz/YThJ8x4zMzMzGz2eu8sr9ebs867UeqlqzM1JU2dDOHkaaHV06Se7mkhD0A5TsAFyL/goituQI/7b3n8QeyFQghIXbe0XPn33fHog+jx2y+g63kikQ/ItzzvErk7TDV64CLcj1fxPSZiMWqRzsvPK83rk3dA3lF5p4hUD5ShL/bE3jgZp+ImPIy38A6+RD8MwExZ21Os7M0kTqaHU5nAHeiDO3EIvmQG/TgJw7ALilTiVGaE2y7cMcLdD6Wis0O4rbgN5aJ3hOidLBJtIrFCJE4VvZOZFglfuOsxiL5wS8XfngxExhNOu3COFbk9mRUfIqfcVuEehgKRmYsBqjlTtNQWioWTwDBuQZ5oVSNffueK7lqR35MJoRLi0RP+nShBGnuqxCKcrLyXmMMr8s3f5W2xiF0qdzsGZ+I4gI0cxDr5HYcDcalIVOAvhMT9p4JprgIk8j8ZnK6/cyO8pFwb/wTbsrNy9QNwtuXN5BaAq7lFqKY/wYePPS29HiZ+gDUcLitl9NBtcD1bAWaUDpjhAsNvp68eFcLP8VuRTYonheoUMsHFnKvaDmxsgbHtM2NtBVYiDfXBWiAcPXMMrLSPnYApih1m6UVsJ4Wus9my2lyMHlYzoLUNMCG/DtRm4PD03gXr9a80lJPOa1tioj630B5TMPFzm65nAWhUGsVKWHA7wCxFQ75p1tJ3PfJ11nE+RzMAWMnNHAmggGKXCc3FdpXy6PNuDuf3/N5qjgZYzX4cyc0AF7JypznJRSuMvvKdijS6xXZPfLYALpiN2nZ3mj71nvplCgrD0b10dYTeyFpDZRzbWUSg6w6tO9GOQnQTzHCc5rsK4Idczw1cyCTHMcWpfFUrdaU5fXZhv2hcAKtEfjhXcCXncIzm0MJ5vjRT9GFrkSsc2RqbOd/Jclz1/mSpSoA5vdJ6Pw1hHM66neoH6kkr2Y2Ys7UbMyjjD8JLa0VPAbiOn6tUJaD1faBwo1Whq5zKBVwv0iu4io2hqshRd5dpE2mOtfS1Izw2sUq1t5oIdrF6uciJU8nKmLMqGZKLARe6Ha29GQtbg0nNTLFcbg81ObW5i1ubqlcaf74zAXRWuVyGboIf0mOlnhJ7cZvQHzvr7TBnIJ90Vsbr79cczC0dzAafCc1JEYroh08CQyvFSnoMnHxrnYhW1mYz/PTH6fc42z0HcWl4LwwXbAVwnnAuZ8acKafhQOyFI3AQKuzkMv840WYayjlUJ4lbCZlIT7WjB06ABTdHu9HK9wDsBGijT9/46hGpkxRUw3RAInKq2RpkwNUmyiytBJMmwk7h9oZ6znaoMc/0spmuuqsxE5GJsDU2M6DErbO8BcW2W12UscrqDLHvKGuYsfwIuIjPWT+dgr6+IfdllvPYbJ64UfnqSVHBNM5zc4OD0Fc721NpP1rfkenY1tW+pB+RLgdc2Lq0/kJ1kgrVd87EcBB6RHbAUkxFLdgOrEWnYNKNqmsabAtFuI5B6KkD1oZPFhym/7YHXbAFcOd8yPaW6PucM1uPZcbaDjp+OiaMnOYzKs2tIY++iX1PuFAtpvH3o1CfxcxLjucmejoxK7XrnrTTu0PQ06sW/x/E1/GGP2aJPmst530upc+RANfoOfEgugA2mvPEw98CW7lNAHdwB8Bp+s7WTJ/bALZyHH/p/FQQ+SZ63M5GNlueF3lDT+uZXKDzq4U+5Z4/a3X6MTYD1QAbmDE29N7BNXF1i98JNqrvqd23yAQT7qkVD83JfPekiJ3VyucULkcMuF4zuwrdhGplO35KrnRmooK1dorEVgibmInrtzgN/IXgO1zQRcmgaxXAKks9z7mGqmRa+TFgNu5dIH6m6fQMYmZaJ36YdONla3R2q2WffmS2dla/yUhe8lHxa1vzrLaJAQBh9Py2jc/WGnHapIlt21a5jG3bubVcQqaLds7z1mNUGVUNBe2o6LpD1b2ecaPnvOBBJ7zjEe/73Eu+iuO+iRO+i5N+iFN+itOGjPrWuAmTZjW1TGnr6Jo2ox2zemnOvAWLBpbssWzFqoI1bNqybRfsYDcKUVVSVlF1VM1Z6hrG/6ZpMkV0IuTtb5oWIyyb1bf6N5vWUgRAVZ7/ZtN+1Qj1oGBLxZm3EesxZNghow5H0REJHHBQsi/CCNj7d8YMu8hozBqPBdOxYC5WLMeC1ViwHnM2YthmjNiKku0YtRNluzGmEPOKsaisalktltRjXDNWtWJCV09bP5YNom1PLNsbbfuiY3+sORDrDkbdoZh0OBqOxJSj0XWW8826MBZc7FILLnOlEVfFmKtjyLVutugWd1h2p3usutd92u73qGWPe8q6pz1jyvNesOSEk2acclrfy7HrlSh4NSpei6LXvaHqzah5K6a9HTXvxLT3o+WD6PkwWj6Kno99punz2PJ1DHwbA9/HwA+x46fY9Ets+CO2/RmrfwE9c0UteAFjYGRgYOBiSGCYx8Dk4uYTwsCXk1iSxyDFwMIABP//g+XFGJgdo1wVGOScg0IUGDRCgrwVGIygsoxgmgkIWaEsZgY2KIuFgZ2BIyczPZHBCCeZn5zDYIFMMjBATAVjJqBZUlAeBxCzMZwCy6oxsMDFpEA0WCUPQyWQVmaoApKGYHXKAFZmI2wAAAB4AS2HNwGDYBSEv3svPVMc/BpiIH2KgUwRxEL1hAVGFOCADtcPV6mELVhkP+C7tP5cFcDOW/eNm23GswfEisf7+eEOXW9NvwOvdAkoB1BtBRDwATBlDPMAeAFdywOMHQsAAMB5iD9qY3Oxatu2bdu2bdt2G9W2fd7YdeNe7YkxiCKGbGIiyCYeeYgKAnH58K/CyqqinS7GmeSce9K99DFSNNI8ejWaFE0JsgV5ggJBQlAuqBGMCA4kZEnI/vEj+FegiHKqaq+r8bY4777Qq8+72fedNcgV5Pu8y/6xIx/fowEfL/HxDHwcw8fRfCxFRraM+IeP4VwIJ4Ydw2lh1TBX2qm0YyKgDFqgE3oZgD24jBBAkjR/i5lnsaMuuG68RbbaZr/bZltnlt3mmOuB+x5a4Ijt9trhsS2e22mPXZ54ab4rLjrvkn0OSHHIKWs8td5hqc555LgTEj1z0CbpjjltrRtuumWhluqoq576GmjorkaaaKqZ5lpoZYjR+uirn/4GGOieQYYaZrgRRhpjlMYGW2KpFZZZbuUn89p7YwAAAAAUAEQAUgBWAAAADP8zAAwB5gAMAgYADAI+AAwCfgAMApAADALIAAx4AV1QRVoDMRROvT0BUsv7cKjs2WCZwd3ycLcz4E7tLI9d2M3BsFeXkfz6xYwQiVkjQkv60+UqoHH9vRihol8iJDz7e0kjXAkprUtFrgMW7gQbg8DMk5A2eXrsFd2FMiMz0ycZacuLwxPy9pSQg9MMpiWJVX3J45oGGsVwjZ4iDvM83uI83tI8GeQZriozMLLFE/xwyZeYleTpXdLLmu5VmEYVhgGkRc6SJkeFAZFb/tpOGW8u2yt7DiQ4GmQSLM+yqmk0TAIzmbLqArrPZMKZLqxqI5wWwyVajdGKYUTR4ZuwjOt+iSOGLggXjS7oAgwDKl47lJhd1ZYKA2DyHyTqa+QAAAAAAAH//wACeAGlvAd8G8eVODyzAAGwAURddOxi0TuxKKwAOymxSGIRSRValuTIiuz02BfX5Cy3FLfUixOlfne+L7Yv5VxkpRenJ7p+5zi5Oxel5+78pccCvzezgyWIgI7+99dP2Bk+7M689+a1eW8WqA2tI8QVuXcgDdKhdmRCdnQCPY7Q5Bpcsp6zCKGeyjoDcATAIW1l/QnUBh1ktvQ/gXRqT6/2DKx3FnUg/MsnUDf7GwbqOYssAMv1WkWzaI2aRbPemsWSCVurWArgdfx0LfT5m/HaTYu7CtrS7tO9Yz8dy3PvuHgVXv5RLY+/86PahdMnTpzGuzauvXYD0IORNXDxcfcA/ma0XMfeSJA1NmDfRQBdSKsCzARgZuTAN0aKeifpUVQ7s2dRO0UVRzM4KhmxCfOSXrLKVgm/OmARerQmwRJ46r+qKz/Y+ME67jxevrq//+ry8fGvcfdcfBXBTYMOAG4hwM2KvCiMDu7IWysBWBVkJPjGSlGQsk8AByVkU3nnAYQADj0v7VmhF1b4Kef9nN2m09v92G4zcklszZeKhUgS29XOgS/llwbFvoGZvaGxQ+W/+9uVXfPzV7xmZePQ6mu4eyyR4WTvHpO2c2EiN5Nz4j8bKJey/1z7RnV8uP+3CGFU2Pwl5+Heh0TUX6ejjaDd1kCHjwB8CpfbALcuiqUPeg4Fy0KFk/MOXp/BUtAI+Prhz1KZN8KfGa6QWjhVHTrgy7qriexcye8rzGZz00I2erRaOTmfxDN7blrL9gpFrxiePDpcOTYViUhyOp9ZvRnYTHldAF53ACOLO3K6gwA6CIZnUSfBjvLVXOehQ+FdmbHMfOCj7zkTW3jN7gfOvBb+zXL3PPD+hetX0hNvvvHG0xeBLQiTefEPYN5ONL/jrJ0E0En5QufTUL3Qwz2ddG31WeAjxaYTvjUwDbGLdtku2aE9gK+r/dvvfod7azdw90x9Z/oX00id+x+p7I/vOHc7AbS3mrt929xbM5olOt/vf78Ek/1gqvYUqq9/BNY/jOZ3XH+RAMTm9Rfp+pNZfXC3SGf1wayel5KJolSUzUwu0vteMTZx0Jd1jebze/sCvvK+olS1veVfhJNjo1ctpBTBkP1Ft1QXDLvlY5O1n4lJkA3GJy5P12jXJazRE6hdXZl2wFPD5ORxpD0P90GrP0+kBayXbJbgKh1Ywr3Ly7XvAsNqP8fWi6/Cxdo3lXnRQzCvBoV3nFdDABoyLx0TxlpiBgSjwOYv8Z3wfA9K1p/Xk9v16vOqqaMSraEcLfsxcLAIyIEwS8HoQUefVCws25OTuesHbo9Ko7P4q7XexGyfcJLxpgpzdL2EfdISgPalzKmBAAwK9zh4touaUxA7ZIAepUzGst4qafR2IFGDzRvf/vnhL7+acOwxvOv3tZfjlVvP12X6W4BPG0ruiA8TPEozx3TXTPRkCQdqdwH7HptG9XW/EcZyoVfuOJaOAHQNABcBuJC2hQY9oXhKSptR7Znhbh3t2WBkF6PXystlsp7mQqks6TWSJiqBaYb1PXpzTpu56egevVZbvCl7c0GrMcwDFz542234CEjOydTLU/fV/hqv3ged2rtVOtLUj+y8RmwFWjmVFhKuhY6B0UH8jYpzBhPB2cJ2/8+ntRr9nv2/mNZq9XsAz+N35E8V8BJg+v7b8ycKtb9GHLUNV4BtsKKoanuZqVURUlGmq2ajOtUBlsFP168N9LxILYGfIzY4mq/A3xmubhLwXOXE7nh894lK5cRsPD57opKalr1eeTrF2q7c+s179ty4msmu3riw5+b13K7I9LHK8NGJcHiC2ITpCPMRduCjEfGtLWaTRj0BMo5YQNAFlsDELIHqz+rogquQrKrT+HrpQFWSRtZL1yx9bv/S8spV3D229O5SYS7vqr2IZ0anpsu1PyDKt0PcaynfCmi2jk+ITB9qwMdNAG6CzznUizDgrk0CtKfuM0I9DDOAPY4yxDZRI7rFQUwMaSSa0TSyWE/NRPnNKhNZi39syvsOXucrrVWDdYaLw/sL/HLJIRrxbPWKmUhk5ooq4+3FZ9s6Vif0pbVrZ+rsn379ark9ldBp635ZAJ4bkPPS4zNV6SjRGGHUQYk2Mvf4BOIJjIktYzzWSKBu9UDI+tAr5hcX51+B7WeuXfp2fmU4uHf8L/F6dWS0cPFT2F77KXePNTUlzx8y154hOpYAW/tfsBbylo9wEBwcDVjmCCC3ZWtJwOhAOSofDkDNR2FhuMUBMNUSwypkscJ81dPpiFQHsF9DloIsziHpMnl0xuyL8Rt8LuIMlGfTg8eltH9XpjxuFSK2w958yB4YWCxMXpO8O5wezLkkl60z3uGNlUPhkaw7HSr5RDntFF1WQ6wrEC+HohOyr5AmtHnh8iKsgR6o2xbOq7SpjoWuAAe0ICpRbUSapKJIooJnnsT/8eQyt3tq6uKjSvyxCnqfofZ1aEdt6iaA7gaAjQBsChP1dePNBFZ1WmANVpec8T5B7Es494aq+/P51WoIzM+Ng3tzdntu7yC+pXbznqvGfb7xq/YALnQzoAFcutHIn4r+mpGr7wIQ6qbLZpatfszLFQwGXCOtn7R4rFqL13xy5fzlgMGZweMDA8cH8VFw0lQbEfd7mFdE11z6vIwHrSMnCuAJgN/CTKTG2gQP8bTnRaIq/yq22/DWKNsWcD38/sl2W7u209l5cO5gl7NL227pmNx7x/Fj7SaD1mA2XH4MyLq/dGWpdLKIr6jdX3i50gMb/5bIrgj8r72errcJLi/Q9Vb9ECZ44j+Ko17Kn5oIwKSQpoNnXZQgK4A6FYI0Mq/QUcWyxiqp+y/pO588tN7Nm7RGe9fq2qc+d+hIt6dHa3R3XbaGl/DwA3zS50vyD9Q+V/vEJ5yy3y87P0HldHMTLt8HvE1o5NLxbkbTpKj0NuTs0pe/sr7S4zdrzUHT8upXlmv4vrOhqXB4KnS29ooa4RlRwP+hMfow2h7yNftuRSV0io85D7f1AIaqFpaxCGqIIcpJ4/Xak/jDtb/Eh2tDec451XvxR9N0LzwANuwx/ALyoChI43Y6DDsvCJNGFuCY6IIQLOFZ1d7a4QYThbsAFqBo2aEXpD2CqqQEnxDuUK+j00crmOyozFJUov6nbHPQyN6oGbj7fe18xJ/c5RODRwfX907oOWlfurJvX29IqkrDy7Ljpr0rZj4VtAu2fkf86oO1fx7wxWdmEiFRuN3g8EeB1kWg9Q/c15ADRdAtTZpnaY7utojvIYAelXjVyKvRXQ+Txh4a0yl7bJ4S2QM9N+3poBdSCY8wI2bXE4L9XN2iQ7aAOADVD+sXFzlxIZ6d6XUFB+YSwV2CRj+2KFSdfjlsNwZ6g4PT3Ne+uOaDCP1gX+XohBTw5a7y8p7emXRsLOseU2xuBuj+OH6hxT7ccMn78PJL7sMzkZGVbG7SFXdkfWJfjHdE+0Sp3xEW96Vyy8PSC8OXjQUlvmh1evKTicREr8dnK/mCwdHLqAxmAb9/g3WxIQndsiOGzCJuLZSFACxI29I0G+CbNhattiELXRcLUOOkdBnVzIhBzYwQqgPbaI0W5T+SRIX8rASrkagI2qUJvcY3mzx09crUSHDIH+j3++SI44XBy8ZCfvfSuYt9/d746evXjorBGlmWCUIvXY9vUJ0LoX3N9LY0513wDc9iSvgGMO3ZhjPAmXwx7KlM6UU/blikcCFCcNdnwsMLicyuXrem9njb6pQ06PEJC/nelWoIL1f73dmIS/vC0GVjkjC0XPBY9hyym0s2vzS6cfnsHnu0P6zIlB8uVe4fQcn370gBWyBLq8QCXQxOJcJAe53nyV+wPGC8MCDrwzS3YJaK5WKhRJYEpz3eiRMnFsfGJnfxcTN3a9fG/HjtMP5IpW1+dn+7nubQ0psC/hnw1wxx2Tq6uo5fmMwebsBvkgAmkW4rNhuC7ycpn4ey9WxPGBCKn4cPQBIUMgQQGSAyQHaze86iRcp9nshJhaPRJSg0ONMClZoia5hoAV1+XNd6I/QcgxgMHwtkrMoILOR7/eyY1B9z5HPLRrHH7ZyJiBZX0OqM+vLjYZtk/OvjZt6bGw1bIuZuSyi/sbISHDnYn95d9H3dHvb1DCWTQz2+8GI0Y/aErEG3ps2aEAJ9vLYz6gtkurW28VRsJO3Ut1XNhUC0mnS2Gzo9Nt5fHg4OZzzmYA4v9PiTntLgYMmT9PeQtV+Ey79QfV1DOxlN1ay29hg6lqKEFaem0aSaRotqGuu+wEw5oTdLi4saaU6en15M5SID4UWwfMfF7JGN2rdwfHI4Eq59EHFoHCZ9hPtb1AlDptH2rYKh5X5NR2QmCVMGI4qQka0AKP4NTlF0wmcMbPBuP+/0+Zy8/3jNTcPGzX/djNJ5TKDF5SZfYmjOOuhgHrIrgHmYdVWDWB0RABb0N2KwtDi0iw91m9ud3SFngqDCiyK+cPHZ3WuGtoqmLZ/m3lRHqr4mF2BN1JyMmoLZeU0uPSeDGnIyVpKTiUJOZnGf5rsHP3L2wF0H93Ffq/kx+lLt+z95+RupfSC+Fv0r4GNCSfSnpwcmqRxxyANYMVmR4uKisV2r13caHOZsjPvaxXfYeriKVttXQoxm7n+oX3vPn5JDw85hM4ttWuf5OTWObmc9Iq+A+vmzyE5ltl119jq2nVM2QwFGEC9TOea3SbN+q10EDyLMpvqqpsh8enbXYgo2b4upLFzwhSkp25tKyHURn619kDUIqXoI9NvQFZdOP9PD1jkeIJcpJlNKRoxFJaZZKSn+0ryqlRTr7Uqp+rz/BlxpvqfZY7QIoGFJVOcgbfdrzEiWmzIWmXBlXzq9rxKutyu2cCEQKIRt9ZaGIcGxy4aHLhslwccQiUTAKScSk3kPRCaIxogy/gPg6QA8b0Hb/ZahOQNraC516VrWZ9rgWSsVIgeA2tW9P69WZNyqKIVUqqOKPwEHCGzfHiZuixF1+o+SNRD3JLMzOSVMHA9yYkWNEQPVUPyj3Kf6vDGIEvsrl09IfvfIVpAYFJQYTMb/RmVJQjfvuD5MVgzNIZeuyeezGIxFXrAJZdFYPf4yXVL8JZdl4vp3jL9GRM4wtqT1zSZaxV8+X+bi/EP7PdHt8RfTmw0aHyxeut40WwXzli14WlES0zYloZGyqiGu6YSP7+mymQLjLlCRtUypY0arzVdqXyO5FtCNWwCf3q08kp3MZm+YPkMAme15pAyyUyTshIcUFoJnMk15pGirNBK/lUWaFMZTuZwtELO5C3E3n6xEM/PeqKskZJIWf9TuLiY8fHo8WdiQXuEWUoLVa+8xtPNiKuAD+RL4gtWZ8JpdVpOhwyHmgv5CjI/4FPvkBLomuVeAJk828bmlOaI2p13dyxLWaqC1AbQboDSrVMUyrTfZHA2BbfHIzFzH5C23iIluf1ePLQuBu2gkIeEdd4zXXkj3gsPUdzoTQyHAaRZw+g6+wGxm66VW1b157Vs6TjPJqrKqdTv1mKAWWzU5meFq3tpLFGeXkmArqWAEZ+WFaXyh9i+Tw9Ek7NLdc7HckQ2Ff11w+Sq+0CI/1bwrYoCd81Pb8zxdK4c6+U5tp6Pz0L4Hj8DsPwvNSNJMCNtqbrJuCHGPwryQn7r0edlCGi41PyWy0r8JQDaWn0JIuMT8lHN5xmDUaw097bN759rNkJMyGqYX/vz4VLupHaAdE4SqH0pg7seD2KX0xiTouXGbBEnoSan2IuWxES4fxxda5KdaWL1LzE9B7MlqO5jVdsrRhhSQnm/ITz309vXRTmc3WYvB/W8/sz7d7TZqu51dY/trPzpli9vg/6n/+dWrHSm7Pcm/msnFZpbKhQ9NoW2B2x/vvZpxbtzsdCMDQ4+nxqKKNY3cNmqoq+k12NoTRVPnl1aOdbpAZmwdq3sf6xZSfWGzt6vnNq5tMB3CP6z9f8J0UJwRcPfFF0KVtLutbVTBNQCXu/GFFrksw/9hLktiuawARs/huU2EDbV7Uvja8VTtThL0b24q5zo0eS6CtADQo//EppbwZ3eAP78NLqvwH26Dp1X4T7bBT6nwnwEcQEAS0rTRuqQd3YC26eelF0a71SAcemqBuU11ngrDupn/IUIH60rvs2XrgRs2SzoTbrOLRUgw4fopAXyvtlga5oTaW79zbA2v4ftrP/nRjzCH506duubc9Oc+p5wbAPoKm0lyjgPo01H6foFXKd30HAPln4Hx29sS/uwO8Oe3wdMq/CcMTv49Su/vUsZHt1D4Xrhf1wB/VoFv/gHgnQ3w5xn8RYB3aNIq/CcKXKmR0nU2s3WeonBam6TjWBhdcy3hz+4Af34bXFbhP9wGT6vwnxA4SFF+8xnuFPcw6gR5yaFb6xKTIvKQapAYPwH4FQFB0EnBcqtRAdzppwLTxXowAAiHkaVkyd00ElXvE1VYSO1l2LcsjqDZiCyOKJk/Hm+lLvTRcj2pwZf5866RCKkHmnuC8lxqRvYG8Z8FglB6HS6sS9XMlED6lSunq5nXuX2k2JrbK+RLpNp6ZAK/OfnaGVKFFYWl0MHUNbQ0WN0nkZonMMcPemRBHrTSZEVa1WAV2+FktoOciNKwimxHtp5qMqrxdxepV8G2u+HwDN1dQEbGDNFS4YG/wB+gmiHUlmZfv5jcf921udy11+HZ9z9Az9WkV65fuOlOznCari2tb1FZcDLZGWsJf3YH+PPb4GkV/hMFrtQl6P1eJvu/oLZGhMtueibDg9brPPIQlnhQT4vMdb1a52EmhHDJgxwql9pZtf1x5ARb3A6tC/gUZnVTyihr/ewTCdLN2EjPO73znfvY+ac9e+iZp6ezF7k55QwUHk1qsQATN9TVe5AL9beqgjWl3s/BfZgsb/Ic4qDnhB5NE23V0+3S1iGspWv+uKD+CltmV6kwn3e+66mR6elS7UXCZ1r/4L4B/PRTfv4H7kat4M/sAH9uG/whFX5hG/xhFf7jbfB/UeE/JXCkIXDup/gFoHQETaP70U7u3EkAzgZAngDyhFXbzQYF5NVTXH0oz7Lhfdn6acsADJSisEAWnmMnBJUU7TDtOaE3QXsB6E1Tp8J2ZBDJQHlGU09yqJlnjRqgR6L0Wtwq87QpyYO39IaCe0VDBx7YJ/NDwSOl0pTJty+fnsw6NbWvYFesFJD6PH5hLj+86M1fka3su+s9BigCxWZ8nDzYJ+/Sn06EoiFr0esiFR7sdMTLuVxZGFopeC2RSsrJm/tt/nIonZkZU4tDx2f3Ls0o8UJmM0nqIrAGAl2Dn+MeBof8PF3jIJOJ9pbwZ3aAP7cN/rAK/zGDEy/0aXr/Cejp0H9cjxSZgOYnDfBnFPjm76F5sQH+HIP/EiHMkfEZ/MfXq3kh9FF8ocUZsuaExdYZssVF2JbW3BQPmleispxkshxS6flvikeK8cXUEv7MDvDntsEfUuEXtsEfVuE/JnCkAfiPIXd0kWRRUGZrjxwnJMQbaPISgJf5RejEqV/kmHWLIy+V8biafy2zVH0Lz9bo2Bzg2T4ZLQWKUbvTE+4LFCMOHodcPpJn2jPUGy4p3ciMLx86HIUkQ3w8G415eifj4xnsFlZLJN3UO+QbFVZKJAPlcg37GF2/1Ig0JzaExtGr0fZIfYuuPgLo28qTKgVVE+uBTp6He6CVwddFQJtlSmckCxC4L6Xq7QjtRaA3TulvzqDpKbk7arFOLSLp9H9uC+X9/nzIVm+XQ0N7U30bbu96KT2Vc2lqX8CueDkgDYD+LuT7N7z916X2DEncxyp98m79C02ZtwThS2Wof7RJd0f63wB5uiNUZWlOmZO4v0dldLwpCtjKKZcJoNwyvQs7HuhoqFAIVCgk9vcTKMt6SuiQZ8dMS8y5wMevrYc8amZFWyyoZRpQoJdtRIcTdq/k9aXklM+RGAwnJgKiW/YlEhTiEb3xRGlwkdNsvJqPFf2RYirudgoud6gUE0ox3s0POOw5wRfxuoS0FMylMkKkd36h1kV1g+YtqC7lme71tIQ/swP8uW3wh1X4jxmc7inp/SVlHPRCA/xhFf5jAkccKsI4z3NPIRvq24o3ZMJrGVlaVJXPkXwmSTMnn0A5lKaVEJ96ZigN0msHKc5BKz+ttCUab1AOR7eKzta6hma4unZmMNFWujCgqdgUzAfNGItjQbEQsjqi+VNDc4H+Pb2JhaGQUJ5NRabcxnLkbc6bC0OR0JRwl7PPV9iVMnq7U35najAIh9qdXHj4gWuHjk1HheHVkrw0KAjByqHIrvhoTCjYUp5dcYRYrpidJ4ii1dY7cdXSqkxR8wutcqb1TLBBrebykvmPTgnwxeb87yIWIG+0K0/Tv9KYiLenf7+teecXl9zhevrX61565/b8L63ncgehnhtDN+wYc7ACqqVlyB1W67kWWDyRJPDgYwGoSa3t6uFjhG87zit3maENQ+s4ryQwm6q+2yq/eqJn8BX7k9vtc09CIfgbUAqenXNmLWa/FIp9o6kqrFSGR+bWOgyVbjlXYTViQi/YkkmgN4nuRNuDzeaUVut6WfNJbOWsR7eSxz9P/oKsELSmLIyjZvgJ1A/QIIupJZbniFHiwyyd2YJyILzcmPnkJlma8+F6KdwChfFPdkycPh0kWVCdkvDE3Q218SpwRc2IEv8jAA8WufcB1h7wqm+t86FAiCo0ROFBAggibUtHW0UBqsxBNQ1qAEiVvRsTRAX2bgysNih2Ae5y0btS8K2Z3jWk9qbYk8w1KdGlEVtt4JM0iqXFFFJX/y0fzY6ngu8qy0ZMeq+ySVlPsOLLrUtSdt5sdMQDFp8YTuGbpblY5PBAelfRZw9lXTbeEesLDk+4E0WPsJr1Z03ztr5osOL5G09aMHtdieNF7zBOewb8poAYdviSpYtf7wh7g3FPbjwmljJRc3w+RM4czgwIpWzcJm9E40f6R68p8IJBsRMpsJVf4d4MZPpQsSkTbmx51rlLLSg6FCvQUBnefs7Z8bvLr7rq8iNXXXXEmxYtFjHt9WZEy2TXgx/60AMPfOhDD44FRq+Ymjpe9furx6feejOV/T1weYy7CWYZb3pHwNhC1OtZOD1go4VeXc1xD5H0XC8t0hC2g6ES99z3utSou3q69p4J/M9FA99z8Uml5hKCy5eBBz403HTKzNhCrOr6wp9XcoEmNZ9eZoKgnhUj617Bg9guxmZeNRcbGo5NeLOxA9X1l0UnL+t397nO1o73zrzu2FK0PJUWsqniiZX0wctPDnLaaYKXG9bms4CXjJaads3Gl6y9eFBGfemrS91J25QqDEA6m6owUpEum4q5+u6Vpu7OohO++f6B3fbcQr88mbQMyrFxXyq83p+Ykr0jBV8+Yhf751KRaiHec9qfqZa82aClahFSrmwp5CtIaUd8KDqw26bVe+IlMTGRdZlcgoXKnxkI/SdY73bAb7XJuhtbFA7OkWwREcWkYqa1oLiIrDYz353MevWcV+BW4qnJ9o6mtNskO6RqixjkAec0nLbLbXwnvnp1ofaf2Hxo9YWz7mk3n+KfGvv4x0v4PbXXkRw5rMEXAL8Imm6yxcYW1YT6GvCohxXk+SzgBDj2sEo14/sWv62iXWI+VOX2qHe+j9QKd1fErC9lx9XfmPkMnBgsBUJD84lIVY73vAngkZI4tttm9WJ5+jNdRmfvbKGwt+wzOQNmRbdzgPvXQX66UWZLm5IE2WQr7Fkew0B5K6rxT1J9oyy8hbYfW5W4h6LuaA57cJcz5jcPDPmnBcP0521lvziU8WL/3ujSMX9hIprdOxiMTazmCrOeoi0sC9WZaHrjcvwD3uMdOTEXTh49lpkt+SPzr50/cvtKNOoDrFKk3gTrYEJyK7vQZKXgS2byFTsgVzRlct7UqNFLn9A6k+PZ2hfx/f3TCat25cC7br9pbFqeOX3nXxxW3nOMA9++BnwzIR+aRKtNuc6t+UYJYJSuu1plG2XZiu6sIoMpZicBQnwqkQRoS08zN8LKoEZcrmD1VBXd7KgnsphgNKrja+3hvK9v0purBiPy3wrSQDE94IwX/9MuJnmp5M2NRRNjWZfZFTDZYwGrvzSXSc+XhY95UqKlkg+kRI/RPXmrb0TKjfZlfXHB9eVAMSl2aV3Tcngo6XSlq2FXJiZ0arsC0bwQqmTc7uwI4lAM+PL5Jnli4qNyRRUwKk+IaJAqT7omebJSJ8p8hV5RC6WizghXo8qfBkCQZob5OAjVAohTUBGn2NJRvzwZye4bFIk4yfNuXIpmNo6QQxXVmW/yXu/olfOR1HZxivgIVqObvwQKXgPW0olanADUkAWqezXAYFRyOSXJ6ZLyPknyeSSJjJHd3I+ehDHMdIymV7mI2oNzhGXVSY2nt7KJnEbvlEIuITw1fJ21Esc+rydQSI8ecZM9DcULf4+LottgipNIBy2B07m2wW+nOr4C93+LOwNzmpEbtTiq0q6wOipDZVLfQNFJbonr7A/U6eJGNjZqt71DpU4ZewZ9C7/QcmzD1tgBLEFNrmzbovNxNjaMS8Y/sLGBr1OH5lAE5OgN3P9Lz3XHmk5AUrExq2/IhLcOXkQq2mbvquUbKfqgzuyPBvnenDjIZ4XF0sJhpxQVLdrDQRfBwxV8MlotZtIZT6w/4EmGsiuzYimXzfdLKmrUZg7R8w6tcDMQ3DoAN3MTbthh1NrVsr8iv9poI0dwe1sPQS7XKw7wuTpygkW7UWcSFih2WXesz+9JSrkW2BGZ24v/Bj3AfRLWpBOREpwWAly6vCYsLWnWND15bv+hQ39FZQbuxWtwbwTdRfbHcP04HaMfP4jOcRD0giafQ21s7wujGLGdsZPlRM9FIhqdTfTzfrFU4O519SWx2+X2ZVJDa34Fnyrg8wXuHBurEyBtylh/JPjVSFjTZhf8vC9YLuB/cvWlsNvp9mdTQ6t+RPEFvPApwCuC7iH4wvXjFA5z4Fdz5wB+L4HDlcE3X4W+gG7gCARd/EIdhl8N+8Qog9G7oVXeXRhD/4YfRTrGOx3lXblY5vW8Pn1H6JprQt+v3O9/3xmvQlsJ7n+m4f425f5ombfrow9eI915p/Ti+/z3V+73UvzJ+PhauD+C3kXwhOsZCi8B/I0U/m4Ch+sZOn4cfQ+bsBtpKO8wwoirVw6k+IUL2F1VxmX3cTr0Xvbcv+PHsLP5OfrGcxzfVXsNdvaz+7jf7HQf952LebiP2hysAZvzJk5PbYuJzhFDd2MDfiOyUto7AEZpb3qv8T18LGCxBGK8I07a+N1mIeXxpEk/7fGkBDM7E/Dn+EebZxkeBIIpHqToi39UC02Xy+xc5l7czj2NjHROHZuTVw4o8/QkMq//58rUVEUe6O8fkD9+/AenT3//CueR7732td874qRjRDb3ou+rY+hhRjoG8a40C0MkMRqhj9OB3sMevuL7p0//4LgyxsbmldjMfRme5pl+6ADf5hUyw2fjU69//ae4L0MZKUvfb74S/Z36HGK6cA5pVYp52I7I8Hn76z75yZNcLfviR8lzQTZfL0qq/CHPeaH10V4P9Ew0PxWDlrzTQbjHHGVjCA+cgu0GDSMc6u8nkD9kbE4PDmg0GoMVzGHQamgfGEx74h6dVxC8Ok/8ZLZv3tObybhMHoexYLR7jfJaeb4vm+jNLMu9cru2vZCTlzO9CYVO3A34yiq+vRRLB/SstCcSqii+HdCKFF+eFsDUs+ZKNazINhtkae00EKhXzIr9cU+bRxDgEnen/CmfsT2Ud4k2A9Bg9AHgJEUtVwDU5F6KWrbbL0fLa7IviwtGh8fkymR6PVHZ351V9Llj8wjexz0Jcuhsde6XA3+NZSxh+1DtV0Pcky++F545tKlD7+J+gTREnuj6U3sLi3hoYID7xYtOgBlh3OWXHtcK4xoHcecgHRZxDePq2cht6sggJWYYnYxfK87Xcvi8gv8A/ii+EnJ6RmRrijcstFRCLFNZ5qk5i+qP9l9uell7b/sJ0+V90elp/IWNWMZ11SlnJrbh27+f4JDe/Dzsex8EiZqmEpulek/Wb4KtJMT8FcwiVJrhrB/S0xLh8uNWeedSQ3ECj/gCsaHpgE+IDU/POWHvF45XpivxbMTY40wOSqW1QN41EIul48Mzw/FcxGS2RvsiuXnfF4wRMZgSvZaeWEBMCj58yuYOuz0hvz85nMkNOsxRN2T27REx7fTGfN6I35cYzqRG7D0xiFVD9iBP6Itvfhyfp79X0IPCf+r9aACQ5TdLGmbjoMSOb1heXi0dHo+ERtcKKXgRX8B5V3Gxf3Cf7Kgt0r1C7+avuMe4t9Gzw5NIBi4OkOiScjENvQINFdI09q/vSEZJxfpp9VDmnzg7zDcdkGw8XdhbWLtmdPT164XC+utHR69ZK2RTs8f7+4/PpurtywLyeCgx6JRscXd5fLQk5Bx+byUdGct78fjcdfszmf3Xzc1fv5rNrl4/P3ZqPpGYPzU2emouHp87NZqcyrtFZ9Rmnx0cm3dZE07RI+8isqhFPK7ie7ivszPTw3SXVMiSwhe0QJ/t6fr+MU7yik/X67l5VKC1L1IUPotKFJaFNk9g9Mne80odbFDhzp94k17DogteikqkyK4cAfiH6AR5YXIiGp0k7WQ0UI7xfKwcEEpRno+WYnNCxmbOZKR4fDxp4b246ivMpFIzBV+9DfOJwZA0mHQSMQ0NJnhs8IUsAd3S1NTLgibLOJEvefMKjZe7DdZ9GVVh5QfVlY9Cz0R7u6DnoD0/id5or41ZzCdQECDsLEvTjoedxtw6bazXyU2kUzlQCl9UKKpY2UzJ46999/LcG1Yy4YHpYLy/jXONiPKkJz+dcseMXGc56l92ZqNOaerKifGbjg6m9149klkMOlZO7n7jZWVsiPntabfP6Lf5s6L5m8c/+rpqbvW63WPHxoORiC86MxlbGk/wrtjXXu8tzvcOndidKB+/a23XaxbiLluvnDl422ou4Q18gLcK+SHgEdko/QPsGT0ogO5vqkM0Z/QA0HwYpOWvBmhYBugc0kNPSzVMz1L3pDVCq2U5IA20FpYLsrHWyeBedp+PZArBMdOPrKcfu0Q/Uhk+Vri8du+qZeUQX+RvdRad+0jfVXSedpWespy2fKv/HQOPwb+Bd/R/61vfwm3veAdSzsngx9HPNBVOjx5DCPoEFsOP4xHNboA9rsIkiHurFHZWhSXgvr302ScUGOEl7Jvm8X3gL3i6N9FlXzI9a2mRni3VkwyOxyf27ZsYX1wcj+RykWgu13XV8WOnTh07fpW8Z2Fhfn5hYY/yTjxcf4Zvg9UKshmbzgEqcyFFWVmyVXpwRUzZCiO4kNBZTbXrqF9BeARwd6MUopWH5gooO/J0KSlVyOUW0lMZZ0xyJcwRfjw2OO2IFgP/T+2X4fhoKeuIpESnFEyMVLyFYp9AY0vKuyrMn0AVmJ+vzx8l80fJ/CRUaafVgig7kMJeZmTHUNpRx7a0aUQqQtOUNG30iQuWfEDIjPTL5ajgiIHDGo670kFbxC/Kjojsze3mL/cEQsGgFI/5vUGPLeiTjJ6oS8oaDb1RTzLQkxQRB/givBf43waIzCHirVr8ogTa9osSnPoDP23nlWNSevbuYhdTANN5uu2RTJgnCVGpSOo3T8OZ/+FnBwY++dGZgXdNXb3vaObKK8WL/w2yGMYcTnDLnA59hsqmH/5eoH9/lsol+x7WMQg2kIgjSZKeQzoWv8JfW6fnWTUVAcTNcqUIuZWj1cqmM4O31tuH6cH0aL5cpPH/HenSZVdit+Qu23jXZHSonNdrbfnuaHJ+GLAqJIvlDiFh7070eKZGvhy2ePKxlfaI1+2keDK8kRWW04PIWTKCa+Ovt7hb+5wGtXk0UE44nYlyoN4mMplEIpvFHJ+sxmLVJF9v5dFiaWSkVBylc8fxQ1jiHiJz01jLptqwDui5offS8/6lKy3ZbFLaVW/FREKED37IAjliP+TbLcGs35cVLeDVwplMWMrQed3oBJyXSSMRFVGs6Z3Djnrp2w/TK6/p9lIGVDGN0Ek8WVLOmIBcUy+rpwWuqL45SHGPmK1W80hqIlf1DfATqVGzzWYeTU3wA76R7MSQVUiBIxWsVgEcakqwvmOUH5CkQX4sPRW2WqbSY/ygJA3wo+lJqzUy+Yy/N2S1hnr9rFV+r+s+/Aj3M5RGh9F23TW2eK9pS4lhHwmgqPLuO/MJVlX2fPC8lX4Xhe8SoBZ4W4QLypHBTdFtUbb7MY9v8BbjmVK8sqsSX5oZyh0QIsJ8pq9MAXvGhnLjHq7L4y+kQhlJzE6WZg51XHGsLenr90b7srFiNJiq9o6vdx4/1pag+cgiQrQOl0LrTe/ZGFuciwKDq75Tw1H8Q1nF/4WANkktandmt97AgVge9tvKRphofAY3v3qjvvoLhOOv1F4tl7Wa4+OWDjM5DVOadeesuWCp39T+gfs4jeCfns57L+s3eowFrTAsJCZy7qSQ5f2jBWFQ6DN6TMWpgFbZt/BweQpy9z50Gm0/fGFsfq/F2PKdXC3ysfcwXKxHPAeIsHrMqPs8tMy1O1nq3dlTfw/STL9Vf6Kt4RcvgGA7fKgRlMlHc/WCydejNftMC8sfvesjH/nI4oc//OG3fYS76WPiuCSNix+rfT47mjtzJjeaxSOENvArP4V1K6Krm85IGVtU4pU9QYaqW4ZVrdLQdrA2/LQSrYTVt+2sDHVBfQk531iPJysZKRSHcBk6Fc3WahInqbPZvVhZUxIr4p+Ky+X0/FC8fWQQ48GR9kR1qTe94O6154WxfV21Wg3jp/6pe74SLrv3pXrFkYODSW27pd3QY0i1DR4eDabFrFs8utqrNxo6rO25lSMhidqY4c1/x0/CHkiHwqgMti3I3NE5omVIZPl/xbUHodURt8QWzQ0tT/SOmvcANFTp6GthTNWAoKgZ4gqsK49U+j/vm4kWyBvkgdLuRCE1wsdsabdQsOHvd6+Nja91t7uDmhNXtoWHlouFxaGw7uUnuKCjYHFwtQc4j4X+fso65Mz/wL2vxX6dmg4M+3U9lvFf4Zf31d7VxT1wcY3Gc/XnwO89Sf4mMYXGAeNE0aPkFD+0T+JX0wzfo/jrCNF7RuAcw8/pPY9t/pjdc5De8wR+f8M9v2u6Z4re8zgmWLG5uBfpPWfVud7IxnmQ3EMzLlfi73IpWsvfipKVWFfHXL+BnCgnsi4rcv+vjz46Cv9xsvKlL1W+xMbBSfxdfLJl3jCIH6ot4ZMVcl8e3YQ/x9lRJ81gtNHwiTiPaLmeg8jff3/q/vem4Pre+z/2gTPpD3wgfeYD6Q/S9241gO8RwFem8awHiTCXCUa1U8nRqf7R1eAft0JZTUPiPwp4/Y4XBJ4PBBboleeFD1JEv+Z38n4/7/TXvqO0/p9WKD+D6AtA4z9yEfRtdArp4PouxsNbACc70GKnWRkD0AS0QYuVGCVsN2F7ANvDwIl/uO22f4BJrn5geiGvzS9MP4BajAFPsjEwe1MdtxWruJjFxTb83doSGQSCg2TjIByMgdhamlAf+qMzTTsvL9zFDpFuX2g9MIkt9qOUNbnKZz9b+cRoRcF5CHvBv54BnN3NmZLGV+CtEv19W+kfd71nZrRXmx+d4c7Ubvr4rbd+HMYIYYy/jd8OqxlF5yjdJoqgieUDNKpfMjGttyqBc0NGg5iukGdM8GfjEbsjYs409HF/LuEW3D4hmIvTlsk8kdU2VVa1TFYhzUqkdaz2Tnyytvklml/uw/+ILwB+IdT89qRawrUhM30duPXrkVHeLi2St8n7yRuSnG8y4cuY2xa3vSQZTZvdB+4FGfsgN4z3ct+on0amcvcdgN1EYc+osLsAdg+FPafCTnPj+L3cQwC7oMLu4Obwu7iHAfZjFfYObh2/mbsPYD9SYZ/k/Pgq7l8A9lMV9vlNQA5p6Elyu2IvNu/lhtEozKvaQdVVddXtoEbmpVfuP1DlvvE25ZkbuS8iWZNX86EakG3ltls21iY1+VfBPZ+DcV+5fVwmS3RcjowrlWWuemA/N8zG/TaMe4s6Lsdy/1a4zTK5tsF9kYx7M4z7JsqnctNJJRi3BQG6HgA+rbxqqqNLSvAkP2QUnLhsaFw7PgRkHTxI5/8zmP8OTZ7m7iFmpDqrjKB/GkYFWdWUlYfvOLL3yDg8DDi9amKCPPt6bhy9BdbKgUZb0duougA4h6xk9G0/AKgocSeZpa4RZbmuEHrpLdpY2Cp4nF3e9gzpBjzObuhy4+nhHmuPrFwVPr6S+yq6VyMDFRLQgZCZ2Z4uaLdMRBe0NqJ/TOMkVQ/JtEM2weM2Gh3aTFsq1NDnvtpj64H86jA0HlemQua7lVtH13P30XUj3K+LjV26/te/5u7rR4jbfAM3h24HuaV5i1Yc0rFDPF3s1WoLYUSDsZcb+vsdHo/D7vW+Gz526HNzXjt0Gz5ImfOb6HZNWvEtajXp/2Kab9KGgFiL6Dwf4fzoMOgamccK83TRech8FirB5h3SL/FCIU4+Nq/XBh/OLycS+XwiIYf9bpff73L76+OfQ4c1p/5X48eKxZjV67WS8c9l47FcLhbPhjwu3uvlXR5FXj62mUQTsHpmOjrH5KUTddDRozJLaOijysbBWoRStTNW9BpivaJfUzQIAYfg6NbE23ibzW5oMSZcOTpmB+r8344J9ov9dg7EXhGE/n/1rmepAAEAAAABDMx3DJjcXw889QAJA+gAAAAAzZeApQAAAADNl+MW/0D+vQSIA7gAAAAJAAIAAAAAAAB4AWNgZGBgvvHvPQMDy6r/Dv/FWDqAIqjgIwClcgdKAHgBbdMDjF9BGATw+fZs27Zt27UR1Ahq23ZY241RhLXtBkVQm6+TzfkuyS/zPe7OH2o1+gMALhGpQJioWgxQOUhVI5kOGCDvmKN5vIqzBb1AgCphvuL5+7SUtvF6YXOWMd0RrTzhqxajrzIBTJzhoQQO8sQwlA/iFJCtYtBDLiCemSAfEC+h8JcfvBaIHviGEnw3Hsh7zgZ6mOShhwohf/B+ZjifGcjnd8NXesCTz9TIbdiqa/CUM7AXznIQAbpPN0zSYak7sl977AdqFAvjN/1p7dZVskkDk/3a0/3mIEhV8Dl27I6JE7KFvdm5A3a2pWx8M37RF86tnbtjUsEs5HV2b093JxWENP1Z5cNf9aDhCJSPiKUGCiVvciIvSmy+FkWR+IkiSiCdEoTe2kGEUyLPNTIbmVlUwDmLCqBnckAcpctR5lFmBqK0F4gyCUQU+F5JRKR8ga0koAeF0xAKoGCdQ2CNjxjEtNe5C9nyEXEqEVHqDZJkPzz4HaTwXIhMRF+KpBCKpq7n+iNMxsO/NWt0V29KpTTyaJaHPkZ/kEkmQkzsUGgSSb4I4bXgZsk6u8qlUApW4fq720HXaA0tomW0iU7ROcBYT7PpLF2luTSNJtNEWkzTm+1udli7BID/k/8nHMTsAAAAeAEtwQPAtDAAANCz1artWLvOa9dv27Zt27Zt27Zt27ZtG997KpWqbsrVaqhuqJ6rvq2Bmvaa1ZrTWo22qLa19rwurOus261Pqx+of2/AhoaGuUa3sb1xrfGxqabpqpmY51p0lrJWjbW0tb/1uI2zFbQdtPvtre3PHXUdS53QWdu50fnW1di1l3EzJZmqTEOmLdOTGcpMZL4CAwBAAARkBHlBSVAVNARtQU8wFEwEc8FKNis7mp3OLmbXs7vZ41yYa8y153pzw7nJ3HxuNd+Qb8v35IfyE/m5/Ep+K3+QP8u/5X9CC4RQggrMCheinGg7OozOo9voOfrs1rkHup976nseeqd67/r8vsa+q37iH+x/GQgHVgtQyCzUFVoKXYXFwnlREtuKg8Xx4lbxtvhc/Iw12IXjODcujivj5rg9Hoyn44X4OL4d1ATDwdzB6sH1kksiUnapotRamhwyhbKGNoYzh49GCkfKRqpHGkZaR5ZGTkYd0abR5dH10e0xV6xkbGjsZbxm/Hgia2I74UhqkpkUJdVJS9KbzCebyWUZypKcX24oz5R3y1+pixamZWl12pC2pl1pfzqSTqZr6X56k75OCsnsyb7JrcnjikHJqjRVpir7lZd/fvwNsPeMZgAAAAEAAADxAFoABwBxAAUAAQAAAAAACgAAAgABcwADAAF4AY2Og07GYRjFf9ndQW7I9pBt1/TZxh/zuuymzp5l79V5jvYCTdxTR019CzV0wzOupVOTYd2HjD7j+neeBh7wnnEjwzw+4052awa5IE+VEmGi9GoKkqMsdEpJyrnYuPSM+BJTjDPJnM4Ka2yyzpbQ+4bn/HN67GP+OXViuWspJTmT8uXo/dB9qp7eHxxaciTEVQib4vCs6SyYmiVImqh5YoyTUUeIaaE5O0vW8v9/J6hoFVhmQsu1NS4lIm+IqHBYKCutgi9fVChjDUG945bPknkC7V1CCAB4AWzBRQHCAAAAwNtWBXeG2w93d2ISjARQgDsh8H2L/fMhCINIJCEpJS0jKyevoKikrKIqVlPX0NTS1tHV0zcwNDI2MTUzt7C0sraxtbN3cHRydnF1c/fw9PoRBA/dQihgAADne+/atm/9jLzMWmbb9sl2q2zbtnky19m1bsY+t9ywX0+9LNPbHX3cdNsDd91zX1+PPfTIAf0sd9AzTzzV3ycffDHQAIMMMdhQGw0zwnAjjTLGaGON89x4E00wyRSTbXLIfNNMNd0Mn5320ddIiESp8hUoVKRYiVJlylXIla1SlQRb1NksTaLjTjjqmKuuOe+CXXbLiSSHHXHFXM2ly9BYU111l6ebJt7LNMc8Cy1QLymSIyVSIy3SIyMyIyuyrbM+ciLXTOckm21N5JnlbORL8cdvf2X55ruTalSrtVIzPWyNAi0s9sJLS7zy1rsojKIojpIojbIoj4qojKqojpqojbqojwbRMKFN51at/mvR4f++3f41NRZGDMMwACyTZgnjcXGPMrdhXD5gmRLb/ye8dzedZ/+Xrm/2i+B/U5MsTBZpERBZpc/4hrHlPchitPoWB+r1H/z6BpS+oRdsIIoNvc1OpG+GMyeHLA76laZpOLh6i7R0HSjVyymkMQMopZgJlFLMAkopZlP6QCnlNkApYqYHlDymb/TpS03jTwtov73SAAAAsAArALIBAQIrAbICAgIrAbcCRDYqIRQACCu3A0A2KiEUAAgrALcBUUM0JBcACCsAsgQIByuwACBFfWkYREuwYFJYsAEbsABZsAGOAA==) format('woff')}@font-face{font-family:'Source Sans Pro';font-weight:600;src:url(data:application/font-woff;base64,d09GRgABAAAAAEPkABEAAAAAh/gAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABgAAAADYAAABAA0QDckdQT1MAAAG4AAAHKwAAGWjKVuYJR1NVQgAACOQAAACJAAAA4PFn1ldPUy8yAAAJcAAAAFMAAABgW3yVKGNtYXAAAAnEAAABRQAAAebzMPm1Y3Z0IAAACwwAAAAoAAAAKA27ATZmcGdtAAALNAAAAQIAAAFzBlmcN2dhc3AAAAw4AAAACAAAAAj//wADZ2x5ZgAADEAAADAwAABc4JLzHaJoZWFkAAA8cAAAADYAAAA2/iCz02hoZWEAADyoAAAAHwAAACQHrwOeaG10eAAAPMgAAAHjAAADxLPOIg9sb2NhAAA+rAAAAeEAAAHkDsgmwm1heHAAAECQAAAAIAAAACADCwIzbmFtZQAAQLAAAADgAAAB3CQ5P0Rwb3N0AABBkAAAAgUAAAM5bFBIb3ByZXAAAEOYAAAASwAAAEte3My2eAENwbEBQDAQAMADgH10RmAMHSUyaqbJ36mwAmqj1qayx8rhVLli5faofLHyS2o5VgWBWga3AAB4AbSOM8BXcRiFn3s//G1jybZbclNbmLNda1rD1pptt2S3ZNu23dvJNYbv5/ec5z334gB+GtIRt2PnLt0ID+01djh5SgHM5PPzrekO6Td6OOGvJ0jRFuGCtt9pq6q6auR2ZiDjmc1adnOW27xVgN/JOpWdujjyfSTIUaAJzejBSKYwjZksYgk78VKwmzSxGzSzK/QjQn8qM9Mes8j2sooQZYiQ+1jqO6k7pWbwqueV1Ef0IUw/atKf1iLeiDgnoimlIu6JeCn1NMt0l5Kzt/Lvq3onpiplUp4rX4xUMVKzlIg5zpIvXQe/51Aq5Qqr5DjqOomrc6N694vPySloN1OKS3+CPCb6hTisbmlMsPdM0naTUymF5MV0XVpRH+yxrbe7OjdTpKMILxU47OR/zrtne4Ei/3nYVtC58TftpuaF/5A9345oTgea/tDW22Gd2+y4Xtf/wzdeEv3xPv39/7Vv22qidv+vMk/bs0/cmDWYKzcQx8PMzMxUp0pfpi9Spg1U6cLMTbgJMzMz02Omtc+89ls7+8nJPn/fL7uTsU5aK2e/C2dM0mo085f0nxndFXyRdiaMEWEbRjyKMOJJ1lDnaTsvIeIT3iEhCVtWbJnaNiR2ZmY1evJdyT8WvXrVOWSYoO2+jKXEnu06hs3atvjz6N7TzsssCsFBNImfeF7Pn+cLjVI/mYnbCYaOoO5QY5O0emNrpL99U8u/v2fgrDcttIlkpMNqTNknN7GZL3hR1rWB5TwyRkkkK/5Jd/UiXuIBHrDzXmEVt/MITZZL/0eqwT0Z+RzDuOfG0uJbZ+7C0OpYoe20U9+24HN1DQOxLx/6zonuGdpv4cpGKmopKdlOAyfltekXnJ/h/H6ZeCIxQ7WIcO5mDVv4kTrv0edtrsufrdDccwivTHB4jpgtvMR3rOBz3mHNhG3jMKePmeci33mazcB+G1Kfvb7ws21tKE5XXsI4vnLUTpAoDYjGpjC5jJjPJBqT/PsBEno0SXgr7y3hIfGXC9/vcAxfeozQs2MDb/MRa1jCd3xBxVnRiKTQZKBIR1QwxAyI2Spn0FP/h9hZI4tN5lClGs5V1G2/E14zP07Ec3sqZ+JxbpL70C4TMdT02LhvgKUyh9rveohEZwFh3SJqw9002SB54niqxFzrDdt44eLS8xfyz13cS4MlC9YGQ4TN9W60MrLtKjUyh6kdlCfKmV1kzJdz5HMOqVo/1mGZ1Avr8by8fwFDzSnn73BEfuc7Nm/tkTMzw1ZFGT1LepH1fhaZH1f0dETzINtyu6fZ0dWhWFR98a9RMcBg1NeeHk9ayuJhMUZazqwW74eKMsXqUFdWTxGNDD+HTBG+0GgW72T2VPUsfZ46mLoFboY6q60sH4UqPnO2dZ+9o6wjCd2rWMJVXn/KCjByQ96HlDfp6l7X5PlHqrIHg5wr3t4IT/bRyBZWYXx+e9WxN1tcYkqae/nj3jlbfw5P+pbfI0VwrDMa+VmDvmfBIqYy9cy7i7j71G2UKHI+IXP4nVFx8soRNr9kvh+f9X6dd2wv8e9zXm09hWHAduZj9WsrGSNfy18NRrEfEKqUGn1/v2hVo1N45wXW0aSCkYqfcEWYmeW7BDExTclg/zuhXuqvXbSlrkZwyz65gR8ki0ic0+bSmewkmk/iv20PzJifDCUfv0PCgC5G6thW3pstJ9Ahpk/L2YHYYVHiVm/amKKvt/owc1M62t5CojZC2fzCMKLF7yCJ+H97ey0SOegjkul+6PDWZK4IS6HLZ/QW0FgjO/vFrOsM/20WXiXfezVRhQr9+XYYtbfe+kx45v5Svl/HR9ubRQMMCOLlct7R1gYGs6yDNPSftHBNk17m17RZ7VMv46XpV1Wy8T3m18LrWU8OKACj+Bnbtj2zRpw2aWLbtlWuJrZt59XyCPnKeO//d9rL6k7/q/npPX++FyMezOSBIhUy1KVAk20U2c0p1nOGc+xnglsc4i5PucALGeeVTPBGJnknU3yQLgbMvMaKDTtuqtRwUKdBEycu6uKmpTx48eGnQ4AegoQIkyECxEmQJA1ACkhLRorkyFOgyCglZlCmgvUnVexKpCGCVvtJFb8IQdy0Cf8kTkSJAAAU0co/idNPEZGyQIYEBf4+YhIVA0aGMDMsWUZQAAwwiKIPERMAvT9lwcgCzOLGKj6c4sMjIYLiIyw+ouIhJkbiYiIhWZJiJiVl0mIhI16y4idPkSAlCVAW3aGEqYmNJi3qtCVIR+r0SJBeqdMnDfqlyIBEGJQSQ2JnWCqMiINRaTKDubiZLz4WshgfS1iOiRViYaUYWM1G/GxiG0G2s4swu9lDnb0cJshRThDhJKdwcJZzBJhgEhdTdGlzUVJckjSXJccVyXCVa+S5LgVuiJObUuCWOLkrNe5Ji/tS44G0eMgTqjyVOC+lw2vp8FY6vJMkHyTGJ4nyTRJ8l/APlZVMdQB4AWNgZGBg4GJIYJjHwOTi5hPCwJeTWJLHIMXAwgAE//+D5cUYmB2jXBUY5JyDQhQYNEKCvBUYjKCyjGCaCQhZoSxmBjYoi4WBnYEjJzM9kcEIJ5mfnMNggUwyMEBMBWMmoFlSUB4HELMxnALLqjGwwMWkQDRYJQ9DJZBWZqgCkoZgdcoAVmYjbAAAAHgBJcWxYQFQEADQd3dJApUN/gwmACoLqAykZTOgNIENAF7zVOxz6Zdcv2bxPVaG0cj+X9VPZf40/hEA49lkDvdHXh9/1DkGTWwgLrlFU08ZkAyKAHgBXcsDjB0LAADAeYg/amNzsWrbtm3btm3bdhvVtn3e2HXjXu2JMYgihmxiIsgmHnmICgJx+fCvwsqqop0uxpnknHvSvfQxUjTSPHo1mhRNCbIFeYICQUJQLqgRjAgOJGRJyP7xI/hXoIhyqmqvq/G2OO++0KvPu9n3nTXIFeT7vMv+sSMf36MBHy/x8Qx8HMPH0XwsRUa2jPiHj+FcCCeGHcNpYdUwV9qptGMioAxaoBN6GYA9uIwQQJI0f4uZZ7GjLrhuvEW22ma/22ZbZ5bd5pjrgfseWuCI7fba4bEtnttpj12eeGm+Ky4675J9DkhxyClrPLXeYanOeeS4ExI9c9Am6Y45ba0bbrploZbqqKue+hpo6K5GmmiqmeZaaGWI0froq5/+BhjonkGGGma4EUYaY5TGBltiqRWWWW7lJ/Pae2MAAAAAFABeAHQAeAAAAAz/MwAMAeYADAIGAAwCPgAMAn4ADAKQAAwCyAAMeAFdUEVaAzEUTr09AVLL+3Co7NlgmcHd8nC3M+BO7SyPXdjNwbBXl5H8+sWMEIlZI0JL+tPlKqBx/b0YoaJfIiQ8+3tJI1wJKa1LRa4DFu4EG4PAzJOQNnl67BXdhTIjM9MnGWnLi8MT8vaUkIPTDKYliVV9yeOaBhrFcI2eIg7zPN7iPN7SPBnkGa4qMzCyxRP8cMmXmJXk6V3Sy5ruVZhGFYYBpEXOkiZHhQGRW/7aThlvLtsrew4kOBpkEizPsqppNEwCM5my6gK6z2TCmS6saiOcFsMlWo3RimFE0eGbsIzrfokjhi4IF40u6AIMAypeO5SYXdWWCgNg8h8k6mvkAAAAAAAB//8AAngBpXwHeBvH0ejuAQTABhBEOfR2AA69HQEQbGBvIiVKoihRxZYpOZLjKom2XCT3luZUx46S+OVP783/79hMeXHs9Kb0/C9Kc9yV3osFvtm9xRGgwFR9OtxicLc7MzttZ3aJmtA8QtwM9zqkQhrUjAzIgg6hhxEa2wUfaecjCKGOgUUG4AiAQ+qBxRXUBA1k7CytII3S0iotHWs9gloQ/v0KamffoaOOR1AnwDJZk9/oN4lGv1FrSmPBgE1lLHjxPP5zRffolXj7FcnulDpduiyVeTSb4l537jI8/43KfvyWb1T+dPjgwcN4/ILFxQsQAvRXK/CR4e4B/I1AD0NWT5DV12DfRgBtSK0AjARgZOTAL3qKeqvcAlRb04+gZooqFlNYFPTYgHlBK5gkk4Av4TvcerXe1cE/eka65jt3f+d63DbXdUWheHnX3J6PcPecO0pwU6HtDDcTcqEQ2rMhb00EYJKREeAXE0VBSK+gFvhuVnjnBIQADi0XbZmgFZL5KeU8nMWs0Vo82GLWc3FsyhXyXeE4tiiN7d+QdpSFYmliU3hkb+GJz4/1Dw7uWJrcPDO9xN1jDA0k03MGddvscHYyw+Obcqlk5H8qPyh0S9lvIoRRdvX3XJg7hbyoVKWjiaDdVEOHkwCcMpebALc2iqUTWmYZy64BTspZeW0KCwE94GuVcoUir4dvKS6b3nZFue9C75C9FI6OZJ0XbU8MuYYiLxooXzaXwmNzN+1K5fyTdre/f1fxqpsC3olUNrFwI5EByudB4HMLMDG/IZdbCKCFYPcIagV8ZJ4aq/yzynwrMnYZt3/o9e+Obz029T/vOnjpZQck7p63vnXrie2xC45dd+0VlRT0jMm4+BkYtxXNbjhqKwG0klHl8VRUJ7TwTCudV20aeEixaYVfdUw7LH6LZBEscN+OX1p5/uxZ7Ktcxt2z9K6lR5eQMvaPqNyPbDh2MwE0Nxq7uW7stRGNwnb8sspzv/zlMgz2yFLlV6g69wWY+xCa3XDu/QTgX5t7PaXJDy0rHdUNT/vpqG4Y1VkjDx5unUDkhbxkZEKRmT86HOjJRAyD9t5UYkJyXXzAles48WTgitGhq7YxudDzno5Jm0eWDIP+FRdWfuJPgnAwPnGjdI6m/ok5WkHNysw0A54qJicPI/VpeA7u2tNEWsBySUYBPoXty7h5ebnyJ2BY5S9Ye+4oDlR+JI+LPg/jqlBow3FVBKAi49I+oa9lZjwwsq3+Hj8A7xtRvPq+ljyuVd5XzByVaBXlaNGDgYN5QA6EWQiExcPmLk8qtWzNTEtjF3ofCPq7x/HHKuXklt5Ad7HKnwUYp+3v2Cc1Aaj/njnVEYBO5iAH77ZRcwqih3TQotRJJklrElRaC5Cpeu6WL569+SMXEKZ9D0dXK0dw19X/V5Hr7wE+TSi+IT5M+CjdHNNfI9GVZdxROQEs/PYSqtL2eujLja7csC8NAWhqAG4CcCN1Ay1akT0lpa1DaZngaQ1tWaFnN6PXxEtFMqdGkPGioFUJKlHwqCwwyTdfE1fHlm++XKXWcOr8kcKRLjXX1KS6FHjxqRMn8DCIUJwHLeDvJby5l88XC3zlO1V6+ok/2Xiu2Ew0dC4NpF0NDR3FXc/8DsM9hYkQUaQxQfr4V0bU6tYr2Q1QfdXLpKuL2AXIXvWyrqPFyhOIo3biKNgJExJQHtWZXQUhBWUYX/YL8AjzaJlsE+h8nlkFYo/F3AB8DwsBjcXs4fDWwRfPxGIzLx4cfPFsLDb74kFxIGHbs3nzHltioC29eOPclpMLqdTCyS1zNy6mLw6UF4tXnDx5RXGxHFD8RAT4p0fWDaxmvVatgIwjFhC0gTUwKF7DUut1iduwCKYcc7NnCnuGw+GRPfnDy9961eTkhdw95sxsqTCXt/8azx4fSD+JEOXVDu6lwCsDyqBNVVx8ZGhfDS5WArASXD6OkgiTeYwDtKPqM3zQaqctK9imGLFNwCdJ4RrWY2IFUqpapmqpkSi+G5gmVhmInzAWAku3+/ILA/4qg709W3P81owVgp/p0t7BQGBwbwn4ee5Dupa9k83FheWJKrtHjyx0t4oBTRNS+NwDfNYh2z8fkymKRonFCKMWSqyeucUVxBMYE1HmpU0qAVSsGvyYPr9vZGJiZN9vXnFw+YfSrsHQRO+b8N7u7lLi3NcxWkUwE8np4qa9ZubXwmBjz8EcSIpvYAyvwTJDAJk1G0uCRCvKABryPLgpLASPWAGmWOCuFJfGMttZxMNk2IvBCjDj/GJhd7K7x+AMWXfb0mHe37M1U3pxuNc3Gkt3G91h86I7I5h9/TuLwzek3+aPpsJml6WzWWzxxLqD4kjOlQxPOzzRoMlpMepCbb5YMRgZz3tzKUKbAyFOB3OgBerqQ3hGm+JQ5BnggBZEJamJSJGQ95No4HufxL/85DKXWFo692057pgDHS9Dv3bUt6EGtRNAew3ATABmmYnaqsFmgqo4K9D8uWWTkPF4ckHzlYGhPaXSniEBTM3d+amEyZSYyuNrK6/eemTS7588shVwWX0BaHQBLu1o8B9EfecjxyJ/hNrptKkkkwfz0gAGo60S1Dfs7bAb1Ua7Yc/xL78EUPhY98FS6WA3ngLvjFEnQqp2GNePrvnnx2U8aBwxUQBPAPwaZn5qmA3wEk9bLuRX5F/Btg5vlbxUAW/Tebyk69CpWy2tcxfPtVpb1TqDrnTlyVdt1Rk0aq1Bt+UVQNVD+cNdXZd04enKQ9IlefiCp8Ge7wpvCsP/yrvpfLcjEA0634rPwQRPfH78tAawE4C9hjQDARhk0jTwrl32nABqlQlSSbxMRxlLKpOgrLmEz3/wlsk2HhYmltaxGz64cstcu92gbufbNt+At+L+V/MZtzvDv7ryqcp/n3JIHo/kOIUQkw38F8DbgAb/ebzXo2mQVboOOYvw6cdPTMDiT93h0U9c99nl3+E3vis0Fg6Phd5Vufh3hGcR+Pgjjc37UX2ot95Pyyqhkf3KaXgMtBArWljEflBDDJFNBPdVfoo/VHkMlyu5cfz00njFsURtbRfYsC/gPyEnEkEa6+nQbTwhTBpZUGOgE0KwhHcVe2uBBwwUbgeYl6JlgVYAWjKqghx0dhWKYO/AwGnFAUxWUkZBFKjnKbKIXq/qeul9LRa/PTTmi0Uv7t23+4JmlX9TYmjrVaLXk3PmRqOdt29ZMJoEp8FlnnWELz9U+WHBFd4y7fe4HFdozHYf0DoFtqeZ+wyyAgq3rdO8zvUR3RrxHQTQgdTrvaoS0XUwaeygcZy8ruYpkR3QctCWBlo+hfAAM2IWLSHYw1UtOmQImAMg7lc7dUzlnQqnJnMOoWcmGhjzqlr2HHNI1uxwrrurm/vMQ9ucAWFob6l8YDTo8+RO8J075qe30bg8CvP6MZhX9/r1doNZrV9v25UsS7HO+9Qvt6OR0cVsZsJeNEf4vumpHo9kzgjzsdzOweAf+vcPC37blNG0uGl60W6c8voCw/uB/3HA6Szw3wycuG1DrJjlW5uQTgLoROqGJlgHvzSxCLQJdVL+dwIFNkqLXuG/juU/ZEo9dfSJeale4op5SnFc6NkUifW41ccu1Kmdk5F9lx8vFYoZR9aeGf5D3/7hoNu+5aPnBvLO8F03zG0vPsV3zlOdIrz/LtWpANq6EZ115roNfuFZnAi/AIYdMq6K4rQpUlNUQjGt34NrpiREQzZtNFzeFs/NSI6myhd0s32+olMUFtLSruEQnspKiWEdRT0wsCAFzCOTJuMk7xGGL7ykd7g8gDDgjPBu7muA2MKGmDPl72yYKDADopyCvI62Wk+TbzAdYJQwIOnGNFdgFPLFaoSDR53Ogf37jw0Pd0b6E460kTvUtj2/p3IVftVCkxQeSjtatCQnFlkV8V+Bt+1oDHh7Oap3hms4DhHAENKsxV3d8PsQ5XE3S87Ba4BU6DRcAAlTSDdA0gBJA2RcyUHMUM7zIBsscAyIGm1XgchJnt0YHUCYB1fVWQ8tay8Gi8YiFDmZJivT3SO9fVPJ2EjEGd4fDJjsfotd9OXHRXPI8OHdbR3u3KjYETC2tgvJnYt7QqMX9SU2FbzfFHLZYDCXm/SH0j63WevOe/MWdVvE404b1KaRRGw4yWubdujjjuhAzKrRtpiNFltpMDiYcnb4MniflElJuXRGor55Cj5+TvVxF9rI+CnmsbHl17D0IswwNXEGxcR11pk4YtONsnwahaljau8macvksWDUl/UeAzt2oTt58QWVr+BQf87rqbwHcagPBv0S9zhqhS6TqD7k1zVca2lIzjQOQwbCMCcwJAnpQbFv80YiXrgWjx3jhoJudyjkdgcXKhY5//vN1Rz6CoxjRG5UXOcTdOsjXQ0Zx4Y6YRyAsSwVjKikS5jxrkVherkj1Bd3inpri6XVz/d4olGPNxrFZ8+9IA5nnbqmBVVTMs7trWImzwvHwbwoORUlhbLxvPzzORVUk1MpkpyKCDmVqSOqh2589yMnb5w9wn2mMvHEFyo//M6ukwgTn4n+CLjoz8undDbICICdUDhilXpwkdcIwIupYx3Naq3WwIcHctxnzj1kMXILLdODTAZVOtBnP7r/H8mgbuPQl3myxvl5TomFm1mLyCpw7LS8EAf7pThsDVuSyQsaLyOGl6gM83WSrF27T4F/8E6lCkNGYSa9eXo5FE6VjoVE+MBnR/ypVCycq4p3f+U97IbWdBDoN6OL/3n6FR2snwGFXKaUTCEZMZ0KMfUKyfD3zyoaCVgP+1J1Cqn4tRcAVxMwJt/QO9QHwcBUxRG4630Xs4fU2zKTGA0PzadSOwZFcXBHKjU/FL6mv1AYGCgU+mlAIQzv74d7IAD36cXF6U2Li4jGdANcM+DEYro6f6RbnynVrS9HaRrWUJrgXRPLyJlYrtuqiIhJiSmalZiuSfHOPDg2YHF9WFcb033iGBGWmWhqQo7qfIMelTPHQrrIKe4DXY6wMExiupGgy7FNCelo/DSAz1JZ8aGbNuQ/kwXd+nBJ0yCggviJRU2wUGSRVDV2MvxTsROYEOLGN4idyjRiVTvHw+fHTp70ue3/tdURUGInRR+OA41GtO2f14f12m5c0/EzsvAb6oSfRbRM8h3TomAxt1s73IM8iP7OjNRyUK1Odle+zfLorwZ80oBZvaisDZ8ggER9jidBJUiRFoAF4J3EuhwPzawV6jM8/FqCZ7N/KBwJd3rCJltOtO/fLm3z5B15d0gwukJmey5qt2cmU9KB4K0Ot99h4DsNumZbIOkd2hbkx028lzdYOvTaZj6QDfiKUZvoBnrMQM9u7gjM8Vhj/tbpgmxDmtn6UmapCu4mtlqgmZ4ylmi1yVwTi+Z3j8/o9588OdbubDWbJXe8N2Qkkdydd+55fkHTtEPb6kgPRQCfUcDnZ/isYv/qp7exOivz3dD5GVl2k1SOm6nXAxVgoSjzTARR41rMnx89Foz4sq5jsKj0kcAEn638oD/nC+K5imUyDJZQttVtCBFcG+SL1q9eGGDjfJGpNiUgtF29ucXSom41t26+8l134LOrwelweDq4WrHItRyEuG/DuJAv+ufHZT5C98/mi/ys/G4AkJnlixDyrce2Du+1NIttuVfXqoXMUPPgJUPNkDnStml7Lzv5kpJOD229rvt2oEoYDwbHhVV2B+qeEsZEcVx4EmFSP8Dfwmcb5IkaW7Z/Jk8EsSMlxMwyRkCIWIO/lq8h4L2vOVFs5VvVzebm9PFXv+lEX5u9naS/uq7F6Ow+c9wM//f9+XcXWRMWS5y/SJaH1QKVBzcaR/VBl269rV2Pc+3CpB3pGHo8NQxlXJcw0quoCylrTc1iuq350at3tFoAT1Pz9JUfNQTT/RGzR295EdeUDnnxU7/xTQiBCf9vzq3GhtP2pqbdBM8AfLwVn22QT9L9a/kkmtZVSRZ/4Nmv4Que+VXl5gG8bU9/5YN7EVpdlfdSqLq4MFIjhLTop9jSEP6zDeBP1cH7FfgzdfBeBf58HXxZgf8c4ABaPYeQysDdA/RY0YkN6whMQRsX7Dh4tJ0KUQtrEVUDFikBgFZJNHSwrJeehQwWeM4k7wExUiFrsvjzkOHB1fI8fqnaluIlzl654dPbr8cT+NHKX374Q9yEM9LBzCVvW3rve+WSPdAI9R6R7J8AGjWUxl/i3Qjg8h4CykMd47m3IfxnG8CfqoP3KvDnZfjqX+HbV+jzbXL/6BX0+c3wvK8G/jMZvvoHgIdq4E/VwXsV+PMyXK5J0rk2srmepnBaC6T9dDK6tjaE/2wD+FN18H4F/kwdvFeBP0/goBHJ1TPczdxHUTPAE+iOqtREiExEGu5VWUEIGhEaOXFUBJrhSacsNKwFHTCPuYI62dOQMVSecyswn9KKsV9B76xyRS6NWeqAx9Ucg6gVi9XkA1/kfxHtJxU5Uzw+uHlPayu+1u0h1bgDkwP9Y14/qXgemhwcuDVKSpujXZPxk1eEh534wtgVo6TiWdyS2ZU4RqtyPdsySK4zAlNIzdMMOO7YYPdGfYq0SYkbtaflzFaHUgFtUbKidlltTrNdHDU7VuRlQS82Aq3Z9MJ10/gdlb+eOQOl+U2Tx+fTx4++KBi8+Bie3nrd1ijd0BLbfmLr9UfxNy+HeZXrS1QObExuJlAj+M82gD9VB+9V4M/LcLkuQJ93MbknGsIhD3zsovsgnGgR1QtJo3pstVrm7JCNCZEVJ7IqBqaZVbYfRjaww81wtwOvQiz9RHllqu45IgG4EXceuOzyJeklL+m5D/YdLU/99+ws3Wv0qd7/5TLy3iNcLKqxC9XVsjuAvlJ9VNFwXj+ObNRLq+MfRxy0eGjR9E5tDbtYrWEvHz6viH3Ykp4tFee67Hc/dbw/8yTwUq49cF8HXnooL3+CTagR/IkN4E/WwVcU+NMbwJ+rgz+pwM8SOFIROEfW4m2oD42gN6J6JujWz+IaIEUAKcKmenNBASll51QXSjFd6GK6QDTfiiJsX5NsJfyKjpRoywqtQdpyQ2uEOhRWF0nhNNZjlZKcOD8rHBbpJ0lSplRkydMkr/nfLnrdg06dFnMXbN4bXSpIYybPpkx6ImNvqnxbs6ng63JGhR3pnnm3dGFmaOtL7sXNZr9dGPNxfCYT69e+wu/xOo0JK9+ew2pHOB2LpYWBnZBRHhy3GGd5T1ZIRLYMbdnZIVdkNvWNDPdT3kdXRVKXAN77KO9/ga0MDjlzOrcBJgutDeFPbAB/sg6+osCfY3AM375Knz8k949uo/ASQripBv4EgTN/ZayBP8ngZwHu5FYU+HMAZ+tW9Al8tsF+rfOSDsp+raljsNys0BiG5WJJv3EmwwmFnhcoHgnGl86G8Cc2gD9ZB19R4E9vAH+OwJEK4M8CXA0QA4qtrX3DhIRwDU02ArAxPwiNcI0f1MGTNirb4Q5Z3mVfVlhzZWueLLzOlX3ekyyM6v3eRGFEjz0ON8kKhcbdqVTO4SaZotnedOoyz+JkXsy7d092iZjz7SyQBJHV2h3s9y2QtpAoBSktv1c5QK9bQQN70RFUv4rUbbhXQ15LNrMWLJhOwzN0n8wK8iIexSht3jRAWDFTrt1005YXWr012pqXNQ/oq9VWTQN1VUo5D/TlpL4+Kdd3dWhgW6qw0+leyKansnYVV/mGZqbgy4OSLmSkXW7PjlJqW3+Qe0U2E+/X/oHlxzYRFhRy2SJRTaOrmelmMSvlIYt2SNZJOcfLDXFfAbwP1nv4mhxvNwDgAXXDAMgADTl15KcTH2TfV1CGteSwQKrbtgnzDpdHXZ1/Jhopdb5L2bUFSuIqbC0kx9K8w+dwRbNRF58YisbGAzEp5w+6opmoy+62h4K5/FH8h8yWktcR7/aGpGjYanbZ7KGC6C/FbHbrFjHssPkcNn8y4E1G455AYnKs4gL5p/kHqi85pl88Qg3gT2wAf7IOvqLAn2NwWCOy5wvMvqzWwFcU+HMEjjiUgX5+y/0YdaA8WkT1gtnZYOH7cSJ3ZLUSXwENjtLKhF3ZixMFaTWC1Cbgnjkj3yUljhCVnJOJaaGyfc7Kp7CskCR5CeqIO4t9GDtLnuKgyZ840jvj653vSs4NwI7QaQgneX0mcI/3ej7nkoLjvld6t+3Z3m5vC9imh/z5kJlr7nvnsb6LpyJki23XfK/X7pY2+yfzA0aPYcIUd0wVEMvlKvX5nRuuqJkVVZjB8gQN85u1GUyfnJ8VjOdV3fl8XX72GOceC6enJSdJzwr9bk5Jz35GdftDszZfNT3rsG1+u5KflWun3J1QO02iE2ij5X+MAGKos0EJC35UaqedMFFBknCDqxOgBqWOqoVLD7+2kF/Y9MZomMjqFfUV1roqK8sywo9VAHeV01GGsusjUHg1RweTrlynVe8ye4OPkDKsM1VThpVLsfHwYMrRolvQaqKxuFKXBdrBhhwD2hPoJf8obtqoZlW3oxleV/YmNp8m3yC7A3dDGvph+z1lqA+gQRYjh1jOIkoZEZKNaT9ezwWZQcWatCV3zCx5YpCkfNtaBdoKFen36C+88cZxSGJq5Izl83UF6QXgDctoEp/pAh5cwZ1CrYhHY+jl9dpbE1V7CMCD1A2daC9yUiX2KGGjBiC9rH7lQRm2uRxYcQb6ZptJV1AENbN0V7fSGmZv1gSM4GBMZuJ+2AZPCqnqveJ/rWyfJzxXlPSYtF46mHXm7YktPpcwpm+LdtvcXhHfGp6Likt92dmC2ywk7dBxoi88MmuLSE7PbMwZ07+4szvs63M8mOq2dIa2J/ksFvm0w1EIO4Kpc99tFhy+sDs7IgYKyZAxPBkI9CYcM2VfIRkxRWd9wZ3SwFHJ7CTnhSJgF7/P3Y1aae2pPnOtb7g3uE0p6FnZ3uC1oizdF0z3HFEXxGF0cHn5ILl40dXR4RJ5dm9731vf+o53/tdb37fb3rN3dGRPN8937xkZ3dtjRxhNI4S/yJ2AkUbW7bHXNxBxlkmjdWM1tKqqjjuIhGeytJBC+A0Gyj99z9WZMc/CycotF+C3TjbbO889cQG1MX74+DbwwYv61+3W0jeQr6qe2E/L+TyDkgcvDqhYQCKuGUOwi73Y4g8PX7opUpBC/Y7BxL7h1HSXO1SeT9t7+DdVto0MXrZ3Lij1xtyDacklTSZiOxcvyHFNS6xW8nXALY22r1sB6/9urcSJEsqhqfozPxBKAKR1XdVEyMP01WGvnF1iFn2LezrTNdQZnyyky6Ixlwj3u3uj+0qpmYKnt6u0Sejfmtp8OHC/Oyol7RGXYd7gDFsjSb9zIpiwxsuRwqRZbZgrJ0bT9q6wvF7uJPkxmOtmwGvnOuuub5BL/DjZG0BEMS6baTUoKyIzzcx3K7NYHadluIl4ZbJMkzOFggVSrXlSosdxFaduc+iX8YG5yyrPvPCSuXOvcs247BnHRxbf/vbNcLznFsR4/w3AL4Am6jOdNdh5CcCr8J4lK70sWQn4nFlLZCr8Zmw2+S3COrc5657MFMYG896Yu2jC879uN4v26FTBFxrYmpi9RHi9MzrYUxowmgTcc9GpllZbdpPUta3bnRcIPxOA7w+oPqfWtCdOEIwjfYOaxcfp9m0N5adfiW/iSmkvpKDqwSb5OJCCbn1gg42RYlfG2edsLn/YWHR78hE759okRMeyTl9hKpqeL4diY7uzPdvdMwO9pW5/ZMcM/qDFaSvuHhYitux0LjNb8IRmrpnbd/t8JOYBrBL0zMkJwERqbAfqLBP8yCpUst5LA6oi2aepV2mFd6ut0cFU5Vv4fd1jEaP66h2vuPm6wYvGx07e9rp9iK3LZDvYDlozokgiCytqxisTQLlez8ooxvSMZN7gJaZl8skk/xlZMrvOyO6CVSj1uEidBYsUyfpF2e/EuAuap6py+Z7RbKrXFu32ecQ3O5zZZChnFlLPmHxRPiS5pYlofCzrMNi8BmvE0+ntnsukt5T8H8qUcklHyG1rt/a91lHwxIrJmC3o5L/myUZ8bWrHpESSmvZkWbTFQ95WdZs3kvOFyuBHUoNElkLAk28pslQvOgpHmHDJskQ0hldkSbNOlkzUU8qpVnrYAUhec5K1KvB7Z7+jeSgldud3dlZFaUaIjsqilAJRio7tyZbm3Xg8EJmfGegplR5noiQyUXKHZq4GUdoBokSwKsN+m2+jVwEtNtRgd52KTE7ViwEG5bjPF4v5fPHhQCIRgIv0kV6dRk9CH5Dbq98/JPfRAn2QAECo3SiVzuSb3cmkL5rcuum4aTAcCvjEge7pyxyrqzJO+GlORHdC9wtIQ+4Ap+PUwe+Sc4vw/LPcAyDjRuRADbaFNMtsFiWtStDWUHNItazumA1WaeISR49W3nevQpnc9yx6Fv+J9t2OPo4MbPWTyXqxUDRJRdIRo+vb6mUVdEe6Il0OHT2Kd7xeSCQEP/TGYopXce9FbhRFkSqeIsFTlMXEqJwiEVliDrgfHlCf5znVfC0Zb24yuEN+azzszVkHfbuKkdGckw+IXpP6RBWX7wX7cvFohA8mXLbecNYSG4j6pFQs1eVTyOWgFjGE7wVaCX480NoC6BhpBlaElp9Sja16df3OXJBRtVjLBuxr6qD4iJ6cdci/qxAdIfiEvZ0qgk+MXNgc6gWEolYh6ebrEVI4RrVmM74bPc59GPjfikh5TA3BKp1KAxa2qa9R8cNc+ciRj1H5gGfxa+HZMLqHrHHh8+u0j258IzrNfQR10BlsAkirnE3WY4vCRUC+eDoWVWmcYsgdjpzgbrSVktjv9oV68rcN0X5G8F3oDLfC+mkFSJPcDwi3pka6tSOkH4cYdIejJ/FKfT+AJ+CD/w/gE0avInjC59cpHPrH7+NWAP5qAodPBl89iM6g+zgCQed+WIXh96H7kMhg9Gm4U31E/aAnH0caxjMN5VkxX+S1vDZ9V+rYsdQ35+8NnXpjUOZxAZ4/W/N8k/y8WOQtWvHdR5N335184VTo3vl7g4iODf3j18DzYfR6gid8fpDCCwB/A4XfR+H3ETj0H0PfxDYcRirKN4ww4qrZfCH2y1/i8LzcL3uO06BT7L3v4odw4Pz3oGoTwy+vLOPADHuO++NGz3FfOddFniN2Bf0O7MorgY93kgHpGBF0B3bjlyATpb0FIZn2dbb3rfak32TyJ+32ZMBkCiTvMAlptysTMJsDGZc7LZhYff4G/OzqpxkeBIIpHqQOi5+tBAfHxih0dnUztnM/QnoyprKPlJf39vJ0Ey+v/UZ5bKw8WcrnS5MfPXTmjjvOHPJd9P0rr/z+RT7aR3x1M/qT0ocWtct9UNdJUijExopx+jrt6DXsZdYV7WP36mHs4R5HWqrtTSzeWT9DRrh2P3755Y9zj5fPBcr0HNph9LzyHpL1AN5TKxTzsLSQ4Hrt5Y8/vsj9qPzC28l7PjZeEcUV/pD3AnAXaKuTWZwVlII72SdGuFf1hDWxuAyAxYPsNZVwnHyRsKfgS3kMKpVKB0UEu9+sg6bBk/IV+KC9yScIviZ7cLHc7stHnMlY3KZ3WvTjeotDb4/Hks5I3tdeFpLxrZlCrlndLOUzW+NJAcl042Ad/gWKtQVaZtoKkaco/i1wD1H8eRopKtu25YJaviYhyUmWAG1QAPzWv4YlX/Cmz6ck7S0sUgzzEmCYK1AMFXrsegelx6m3KfQw+dyPR7jPgXzaGu2t5c6QCVcJf9j83Gbucy/IOri4qkHv5H6JVFTOMOJk+wuTuzg7y/3yBRvALNDv7D/q1yRYNj9f7Zer6VfLem5SegbpMULvpP/K4L5KHn9ZtlMZfDO+k/sCyLz5/LwxvEosVlHiqZkTtXt7XmS6vLXceoXpUI+0aQa/YynZZz9yxN6XXArt3CnvL30IYtsPIQMap5KcJvaAzuMIO7sqn08FrWKhKQvTaBKZRP7rN1ym1GSew9XQ/+XB4rCH5wXyaXuWL3oEb7Rvsi+aENv0ttRwtLDkG7V3C0Io0jfWF0mGPZZYfyS/w4MXXCGP09gWcLhEj8OoD6yYnS4r73V7E+VkaoDvEO0w19ZIuM/u9NlsPpcnPpCODlv9WU9X2BrmaazxYfwnep6/Y4MT/fWbU8CzA+dVgapiGfFty8t7C/vGRKiqdiXgkHoAB53d86X+7RJfGadrg/TqH7lP0PPdIhpDEnCwB2E0RDmYhFYXDWWSEOMLZ6orkCH4Zj+jRL7r9+cSbajJlfB/50RQOr94fHj4msV8fvGa4eHji/me2NRSd/fSZCw2Se5TsT3e/Ggo2WdLmoK81FPKuGNm0VWORcZybjw6e/1CMrlw/ezsDTvT6Z03zJYPT0ej04fL5UNTkcjUocHERNYZsOU6TaNdPaNWYzfvdUqk6IbUyIV78QPclyjdeTRAV0Vdach+p+EO9JnP1ObDPWeqtdcccCRKV0Q5yDwUKCwN9xzA5Dezp8k7j6A+mTvrvND6LJKKxdK8IAqkGM5q9T+EU7UeT348Ur17C6LVKha8vjy555Pb/WlTZyLvD+SmjGajyYZ73XBkNTGdd7vz0wk4uOqO89GSIPTEeD7WIwilKI85j2D06LYMDV3coW839lEdzq3uVfVwr0GjaBsqs9kfpLMfhpaBtiahZaUtD4nmoCV7mgy1kn6AeORsy7qVDtspSaE8OTegkdZRT0VBQ6tXVC7KOEx+yo1f88CuzTctZsSBmVC8T8vxw77ihLtrMtHl4Nqn4vZNfCpo9Y8cnhg+sb83NXdZf2yb37nrkuk7DvbioN/VGeYdVr8j6jZ85dC7jg/l9tw0M3LxaCAc9oRnx+I7RuPZ6LcuduYmkqWDmxKFAy9bmL5qU5jvHMyl9r1kXzRkc71W8KTk/dMC2b/JvRQ5kRddh87bksEADgJwyAsAFcvhfBxpoaWm+qNlyXdy18NdzbI4Krh3smyOmd1tDO5iz7lJng9cMb0kLb0sAr2EIlwm+LhqdkfH1j3WMcvV/KhlbrFjxwX8OH+1bfxMxzUdj83cMvNB+Ae3xx57DBtuuYXEUdvww2hVNcBpEcTd0CYwP34YT6imAfawAvPi9+NJCntEgYXhuf303RUKo3tZYF20gF9D89gRRNKsG6dXlYMFfF16de2Y08rYzp1j5PJFoz642i4/cODSSw8cuHxkZHp6aGh6eoT6km0I0MF3Ih0KKCPWbUuUx0KyIrJEqfDuhVCCH9uJwwWNyVi5nuVHJwB3G0ogko87b69r9cwHOi1vC9awdCiuWdSx/DfkYTPzFwhea9iYc0zGBjZBpOl4ReW5zORS1B9286lQdLjsyGRzzilKgxf4NgljR9EAIoJVXzNnfx5GR7P7YaXSY2VphxXkgd+alQ1flIN5uNWtNevd3I5g0SUWu+JZn9McMuTcZShbBC2wtknykbw7t9l2Wd7jcnsln4t32UwJX8DgjNj9SUNrOuRMeI2xoLymDgPy+4H3TagDzSDihRr8RQVU9xcVOOUP2zSdlrcpadkZvzYm+obTdIkjGDBPkplCnpRb/rdJnZSe7pl98P7DPaeKR7Ys9l54YeJvP6fyinCeW+A06JNULp3wfQf9/imKI/sd5tWPehERRVJgImskjNwUrQ6lVq9oINtnJv9lFbYDmG2GwQpT3ZjuAxfpOWNg7R2pwp4XcTYPn+U97gnx5IhObc5wRp+xnMWoGMuXmp0+U3ux03HPp4VO+6CxY1zr5yMER4YzPR/jRGQfF8Gz9q+UOBr7kRp1edjXnbDbE90+b4ncS14hFBLgwohPlCPRwQTPJwajkXKCHymlMqVSJlWi/IngN+MC9wEyNo2dzIrlaoGWA1p/f9wPubIhiyWUdVXvPlH0wYXfbA7m6J9UqN574wEhHhcCcTouj/bh57kCzEoSRdad02shchNiEyQfYY1TBpQxjbxJfFigmz2oTGtImCjvgtDWbpngB4ydncaB3tFMT09mrLdMvpV7x/iiy1Xgxyb7YvHe3nis7w1lvuQPlPhy37jXO94H3wJ++q3DYOgY/3lXudyVL5eJGCdXX4M/yz0PGO9bl/XRNzwDIa6V4gAksnMdxOrXnu9wI5Gd+RDZ31LBddEpKEEKr4tM85LFg3l8s6srGE1F+sf7I3Ob5lL7Al3CZDSbkQGjc5kxF/6DwxkVPaLfnxrJTV6kP3SwSfRtdgWSUX8i6E+W0+P72l50UBN3E/ok+DgD+dD42g4CCyEHsv0N9iKBUWWVFgviKP7BtOzhgsqultbTctbYovylnRCsoeXFLdFsmM71kSE5CSuvA/GZys5MD8cdHGnTtZOIprjFM2iKudNdLdpTL1E5nEtLE7adxVZT61iTQ7InJ3POWLBstRcle9o2bbEk+9xqgiOtwfwK8uxudDuq/2tL+vWbt/UNz6uqkZudbbCzFtlKsXbwHTzCabgzx20jJobcO6o+w0h/Vf78WM0mfVJIhosaOolcqpcM6h0Gtd6pH1x+4+0PPPDA8pvf/ObbTnEn7vOTM2z++ypf6+3tu//+vt5enCMxGfiOv8CcSehyVO+z9A2q4x8nioQSVK0SzPjG4a5jd+GMjLoAyAYU1Hn45lEO6GbWauR0EsNd+T5chMaAam0iSWiiMVtcmEynvFEJ/0XYWdx7m7lcwrhUbkkOz+eS29xD5qS7MN72/K8x/sKj+lHJl+O3JjPXH+smfzlEp9d2q3t2D/hhXnnX3GS/pk3bbNT1j8/6vIhD/as/oH+zSwO4FsF2eZmrkfeceiiVng7ZZXvhrmEup51ODkgl0TWapvHCjSoalUemXtqwaIRwAbencrnUl9wTweELjvqKM8nhWJkvCKIj0Yk/b5jp7p7Riw5v09KBthsv69oxIGovXuJ8/IRXXfkKZzMghFd3whxx3KkG62pqJjCsq7VYwu/BF49X3tTGvfPcbppHq74Hvuxz5DuJr1Rx6EdED0G3Orh/Dh+nGbqH8HcQknOS8IyWPvOx1afZM/vpMyv4XTXP6Nc9s5k+8zD+PELKWCb6zCOrq+yZk6yfB+Eunw85gH/MZVErcqO1mBekhPEaUcmiYZeUl2Q5/86DD+6E/9g3/8lPzn+S9YN9+Md4uWHeL4A/WRnBy/PkuSw6jk9zLhivFdHTljTTUAbjWM0VZF/3ut7X3dt7L/x/3YOvv4/oyX2v77vvPrkGFkDzgG+ZxqdOlFu3G6BtYFEuC1ioCNmZ+6uNH3FNrKqqSdyLgChGblF0w3UR/YTrFEX9E+RwNjmkXfl/8j30t3ngbwB9BGgme4m/iq5CGvh8NePp1YCjALRZgEZMwz6gFe5YjkNCFgO2eLElBJx59K67HoUhrvnARaMZdWb0og+gBn3Am6wPzE5w46Z8GefTON+Ef1wZIZ3gT+JAbScc9IHY3BpQNzpvz9HG0w1PsSML9ROvBR6xyX+QMia+/cEHt7+pb7uMcxHz+EvcA4CzY32Wo/Z4uEmgf7dVOL3/1EXFjDpdvIh7oHLdfXfddR/JS6I/45/gNwHVIsiRju07l6uTpjOyL5Y9lJ5pv0mOjqVcNRkhBLQ+34TgSUdClpiv3zdebeLerqTdY+9NS/SmyD+R2yZFbtVMbiFlSiR3Z+UUXq786pN0r2Q3fg6fRTwqo7pjaQ0Pn7OTiIC/kW2INLNW45OIIm+BXcGhiC/jJseE+XLYFtZrjtWfR4y0mXffTXT7VVw/3st9Xd4VzPT9CwB7HYU9ocBuB9hbKOxJBXYdN0RrG7ATWIFdz43g91LYcwxGxpiH/t4CsGcV2Ps4P76OexJgZxXYw6siPoBUbEc3lYXVG7l+NA3jKvZScWFtVXupknjh1qO7F7ivv0Z+53Lu0wjOYyj5TRXIvPzYa2+c36fqWoZnHoJ+b63vl8lYm5LfFIpS28Luo1w/6/dR6Pe1Sr8cy/GTxzz75m/kPk36hafRPZRPxfNtSgMCNB0APCMf69TAnMp4kj/ik9q7f65P3TcHZO3YQcc/DOPfD+OTHD3EkVSX5R60Z8iZWni3KL98//6r9u+GlwGn5cFB8u4hbgi9gSOrmKEG9NapNDV+JtJ73R+/k5W7lYyiaIqkaIrwBk02bPK7bG2Oln5omn1OO2lyQ6VBQ6chI3/KfFziHkMPqPqBCgGtAMTIbFIb3NdMRxtdgckrDhJFCNVRC2TYAbPfbW/Xm5v6NV3hmjb3mMFksNl7yHC8vTQky9A8zPVb6LwR7jOxASN0K1ZXXuDeMkPXsMDfEXQ/t8JyFg24pGEbcNrYMeZOwowaPyDVtPc5/H6HIxB4C1wOaHMjAYc9ELDDV3ZnYz4Oc9or+yGlcvQfDPM4fJWb7C7T9gbOjw5xT9JxTDBOGx2HjNdJpdi4QeolVSqlyGWy201wcf58IpnPJxP5lIvn3W6ed7H+gW9Qev+3+k/29CQ7HY5O0v9KLh6TpFg8l7LxVrvdyttkmXnrqohmYQaNtHeOyUwraqG9ixLLaGhFeUHhHuftAUeyx9cSzrh9qmyz32sL2CCT2GQxd1h1jfo0Io722YJa/90+wYaxvznDaUhG5P8DKD5xZwABAAAAAQzMTspD0l8PPPUACQPoAAAAAM2XgKcAAAAAzZfjGP8s/p8EqAPDAAAACQACAAAAAAAAeAFjYGRgYL7x7z0DA8ux/zr/vrOsAIqggo8AuyEIUgB4AW3TA4xeQRTF8TN3atu2bdu2bdt2o25UREVcRrWjmjGLqLb1+n+zVvLLucO8O5vPDmmY4r97gDWI/tlwDbVOamyLyTIa6v6Ryxjvoy6DdyoV73HfmH+JBBxhfVBSjiQrq4ZVURnbo0FWKvrjK6mIlVd+9zX6Y+VUy/KqmTVTb3dftcm67rdqu/Yqa7lYa8B8DrVzOaPHloe6oHr7vuptjVA93k825sxy1k6plJutYlZP3dx75bM3jJ+pQFy7i6oS+smC7xL9DT3ST1quTPSLHgeQX2MpvWVW348l6S+t0N9uVbBRnKPHrPjCahb3Tc/p5ZBDa5cj+orX1Ck9Z8V3IMexTu9phd5htdQovNUs3nULNqqcOdVCH1RGKRRJynpJaqO6K6qOaJiSbTU0uKRaqM/cAPccRdUKXalboWtijTJqiBbuJnmT7KI6wVPV8ZXI+N7u3PWFb+2m/qiLMaiBSiFHKZ8ijSaLhzyjRuH7RtLDa+69qHJ+rpowV9Ut0xBURkXUQIVMc5OoF6tsSo4Lb1ES9dEURVAC7TU2GhnzvVXBl1Ur3xY1VYG1Kkkah8ysJSqFukn43yXgJrZhNdYkzR3DOSnagPk4g+tYglmYiSlJ67OSHEhyJLgnxb+T/+MKsQQAeAEtwQMUIzkAANDaSmZSt0kxSnK3to2HtW3bts2HtW3btm3btv7X6XT1/12p9+sb6mfqrxtEQ3vDcsNxo8FY2tjWeNGkmrqatpu5ua/5uSVsqWuZaoXW1tal1tu2qrbj9qB9pP2xI7fjpvN/Z33nUudrV9I12/XRXdi90UM97T2XvUlvV+9an+zr6zsLbEAEBHCQExQFZcF6sBscB5fBffAafIcO6IdJ+D/MDYvD8rA2bA5fC6WFykJ9obXQXRgsXBeJyMWcYlGxrFhTbIrCSEaZUX5UGlVG9VFr1B0NRvPRarQdHUbn0W303F/RfzdQP9A60D0wODA+MD2wOKgGh4YsoaHh7OG+4c0RW6RzZH80Hh0Y3RwrGlscOxp7G9fFXfHc8brxlfHP2IejuCBuiNvinngwnoiX4r34JL6KX+KPxEUIoaQsaUgGk4VkL7mZKJiYmdicOJq4n9Ql1eTk5MFU0dTBdNX03vTx9MX07fRTSZZKS0Olu7Is/y9nl4fK+xWDUlM5quZU52tpras2UZuqLdd2aqe1u9RGozQv7U9H05X0NHMwzhqy7mwnO8zOsuvsIXvNvnIL93HK8/LKvCmfzNfyz/9l/q/4f33/W/vf3f/F//P/3/TXtj8B7WaMQAAAAAABAAAA8QBVAAcAaQAFAAEAAAAAAAoAAAIAAXMAAwABeAGVkLN2RGEUhb/YeYbbx7bZxE41tueyTZ2Hzl5nxZ7162wOgB7u6aCts482JuF5bmdYyGbdx6w/z53vPF088Pg8dzPaNvw8D3PYtswFVTwaJEnjCMWp0NR0Kq5qTJoyeRJCJVKcC2fxNMflmGKcSeZ0Vthil232NL3vfG587hv73PecO7HktbSGvHlpFZwP7adKOj84tOTIiXNJmuLzrOksmFomTpG0eTJiS/YdpjXN2Vmylla+eQ5Xq8YyE1qBrXHiUqrypDUnNZWluUTypTWV7N+L6x23fJnSE374RZx4AWzBRQHCAAAAwNtWBXeG2w93d2ISjARQgDsh8H2L/fMhCINIJCEpJS0jKyevoKikrKIqVlPX0NTS1tHV0zcwNDI2MTUzt7C0sraxtbN3cHRydnF1c/fw9PoRBA/dQihgAADne+/atm/9jLzMWmbb9sl2q2zbtnky19m1bsY+t9ywX0+9LNPbHX3cdNsDd91zX1+PPfTIAf0sd9AzTzzV3ycffDHQAIMMMdhQGw0zwnAjjTLGaGON89x4E00wyRSTbXLIfNNMNd0Mn5320ddIiESp8hUoVKRYiVJlylXIla1SlQRb1NksTaLjTjjqmKuuOe+CXXbLiSSHHXHFXM2ly9BYU111l6ebJt7LNMc8Cy1QLymSIyVSIy3SIyMyIyuyrbM+ciLXTOckm21N5JnlbORL8cdvf2X55ruTalSrtVIzPWyNAi0s9sJLS7zy1rsojKIojpIojbIoj4qojKqojpqojbqojwbRMKFN51at/mvR4f++3f41NRZGDMMwACyTZgnjcXGPMrdhXD5gmRLb/ye8dzedZ/+Xrm/2i+B/U5MsTBZpERBZpc/4hrHlPchitPoWB+r1H/z6BpS+oRdsIIoNvc1OpG+GMyeHLA76laZpOLh6i7R0HSjVyymkMQMopZgJlFLMAkopZlP6QCnlNkApYqYHlDymb/TpS03jTwtov73SAAAAsAArALIBAQIrAbICAgIrAbcCMCYeFw4ACCu3Ay4mHhcOAAgrALcBOzAmHBIACCsAsgQIByuwACBFfWkYREuwYFJYsAEbsABZsAGOAA==) format('woff')}</style><link type="text/css" href="optanon.css" rel="stylesheet"><style>#optanon ul#optanon-menu li { background-color: #F7FBFE !important }#optanon ul#optanon-menu li.menu-item-selected { background-color: #FFFFFF !important }#optanon #optanon-popup-wrapper .optanon-white-button-middle { background-color: #3365A4 !important }.optanon-alert-box-wrapper .optanon-alert-box-button-middle { background-color: #3365A4 !important; border-color: #3365A4 !important; }#optanon #optanon-popup-wrapper .optanon-white-button-middle a { color: #ffffff !important }.optanon-alert-box-wrapper .optanon-alert-box-button-middle a { color: #ffffff !important }#optanon #optanon-popup-bottom { background-color: #F7FBFE !important }#optanon.modern #optanon-popup-top, #optanon.modern #optanon-popup-body-left-shading { background-color: #F7FBFE !important }.optanon-alert-box-wrapper { background-color:#F7FBFE !important }.optanon-alert-box-wrapper .optanon-alert-box-bg p { color:#333333 !important }</style><link rel="preload" href="integrator.js"><script type="text/javascript" src="integrator.js"></script><link rel="preload" href="integrator_002.js"><script type="text/javascript" src="integrator_002.js"></script><script src="pubads_impl_216.js" async=""></script><style type="text/css">.MathJax_Hover_Frame {border-radius: .25em; -webkit-border-radius: .25em; -moz-border-radius: .25em; -khtml-border-radius: .25em; box-shadow: 0px 0px 15px #83A; -webkit-box-shadow: 0px 0px 15px #83A; -moz-box-shadow: 0px 0px 15px #83A; -khtml-box-shadow: 0px 0px 15px #83A; border: 1px solid #A6D ! important; display: inline-block; position: absolute}
.MathJax_Menu_Button .MathJax_Hover_Arrow {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 4px; -webkit-border-radius: 4px; -moz-border-radius: 4px; -khtml-border-radius: 4px; font-family: 'Courier New',Courier; font-size: 9px; color: #F0F0F0}
.MathJax_Menu_Button .MathJax_Hover_Arrow span {display: block; background-color: #AAA; border: 1px solid; border-radius: 3px; line-height: 0; padding: 4px}
.MathJax_Hover_Arrow:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_Hover_Arrow:hover span {background-color: #CCC!important}
</style><style type="text/css">#MathJax_About {position: fixed; left: 50%; width: auto; text-align: center; border: 3px outset; padding: 1em 2em; background-color: #DDDDDD; color: black; cursor: default; font-family: message-box; font-size: 120%; font-style: normal; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 15px; -webkit-border-radius: 15px; -moz-border-radius: 15px; -khtml-border-radius: 15px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_About.MathJax_MousePost {outline: none}
.MathJax_Menu {position: absolute; background-color: white; color: black; width: auto; padding: 2px; border: 1px solid #CCCCCC; margin: 0; cursor: default; font: menu; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
.MathJax_MenuItem {padding: 2px 2em; background: transparent}
.MathJax_MenuArrow {position: absolute; right: .5em; padding-top: .25em; color: #666666; font-size: .75em}
.MathJax_MenuActive .MathJax_MenuArrow {color: white}
.MathJax_MenuArrow.RTL {left: .5em; right: auto}
.MathJax_MenuCheck {position: absolute; left: .7em}
.MathJax_MenuCheck.RTL {right: .7em; left: auto}
.MathJax_MenuRadioCheck {position: absolute; left: 1em}
.MathJax_MenuRadioCheck.RTL {right: 1em; left: auto}
.MathJax_MenuLabel {padding: 2px 2em 4px 1.33em; font-style: italic}
.MathJax_MenuRule {border-top: 1px solid #CCCCCC; margin: 4px 1px 0px}
.MathJax_MenuDisabled {color: GrayText}
.MathJax_MenuActive {background-color: Highlight; color: HighlightText}
.MathJax_MenuDisabled:focus, .MathJax_MenuLabel:focus {background-color: #E8E8E8}
.MathJax_ContextMenu:focus {outline: none}
.MathJax_ContextMenu .MathJax_MenuItem:focus {outline: none}
#MathJax_AboutClose {top: .2em; right: .2em}
.MathJax_Menu .MathJax_MenuClose {top: -10px; left: -10px}
.MathJax_MenuClose {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; font-family: 'Courier New',Courier; font-size: 24px; color: #F0F0F0}
.MathJax_MenuClose span {display: block; background-color: #AAA; border: 1.5px solid; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; line-height: 0; padding: 8px 0 6px}
.MathJax_MenuClose:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_MenuClose:hover span {background-color: #CCC!important}
.MathJax_MenuClose:hover:focus {outline: none}
</style><style type="text/css">.MJX_Assistive_MathML {position: absolute!important; top: 0; left: 0; clip: rect(1px, 1px, 1px, 1px); padding: 1px 0 0 0!important; border: 0!important; height: 1px!important; width: 1px!important; overflow: hidden!important; display: block!important; -webkit-touch-callout: none; -webkit-user-select: none; -khtml-user-select: none; -moz-user-select: none; -ms-user-select: none; user-select: none}
.MJX_Assistive_MathML.MJX_Assistive_MathML_Block {width: 100%!important}
</style><style type="text/css">#MathJax_Zoom {position: absolute; background-color: #F0F0F0; overflow: auto; display: block; z-index: 301; padding: .5em; border: 1px solid black; margin: 0; font-weight: normal; font-style: normal; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; -webkit-box-sizing: content-box; -moz-box-sizing: content-box; box-sizing: content-box; box-shadow: 5px 5px 15px #AAAAAA; -webkit-box-shadow: 5px 5px 15px #AAAAAA; -moz-box-shadow: 5px 5px 15px #AAAAAA; -khtml-box-shadow: 5px 5px 15px #AAAAAA; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_ZoomOverlay {position: absolute; left: 0; top: 0; z-index: 300; display: inline-block; width: 100%; height: 100%; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
#MathJax_ZoomFrame {position: relative; display: inline-block; height: 0; width: 0}
#MathJax_ZoomEventTrap {position: absolute; left: 0; top: 0; z-index: 302; display: inline-block; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
</style><style type="text/css">.MathJax_Preview {color: #888}
#MathJax_Message {position: fixed; left: 1px; bottom: 2px; background-color: #E6E6E6; border: 1px solid #959595; margin: 0px; padding: 2px 8px; z-index: 102; color: black; font-size: 80%; width: auto; white-space: nowrap}
#MathJax_MSIE_Frame {position: absolute; top: 0; left: 0; width: 0px; z-index: 101; border: 0px; margin: 0px; padding: 0px}
.MathJax_Error {color: #CC0000; font-style: italic}
</style></head>
    <body><button href="javascript:;" title="focus catcher" class="js-focus-catcher u-screenreader-only" tabindex="-1"></button><div id="popup-references" class="u-composite-layer popup-base-theme" style="top:-60px;left:-60px;visibility:hidden;opacity:0;" aria-hidden="true" aria-label="popup"><svg class="popup-arrow" xmlns="http://www.w3.org/2000/svg" width="33.7" height="18.4" viewBox="0 0 33.7 18.4"><path class="fill" fill="#F7FBFE" d="M1.4 18.4h30.9l-15.5-16.9z"></path><path class="stroke" fill="#98BED7" d="M0 18.4h1.4l15.4-16.9 15.5 16.9h1.4l-16.9-18.4z"></path></svg><div class="popup-arrow popup-arrow-shadow icon--popup-arrow-shadow"></div><div class="popup__references"><div class="popup-base-theme__inner" data-component="SpringerLink-Popup-inner"></div></div><button tabindex="-1" class="popup-close icon--close-btn" data-component="SpringerLink-Popup-close">close</button></div><div id="popup-article-dates" class="u-composite-layer popup-base-theme" style="top:-60px;left:-60px;visibility:hidden;opacity:0;" aria-hidden="true" aria-label="popup"><svg class="popup-arrow" xmlns="http://www.w3.org/2000/svg" width="33.7" height="18.4" viewBox="0 0 33.7 18.4"><path class="fill" fill="#F7FBFE" d="M1.4 18.4h30.9l-15.5-16.9z"></path><path class="stroke" fill="#98BED7" d="M0 18.4h1.4l15.4-16.9 15.5 16.9h1.4l-16.9-18.4z"></path></svg><div class="popup-arrow popup-arrow-shadow icon--popup-arrow-shadow"></div><div class="popup__article-dates"><div class="popup-base-theme__inner" data-component="SpringerLink-Popup-inner"></div></div><button tabindex="-1" class="popup-close icon--close-btn" data-component="SpringerLink-Popup-close">close</button></div><div id="popup-search" class="u-composite-layer popup-search-theme" style="top:-60px;left:-60px;visibility:hidden;opacity:0;" aria-hidden="true" aria-label="popup"><svg class="popup-arrow" xmlns="http://www.w3.org/2000/svg" width="33.7" height="18.4" viewBox="0 0 33.7 18.4"><path class="fill" fill="#F7FBFE" d="M1.4 18.4h30.9l-15.5-16.9z"></path><path class="stroke" fill="#98BED7" d="M0 18.4h1.4l15.4-16.9 15.5 16.9h1.4l-16.9-18.4z"></path></svg><div class="popup-arrow popup-arrow-shadow icon--popup-arrow-shadow"></div><div class="popup__search"><div class="popup-base-theme__inner" data-component="SpringerLink-Popup-inner"></div></div><button tabindex="-1" class="popup-close icon--close-btn" data-component="SpringerLink-Popup-close">close</button></div>
        <noscript><iframe src="//www.googletagmanager.com/ns.html?id=GTM-WCF9Z9"
                      height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>

    <nav class="skip-to">
    <a class="skip-to__link skip-to__link--article" href="#main-content">Skip to main content</a>
        <a class="skip-to__link skip-to__link--contents" href="#article-contents">Skip to sections</a>
</nav>
        <div class="page-wrapper">
            <noscript>
    <div class="nojs-banner u-interface">
        <p>This service is more advanced with JavaScript available, learn more at <a
                href="http://activatejavascript.org" target="_blank" rel="noopener">http://activatejavascript.org</a>
        </p>
    </div>
</noscript>
                    <div id="leaderboard" class="leaderboard u-hide" data-component="SpringerLink.GoogleAds" data-namespace="leaderboard"><div class="leaderboard__wrapper"><p class="leaderboard__label">Advertisement</p><button class="leaderboard__hide" title="Hide this advertisement">Hide</button><div id="doubleclick-leaderboard-ad" class="leaderboard__ad"><div id="google_ads_iframe_270604982/springerlink/40537/article_0__container__" style="border: 0pt none;"><iframe id="google_ads_iframe_270604982/springerlink/40537/article_0" title="3rd party ad content" name="google_ads_iframe_270604982/springerlink/40537/article_0" scrolling="no" marginwidth="0" marginheight="0" style="border: 0px none; vertical-align: bottom;" srcdoc="" width="728" frameborder="0" height="90"></iframe></div></div></div></div>

                <header id="header" class="header u-interface">
        <div class="header__content">
            <div class="header__menu-container">
                    <a id="logo" class="site-logo" href="https://link.springer.com/" title="Go to homepage">
                <div class="u-screenreader-only">SpringerLink</div>
    <svg class="site-logo__springer" width="148" height="30" role="img" focusable="false" aria-hidden="true">
        <image width="148" height="30" alt="" src="/springerlink-static/481091012/images/png/springerlink.png" xlink="http://www.w3.org/1999/xlink" xlink:href="/springerlink-static/481091012/images/svg/springerlink.svg"></image>
    </svg>

    </a>


                    <nav id="search-container" class="u-inline-block">
                        <div class="search">
                            <div class="search__content">
                                <form class="u-form-single-input" action="/search" method="get" role="search">
    <input aria-label="Search SpringerLink" name="query" autocomplete="off" placeholder="Search SpringerLink" type="text">
    <input class="u-hide-text" value="Submit" title="Submit" type="submit">
    <svg class="u-vertical-align-absolute" width="13" height="13" viewBox="222 151 13 13" version="1.1" xmlns="http://www.w3.org/2000/svg" focusable="false" aria-hidden="true" role="presentation">
        <path d="M227 159C228.7 159 230 157.7 230 156 230 154.3 228.7 153 227 153 225.3 153 224 154.3 224 156 224 157.7 225.3 159 227 159L227 159 227 159 227 159ZM230 160.1L231.1 159 233.9 161.7C234.2 162.1 234.2 162.6 233.9 162.9 233.6 163.2 233.1 163.2 232.7 162.9L230 160.1 230 160.1 230 160.1 230 160.1ZM227 161L227 161C224.2 161 222 158.8 222 156 222 153.2 224.2 151 227 151 229.8 151 232 153.2 232 156 232 158.8 229.8 161 227 161L227 161 227 161 227 161 227 161Z" stroke="none" fill-rule="evenodd"></path>
    </svg>
</form>
                            </div>
                        </div>
                    </nav>

                    <nav class="nav-container u-interface">
    <div class="global-nav__wrapper">
        <div class="search-button">
            <a class="search-button__label" href="#search-container" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-search">
                <span class="search-button__title">Search</span><svg width="12" height="12" viewBox="222 151 12 12" version="1.1" xmlns="http://www.w3.org/2000/svg" focusable="false" aria-hidden="true" role="presentation">
                    <path d="M227 159C228.7 159 230 157.7 230 156 230 154.3 228.7 153 227 153 225.3 153 224 154.3 224 156 224 157.7 225.3 159 227 159L227 159 227 159 227 159ZM230 160.1L231.1 159 233.9 161.7C234.2 162.1 234.2 162.6 233.9 162.9 233.6 163.2 233.1 163.2 232.7 162.9L230 160.1 230 160.1 230 160.1 230 160.1ZM227 161L227 161C224.2 161 222 158.8 222 156 222 153.2 224.2 151 227 151 229.8 151 232 153.2 232 156 232 158.8 229.8 161 227 161L227 161 227 161 227 161 227 161Z" stroke="none" fill-rule="evenodd"></path>
                </svg>
            </a>
        </div>

        <ul class="global-nav" data-component="SpringerLink.Menu" data-title="Navigation menu" data-text="Menu">
            <li>
                <a href="https://link.springer.com/">
                    <span class="u-overflow-ellipsis">Home</span>
                </a>
            </li>
            <li>
                <a href="https://link.springer.com/contactus">
                    <span class="u-overflow-ellipsis">Contact us</span>
                </a>
            </li>

                <li class="global-nav__logged-out">
                    <a class="test-login-link" href="https://link.springer.com/signup-login?previousUrl=https%3A%2F%2Flink.springer.com%2Farticle%2F10.1186%2F2196-1115-1-2">
                        <span class="u-overflow-ellipsis">Log in</span>
                    </a>
                </li>

        </ul><div class="main-menu u-composite-layer c-button-dropdown c-button-dropdown--ghost" data-component="SV.Dropdown" data-namespace="Menu" aria-label="button with dropdown options" style="width: 75.3333px;"><button type="button" title="Navigation menu" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown.Menu-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">Menu</span><span class="c-button-dropdown__icon"></span></button><div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown.Menu-dropdown"><ul class="main-menu__content" data-role="button-dropdown__content"><li>
                <a href="https://link.springer.com/" tabindex="-1">
                    <span class="u-overflow-ellipsis">Home</span>
                </a>
            </li><li>
                <a href="https://link.springer.com/contactus" tabindex="-1">
                    <span class="u-overflow-ellipsis">Contact us</span>
                </a>
            </li><li class="global-nav__logged-out">
                    <a class="test-login-link" href="https://link.springer.com/signup-login?previousUrl=https%3A%2F%2Flink.springer.com%2Farticle%2F10.1186%2F2196-1115-1-2" tabindex="-1">
                        <span class="u-overflow-ellipsis">Log in</span>
                    </a>
                </li></ul></div></div>
    </div> 
</nav> 
            </div>

        </div>
    </header>

            

            <main id="main-content" class="main-wrapper" tabindex="-1">
                <div class="main-container uptodate-recommendations-off">
                    <aside class="main-sidebar-left">
                        <div class="main-sidebar-left__content">
                            <div class="test-cover cover-image" itemscope="">
        <a class="test-cover-link" href="https://link.springer.com/journal/40537" title="Journal Of Big Data">
            <img class="test-cover-image" src="1.jpg" alt="Journal Of Big Data" itemprop="image">
        </a>

</div>
                        </div>
                    </aside>
                    <div class="main-body" data-role="NavigationContainer">
                                <div class="cta-button-container cta-button-container--top cta-button-container--stacked u-mb-16 u-hide-two-col">
                    <a href="https://link.springer.com/content/pdf/10.1186%2F2196-1115-1-2.pdf" target="_blank" class="c-button c-button--blue c-button__icon-right gtm-pdf-link" title="Download this article in PDF format" rel="noopener">
            <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" version="1.1"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g fill="#fff"><g transform="translate(12.000000, 5.000000)"><path d="M7 7.3L7 1C7 0.4 6.6 0 6 0 5.4 0 5 0.4 5 1L5 7.3 3.5 5.7C3.1 5.3 2.5 5.3 2.1 5.7L2.1 5.7C1.7 6.1 1.7 6.7 2.1 7.1L5.3 10.3C5.7 10.7 6.3 10.7 6.7 10.3L9.9 7.1C10.3 6.7 10.3 6.1 9.9 5.7L9.9 5.7C9.5 5.3 8.9 5.3 8.5 5.7L7 7.3 7 7.3ZM0 13C0 12.4 0.5 12 1 12L11 12C11.6 12 12 12.4 12 13 12 13.6 11.5 14 11 14L1 14C0.4 14 0 13.6 0 13L0 13Z"></path></g></g></g></svg>
            <span class="hide-text-small">Download</span>
            <span>PDF</span>
        </a>

        </div>



                        <article class="main-body__content">
                            <div xmlns="http://www.w3.org/1999/xhtml" class="FulltextWrapper"><div class="ArticleHeader main-context"><div id="enumeration" class="enumeration"><p><a href="https://link.springer.com/journal/40537" title="Journal Of Big Data"><span class="JournalTitle">Journal Of Big Data</span></a></p><p class="icon--meta-keyline-before"><span class="ArticleCitation_Year"><time datetime="2014-12">December 2014</time>, </span><span class="ArticleCitation_Volume">1:2</span><span class="u-inline-block u-ml-4"> | <a class="gtm-cite-link" href="#citeas">Cite as</a></span></p></div><div class="MainTitleSection"><h1 class="ArticleTitle" lang="en">A review of data mining using big data in health informatics</h1></div><div class="authors u-clearfix authors--enhanced" data-component="SpringerLink.Authors"><ul class="u-interface u-inline-list authors__title" data-role="AuthorsNavigation"><li><a href="#authors" class="gtm-tab-authors selected">Authors</a></li><li><a href="#authorsandaffiliations" class="gtm-tab-authorsandaffiliations">Authors and affiliations</a></li></ul><span class="marker" style="width: 160px;"></span><div class="authors__list" data-role="AuthorsList" id="authors"><ul class="test-contributor-names"><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors__name">Matthew&nbsp;Herland</span></li><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors__name">Taghi&nbsp;M&nbsp;Khoshgoftaar</span></li><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors__name">Randall&nbsp;Wald</span><span class="author-information"><span class="authors__contact"><a href="mailto:rwald1@fau.edu" title="rwald1@fau.edu" itemprop="email" class="gtm-email-author">Email author</a></span></span></li></ul></div><div class="authors__affiliations" id="authorsandaffiliations"><div class="authors-affiliations u-interface"><ul class="test-contributor-names"><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors-affiliations__name">Matthew&nbsp;Herland</span><ul class="authors-affiliations__indexes u-inline-list" data-role="AuthorsIndexes"><li data-affiliation="affiliation-1">1</li></ul></li><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors-affiliations__name">Taghi&nbsp;M&nbsp;Khoshgoftaar</span><ul class="authors-affiliations__indexes u-inline-list" data-role="AuthorsIndexes"><li data-affiliation="affiliation-1">1</li></ul></li><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors-affiliations__name">Randall&nbsp;Wald</span><ul class="authors-affiliations__indexes u-inline-list" data-role="AuthorsIndexes"><li data-affiliation="affiliation-1">1</li></ul><span class="author-information"><span class="author-information__contact u-icon-before icon--email-before"><a href="mailto:rwald1@fau.edu" title="rwald1@fau.edu" itemprop="email" class="gtm-email-author">Email author</a></span></span></li></ul><ol class="test-affiliations"><li class="affiliation" data-test="affiliation-1" data-affiliation-highlight="affiliation-1" itemscope="" itemtype="http://schema.org/Organization"><span class="affiliation__count">1.</span><span class="affiliation__item"><span itemprop="name" class="affiliation__name">Florida Atlantic University</span><span itemprop="address" itemscope="" itemtype="http://schema.org/PostalAddress" class="affiliation__address"><span itemprop="addressRegion" class="affiliation__city">Boca Raton</span><span itemprop="addressCountry" class="affiliation__country">USA</span></span></span></li></ol></div></div></div><div class="main-context__container" data-component="SpringerLink.ArticleMetrics"><div class="main-context__column"><span><div id="open-choice-icon" class="open-access u-text-separator-after-micro"><abbr title="This content is freely available to anyone, anywhere, at any time">Open Access</abbr></div><span class="test-render-category">Survey paper</span></span><div class="article-dates article-dates--enhanced" data-component="SpringerLink.ArticleDates"><div class="article-dates__entry"><span class="article-dates__label">First Online: </span><span class="article-dates__first-online"><a href="#article-dates-history" class="gtm-first-online" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-article-dates">24 June 2014</a></span></div><div class="article-dates__history" id="article-dates-history"><div class="article-dates__entry"><span class="article-dates__label">Received: </span><span><time datetime="2013-10-12">12 October 2013</time></span></div><div class="article-dates__entry"><span class="article-dates__label">Accepted: </span><span><time datetime="2014-02-20">20 February 2014</time></span></div></div></div></div><div class="main-context__column">    <ul id="book-metrics" class="article-metrics u-sansSerif">
            <li class="article-metrics__item">
                    <a class="article-metrics__link gtm-socialmediamentions-count" href="http://www.altmetric.com/details.php?citation_id=2460578&amp;domain=link.springer.com" target="_blank" rel="noopener" title="Visit Altmetric for full social mention details" id="socialmediamentions-link">
                            <span id="socialmediamentions-count-number" class="test-metric-count c-button-circle gtm-socialmediamentions-count">32</span>
                       <span class="test-metric-name article-metrics__label gtm-socialmediamentions-count">Shares</span>
                    </a>
            </li>
            <li class="article-metrics__item">
                     <span class="article-metrics__views">47k</span>
                     <span class="article-metrics__label">Downloads</span>
            </li>
            <li class="article-metrics__item">
                    <a class="article-metrics__link gtm-citations-count" href="https://citations.springer.com/item?doi=10.1186/2196-1115-1-2" target="_blank" rel="noopener" title="Visit Springer Citations for full citation details" id="citations-link">
                            <span id="citations-count-number" class="test-metric-count c-button-circle gtm-citations-count">53</span>
                       <span class="test-metric-name article-metrics__label gtm-citations-count">Citations</span>
                    </a>
            </li>
    </ul>
</div></div></div><section class="Abstract" id="Abs1" tabindex="-1" lang="en"><h2 class="Heading">Abstract</h2><p class="Para">The
 amount of data produced within Health Informatics has grown to be quite
 vast, and analysis of this Big Data grants potentially limitless 
possibilities for knowledge to be gained. In addition, this information 
can improve the quality of healthcare offered to patients. However, 
there are a number of issues that arise when dealing with these vast 
quantities of data, especially how to analyze this data in a reliable 
manner. The basic goal of Health Informatics is to take in real world 
medical data from all levels of human existence to help advance our 
understanding of medicine and medical practice. This paper will present 
recent research using Big Data tools and approaches for the analysis of 
Health Informatics data gathered at multiple levels, including the 
molecular, tissue, patient, and population levels. In addition to 
gathering data at multiple levels, multiple levels of questions are 
addressed: human-scale biology, clinical-scale, and epidemic-scale. We 
will also analyze and examine possible future work for each of these 
areas, as well as how combining data from each level may provide the 
most promising approach to gain the most knowledge in Health 
Informatics.</p></section><div class="KeywordGroup" lang="en"><h2 class="Heading">Keywords</h2><span class="Keyword">Big data&nbsp;</span><span class="Keyword">Health informatics&nbsp;</span><span class="Keyword">Bioinformatics&nbsp;</span><span class="Keyword">Neuroinformatics&nbsp;</span><span class="Keyword">Clinical informatics&nbsp;</span><span class="Keyword">Public health informatics&nbsp;</span><span class="Keyword">Social media&nbsp;</span></div><div class="article-actions--inline" id="article-actions--inline" data-component="article-actions--inline"><div class="citations u-interface c-button-dropdown" data-component="SV.Dropdown" data-namespace="citations--inline" aria-label="button with dropdown options">
        <button type="button" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown.citations--inline-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">
    <span>Cite</span>
    <span class="hide-text-small">article</span>
</span><span class="c-button-dropdown__icon"></span></button>
<div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown.citations--inline-dropdown"><ul class="citations__content" data-role="button-dropdown__content">
    <li>
        <a href="#citeas" class="gtm-cite-dropdown" tabindex="-1">How to cite?</a>
    </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1186/2196-1115-1-2?format=refman&amp;flavour=citation" title="Download this article's citation as a .RIS file" class="gtm-export-citation" data-gtmlabel="RIS" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="RIS">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .RIS
                </span>
                <span class="citations__types">
                        <span>
                            Papers
                        </span>
                        <span>
                            Reference Manager
                        </span>
                        <span>
                            RefWorks
                        </span>
                        <span>
                            Zotero
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1186/2196-1115-1-2?format=endnote&amp;flavour=citation" title="Download this article's citation as a .ENW file" class="gtm-export-citation" data-gtmlabel="ENW" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="ENW">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .ENW
                </span>
                <span class="citations__types">
                        <span>
                            EndNote
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1186/2196-1115-1-2?format=bibtex&amp;flavour=citation" title="Download this article's citation as a .BIB file" class="gtm-export-citation" data-gtmlabel="BIB" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="BIB">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .BIB
                </span>
                <span class="citations__types">
                        <span>
                            BibTeX
                        </span>
                        <span>
                            JabRef
                        </span>
                        <span>
                            Mendeley
                        </span>
                </span>
            </a>
        </li>
</ul></div>
    </div></div><div id="body"><section id="Sec1" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Introduction<span class="section-icon"></span></h2><div class="content"><p class="Para">The
 field of Health Informatics is on the cusp of its most exciting period 
to date, entering a new era where technology is starting to handle Big 
Data, bringing about unlimited potential for information growth. Data 
mining and Big Data analytics are helping to realize the goals of 
diagnosing, treating, helping, and healing all patients in need of 
healthcare, with the end goal of this domain being improved Health Care 
Output (HCO), or the quality of care that healthcare can provide to end 
users (i.e. patients).</p><p class="Para">Health Informatics is a 
combination of information science and computer science within the realm
 of healthcare. There are numerous current areas of research within the 
field of Health Informatics, including Bioinformatics, Image Informatics
 (e.g. Neuroinformatics), Clinical Informatics, Public Health 
Informatics, and also Translational BioInformatics (TBI). Research done 
in Health Informatics (as in all its subfields) can range from data 
acquisition, retrieval, storage, analytics employing data mining 
techniques, and so on. However, the scope of this study will be research
 that uses data mining in order to answer questions throughout the 
various levels of health.</p><p class="Para">Each of the studies done in a particular subfield of Health Informatics utilizes data from a particular level of human existence
[<span class="CitationRef"><a href="#CR1" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1</a></span>]:
 Bioinformatics uses molecular level data, Neuroinformatics employs 
tissue level data, Clinical Informatics applies patient level data, and 
Public Health Informatics utilizes population data (either from the 
population or on the population). These subfields do sometimes overlap 
(for example, a single study might consider data from two adjacent 
levels), but in the interest of minimizing confusion we in this work 
will classify a study based on the highest data level used (as this 
paper will be structured according to data usage). In addition, within a
 given data level we will break down studies based on the type (i.e., 
level) of question a study attempts to answer, where each question level
 is of a relatively comparable scope to one of the data levels. The 
tissue level is of analogous scope to the human-scale biology questions,
 the scope of patient data is related to clinical questions and the 
scope of the population data is comparable to the epidemic-level 
questions.</p><p class="Para">The scope of data used by the subfield 
TBI, on the other hand, exploits data from each of these levels, from 
the molecular level to entire populations
[<span class="CitationRef"><a href="#CR1" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1</a></span>].
 In particular, TBI is specifically focused on integrating data from the
 Bioinformatics level with the higher levels, because traditionally this
 level has been isolated in the laboratory and separated from the more 
patient-facing levels (Neuroinformatics, Clinical Informatics, and 
Population Informatics). TBI and the idea of combining data from all 
levels of human existence is a popular new direction in Health 
Informatics. The main level of questions that TBI ultimately tries to 
answer are on the clinical level, as such answers can help improve HCO 
for patients. Research throughout all levels of accessible data, using 
various data mining and analytical techniques, can be used to help the 
healthcare system make decisions faster, more accurately, and more 
efficiently, all in a more cost-effective manner than without using such
 methods.</p><p class="Para">This paper is organized as follows: Section
 “Big data in health informatics” provides a general background on Big 
Data in Health Informatics. Section “Levels of health informatics data” 
delineates and discusses the subfields: Bioinformatics, 
Neuroinformatics, Clinical Informatics, and Public Health Informatics, 
as well as discussing their corresponding data and question levels (used
 to structure Sections “1” through 1). Section “Using micro level data –
 Molecules” looks at studies using data from the micro (molecular) level
 while the studies in Section “Using tissue level data” use data from 
the tissue level, Section “Using patient level data” covers research 
being done on the patient level, and the research efforts in Section 
“Using population level data – Social media” employ data gathered at the
 population level. The sub-field of TBI is discussed in Section 
“Translational bioinformatics”. Section “Analysis and future works” 
analyzes the works shown and discusses both possible future work and the
 possible directions each line of research could take. Section 
“Conclusion” contains our conclusion.</p></div></section><section id="Sec2" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Big data in health informatics<span class="section-icon"></span></h2><div class="content"><p class="Para">The term Big Data is a vague term with a definition that is not universally agreed upon. According to
[<span class="CitationRef"><a href="#CR2" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2</a></span>], a rough definition would be any data that is around a petabyte (10<sup>15</sup>
 bytes) or more in size. In Health Informatics research though, Big Data
 of this size is quite rare; therefore, a more encompassing definition 
will be used here to incorporate more studies, specifically a definition
 by Demchenko et al.
[<span class="CitationRef"><a href="#CR3" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">3</a></span>]
 who define Big Data by five V’s: Volume, Velocity, Variety, Veracity, 
and Value. Volume pertains to vast amounts of data, Velocity applies to 
the high pace at which new data is generated, Variety pertains to the 
level of complexity of the data, Veracity measures the genuineness of 
the data, and Value evaluates how good the quality of the data is in 
reference to the intended results.</p><p class="Para">Data gathered for 
Health Informatics research does exhibit many of these qualities. Big 
Volume comes from large amounts of records stored for patients: for 
example, in some datasets each instance is quite large (e.g. datasets 
using MRI images or gene microarrays for each patient), while others 
have a large pool with which to gather data (such as social media data 
gathered from a population). Big Velocity occurs when new data is coming
 in at high speeds, which can be seen when trying to monitor real-time 
events whether that be monitoring a patient’s current condition through 
medical sensors or attempting to track an epidemic through multitudes of
 incoming web posts (such as from Twitter). Big Variety pertains to 
datasets with a large amount of varying types of independent attributes,
 datasets that are gathered from many sources (e.g. search query data 
comes from many different age groups that use a search engine), or any 
dataset that is complex and thus needs to be seen at many levels of data
 throughout Health Informatics. High Veracity of data in Health 
Informatics, as in any field using analytics, is a concern when working 
with possibly noisy, incomplete, or erroneous data (as could be seen 
from faulty clinical sensors, gene microarrays, or from patient 
information stored in databases) where such data needs to be properly 
evaluated and dealt with. High Value of data is seen all throughout 
Health Informatics as the goal is to improve HCO. Although data gathered
 by traditional methods (such as in a clinical setting) is widely 
regarded as High Value, the value of data gathered by social media (data
 submitted by anyone) may be in question; however, as shown in Section 
“Using population level data – Social media”, this can also have High 
Value.</p><p class="Para">It should be mentioned that not all the 
studies covered here or in the field of Health Informatics fit all 5 of 
the qualities in Demchenko et al.’s. definition of Big Data. Even so, 
many still impose significant computational constraints that need to be 
addressed in one way or another. Offline storage of datasets such as 
Electronic Health Records (EHR) can be difficult even if the data does 
not exhibit Big Velocity or Variety, and high-throughput processing seen
 in real-time continuous data requires capable and efficient techniques 
even when each individual data instance does not have Big Volume. Data 
that has Big Value without Big Veracity may need complex methods to find
 a consensus among various models, or which require time-consuming 
adjustments to the data which could expand the size of the dataset. 
Therefore, even datasets lacking Big Volume can still have Big Data 
problems, meaning that the Big Data definitions mainly focusing on 
Volume and Velocity may not be considering enough qualities of the 
dataset to fully characterize it.</p><p class="Para">It is noted in
[<span class="CitationRef"><a href="#CR4" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">4</a></span>] and
[<span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">5</a></span>]
 that just in the United States, using data mining in Health Informatics
 can save the healthcare industry up to $450 billion each year. This is 
because the field of Health Informatics generates a large and growing 
amount of data. As of 2011, health care organizations had generated over
 150 exabytes of data
[<span class="CitationRef"><a href="#CR4" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">4</a></span>]
 (one exabyte is 1000 petabytes). This data needs to be sifted through 
and efficiently analyzed in order to be of any use to the health care 
system. As mentioned, health information can be stored in EHRs, which 
can store 44+ petabytes of patient data, which on top of this health 
data can come in many other forms. This explosion of data that is seen 
in Health Informatics has also been noticed in Bioinformatics as well, 
where genomic sequencing can generate many terabytes of data. With data 
coming from many different places and in many different forms, it is up 
to the Health Informatics community to find ways of dealing with all 
this data. It would seem that it is becoming more and more popular to 
integrate and combine different sources of data, even across different 
subfields (i.e. Translational Bioinformatics), and even across Health 
Informatics and Bioinformatics. Successful integration of this huge 
amount of data could lead to a huge improvement for the end users of the
 health care system, i.e. patients.</p></div></section><section id="Sec3" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Levels of health informatics data<span class="section-icon"></span></h2><div class="content"><p class="Para">This
 section will be describing various subfields of Health Informatics: 
Bioinformatics, Neuroinformatics, Clinical Informatics, and Public 
Health Informatics. The lines between each subfield of Health 
Informatics can be blurred in terms of definition, confusing which 
subfield a study should fall under; therefore, this paper will be 
deciding subfield membership by the highest level of data used for 
research and will be the organizing factor for Sections “1” through 1. 
The works from the subfield of Bioinformatics discussed in this study 
consist of research done with molecular data (Section “Using micro level
 data – Molecules”), Neuroinformatics is a form of Medical Image 
Informatics which uses image data of the brain, and thus it falls under 
tissue data (Section “Using tissue level data”), Clinical Informatics 
here uses patient data (Section “Using patient level data”), and Public 
Health Informatics makes use of data either about the population or from
 the population (Section “Using population level data – Social media”).</p><div class="Para">In
 Health Informatics research, there are two sets of levels which must be
 considered: the level from which the data is collected, and the level 
at which the research question is being posed. The four subfields 
discussed in this study correspond to the data levels, but the question 
level in a given work may be different from its data level. These 
question levels are of similar scope to the data levels: the tissue 
level data is of similar scope to human-scale biology questions, the 
patient level data is of comparable scope to clinical questions, and the
 population level data is of proportionate scope to epidemic-scale 
questions. Each section will be further sub-sectioned by question level 
starting with the lowest to the highest. Table
<span class="InternalRef"><a href="#Tab1">1</a></span> summarizes the 
breakdown of Sections “1” through “Using population level data – Social 
media” including the sections (by data levels used), subsections (by 
question level used), and the questions the studies discussed are 
attempting to answer.<div id="Tab1" class="Table"><div class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Table 1</span> <p class="SimplePara"> <strong class="EmphasisTypeBold ">Summary of Studies Covered</strong> </p> </div></div><div class="u-scroll-horizontal"><table><colgroup><col class="tcol1 align-left"><col class="tcol2 align-center"><col class="tcol3 align-center"><col class="tcol4 align-center"><col class="tcol5 align-center"></colgroup><thead><tr><th> <p class="SimplePara">Sections</p> </th><th> <p class="SimplePara">Data level(s) used</p> </th><th> <p class="SimplePara">Subsections</p> </th><th> <p class="SimplePara">Question level(s) answered</p> </th><th> <p class="SimplePara">Questions to be answered</p> </th></tr></thead><tbody><tr><td> <p class="SimplePara">Using Micro Level Data – Molecules</p> </td><td> <p class="SimplePara">Molecular</p> </td><td> <p class="SimplePara">Using Gene Expression Data to Make Clinical Predictions</p> </td><td> <p class="SimplePara">Clinical</p> </td><td> <p class="SimplePara">What sub-type of cancer does a patient have?
[<span class="CitationRef"><a href="#CR6" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">6</a></span>]</p> </td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td> <p class="SimplePara">Will a patient have a relapse of cancer?
[<span class="CitationRef"><a href="#CR7" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">7</a></span>]</p> </td></tr><tr><td> <p class="SimplePara">Using Tissue Level Data</p> </td><td> <p class="SimplePara">Tissue</p> </td><td> <p class="SimplePara">Creating a Connectivity Map of the Brain Using Brain Images</p> </td><td> <p class="SimplePara">Human-Scale Biology</p> </td><td> <p class="SimplePara">Can a full connectivity map of the brain be made
[<span class="CitationRef"><a href="#CR8" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">8</a></span>, <span class="CitationRef"><a href="#CR9" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">9</a></span>]?</p> </td></tr><tr><td>&nbsp;</td><td> <p class="SimplePara">Patient</p> </td><td> <p class="SimplePara">Using MRI Data for Clinical Prediction</p> </td><td> <p class="SimplePara">Clinical</p> </td><td> <p class="SimplePara">Do particular areas of the brain correlate to clinical events?
[<span class="CitationRef"><a href="#CR10" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">10</a></span>]</p> </td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td> <p class="SimplePara">What level of Alzheimer’s disease does a patient have?
[<span class="CitationRef"><a href="#CR11" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">11</a></span>]</p> </td></tr><tr><td> <p class="SimplePara">Using Patient Level Data</p> </td><td> <p class="SimplePara">Patient</p> </td><td> <p class="SimplePara">Prediction of ICU Readmission and Mortality Rate</p> </td><td> <p class="SimplePara">Clinical</p> </td><td> <p class="SimplePara">Should a patient be released from the ICU, or would they benefit from a longer stay?
[<span class="CitationRef"><a href="#CR12" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">12</a></span>, <span class="CitationRef"><a href="#CR13" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">13</a></span>, <span class="CitationRef"><a href="#CR14" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">14</a></span>]</p> </td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td> <p class="SimplePara">What is the 5 year expectancy of a patient over the age of 50?
[<span class="CitationRef"><a href="#CR15" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">15</a></span>]</p> </td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td> <p class="SimplePara">Real-Time Predictions Using Data Streams</p> </td><td> <p class="SimplePara">Clinical</p> </td><td> <p class="SimplePara">What ailment does a patient have (real-time prediction)
[<span class="CitationRef"><a href="#CR16" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">16</a></span>, <span class="CitationRef"><a href="#CR17" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">17</a></span>]</p> </td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td> <p class="SimplePara">Is an infant experiencing a cardiorespiratory spell (real-time)?
[<span class="CitationRef"><a href="#CR18" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">18</a></span>]</p> </td></tr><tr><td> <p class="SimplePara">Using Population Level Data – Social Media</p> </td><td> <p class="SimplePara">Population</p> </td><td> <p class="SimplePara">Using Message Board Data to Help Patients Obtain Medical Information</p> </td><td> <p class="SimplePara">Clinical</p> </td><td> <p class="SimplePara">Can message post data be used for dispersing clinically reliable information?
[<span class="CitationRef"><a href="#CR19" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">19</a></span>, <span class="CitationRef"><a href="#CR20" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">20</a></span>]</p> </td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td> <p class="SimplePara">Tracking Epidemics Using Search Query Data</p> </td><td> <p class="SimplePara">Epidemic-Scale</p> </td><td> <p class="SimplePara">Can search query data be used to accurately track epidemics throughout a population?
[<span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">5</a></span>, <span class="CitationRef"><a href="#CR21" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">21</a></span>]</p> </td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td> <p class="SimplePara">Tracking Epidemics Using Twitter Post Data</p> </td><td> <p class="SimplePara">Epidemic-Scale</p> </td><td> <p class="SimplePara">Can Twitter post data be used to accurately track epidemics throughout a population?
[<span class="CitationRef"><a href="#CR22" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">22</a></span>, <span class="CitationRef"><a href="#CR23" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">23</a></span>]</p> </td></tr></tbody></table></div></div></div><section id="Sec4" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Bioinformatics</h3><p class="Para">Research
 in Bioinformatics may not be considered as part of traditional Health 
Informatics, but the research done in Bioinformatics is an important 
source of health information at various levels. Bioinformatics focuses 
on analytical research in order to learn how the human body works using 
molecular level data in addition to developing methods of effectively 
handling said data. The increasing amount of data here has greatly 
increased the importance of developing data mining and analysis 
techniques which are efficient, sensitive, and better able to handle Big
 Data.</p><p class="Para">Data in Bioinformatics, such as gene 
expression data, is continually growing (due to technology being able to
 generate more molecular data per individual), and is certainly 
classifiable as Big Volume. The issue of Big Volume within molecular 
data leads to research such as McDonald et al.
[<span class="CitationRef"><a href="#CR24" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">24</a></span>]
 who created a Bioinformatics suite of software tools they call khmer. 
This suite seeks to solve hardware computational problems through 
software. The tools in this suite pre-process Big Volume genomic 
sequence data by breaking up long sequences into relatively short 
strings which can be stored in a Bloom filter-based hash table, helping 
both the ability and efficiency of analysis of Bioinformatics data.</p></section><section id="Sec5" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Neuroinformatics</h3><p class="Para">Neuroinformatics
 research is a young subfield, as each data instance (such as MRIs) is 
quite large leading to datasets with Big Volume. Only recently can 
computational power keep up with the demands of such research. 
Neuroinformatics concentrates its research on analysis of brain image 
data (tissue level) in order to: learn how the brain works, find 
correlations between information gathered from brain images to medical 
events, etc., all with the goal of furthering medical knowledge at 
various levels. We chose the field of Neuroinformatics to represent the 
broader domain of Medical Image Informatics because by limiting the 
scope to brain images, more in-depth research may be performed while 
still gathering enough information to constitute Big Data. From this 
point on Neuroinformatics research using tissue level data will be 
referenced by data level rather than the subfield.</p></section><section id="Sec6" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Clinical informatics</h3><p class="Para">Clinical
 Informatics research involves making predictions that can help 
physicians make better, faster, more accurate decisions about their 
patients through analysis of patient data. Clinical questions are the 
most important question level in Health Informatics as it works directly
 with the patient. This is where a confusion can arise with the term 
“clinical” when found in research, as all Health Informatics research is
 performed with the eventual goal of predicting “clinical” events 
(directly or indirectly). This confusion is the reason for defining 
Clinical Informatics as only research which directly uses patient data. 
With this, data used by Clinical Informatics research has Big Value.</p><p class="Para">Even with all research eventually helping answer clinical realm events, according to Bennett et al.
[<span class="CitationRef"><a href="#CR25" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">25</a></span>]
 there is about a 15 ±2 year gap between clinical research and the 
actual clinical care used in practice. Decisions these days are made 
mostly on general information that has worked before, or based on what 
experts have found to work in the past. Through all the research 
presented here as well as with all the research being done in Health 
Informatics, the healthcare system can embrace new ways that can be more
 accurate, reliable, and efficient.</p></section><section id="Sec7" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Public health informatics – Social media</h3><p class="Para">Public
 Health Informatics applies data mining and analytics to population 
data, in order to gain medical insight. Data in Public Health 
Informatics is from the population, gathered either from “traditional” 
means (experts or hospitals) or gathered from the population (social 
media). In either event, population data has Big Volume, along with Big 
Velocity and Big Variety. Data gathered from the population through 
social media could possibly have low Veracity leading to low Value, but 
techniques for extracting the useful information from social media (such
 as Twitter posts), this line of data can also have Big Value. cdot</p></section></div></section><section id="Sec8" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Using micro level data – Molecules<span class="section-icon"></span></h2><div class="content"><p class="Para">The
 studies covered in this section use data gathered from the molecular 
level, and answer the clinical question of using gene expression data 
for prediction of clinical outcome. Molecule-level data frequently 
experiences the problem of “high dimensionality,” where the data has a 
large number of independent attributes; this is because molecule-level 
data tends to have thousands (or tens of thousands) of possible 
molecules, configurations of molecules, or molecule-molecule 
interactions, and these are represented in datasets as features. This 
high dimensionality can stymie approaches which do not consider feature 
selection to address this form of Big Data. Although we only focus on 
one form of clinical question in this paper, other applications of 
micro-level data exist, such as cheminformatics
[<span class="CitationRef"><a href="#CR26" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">26</a></span>], high-throughput screening
[<span class="CitationRef"><a href="#CR27" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">27</a></span>], and DNA sequence analysis
[<span class="CitationRef"><a href="#CR28" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">28</a></span>].</p><section id="Sec9" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Using gene expression data for prediction of clinical outcome</h3><p class="Para">As
 stated, the studies in this subsection use gene expression data to 
answer clinical questions. Two research efforts are reviewed in this 
subsection, both of which focus on cancer: the first uses gene 
expression profiling to categorize leukemia into two different 
subclasses, while the second study uses gene expression data to predict 
relapse among patients in the early stages of colorectal cancer (CRC). 
Both of these studies (as well as similar studies) can help physicians 
guide, advise and treat their cancer patients.</p><p class="Para">Haferlach et al.
[<span class="CitationRef"><a href="#CR6" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">6</a></span>]
 formulated a gene expression profiling classifier to place patients 
into 18 different subclasses of either myeloid or lymphoid leukemia. 
This study used 3,334 patients where about two-thirds were used for 
training (2,143 patients) and the rest one-third for testing (1,191 
patients), and from each patient 54,630 gene probe set samples were 
taken (3,334 × 54,630 ≈ 182 million). The authors chose to use an 
all-pairwise classification design using the trimmed mean of the 
difference between perfect match and mismatch intensities with quantile 
normalization, all to handle the multiclass nature of this research. 
This technique, called Difference of Quantile Normalized values (DQN), 
is explained in greater detail in
[<span class="CitationRef"><a href="#CR29" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">29</a></span>]
 by Liu et al. There are 153 class pairs created due to there being 18 
distinct classes ((18 × 17)/2 = 153) where for each pair there will be a
 linear binary classifier created using Support Vector Machines (SVM)
[<span class="CitationRef"><a href="#CR30" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">30</a></span>].
 They tested their method on the training pool using 30-fold 
cross-validation, where each of the 30 runs used the top 100 probe sets 
with the highest t-statistic for each class pair. The results of this 
testing yielded a mean specificity of 99.7% and an accuracy of 92.2%.</p><p class="Para">The
 second part of their study was to get the results for the testing pool 
of patients where Haferlach et al. achieve a median specificity of 99.8%
 and a median sensitivity of 95.6% for the classification of 14 
subclasses of acute leukemia where 6 are lymphoid and 8 are myeloid. The
 authors also claim that using their methods with microarray data, on 
57% of the discrepant instances they were able to achieve better results
 than with routine diagnostic methods. This research has shown promise 
that using microarray gene expression data patients can be reliably 
classified into different forms of leukemia, yet this study could have 
done more by exploiting any number of feature selection techniques (or 
in this case gene probe selection). Using feature selection techniques 
in this study could also lead to determining which gene probes have a 
strong correlation between different forms of leukemia.</p><p class="Para">Salazar et al.
[<span class="CitationRef"><a href="#CR7" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">7</a></span>]
 construct a gene expression classifier seeking to predict, within a 
five-year period, whether or not a patient will relapse back into having
 CRC. The authors gathered a total of 394 patients where 188 were used 
for training and 206 were used for validation. The training pool was 
accumulated over a 19 year period (1983-2002) from three different 
institutions from different counties, and the validation pool was 
assembled over 8 years (1996-2004) from another institution from a 
different country. They decided on a set of 33,834 gene probes that were
 found to have variation within the patients from the test group (33,834
 × 188 ≈ 6.4 million not counting the probes that showed no 
correlation). The authors decided for feature selection (gene probe 
selection) to use a leave-one-out cross-validation method to determine 
which gene probes were strongly correlated with the 5-year distant 
metastasis-free survival (DMFS) with a t-test as the deciding factor. 
After this selection method they ended up with an optimal set of 18 gene
 probes.</p><p class="Para">For classification, Salazar et al. used a 
Nearest Centroid-Base Classifier (NCBC) named ColoPrint. A basic 
explanation of a centroid based classification system would be: when a 
new instance is ready to be classified it is assigned to the class 
consisting of the training samples whose average centroid is closest to 
the new instance. ColoPrint also incorporates feature reduction to 
alleviate the problem of high dimensionality, by using a predefined set 
of 18 genes which were previously found to be useful in identifying CRC.
 With ColoPrint, a patient can be classified as either low or high risk.
 The authors note that in a similar study (for breast cancer prognostic 
tests
[<span class="CitationRef"><a href="#CR31" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">31</a></span>])
 that centroid-based classifiers have shown to give reliable results. 
The authors compared ColoPrint to an assortment of clinical factors 
through multivariate analysis to see which could best predict Relapse 
Free Survival, and ColoPrint was found to be one of the most significant
 factors with a hazard ratio of 2.69 at a confidence interval of 95% and
 a P value of 0.003. The hazard ratio shows the proportion of relapse 
rates of predicted-relapse and predicted-RFS patients). Through this 
study, they were able to find which patients were having a generally 
higher probability of being marked for high risk than other patients 
(e.g. males are more often marked as high risk than females).</p><p class="Para">The
 approach the authors took by gathering data from different sources for 
the training set (giving variation to the learning) and even another 
source for validation (giving variation to the testing) is a notably 
good strategy and is a strategy future research should employ. One place
 variation could have been added, though, is using more probe selection 
techniques: this would help test if the ColoPrint feature subset was 
indeed “optimal”, especially since there were 33,834 probes to start 
with.</p><p class="Para">Both of the studies discussed in this 
subsection are showing the usefulness of microarray gene expression data
 as they can be used for determining both: if a patient will relapse 
back into cancer as well as which subtype of cancer a patient has. The 
results of these research efforts lead to the idea that microarray data 
could give similar results if these procedures were applied to other 
types of cancers in order to help physicians both diagnose and begin to 
treat their patients.</p></section></div></section><section id="Sec10" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Using tissue level data<span class="section-icon"></span></h2><div class="content"><p class="Para">All
 the studies covered in this section will cover data at the tissue level
 and venture to answer human-scale biology questions including: creating
 a full connectivity map of the brain, and predicting clinical outcomes 
by using MRI data. This level, which incorporates imaging data, brings 
in a number of additional Big Data challenges, such as feature 
extraction and managing complex images. Studies which combine imaging 
data with other data sources also exemplify the Variety aspect of Big 
Data, by building models which incorporate diverse data sources.</p><section id="Sec11" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Creating a connectivity map of the brain using brain images</h3><p class="Para">This
 subsection will be covering two studies with the goal of answering 
human-scale biology questions attempting to develop a comprehensive 
connectivity diagram of the human brain. A diagram such as this can 
provide many opportunities for health information gain for physicians 
for prognosis, diagnosis, treatments, etc. through the execution of 
analysis and novel data mining.</p><p class="Para">There is a huge and 
ongoing project called the Human Connectome Project (HCP) led by the 
WU-Minn HCP consortium where the goal is to eventually map the human 
brain by making a comprehensive connectivity diagram. According to
[<span class="CitationRef"><a href="#CR9" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">9</a></span>]
 HCP is looking to find a map of the neural pathways that make up the 
brain in order to advance current knowledge of how the brain functions 
and behaves region-to-region. The project is broken into two phases: 1.)
 in the first 2 years (Fall 2010- Spring 2012), methods for data 
acquisition and analysis were improved, 2.) in the last 3 years (summer 
2012-Summer 2015), these methods will be applied to 1200 healthy adults 
(between the ages of 22 and 35 from varying ethnic groups) using top of 
the line methods of noninvasive neuroimaging. The subjects used are 
twins and non-twin siblings in order to also determine variability and 
heritability factors in brain structure and connectivity throughout such
 cases. This project is a five year long project that was started in 
fall 2010 and should be finished in 2015. The HCP data generated is 
being made freely accessible to the public, and the first and second 
quarterly data are now available at: http://www.humanconnectome.org/ 
containing the data generated from a total of 148 of the 1200 
participants (about 12% of the total). The data available includes image
 data (T1w and T2w MRI, rfMRI, tfMRI, dMRI) as well as behavioral 
measures for a current total of about 4.5 TB. Data will be further 
released at a quarterly rate with each quarter including about 100 new 
participants.</p><p class="Para">According to Van Essen et al.
[<span class="CitationRef"><a href="#CR9" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">9</a></span>],
 the project looks very promising because new and extremely important 
information can be gleaned from mining the HCP data coming out from the 
HCP consortium’s research. Creating a full connectivity map of the brain
 could lead to information that could help in determining the reasons 
why people have certain brain disorders at a level previously 
unattainable, giving physician a possibility for easier diagnosis, early
 detection of future illnesses or maybe even prevention of mental or 
physical ailments. Considering this data is only recently released, the 
research applying this data is all in the future, but with the 
technology being so advanced there is endless possibility for studies 
employing this data for heath information gain. Once all the data for 
the 1200 patients have been generated, there could be similar data 
created on patients with various ailments and various ages to find the 
differences between such brains through data mining and analysis.</p><p class="Para">Annese
[<span class="CitationRef"><a href="#CR8" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">8</a></span>]
 develops a system to link MRI (primarily diffusion tensor imaging 
(DTI)) measurements to that of an actual brain sample gathered from 
histological methods (physical studying of tissue) arguing that it is 
not enough to just depend on neuroimaging for making a comprehensive 
connectivity map of brain connectivity, but should also include 
histological methods of studying of actual brain tissue. The author 
mentions that the difficulties of matching up MRI measurements with that
 of anatomical measurements is a considerable issue interfering with 
making a comprehensive connectivity model of the human brain. MRIs are 
large and have high resolution, a higher resolution then histological 
methods which give results that are neither the same size nor quality. 
Annese also comments that it would be highly beneficial to validate MRIs
 with studies of actual brain tissue. The study was carried out on one 
patient that had died at the age of 88, and as the patient had a 
pacemaker (which prevented the acquisition of MRI images while the 
patient was alive), all images of the brain were taken after death and 
compared to the histological study of the brain tissue using the 
author’s method.</p><p class="Para">To ensure that the histological 
images match the Big Volume found in typical MRI images, Annese takes 
stained histological slices of the brain at 20x magnification, with a 
resolution of around 0.4 <em class="EmphasisTypeItalic ">μ</em> m/pixel. This creates 334,500 × 266,200 images. In addition, multiple MRI images were taken of the brain <em class="EmphasisTypeItalic ">in situ</em>
 prior to the histological imaging. The author was able to show that 
MRIs (and DTIs in particular) do show tissue properties that can be 
correlated to stained tissue (histological images), making comparison 
possible. The author argues that histological comparison to MRIs can 
help validate MRIs, localize neuropathlogical phenomena that show as MRI
 abnormalities, and create the full connectivity map of the human brain.
 Data mining and analysis of both types of images in parallel could 
offer more gains than simply using MRI data alone, offering a much more 
powerful set of data.</p><p class="Para">As technology only recently 
could handle the endeavor of creating a full connectivity map of the 
brain this line of research is very new. The HCP is creating and 
releasing Big Volumes of data with various amounts of different MRIs and
 behavioral measurements data, allowing for more analysis and novel data
 mining to be executed. Along with the HCP, there is discussion and 
testing being conducted for comparing MRIs to histological data to help 
validate MRI data, to help create the connectivity map of the brain and 
offer more power to the datasets being created for novel data mining. 
The HCP could benefit from employing a comparison to histological image 
data.</p></section><section id="Sec12" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Using MRI data for clinical prediction</h3><p class="Para">This
 subsection will be covering two studies with the goal of answering 
clinical level questions. The first study uses both MRI data and a list 
of clinical features with the goal to find correlations between physical
 ailments to that of different locations of the brain. The second looks 
to take MRI data and determine the amount a patient has Alzheimer’s 
disease. Research using MRIs can be beneficial to clinical diagnosis and
 predictions giving physicians another option with which to make 
decisions.</p><p class="Para">Yoshida et al.
[<span class="CitationRef"><a href="#CR10" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">10</a></span>]
 propose a novel method combining patients’ clinical features with that 
of MRI image intensities consisting of millions of voxels (an element of
 volume representing a point on a grid in three dimensional space). The 
method the authors create is based on the algorithm of radial basis 
function-sparse partial least squares (RBF-sPLS) giving their method an 
advantage over similar methods granting the ability to select not only 
clinical characteristics but also determine effective brain regions. 
This is to say that by creating sparse, linear combinations of 
explanatory variables, the developed approach concurrently performs 
feature selection and dimensionality reduction (that is, creating new, 
more condensed features). Simultaneously doing both of these tasks is 
problematic for other techniques especially for vast amounts of data, 
but employing RBF-sPLS allows the authors to manage large amount of data
 efficiently. Yoshida et al. do a comparison of their RBF-sPLS to that 
of the original sPLS on a simulated dataset, and demonstrate that their 
technique outperforms the original in terms of sensitivity, specificity 
and c-index scores. The simulated dataset does appear to be the only 
testing of the RBF-sPLS and even with the results garnered being quite 
promising this method should be tested on real-world data so that their 
results can be considered clinically significant.</p><p class="Para">They
 do present an example of their prospective method using a dataset of 
102 chronic kidney patients with the goal of attempting to identify if 
any locations of the brain are correlated to patients with chronic 
kidney disease. For each of these 102 patients they gathered 73 clinical
 features and around 2.1 million voxels from the MRI data. Through 
experiments Yoshida et al. found that there is a strong correlation 
between clinical variables related to chronic kidney disease and the 
bilateral temporal lobe area of the brain. They also determined that the
 bilateral temporal lobes are closely related to aging and arterial 
stiffness, while the occipital lobes correspond to clinical markers for 
anemia. As this is only a preliminary study, further testing will be 
needed to confirm the accuracy of these results.</p><p class="Para">Estella et al.
[<span class="CitationRef"><a href="#CR11" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">11</a></span>]
 introduce a method with the goal of predicting to what degree a patient
 has Alzheimer’s disease with three levels of classification: completely
 healthy, Mild Cognitive Impairment (MCI), and already has Alzheimer’s. 
They gathered around 240GB of brain image data for 1200 patients stored 
by the Alzheimer’s Disease Neuroimaging Initiative (ADNI). There are a 
number of steps to their devised method, which include spatial 
normalization, extraction of features, feature selection and patient 
classification. After the feature extraction step, the authors found two
 different subgroups of features: morphological and mathematical where 
332 and 108 were respectively gathered. Examples of features from the 
morphological subgroup are Area Centroid, Major Axis Length, Whole 
Matter Volumes, etc. Examples of features from the mathematical subgroup
 are Mean, Cosine Transform Coefficients, Euclidean distance, etc. The 
authors also decided to have a third group that consists of features 
from both subgroups and deemed this group as “mixed”. For the feature 
selection step they used a method based on Mutual Information (MI) along
 with some influence from the minimal Redundancy-Maximal-Relevance 
criterion (mRMR). MI will assist in determining the dependence between 
two given variables while the mRMR looks to acquire applicable features 
correlated to the final prediction while simultaneously removing 
redundancies from the model.</p><p class="Para">Fuzzy Decision Tree 
(FDT) classifiers work best with Estella et al.’s feature extraction and
 feature selection combination, getting better results than other 
classifiers tested in terms of classifying efficiency (able to make 
reliable diagnosis on a minimal set of features). FDT is an extension to
 the traditional decisions with the additional ability to handle fuzzy 
data. The authors found that by using their method they were able to get
 a classifying efficiency of 0.94 when using 75% of the mixed feature 
subset and when only using 10% of the mixed features could still get a 
reliable efficiency of 0.75. Estella et al. conclude that using their 
method while using only a minimal number of both morphological and 
mathematical variables they can efficiently and reliably classify 
patients into three levels of Alzheimer using only MRI data. Even though
 this study compared various classification techniques there was only 
one feature selection technique tested; there could be other feature 
selection methods that could have worked better with the FDT or one of 
the other classification techniques tested.</p><p class="Para">The 
studies shown in this section using MRI data have shown that they can be
 useful in answering clinical questions as well as making clinical 
predictions. More research (on real world data) will be needed before 
the brain regions, determined by Yoshida et al.
[<span class="CitationRef"><a href="#CR10" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">10</a></span>] to have correlation with kidney disease, anemia and aging, can be determined as clinically significant. In Estella et al.
[<span class="CitationRef"><a href="#CR11" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">11</a></span>],
 various features extracted from MRIs were shown here to have the 
ability to classify patients into varying degrees of dementia. This 
process could eventually be improved upon with the goal of adding more 
classification levels for Alzheimer patients leading to such patients 
being detected both earlier and more efficiently. This is an interesting
 line of research that could be very beneficial if a physician could one
 day have methods that can look at MRIs of the brain and be able to 
determine whether a patient has kidney disease, more likely to have 
kidney disease, determining how far along a patient is down the road to 
dementia, etc. With the promising results shown in these studies, 
further research should extend these research efforts to see what other 
correlations between MRIs and other diseases can be discovered, giving 
physicians one more tool to diagnose and treat a patient earlier and 
more accurately.</p></section></div></section><section id="Sec13" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Using patient level data<span class="section-icon"></span></h2><div class="content"><p class="Para">All
 the studies presented in this section will cover data at the patient 
level and venture to answer clinical level questions including: 
Prediction of ICU readmission, prediction of patient mortality rate 
(after ICU discharge and 5 years) and making clinical predictions using 
data streams. Due to this use of streaming data, ICU data exemplifies 
the Velocity aspect of Big Data, as well as Variety, as clinical 
information can contain many different types of features. As with the 
molecular-level data, feature selection can help choose the most 
important features, which also helps in making quicker clinical 
decisions.</p><section id="Sec14" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Prediction of ICU readmission and mortality rate</h3><p class="Para">The
 focus of the research covered in this subsection will be on predicting 
Intensive Care Unit (ICU) readmission, mortality rate after ICU 
discharge as well as predicting a 5 year life expectancy rate. The five 
year life expectancy rate will be looking to see how likely a patient 
will survive within a 5 year period. This is a useful line of research 
in that it can potentially help physicians know what to look for in 
their patients, determine which patients should have their ICU stay 
extended, and better tell which patients should receive particular 
treatments.</p><p class="Para">The study done by Campbell et al.
[<span class="CitationRef"><a href="#CR12" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">12</a></span>]
 focuses on ICU patients that were discharged and expected to both live 
and not return too early afterwards. Their research used 4376 ICU (out 
of 6208 total) admissions from a database containing admissions from one
 ICU from January 1995 to January 2005 (a 10 year period). There was a 
total of 16 attributes chosen for each ICU admission, where among these 
attributes there were a few well-known scores used for the prediction of
 ICU readmission and death after discharge including: Acute Physiology 
and Chronic Health Evaluation II (APAHACE II) score
[<span class="CitationRef"><a href="#CR32" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">32</a></span>], Simplified Acute Physiology Score II (SAPS II)
[<span class="CitationRef"><a href="#CR33" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">33</a></span>], and the updated Therapeutic Intervention Scoring System (TISS)
[<span class="CitationRef"><a href="#CR34" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">34</a></span>]
 score. The APACHE II score is a popular and well tested score based 
severity of disease classification system (SoDCS), described in
[<span class="CitationRef"><a href="#CR32" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">32</a></span>]
 by Knaus et al., which uses a simple set of 12 physiological variables 
for prediction. The SAPS II score is another popular and well tested 
SoDCS using 17 total features, including age, 12 physiology features, 
admission type, and 3 disease type features, which is discussed further 
by Gall et al.
[<span class="CitationRef"><a href="#CR33" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">33</a></span>].
 The TISS score is a third popular and well tested SoDCS where 
originally 57 therapeutic intervention measurements were used but was 
updated where some features were added and some removed, while test 
results stayed the same. The updated TISS score is covered by Keene et 
al.
[<span class="CitationRef"><a href="#CR34" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">34</a></span>].</p><p class="Para">Campbell
 et al. decided upon three research directives for prediction: death 
after ICU discharge (but before hospital discharge), readmission to the 
ICU within 48 hours of ICU discharge (again, before hospital discharge),
 and readmission to the ICU at any point after ICU discharge (and before
 hospital discharge). It should be noted that each patient could 
potentially fall into more than one of these categories, but this is not
 an issue due to there being three separate binary models built. The 
importance these prediction models could potentially have would be to 
help physicians determine which of these patients fall into these three 
groups and most significantly why they fall into these groups. If 
physicians can know why then they can determine which patients need 
their ICU stay extended. The feature selection method decided upon was 
simple logistic regression for all three models to determine which of 
the 16 attributes had a strong correlation to each prediction (P ≤ 0.2).
 Multiple Logistic regression was chosen for building the prediction 
models. All three of their models were tested using the Hosmer and 
Lemeshow goodness-of-fit (HLgof) test
[<span class="CitationRef"><a href="#CR35" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">35</a></span>]
 and determined that the calibration was good and no more calibration 
was necessary. The HLgof is used to determine if logistic regression 
models have sufficient calibration (discussed by Hosmer et al.
[<span class="CitationRef"><a href="#CR35" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">35</a></span>]).</p><p class="Para">The
 Area Under the (ROC) Curve (AUC) was used in this study to determine 
the classification and discrimination performance of the three models. 
According to Bradley
[<span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">36</a></span>],
 AUC is the best criteria for measuring the classification performance 
of a binary classifier such as logistic regression. AUC is a fitting 
metric to use on ICU admission data because the positive-class instances
 are a small percentage (i.e. only 3.3% of the patients were readmitted 
to the ICU within 48 hours). For determining the quality of the three 
models (using the chosen set of features) the AUC results for each model
 are compared to results obtained from the APACHE II score for each 
prediction. The first model, predicting death after ICU discharge 
(before hospital discharge) had an AUC value of 0.74 compared to AUC 
garnered from APACHE II of 0.69. The second model, for predicting 
readmission (before hospital discharge) obtained an AUC of 0.67 while 
APACHE II received an AUC of 0.63. The third model for predicting 
readmission within 48 hours of ICU discharge acquired an AUC value of 
0.62 while APACHE II earned an AUC of 0.59.</p><p class="Para">The three
 models with the chosen set of features only achieved minimal 
improvement over APACHE II alone for the prediction of ICU readmission 
and mortality rate for ICU patients. Campbell et al. note that this 
minimal improvement could be because the APACHE II score already uses 
physiological variables, and it is these variables which are generally 
most useful for predicting ICU readmission and the mortality rate after 
ICU discharge. However, one non-physiological variable was shown to be 
highly correlated to both ICU readmission and death predictions: 
increasing age.</p><p class="Para">Only about 23.3% of patients used 
fall into one of these three model’s positive class and should not have 
been released from the ICU where if their ICU stay was extended, maybe 
some of these patients could have been saved. One item to note is that 
Campbell et al. only used one feature selection method as well as only 
one prediction model for their research. There is a possibility if other
 methods were tested there could have been models built that were able 
to outperform APACHE II. Even though the data was gathered from a 
database containing a 10 year’s worth of ICU patients, the data was 
still only from one ICU, and for the purpose of validation a larger 
variation (collecting from various ICUs) would have been beneficial.</p><p class="Para">Ouanes et al.
[<span class="CitationRef"><a href="#CR14" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">14</a></span>]
 conducted their research with the goal of predicting whether a patient 
would die or return to the ICU within the first week after ICU 
discharge. Research was performed on 3462 patients (out of 5014) 
admitted to an ICU for a minimum of 24 hours, gathered from 4 different 
ICUs from the Outcomerea database. As Campbell et al. found in their 
data, the positive class is very small compared to the overall 
population with only about 3% (where 0.8% died and 2.1% were readmitted 
within 7 days). The feature selection method chosen was univariate 
analysis selecting variables with (P &lt; 0.2) to add to the final 
model, and used the Akaike Information Criterion (AIC)
[<span class="CitationRef"><a href="#CR37" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">37</a></span>]
 to identify the best model. AIC is a metric used to verify the overall 
quality of statistical models. The model created was then subjugated to a
 few validation and verification steps for testing (clinical relevancy, 
variable inter-correlation and co-linearity between variables) in order 
to end up with their final set of 6 variables from the original 41. The 
six variables chosen were age, SAPS II, the need for a central venous 
catheter, SIRS score during ICU stay, SOFA score, and discharge at 
night.</p><p class="Para">These variables were used to make the final 
prediction model using multivariate logistic regression, which will be 
used to develop their Minimizing ICU Readmission (MIR) score. The MIR 
score will be a quantitative measurement for determining whether a 
patient should be discharged from an ICU or not. The predictive results 
of MIR are compared to the results garnered from both SAPS II and 
Stability and Workload Index for Transfer (SWIFT)
[<span class="CitationRef"><a href="#CR38" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">38</a></span>],
 another SoDCS using a small set of commonly available variables. 
Through MIR, Ouanes et al. were able to achieve good results with good 
calibration decided by the HLgof test and an AUC of 0.74 at a 95% 
confidence interval. The result SAPS II received for AUC was 0.64 and 
SWIFT getting an AUC score if 0.61, which shows that MIR performed 
considerably better. In this research only one feature selection method 
as well as only one classification method is used. The MIR score 
possibly could have yielded better results if more than one feature 
selection technique was used to determine which of the 41 variables 
would make the best model. In conjunction, MIR could possibly be further
 benefitted by testing out a number of other classifiers to develop the 
final model with the highest predictive power for this line of research.
 To fully test these results, one more comparison that should be made is
 their MIR score to that of either APACHE II or APACHE III scores, as 
shown in Campbell et al.
[<span class="CitationRef"><a href="#CR12" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">12</a></span>] and Failho et al.
[<span class="CitationRef"><a href="#CR13" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">13</a></span>].</p><p class="Para">Failho et al.
[<span class="CitationRef"><a href="#CR13" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">13</a></span>]
 also seek to predict patients that will be readmitted to the ICU, but 
with the goal of only using a small amount of physiological variables, 
following the findings of Campbell et al.
[<span class="CitationRef"><a href="#CR12" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">12</a></span>].
 The prediction that this research is focusing on making is determining 
if patients will be readmitted within 3 days after discharge. The 
dataset they gathered was from the MIMIC II database
[<span class="CitationRef"><a href="#CR39" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">39</a></span>]
 choosing only the patients over the age of 15 having an ICU stay of 
more than 24 hours giving a sample of 19,075 adults that were admitted 
to one of four ICUs. These 19,075 patients were further reduced by 
considering only patients for whom the researchers have all the 
variables available, leaving 3,034 patients. Further preprocessing 
reduced this to 1,267 patients. Finally, out of these 1,267 patients 
only 1,028 survived, giving a final dataset with 1,028 instances (and 13
 members of the positive (readmittance) class).</p><p class="Para">Failho
 et al. decided upon two possible methods for feature selection: 
Sequential Forward Selection (SFS), or bottom-up approach, and 
Sequential Backwards Elimination (SBE), or top-down feature selection 
where both are discussed in
[<span class="CitationRef"><a href="#CR40" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">40</a></span>].
 SFS works on the original pool of features and starts with one feature 
then during each iteration one more feature will be added and the 
iterations will be stopped when the current model is deemed the best 
possible. SBE takes a different approach starting with the whole set of 
the original features and removing one feature at each iteration until 
the model is deemed the best possible. Failho et al. compared both of 
these methods in combination with their classifier, and found SFS to 
have better results than SBE in terms of AUC. Thus, it was chosen as 
their feature selection method. Using SFS, 6 out of the original 24 
physiological variables were chosen: mean heart rate, mean temperature, 
mean platelets, mean blood pressure, mean SpO2, and mean lactic acid.</p><p class="Para">The classifier method used in this study was fuzzy modeling with sequential forward selection
[<span class="CitationRef"><a href="#CR40" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">40</a></span>, <span class="CitationRef"><a href="#CR41" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">41</a></span>],
 specifically Takagi-Sugeno (TS) fuzzy modeling. Fuzzy models use 
linguistic interpretations to formulate rules and logical connectives in
 order to make connections between features and the final prediction. 
With the use of linguistic interpretation, fuzzy modeling is a good 
choice for this line of research as clinical data needs to be 
interpreted, as a physician would do in a clinical setting. The only 
worry would be that with the rule-based side of the models there could 
possibly be too much rigidity causing the absence of physician 
discernment in the final model.</p><p class="Para">Failho et al. 
compared the results they got from their set of 6 physiological 
variables (determined by SFS) in conjunction with TS to that of the sets
 determined by APACHE II and APACHE III
[<span class="CitationRef"><a href="#CR42" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">42</a></span>]
 scores (also in conjunction with TS). APACHE III is another SoDCS, and 
was created with the goal of improving some of the problems in APACHE 
III as discussed in
[<span class="CitationRef"><a href="#CR42" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">42</a></span>]
 by Knaus et al. The results show a significant advantage in favor of 
Failho et al.’s. set securing an AUC score of 0.72 ± 0.04 while APACHE 
II and APACHE scored an AUC of 0.62 ± 0.03 and 0.64 ± 0.04 respectively.
 The SFS set also scored better in terms of specificity, sensitivity, 
and accuracy. This result shows that good prediction performance can be 
reached by using a small set of physiological variables. Even with the 
promising results there is more that could have been done in this 
research, one being that more variables, either physiological or not, 
could have been added to the original pool to see if results could have 
been improved. It might also have been beneficial to see if other 
feature selection techniques (tree-based or otherwise) may have improved
 upon the results achieved by SFS. Failho et al. do mention that fuzzy 
modeling has been shown to work comparably well to other classification 
methods for medical data yet there still could be benefit to this 
research if other classifier methods were tested to see if TS does yield
 the best results.</p><p class="Para">The research objective of Mathias et al.
[<span class="CitationRef"><a href="#CR15" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">15</a></span>]
 is along a slightly different line where instead of trying to predict 
ICU readmission they look to predict a 5 year mortality rate through the
 construction of an Ensemble Index (EI). They used a group of 7463 
patients taken from an Electronic Health Record (EHR) along with 980 
attributes for each patient. There were two requirements for the 
patients to be used in this study: must be over the age of 50 (due to 
increasing age being a huge prediction factor for this line of study) 
and had at least 1 hospital visit within the year 2003. Due to the large
 amount of attributes in the original set of variables the feature 
selection method chosen was Correlation Feature Selection (CFS)
[<span class="CitationRef"><a href="#CR43" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">43</a></span>]
 along with greedy stepwise search which will be used to create their 
EI. The CFS method finds variables that are both strongly correlated to 
the final prediction and that are weakly correlated between them. The 
CFS method with greedy stepwise search found a subset of 52 features 
which was broken down further by manual reduction followed by another 
round of CFS bringing the subset down to 23. This subset was then 
populated by one variable (gender) giving the final EI subset of 24 
variables. The top 6 attributes in the EI (ranked by information gain) 
are age, comorbidity count, amount of hospitalization a year prior to 
admission, high blood urea nitrogen levels, low calcium, and mean 
albumin.</p><p class="Para">Rotation Forest Ensembling (RFE)
[<span class="CitationRef"><a href="#CR44" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">44</a></span>] with Alternating Decision Tree (ADT)
[<span class="CitationRef"><a href="#CR45" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">45</a></span>]
 was used to create their predictive model and was evaluated with 
tenfold cross-validation. Mathias et al. tested this technique with many
 other methods and found this technique to perform better (the reason it
 was used). The RFE algorithm is an ensemble of decision trees creating 
variation by assigning each tree a subset of features randomly chosen 
where Principle Component Analysis (PCA) is applied to each subset 
before each tree model is built. The ADT is a decision tree which 
instead of having a single class prediction located in its base leaf 
nodes, has a “probability of class membership” prior to each terminal 
node, and the sum of all these values are along an instance’s whole path
 in order to predict its class value. RFE and ADT is a good combination 
as RFE brings accuracy and diversity to the model, and ADT allows for 
more information to be gained about an instance as it goes along the 
tree.</p><p class="Para">The Ensemble Index was able to achieve quite 
good results scoring better in recall, precision c-statistic, etc. than 
both the modified Walter Life Expectancy Index (WLEI)
[<span class="CitationRef"><a href="#CR46" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">46</a></span>] and the modified Charlson Comorbidity Index (CCI)
[<span class="CitationRef"><a href="#CR47" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">47</a></span>].
 The modified WLE and CCI are two well-tested and better-known life 
expectancy indices that are used for prediction in similar research. 
Even though these good results were achieved for the EI, there was only 
one feature selection process used where if more techniques were used a 
better subset could have been created, which is possible with there 
being 980 attributes in the original set of features. Again more 
variation of data could be added to this research as all the data was 
gathered from one source.</p><p class="Para">The studies shown in this 
section have the potential to improve clinical discharge procedure, 
determining which patients should be released from the ICU and which 
patients should receive a particular line of treatment. The goal of 
these studies is to find which attributes are the most correlated to why
 patients return to ICUs early or do not survive after discharge. 
Looking at the research efforts of Campbell et al.
[<span class="CitationRef"><a href="#CR12" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">12</a></span>], Failho et al.
[<span class="CitationRef"><a href="#CR13" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">13</a></span>], and Ouanes et al.
[<span class="CitationRef"><a href="#CR14" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">14</a></span>]
 that covered prediction of ICU readmission and death rate after 
discharge, the top variables of why are age, APACHE scores, various 
physiological variables (e.g. heart rate), amount of organ dysfunction, 
as well as a few others. If physicians can better predict which of their
 patients will return to the ICU or not survive then they would know 
which patients to keep in the ICU longer and to give more focused care 
potentially saving, if not, at least prolonging a life. According to 
Ouanes et al., between day one and day seven after discharge the 
readmission rate and death rate go down drastically, meaning that 
keeping a patient just a little longer could be beneficial, but could 
also take away an ICU bed from another patient that needs intensive 
care. These are the reasons that studies attempting to figure out why 
patients return early or die soon after ICU discharge are quite 
important as lives can potentially be saved. The target of this research
 of course is on those patients that have preventable death as not all 
death will be preventable.</p><p class="Para">The variables that were shown to be most telling for the 5 year survival rate (Mathias et al.
[<span class="CitationRef"><a href="#CR15" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">15</a></span>])
 happened to be similar with age and physiological variables being at 
the top of the list. One benefit of this research is by looking well 
into a patient’s future can help physicians advise their patients better
 as far a treatment options. An example of this would be a physician 
could advise a patient whether or not to go through a particular rough 
line of treatment when they may not live long enough to reap the 
benefits of such treatment. This research is especially important for 
patients with increasing age as the older a patient is the less likely a
 harsh treatment would be beneficial.</p></section><section id="Sec15" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Real-time predictions using data streams</h3><p class="Para">The
 studies covered in this section are also on data gathered from the 
patient level and again have the intention of answering clinical level 
questions. Instead of predicting the patient’s condition in the future 
(i.e. ICU readmission or 5 year survival), the research here will be 
using data streams in order to predict patient’s conditions in 
real-time. Data streams are never ending torrents of data that requires 
continuous analysis giving the possibility for real-time results (a 
feature not available when using static data sets). This section will 
sample two different categories of data steam studies: making prognosis 
and diagnosis predictions for patients, and detecting if a new born is 
experiencing a cardiorespiratory spell both in a real time. The 
researchers here are attempting to develop methods that use these 
constant streams of data and make predictions in a continuous manner 
while keeping satisfactory accuracy and precision.</p><p class="Para">Zhang et al.
[<span class="CitationRef"><a href="#CR17" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">17</a></span>]
 develop a clinical support system using data stream mining with the 
goal of analyzing patient data in order to make real-time prognosis and 
diagnosis. In order to handle the continuous stream of data an algorithm
 that can handle high-throughput data will be necessary leading the 
authors to choose Very Fast Decision Tree (VFDT). The VFDT algorithm is 
quite efficient as it was built to handle thousands of instances per 
second using basic hardware (discussed by Domingos et al.
[<span class="CitationRef"><a href="#CR48" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">48</a></span>]).
 They discuss that VFDT has many advantages over other methods (e.g., 
rule based, neural networks, other decision trees, Bayesian networks) 
such as VFDT can make prediction both diagnostically and prognostically,
 can handle a changing non-static dataset, not using rigid rules (can be
 difficult for experts to put their knowledge into rules).</p><p class="Para">VFDT
 alone, though, is not able to give future predictions of a patient’s 
status only the current status; therefore, Zhang et al. decided to 
modify VFDT. For the modified VFDT, one or more pointer(s) were added to
 each of the terminating leaf nodes, where each base node corresponds to
 a distinct medical condition and each pointer corresponds to one 
medical records of a previous patient. To connect each stored medical 
record to its corresponding pointer, the authors created a mapping table
 so when the VFDT runs through and ends up on a base node the map will 
connect the leaf to its pointer(s) (corresponding medical records). 
These medical records, through Natural Language Processing (sentence and
 semantic similarity
[<span class="CitationRef"><a href="#CR49" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">49</a></span>, <span class="CitationRef"><a href="#CR50" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">50</a></span>]),
 will then be used to make a prediction about the patient’s future and 
give physicians the ability to better treat and advise their patients 
based on previous similar situations. The VFDT and the mapping table are
 updated as necessary (i.e. when a physician makes a new diagnosis or 
when a new medical record is added to the map).</p><p class="Para">Zhang et al. compare their method to that of IBM’s similar data stream mining technique covered by Sun et al.
[<span class="CitationRef"><a href="#CR51" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">51</a></span>].
 To test their method, IBM used 1500 ICU patients from the MIMIC II 
database along with various physiological waveforms (taken from an 
assortment of medical devices) and clinical data on each patient. IBM’s 
method consists of three main parts: 1.) physiological stream 
processing, 2.) offline analysis, and 3.) online analysis. For the steam
 processing part, a correlation base technique was chosen as such 
techniques are able to correlate well among sensors and are able to 
efficiently handle missing data (estimating missing values by way of 
linear regression models using other sensors during that period of 
time). Better results could possibly have been attained if techniques 
other than linear regression models were used to estimate missing 
values. A correlation based technique was not the only technique tested,
 they also tried a window based technique which estimates missing values
 for a sensor by using an averaged value during a small window of time 
from that sensor and imputing that value for the missing time. The 
correlation based technique found better results and therefore was used 
in the final method.</p><p class="Para">During offline analysis, Sun et 
al. use a method they created called Locally Supervised Metric Learning 
(LSML), which learns an adjustable distance metric by using knowledge 
from the current domain (in this study, clinical knowledge). The last 
step of online analysis takes place when a new patient is ready for 
prediction of prognosis where the system will find the set of similar 
cases by way of temporal alignment, and this is followed by applying a 
regression model to account for the uniqueness of patients. Sun et al. 
ran a comparison of their LSML to that of PCA and Linear Discriminant 
Analysis (LDA) in terms of both precision and accuracy with LSML scoring
 considerably better in both.</p><p class="Para">Mentioned by Zhang et 
al., their system will use fewer computer resources to run compared to 
IBMs as offline analysis will not needed. This comes down to a 
comparison of the complexity between LSML and VFDT, where LSML is quite 
complex and does not allow the real-time nature that can be offered by 
VDFT (the most complex calculation of Zhang et al.’s. method). Zhang et 
al.’s. method is not tested on real world data and would need to be 
before its usefulness can be determined and can be legitimately compared
 to IBM’s method. The real-time nature does offer the ability for making
 quick prognosis for patients, and if the results can be similar to the 
results found in IBM’s method, then predictions could not just be made 
quickly but with good accuracy and precision.</p><p class="Para">Thommandram et al.
[<span class="CitationRef"><a href="#CR18" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">18</a></span>]
 use data streams with a different goal attempting to detect and 
eventually classify neonatal cardiorespiratory spells (a condition that 
can be greatly helped by being detected and classified in real time). A 
cardiorespiratory spell is classified as some combination of a pause in 
breathing, drop in blood oxygen saturation, and a decrease in heart 
rate. The name of their system is called Artemis and is designed to use a
 steady stream of physiological data from the new born patient and both 
detect and ascertain which type of cardiorespiratory spell the patient 
is experiencing all in real-time. The real-time manner could potentially
 save the lives of these infants giving physicians more time to fix what
 is wrong as the need for human diagnosis will be less. The actual 
stream processing part of Artemis is handled by the middleware system 
developed by IBM called InfoSphere Streams
[<span class="CitationRef"><a href="#CR16" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">16</a></span>],
 built to handle multiple high-throughput data streams. Middleware is 
software that works to connect two programs that are otherwise not 
connected. Three different data streams (from three different sensors) 
are used, which correspond to the three different conditions for a 
cardiorespiratory spell: a respiratory impedance wave, a decrease in 
blood oxygen saturation, and a decrease in heart rate.</p><p class="Para">The
 authors wanted to develop a system that will improve upon current 
machines used today that use the absolute change method (i.e. if a 
machine detects heart rate under or over a cutoff, then notify 
physicians). Artemis will use a relative change method, where instead of
 a cutoff; there will be a sliding baseline that will be continuously 
updated in real-time as the patients “normal” readings change over time.
 The sliding baseline method can give a more reliable reading as it 
adapts to the unique reading of each patient as well as allow for more 
accurate spell detection. For classification, the reading from the three
 streams will be analyzed through a hierarchal rule based temporal model
 to determine which of the many cardiorespiratory spells the infant is 
experiencing.</p><p class="Para">The detection part of Thommandram et 
al.’s. system was tested on one patient in a Neonatal ICU during a 24 
hour period where the sliding baseline method was found to alert 
physicians as often as found in the cutoff method for both heart rate 
and SpO2 readings. From the results for this one tested patient, they 
showed that their sliding baseline method could achieve clinically 
significant results for heart rate detection, with a specificity of 
98.9% and a sensitivity of 100%. They did not test their classification 
method as mentioned by the authors as their future work. The detection 
part of this method will need further testing on many more patients 
before this method can be considered as clinically acceptable. When the 
classification part is tested it could be beneficial if they were to 
compare other methods to their hierarchal rule based temporal model.</p><p class="Para">Data
 stream mining presented here has shown the potential to be beneficial 
for clinical practice as it can be extended to be used in real time by 
use of efficient algorithms and methods (that are not previously used in
 the clinic). By using these data stream diagnosis for prognosis and 
spell detection, physicians could make faster and more accurate 
decisions and start solving the problem without spending as much time 
developing a plan. This line of research is fairly young and more 
studies will be needed through the development and testing of various 
new methods for using data stream mining for medical data with the goal 
of outputting results in real-time.</p></section></div></section><section id="Sec16" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Using population level data – Social media<span class="section-icon"></span></h2><div class="content"><p class="Para">All
 the studies covered in this section will use data at the population 
level, specifically social media data (data that can be from anyone) and
 venture to answer both clinical questions and epidemic-scale questions.
 Data in Health Informatics is “traditionally” gathered from the 
doctors, clinics, hospitals and such, but recently people all around the
 world are starting to document health information all over the 
internet. This data could be from Twitter, internet query data (e.g. 
Google search data), message boards, or anywhere else people put 
information on the internet. This form of Big Data, which brings 
additional challenges such as text mining and handling potentially 
malicious noise, could possibly lead to finding many new breakthroughs 
in the field of medicine.</p><p class="Para">The first line of research 
presented is determining whether message board data can be useful to 
help patients find information on a given ailment. The second line of 
study is testing if using search query data or Twitter post data can 
effectively track an epidemic across a given population (in real-time). 
As previously mentioned above, the challenge with social media data is 
that although it is clearly High Volume, Velocity, and Variety, it could
 have both low Veracity and Value (data coming in could be unreliable, 
as discussed by Hay et al.
[<span class="CitationRef"><a href="#CR52" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">52</a></span>]). However, through analyzing and mining the data, the useful parts can be extracted and used to gain medical insight.</p><p class="Para">There
 is much that can be learned by employing research on social media data 
including but not limited to: spatiotemporal information of disease 
outbreaks, real-time tracking of a harmful and infectious diseases, 
increasing the knowledge of global distribution for various diseases, 
and creating an extremely accessible way of letting people get 
information about any medical questions they might have.</p><section id="Sec17" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Using message board data to help patients</h3><p class="Para">Research
 in this subsection concentrated on determining if message board entries
 could possibly be a useful source of data for helping people find 
health information that is beneficial in appropriating reliable medical 
knowledge (answering clinical questions). Even though this line of 
research does not appear to be all that popular, it could offer keys to 
unlocking boundless reliable health information (as 59% of US adults use
 look for health information on the internet
[<span class="CitationRef"><a href="#CR19" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">19</a></span>]), where one in five use social media).</p><p class="Para">Ashish et al.
[<span class="CitationRef"><a href="#CR19" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">19</a></span>]
 created a platform called Smart Health Informatics Program (SHIP) with 
the goal of helping patients connect to the medical experiences of other
 patients posted throughout the internet via message boards (i.e. 
forums, blogs etc.). This study used a pool of 50,000 discussions 
including over 400,000 posts from four different message board websites 
(inspire.com, medhelp.com, and 2 others). SHIP uses a pipeline structure
 taking in all the message board entries (discussions and posts) from 
these four websites and go through various steps to find the entries 
with useful medical information and store them in a database for patient
 retrieval.</p><p class="Para">The first step of SHIP’s pipeline is 
Elementary Extraction, which will execute some basic text processing for
 each entry: parsing through the HTML and extracting information on each
 entry (number of replies, when discussion was last updated, etc.) and 
giving a unique ID to each discussion and post. This is followed by the 
Entity Extraction step looking to determine which of the entries have 
medical significance related to health (such as treatments, 
side-effects, hospitals, drugs, etc.). Ashish et al. decided to use the 
XAR system
[<span class="CitationRef"><a href="#CR53" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">53</a></span>]
 (an information extraction system of free text), which they expanded by
 incorporating ontologies from UMLS (Unified Medical Language System)
[<span class="CitationRef"><a href="#CR54" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">54</a></span>]
 as well as other vocabularies and ontologies. Then SHIP enters 
Expression Distillation which looks at each post and determines whether 
or not the post includes any number of many expression of interests 
broken into 5 categories: 1.) personal experience, 2.) advice, 3.) 
information, 4.) support, 5.) outcome. This step requires that 5 
classifiers be built for each expression of interest and they decided 
upon the J48 decision tree algorithm using the WEKA tool
[<span class="CitationRef"><a href="#CR55" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">55</a></span>].
 The next step is Aggregation using the data from the previous steps 
(done at the post level) and aggregates them at the discussion level. 
After the entries have gone through this process the numerous facts and 
expressions (over 20 million) found in each entry are stored in a 
database.</p><p class="Para">For retrieval from the database the authors extended Lucene
[<span class="CitationRef"><a href="#CR56" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">56</a></span>],
 an open-source Java-based text search engine library, by adding 
optimizations for indexing, dynamic score boosting, and local caching. 
They tested their system on a test case of a patient experiencing severe
 cough since starting Tarceva (a type of chemotherapy). This was chosen 
as a test case due to coughing not being listed as a possible 
side-effect on the Tarceva website
[<span class="CitationRef"><a href="#CR57" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">57</a></span>]
 or on another website depicting a clinical trial for this drug. Through
 the authors system (website) a search of Tarceva shows that cough is 
actually the third most common side-effect. This system could also find 
advice (through search filters) on how to help with the cough due to 
Tarceva. As this system has been shown to be beneficial for patients 
there could also be benefits to physicians as they may be looking to 
clinical trials for answers they could use this system (employing 
message board data) to find other cases similar to their patients and 
find information helping to make diagnosis, treatment options, etc. The 
authors tested extraction accuracies for each of the 5 expression of 
interest categories in terms of precision and accuracy: personal 
experience (0.87 and 0.82), advise (0.91 and 0.62), Information (0.93 
and 0.91), support (0.89 and 0.90), and outcome (0.80 and 0.58).</p><p class="Para">There
 will need to be more testing with more test cases to see if the data 
they are presenting through search is correct rather than just 
speculation as this data can be posted by anyone and even if it is 
predominantly posted by medical professionals, this does not 
automatically make the information correct. Martin
[<span class="CitationRef"><a href="#CR2" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2</a></span>]
 mentions that traditional health data should be used in the development
 of SHIP. This would help determine whether the data they gathered has 
medical importance or not, rather than relying on testing SHIP after it 
is constructed. From a data mining stand point the step of expression 
distillation could have been improved by using more classifiers other 
than J48 with which they could have improved the results they got for 
extraction accuracies.</p><p class="Para">Rolia et al.
[<span class="CitationRef"><a href="#CR20" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">20</a></span>]
 devise a new system to use social health forums to help patients learn 
about their condition from posts by other patients with similar 
conditions. Their method consists of three steps: 1.) determining the 
patient’s current medical condition from the personal health record 
(PHR), 2.) the system will ascertain which other users have a similar 
condition, 3.) a metric will be implemented evaluating and ranking the 
forum topics to determine the most relevant to present to the user. They
 describe their system implementation for a test case of type II 
diabetes mellitus (DM II). Due to privacy issues it was decided to use 
synthesized electronic medical health records (EMR) and PHRs with help 
from a medical professional. Also created is an aggregated clinical 
pathway that a patient can follow from the moment the patient is 
diagnosed onward where the path is determined by clinical factors (e.g. 
glucose levels) depicting a sequence of clinical events and was based on
 information found in medical research
[<span class="CitationRef"><a href="#CR58" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">58</a></span>].
 Using the aggregated clinical pathway they created 1000 synthetic 
patients (including their path and a few clinical variables) akin to 
ones that could be found in typically PHRs and EMRs as well as manually 
selecting 100 relevant posts from the websites: www.diabetesforum.com 
and www.diabetesdaily.com.</p><p class="Para">The system starts by 
creating and assigning the states to the synthesized patients and for DM
 II they decided upon 6 states (also described in medical research
[<span class="CitationRef"><a href="#CR58" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">58</a></span>]). Assigning the patients into these six classes is done through rules (created by using NICE Pathways
[<span class="CitationRef"><a href="#CR59" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">59</a></span>]
 and a medical expert who worked with Rolia et al.) guiding the patient 
along the pathway. The next step consists of appointing weights to forum
 topics (as for this example they only used the topic and not the 
content within). A medical expert will give the original set of weights 
to topics (for each of the six states) as percentages of likelihood that
 a patient with a given state would be benefitted by that topic, which 
will give the initial population model. This system will also offer the 
new user the ability to determine how relevant they want posts to be 
compared to their current state, which is implemented by calculating the
 cosine similarity between each state using the weights determined by 
each of the 100 posts. Now the system moves onto ranking the forum 
topics by correlation to the current user’s state. The authors, in order
 to rank the topics used a metric deemed adjusted weights using the 
similarity between classes, patient preference (0 through 1), and the 
weights between post and clinical state; the higher the adjusted weight 
the higher the rank. The best feature about this system is its ability 
to learn as user feedback can adjust the weights of a post has to the 
given level of the current user (assuming using a similar method used by
 the expert). The user will also be able to change their current state 
as well as modify their patient preference.</p><p class="Para">The 
authors did build a prototype of this system and found that their system
 did behave as it was designed. One issue is that this system was not 
tested nor appears to be built with real data. A system such as this one
 is difficult to test outside letting real patients test this system in 
numerous field tests. As the authors mentioned they could possibly 
improve upon the cosine similarity by testing other such methods for 
defining weights. Another note, is that instead of a medical expert 
deciding the initial weights for posts through rules they could use a 
technique that would need to be tested and could then be approved by an 
expert, which could help with efficiency as well as the quality of the 
initial weights. This efficiency could help production of their classes 
as there are numerous diseases and ailments out there, and this research
 only discusses the construction for one.</p><p class="Para">Social media and the internet are becoming more and more popular for looking up and sharing medical data as mentioned in both
[<span class="CitationRef"><a href="#CR19" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">19</a></span>] and
[<span class="CitationRef"><a href="#CR20" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">20</a></span>].
 The results found in these studies do show that there is a possibility 
that message board data could be used, but there is no real field 
testing to show that it works in the real world. If the research in 
these studies are found to be useful they could even be extended to help
 advise physician’s diagnosis and treatments of their patients as other 
physicians and other patients with similar experiences to their current 
patients are located in these message board discussions.</p></section><section id="Sec18" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Using search query data to track epidemics</h3><p class="Para">The
 focus of the studies covered in this subsection will be discussing 
research done using search query data gathered from two popular search 
engines: Google (google.com) and Baidu (baidu.com) with the goal of 
predicting whether such data can be used to predict the occurrence and 
movement of Epidemics in a given population (all in real-time). The 
Center of Disease Control and Prevention (CDC) produces their results 
for influenza-like illness (ILI) epidemics, but there is generally a one
 to two week delay for this information. The research here is attempting
 to use search query data to get ILI epidemic information out to the 
public quicker than by the traditional method of the CDC reports. This 
research can help physicians and hospital to know both when and where an
 Epidemic is happening with real-time updates allowing them to act 
quicker in stopping the spread of the disease as well as help the 
patients already infected.</p><p class="Para">Ginsburg et al.
[<span class="CitationRef"><a href="#CR21" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">21</a></span>]
 developed an automated method that can analyze a Big Volume of search 
queries from Google with the goal of tracking ILI within a given 
population. They conducted their research on Google search queries taken
 from historical logs during a 5 year period (between 2003 and 2008) 
using 50 million of the most popular searches as well as using data from
 the CDC historical data. These queries were taken directly, without 
combination, spelling correction, translation, or any other 
modification. They built and trained their model based on data from the 
years 2003-2007 and the validation was done on data from 2008. As a 
note, the CDC splits the US into 9 regions and this study looked to make
 predictions also using these regions of separation.</p><p class="Para">The
 model the authors built looks to find the probability that a patient 
visiting a physician is related to an ILI for a particular region using a
 single explanatory variable: the probability that a given search query 
is related to an ILI within the same region. To accomplish this the 
authors fit a linear model using both the log-odds for ILI physician 
visits and ILI related search queries giving: <em class="EmphasisTypeItalic ">l</em> <em class="EmphasisTypeItalic ">o</em> <em class="EmphasisTypeItalic ">g</em> <em class="EmphasisTypeItalic ">i</em> <em class="EmphasisTypeItalic ">t</em>(<em class="EmphasisTypeItalic ">I</em>(<em class="EmphasisTypeItalic ">t</em>))=<em class="EmphasisTypeItalic ">α</em>×<em class="EmphasisTypeItalic ">l</em> <em class="EmphasisTypeItalic ">o</em> <em class="EmphasisTypeItalic ">g</em> <em class="EmphasisTypeItalic ">i</em> <em class="EmphasisTypeItalic ">t</em>(<em class="EmphasisTypeItalic ">Q</em>(<em class="EmphasisTypeItalic ">t</em>))+<em class="EmphasisTypeItalic ">ε</em>, where <em class="EmphasisTypeItalic ">I</em>(<em class="EmphasisTypeItalic ">t</em>) is the percentage of ILI physician visits, <em class="EmphasisTypeItalic ">α</em> is a coefficient, <em class="EmphasisTypeItalic ">Q</em>(<em class="EmphasisTypeItalic ">t</em>) stands for the fraction of queries related to ILI at time <em class="EmphasisTypeItalic ">t</em>, <em class="EmphasisTypeItalic ">ε</em> stands for the error in the formula (and <em class="EmphasisTypeItalic ">l</em> <em class="EmphasisTypeItalic ">o</em> <em class="EmphasisTypeItalic ">g</em> <em class="EmphasisTypeItalic ">i</em> <em class="EmphasisTypeItalic ">t</em>(<em class="EmphasisTypeItalic ">p</em>)=<em class="EmphasisTypeItalic ">l</em> <em class="EmphasisTypeItalic ">n</em>(<em class="EmphasisTypeItalic ">p</em>/1-<em class="EmphasisTypeItalic ">p</em>)).</p><p class="Para">The single explanatory variable <em class="EmphasisTypeItalic ">Q</em>(<em class="EmphasisTypeItalic ">t</em>)
 is determined through an automated technique that does not need any 
prior knowledge of influenza. The authors tested each of the 50 million 
stored queries alone as <em class="EmphasisTypeItalic ">Q</em>(<em class="EmphasisTypeItalic ">t</em>)
 to see which queries fit best with the CDC ILI visit percentage for 
each region (presumably univariate analysis). The top 45 search queries,
 sorted by Z-transformed correlation throughout the nine regions, were 
chosen to belong to <em class="EmphasisTypeItalic ">Q</em>(<em class="EmphasisTypeItalic ">t</em>)
 as the top 45 scored the best after they tested (through 
cross-validation) the top 1 search query through the top 100 search 
queries. Examination of these top queries showed many connections to 
influenza symptoms, complications, and remedies, consistent with 
searches made by an individual affected by influenza. The model was 
trained using weekly ILI percentages from 2003 through part of 2007 for 
all nine regions jointly to end up with a coefficient <em class="EmphasisTypeItalic ">α</em>
 that is region independent. Ginsburg et al. were able to obtain good 
fit compared to that of the reported CDC ILI percentages scoring a mean 
correlation of 0.90 throughout all nine regions.</p><p class="Para">Validation
 was carried out on the data gathered for part of 2007 through 2008 on 
42 points per region and was able to achieve a mean correlation of 0.97 
compared to the reported CDC ILI percentages. This correlation is quite 
high and through the authors’ method was able to be reported 1 to 2 
weeks prior to the CDC reports; thus, showing that search query data can
 be used to determine an ILI epidemic in a more real-time manner. This 
study only used one feature selection method narrowing down the 50 
million most popular search queries down to 45; the results achieved 
maybe could have been improved upon if other techniques were used to 
determine an optimal set other than the chosen 45.</p><p class="Para">Yuan et al.
[<span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">5</a></span>]
 developed a similar system using search query data, but this study uses
 search queries gathered from Baidu (baidu.com) with the goal of 
tracking ILI epidemics across China. The author gathered their data from
 Baidu’s database (http://index.baidu.com/) which stores the online 
search query since June 2006. For this study they only gathered data 
from March 2009 to August 2012, which was during the H1N1 epidemic and 
compare their results to that of China’s Ministry of Health (MOH). 
Similar to the United State’s CDC the MOH also releases their data with a
 1 to 2 week delay. Baidu releases their search query data on a daily 
basis allowing for methods exploiting this data the ability to give 
answers in near real-time.</p><p class="Para">Yuan et al.’s system is 
split into four main parts: 1.) choosing keywords, 2.) filtering these 
keywords, 3.) defining weights and composite search index, and 4.) 
fitting the regression model with the keyword index to that of the 
influenza case data. For choosing keywords they reference Ginsburg et 
al. mentioning the benefit of choosing an optimal set of keywords, but 
want a more efficient way of determining such a set. Yuan et al. 
determined the key words using a tool from a Chinese website 
(http://tool.chinaz.com/baidu/words.aspx) which determined keywords 
through recommendations from Baidu as well as from others mined using 
semantic correlation analysis from sources such as portal websites, 
online reports, and blogs. The source keyword put into the tool was the 
Chinese symbol for Flu giving a total of 94 related keywords (assuming a
 total of 95 including Flu is the current set). The current set of 
keywords is then taken into the filtering step where three conditions 
must be met: (i) the keyword must include aspects that could impact the 
influenza epidemic, (ii) the data for a keyword must be presented with a
 time series with a resolution (daily, weekly, or monthly), and finally,
 (iii) the time series data for the chosen keywords must have a maximum 
cross-correlation coefficient no less than 0.4 to that of the influenza 
case data.</p><p class="Para">The next step starts with defining the 
weights for the set of keywords that passed filtering. The authors 
tested two techniques for determining weights for the keywords: 
systematic assessment (SA)
[<span class="CitationRef"><a href="#CR60" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">60</a></span>, <span class="CitationRef"><a href="#CR61" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">61</a></span>] and strength of the correlation coefficient (SCC)
[<span class="CitationRef"><a href="#CR21" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">21</a></span>, <span class="CitationRef"><a href="#CR62" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">62</a></span>].
 SA entails using the principal of prior evaluations to rate the 
keywords and assign these ratings as weights, while SCC compares the 
influenza epidemic curve to that of the keyword frequency curve 
determining the correlation coefficient between these and assigning the 
weights accordingly. Yuan et al. decide to determine the optimal set of 
keywords with the formula: <em class="EmphasisTypeItalic ">y</em>=<em class="EmphasisTypeItalic ">α</em><sub>0</sub>+<em class="EmphasisTypeItalic ">α</em><sub>1</sub>×<em class="EmphasisTypeItalic ">i</em> <em class="EmphasisTypeItalic ">n</em> <em class="EmphasisTypeItalic ">d</em> <em class="EmphasisTypeItalic ">e</em> <em class="EmphasisTypeItalic ">x</em><sub> <em class="EmphasisTypeItalic ">j</em> </sub>+<em class="EmphasisTypeItalic ">ε</em>, where <em class="EmphasisTypeItalic ">α</em><sub>0</sub> is the intercept, <em class="EmphasisTypeItalic ">α</em><sub>1</sub> is the coefficient, and <em class="EmphasisTypeItalic ">ε</em> is the error. The <em class="EmphasisTypeItalic ">i</em> <em class="EmphasisTypeItalic ">n</em> <em class="EmphasisTypeItalic ">d</em> <em class="EmphasisTypeItalic ">e</em> <em class="EmphasisTypeItalic ">x</em><sub> <em class="EmphasisTypeItalic ">j</em> </sub> is equal to
<span class="InlineEquation" id="IEq1"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow>
                        <munderover>
                          <mrow>
                            <mo>∑</mo>
                          </mrow>
                          <mrow>
                            <mi>i</mi>
                            <mo>=</mo>
                            <mn>1</mn>
                          </mrow>
                          <mrow>
                            <mi>j</mi>
                          </mrow>
                        </munderover>
                        <msub>
                          <mrow>
                            <mi>ω</mi>
                          </mrow>
                          <mrow>
                            <mi>i</mi>
                          </mrow>
                        </msub>
                        <msubsup>
                          <mrow>
                            <mi>x</mi>
                          </mrow>
                          <mrow>
                            <mi>i</mi>
                          </mrow>
                          <mrow>
                            <mi>l</mi>
                          </mrow>
                        </msubsup>
                      </mrow><annotation-xml encoding="application/xhtml+xml" name="alternate-representation"><span xmlns="http://www.w3.org/1999/xhtml"><a href="https://static-content.springer.com/image/art%3A10.1186%2F2196-1115-1-2/MediaObjects/40537_2013_Article_2_IEq1_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="40537_2013_Article_2_IEq1_HTML.gif" alt=""></a></span></annotation-xml></semantics></math></span>, where <em class="EmphasisTypeItalic ">ω</em><sub> <em class="EmphasisTypeItalic ">i</em> </sub> is the weight for the <em class="EmphasisTypeItalic ">i</em> th keyword and
<span class="InlineEquation" id="IEq2"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow>
                        <msubsup>
                          <mrow>
                            <mi>x</mi>
                          </mrow>
                          <mrow>
                            <mi>i</mi>
                          </mrow>
                          <mrow>
                            <mi>l</mi>
                          </mrow>
                        </msubsup>
                      </mrow><annotation-xml encoding="application/xhtml+xml" name="alternate-representation"><span xmlns="http://www.w3.org/1999/xhtml"><a href="https://static-content.springer.com/image/art%3A10.1186%2F2196-1115-1-2/MediaObjects/40537_2013_Article_2_IEq2_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="40537_2013_Article_2_IEq2_HTML.gif" alt=""></a></span></annotation-xml></semantics></math></span>
 denotes the sequence after alignment (not used in this study, 
presumably set to 1). Each keyword will be brought into the model in a 
stepwise manner in an order determined by their correlation coefficient 
and then use a partial F test to evaluate goodness of fit after each 
keyword is brought into the model (this will be repeated until the 
goodness of fit stops improving). After all the keyword selection 
processes are concluded they are left with 8 keywords left for the 
optimal set.</p><p class="Para">The authors decided to use the following regression model: <em class="EmphasisTypeItalic ">I</em> <em class="EmphasisTypeItalic ">C</em> <em class="EmphasisTypeItalic ">D</em>[<em class="EmphasisTypeItalic ">t</em>]=<em class="EmphasisTypeItalic ">β</em><sub>0</sub>×<em class="EmphasisTypeItalic ">I</em> <em class="EmphasisTypeItalic ">C</em> <em class="EmphasisTypeItalic ">D</em>[<em class="EmphasisTypeItalic ">t</em>-1]+ <em class="EmphasisTypeItalic ">β</em><sub>1</sub>×<em class="EmphasisTypeItalic ">i</em> <em class="EmphasisTypeItalic ">n</em> <em class="EmphasisTypeItalic ">d</em> <em class="EmphasisTypeItalic ">e</em> <em class="EmphasisTypeItalic ">x</em>[<em class="EmphasisTypeItalic ">t</em>]+ <em class="EmphasisTypeItalic ">β</em><sub>2</sub>×<em class="EmphasisTypeItalic ">i</em> <em class="EmphasisTypeItalic ">n</em> <em class="EmphasisTypeItalic ">d</em> <em class="EmphasisTypeItalic ">e</em> <em class="EmphasisTypeItalic ">x</em>[<em class="EmphasisTypeItalic ">t</em>-1]+ <em class="EmphasisTypeItalic ">ε</em>, where ICD is the influenza case data, <em class="EmphasisTypeItalic ">β</em><sub>0</sub>, <em class="EmphasisTypeItalic ">β</em><sub>1</sub>, and <em class="EmphasisTypeItalic ">β</em><sub>2</sub> are coefficients, <em class="EmphasisTypeItalic ">index</em> represents the optimal set, and again <em class="EmphasisTypeItalic ">ε</em> represents the error. The variable <em class="EmphasisTypeItalic ">t</em>
 in this formula stands for time and can be broken into month time 
blocks. The formula therefore would be estimating ICD using the ICD from
 the previous month and on the optimal index for the current and 
previous months. The model was trained on the data from March 2009 to 
December 2011 and validated on the time period of January 2012 to August
 2012. They achieved an R-squared of 0.95, an AIC of 18.50 and found 
that autocorrelation was not an issue due to the Durbin-Watson test 
results of 1.89. The validation of the model tested over the validation 
time period resulted in a mean absolute error of 10.6%, but in a more 
real-time manner of up to 1-2 weeks earlier. One area that could have 
improved in this study is the starting pool was very small compared to 
the study done by Ginsburg et al.
[<span class="CitationRef"><a href="#CR21" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">21</a></span>], depending on other sources and not their techniques to determine the staring pool of keywords.</p><p class="Para">These
 studies have shown that search query data can be a useful tool for 
quickly and accurately detecting the occurrence of an ILI epidemic which
 could even be extended to tracking an epidemic. Additionally, it would 
be interesting to see if the formula created on this study could work as
 well for ILI epidemics that happen many years in the future from the 
time period used to train the model or in a different population that 
uses the same language. This line of research can help health officials,
 physicians, hospitals in reacting to epidemics faster and work to stop 
them better (faster) than with traditional methods used today such as 
the CDC or MOH reports.</p></section><section id="Sec19" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Using twitter post data to track epidemics</h3><p class="Para">The
 studies discussed in this subsection have a similar research goal to 
the previous subsection of attempting to detect and track ILI epidemics,
 but instead of using search query data the researchers use Twitter post
 data. One advantage to Twitter data over search query data is that 
Twitter posts come with context
[<span class="CitationRef"><a href="#CR23" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">23</a></span>]
 (as opposed to not knowing why a person is searching for a topic). 
Twitter is a social networking site that allows its user to post any 
messages they like in 140 words or less (known as tweets), and currently
 has about 554 hundred million members worldwide with around 58 million 
tweets per day
[<span class="CitationRef"><a href="#CR63" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">63</a></span>].
 With this Big Volume of people (sensors) there is a high probability 
that there can be useful ILI epidemic information being posted, but, of 
course, there will be noisy sensors and only through data mining 
techniques and analysis can the useful information be found. Another 
issue that could make Twitter post data potentially not reliable is that
 it is difficult to directly ascertain the age of a given Twitter user, 
and in fact a user may be discussing a disease symptom from a family 
member who differs in age from the user anyway, but research shows
[<span class="CitationRef"><a href="#CR22" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">22</a></span>] that this is not a deterrent to age-based prediction.</p><p class="Para">Signorini et al.
[<span class="CitationRef"><a href="#CR23" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">23</a></span>],
 in their research, employ Twitter post data across the United States by
 searching through particular spatiotemporal areas and analyzing the 
data in order to predicate weekly ILI levels both across and within 
these regions (CDC’s ILI regions). The focus of their efforts is on the 
time period when the H1N1 epidemic was happening in the US as they 
gathered a large amount of Tweets from October 1, 2009 - May 20, 2010 
using Twitter’s streaming application programmer’s interface (API)
[<span class="CitationRef"><a href="#CR64" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">64</a></span>]. The tweets were sifted through looking for posts containing a preset of key words correlation to H1N1 (<em class="EmphasisTypeItalic ">h1n1</em>, <em class="EmphasisTypeItalic ">flu</em>, <em class="EmphasisTypeItalic ">swine</em>, <em class="EmphasisTypeItalic ">influenza</em>).
 Twitter’s posts stream does come with its own filters according to the 
API documentation; therefore, not all tweets in the United States were 
used, only the subset to come through Twitter’s filtered stream.</p><p class="Para">Tweets
 containing the following attributes were not used for analysis: if 
located outside the United States, from a user with a time zone outside 
the US, containing less than five words, not in English, not containing 
ASCII characters, and those submitted through the “API”. The tweets 
leftover are used to create a dictionary of English words, from which 
items such as (#hashtags, @user, and links) are not used and also words 
were brought to their root form through Porter’s Stemming Algorithm
[<span class="CitationRef"><a href="#CR65" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">65</a></span>]
 to make the dictionary as efficient to use as possible. Using this 
dictionary Signori et al. gathered daily and weekly statistics (such as 
amount of tweets a word is present within) for each word both in the 
dictionary throughout the US and within each of the CDC’s 10 regions.</p><p class="Para">The
 authors use the weekly statistics in order to estimate the weekly ILI 
epidemic status through a more general class of SVM called Support 
Vector Regression
[<span class="CitationRef"><a href="#CR66" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">66</a></span>].
 This is a generalization of Support Vector Machines, a type of 
classifier which attempts to find a minimal-margin separator, which is a
 hyperplane in the space of instances such that one class is on one side
 of the hyperplane and the other class is on the other side, with the 
distance between each class’s instances and the hyperplane being 
maximized. As data is rarely linearly separable in feature space, a 
kernel function is used to transform the data into a higher-dimensional 
space, which generally assures that such a hyperplane exists. Signorini 
et al. used a polynomial kernel function for this purpose. To extend 
SVMs to the problem of regression (as opposed to classification), the 
model disregards any training data points which are already within a 
threshold <em class="EmphasisTypeItalic ">ε</em> of the model prediction
 (just as a SVM classification model disregards points which lie outside
 such a margin, as those points cannot help in determining the optimal 
hyperplane), and then builds a nonlinear model to minimize a preselected
 linear-error-cost function. In this model, each point (or instance) is a
 tweet, and the features each represent dictionary terms which occur 
more than 10 times per week. The value of each feature is the fraction 
of total tweets within the given week which contain the corresponding 
dictionary word (after stemming).</p><p class="Para">Determining if 
Twitter data can indeed detect ILI epidemics by accurately estimating 
CDC ILI values was done on a weekly basis on a national level and a 
regional level (where region 2 is presented by the authors). For the 
national estimation they trained their method using 1 million of the 
tweets from October 1, 2009 - May 20, 2010 throughout the United States,
 with the objective being CDC’s ILI values throughout the nation. 
Leave-one-out cross-validation was used to determine the accuracy of the
 model. The results found for the national level were quite accurate, 
scoring an average error of 0.28% with a standard deviation of 0.23%. In
 order to estimate the ILI levels in a particular region, with the goal 
being real-time prediction (as again, the CDC has a 1 to 2 week delay), 
they took the tweets that had geolocation information and fit these 
tweets to CDC region ILI readings from 9 of the 10 regions in order to 
construct the model to determine the results for last region (region 2).
 The authors argue that perhaps the smaller amount of tweets containing 
the geolocation information could have generated the slightly higher 
error rating of 0.37% with a standard deviation of 0.26%.</p><p class="Para">This
 study could have incorporated more words to include in their tweet 
searches rather than just the 4 they used as well as use methods to 
determine the most affective set. Also there was one prediction model 
applied, where as with more tested they could have found that a 
different model generated better results with less error across the US 
and within the tested region. As a note for this paper, Signorini et al.
[<span class="CitationRef"><a href="#CR23" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">23</a></span>]
 also (and primarily) used the tweets to follow public concern for ILI 
epidemics throughout daily and monthly trends of tweets, but the scope 
of this survey does not look to cover such findings.</p><p class="Para">Achrekar et al.
[<span class="CitationRef"><a href="#CR22" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">22</a></span>]
 devises a system deemed Social Network Enabled Flu Trends (SNEFT) that 
continuously monitors tweets with the goal of detecting and tracking the
 spread of ILI epidemics. This study uses a dataset of tweets and 
profile details of the Twitter users who have commented on flu keywords 
starting on October 18, 2009. The SNEFT network uses an OSN Crawler (bot
 that systematically searches online social networks) they developed to 
retrieve tweets from the internet using keywords <em class="EmphasisTypeItalic ">flu</em>, <em class="EmphasisTypeItalic ">H1N1</em>, and <em class="EmphasisTypeItalic ">swine flu</em>
 storing important information about the tweets (e.g. location, further 
determined by Google to determine geolocation), and relative keyword 
frequency in a spatio-temporal database. Also gathered are CDC ILI 
reports along with other influenza related data are downloaded from the 
internet.</p><p class="Para">The authors found that in one of their 
previous works using tweets from 2009–2010 (when influenza was a larger 
issue), the correlation coefficient was 0.98, but a huge drop was seen 
in this measure since for 2010-2011 at 0.47. Due to the H1N1 being less 
of an issue the noisy tweets are much more prominent resulting in a 
lower correlation coefficient, which leads to give something to think 
about: maybe this measure could be used to detect ILI in an area where 
the higher it is (without using methods to correct such noise) the more 
likely it is that an ILI epidemic is spreading through that region. The 
authors do try to alleviate the noise using text classification 
determining whether or not a tweet is correlated to a flu event or not.</p><p class="Para">For
 text classification they tested three different methods: decision tree,
 SVMs, and Naive Bayes, along with a few configurations for efficient 
learning. SVM was determined as the best classification method, beating 
the other two methods in both precision and recall. These classifiers 
were trained on a dataset of 25,000 tweets manually classified by the 
use of Amazon Mechanical Turk (an internet marketplace to perform such 
tasks by the coordinated use of human intelligence). The results of 
adding this text classifier caused the correlation coefficient to rise 
back up for the US as a whole and within each region. The authors also 
performed data cleansing from both retweets and multiple tweets posted 
by the same user during a single bout with the same illness (different 
time windows were tested for the length of a single bout). Results found
 that only retweets should be considered.</p><p class="Para">The 
prediction model they used is Logistic Autoregression with exogenous 
inputs (ARX), which has the goal of predicting the CDC ILI statistic 
during a given week using both the tweets details and the CDC data 
(percent of physician visits from preceding weeks). The autoregression 
part of the model is the prediction of current ILI activity employing 
ILI activity from past weeks, and the exogenous inputs come from the 
tweets from the previous weeks. As simple Linear ARX could not handle 
the fact that the number of Twitter users is only bounded below by 0 
(rather than being completely bounded between 0% and 100%), and so the 
authors introduce a logit link function for the CDC data and use a 
logarithmic transformation of the Twitter data. The purpose of the model
 is to produce timely updates estimating the percentage of physician 
visits for the week. The model can optionally use two distinct variables
 to determine how many previous weeks of data are used for each data 
type, with <em class="EmphasisTypeItalic ">m</em> referring to CDC data and <em class="EmphasisTypeItalic ">n</em>
 referring to Twitter data. The authors used 10-fold cross-validation on
 33 weeks of data (from October 3, 2010 - May 15, 2011) to determine the
 (<em class="EmphasisTypeItalic ">m</em>,<em class="EmphasisTypeItalic ">n</em>)
 model that gave the least root mean squared error, and the (2,1) model 
(using only the current twitter data and the two most recent instances 
of CDC data) gave the best results.</p><p class="Para">Using their 
methods they were able to achieve a higher correlation to the actual CDC
 percent of visits ILI-related compared to the Twitter data alone for 
the 33 week period in 2010-2011, yet the authors do mention that Twitter
 data alone would still predict higher toward the beginning and end of 
the annual flu season (and, of course, during an epidemic). Achrekar et 
al. also found that they could get good results for regions 1, 6, and 9 
determining that Twitter data and percentage of visits ILI-related are 
correlated across regions. One issue with this research is that only one
 prediction method is used; it would have been beneficial to see if the 
jump in correlation could also be seen with other prediction models or 
that results could be better with other models.</p><p class="Para">Results
 shown here are impressive, showing that Twitter data can be used to 
detect and possibly track Influenza like epidemics in real time. Both 
research efforts have shown that they can get very little error for 
predicting ongoing ILI epidemics, and produce results up to 1-2 weeks 
faster than the CDC posts their results. One issue not looked at by 
either paper discussed here, but discussed by Doan et al.
[<span class="CitationRef"><a href="#CR67" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">67</a></span>],
 is that results could be improved if the keywords used were more broad,
 using a more knowledge-based method with 37 symptom keywords under 
respiratory syndromes from the BioCaster Ontology (BCO) [79] plus the 
word <em class="EmphasisTypeItalic ">flu</em>. This leads to wondering 
what results could be found for ILI prediction if other keyword 
selection methods were chosen. It would be interesting if more areas 
across the globe were researched in order to see if the results shown 
here can be achieved worldwide. An example of such research could be to 
see if similar results can be garnered using a set of keywords used to 
achieve good prediction results in a given population could get the same
 results in another. This line of research, with the promising results 
shown, could possibly one day create a system that could continually be 
tracking twitter posts in order to create a worldwide ILI epidemic map 
in real time helping people and the health care systems stay one step 
ahead.</p></section></div></section><section id="Sec20" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Translational bioinformatics<span class="section-icon"></span></h2><div class="content"><p class="Para">Many
 authors argue that although it is a young field, Translational 
BioInformatics (TBI) is the way of the future for Health Informatics. 
TBI is an interdisciplinary subfield that deals with High Volumes of 
biomedical data and genomic data, where current research areas include 
developing new techniques for integrating biological data and clinical 
data as well as improving clinical methodology by including findings 
from biological research
[<span class="CitationRef"><a href="#CR68" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">68</a></span>]. According to Chen et al.
[<span class="CitationRef"><a href="#CR1" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1</a></span>],
 the scope of TBI encompasses all the same levels of Health Informatics 
in general: Micro Level (i.e. Molecules), Tissue Level, Patient Level, 
and Macro Level (i.e. Population). The main goal of TBI is answering 
various questions at the clinical level. However, TBI does not seem to 
have a universally accepted definition. Also confused is the dividing 
line between what is included as clinical information and overall Health
 Informatics (where both seem to be used interchangeably as biomedical).
 When clinical information is referenced, it does seem to include all 
Health Informatics’ levels from the tissue, patient, and population 
levels (biomedical data), as they are all connected to health. TBI is 
the field to bridge the gap between these fields and the molecular 
level, by way of developing tools that can better link this disparate 
data as well as develop and test techniques that can efficiently and 
accurately analyze such data together with the end goal of improving HCO
[<span class="CitationRef"><a href="#CR69" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">69</a></span>].</p><p class="Para">TBI
 looks to use the health information from the discussed levels, combine 
the aggregated information in order to provide the most health gains, 
and help in offering the best of modern health practices. It makes sense
 that any field of research containing numerous subfields (including 
Health Informatics) would eventually converge into “translation” between
 its subfields, and this is what appears to have happened in the 
construction of TBI. To fully or even partially encompass TBI, many 
discussions and research examples would need to be presented, but for 
the scope of this paper the essence of “translation” is what will be 
shown. Thus, this section will present a small sample of discussions 
(such as editorials, perspectives, and highlights) from JAMIA (the 
Journal of the American Medical Informatics Association) to give the 
overall feel of TBI.</p><p class="Para">Butte et al.
[<span class="CitationRef"><a href="#CR70" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">70</a></span>],
 in their 2011 editorial, discuss several Translational Bioinformatics 
studies featured in JAMIA which combine biological data (e.g. microarray
 data) with medical records to achieve medical gains as more data angles
 are tested in tandem. The authors acknowledge that TBI started from 
research done by a small group who found how to bridge the gap between 
computational biology and medicine. They also comment that they expect 
TBI to exploit Big Data in the not-too-distant future, with the goal of 
answering various clinical level questions and leading to a number of 
clinically applicable decisions.</p><p class="Para">Sarkar et al.
[<span class="CitationRef"><a href="#CR69" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">69</a></span>]
 in a perspective article (released in 2011) discuss three areas as 
being the primary research objectives of TBI: 1.) determining the 
molecular level (genotype) impacts on the evolution of diseases, 2.) 
learning the impact of therapeutic procedures as can be measured by 
molecular biomarkers, and 3.) understanding the overall consistency 
between the molecular, the phenotype, and environmental correlations 
across different populations. The authors believe that with the 
explosion of both the molecular level data and biomedical data, as well 
as with the technological advances, TBI is in a prime position to 
possibly determine many of the mysteries of complex diseases or any of 
the other primary research objectives mentioned.</p><p class="Para">Shah et al.
[<span class="CitationRef"><a href="#CR71" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">71</a></span>],
 in their 2012 editorial, discuss JAMIA’s focus of Big Data in TBI. The 
goal of their 2012 summit (as well as TBI in general) was bringing 
molecular level data into Health Informatics, which is now possible due 
to the explosion of computational power now available. With this new 
computational power, molecular level data can now be used to improve 
medical understanding, as the farther one deviates from the patient 
level, the more data will be generated. This editorial acknowledges many
 works that are implementing TBI methodologies with success, including 
Liu et al.
[<span class="CitationRef"><a href="#CR72" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">72</a></span>],
 who combined chemical, biological, and phenotype properties of drugs to
 improve predictions of adverse drug reactions by 5%, compared to only 
using chemical properties of the drugs. The authors argue that data for 
TBI should include a Big Volume of molecular data and a small amount of 
patient measurements from a Big Volume of patients. This leads to us 
wondering why are the authors limiting the scope of TBI research to only
 these two levels, and why not (as shown in
[<span class="CitationRef"><a href="#CR1" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1</a></span>])
 include all levels of accessible human existence. The authors also 
mention, in closing, that the TBI Summit and the Clinical Research 
Informatics Summit are increasing their synergy, which is a good start, 
but eventually study groups from all levels of accessible human 
existence should come together in the same way.</p></div></section><section id="Sec21" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Analysis and future works<span class="section-icon"></span></h2><div class="content"><p class="Para">This section will cover analysis and future work that could be beneficial from the lines of research presented in this survey.</p><section id="Sec22" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Covering molecule-level</h3><p class="Para">For the subsection “Using Gene Expression Data to Make Clinical Predictions” (covering studies
[<span class="CitationRef"><a href="#CR6" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">6</a></span>] and
[<span class="CitationRef"><a href="#CR7" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">7</a></span>]),
 the main challenge is handling the Big Volume of the features (probes).
 A particular concern, with gene expression microarray data only getting
 larger, is developing and testing probe selection techniques. Future 
work will need to be put into making these probe selection techniques 
able to handle these vast amounts of gene probes and select the subset 
that has the best correlation to the final prediction in a way that is 
as fast, accurate, and efficient as possible.</p></section><section id="Sec23" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Covering tissue-level</h3><p class="Para">For the “Creating a Connectivity Map of the Brain Using Brain Images” subsection (presented by the studies
[<span class="CitationRef"><a href="#CR8" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">8</a></span>, <span class="CitationRef"><a href="#CR66" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">66</a></span>], and
[<span class="CitationRef"><a href="#CR9" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">9</a></span>]),
 actual data mining analysis of the connectivity map remains entirely in
 the scope of future work. With the Big Volume of data being created by 
the HCP and the study done by Annese
[<span class="CitationRef"><a href="#CR8" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">8</a></span>]
 there is a lot of opportunity for novel data mining techniques to be 
employed, allowing for the possible discovery of previously-unattainable
 knowledge about the brain and how it connects to the health of the 
human body, leading to possibly giving physicians more accurate 
diagnosis methods, earlier detection of diseases, etc.</p><p class="Para">For the subsection “Using MRI Data for Clinical Prediction” (which covered research efforts
[<span class="CitationRef"><a href="#CR11" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">11</a></span>] and
[<span class="CitationRef"><a href="#CR10" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">10</a></span>]),
 the difficulty is handling the Big Volume of the high-resolution MRI 
data. As each instance in an MRI dataset is quite large, and numerous 
MRI samples are needed to perform acceptable research, attempting to 
analyze the whole brain MRI datasets can be of quite a Big Volume 
challenge; luckily, recent computational power has entered an era that 
can handle such data (meaning this line of research is fairly young). 
More testing will be needed here, and the way forward for using MRI data
 for clinical predictions is to create and test new machine learning 
methods that can accurately locate brain regions that best correlate to 
specific ailments to help physicians make more reliable predictions and 
diagnosis.</p></section><section id="Sec24" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Covering patient-level</h3><p class="Para">For the subsection “Prediction of ICU Readmission and Mortality Rate” (presented by the studies
[<span class="CitationRef"><a href="#CR12" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">12</a></span>, <span class="CitationRef"><a href="#CR13" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">13</a></span>, <span class="CitationRef"><a href="#CR15" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">15</a></span>], and
[<span class="CitationRef"><a href="#CR14" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">14</a></span>]),
 although all the studies had similar goals, none started with the same 
pool of variables; it could be beneficial in future work if all these 
variables and even variables from other data levels could be used in 
this area of research (as data from all 4 levels discussed here can be 
used to answer clinical questions). Also, data from the other studies 
presented previously could be used to expand the size of the starting 
pool (such as MRI data seen in
[<span class="CitationRef"><a href="#CR10" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">10</a></span>], using the relative baseline method from
[<span class="CitationRef"><a href="#CR18" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">18</a></span>], or even message board data from
[<span class="CitationRef"><a href="#CR19" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">19</a></span>] and
[<span class="CitationRef"><a href="#CR20" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">20</a></span>]).
 Another item to note is that throughout these studies only one feature 
selection technique was tested with minimal overlap; therefore, future 
work should look to test multiple feature selection techniques in order 
to find which one works the best with medical data. Classifiers that are
 employed for this line of research should be able to make decisions as a
 physician would do, that is, be able to look at a patient’s medical 
attributes and make subjective decisions. Thus, there are three main 
concerns for making these predictions (starting variables, feature 
selection, and learning algorithms) and future research should be to 
test as many combinations of these three to find the best one that can 
make the most accurate and efficient predictions.</p><p class="Para">For the “Real-Time Predictions Using Data Streams” subsection (covered by the research efforts
[<span class="CitationRef"><a href="#CR16" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">16</a></span>, <span class="CitationRef"><a href="#CR18" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">18</a></span>], and
[<span class="CitationRef"><a href="#CR17" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">17</a></span>]),
 we see that this area of research is relatively new due to technology 
only very recently being able to handle the high-throughput processing 
necessary from the Big Velocity of data streams. As such, research is 
lacking testing and validation of developed methods as seen for Zhang et
 al.
[<span class="CitationRef"><a href="#CR17" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">17</a></span>] and Thommandram et al.
[<span class="CitationRef"><a href="#CR18" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">18</a></span>].
 As sensors are not perfect (creating missing or erroneous data during a
 given time period), especially when being used for real-time analysis, 
future work will need to focus on developing and testing methods that 
can handle such data in the most reliable and efficient way. Another 
issue for future work in this line research is the need to devise and 
test numerous classification methods to find the best for both prognosis
 and acute problem detection (knowing that the best choices could be 
different for different ailments). It could also be beneficial if 
numerous sensors were checked to find the best set of sensors for each 
ailment prediction.</p></section><section id="Sec25" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Covering population-level</h3><p class="Para">For the “Using Message Board Data to Help Patients Obtain Medical Information” subsection (presented here by
[<span class="CitationRef"><a href="#CR19" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">19</a></span>] and
[<span class="CitationRef"><a href="#CR20" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">20</a></span>]),
 we find that the existing work does show that message board data does 
have the potential to supply patients with reliable medical information.
 More real world testing will be needed before this question can be 
fully answered, however. Message board data could be helpful if used in 
conjunction with studies such as Zhang et al.’s.
[<span class="CitationRef"><a href="#CR17" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">17</a></span>] research on real-time diagnosis and prognosis, if it was found to give reliable medical data.</p><p class="Para">For the subsection “Tracking Epidemics Using Search Query Data” (shown by
[<span class="CitationRef"><a href="#CR21" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">21</a></span>] and
[<span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">5</a></span>]),
 the results shown are promising as they could detect the occurrence of 
ILI epidemics in a given region in a relatively real-time manner with 
minimal error. Future work for this line of research will need to find 
ways of finding the optimal set of keywords/queries to use for 
predicting the occurrence of an ILI epidemic. Also work should be done 
to determine whether research done in one area of the world can be 
translated to another (e.g. could the results found from Ginsburg et al.
[<span class="CitationRef"><a href="#CR21" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">21</a></span>] translate to China’s region or visa-versa), or across different languages or different areas with the same language.</p><p class="Para">For the “Tracking Epidemics Using Twitter Post Data” subsection (covered by
[<span class="CitationRef"><a href="#CR22" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">22</a></span>] and
[<span class="CitationRef"><a href="#CR23" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">23</a></span>]),
 we see that Twitter data can be quite useful for detecting or even 
tracking ILI epidemics. In the future there should be more work on 
developing methods to best determine what keywords to use for study 
(filtering methods), as well as testing more text classification methods
 in order to reduce the noisy tweets in the collected datasets. More 
research should also be concentrated on determining the best model for 
using Twitter post data to predict the CDC’s percentage of ILI related 
visits, as according to
[<span class="CitationRef"><a href="#CR22" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">22</a></span>]
 this is one of the ways the CDC monitors ILI epidemics and Twitter is 
quite highly correlated to this statistic. This line of research, with 
the promising results shown, could be possibly one day help create a 
system that could continually track Twitter posts in order to create a 
worldwide ILI epidemic map in real time, helping people and the health 
care systems stay one step ahead. Also, this leads one to wonder what 
else Twitter data can be used to determine for the health care system.</p></section><section id="Sec26" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading">Covering translational bioinformatics</h3><p class="Para">As discussed by reviews and editorials by
[<span class="CitationRef"><a href="#CR1" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1</a></span>, <span class="CitationRef"><a href="#CR69" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">69</a></span>, <span class="CitationRef"><a href="#CR70" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">70</a></span>], and
[<span class="CitationRef"><a href="#CR71" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">71</a></span>],
 the way forward for TBI and Health Informatics as a whole is for the 
translational approach to increasingly envelop the entire scope of 
Health Informatics and continue to combine data from all levels of human
 existence (as diseases and ailments handled by the healthcare system 
are very complex). Through this combination, questions throughout all 
levels can be more precisely answered and results can be validated both 
more quickly and more accurately. All future work in Health Informatics 
should look to take the translational approach shown by TBI, not just 
focusing on combining the molecular level with the other levels, but 
attempting to make connections across as many levels of data as 
possible. This combination of data would offer Big Volume, Velocity, 
Variety, Veracity, and, of course, Value, which could provide an 
unprecedented degree of medical knowledge gain.</p></section></div></section><section id="Sec27" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Conclusion<span class="section-icon"></span></h2><div class="content"><p class="Para">This
 survey discussed a number of recent studies being done within the most 
popular sub branches of Health Informatics, using Big Data from all 
accessible levels of human existence to answer questions throughout all 
levels. Analyzing Big Data of this scope has only been possible 
extremely recently, due to the increasing capability of both 
computational resources and the algorithms which take advantage of these
 resources. Research on using these tools and techniques for Health 
Informatics is critical, because this domain requires a great deal of 
testing and confirmation before new techniques can be applied for making
 real world decisions across all levels. The fact that computational 
power has reached the ability to handle Big Data through efficient 
algorithms (as well as hardware advances, of course) lets data mining 
handle the Big Volume, Velocity, Variety, Veracity, and Value of the 
data generated by Health Informatics (traditional or otherwise). The use
 of Big Data provides advantages to Health Informatics by allowing for 
more tests cases or more features for research, leading to both quicker 
validation of studies and the ability to accrue enough instances for 
training when only a small fraction of instances exist within the 
positive class. The way forward for Health Informatics is definitely 
exploiting the Big Data created throughout all the various levels of 
medical data and finding ways to best analyze, mine and answer as many 
medical questions as possible.</p><p class="Para">As mentioned, the 
overall goal of answering any medical question, whether it be on the 
level of human-scale biology, clinical itself, or population, is to 
eventually improve healthcare for patients. The human body is a compound
 and complex system, containing many levels (not all of which may even 
be accessible as of now). Therefore, research needs to be done on data 
at all of these levels in order to answer the ever-growing list of 
medical questions on all of these levels. The studies covered here seek 
to answer questions on all four levels from data through analysis of 
data from the molecular, tissue, patient, and population levels.</p><p class="Para">TBI
 has the scope to use all these data levels with the goal of answering 
clinical questions, but this objective could benefit from not just 
attempting to answer clinical questions but questions on all levels. 
Perhaps future research in TBI and Health Informatics overall could 
focus on using data from all levels in order to find correlations and 
connections between them, possibly giving physicians more ways of 
diagnosing, treating, and helping their patients. All future work in 
Health Informatics should have a translational approach of using data 
from all levels of human existence.</p><p class="Para">All of the 
techniques discussed here show promise, provide inspiration for future 
work, and show the importance of using all accessible levels of data in 
Health Informatics. Each of the techniques presented in these studies 
can be both expanded and further tested using datasets with Big Volume 
and Variety (potentially on data from other levels) to see if their 
results are the same through different populations.</p><p class="Para">These
 studies are only a taste of the future possibilities that could be 
achieved through data mining and analysis of Big Data for Health 
Informatics. As computational power increases, more efficient and 
accurate methods will be developed. This could lead to possibly new 
levels of human existence becoming available for analysis, such as below
 the level of molecular data (e.g. the atomic scale) where the data 
gathered from each patient would be Really Big Data (RBD). Who knows, 
maybe a missing electron in a given chromosome could indicate a patient 
will be susceptible to cancer.</p><p class="Para">Even with the research
 discussed here and even with the promising results shown, it is up to 
future research to improve medical knowledge and create more advanced 
methods, as well as the readiness of the healthcare field to accept and 
apply these findings and techniques to improve HCO.</p></div></section></div><section id="Notes" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Notes<span class="section-icon"></span></h2><div class="content"><aside class="ArticleNote ArticleNoteMisc"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Competing interests<span class="section-icon"></span></h2><p class="SimplePara">The authors declare that they have no competing interests.</p></aside><aside class="ArticleNote ArticleNoteMisc"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">Authors’ contributions<span class="section-icon"></span></h2><p class="SimplePara">MH
 performed the primary literature review and analysis for this work, and
 also drafted the manuscript. RW worked with MH to develop the article’s
 framework and focus. TMK introduced this topic to MH and RW, and 
coordinated the other authors to complete and finalize this work. All 
authors read and approved the final manuscript.</p></aside></div></section><section class="Section1 RenderAsSection1" id="Bib1" tabindex="-1"><h2 class="Heading" data-role="collapsible-handle" tabindex="0">References<span class="section-icon"></span></h2><div class="content"><ol class="BibliographyWrapper"><li class="Citation"><div class="CitationNumber">1.</div><div class="CitationContent" id="CR1">Chen J, Qian F, Yan W, Shen B: <strong class="EmphasisTypeBold ">Translational biomedical informatics in the cloud: present and future.</strong> <em class="EmphasisTypeItalic ">BioMed Res Int 2013</em> 2013, 8. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://dx.doi.org/10.1155/2013/658925"><span class="RefSource">http://dx.doi.org/10.1155/2013/658925</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Translational%20biomedical%20informatics%20in%20the%20cloud%3A%20present%20and%20future&amp;author=J.%20Chen&amp;author=F.%20Qian&amp;author=W.%20Yan&amp;author=B.%20Shen&amp;pages=8&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">2.</div><div class="CitationContent" id="CR2">Martin M: <strong class="EmphasisTypeBold ">Big Cdata/social media combo poised to advance healthcare.</strong> <em class="EmphasisTypeItalic ">HPC Source</em> 2013, 33–35. <span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.scientificcomputing.com/digital-editions/2013/04/hpc-source-big-data-beyond"><span class="RefSource">http://www.scientificcomputing.com/digital-editions/2013/04/hpc-source-big-data-beyond</span></a></span>
<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Big%20Cdata%2Fsocial%20media%20combo%20poised%20to%20advance%20healthcare&amp;author=M.%20Martin&amp;pages=33-35&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">3.</div><div class="CitationContent" id="CR3">Demchenko Y, Zhao Z, Grosso P, Wibisono A, de Laat C: <strong class="EmphasisTypeBold ">Addressing Big Data challenges for Scientific Data Infrastructure.</strong> In <em class="EmphasisTypeItalic ">IEEE 4th International Conference on Cloud Computing Technology and Science (CloudCom 2012)</em>. Taipei, Taiwan: IEEE Computing Society, based in California, USA; 2012:614–617.<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/CloudCom.2012.6427494"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Addressing%20Big%20Data%20challenges%20for%20Scientific%20Data%20Infrastructure&amp;author=Y.%20Demchenko&amp;author=Z.%20Zhao&amp;author=P.%20Grosso&amp;author=A.%20Wibisono&amp;pages=614-617&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">4.</div><div class="CitationContent" id="CR4">Huan JL, Pai V, Teredesai AM, Yu S(Eds): <em class="EmphasisTypeItalic ">IEEE Workshop on BigData In Bioinformatics and Health Care Informatics</em>. 2013. <span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.ittc.ku.edu/%7Ejhuan/BBH/"><span class="RefSource">http://www.ittc.ku.edu/~jhuan/BBH/</span></a></span>
<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=IEEE%20Workshop%20on%20BigData%20In%20Bioinformatics%20and%20Health%20Care%20Informatics&amp;author=JL.%20Huan&amp;author=V.%20Pai&amp;author=AM.%20Teredesai&amp;author=S%28Eds%29.%20Yu&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">5.</div><div class="CitationContent" id="CR5">Yuan Q, Nsoesie EO, Lv B, Peng G, Chunara R, Brownstein JS: <strong class="EmphasisTypeBold ">Monitoring influenza epidemics in China with search query from Baidu.</strong> <em class="EmphasisTypeItalic ">PLoS ONE</em> 2013,<strong class="EmphasisTypeBold ">8</strong>(5):e64323. [doi: 10.1371/journal.pone.0064323] [doi: 10.1371/journal.pone.0064323] 10.1371/journal.pone.0064323<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1371/journal.pone.0064323"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Monitoring%20influenza%20epidemics%20in%20China%20with%20search%20query%20from%20Baidu&amp;author=Q.%20Yuan&amp;author=EO.%20Nsoesie&amp;author=B.%20Lv&amp;author=G.%20Peng&amp;author=R.%20Chunara&amp;author=JS.%20Brownstein&amp;journal=PLoS%20ONE&amp;volume=8&amp;issue=5&amp;pages=e64323&amp;publication_year=2013&amp;doi=10.1371%2Fjournal.pone.0064323"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">6.</div><div class="CitationContent" id="CR6">Haferlach
 T, Kohlmann A, Wieczorek L, Basso G, Kronnie GT, Béné MC, De Vos J, 
Hernández JM, Hofmann WK, Mills KI, Gilkes A, Chiaretti S, Shurtleff SA,
 Kipps TJ, Rassenti LZ, Yeoh AE, Papenhausen PR, Wm Liu, Williams PM, Fo
 R: <strong class="EmphasisTypeBold ">Clinical utility of 
microarray-based gene expression profiling in the diagnosis and 
subclassification of leukemia: report from the international microarray 
innovations in leukemia study group.</strong> <em class="EmphasisTypeItalic ">J Clin Oncol</em> 2010,<strong class="EmphasisTypeBold ">28</strong>(15):2529–2537. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://jco.ascopubs.org/content/28/15/2529.abstract"><span class="RefSource">http://jco.ascopubs.org/content/28/15/2529.abstract</span></a></span>] 10.1200/JCO.2009.23.4732<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1200/JCO.2009.23.4732"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Clinical%20utility%20of%20microarray-based%20gene%20expression%20profiling%20in%20the%20diagnosis%20and%20subclassification%20of%20leukemia%3A%20report%20from%20the%20international%20microarray%20innovations%20in%20leukemia%20study%20group&amp;author=T.%20Haferlach&amp;author=A.%20Kohlmann&amp;author=L.%20Wieczorek&amp;author=G.%20Basso&amp;author=GT.%20Kronnie&amp;author=MC.%20B%C3%A9n%C3%A9&amp;author=J.%20De%20Vos&amp;author=JM.%20Hern%C3%A1ndez&amp;author=WK.%20Hofmann&amp;author=KI.%20Mills&amp;author=A.%20Gilkes&amp;author=S.%20Chiaretti&amp;author=SA.%20Shurtleff&amp;author=TJ.%20Kipps&amp;author=LZ.%20Rassenti&amp;author=AE.%20Yeoh&amp;author=PR.%20Papenhausen&amp;author=Liu.%20Wm&amp;author=PM.%20Williams&amp;author=R.%20Fo&amp;journal=J%20Clin%20Oncol&amp;volume=28&amp;issue=15&amp;pages=2529-2537&amp;publication_year=2010&amp;doi=10.1200%2FJCO.2009.23.4732"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">7.</div><div class="CitationContent" id="CR7">Salazar
 R, Roepman P, Capella G, Moreno V, Simon I, Dreezen C, Lopez-Doriga A, 
Santos C, Marijnen C, Westerga J, Bruin S, Kerr D, Kuppen P, van de 
Velde C, Morreau H, Van Velthuysen L, Glas AM, Van’t Veer LJ, Tollenaar 
R: <strong class="EmphasisTypeBold ">Gene expression signature to improve prognosis prediction of stage II and III colorectal cancer.</strong> <em class="EmphasisTypeItalic ">J Clin Oncol</em> 2011, <strong class="EmphasisTypeBold ">29:</strong> 17–24. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://jco.ascopubs.org/content/29/1/17.abstract"><span class="RefSource">http://jco.ascopubs.org/content/29/1/17.abstract</span></a></span>] 10.1200/JCO.2010.30.1077<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1200/JCO.2010.30.1077"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Gene%20expression%20signature%20to%20improve%20prognosis%20prediction%20of%20stage%20II%20and%20III%20colorectal%20cancer&amp;author=R.%20Salazar&amp;author=P.%20Roepman&amp;author=G.%20Capella&amp;author=V.%20Moreno&amp;author=I.%20Simon&amp;author=C.%20Dreezen&amp;author=A.%20Lopez-Doriga&amp;author=C.%20Santos&amp;author=C.%20Marijnen&amp;author=J.%20Westerga&amp;author=S.%20Bruin&amp;author=D.%20Kerr&amp;author=P.%20Kuppen&amp;author=C.%20van%20de%20Velde&amp;author=H.%20Morreau&amp;author=L.%20Van%20Velthuysen&amp;author=AM.%20Glas&amp;author=LJ.%20Van%E2%80%99t%20Veer&amp;journal=J%20Clin%20Oncol&amp;volume=29&amp;pages=17-24&amp;publication_year=2011&amp;doi=10.1200%2FJCO.2010.30.1077"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">8.</div><div class="CitationContent" id="CR8">Annese J: <strong class="EmphasisTypeBold ">The importance of combining MRI and large-scale digital histology in neuroimaging studies of brain connectivity and disease.</strong> <em class="EmphasisTypeItalic ">Front Neuroinform</em> 2012, <strong class="EmphasisTypeBold ">6:</strong> 13. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://europepmc.org/abstract/MED/22536182"><span class="RefSource">http://europepmc.org/abstract/MED/22536182</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.3389/fninf.2012.00013"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20importance%20of%20combining%20MRI%20and%20large-scale%20digital%20histology%20in%20neuroimaging%20studies%20of%20brain%20connectivity%20and%20disease&amp;author=J.%20Annese&amp;journal=Front%20Neuroinform&amp;volume=6&amp;pages=13&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">9.</div><div class="CitationContent" id="CR9">Van Essen DC, Smith SM, Barch DM, Behrens TE, Yacoub E, Ugurbil K: <strong class="EmphasisTypeBold ">The WU-Minn human connectome project: an overview.</strong> <em class="EmphasisTypeItalic ">NeuroImage</em> 2013,<strong class="EmphasisTypeBold ">80</strong>(0):62–79. . [Mapping the Connectome] [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.sciencedirect.com/science/article/pii/S1053811913005351"><span class="RefSource">http://www.sciencedirect.com/science/article/pii/S1053811913005351</span></a></span>]. [Mapping the Connectome]<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/j.neuroimage.2013.05.041"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20WU-Minn%20human%20connectome%20project%3A%20an%20overview&amp;author=DC.%20Van%20Essen&amp;author=SM.%20Smith&amp;author=DM.%20Barch&amp;author=TE.%20Behrens&amp;author=E.%20Yacoub&amp;author=K.%20Ugurbil&amp;journal=NeuroImage&amp;volume=80&amp;issue=0&amp;pages=62-79&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">10.</div><div class="CitationContent" id="CR10">Yoshida H, Kawaguchi A, Tsuruya K: <strong class="EmphasisTypeBold ">Radial basis function-sparse partial least squares for application to brain imaging data.</strong> <em class="EmphasisTypeItalic ">Comput Math Methods Med</em> 2013, <strong class="EmphasisTypeBold ">2013:</strong> 7. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://dx.doi.org/10.1155/2013/591032"><span class="RefSource">http://dx.doi.org/10.1155/2013/591032</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=3056529"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1155/2013/591032"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Radial%20basis%20function-sparse%20partial%20least%20squares%20for%20application%20to%20brain%20imaging%20data&amp;author=H.%20Yoshida&amp;author=A.%20Kawaguchi&amp;author=K.%20Tsuruya&amp;journal=Comput%20Math%20Methods%20Med&amp;volume=2013&amp;pages=7&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">11.</div><div class="CitationContent" id="CR11">Estella F, Delgado-Marquez BL, Rojas P, Valenzuela O, San Roman B, Rojas I: <strong class="EmphasisTypeBold ">Advanced system for automously classify brain MRI in neurodegenerative disease.</strong> In <em class="EmphasisTypeItalic ">International Conference on Multimedia Computing and Systems (ICMCS 2012)</em>. Tangiers, Morocco: IEEE, based in New York, USA; 2012:250–255.<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/ICMCS.2012.6320281"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Advanced%20system%20for%20automously%20classify%20brain%20MRI%20in%20neurodegenerative%20disease&amp;author=F.%20Estella&amp;author=BL.%20Delgado-Marquez&amp;author=P.%20Rojas&amp;author=O.%20Valenzuela&amp;author=B.%20San%20Roman&amp;author=I.%20Rojas&amp;pages=250-255&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">12.</div><div class="CitationContent" id="CR12">Campbell AJ, Cook JA, Adey G, Cuthbertson BH: <strong class="EmphasisTypeBold ">Predicting death and readmission after intensive care discharge.</strong> <em class="EmphasisTypeItalic ">British J Anaesth</em> 2008,<strong class="EmphasisTypeBold ">100</strong>(5):656–662. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://europepmc.org/abstract/MED/18385264"><span class="RefSource">http://europepmc.org/abstract/MED/18385264</span></a></span>] 10.1093/bja/aen069<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1093/bja/aen069"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Predicting%20death%20and%20readmission%20after%20intensive%20care%20discharge&amp;author=AJ.%20Campbell&amp;author=JA.%20Cook&amp;author=G.%20Adey&amp;author=BH.%20Cuthbertson&amp;journal=British%20J%20Anaesth&amp;volume=100&amp;issue=5&amp;pages=656-662&amp;publication_year=2008&amp;doi=10.1093%2Fbja%2Faen069"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">13.</div><div class="CitationContent" id="CR13">Fialho AS, Cismondi F, Vieira SM, Reti SR, Sousa JMC, Finkelstein SN: <strong class="EmphasisTypeBold ">Data mining using clinical physiology at discharge to predict ICU readmissions.</strong> <em class="EmphasisTypeItalic ">Expert Syst Appl</em> 2012,<strong class="EmphasisTypeBold ">39</strong>(18):13158–13165. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.sciencedirect.com/science/article/pii/S0957417412008020"><span class="RefSource">http://www.sciencedirect.com/science/article/pii/S0957417412008020</span></a></span>] 10.1016/j.eswa.2012.05.086<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/j.eswa.2012.05.086"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Data%20mining%20using%20clinical%20physiology%20at%20discharge%20to%20predict%20ICU%20readmissions&amp;author=AS.%20Fialho&amp;author=F.%20Cismondi&amp;author=SM.%20Vieira&amp;author=SR.%20Reti&amp;author=JMC.%20Sousa&amp;author=SN.%20Finkelstein&amp;journal=Expert%20Syst%20Appl&amp;volume=39&amp;issue=18&amp;pages=13158-13165&amp;publication_year=2012&amp;doi=10.1016%2Fj.eswa.2012.05.086"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">14.</div><div class="CitationContent" id="CR14">Ouanes I, Schwebel C, Franais A, Bruel C, Philippart F, Vesin A, Soufir L, Adrie C, Garrouste-Orgeas M, Timsit JF, Misset B: <strong class="EmphasisTypeBold ">A model to predict short-term death or readmission after intensive care unit discharge.</strong> <em class="EmphasisTypeItalic ">J Crit Care</em> 2012,<strong class="EmphasisTypeBold ">27</strong>(4):422.e1–422.e9. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.sciencedirect.com/science/article/pii/S0883944111003790"><span class="RefSource">http://www.sciencedirect.com/science/article/pii/S0883944111003790</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/j.jcrc.2011.08.003"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20model%20to%20predict%20short-term%20death%20or%20readmission%20after%20intensive%20care%20unit%20discharge&amp;author=I.%20Ouanes&amp;author=C.%20Schwebel&amp;author=A.%20Franais&amp;author=C.%20Bruel&amp;author=F.%20Philippart&amp;author=A.%20Vesin&amp;author=L.%20Soufir&amp;author=C.%20Adrie&amp;author=M.%20Garrouste-Orgeas&amp;author=JF.%20Timsit&amp;author=B.%20Misset&amp;journal=J%20Crit%20Care&amp;volume=27&amp;issue=4&amp;pages=422.e1-422.e9&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">15.</div><div class="CitationContent" id="CR15">Mathias JS, Agrawal A, Feinglass J, Cooper AJ, Baker DW, Choudhary A: <strong class="EmphasisTypeBold ">Development of a 5 year life expectancy index in older adults using predictive mining of electronic health record data.</strong> <em class="EmphasisTypeItalic ">J Am Med Inform Assoc</em> 2013,<strong class="EmphasisTypeBold ">20</strong>(e1):e118-e124.
 [ttp://jamia.bmj.com/content/20/e1/e118.abstract] 
[ttp://jamia.bmj.com/content/20/e1/e118.abstract] 
10.1136/amiajnl-2012-001360<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1136/amiajnl-2012-001360"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Development%20of%20a%205%20year%20life%20expectancy%20index%20in%20older%20adults%20using%20predictive%20mining%20of%20electronic%20health%20record%20data&amp;author=JS.%20Mathias&amp;author=A.%20Agrawal&amp;author=J.%20Feinglass&amp;author=AJ.%20Cooper&amp;author=DW.%20Baker&amp;author=A.%20Choudhary&amp;journal=J%20Am%20Med%20Inform%20Assoc&amp;volume=20&amp;issue=e1&amp;pages=e118-e124&amp;publication_year=2013&amp;doi=10.1136%2Famiajnl-2012-001360"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">16.</div><div class="CitationContent" id="CR16">Ballard C, Foster K, Frenkiel A, Gedik B, Koranda MP, Nathan S, Rajan D, Rea R, Spicer M, Williams B, Zoubov VN: <em class="EmphasisTypeItalic ">IBM Infosphere Streams: Assembling Continuous Insight in the Information Revolution</em>. 2011. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.redbooks.ibm.com/abstracts/sg.pages=247970html"><span class="RefSource">http://www.redbooks.ibm.com/abstracts/sg.pages=247970html</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=IBM%20Infosphere%20Streams%3A%20Assembling%20Continuous%20Insight%20in%20the%20Information%20Revolution&amp;author=C.%20Ballard&amp;author=K.%20Foster&amp;author=A.%20Frenkiel&amp;author=B.%20Gedik&amp;author=MP.%20Koranda&amp;author=S.%20Nathan&amp;author=D.%20Rajan&amp;author=R.%20Rea&amp;author=M.%20Spicer&amp;author=B.%20Williams&amp;author=VN.%20Zoubov&amp;publication_year=2011"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">17.</div><div class="CitationContent" id="CR17">Zhang Y, Fong S, Fiaidhi J, Mohammed S: <strong class="EmphasisTypeBold ">Real-time clinical decision support system with data stream mining.</strong> <em class="EmphasisTypeItalic ">J Biomed Biotechnol</em> 2012, <strong class="EmphasisTypeBold ">2012:</strong> 8. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://dx.doi.org/10.1155/2012/580186"><span class="RefSource">http://dx.doi.org/10.1155/2012/580186</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Real-time%20clinical%20decision%20support%20system%20with%20data%20stream%20mining&amp;author=Y.%20Zhang&amp;author=S.%20Fong&amp;author=J.%20Fiaidhi&amp;author=S.%20Mohammed&amp;journal=J%20Biomed%20Biotechnol&amp;volume=2012&amp;pages=8&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">18.</div><div class="CitationContent" id="CR18">Thommandram A, Pugh JE, Eklund JM, McGregor C, James AG: <strong class="EmphasisTypeBold ">Classifying neonatal spells using real-time temporal analysis of physiological data streams: Algorithm development.</strong> In <em class="EmphasisTypeItalic ">IEEE Point-of-Care Healthcare Technologies (PHT 2013)</em>. Bangalore, India: IEEE, based in New York, USA; 2013:240–243.<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/PHT.2013.6461329"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Classifying%20neonatal%20spells%20using%20real-time%20temporal%20analysis%20of%20physiological%20data%20streams%3A%20Algorithm%20development&amp;author=A.%20Thommandram&amp;author=JE.%20Pugh&amp;author=JM.%20Eklund&amp;author=C.%20McGregor&amp;author=AG.%20James&amp;pages=240-243&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">19.</div><div class="CitationContent" id="CR19">Ashish N, Biswas A, Das S, Nag S, Pratap R: <strong class="EmphasisTypeBold ">The Abzooba smart health informatics platform (SHIP)™– from patient experiences to big data to insights.</strong> <em class="EmphasisTypeItalic ">CoRR abs/1203.3764</em> 2012, 1–3.<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20Abzooba%20smart%20health%20informatics%20platform%20%28SHIP%29%E2%84%A2%E2%80%93%20from%20patient%20experiences%20to%20big%20data%20to%20insights&amp;author=N.%20Ashish&amp;author=A.%20Biswas&amp;author=S.%20Das&amp;author=S.%20Nag&amp;author=R.%20Pratap&amp;pages=1-3&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">20.</div><div class="CitationContent" id="CR20">Rolia J, Yao W, Basu S, Lee WN, Singhal S, Kumar A, Sabella S: <strong class="EmphasisTypeBold ">Tell me what i don’t know - making the most of social health forums.</strong> <em class="EmphasisTypeItalic ">Tech. Rep</em> 2013, HPL-2013–43. Hewlett Packard Labs [<span class="ExternalRef"><a target="_blank" rel="noopener" href="https://www.hpl.hp.com/techreports/2013/HPL-2013-43.pdf"><span class="RefSource">https://www.hpl.hp.com/techreports/2013/HPL-2013–43.pdf</span></a></span>] Hewlett Packard Labs<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Tell%20me%20what%20i%20don%E2%80%99t%20know%20-%20making%20the%20most%20of%20social%20health%20forums&amp;author=J.%20Rolia&amp;author=W.%20Yao&amp;author=S.%20Basu&amp;author=WN.%20Lee&amp;author=S.%20Singhal&amp;author=A.%20Kumar&amp;author=S.%20Sabella&amp;pages=HPL-2013%E2%80%9343&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">21.</div><div class="CitationContent" id="CR21">Ginsberg J, Mohebbi MH, Patel RS, Brammer L, Smolinski MS, Brilliant L: <strong class="EmphasisTypeBold ">Detecting influenza epidemics using search engine query data.</strong> <em class="EmphasisTypeItalic ">Nature</em> 2009,<strong class="EmphasisTypeBold ">457</strong>(7232):1012–1014. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://dx.doi.org/10.1038/nature07634"><span class="RefSource">http://dx.doi.org/10.1038/nature07634</span></a></span>] 10.1038/nature07634<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1038/nature07634"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Detecting%20influenza%20epidemics%20using%20search%20engine%20query%20data&amp;author=J.%20Ginsberg&amp;author=MH.%20Mohebbi&amp;author=RS.%20Patel&amp;author=L.%20Brammer&amp;author=MS.%20Smolinski&amp;author=L.%20Brilliant&amp;journal=Nature&amp;volume=457&amp;issue=7232&amp;pages=1012-1014&amp;publication_year=2009&amp;doi=10.1038%2Fnature07634"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">22.</div><div class="CitationContent" id="CR22">Achrekar H, Gandhe A, Lazarus R, Yu SH, Liu B: <strong class="EmphasisTypeBold ">Twitter improves seasonal influenza prediction.</strong> In <em class="EmphasisTypeItalic ">International Conference on Health Informatics (HEALTHINF’12)</em>. Vilamoura, Portugal: Nature Publishing Group, based in London, UK; 2012:61–70.<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Twitter%20improves%20seasonal%20influenza%20prediction&amp;author=H.%20Achrekar&amp;author=A.%20Gandhe&amp;author=R.%20Lazarus&amp;author=SH.%20Yu&amp;author=B.%20Liu&amp;pages=61-70&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">23.</div><div class="CitationContent" id="CR23">Signorini A, Segre AM, Polgreen PM: <strong class="EmphasisTypeBold ">The use of twitter to track levels of disease activity and public concern in the U.S. during the influenza A H1N1 pandemic.</strong> <em class="EmphasisTypeItalic ">PLoS ONE</em> 2011,<strong class="EmphasisTypeBold ">6</strong>(5):e19467. doi:10.1371/journal.pone.0019467 doi:10.1371/journal.pone.0019467 10.1371/journal.pone.0019467<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1371/journal.pone.0019467"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20use%20of%20twitter%20to%20track%20levels%20of%20disease%20activity%20and%20public%20concern%20in%20the%20U.S.%20during%20the%20influenza%20A%20H1N1%20pandemic&amp;author=A.%20Signorini&amp;author=AM.%20Segre&amp;author=PM.%20Polgreen&amp;journal=PLoS%20ONE&amp;volume=6&amp;issue=5&amp;pages=e19467&amp;publication_year=2011"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">24.</div><div class="CitationContent" id="CR24">McDonald E, Brown CT: <em class="EmphasisTypeItalic ">khmer: Working with big data in Bioinformatics</em>. 2013. CoRR abs/1303.2223: 1–18 CoRR abs/1303.2223: 1–18<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=khmer%3A%20Working%20with%20big%20data%20in%20Bioinformatics&amp;author=E.%20McDonald&amp;author=CT.%20Brown&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">25.</div><div class="CitationContent" id="CR25">Bennett C, Doub T: <strong class="EmphasisTypeBold ">Data mining and electronic health records: selecting optimal clinical treatments in practice.</strong> <em class="EmphasisTypeItalic ">CoRR abs/1112</em> 2011, 1668.<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Data%20mining%20and%20electronic%20health%20records%3A%20selecting%20optimal%20clinical%20treatments%20in%20practice&amp;author=C.%20Bennett&amp;author=T.%20Doub&amp;pages=1668&amp;publication_year=2011"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">26.</div><div class="CitationContent" id="CR26">Ertl P: <strong class="EmphasisTypeBold ">Cheminformatics
 analysis of organic substituents: identification of the most common 
substituents, calculation of substituent properties, and automatic 
identification of drug-like bioisosteric groups.</strong> <em class="EmphasisTypeItalic ">J Chem Inform Comput Sci</em> 2003,<strong class="EmphasisTypeBold ">43</strong>(2):374–38. . [PMID: 12653499]. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://pubs.acs.org/doi/abs/10.1021/ci0255782"><span class="RefSource">http://pubs.acs.org/doi/abs/10.1021/ci0255782</span></a></span>]. [PMID: 12653499]. 10.1021/ci0255782<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1021/ci0255782"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Cheminformatics%20analysis%20of%20organic%20substituents%3A%20identification%20of%20the%20most%20common%20substituents%2C%20calculation%20of%20substituent%20properties%2C%20and%20automatic%20identification%20of%20drug-like%20bioisosteric%20groups&amp;author=P.%20Ertl&amp;journal=J%20Chem%20Inform%20Comput%20Sci&amp;volume=43&amp;issue=2&amp;pages=374-38&amp;publication_year=2003&amp;doi=10.1021%2Fci0255782"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">27.</div><div class="CitationContent" id="CR27">Zhang J, Chung TDY, Oldenburg KR: <strong class="EmphasisTypeBold ">Validation of high throughput screening assays.</strong> <em class="EmphasisTypeItalic ">J Biomolecular Screening</em> 1999,<strong class="EmphasisTypeBold ">4</strong>(2):67–73. 10.1177/108705719900400206<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1177/108705719900400206"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Validation%20of%20high%20throughput%20screening%20assays&amp;author=J.%20Zhang&amp;author=TDY.%20Chung&amp;author=KR.%20Oldenburg&amp;journal=J%20Biomolecular%20Screening&amp;volume=4&amp;issue=2&amp;pages=67-73&amp;publication_year=1999&amp;doi=10.1177%2F108705719900400206"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">28.</div><div class="CitationContent" id="CR28">Tamura K, Dudley J, Nei M, Kumar S: <strong class="EmphasisTypeBold ">MEGA4: Molecular evolutionary genetics analysis (MEGA) Software version 4.0.</strong> <em class="EmphasisTypeItalic ">Mol Biol Evol</em> 2007,<strong class="EmphasisTypeBold ">24</strong>(8):1596–1599. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://mbe.oxfordjournals.org/content/24/8/1596.abstract"><span class="RefSource">http://mbe.oxfordjournals.org/content/24/8/1596.abstract</span></a></span>] 10.1093/molbev/msm092<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1093/molbev/msm092"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=MEGA4%3A%20Molecular%20evolutionary%20genetics%20analysis%20%28MEGA%29%20Software%20version%204.0&amp;author=K.%20Tamura&amp;author=J.%20Dudley&amp;author=M.%20Nei&amp;author=S.%20Kumar&amp;journal=Mol%20Biol%20Evol&amp;volume=24&amp;issue=8&amp;pages=1596-1599&amp;publication_year=2007&amp;doi=10.1093%2Fmolbev%2Fmsm092"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">29.</div><div class="CitationContent" id="CR29">Liu W, Li R, Sun JZ, Wang J, Tsai J, Wen W, Kohlmann A, Williams PM: <strong class="EmphasisTypeBold ">PQN and DQN: Algorithms for expression microarrays.</strong> <em class="EmphasisTypeItalic ">J Theor Biol</em> 2006,<strong class="EmphasisTypeBold ">243</strong>(2):273–278. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.sciencedirect.com/science/article/pii/S0022519306002530"><span class="RefSource">http://www.sciencedirect.com/science/article/pii/S0022519306002530</span></a></span>] 10.1016/j.jtbi.2006.06.017<span class="Occurrences"><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=2278026"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/j.jtbi.2006.06.017"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=PQN%20and%20DQN%3A%20Algorithms%20for%20expression%20microarrays&amp;author=W.%20Liu&amp;author=R.%20Li&amp;author=JZ.%20Sun&amp;author=J.%20Wang&amp;author=J.%20Tsai&amp;author=W.%20Wen&amp;author=A.%20Kohlmann&amp;author=PM.%20Williams&amp;journal=J%20Theor%20Biol&amp;volume=243&amp;issue=2&amp;pages=273-278&amp;publication_year=2006&amp;doi=10.1016%2Fj.jtbi.2006.06.017"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">30.</div><div class="CitationContent" id="CR30">Bennett KP, Campbell C: <strong class="EmphasisTypeBold ">Support vector machines: hype or hallelujah.</strong> <em class="EmphasisTypeItalic ">SIGKDD Explor Newslett</em> 2000,<strong class="EmphasisTypeBold ">2</strong>(2):1–13. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://doi.acm.org/10.1145/380995.380999"><span class="RefSource">http://doi.acm.org/10.1145/380995.380999</span></a></span>] 10.1145/380995.380999<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/380995.380999"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Support%20vector%20machines%3A%20hype%20or%20hallelujah&amp;author=KP.%20Bennett&amp;author=C.%20Campbell&amp;journal=SIGKDD%20Explor%20Newslett&amp;volume=2&amp;issue=2&amp;pages=1-13&amp;publication_year=2000&amp;doi=10.1145%2F380995.380999"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">31.</div><div class="CitationContent" id="CR31">Glas
 A, Floore A, Delahaye L, Witteveen A, Pover R, Bakx N, Lahti-Domenici 
J, Bruinsma T, Warmoes M, Bernards R, Wessels L, Van ’t Veer L: <strong class="EmphasisTypeBold ">Converting a breast cancer microarray signature into a high-throughput diagnostic test.</strong> <em class="EmphasisTypeItalic ">BMC Genomics</em> 2006, <strong class="EmphasisTypeBold ">7:</strong> 278. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.biomedcentral.com/1471-2164/7/278"><span class="RefSource">http://www.biomedcentral.com/1471–2164/7/278</span></a></span>] 10.1186/1471-2164-7-278<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1186/1471-2164-7-278"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Converting%20a%20breast%20cancer%20microarray%20signature%20into%20a%20high-throughput%20diagnostic%20test&amp;author=A.%20Glas&amp;author=A.%20Floore&amp;author=L.%20Delahaye&amp;author=A.%20Witteveen&amp;author=R.%20Pover&amp;author=N.%20Bakx&amp;author=J.%20Lahti-Domenici&amp;author=T.%20Bruinsma&amp;author=M.%20Warmoes&amp;author=R.%20Bernards&amp;author=L.%20Wessels&amp;author=L.%20Van%20%E2%80%99t%20Veer&amp;journal=BMC%20Genomics&amp;volume=7&amp;pages=278&amp;publication_year=2006&amp;doi=10.1186%2F1471-2164-7-278"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">32.</div><div class="CitationContent" id="CR32">Knaus WA, Draper EA, Wagner DP, Zimmerman JE: <strong class="EmphasisTypeBold ">APACHE II: A severity of disease classification system.</strong> <em class="EmphasisTypeItalic ">Crit Care Med 00003246–198510000–00009</em> 1985,<strong class="EmphasisTypeBold ">13</strong>(10):818–829. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://journals.lww.com/ccmjournal/Fulltext/1985/10000/APACHE_II__A_severity_of_disease_classification.9.aspx"><span class="RefSource">http://journals.lww.com/ccmjournal/Fulltext/1985/10000/APACHE_II__A_severity_of_disease_classification.9.aspx</span></a></span>] 10.1097/00003246-198510000-00009<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1097/00003246-198510000-00009"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=APACHE%20II%3A%20A%20severity%20of%20disease%20classification%20system&amp;author=WA.%20Knaus&amp;author=EA.%20Draper&amp;author=DP.%20Wagner&amp;author=JE.%20Zimmerman&amp;journal=Crit%20Care%20Med%2000003246-198510000-00009&amp;volume=13&amp;issue=10&amp;pages=818-829&amp;publication_year=1985&amp;doi=10.1097%2F00003246-198510000-00009"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">33.</div><div class="CitationContent" id="CR33">Le Gall J, Lemeshow S, Saulnier F: <strong class="EmphasisTypeBold ">A new simplified acute physiology score (SAPS II) based on a European/North American multicenter study.</strong> <em class="EmphasisTypeItalic ">JAMA</em> 1993,<strong class="EmphasisTypeBold ">270</strong>(24):2957–2963. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://dx.doi.org/10.1001/jama.1993.03510240069035"><span class="RefSource">http://dx.doi.org/10.1001/jama.1993.03510240069035</span></a></span>] 10.1001/jama.1993.03510240069035<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1001/jama.1993.03510240069035"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20new%20simplified%20acute%20physiology%20score%20%28SAPS%20II%29%20based%20on%20a%20European%2FNorth%20American%20multicenter%20study&amp;author=J.%20Le%20Gall&amp;author=S.%20Lemeshow&amp;author=F.%20Saulnier&amp;journal=JAMA&amp;volume=270&amp;issue=24&amp;pages=2957-2963&amp;publication_year=1993&amp;doi=10.1001%2Fjama.1993.03510240069035"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">34.</div><div class="CitationContent" id="CR34">Keene AR, Cullen DJ: <strong class="EmphasisTypeBold ">Therapeutic intervention scoring system: Update 1983.</strong> <em class="EmphasisTypeItalic ">Crit Care Med 00003246–198301000–00001</em> 1983, <strong class="EmphasisTypeBold ">11:</strong> 1–3. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://journals.lww.com/ccmjournal/Fulltext/1983/01000/Therapeutic_Intervention_Scoring_System__Update.1.aspx"><span class="RefSource">http://journals.lww.com/ccmjournal/Fulltext/1983/01000/Therapeutic_Intervention_Scoring_System__Update.1.aspx</span></a></span>] 10.1097/00003246-198301000-00001<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1097/00003246-198301000-00001"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Therapeutic%20intervention%20scoring%20system%3A%20Update%201983&amp;author=AR.%20Keene&amp;author=DJ.%20Cullen&amp;journal=Crit%20Care%20Med%2000003246-198301000-00001&amp;volume=11&amp;pages=1-3&amp;publication_year=1983&amp;doi=10.1097%2F00003246-198301000-00001"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">35.</div><div class="CitationContent" id="CR35">Hosmer DW, Lemesbow S: <strong class="EmphasisTypeBold ">Goodness of fit tests for the multiple logistic regression model.</strong> <em class="EmphasisTypeItalic ">Commun Stat - Theory Methods</em> 1980,<strong class="EmphasisTypeBold ">9</strong>(10):1043–1069. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.tandfonline.com/doi/abs/10.1080/03610928008827941"><span class="RefSource">http://www.tandfonline.com/doi/abs/10.1080/03610928008827941</span></a></span>] 10.1080/03610928008827941<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1080/03610928008827941"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Goodness%20of%20fit%20tests%20for%20the%20multiple%20logistic%20regression%20model&amp;author=DW.%20Hosmer&amp;author=S.%20Lemesbow&amp;journal=Commun%20Stat%20-%20Theory%20Methods&amp;volume=9&amp;issue=10&amp;pages=1043-1069&amp;publication_year=1980&amp;doi=10.1080%2F03610928008827941"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">36.</div><div class="CitationContent" id="CR36">Bradley AP: <strong class="EmphasisTypeBold ">The use of the area under the ROC curve in the evaluation of machine learning algorithms.</strong> <em class="EmphasisTypeItalic ">Pattern Recognit</em> 1997,<strong class="EmphasisTypeBold ">30</strong>(7):1145–1159. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.sciencedirect.com/science/article/pii/S0031320396001422"><span class="RefSource">http://www.sciencedirect.com/science/article/pii/S0031320396001422</span></a></span>] 10.1016/S0031-3203(96)00142-2<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/S0031-3203%2896%2900142-2"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20use%20of%20the%20area%20under%20the%20ROC%20curve%20in%20the%20evaluation%20of%20machine%20learning%20algorithms&amp;author=AP.%20Bradley&amp;journal=Pattern%20Recognit&amp;volume=30&amp;issue=7&amp;pages=1145-1159&amp;publication_year=1997&amp;doi=10.1016%2FS0031-3203%2896%2900142-2"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">37.</div><div class="CitationContent" id="CR37">Akaike H: <strong class="EmphasisTypeBold ">A new look at the statistical model identification.</strong> <em class="EmphasisTypeItalic ">IEEE Trans Automatic Control</em> 1974,<strong class="EmphasisTypeBold ">19</strong>(6):716–723. 10.1109/TAC.1974.1100705<span class="Occurrences"><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=423716"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/TAC.1974.1100705"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20new%20look%20at%20the%20statistical%20model%20identification&amp;author=H.%20Akaike&amp;journal=IEEE%20Trans%20Automatic%20Control&amp;volume=19&amp;issue=6&amp;pages=716-723&amp;publication_year=1974&amp;doi=10.1109%2FTAC.1974.1100705"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">38.</div><div class="CitationContent" id="CR38">Gajic O, Malinchoc M, Comfere TB, Harris MR, Achouiti A, Yilmaz M, Schultz MJ, Hubmayr RD, Afessa B, Farmer JC: <strong class="EmphasisTypeBold ">The
 stability and workload index for transfer score predicts unplanned 
intensive care unit patient readmission: initial development and 
validation *.</strong> <em class="EmphasisTypeItalic ">Crit Care Med</em> 2008,<strong class="EmphasisTypeBold ">36</strong>(3):676–682. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://journals.lww.com/ccmjournal/Fulltext/2008/03000/The_Stability_and_Workload_Index_for_Transfer.2.aspx"><span class="RefSource">http://journals.lww.com/ccmjournal/Fulltext/2008/03000/The_Stability_and_Workload_Index_for_Transfer.2.aspx</span></a></span>] 10.1097/CCM.0B013E318164E3B0<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1097/CCM.0B013E318164E3B0"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20stability%20and%20workload%20index%20for%20transfer%20score%20predicts%20unplanned%20intensive%20care%20unit%20patient%20readmission%3A%20initial%20development%20and%20validation%20%2A&amp;author=O.%20Gajic&amp;author=M.%20Malinchoc&amp;author=TB.%20Comfere&amp;author=MR.%20Harris&amp;author=A.%20Achouiti&amp;author=M.%20Yilmaz&amp;author=MJ.%20Schultz&amp;author=RD.%20Hubmayr&amp;author=B.%20Afessa&amp;author=JC.%20Farmer&amp;journal=Crit%20Care%20Med&amp;volume=36&amp;issue=3&amp;pages=676-682&amp;publication_year=2008&amp;doi=10.1097%2FCCM.0B013E318164E3B0"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">39.</div><div class="CitationContent" id="CR39">Saeed M, Lieu C, Raber G, Mark R: <strong class="EmphasisTypeBold ">MIMIC II: a massive temporal ICU patient database to support research in intelligent patient monitoring.</strong> In <em class="EmphasisTypeItalic ">Computers in Cardiology</em>. Memphis, Tennesee, USA: IEEE Computer Society, based in California, USA; 2002:641–644.<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/CIC.2002.1166854"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=MIMIC%20II%3A%20a%20massive%20temporal%20ICU%20patient%20database%20to%20support%20research%20in%20intelligent%20patient%20monitoring&amp;author=M.%20Saeed&amp;author=C.%20Lieu&amp;author=G.%20Raber&amp;author=R.%20Mark&amp;pages=641-644&amp;publication_year=2002"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">40.</div><div class="CitationContent" id="CR40">Mendonça LF, Vieira SM, Sousa JMC: <strong class="EmphasisTypeBold ">Decision tree search methods in fuzzy modeling and classification.</strong> <em class="EmphasisTypeItalic ">Int J Approximate Reason</em> 2007,<strong class="EmphasisTypeBold ">44</strong>(2):106–123. . [Fuzzy Decision-Making Applications] [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.sciencedirect.com/science/article/pii/S0888613X06000843"><span class="RefSource">http://www.sciencedirect.com/science/article/pii/S0888613X06000843</span></a></span>]. [Fuzzy Decision-Making Applications] 10.1016/j.ijar.2006.07.004<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/j.ijar.2006.07.004"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Decision%20tree%20search%20methods%20in%20fuzzy%20modeling%20and%20classification&amp;author=LF.%20Mendon%C3%A7a&amp;author=SM.%20Vieira&amp;author=JMC.%20Sousa&amp;journal=Int%20J%20Approximate%20Reason&amp;volume=44&amp;issue=2&amp;pages=106-123&amp;publication_year=2007&amp;doi=10.1016%2Fj.ijar.2006.07.004"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">41.</div><div class="CitationContent" id="CR41">Takagi T, Sugeno M: <strong class="EmphasisTypeBold ">Fuzzy identification of systems and its applications to modeling and control.</strong> <em class="EmphasisTypeItalic ">IEEE Trans Syst Man Cybernet SMC-15</em> 1985, 116–132.<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Fuzzy%20identification%20of%20systems%20and%20its%20applications%20to%20modeling%20and%20control&amp;author=T.%20Takagi&amp;author=M.%20Sugeno&amp;pages=116-132&amp;publication_year=1985"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">42.</div><div class="CitationContent" id="CR42">Knaus WA, Wagner DP, Draper EA, Zimmerman JE, Bergner M, Bastos PG, Sirio CA, Murphy DJ, Lotring T, Damiano A: <strong class="EmphasisTypeBold ">The APACHE III prognostic system. risk prediction of hospital mortality for critically ill hospitalized adults.</strong> <em class="EmphasisTypeItalic ">CHEST Journal</em> 1991,<strong class="EmphasisTypeBold ">100</strong>(6):1619–1636. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://dx.doi.org/10.1378/chest.100.6.1619"><span class="RefSource">http://dx.doi.org/10.1378/chest.100.6.1619</span></a></span>] 10.1378/chest.100.6.1619<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1378/chest.100.6.1619"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20APACHE%20III%20prognostic%20system.%20risk%20prediction%20of%20hospital%20mortality%20for%20critically%20ill%20hospitalized%20adults&amp;author=WA.%20Knaus&amp;author=DP.%20Wagner&amp;author=EA.%20Draper&amp;author=JE.%20Zimmerman&amp;author=M.%20Bergner&amp;author=PG.%20Bastos&amp;author=CA.%20Sirio&amp;author=DJ.%20Murphy&amp;author=T.%20Lotring&amp;author=A.%20Damiano&amp;journal=CHEST%20Journal&amp;volume=100&amp;issue=6&amp;pages=1619-1636&amp;publication_year=1991&amp;doi=10.1378%2Fchest.100.6.1619"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">43.</div><div class="CitationContent" id="CR43">Hall M: <em class="EmphasisTypeItalic ">Correlation-based feature selection for machine learning. PhD thesis</em>. Hamilton, New Zealand: The University of Waikato; 1997.<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Correlation-based%20feature%20selection%20for%20machine%20learning.%20PhD%20thesis&amp;author=M.%20Hall&amp;publication_year=1997"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">44.</div><div class="CitationContent" id="CR44">Rodriguez JJ, Kuncheva LI, Alonso CJ: <strong class="EmphasisTypeBold ">Rotation forest: a new classifier ensemble method.</strong> <em class="EmphasisTypeItalic ">IEEE Trans Pattern Anal Mach Intell</em> 2006,<strong class="EmphasisTypeBold ">28</strong>(10):1619–1630.<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/TPAMI.2006.211"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Rotation%20forest%3A%20a%20new%20classifier%20ensemble%20method&amp;author=JJ.%20Rodriguez&amp;author=LI.%20Kuncheva&amp;author=CJ.%20Alonso&amp;journal=IEEE%20Trans%20Pattern%20Anal%20Mach%20Intell&amp;volume=28&amp;issue=10&amp;pages=1619-1630&amp;publication_year=2006"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">45.</div><div class="CitationContent" id="CR45">Freund Y, Mason L: <strong class="EmphasisTypeBold ">The alternating decision tree learning algorithm.</strong> In <em class="EmphasisTypeItalic ">Proceedings of the Sixteenth International Conference on Machine Learning, ICML ’99</em>. San Francisco, CA, USA: Morgan Kaufmann Publishers Inc; 1999:124–133. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://dl.acm.org/citation.cfm?id=.pages=645528657623"><span class="RefSource">http://dl.acm.org/citation.cfm?id=.pages=645528657623</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20alternating%20decision%20tree%20learning%20algorithm&amp;author=Y.%20Freund&amp;author=L.%20Mason&amp;pages=124-133&amp;publication_year=1999"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">46.</div><div class="CitationContent" id="CR46">Perkins AJ, Kroenke K, Unützer J, Katon W, Williams JW, Hope C, Callahan CM: <strong class="EmphasisTypeBold ">Common comorbidity scales were similar in their ability to predict health care costs and mortality.</strong> <em class="EmphasisTypeItalic ">J Clin Epidemiol</em> 2004,<strong class="EmphasisTypeBold ">57</strong>(10):1040–1048. [<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.sciencedirect.com/science/article/pii/S0895435604000812"><span class="RefSource">http://www.sciencedirect.com/science/article/pii/S0895435604000812</span></a></span>] 10.1016/j.jclinepi.2004.03.002<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/j.jclinepi.2004.03.002"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Common%20comorbidity%20scales%20were%20similar%20in%20their%20ability%20to%20predict%20health%20care%20costs%20and%20mortality&amp;author=AJ.%20Perkins&amp;author=K.%20Kroenke&amp;author=J.%20Un%C3%BCtzer&amp;author=W.%20Katon&amp;author=JW.%20Williams&amp;author=C.%20Hope&amp;author=CM.%20Callahan&amp;journal=J%20Clin%20Epidemiol&amp;volume=57&amp;issue=10&amp;pages=1040-1048&amp;publication_year=2004&amp;doi=10.1016%2Fj.jclinepi.2004.03.002"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">47.</div><div class="CitationContent" id="CR47">Walter LC, Covinsky KE: <strong class="EmphasisTypeBold ">Cancer screening in elderly patients: A framework for individualized decision making.</strong> <em class="EmphasisTypeItalic ">JAMA</em> 2001,<strong class="EmphasisTypeBold ">285</strong>(21):2750–2756. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://dx.doi.org/10.1001/jama.285.21.2750"><span class="RefSource">http://dx.doi.org/10.1001/jama.285.21.2750</span></a></span>] [
] 10.1001/jama.285.21.2750<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1001/jama.285.21.2750"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Cancer%20screening%20in%20elderly%20patients%3A%20A%20framework%20for%20individualized%20decision%20making&amp;author=LC.%20Walter&amp;author=KE.%20Covinsky&amp;journal=JAMA&amp;volume=285&amp;issue=21&amp;pages=2750-2756&amp;publication_year=2001&amp;doi=10.1001%2Fjama.285.21.2750"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">48.</div><div class="CitationContent" id="CR48">Domingos P, Hulten G: <strong class="EmphasisTypeBold ">Mining high-speed data streams.</strong> In <em class="EmphasisTypeItalic ">Proceedings of the Sixth ACM SIGKDD International Conference on Knowledge Discovery and Data Mining, KDD ’00</em>. New York, NY, USA: ACM; 2000:71–80. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://doi.acm.org/10.1145/347090.347107"><span class="RefSource">http://doi.acm.org/10.1145/347090.347107</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/347090.347107"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Mining%20high-speed%20data%20streams&amp;author=P.%20Domingos&amp;author=G.%20Hulten&amp;pages=71-80&amp;publication_year=2000"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">49.</div><div class="CitationContent" id="CR49">Achananuparp P, Hu X, Shen X: <strong class="EmphasisTypeBold ">The evaluation of sentence similarity measures.</strong> In <em class="EmphasisTypeItalic ">Data Warehousing and, Knowledge Discovery, Volume 5182 of Lecture Notes in Computer Science</em>. Edited by: Song IY, Eder J, Nguyen T. Springer Berlin, Heidelberg; 2008:305–316. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://dx.doi.org/10.1007/978-3-540-85836-2_29"><span class="RefSource">http://dx.doi.org/10.1007/978–3-540–85836–2_29</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20evaluation%20of%20sentence%20similarity%20measures&amp;author=P.%20Achananuparp&amp;author=X.%20Hu&amp;author=X.%20Shen&amp;pages=305-316&amp;publication_year=2008"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">50.</div><div class="CitationContent" id="CR50">Thiagarajan R, Manjunath G, Stumptne M: <strong class="EmphasisTypeBold ">Computing semantic similarity using Ontologies.</strong> <em class="EmphasisTypeItalic ">Tech. Rep.</em> 2008, HPL-2008–87. Hewlett Packard Labs [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.hpl.hp.com/techreports/2008/HPL-2008-87.pdf"><span class="RefSource">http://www.hpl.hp.com/techreports/2008/HPL-2008–87.pdf</span></a></span>] Hewlett Packard Labs<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Computing%20semantic%20similarity%20using%20Ontologies&amp;author=R.%20Thiagarajan&amp;author=G.%20Manjunath&amp;author=M.%20Stumptne&amp;pages=HPL-2008%E2%80%9387&amp;publication_year=2008"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">51.</div><div class="CitationContent" id="CR51">Sun J, Sow D, Hu J, Ebadollahi S: <strong class="EmphasisTypeBold ">A system for mining temporal physiological data streams for advanced prognostic decision support.</strong> <em class="EmphasisTypeItalic ">IEEE 10th International Conference on Data Mining (ICDM 2010)</em> 2010, 1061–1066. doi:10.1109/ICDM.2010.102 doi:10.1109/ICDM.2010.102<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/ICDM.2010.102"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20system%20for%20mining%20temporal%20physiological%20data%20streams%20for%20advanced%20prognostic%20decision%20support&amp;author=J.%20Sun&amp;author=D.%20Sow&amp;author=J.%20Hu&amp;author=S.%20Ebadollahi&amp;pages=1061-1066&amp;publication_year=2010"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">52.</div><div class="CitationContent" id="CR52">Hay SI, George DB, Moyes CL, Brownstein JS: <strong class="EmphasisTypeBold ">Big data opportunities for global infectious disease surveillance.</strong> <em class="EmphasisTypeItalic ">PLoS Med</em> 2013,<strong class="EmphasisTypeBold ">10</strong>(4):e1001413. doi:10.1371/journal.pmed.1001413 doi:10.1371/journal.pmed.1001413 10.1371/journal.pmed.1001413<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1371/journal.pmed.1001413"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Big%20data%20opportunities%20for%20global%20infectious%20disease%20surveillance&amp;author=SI.%20Hay&amp;author=DB.%20George&amp;author=CL.%20Moyes&amp;author=JS.%20Brownstein&amp;journal=PLoS%20Med&amp;volume=10&amp;issue=4&amp;pages=e1001413&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">53.</div><div class="CitationContent" id="CR53">Ashish N, Mehrotra S: <strong class="EmphasisTypeBold ">XAR An integrated framework for semantic extraction and annotation.</strong> In <em class="EmphasisTypeItalic ">Cases on Semantic Interoperability for Information Systems Integration: Practices and Applications</em>. Edited by: IGI Global, Kalfoglou Y, IGI Global . Hershey, PA, USA; 2009:235–254. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://services.igi-global.com/resolvedoi/resolve.aspx?doi=10.4018/978-1-60566-894-9.ch011"><span class="RefSource">http://services.igi-global.com/resolvedoi/resolve.aspx?doi=10.4018/978–1-60566–894–9.ch011</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=XAR%20An%20integrated%20framework%20for%20semantic%20extraction%20and%20annotation&amp;author=N.%20Ashish&amp;author=S.%20Mehrotra&amp;pages=235-254&amp;publication_year=2009"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">54.</div><div class="CitationContent" id="CR54">Bodenreider O: <strong class="EmphasisTypeBold ">The Unified Medical Language System (UMLS): integrating biomedical terminology.</strong> <em class="EmphasisTypeItalic ">Nucleic Acids Res</em> 2004,<strong class="EmphasisTypeBold ">32</strong>(suppl 1):D267–270. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://nar.oxfordjournals.org/content/32/suppl_1/D267.abstract"><span class="RefSource">http://nar.oxfordjournals.org/content/32/suppl_1/D267.abstract</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1093/nar/gkh061"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20Unified%20Medical%20Language%20System%20%28UMLS%29%3A%20integrating%20biomedical%20terminology&amp;author=O.%20Bodenreider&amp;journal=Nucleic%20Acids%20Res&amp;volume=32&amp;issue=suppl%201&amp;pages=D267-270&amp;publication_year=2004"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">55.</div><div class="CitationContent" id="CR55">Hall MA, Frank E, Holmes G, Pfahringer B, Reutemann P, Witten IH: <strong class="EmphasisTypeBold ">The WEKA data mining software: An update.</strong> <em class="EmphasisTypeItalic ">SIGKDD Explor Newslett</em> 2009, <strong class="EmphasisTypeBold ">11:</strong> 10–18. 10.1145/1656274.1656278<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/1656274.1656278"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20WEKA%20data%20mining%20software%3A%20An%20update&amp;author=MA.%20Hall&amp;author=E.%20Frank&amp;author=G.%20Holmes&amp;author=B.%20Pfahringer&amp;author=P.%20Reutemann&amp;author=IH.%20Witten&amp;journal=SIGKDD%20Explor%20Newslett&amp;volume=11&amp;pages=10-18&amp;publication_year=2009&amp;doi=10.1145%2F1656274.1656278"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">56.</div><div class="CitationContent" id="CR56">The Apache Software Foundation: <strong class="EmphasisTypeBold ">Apache Lucene.</strong> 2013.<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://lucene.apache.org/"><span class="RefSource">http://lucene.apache.org/</span></a></span>. [Accessed: 2013-9-18]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">57.</div><div class="CitationContent" id="CR57">OSI Pharmaceuticals: <strong class="EmphasisTypeBold ">Tarceva®;(erlotinib) tablets advanced-stage non-small cell lung cancer treatment possible risks and side effects.</strong> 2013.<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.tarceva.com/patient/considering/effects.jsp"><span class="RefSource">http://www.tarceva.com/patient/considering/effects.jsp</span></a></span>. [Accessed: 2013-9-18]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">58.</div><div class="CitationContent" id="CR58">Centers for Disease Control and Prevention: <em class="EmphasisTypeItalic ">Diabetes report card 2012. Tech. rep</em>. Atlanta, GA: Centers for Disease Control and Prevention, US Department of Health and Human Services; 2012. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.cdc.gov/diabetes/pubs/pdf/diabetesreportcard.pdf"><span class="RefSource">http://www.cdc.gov/diabetes/pubs/pdf/diabetesreportcard.pdf</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Diabetes%20report%20card%202012.%20Tech.%20rep&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">59.</div><div class="CitationContent" id="CR59">National Institute for HealthandCareExcellence: <strong class="EmphasisTypeBold ">NICE pathways.</strong> 2013.<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://pathways.nice.org.uk/"><span class="RefSource">http://pathways.nice.org.uk/</span></a></span>. [Accessed: 2013-9-18]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?&amp;author=HealthandCareExcellence.%20National%20Institute%20for&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">60.</div><div class="CitationContent" id="CR60">Boehm EA: <strong class="EmphasisTypeBold ">The contribution of economic indicator analysis to understanding and forecasting business cycles.</strong> <em class="EmphasisTypeItalic ">Ind Econ Rev</em> 2001, <strong class="EmphasisTypeBold ">36:</strong> 1–36. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.jstor.org/stable/29794223"><span class="RefSource">http://www.jstor.org/stable/29794223</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20contribution%20of%20economic%20indicator%20analysis%20to%20understanding%20and%20forecasting%20business%20cycles&amp;author=EA.%20Boehm&amp;journal=Ind%20Econ%20Rev&amp;volume=36&amp;pages=1-36&amp;publication_year=2001"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">61.</div><div class="CitationContent" id="CR61">Moore GH, Shiskin J: <strong class="EmphasisTypeBold ">Indicators of Business Expansions and Contractions.</strong> <em class="EmphasisTypeItalic ">National Bureau of, Economic Research</em> 1967. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://papers.nber.org/books/moor67-2"><span class="RefSource">http://papers.nber.org/books/moor67–2</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Indicators%20of%20Business%20Expansions%20and%20Contractions&amp;author=GH.%20Moore&amp;author=J.%20Shiskin&amp;publication_year=1967"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">62.</div><div class="CitationContent" id="CR62">Liu Y, Lv B, Peng G, Yuan Q: <strong class="EmphasisTypeBold ">A preprocessing method of internet search data for prediction improvement: application to Chinese stock market.</strong> In <em class="EmphasisTypeItalic ">Proceedings of the Data Mining and Intelligent Knowledge Management Workshop, DM-IKM ’12</em>. New York, NY, USA: ACM; 2012:3:1–3:7. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://doi.acm.org/10.1145/2462130.2462133"><span class="RefSource">http://doi.acm.org/10.1145/2462130.2462133</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20preprocessing%20method%20of%20internet%20search%20data%20for%20prediction%20improvement%3A%20application%20to%20Chinese%20stock%20market&amp;author=Y.%20Liu&amp;author=B.%20Lv&amp;author=G.%20Peng&amp;author=Q.%20Yuan&amp;pages=3%3A1-3%3A7&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">63.</div><div class="CitationContent" id="CR63">Statistic Brain ResearchInstitute publishing as Statistic Brain: <strong class="EmphasisTypeBold ">Twitter statistics – statistic brain.</strong> 2013.<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.statisticbrain.com/twitter-statistics/"><span class="RefSource">http://www.statisticbrain.com/twitter-statistics/</span></a></span>. [Accessed: 2013-9-18]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?&amp;author=ResearchInstitute.%20Statistic%20Brain&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">64.</div><div class="CitationContent" id="CR64">Twitter Inc: <strong class="EmphasisTypeBold ">The streaming APIs.</strong> 2013.<span class="ExternalRef"><a target="_blank" rel="noopener" href="https://dev.twitter.com/docs/streaming-apis"><span class="RefSource">https://dev.twitter.com/docs/streaming-apis</span></a></span>. [Accessed: 2013-9-18]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?&amp;author=Inc.%20Twitter&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">65.</div><div class="CitationContent" id="CR65">van Rijsbergen CJ, Robertson SE, Porter MF: <em class="EmphasisTypeItalic ">New Models in Probabilistic Information Retrieval</em>. British Library research &amp; development reports, Computer Laboratory, University of Cambridge; 1980. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://books.google.com/books?id=WDZ3bwAACAAJ"><span class="RefSource">http://books.google.com/books?id=WDZ3bwAACAAJ</span></a></span>]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=New%20Models%20in%20Probabilistic%20Information%20Retrieval&amp;author=CJ.%20van%20Rijsbergen&amp;author=SE.%20Robertson&amp;author=MF.%20Porter&amp;publication_year=1980"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">66.</div><div class="CitationContent" id="CR66">Drucker H, Burges CJC, Kaufman L, Smola A, Vapnik V: <strong class="EmphasisTypeBold ">Support vector regression machines.</strong> In <em class="EmphasisTypeItalic ">Advances in neural information processing systems</em>. Edited by: Mozer MC, Jordan MI, Petsche T. Cambridge, MA: MIT Press; 1997:155–161.<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Support%20vector%20regression%20machines&amp;author=H.%20Drucker&amp;author=CJC.%20Burges&amp;author=L.%20Kaufman&amp;author=A.%20Smola&amp;author=V.%20Vapnik&amp;pages=155-161&amp;publication_year=1997"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">67.</div><div class="CitationContent" id="CR67">Doan S, Ohno-Machado L, Collier N: <strong class="EmphasisTypeBold ">Enhancing twitter data analysis with simple semantic filtering: example in tracking influenza-like illnesses.</strong> 2012.<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?&amp;author=S.%20Doan&amp;author=L.%20Ohno-Machado&amp;author=N.%20Collier&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">68.</div><div class="CitationContent" id="CR68">American Medical Informatics Association: <strong class="EmphasisTypeBold ">Translational Bioinformatics.</strong> 2013.<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.amia.org/applications-informatics/translational-bioinformatics"><span class="RefSource">http://www.amia.org/applications-informatics/translational-bioinformatics</span></a></span>. [Accessed: 2013-9-18]<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">69.</div><div class="CitationContent" id="CR69">Sarkar IN, Butte AJ, Lussier YA, Tarczy-Hornoch P, Ohno-Machado L: <strong class="EmphasisTypeBold ">Translational bioinformatics: linking knowledge across biological and clinical realms.</strong> <em class="EmphasisTypeItalic ">J Am Med Inform Assoc</em> 2011,<strong class="EmphasisTypeBold ">18</strong>(4):354–357. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://jamia.bmj.com/content/18/4/354.abstract"><span class="RefSource">http://jamia.bmj.com/content/18/4/354.abstract</span></a></span>] 10.1136/amiajnl-2011-000245<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1136/amiajnl-2011-000245"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Translational%20bioinformatics%3A%20linking%20knowledge%20across%20biological%20and%20clinical%20realms&amp;author=IN.%20Sarkar&amp;author=AJ.%20Butte&amp;author=YA.%20Lussier&amp;author=P.%20Tarczy-Hornoch&amp;author=L.%20Ohno-Machado&amp;journal=J%20Am%20Med%20Inform%20Assoc&amp;volume=18&amp;issue=4&amp;pages=354-357&amp;publication_year=2011&amp;doi=10.1136%2Famiajnl-2011-000245"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">70.</div><div class="CitationContent" id="CR70">Butte AJ, Shah NH: <strong class="EmphasisTypeBold ">Computationally
 translating molecular discoveries into tools for medicine: 
translational bioinformatics articles now featured in JAMIA.</strong> <em class="EmphasisTypeItalic ">J Am Med Inform Assoc</em> 2011,<strong class="EmphasisTypeBold ">18</strong>(4):352–353. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://jamia.bmj.com/content/18/4/352.short"><span class="RefSource">http://jamia.bmj.com/content/18/4/352.short</span></a></span>] 10.1136/amiajnl-2011-000343<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1136/amiajnl-2011-000343"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Computationally%20translating%20molecular%20discoveries%20into%20tools%20for%20medicine%3A%20translational%20bioinformatics%20articles%20now%20featured%20in%20JAMIA&amp;author=AJ.%20Butte&amp;author=NH.%20Shah&amp;journal=J%20Am%20Med%20Inform%20Assoc&amp;volume=18&amp;issue=4&amp;pages=352-353&amp;publication_year=2011&amp;doi=10.1136%2Famiajnl-2011-000343"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">71.</div><div class="CitationContent" id="CR71">Shah NH, Tenenbaum JD: <strong class="EmphasisTypeBold ">The coming age of data-driven medicine: translational bioinformatics’ next frontier.</strong> <em class="EmphasisTypeItalic ">J Am Med Inform Assoc</em> 2012,<strong class="EmphasisTypeBold ">19</strong>(e1):e2-e4. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://jamia.bmj.com/content/19/e1/e2.short"><span class="RefSource">http://jamia.bmj.com/content/19/e1/e2.short</span></a></span>] 10.1136/amiajnl-2012-000969<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1136/amiajnl-2012-000969"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20coming%20age%20of%20data-driven%20medicine%3A%20translational%20bioinformatics%E2%80%99%20next%20frontier&amp;author=NH.%20Shah&amp;author=JD.%20Tenenbaum&amp;journal=J%20Am%20Med%20Inform%20Assoc&amp;volume=19&amp;issue=e1&amp;pages=e2-e4&amp;publication_year=2012&amp;doi=10.1136%2Famiajnl-2012-000969"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationNumber">72.</div><div class="CitationContent" id="CR72">Liu M, Wu Y, Chen Y, Sun J, Zhao Z, Xw Chen, Matheny ME, Xu H: <strong class="EmphasisTypeBold ">Large-scale prediction of adverse drug reactions using chemical, biological, and phenotypic properties of drugs.</strong> <em class="EmphasisTypeItalic ">J Am Med Inform Assoc</em> 2012,<strong class="EmphasisTypeBold ">19</strong>(e1):e28-e35. [
<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://jamia.bmj.com/content/19/e1/e28.abstract"><span class="RefSource">http://jamia.bmj.com/content/19/e1/e28.abstract</span></a></span>] 10.1136/amiajnl-2011-000699<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1136/amiajnl-2011-000699"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Large-scale%20prediction%20of%20adverse%20drug%20reactions%20using%20chemical%2C%20biological%2C%20and%20phenotypic%20properties%20of%20drugs&amp;author=M.%20Liu&amp;author=Y.%20Wu&amp;author=Y.%20Chen&amp;author=J.%20Sun&amp;author=Z.%20Zhao&amp;author=Chen.%20Xw&amp;author=ME.%20Matheny&amp;author=H.%20Xu&amp;journal=J%20Am%20Med%20Inform%20Assoc&amp;volume=19&amp;issue=e1&amp;pages=e28-e35&amp;publication_year=2012&amp;doi=10.1136%2Famiajnl-2011-000699"><span><span>Google Scholar</span></span></a></span></span></div></li></ol></div></section><section class="Section1 RenderAsSection1"><h2 class="Heading" id="copyrightInformation" data-role="collapsible-handle" tabindex="0">Copyright information<span class="section-icon"></span></h2><div class="ArticleCopyright content"><div class="ArticleCopyright">©&nbsp;Herland et al.; licensee Springer.&nbsp;2014<div class="CopyrightComment"><p class="SimplePara">This
 article is published under license to BioMed Central Ltd. This is an 
Open Access article distributed under the terms of the Creative Commons 
Attribution License (<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://creativecommons.org/licenses/by/2.0"><span class="RefSource">http://creativecommons.org/licenses/by/2.0</span></a></span>),
 which permits unrestricted use, distribution, and reproduction in any 
medium, provided the original work is properly credited.</p></div></div></div></section></div>
                        </article>
                        <aside class="section section--collapsible" id="AboutThisContent">
    <h2 class="section__heading" id="aboutcontent" data-role="collapsible-handle" tabindex="0">About this article<span class="section-icon"></span></h2>
    <div class="section__content bibliographic-information">
                <div id="crossMark" class="crossmark">
            <a data-crossmark="10.1186%2F2196-1115-1-2" class="gtm-crossmark" target="_blank" rel="noopener" href="https://crossmark.crossref.org/dialog/?doi=10.1186%2F2196-1115-1-2" title="Verify currency and authenticity via CrossMark">
                <span class="u-screenreader-only">CrossMark</span>
                <svg class="CrossMark" id="crossmark-icon" width="57" height="81">
                    <image width="57" height="81" alt="Verify currency and authenticity via CrossMark" src="/springerlink-static/481091012/images/png/crossmark.png" xlink="http://www.w3.org/1999/xlink" xlink:href="/springerlink-static/481091012/images/svg/crossmark.svg"></image>
                </svg>
            </a>
        </div>

        <div class="crossmark__adjacent">
            <dl class="citation-info u-highlight-target u-mb-16" id="citeas">
    <dt class="test-cite-heading">
        Cite this article as:
    </dt>
    <dd id="citethis-text">Herland, M., Khoshgoftaar, T.M. &amp; Wald, R. Journal of Big Data (2014) 1: 2. https://doi.org/10.1186/2196-1115-1-2</dd>
</dl>
                <ul class="bibliographic-information__list bibliographic-information__list--inline">
        <li class="bibliographic-information__item">
            <span class="bibliographic-information__title"><abbr title="Digital Object Identifier">DOI</abbr></span>
            <span class="bibliographic-information__value u-overflow-wrap" id="doi-url">https://doi.org/10.1186/2196-1115-1-2</span>
        </li>
            <li class="bibliographic-information__item">
                <span class="bibliographic-information__title">Publisher Name</span>
                <span class="bibliographic-information__value" id="publisher-name">Springer International Publishing</span>
            </li>
            <li class="bibliographic-information__item ">
                <span class="bibliographic-information__title">Online ISSN</span>
                <span class="bibliographic-information__value" id="electronic-issn">2196-1115</span>
            </li>

        
    </ul>

            <ul class="bibliographic-information__list">
        <li class="bibliographic-information__item">
            <a id="about-journal" class="bibliographic-information__misc-links gtm-about-this" title="Visit Springer.com for information about this article's journal" href="https://www.springer.com/journal/40537/about">About this journal</a>
        </li>
</ul>

        </div>
      
      
          
    </div>
</aside>

                        <div class="section section--collapsible uptodate-recommendations gtm-recommendations">
    <h2 class="uptodate-recommendations__title section__heading gtm-recommendations__title" id="uptodaterecommendations" data-role="collapsible-handle" tabindex="0">Personalised recommendations<span class="section-icon"></span></h2>
    <div class="section__content">
        <div class="uptodate-recommendations__container">
             <iframe style="width: 0px; height: 0px; bottom: 0px; right: 0px; z-index: 100000;" scrolling="no" src="https://recommended.springernature.com/v2197/generated/client.html?source=https://link.springer.com" name="uptodate-client" id="uptodate-client" title="Personalised recommendations" allowtransparency="true" frameborder="0"></iframe>
        </div>
    </div>
</div>
                                    <div class="sticky-banner u-interface u-js-screenreader-only" aria-hidden="true" data-component="SpringerLink.StickyBanner" data-namespace="hasButton">
                <div class="sticky-banner__container"><span class="sticky-banner__title sticky-banner__title--short u-overflow-ellipsis" title="A review of data mining using big data in health informatics">A review of data mining using big data in health informatics</span>
                        <div class="citations c-button-dropdown" data-component="SV.Dropdown" data-namespace="citationsSticky" aria-label="button with dropdown options" style="">
        <button type="button" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown.citationsSticky-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">
    <span>Cite</span>
    <span class="hide-text-small">article</span>
</span><span class="c-button-dropdown__icon"></span></button>
<div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown.citationsSticky-dropdown"><ul class="citations__content" data-role="button-dropdown__content">
    <li>
        <a href="#citeas" class="gtm-cite-dropdown">How to cite?</a>
    </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1186/2196-1115-1-2?format=refman&amp;flavour=citation" title="Download this article's citation as a .RIS file" class="gtm-export-citation" data-gtmlabel="RIS">
                <span class="citations__extension" data-gtmlabel="RIS">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .RIS
                </span>
                <span class="citations__types">
                        <span>
                            Papers
                        </span>
                        <span>
                            Reference Manager
                        </span>
                        <span>
                            RefWorks
                        </span>
                        <span>
                            Zotero
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1186/2196-1115-1-2?format=endnote&amp;flavour=citation" title="Download this article's citation as a .ENW file" class="gtm-export-citation" data-gtmlabel="ENW">
                <span class="citations__extension" data-gtmlabel="ENW">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .ENW
                </span>
                <span class="citations__types">
                        <span>
                            EndNote
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1186/2196-1115-1-2?format=bibtex&amp;flavour=citation" title="Download this article's citation as a .BIB file" class="gtm-export-citation" data-gtmlabel="BIB">
                <span class="citations__extension" data-gtmlabel="BIB">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .BIB
                </span>
                <span class="citations__types">
                        <span>
                            BibTeX
                        </span>
                        <span>
                            JabRef
                        </span>
                        <span>
                            Mendeley
                        </span>
                </span>
            </a>
        </li>
</ul></div>
    </div>

                            <a class="c-button share-this gtm-shareby-sharelink-link test-shareby-sharelink-link" data-test="shareable-link" target="_blank" rel="noopener" href="https://link.springer.com/sharelink/10.1186/2196-1115-1-2"><span>Share</span> <span class="hide-text-small">article</span></a>




                                    <a href="https://link.springer.com/content/pdf/10.1186%2F2196-1115-1-2.pdf" target="_blank" class="c-button c-button--blue c-button__icon-right gtm-pdf-link" title="Download this article in PDF format" rel="noopener">
            <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" version="1.1"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g fill="#fff"><g transform="translate(12.000000, 5.000000)"><path d="M7 7.3L7 1C7 0.4 6.6 0 6 0 5.4 0 5 0.4 5 1L5 7.3 3.5 5.7C3.1 5.3 2.5 5.3 2.1 5.7L2.1 5.7C1.7 6.1 1.7 6.7 2.1 7.1L5.3 10.3C5.7 10.7 6.3 10.7 6.7 10.3L9.9 7.1C10.3 6.7 10.3 6.1 9.9 5.7L9.9 5.7C9.5 5.3 8.9 5.3 8.5 5.7L7 7.3 7 7.3ZM0 13C0 12.4 0.5 12 1 12L11 12C11.6 12 12 12.4 12 13 12 13.6 11.5 14 11 14L1 14C0.4 14 0 13.6 0 13L0 13Z"></path></g></g></g></svg>
            <span class="hide-text-small">Download</span>
            <span>PDF</span>
        </a>

                </div>
            </div>




                    </div>
                    <aside class="main-sidebar-right u-interface">
                        <div data-role="sticky-wrapper">
                            <div class="main-sidebar-right__content u-composite-layer" data-component="SpringerLink.StickySidebar">
                                <div class="article-actions" id="article-actions">
                                    <h2 class="u-screenreader-only">Actions</h2>


                                    <div class="u-js-hide u-js-show-two-col">
                                        

                                                <div class="download-article test-pdf-link">
                                                            <a href="https://link.springer.com/content/pdf/10.1186%2F2196-1115-1-2.pdf" target="_blank" class="c-button c-button--blue c-button__icon-right gtm-pdf-link" title="Download this article in PDF format" rel="noopener">
            <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" version="1.1"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g fill="#fff"><g transform="translate(12.000000, 5.000000)"><path d="M7 7.3L7 1C7 0.4 6.6 0 6 0 5.4 0 5 0.4 5 1L5 7.3 3.5 5.7C3.1 5.3 2.5 5.3 2.1 5.7L2.1 5.7C1.7 6.1 1.7 6.7 2.1 7.1L5.3 10.3C5.7 10.7 6.3 10.7 6.7 10.3L9.9 7.1C10.3 6.7 10.3 6.1 9.9 5.7L9.9 5.7C9.5 5.3 8.9 5.3 8.5 5.7L7 7.3 7 7.3ZM0 13C0 12.4 0.5 12 1 12L11 12C11.6 12 12 12.4 12 13 12 13.6 11.5 14 11 14L1 14C0.4 14 0 13.6 0 13L0 13Z"></path></g></g></g></svg>
            <span class="hide-text-small">Download</span>
            <span>PDF</span>
        </a>

                                                </div>


                                            <div class="citations c-button-dropdown" data-component="SV.Dropdown" data-namespace="citations" aria-label="button with dropdown options">
        <button type="button" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown.citations-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">
    <span>Cite</span>
    <span class="hide-text-small">article</span>
</span><span class="c-button-dropdown__icon"></span></button>
<div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown.citations-dropdown"><ul class="citations__content" data-role="button-dropdown__content">
    <li>
        <a href="#citeas" class="gtm-cite-dropdown">How to cite?</a>
    </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1186/2196-1115-1-2?format=refman&amp;flavour=citation" title="Download this article's citation as a .RIS file" class="gtm-export-citation" data-gtmlabel="RIS">
                <span class="citations__extension" data-gtmlabel="RIS">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .RIS
                </span>
                <span class="citations__types">
                        <span>
                            Papers
                        </span>
                        <span>
                            Reference Manager
                        </span>
                        <span>
                            RefWorks
                        </span>
                        <span>
                            Zotero
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1186/2196-1115-1-2?format=endnote&amp;flavour=citation" title="Download this article's citation as a .ENW file" class="gtm-export-citation" data-gtmlabel="ENW">
                <span class="citations__extension" data-gtmlabel="ENW">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .ENW
                </span>
                <span class="citations__types">
                        <span>
                            EndNote
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1186/2196-1115-1-2?format=bibtex&amp;flavour=citation" title="Download this article's citation as a .BIB file" class="gtm-export-citation" data-gtmlabel="BIB">
                <span class="citations__extension" data-gtmlabel="BIB">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .BIB
                </span>
                <span class="citations__types">
                        <span>
                            BibTeX
                        </span>
                        <span>
                            JabRef
                        </span>
                        <span>
                            Mendeley
                        </span>
                </span>
            </a>
        </li>
</ul></div>
    </div>

                                                <a class="c-button share-this gtm-shareby-sharelink-link test-shareby-sharelink-link" data-test="shareable-link" target="_blank" rel="noopener" href="https://link.springer.com/sharelink/10.1186/2196-1115-1-2"><span>Share</span> <span class="hide-text-small">article</span></a>




                                    </div>
                                </div>
                                <nav class="toc" aria-label="article contents">
    <h2 class="u-h4 u-screenreader-only">Table of contents</h2>
    <ul id="article-contents" class="article-contents" role="menu">
            <li role="menuitem">
                <a title="Article" href="#enumeration"><span class="u-overflow-ellipsis">Article</span></a>
            </li>
            <li role="menuitem">
                <a title="Abstract" href="#Abs1"><span class="u-overflow-ellipsis">Abstract</span></a>
            </li>
            <li role="menuitem">
                <a title="Introduction" href="#Sec1"><span class="u-overflow-ellipsis">Introduction</span></a>
            </li>
            <li role="menuitem">
                <a title="Big data in health informatics" href="#Sec2"><span class="u-overflow-ellipsis">Big data in health informatics</span></a>
            </li>
            <li role="menuitem">
                <a title="Levels of health informatics data" href="#Sec3"><span class="u-overflow-ellipsis">Levels of health informatics data</span></a>
            </li>
            <li role="menuitem">
                <a title="Using micro level data – Molecules" href="#Sec8"><span class="u-overflow-ellipsis">Using micro level data – Molecules</span></a>
            </li>
            <li role="menuitem">
                <a title="Using tissue level data" href="#Sec10"><span class="u-overflow-ellipsis">Using tissue level data</span></a>
            </li>
            <li role="menuitem">
                <a title="Using patient level data" href="#Sec13"><span class="u-overflow-ellipsis">Using patient level data</span></a>
            </li>
            <li role="menuitem">
                <a title="Using population level data – Social media" href="#Sec16"><span class="u-overflow-ellipsis">Using population level data – Social media</span></a>
            </li>
            <li role="menuitem">
                <a title="Translational bioinformatics" href="#Sec20"><span class="u-overflow-ellipsis">Translational bioinformatics</span></a>
            </li>
            <li role="menuitem">
                <a title="Analysis and future works" href="#Sec21"><span class="u-overflow-ellipsis">Analysis and future works</span></a>
            </li>
            <li role="menuitem">
                <a title="Conclusion" href="#Sec27"><span class="u-overflow-ellipsis">Conclusion</span></a>
            </li>
            <li role="menuitem">
                <a title="Notes" href="#Notes"><span class="u-overflow-ellipsis">Notes</span></a>
            </li>
            <li role="menuitem">
                <a title="References" href="#Bib1"><span class="u-overflow-ellipsis">References</span></a>
            </li>
            <li role="menuitem">
                <a title="Copyright information" href="#copyrightInformation"><span class="u-overflow-ellipsis">Copyright information</span></a>
            </li>
            
            <li role="menuitem">
                <a title="About this article" href="#aboutcontent"><span class="u-overflow-ellipsis">About this article</span></a>
            </li>
    </ul>
</nav>

                            </div>
                                <div class="skyscraper-ad u-hide" data-component="SpringerLink.GoogleAds" data-namespace="skyscraper"></div>

                        </div>
                    </aside>
                </div>
            </main>
                <footer class="footer u-interface">
        <div class="footer__aside-wrapper">
            <div class="footer__content">
                <div class="footer__aside">
                    <p class="footer__strapline">Over 10 million scientific documents at your fingertips</p>
                                <div class="footer__edition c-button-dropdown c-button-dropdown--ghost" data-component="SpringerLink.EditionSwitcher" aria-label="button with dropdown options">
                                    <button type="button" title="Switch between Academic &amp; Corporate Edition" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="EditionSwitcher.4119061118-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">Academic Edition</span><span class="c-button-dropdown__icon"></span></button>
                                    <div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="EditionSwitcher.4119061118-dropdown"><ul data-role="button-dropdown__content">
                                        <li class="selected"><a href="https://link.springer.com/siteEdition/link" id="siteedition-academic-link">Academic Edition</a></li>
                                        <li><a href="https://link.springer.com/siteEdition/rd" id="siteedition-corporate-link" tabindex="-1">Corporate Edition</a></li>
                                    </ul></div>
                                </div>
                </div>
            </div>
        </div>
        <div class="footer__content">
            <ul class="footer__nav">
                <li>
                    <a href="https://link.springer.com/">Home</a>
                </li>
                <li>
                    <a href="https://link.springer.com/impressum">Impressum</a>
                </li>
                <li>
                    <a href="https://link.springer.com/termsandconditions">Legal information</a>
                </li>
                <li>
                    <a href="https://link.springer.com/privacystatement">Privacy statement</a>
                </li>
                <li>
                    <a href="https://link.springer.com/cookiepolicy">How we use cookies</a>
                </li>
                <li>
                    <a href="https://link.springer.com/accessibility" class="gtm-footer-accessibility">Accessibility</a>
                </li>
                <li>
                    <a id="contactus-footer-link" href="https://link.springer.com/contactus">Contact us</a>
                </li>
            </ul>
            <a class="parent-logo" target="_blank" rel="noopener" href="https://www.springernature.com/" title="Go to Springer Nature">
                <span class="u-screenreader-only">Springer Nature</span>
                <svg width="125" height="12">
                    <image width="125" height="12" alt="Springer Nature logo" src="/springerlink-static/481091012/images/png/springernature.png" xlink="http://www.w3.org/1999/xlink" xlink:href="/springerlink-static/481091012/images/svg/springernature.svg">
                    </image>
                </svg>
            </a>

            <p class="footer__copyright">© 2017 Springer International Publishing AG. Part of <a target="_blank" rel="noopener" href="https://www.springernature.com/">Springer Nature</a>.</p>

                <p class="footer__user-access-info">
                    <span>Not logged in</span>
                    <span>CAPES MEC (3000197460) - Universidade Tecnologica Federal do Parana (3000201946)</span>
                    <span>168.181.51.234</span>
                </p>
        </div>
    </footer>

        </div>
          <script type="text/javascript">
    (function() {
        document.addEventListener('readystatechange', function () {
            if (document.readyState === 'complete') {
                var linkEl = document.querySelector('.js-ctm');
                if (window.matchMedia && window.matchMedia(linkEl.media).matches) {
                    var scriptMathJax = document.createElement('script');
                    scriptMathJax.async = true;
                    scriptMathJax.src = '/springerlink-static/481091012/js/mathJax.js';
                    var s0 = document.getElementsByTagName('script')[0];
                    s0.parentNode.insertBefore(scriptMathJax, s0);
                }
            }
        });
    })();
</script>

<script type="text/javascript">
    (function() {
        var linkEl = document.querySelector('.js-ctm');
        var scriptsList = [];
        var polyfillFeatures = '';

        if (window.matchMedia && window.matchMedia(linkEl.media).matches) {
            (function(h){h.className = h.className.replace('no-js', 'js')})(document.documentElement);
            
            window.SpringerLink = window.SpringerLink || {};
            window.SpringerLink.staticLocation = '/springerlink-static/481091012';
            
            polyfillFeatures = 'default,fetch,Promise,Object.setPrototypeOf,Object.entries,Number.isInteger,MutationObserver,startsWith,Array.prototype.includes,Array.from';
            
            scriptsList = [
                // 'https://cdnjs.cloudflare.com/ajax/libs/airbrake-js/1.0.7/client.min.js',
                window.SpringerLink.staticLocation + '/js/jquery-3.3.1.min.js',
                'https://cdn.polyfill.io/v2/polyfill.js?features=' + polyfillFeatures,
                window.SpringerLink.staticLocation + '/js/main.js',
                        'https://cdn.cookielaw.org/consent/6b2ec9cd-5ace-4387-96d2-963e596401c6.js'
            ];

            scriptsList.forEach(function(script) {
                var tag = document.createElement('script');
                
                tag.async = false;
                tag.src = script;

                document.body.appendChild(tag);
            });
        }
    })();
</script><script src="jquery-3.js"></script><script src="polyfill.js"></script><script src="main.js"></script><script src="6b2ec9cd-5ace-4387-96d2-963e596401c6.js"></script>

<script type="text/javascript">
    function OptanonWrapper() {
        dataLayer.push({
            'event' : 'onetrustActive'
        });
    }
</script>

<script type="text/javascript" class="optanon-category-2">

   function viewport() {
        if (document.documentElement.clientWidth < 620) {
            size = 'small';
        }
        else if(document.documentElement.clientWidth < 1075 ) {
            size = 'medium';
        }
        else {
            size = 'wide';
        }
        return size;
    };

    function reportForMouseEvents(linkCssSelectors, nolardUrl, experiment, abgroup) {
        var counter = 0;
        linkCssSelectors.forEach(function(cssSelector) {
            $('body').delegate(cssSelector, 'click', function() {
                if(counter == 0) {
                    reportConversion(nolardUrl, experiment, abgroup);
                    counter++;
                }
            });
        });
    };

    function reportConversion(nolardUrl, experiment, abgroup) {
        $.ajax({ url: nolardUrl + '/convert/' + experiment + '/' + abgroup });
    };

    function reportForMouseEvent(linkCssSelector, nolardUrl, experiment, abgroup) {
        var counter = 0;
        var elem = document.querySelector(linkCssSelector)

        if (elem.addEventListener) {
            elem.addEventListener("click", function(e) {
                if(counter == 0) {
                    reportConversion(nolardUrl, experiment, abgroup);
                    counter++;
                }
            });
        } else  {
            elem.attachEvent("click", function(e) {
                if(counter == 0) {
                    reportConversion(nolardUrl, experiment, abgroup);
                    counter++;
                }
            });
        }
    };

    function reportParticipation(nolardUrl, experiment, abgroup, participations) {
        if (participations)
            participations.push(experiment + '/' + abgroup)

        var xhr = new XMLHttpRequest()
        xhr.open('GET', nolardUrl + '/participate/' + experiment + '/' + abgroup);
        xhr.send();
    };

        function recordCurrentExperiment() {
            var participations = [];
            if (document.querySelector('#reprintsandpermissions-bottom-link') !== null) {
                reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-article-reprints-permissions-location', 'featureoff', participations);
                reportForMouseEvent('#reprintsandpermissions-bottom-link', 'https://ab-reporting.live.cf.public.springer.com', 'bunsen-article-reprints-permissions-location', 'featureoff');
            } else if(document.querySelector('#reprintsandpermissions-top-link') !== null) {
                reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-article-reprints-permissions-location', 'featureon', participations);
                reportForMouseEvent('#reprintsandpermissions-top-link', 'https://ab-reporting.live.cf.public.springer.com', 'bunsen-article-reprints-permissions-location', 'featureon');
            }
            return participations;
        };

</script>

<script type="text/javascript" class="optanon-category-2">
    window.onload = function() {
    var linkEl = document.querySelector('.js-ctm');


        if (window.matchMedia && window.matchMedia(linkEl.media).matches) {
          var bookProductType = dataLayer[0].content && dataLayer[0].content.book && dataLayer[0].content.book.bookProductType
          var hasBody = dataLayer[0]['Has Body'] == 'Y';
          var userHasAccess = dataLayer[0]['HasAccess'] == 'Y';
          var liveOrStatic = dataLayer[0].content && dataLayer[0].content.version

          // Baseline reference work page
          var isReferenceWork = window.location.pathname.startsWith('/referencework/');
          var isReferenceWorkEntry = window.location.pathname.startsWith('/referenceworkentry/');
          var isEncOrDict = bookProductType == 'Encyclop(a)edia' || bookProductType == 'Dictionary';

          // Baseline static reference work entry page pdf download
          if (isReferenceWorkEntry && isEncOrDict && $('.test-rwepdf-link').length) {
            reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwepage-enc-pdf-download', 'baseline');
            reportForMouseEvents(['.test-rwepdf-link'], 'https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwepage-enc-pdf-download', 'baseline');
          }
          if (isReferenceWorkEntry && !isEncOrDict && $('.gtm-pdf-link').length) {
            reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwepage-nonenc-pdf-download', 'baseline');
            reportForMouseEvents(['.gtm-pdf-link'], 'https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwepage-nonenc-pdf-download', 'baseline');

            if (($('.test-bookpdf-link').length || $('.test-bookepub-link').length)) {
                reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwpage-nonenc-book-download', 'baseline');
                reportForMouseEvents(['.test-bookpdf-link', 'test-bookepub-link'], 'https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwpage-nonenc-book-download', 'baseline');
            }
          }

          // Baseline static reference work page pdf download
          if (isReferenceWork && isEncOrDict && ($('.test-bookpdf-link').length || $('.test-bookepub-link').length)) {
            reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwpage-enc-book-download', 'baseline');
              reportForMouseEvents(['.test-bookpdf-link', 'test-bookepub-link'], 'https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwpage-enc-book-download', 'baseline');
          }

          if (isReferenceWork && !isEncOrDict) {
            if (($('.test-bookpdf-link').length || $('.test-bookepub-link').length)) {
                reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwpage-nonenc-book-download', 'baseline');
                reportForMouseEvents(['.test-bookpdf-link', 'test-bookepub-link'], 'https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwpage-nonenc-book-download', 'baseline');
            }

            if ($('.test-book-toc-download-link').length) {
                reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwpage-nonenc-pdf-download', 'baseline');
                reportForMouseEvents(['.test-book-toc-download-link'], 'https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwpage-nonenc-pdf-download', 'baseline');
            }
          }

          // Baseline static rwe page fulltext html view with access
          if (isReferenceWorkEntry && isEncOrDict && hasBody) {
            reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwepage-enc-html-download', 'baseline');
            if (userHasAccess)
                reportConversion('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwepage-enc-html-download', 'baseline');
          }
          if (isReferenceWorkEntry && !isEncOrDict && hasBody) {
            reportParticipation('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwepage-nonenc-html-download', 'baseline');
            if (userHasAccess)
                reportConversion('https://ab-reporting.live.cf.public.springer.com', 'bunsen-'+liveOrStatic+'-rwepage-nonenc-html-download', 'baseline');
          }
      }

            if (window.matchMedia && window.matchMedia(linkEl.media).matches) {
                recordCurrentExperiment()
            }
   };

</script>

    <script type="text/javascript" id="googletag-push">
        
            var adSlot = '270604982/springerlink/40537/article';
        

        var definedSlots = [
                {slot: [728, 90], containerName: 'doubleclick-leaderboard-ad'},
                {slot: [160, 600], containerName: 'doubleclick-ad'},
        ];
    </script>


        
        <span id="chat-widget" class="u-hide"></span>
                    <noscript>
                <img aria-hidden="true" role="presentation" src="https://ssl-springer.met.vgwort.de/na/vgzm.415900-10.1186-2196-1115-1-2" width='1' height='1' alt='' />
            </noscript>

    

<div style="display: none; visibility: hidden;"><script id="gpt-control-script" src="gpt.js"></script>

<script id="gpt-control-setup">var googletag=googletag||{};googletag.cmd=googletag.cmd||[];</script></div><script type="text/javascript" id="">window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);
googletag.cmd.push(function(){googletag.pubads().setTargeting("doi",google_tag_manager["GTM-WCF9Z9"].macro(8));googletag.pubads().setTargeting("kwrd",google_tag_manager["GTM-WCF9Z9"].macro(9));googletag.pubads().setTargeting("pmc",google_tag_manager["GTM-WCF9Z9"].macro(10));googletag.pubads().setTargeting("BPID",google_tag_manager["GTM-WCF9Z9"].macro(11));googletag.pubads().setTargeting("edition",google_tag_manager["GTM-WCF9Z9"].macro(12));googletag.pubads().setTargeting("sucode",google_tag_manager["GTM-WCF9Z9"].macro(13));googletag.pubads().setTargeting("eissn",google_tag_manager["GTM-WCF9Z9"].macro(14));googletag.pubads().setTargeting("pissn",
google_tag_manager["GTM-WCF9Z9"].macro(15));googletag.pubads().setTargeting("eisbn",google_tag_manager["GTM-WCF9Z9"].macro(16));googletag.pubads().setTargeting("pisbn",google_tag_manager["GTM-WCF9Z9"].macro(17));googletag.pubads().setTargeting("logged",google_tag_manager["GTM-WCF9Z9"].macro(18));googletag.pubads().setTargeting("ksg",Krux.segments);googletag.pubads().setTargeting("kuid",Krux.uid);googletag.pubads().setRequestNonPersonalizedAds(google_tag_manager["GTM-WCF9Z9"].macro(22)?0:1);googletag.pubads().enableSingleRequest();googletag.pubads().enableAsyncRendering();googletag.enableServices()});</script><button href="javascript:;" title="focus catcher" class="js-focus-catcher u-screenreader-only" tabindex="-1"></button><div id="optanon" class="modern"><div id="optanon-popup-bg"></div><div id="optanon-popup-wrapper" role="dialog" aria-modal="true" tabindex="-1" lang="en-GB"><div id="optanon-popup-top"><a href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Close Button');" class="optanon-close-link optanon-close optanon-close-ui" title="Close Preference Centre"><div id="optanon-close" style="background: url(https://optanon.blob.core.windows.net/skins/default_flat_bottom_two_button_white/v2/images/optanon-pop-up-close.png);width:34px;height:34px;"></div></a></div><div id="optanon-popup-body"><div id="optanon-popup-body-left"><div id="optanon-popup-body-left-shading"></div><div id="optanon-branding-top-logo" style="background-image: url(https://optanon.blob.core.windows.net/logos/5138/5138:link.springer.com/springer.png) !important;"></div><ul id="optanon-menu"><li class="menu-item-on menu-item-about" title="Your Privacy"><p><a href="#">Your Privacy</a></p></li><li class="menu-item-necessary menu-item-on" title="Strictly Necessary Cookies"><p><a href="#">Strictly Necessary Cookies</a></p></li><li class="menu-item-on menu-item-performance" title="Performance Cookies"><p><a href="#">Performance Cookies</a></p></li><li class="menu-item-on menu-item-functional" title="Functional Cookies"><p><a href="#">Functional Cookies</a></p></li><li class="menu-item-on menu-item-advertising" title="Targeting Cookies"><p><a href="#">Targeting Cookies</a></p></li><li class="menu-item-moreinfo menu-item-off" title="More Information"><p><a target="_blank" href="https://link.springer.com/cookiepolicy" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Cookie Policy');">More Information</a></p></li></ul></div><div id="optanon-popup-body-right"><h2 aria-label="true">Privacy Preference Centre</h2><h3></h3><div id="optanon-popup-more-info-bar"><div class="optanon-status"><div class="optanon-status-editable"><form><fieldset><p><input value="check" id="chkMain" checked="checked" class="optanon-status-checkbox" type="checkbox"><label for="chkMain">Active</label></p></fieldset></form></div><div class="optanon-status-always-active optanon-status-on"><p>Always Active</p></div></div></div><div id="optanon-main-info-text"></div></div><div class="optanon-bottom-spacer"></div></div><div id="optanon-popup-bottom"> <a href="https://onetrust.com/poweredbyonetrust" target="_blank"><div id="optanon-popup-bottom-logo" style="background: url(https://optanon.blob.core.windows.net/skins/default_flat_bottom_two_button_white/v2/images/cookie-collective-top-bottom.png);width:155px;height:35px;" title="powered by OneTrust"></div></a><div class="optanon-button-wrapper optanon-save-settings-button optanon-close optanon-close-consent"><div class="optanon-white-button-left"></div><div class="optanon-white-button-middle"><a href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Save Settings');">Save Settings</a></div><div class="optanon-white-button-right"></div></div><div class="optanon-button-wrapper optanon-allow-all-button optanon-allow-all" style="display: none;"><div class="optanon-white-button-left"></div><div class="optanon-white-button-middle"><a href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Allow All');">Allow All</a></div><div class="optanon-white-button-right"></div></div></div></div></div><div class="optanon-alert-box-wrapper  " style="bottom: 0px;"><div class="optanon-alert-box-bottom-top"><div class="optanon-alert-box-corner-close"><a class="optanon-alert-box-close" href="#" title="Close Banner" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Banner Close Button');"></a></div></div><div class="optanon-alert-box-bg"><div class="optanon-alert-box-logo"> </div><div class="optanon-alert-box-body"><p>We
 use cookies to personalise content and ads, to provide social media 
features and to analyse our traffic. We also share information about 
your use of our site with our social media, advertising and analytics 
partners in accordance with our <a href="https://link.springer.com/privacystatement">Privacy Statement</a>. You can manage your preferences in Manage Cookies.</p></div><div class="optanon-clearfix"></div><div class="optanon-alert-box-button-container"><div class="optanon-alert-box-button optanon-button-close"><div class="optanon-alert-box-button-middle"><a class="optanon-alert-box-close" href="#">Close</a></div></div><div class="optanon-alert-box-button optanon-button-allow"><div class="optanon-alert-box-button-middle"><a class="optanon-allow-all" href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Banner Accept Cookies');">OK</a></div></div><div class="optanon-alert-box-button optanon-button-more"><div class="optanon-alert-box-button-middle"><a class="optanon-toggle-display" href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Banner Open Preferences');">Manage Cookies</a></div></div></div><div class="optanon-clearfix optanon-alert-box-bottom-padding"></div></div></div><script type="text/javascript" id="">(function(d,z){function n(){for(var a in dataLayer)if(dataLayer.hasOwnProperty(a)&&dataLayer[a]["Event Category"])return dataLayer[a]["Event Category"];return"warning: no event category"}function v(a,b,c,h){"undefined"!==typeof dataLayer?dataLayer.push({event:"Scroll To Section",eventCategory:n(),eventAction:a,eventLabel:b,eventValue:1,eventNonInteraction:h}):("undefined"!==typeof ga&&ga("send","event",n(),a,b,1,{nonInteraction:1}),"undefined"!==typeof _gaq&&_gaq.push(["_trackEvent",n(),a,b,1,!0]))}
function k(b,g,c,h){if(-1===a[c].cache.indexOf(c+"-"+h)&&document.querySelectorAll(b).length){var f="viewed"===h?document.querySelectorAll(b)[0].getBoundingClientRect().height:0,x=document.querySelectorAll(b)[0].getBoundingClientRect().top+d.pageYOffset;g>=x+f&&(v(p(c)+" "+p(h),b,c,!1),a[c].cache.push(c+"-"+h),q++)}}function r(a){a=document.querySelectorAll(a);return a.length?a[0].offsetHeight||0:0}function m(b,g,c){var f=d.pageYOffset+d.innerHeight;if(q>=w)"handleClick"===g&&d.removeEventListener(b,
t),"throttle"===g&&d.removeEventListener(b,u);else if(c)setTimeout(function(){0<r(a[c].content)&&(a[c].reached&&k(a[c].content,f,c,"reached"),a[c].viewed&&k(a[c].content,f,c,"viewed"))},10);else for(var e in a)a.hasOwnProperty(e)&&0<r(a[e].content)&&(a[e].reached&&k(a[e].content,f,e,"reached"),a[e].viewed&&k(a[e].content,f,e,"viewed"))}function y(b){b=document.querySelectorAll(a[b].content).length;return 0<b?!0:!1}function p(a){return a.charAt(0).toUpperCase()+a.slice(1)}function u(a,b){var c,f,e,
g=null,d=0,k=function(){d=new Date;g=null;e=a.apply(c,f)};return function(){var h=new Date;d||(d=h);var l=b-(h-d);c=this;f=arguments;0>=l?(clearTimeout(g),g=null,d=h,e=a.apply(c,f)):g||(g=setTimeout(k,l));return e}}function t(a,b){return function(c){a(c,b)}}var w=0,q=0,a={recommendations:{content:".gtm-recommendations iframe",clickable:".gtm-recommendations .gtm-recommendations__title",exists:!1,reached:!1,viewed:!0,size:0,cache:[]},"abstract":{content:"Section.Abstract",clickable:null,exists:!0,
reached:!1,viewed:!0,size:0,cache:[]},references:{content:".Bibliography \x3e .content, [id^\x3dBib1] \x3e .content",clickable:".Bibliography \x3e .Heading, [id^\x3dBib1] \x3e .Heading",exists:!0,reached:!0,viewed:!0,size:0,cache:[]},about:{content:"#AboutThisContent",clickable:"#AboutThisContent \x3e #aboutcontent",exists:!1,reached:!1,viewed:!0,size:0,cache:[]}};d.addEventListener("scroll",u(function(){m("scroll","throttle")},500));d.addEventListener("orientationchange",u(function(){m("orientationchange",
"throttle")},500));for(var b in a)if(a.hasOwnProperty(b)&&(a[b].exists&&a[b].size++,a[b].reached&&a[b].size++,a[b].viewed&&a[b].size++,w+=a[b].size,a[b].exists&&y(b)||!a[b].exists)){a[b].exists&&-1===a[b].cache.indexOf(b+"-exists")&&(v(p(b)+" Exists",a[b].content,b,!0),a[b].cache.push(b+"-exists"),q++);if(0<r(a[b].content)){var l=d.pageYOffset+d.innerHeight;a[b].reached&&k(a[b].content,l,b,"reached");a[b].viewed&&k(a[b].content,l,b,"viewed")}a[b].clickable&&(l=document.querySelectorAll(a[b].clickable)[0])&&
(l.addEventListener("click",t(function(a,b){m("click","handleClick",b)},b)),l.addEventListener("click",t(function(a,b){13==a.keyCode&&m("click","handleClick",b)},b)))}})(window);</script><script type="text/javascript" id="gtm-recommendations-script" src="entry-point"></script><script type="text/javascript" id="">window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);(function(){var a=document.createElement("script");a.type="text/javascript";a.async=!0;a.src=("https:"===location.protocol?"https:":"http:")+"//cdn.krxd.net/controltag/KDqyaFZ_.js";var b=document.getElementsByTagName("script")[0];b.parentNode.insertBefore(a,b)})();</script><script type="text/javascript" id="">window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);(function(){function b(a){a="kx"+a;return window.localStorage?window.localStorage[a]||"":navigator.cookieEnabled?(a=document.cookie.match(a+"\x3d([^;]*)"))&&unescape(a[1])||"":""}Krux.user=b("user");Krux.segments=b("segs")&&b("segs").split(",")||[]})();</script><script type="text/javascript" id="">var allowed=google_tag_manager["GTM-WCF9Z9"].macro(38);Krux("consent:set",{dc:allowed,al:allowed,tg:allowed,cd:!1,sh:!1,re:!1},function(a,b){a?console.error(a):console.log("consent flags set ",b)});</script><script type="text/javascript" id="">window.dataLayer.push({ksg:Krux.segments,kuid:Krux.uid});</script><div class="kxhead" data-id="KDqyaFZ_" style="display:none !important;"><span class="kxtag kxinvisible" data-id="28134"><script>
(function() {

    // cient only wants to collect first two paths
    Krux('scrape', {
        'page_attr_Springer_Link_url_path_1': {
            'url_path': '1'
        }
    });
    Krux('scrape', {
        'page_attr_Springer_Link_url_path_2': {
            'url_path': '2'
        }
    });
    // Krux('scrape',{'page_attr_url_path_3':{'url_path':'3'}});
    // Krux('scrape',{'page_attr_meta_keywords':{meta_name:'keywords'}});

    // Krux('scrape',{'page_attr_domain':{url_domain: '2'}});

})();
</script></span></div></body><iframe id="kx-proxy-KDqyaFZ_" src="https://cdn.krxd.net/partnerjs/xdi/proxy.3d2100fd7107262ecb55ce6847f01fa5.html#%21kxcid=KDqyaFZ_&amp;kxt=https%3A%2F%2Flink.springer.com&amp;kxcl=cdn&amp;kxp=" style="display: none; visibility: hidden; height: 0; width: 0;"></iframe></html>