<!DOCTYPE html>
<html class="js" lang="en-gb"><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
        <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1">
<meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=2.5,user-scalable=yes">
    <meta name="citation_publisher" content="Springer US">
    <meta name="citation_title" content="Classy: fast clustering streams of call-graphs">
    <meta name="citation_doi" content="10.1007/s10618-014-0367-9">
    <meta name="citation_language" content="en">
    <meta name="citation_abstract_html_url" content="https://link.springer.com/article/10.1007/s10618-014-0367-9">
    <meta name="citation_fulltext_html_url" content="https://link.springer.com/article/10.1007/s10618-014-0367-9">
    <meta name="citation_pdf_url" content="https://link.springer.com/content/pdf/10.1007%2Fs10618-014-0367-9.pdf">
    <meta name="citation_springer_api_url" content="http://api.springer.com/metadata/pam?q=doi:10.1007/s10618-014-0367-9&amp;api_key=">
    <meta name="citation_firstpage" content="1554">
    <meta name="citation_lastpage" content="1585">
    <meta name="citation_author" content="Orestis Kostakis">
    <meta name="citation_author_institution" content="Labs, F-Secure">
    <meta name="citation_author_institution" content="Aalto University">
    <meta name="citation_author_email" content="orestis.kostakis@f-secure.com">
    <meta name="dc.identifier" content="10.1007/s10618-014-0367-9">
    <meta name="format-detection" content="telephone=no">
    <meta name="description" content="An abstraction resilient to common malware obfuscation techniques is the call-graph. A call-graph is the representation of an executable file as a directed graph with labeled vertices, where the...">
    <meta name="twitter:card" content="summary">
    <meta name="twitter:title" content="Classy: fast clustering streams of call-graphs">
    <meta name="twitter:image" content="https://static-content.springer.com/cover/journal/10618/28/5.jpg">
    <meta name="twitter:image:alt" content="Content cover image">
    <meta name="twitter:site" content="SpringerLink">
    <meta name="twitter:description" content="An abstraction resilient to common malware obfuscation techniques is the call-graph. A call-graph is the representation of an executable file as a directed graph with labeled vertices, where the...">
    <meta name="citation_journal_title" content="Data Mining and Knowledge Discovery">
    <meta name="citation_journal_abbrev" content="Data Min Knowl Disc">
    <meta name="citation_volume" content="28">
    <meta name="citation_issue" content="5-6">
    <meta name="citation_issn" content="1384-5810">
    <meta name="citation_issn" content="1573-756X">
    <meta name="citation_online_date" content="2014/07/10">
    <meta name="citation_cover_date" content="2014/09/01">
    <meta name="citation_article_type" content="Article">
    <meta property="og:title" content="Classy: fast clustering streams of call-graphs">
    <meta property="og:type" content="Article">
    <meta property="og:url" content="https://link.springer.com/article/10.1007/s10618-014-0367-9">
    <meta property="og:image" content="https://static-content.springer.com/cover/journal/10618/28/5.jpg">
    <meta property="og:site_name" content="SpringerLink">
    <meta property="og:description" content="An abstraction resilient to common malware obfuscation techniques is the call-graph. A call-graph is the representation of an executable file as a directed graph with labeled vertices, where the...">

        <title>Classy: fast clustering streams of call-graphs | SpringerLink</title>
        <link rel="canonical" href="https://link.springer.com/article/10.1007/s10618-014-0367-9">
        <link rel="shortcut icon" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon.ico">
<link rel="icon" sizes="16x16 32x32 48x48" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon.ico">
<link rel="icon" sizes="16x16" type="image/png" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon-16x16.png">
<link rel="icon" sizes="32x32" type="image/png" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon-32x32.png">
<link rel="icon" sizes="48x48" type="image/png" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon-48x48.png">
<link rel="apple-touch-icon" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-iphone@3x.png">
<link rel="apple-touch-icon" sizes="72x72" href="https://link.springer.com/springerlink-static/439917616/images/favicon/ic_launcher_hdpi.png">
<link rel="apple-touch-icon" sizes="76x76" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-ipad.png">
<link rel="apple-touch-icon" sizes="114x114" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-114x114.png">
<link rel="apple-touch-icon" sizes="120x120" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-iphone@2x.png">
<link rel="apple-touch-icon" sizes="144x144" href="https://link.springer.com/springerlink-static/439917616/images/favicon/ic_launcher_xxhdpi.png">
<link rel="apple-touch-icon" sizes="152x152" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-ipad@2x.png">
<link rel="apple-touch-icon" sizes="180x180" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-iphone@3x.png">
<meta name="msapplication-TileColor" content="#ffffff">
<meta name="msapplication-TileImage" content="/springerlink-static/439917616/images/favicon/ic_launcher_xxhdpi.png">
        <link rel="dns-prefetch" href="https://fonts.gstatic.com/">
<link rel="dns-prefetch" href="https://fonts.googleapis.com/">
<link rel="dns-prefetch" href="https://google-analytics.com/">
<link rel="dns-prefetch" href="https://www.google-analytics.com/">
<link rel="dns-prefetch" href="https://www.googletagservices.com/">
<link rel="dns-prefetch" href="https://www.googletagmanager.com/">
<link rel="dns-prefetch" href="https://static-content.springer.com/">
        <link rel="stylesheet" href="basic.css" media="screen">
<link rel="stylesheet" href="styles.css" class="js-ctm" media="only screen and (-webkit-min-device-pixel-ratio:0) and (min-color-index:0), (-ms-high-contrast: none), only all and (min--moz-device-pixel-ratio:0) and (min-resolution: 3e1dpcm)">
<link rel="stylesheet" href="print.css" media="print">


            <script type="text/javascript" src="https://beacon.krxd.net/optout_check?callback=Krux.ns.macmillan.kxjsonp_optOutCheck"></script><script async="" src="/springerlink-static/439917616/js/mathJax.js"></script><script type="text/javascript" src="bd339c69-af54-4a21-b4f1-654bcfcd83ca"></script><script id="nativeads-plugin" type="text/javascript" async="" src="plugin.js"></script><script async="" src="controltag.10b38d7fdd1a9bfe7b2b04b85118d005"></script><script type="text/javascript" async="" src="KDqyaFZ_.js"></script><script type="text/javascript" async="" src="analytics.js"></script><script async="" src="gtm.js"></script><script type="text/javascript">
        window.Krux||((Krux=function(){Krux.q.push(arguments);}).q=[]);
        var dataLayer = [{
                    'GA Key':'UA-26408784-1',
                    'Features':["leaderboardadverts","abtesting"],
                    'Event Category':'Article',
                    'Open Access':'N',
                    'Labs':'Y',
                    'DOI':'10.1007\/s10618-014-0367-9',
                    'VG Wort Identifier':'pw-vgzm.415900-10.1007-s10618-014-0367-9',
                    'HasAccess':'Y',
                    'Full HTML':'Y',
                    'Has Body':'Y',
                    'Static Hash':'439917616',
                    'Has Preview':'N',
                    'user':{'license': {'businessPartnerID': ['3000197460', '3000201946'], 'businessPartnerIDString': '3000197460|3000201946'}},
                    'content':{'type': 'article', 'serial': {'eissn': '1573-756X', 'pissn': '1384-5810'}, 'category': {'pmc': {'primarySubject': 'Computer Science', 'primarySubjectCode': 'I', 'secondarySubjects': {'1': 'Data Mining and Knowledge Discovery', '2': 'Artificial Intelligence (incl. Robotics)', '3': 'Information Storage and Retrieval', '4': 'Statistics for Engineering, Physics, Computer Science, Chemistry and Earth Sciences'}, 'secondarySubjectCodes': {'1': 'I18030', '2': 'I21017', '3': 'I18032', '4': 'S14000'}}, 'sucode': 'SC6'}},
                    'Access Type':'subscription',
                    'Page':'article',
                    'Bpids':'3000197460, 3000201946',
                    'Bpnames':'CAPES MEC, Universidade Tecnologica Federal do Parana',
                    'SubjectCodes':'SCI, SCI18030, SCI21017, SCI18032, SCS14000',
                    'session':{'authentication': {'loginStatus': 'N'}, 'attributes': {'edition': 'academic'}},
                    'Keywords':'Clustering, Streams, Call-graphs, Malware, Graph edit distance',
                    'Country':'BR',
                    'Journal Id':'10618',
                    'Journal Title':'Data Mining and Knowledge Discovery',

                    'doi': "10.1007-s10618-014-0367-9",
                    'kwrd': ["Clustering","Streams","Call-graphs","Malware","Graph_edit_distance"],
                    'pmc': ["I","I18030","I21017","I18032","S14000"],
                    'BPID': ["3000197460","3000201946"],
                    'ksg': Krux.segments,
                    'kuid': Krux.uid,

        }];
    </script>

<script type="text/javascript" src="jquery-3.js"></script>

        <script type="text/javascript" src="6b2ec9cd-5ace-4387-96d2-963e596401c6.js" charset="UTF-8"></script><link type="text/css" href="optanon.css" rel="stylesheet"><style>#optanon ul#optanon-menu li { background-color: #F7FBFE !important }#optanon ul#optanon-menu li.menu-item-selected { background-color: #FFFFFF !important }#optanon #optanon-popup-wrapper .optanon-white-button-middle { background-color: #3365A4 !important }.optanon-alert-box-wrapper .optanon-alert-box-button-middle { background-color: #3365A4 !important; border-color: #3365A4 !important; }#optanon #optanon-popup-wrapper .optanon-white-button-middle a { color: #ffffff !important }.optanon-alert-box-wrapper .optanon-alert-box-button-middle a { color: #ffffff !important }#optanon #optanon-popup-bottom { background-color: #F7FBFE !important }#optanon.modern #optanon-popup-top, #optanon.modern #optanon-popup-body-left-shading { background-color: #F7FBFE !important }.optanon-alert-box-wrapper { background-color:#F7FBFE !important }.optanon-alert-box-wrapper .optanon-alert-box-bg p { color:#333333 !important }</style>

<script type="text/javascript">
    function OptanonWrapper() {
        dataLayer.push({
            'event' : 'onetrustActive'
        });
    }
</script>

    <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
                new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
            j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
            'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
    })(window,document,'script','dataLayer','GTM-WCF9Z9');</script>

    <style>@font-face{font-family:'Source Sans Pro';font-weight:400;src:url(data:application/font-woff;base64,d09GRgABAAAAAERcABEAAAAAiPgAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABgAAAADYAAABAA0QDckdQT1MAAAG4AAAHPAAAGXDUveN5R1NVQgAACPQAAACJAAAA4PFn1ldPUy8yAAAJgAAAAFcAAABgWrSUW2NtYXAAAAnYAAABRQAAAebzMPm1Y3Z0IAAACyAAAAAoAAAAKA2ZAPpmcGdtAAALSAAAAQIAAAFzBlmcN2dhc3AAAAxMAAAACAAAAAj//wADZ2x5ZgAADFQAADCcAABd7CMplIhoZWFkAAA88AAAADYAAAA2/hSz4mhoZWEAAD0oAAAAHwAAACQHowOhaG10eAAAPUgAAAHlAAADxKaoJjpsb2NhAAA/MAAAAd4AAAHkOIhQ1G1heHAAAEEQAAAAIAAAACADCwJAbmFtZQAAQTAAAADXAAABxiK6PDBwb3N0AABCCAAAAgUAAAM5bFBIb3ByZXAAAEQQAAAASwAAAEuWBPrreAENwbEBQDAQAMADgH10RmAMHSUyaqbJ36mwAmqj1qayx8rhVLli5faofLHyS2o5VgWBWga3AAB4AdxWVXDkRhB9ZqjyYZiZmfknzJzvMDMz/YThJ8x4zMzMzGz2eu8sr9ebs867UeqlqzM1JU2dDOHkaaHV06Se7mkhD0A5TsAFyL/goituQI/7b3n8QeyFQghIXbe0XPn33fHog+jx2y+g63kikQ/ItzzvErk7TDV64CLcj1fxPSZiMWqRzsvPK83rk3dA3lF5p4hUD5ShL/bE3jgZp+ImPIy38A6+RD8MwExZ21Os7M0kTqaHU5nAHeiDO3EIvmQG/TgJw7ALilTiVGaE2y7cMcLdD6Wis0O4rbgN5aJ3hOidLBJtIrFCJE4VvZOZFglfuOsxiL5wS8XfngxExhNOu3COFbk9mRUfIqfcVuEehgKRmYsBqjlTtNQWioWTwDBuQZ5oVSNffueK7lqR35MJoRLi0RP+nShBGnuqxCKcrLyXmMMr8s3f5W2xiF0qdzsGZ+I4gI0cxDr5HYcDcalIVOAvhMT9p4JprgIk8j8ZnK6/cyO8pFwb/wTbsrNy9QNwtuXN5BaAq7lFqKY/wYePPS29HiZ+gDUcLitl9NBtcD1bAWaUDpjhAsNvp68eFcLP8VuRTYonheoUMsHFnKvaDmxsgbHtM2NtBVYiDfXBWiAcPXMMrLSPnYApih1m6UVsJ4Wus9my2lyMHlYzoLUNMCG/DtRm4PD03gXr9a80lJPOa1tioj630B5TMPFzm65nAWhUGsVKWHA7wCxFQ75p1tJ3PfJ11nE+RzMAWMnNHAmggGKXCc3FdpXy6PNuDuf3/N5qjgZYzX4cyc0AF7JypznJRSuMvvKdijS6xXZPfLYALpiN2nZ3mj71nvplCgrD0b10dYTeyFpDZRzbWUSg6w6tO9GOQnQTzHCc5rsK4Idczw1cyCTHMcWpfFUrdaU5fXZhv2hcAKtEfjhXcCXncIzm0MJ5vjRT9GFrkSsc2RqbOd/Jclz1/mSpSoA5vdJ6Pw1hHM66neoH6kkr2Y2Ys7UbMyjjD8JLa0VPAbiOn6tUJaD1faBwo1Whq5zKBVwv0iu4io2hqshRd5dpE2mOtfS1Izw2sUq1t5oIdrF6uciJU8nKmLMqGZKLARe6Ha29GQtbg0nNTLFcbg81ObW5i1ubqlcaf74zAXRWuVyGboIf0mOlnhJ7cZvQHzvr7TBnIJ90Vsbr79cczC0dzAafCc1JEYroh08CQyvFSnoMnHxrnYhW1mYz/PTH6fc42z0HcWl4LwwXbAVwnnAuZ8acKafhQOyFI3AQKuzkMv840WYayjlUJ4lbCZlIT7WjB06ABTdHu9HK9wDsBGijT9/46hGpkxRUw3RAInKq2RpkwNUmyiytBJMmwk7h9oZ6znaoMc/0spmuuqsxE5GJsDU2M6DErbO8BcW2W12UscrqDLHvKGuYsfwIuIjPWT+dgr6+IfdllvPYbJ64UfnqSVHBNM5zc4OD0Fc721NpP1rfkenY1tW+pB+RLgdc2Lq0/kJ1kgrVd87EcBB6RHbAUkxFLdgOrEWnYNKNqmsabAtFuI5B6KkD1oZPFhym/7YHXbAFcOd8yPaW6PucM1uPZcbaDjp+OiaMnOYzKs2tIY++iX1PuFAtpvH3o1CfxcxLjucmejoxK7XrnrTTu0PQ06sW/x/E1/GGP2aJPmst530upc+RANfoOfEgugA2mvPEw98CW7lNAHdwB8Bp+s7WTJ/bALZyHH/p/FQQ+SZ63M5GNlueF3lDT+uZXKDzq4U+5Z4/a3X6MTYD1QAbmDE29N7BNXF1i98JNqrvqd23yAQT7qkVD83JfPekiJ3VyucULkcMuF4zuwrdhGplO35KrnRmooK1dorEVgibmInrtzgN/IXgO1zQRcmgaxXAKks9z7mGqmRa+TFgNu5dIH6m6fQMYmZaJ36YdONla3R2q2WffmS2dla/yUhe8lHxa1vzrLaJAQBh9Py2jc/WGnHapIlt21a5jG3bubVcQqaLds7z1mNUGVUNBe2o6LpD1b2ecaPnvOBBJ7zjEe/73Eu+iuO+iRO+i5N+iFN+itOGjPrWuAmTZjW1TGnr6Jo2ox2zemnOvAWLBpbssWzFqoI1bNqybRfsYDcKUVVSVlF1VM1Z6hrG/6ZpMkV0IuTtb5oWIyyb1bf6N5vWUgRAVZ7/ZtN+1Qj1oGBLxZm3EesxZNghow5H0REJHHBQsi/CCNj7d8YMu8hozBqPBdOxYC5WLMeC1ViwHnM2YthmjNiKku0YtRNluzGmEPOKsaisalktltRjXDNWtWJCV09bP5YNom1PLNsbbfuiY3+sORDrDkbdoZh0OBqOxJSj0XWW8826MBZc7FILLnOlEVfFmKtjyLVutugWd1h2p3usutd92u73qGWPe8q6pz1jyvNesOSEk2acclrfy7HrlSh4NSpei6LXvaHqzah5K6a9HTXvxLT3o+WD6PkwWj6Kno99punz2PJ1DHwbA9/HwA+x46fY9Ets+CO2/RmrfwE9c0UteAFjYGRgYOBiSGCYx8Dk4uYTwsCXk1iSxyDFwMIABP//g+XFGJgdo1wVGOScg0IUGDRCgrwVGIygsoxgmgkIWaEsZgY2KIuFgZ2BIyczPZHBCCeZn5zDYIFMMjBATAVjJqBZUlAeBxCzMZwCy6oxsMDFpEA0WCUPQyWQVmaoApKGYHXKAFZmI2wAAAB4AS2HNwGDYBSEv3svPVMc/BpiIH2KgUwRxEL1hAVGFOCADtcPV6mELVhkP+C7tP5cFcDOW/eNm23GswfEisf7+eEOXW9NvwOvdAkoB1BtBRDwATBlDPMAeAFdywOMHQsAAMB5iD9qY3Oxatu2bdu2bdt2G9W2fd7YdeNe7YkxiCKGbGIiyCYeeYgKAnH58K/CyqqinS7GmeSce9K99DFSNNI8ejWaFE0JsgV5ggJBQlAuqBGMCA4kZEnI/vEj+FegiHKqaq+r8bY4777Qq8+72fedNcgV5Pu8y/6xIx/fowEfL/HxDHwcw8fRfCxFRraM+IeP4VwIJ4Ydw2lh1TBX2qm0YyKgDFqgE3oZgD24jBBAkjR/i5lnsaMuuG68RbbaZr/bZltnlt3mmOuB+x5a4Ijt9trhsS2e22mPXZ54ab4rLjrvkn0OSHHIKWs8td5hqc555LgTEj1z0CbpjjltrRtuumWhluqoq576GmjorkaaaKqZ5lpoZYjR+uirn/4GGOieQYYaZrgRRhpjlMYGW2KpFZZZbuUn89p7YwAAAAAUAEQAUgBWAAAADP8zAAwB5gAMAgYADAI+AAwCfgAMApAADALIAAx4AV1QRVoDMRROvT0BUsv7cKjs2WCZwd3ycLcz4E7tLI9d2M3BsFeXkfz6xYwQiVkjQkv60+UqoHH9vRihol8iJDz7e0kjXAkprUtFrgMW7gQbg8DMk5A2eXrsFd2FMiMz0ycZacuLwxPy9pSQg9MMpiWJVX3J45oGGsVwjZ4iDvM83uI83tI8GeQZriozMLLFE/xwyZeYleTpXdLLmu5VmEYVhgGkRc6SJkeFAZFb/tpOGW8u2yt7DiQ4GmQSLM+yqmk0TAIzmbLqArrPZMKZLqxqI5wWwyVajdGKYUTR4ZuwjOt+iSOGLggXjS7oAgwDKl47lJhd1ZYKA2DyHyTqa+QAAAAAAAH//wACeAGlvAd8G8eVODyzAAGwAURddOxi0TuxKKwAOymxSGIRSRValuTIiuz02BfX5Cy3FLfUixOlfne+L7Yv5VxkpRenJ7p+5zi5Oxel5+78pccCvzezgyWIgI7+99dP2Bk+7M689+a1eW8WqA2tI8QVuXcgDdKhdmRCdnQCPY7Q5Bpcsp6zCKGeyjoDcATAIW1l/QnUBh1ktvQ/gXRqT6/2DKx3FnUg/MsnUDf7GwbqOYssAMv1WkWzaI2aRbPemsWSCVurWArgdfx0LfT5m/HaTYu7CtrS7tO9Yz8dy3PvuHgVXv5RLY+/86PahdMnTpzGuzauvXYD0IORNXDxcfcA/ma0XMfeSJA1NmDfRQBdSKsCzARgZuTAN0aKeifpUVQ7s2dRO0UVRzM4KhmxCfOSXrLKVgm/OmARerQmwRJ46r+qKz/Y+ME67jxevrq//+ry8fGvcfdcfBXBTYMOAG4hwM2KvCiMDu7IWysBWBVkJPjGSlGQsk8AByVkU3nnAYQADj0v7VmhF1b4Kef9nN2m09v92G4zcklszZeKhUgS29XOgS/llwbFvoGZvaGxQ+W/+9uVXfPzV7xmZePQ6mu4eyyR4WTvHpO2c2EiN5Nz4j8bKJey/1z7RnV8uP+3CGFU2Pwl5+Heh0TUX6ejjaDd1kCHjwB8CpfbALcuiqUPeg4Fy0KFk/MOXp/BUtAI+Prhz1KZN8KfGa6QWjhVHTrgy7qriexcye8rzGZz00I2erRaOTmfxDN7blrL9gpFrxiePDpcOTYViUhyOp9ZvRnYTHldAF53ACOLO3K6gwA6CIZnUSfBjvLVXOehQ+FdmbHMfOCj7zkTW3jN7gfOvBb+zXL3PPD+hetX0hNvvvHG0xeBLQiTefEPYN5ONL/jrJ0E0En5QufTUL3Qwz2ddG31WeAjxaYTvjUwDbGLdtku2aE9gK+r/dvvfod7azdw90x9Z/oX00id+x+p7I/vOHc7AbS3mrt929xbM5olOt/vf78Ek/1gqvYUqq9/BNY/jOZ3XH+RAMTm9Rfp+pNZfXC3SGf1wayel5KJolSUzUwu0vteMTZx0Jd1jebze/sCvvK+olS1veVfhJNjo1ctpBTBkP1Ft1QXDLvlY5O1n4lJkA3GJy5P12jXJazRE6hdXZl2wFPD5ORxpD0P90GrP0+kBayXbJbgKh1Ywr3Ly7XvAsNqP8fWi6/Cxdo3lXnRQzCvBoV3nFdDABoyLx0TxlpiBgSjwOYv8Z3wfA9K1p/Xk9v16vOqqaMSraEcLfsxcLAIyIEwS8HoQUefVCws25OTuesHbo9Ko7P4q7XexGyfcJLxpgpzdL2EfdISgPalzKmBAAwK9zh4touaUxA7ZIAepUzGst4qafR2IFGDzRvf/vnhL7+acOwxvOv3tZfjlVvP12X6W4BPG0ruiA8TPEozx3TXTPRkCQdqdwH7HptG9XW/EcZyoVfuOJaOAHQNABcBuJC2hQY9oXhKSptR7Znhbh3t2WBkF6PXystlsp7mQqks6TWSJiqBaYb1PXpzTpu56egevVZbvCl7c0GrMcwDFz542234CEjOydTLU/fV/hqv3ged2rtVOtLUj+y8RmwFWjmVFhKuhY6B0UH8jYpzBhPB2cJ2/8+ntRr9nv2/mNZq9XsAz+N35E8V8BJg+v7b8ycKtb9GHLUNV4BtsKKoanuZqVURUlGmq2ajOtUBlsFP168N9LxILYGfIzY4mq/A3xmubhLwXOXE7nh894lK5cRsPD57opKalr1eeTrF2q7c+s179ty4msmu3riw5+b13K7I9LHK8NGJcHiC2ITpCPMRduCjEfGtLWaTRj0BMo5YQNAFlsDELIHqz+rogquQrKrT+HrpQFWSRtZL1yx9bv/S8spV3D229O5SYS7vqr2IZ0anpsu1PyDKt0PcaynfCmi2jk+ITB9qwMdNAG6CzznUizDgrk0CtKfuM0I9DDOAPY4yxDZRI7rFQUwMaSSa0TSyWE/NRPnNKhNZi39syvsOXucrrVWDdYaLw/sL/HLJIRrxbPWKmUhk5ooq4+3FZ9s6Vif0pbVrZ+rsn379ark9ldBp635ZAJ4bkPPS4zNV6SjRGGHUQYk2Mvf4BOIJjIktYzzWSKBu9UDI+tAr5hcX51+B7WeuXfp2fmU4uHf8L/F6dWS0cPFT2F77KXePNTUlzx8y154hOpYAW/tfsBbylo9wEBwcDVjmCCC3ZWtJwOhAOSofDkDNR2FhuMUBMNUSwypkscJ81dPpiFQHsF9DloIsziHpMnl0xuyL8Rt8LuIMlGfTg8eltH9XpjxuFSK2w958yB4YWCxMXpO8O5wezLkkl60z3uGNlUPhkaw7HSr5RDntFF1WQ6wrEC+HohOyr5AmtHnh8iKsgR6o2xbOq7SpjoWuAAe0ICpRbUSapKJIooJnnsT/8eQyt3tq6uKjSvyxCnqfofZ1aEdt6iaA7gaAjQBsChP1dePNBFZ1WmANVpec8T5B7Es494aq+/P51WoIzM+Ng3tzdntu7yC+pXbznqvGfb7xq/YALnQzoAFcutHIn4r+mpGr7wIQ6qbLZpatfszLFQwGXCOtn7R4rFqL13xy5fzlgMGZweMDA8cH8VFw0lQbEfd7mFdE11z6vIwHrSMnCuAJgN/CTKTG2gQP8bTnRaIq/yq22/DWKNsWcD38/sl2W7u209l5cO5gl7NL227pmNx7x/Fj7SaD1mA2XH4MyLq/dGWpdLKIr6jdX3i50gMb/5bIrgj8r72errcJLi/Q9Vb9ECZ44j+Ko17Kn5oIwKSQpoNnXZQgK4A6FYI0Mq/QUcWyxiqp+y/pO588tN7Nm7RGe9fq2qc+d+hIt6dHa3R3XbaGl/DwA3zS50vyD9Q+V/vEJ5yy3y87P0HldHMTLt8HvE1o5NLxbkbTpKj0NuTs0pe/sr7S4zdrzUHT8upXlmv4vrOhqXB4KnS29ooa4RlRwP+hMfow2h7yNftuRSV0io85D7f1AIaqFpaxCGqIIcpJ4/Xak/jDtb/Eh2tDec451XvxR9N0LzwANuwx/ALyoChI43Y6DDsvCJNGFuCY6IIQLOFZ1d7a4QYThbsAFqBo2aEXpD2CqqQEnxDuUK+j00crmOyozFJUov6nbHPQyN6oGbj7fe18xJ/c5RODRwfX907oOWlfurJvX29IqkrDy7Ljpr0rZj4VtAu2fkf86oO1fx7wxWdmEiFRuN3g8EeB1kWg9Q/c15ADRdAtTZpnaY7utojvIYAelXjVyKvRXQ+Txh4a0yl7bJ4S2QM9N+3poBdSCY8wI2bXE4L9XN2iQ7aAOADVD+sXFzlxIZ6d6XUFB+YSwV2CRj+2KFSdfjlsNwZ6g4PT3Ne+uOaDCP1gX+XohBTw5a7y8p7emXRsLOseU2xuBuj+OH6hxT7ccMn78PJL7sMzkZGVbG7SFXdkfWJfjHdE+0Sp3xEW96Vyy8PSC8OXjQUlvmh1evKTicREr8dnK/mCwdHLqAxmAb9/g3WxIQndsiOGzCJuLZSFACxI29I0G+CbNhattiELXRcLUOOkdBnVzIhBzYwQqgPbaI0W5T+SRIX8rASrkagI2qUJvcY3mzx09crUSHDIH+j3++SI44XBy8ZCfvfSuYt9/d746evXjorBGlmWCUIvXY9vUJ0LoX3N9LY0513wDc9iSvgGMO3ZhjPAmXwx7KlM6UU/blikcCFCcNdnwsMLicyuXrem9njb6pQ06PEJC/nelWoIL1f73dmIS/vC0GVjkjC0XPBY9hyym0s2vzS6cfnsHnu0P6zIlB8uVe4fQcn370gBWyBLq8QCXQxOJcJAe53nyV+wPGC8MCDrwzS3YJaK5WKhRJYEpz3eiRMnFsfGJnfxcTN3a9fG/HjtMP5IpW1+dn+7nubQ0psC/hnw1wxx2Tq6uo5fmMwebsBvkgAmkW4rNhuC7ycpn4ey9WxPGBCKn4cPQBIUMgQQGSAyQHaze86iRcp9nshJhaPRJSg0ONMClZoia5hoAV1+XNd6I/QcgxgMHwtkrMoILOR7/eyY1B9z5HPLRrHH7ZyJiBZX0OqM+vLjYZtk/OvjZt6bGw1bIuZuSyi/sbISHDnYn95d9H3dHvb1DCWTQz2+8GI0Y/aErEG3ps2aEAJ9vLYz6gtkurW28VRsJO3Ut1XNhUC0mnS2Gzo9Nt5fHg4OZzzmYA4v9PiTntLgYMmT9PeQtV+Ey79QfV1DOxlN1ay29hg6lqKEFaem0aSaRotqGuu+wEw5oTdLi4saaU6en15M5SID4UWwfMfF7JGN2rdwfHI4Eq59EHFoHCZ9hPtb1AlDptH2rYKh5X5NR2QmCVMGI4qQka0AKP4NTlF0wmcMbPBuP+/0+Zy8/3jNTcPGzX/djNJ5TKDF5SZfYmjOOuhgHrIrgHmYdVWDWB0RABb0N2KwtDi0iw91m9ud3SFngqDCiyK+cPHZ3WuGtoqmLZ/m3lRHqr4mF2BN1JyMmoLZeU0uPSeDGnIyVpKTiUJOZnGf5rsHP3L2wF0H93Ffq/kx+lLt+z95+RupfSC+Fv0r4GNCSfSnpwcmqRxxyANYMVmR4uKisV2r13caHOZsjPvaxXfYeriKVttXQoxm7n+oX3vPn5JDw85hM4ttWuf5OTWObmc9Iq+A+vmzyE5ltl119jq2nVM2QwFGEC9TOea3SbN+q10EDyLMpvqqpsh8enbXYgo2b4upLFzwhSkp25tKyHURn619kDUIqXoI9NvQFZdOP9PD1jkeIJcpJlNKRoxFJaZZKSn+0ryqlRTr7Uqp+rz/BlxpvqfZY7QIoGFJVOcgbfdrzEiWmzIWmXBlXzq9rxKutyu2cCEQKIRt9ZaGIcGxy4aHLhslwccQiUTAKScSk3kPRCaIxogy/gPg6QA8b0Hb/ZahOQNraC516VrWZ9rgWSsVIgeA2tW9P69WZNyqKIVUqqOKPwEHCGzfHiZuixF1+o+SNRD3JLMzOSVMHA9yYkWNEQPVUPyj3Kf6vDGIEvsrl09IfvfIVpAYFJQYTMb/RmVJQjfvuD5MVgzNIZeuyeezGIxFXrAJZdFYPf4yXVL8JZdl4vp3jL9GRM4wtqT1zSZaxV8+X+bi/EP7PdHt8RfTmw0aHyxeut40WwXzli14WlES0zYloZGyqiGu6YSP7+mymQLjLlCRtUypY0arzVdqXyO5FtCNWwCf3q08kp3MZm+YPkMAme15pAyyUyTshIcUFoJnMk15pGirNBK/lUWaFMZTuZwtELO5C3E3n6xEM/PeqKskZJIWf9TuLiY8fHo8WdiQXuEWUoLVa+8xtPNiKuAD+RL4gtWZ8JpdVpOhwyHmgv5CjI/4FPvkBLomuVeAJk828bmlOaI2p13dyxLWaqC1AbQboDSrVMUyrTfZHA2BbfHIzFzH5C23iIluf1ePLQuBu2gkIeEdd4zXXkj3gsPUdzoTQyHAaRZw+g6+wGxm66VW1b157Vs6TjPJqrKqdTv1mKAWWzU5meFq3tpLFGeXkmArqWAEZ+WFaXyh9i+Tw9Ek7NLdc7HckQ2Ff11w+Sq+0CI/1bwrYoCd81Pb8zxdK4c6+U5tp6Pz0L4Hj8DsPwvNSNJMCNtqbrJuCHGPwryQn7r0edlCGi41PyWy0r8JQDaWn0JIuMT8lHN5xmDUaw097bN759rNkJMyGqYX/vz4VLupHaAdE4SqH0pg7seD2KX0xiTouXGbBEnoSan2IuWxES4fxxda5KdaWL1LzE9B7MlqO5jVdsrRhhSQnm/ITz309vXRTmc3WYvB/W8/sz7d7TZqu51dY/trPzpli9vg/6n/+dWrHSm7Pcm/msnFZpbKhQ9NoW2B2x/vvZpxbtzsdCMDQ4+nxqKKNY3cNmqoq+k12NoTRVPnl1aOdbpAZmwdq3sf6xZSfWGzt6vnNq5tMB3CP6z9f8J0UJwRcPfFF0KVtLutbVTBNQCXu/GFFrksw/9hLktiuawARs/huU2EDbV7Uvja8VTtThL0b24q5zo0eS6CtADQo//EppbwZ3eAP78NLqvwH26Dp1X4T7bBT6nwnwEcQEAS0rTRuqQd3YC26eelF0a71SAcemqBuU11ngrDupn/IUIH60rvs2XrgRs2SzoTbrOLRUgw4fopAXyvtlga5oTaW79zbA2v4ftrP/nRjzCH506duubc9Oc+p5wbAPoKm0lyjgPo01H6foFXKd30HAPln4Hx29sS/uwO8Oe3wdMq/CcMTv49Su/vUsZHt1D4Xrhf1wB/VoFv/gHgnQ3w5xn8RYB3aNIq/CcKXKmR0nU2s3WeonBam6TjWBhdcy3hz+4Af34bXFbhP9wGT6vwnxA4SFF+8xnuFPcw6gR5yaFb6xKTIvKQapAYPwH4FQFB0EnBcqtRAdzppwLTxXowAAiHkaVkyd00ElXvE1VYSO1l2LcsjqDZiCyOKJk/Hm+lLvTRcj2pwZf5866RCKkHmnuC8lxqRvYG8Z8FglB6HS6sS9XMlED6lSunq5nXuX2k2JrbK+RLpNp6ZAK/OfnaGVKFFYWl0MHUNbQ0WN0nkZonMMcPemRBHrTSZEVa1WAV2+FktoOciNKwimxHtp5qMqrxdxepV8G2u+HwDN1dQEbGDNFS4YG/wB+gmiHUlmZfv5jcf921udy11+HZ9z9Az9WkV65fuOlOznCari2tb1FZcDLZGWsJf3YH+PPb4GkV/hMFrtQl6P1eJvu/oLZGhMtueibDg9brPPIQlnhQT4vMdb1a52EmhHDJgxwql9pZtf1x5ARb3A6tC/gUZnVTyihr/ewTCdLN2EjPO73znfvY+ac9e+iZp6ezF7k55QwUHk1qsQATN9TVe5AL9beqgjWl3s/BfZgsb/Ic4qDnhB5NE23V0+3S1iGspWv+uKD+CltmV6kwn3e+66mR6elS7UXCZ1r/4L4B/PRTfv4H7kat4M/sAH9uG/whFX5hG/xhFf7jbfB/UeE/JXCkIXDup/gFoHQETaP70U7u3EkAzgZAngDyhFXbzQYF5NVTXH0oz7Lhfdn6acsADJSisEAWnmMnBJUU7TDtOaE3QXsB6E1Tp8J2ZBDJQHlGU09yqJlnjRqgR6L0Wtwq87QpyYO39IaCe0VDBx7YJ/NDwSOl0pTJty+fnsw6NbWvYFesFJD6PH5hLj+86M1fka3su+s9BigCxWZ8nDzYJ+/Sn06EoiFr0esiFR7sdMTLuVxZGFopeC2RSsrJm/tt/nIonZkZU4tDx2f3Ls0o8UJmM0nqIrAGAl2Dn+MeBof8PF3jIJOJ9pbwZ3aAP7cN/rAK/zGDEy/0aXr/Cejp0H9cjxSZgOYnDfBnFPjm76F5sQH+HIP/EiHMkfEZ/MfXq3kh9FF8ocUZsuaExdYZssVF2JbW3BQPmleispxkshxS6flvikeK8cXUEv7MDvDntsEfUuEXtsEfVuE/JnCkAfiPIXd0kWRRUGZrjxwnJMQbaPISgJf5RejEqV/kmHWLIy+V8biafy2zVH0Lz9bo2Bzg2T4ZLQWKUbvTE+4LFCMOHodcPpJn2jPUGy4p3ciMLx86HIUkQ3w8G415eifj4xnsFlZLJN3UO+QbFVZKJAPlcg37GF2/1Ig0JzaExtGr0fZIfYuuPgLo28qTKgVVE+uBTp6He6CVwddFQJtlSmckCxC4L6Xq7QjtRaA3TulvzqDpKbk7arFOLSLp9H9uC+X9/nzIVm+XQ0N7U30bbu96KT2Vc2lqX8CueDkgDYD+LuT7N7z916X2DEncxyp98m79C02ZtwThS2Wof7RJd0f63wB5uiNUZWlOmZO4v0dldLwpCtjKKZcJoNwyvQs7HuhoqFAIVCgk9vcTKMt6SuiQZ8dMS8y5wMevrYc8amZFWyyoZRpQoJdtRIcTdq/k9aXklM+RGAwnJgKiW/YlEhTiEb3xRGlwkdNsvJqPFf2RYirudgoud6gUE0ox3s0POOw5wRfxuoS0FMylMkKkd36h1kV1g+YtqC7lme71tIQ/swP8uW3wh1X4jxmc7inp/SVlHPRCA/xhFf5jAkccKsI4z3NPIRvq24o3ZMJrGVlaVJXPkXwmSTMnn0A5lKaVEJ96ZigN0msHKc5BKz+ttCUab1AOR7eKzta6hma4unZmMNFWujCgqdgUzAfNGItjQbEQsjqi+VNDc4H+Pb2JhaGQUJ5NRabcxnLkbc6bC0OR0JRwl7PPV9iVMnq7U35najAIh9qdXHj4gWuHjk1HheHVkrw0KAjByqHIrvhoTCjYUp5dcYRYrpidJ4ii1dY7cdXSqkxR8wutcqb1TLBBrebykvmPTgnwxeb87yIWIG+0K0/Tv9KYiLenf7+teecXl9zhevrX61565/b8L63ncgehnhtDN+wYc7ACqqVlyB1W67kWWDyRJPDgYwGoSa3t6uFjhG87zit3maENQ+s4ryQwm6q+2yq/eqJn8BX7k9vtc09CIfgbUAqenXNmLWa/FIp9o6kqrFSGR+bWOgyVbjlXYTViQi/YkkmgN4nuRNuDzeaUVut6WfNJbOWsR7eSxz9P/oKsELSmLIyjZvgJ1A/QIIupJZbniFHiwyyd2YJyILzcmPnkJlma8+F6KdwChfFPdkycPh0kWVCdkvDE3Q218SpwRc2IEv8jAA8WufcB1h7wqm+t86FAiCo0ROFBAggibUtHW0UBqsxBNQ1qAEiVvRsTRAX2bgysNih2Ae5y0btS8K2Z3jWk9qbYk8w1KdGlEVtt4JM0iqXFFFJX/y0fzY6ngu8qy0ZMeq+ySVlPsOLLrUtSdt5sdMQDFp8YTuGbpblY5PBAelfRZw9lXTbeEesLDk+4E0WPsJr1Z03ztr5osOL5G09aMHtdieNF7zBOewb8poAYdviSpYtf7wh7g3FPbjwmljJRc3w+RM4czgwIpWzcJm9E40f6R68p8IJBsRMpsJVf4d4MZPpQsSkTbmx51rlLLSg6FCvQUBnefs7Z8bvLr7rq8iNXXXXEmxYtFjHt9WZEy2TXgx/60AMPfOhDD44FRq+Ymjpe9furx6feejOV/T1weYy7CWYZb3pHwNhC1OtZOD1go4VeXc1xD5H0XC8t0hC2g6ES99z3utSou3q69p4J/M9FA99z8Uml5hKCy5eBBz403HTKzNhCrOr6wp9XcoEmNZ9eZoKgnhUj617Bg9guxmZeNRcbGo5NeLOxA9X1l0UnL+t397nO1o73zrzu2FK0PJUWsqniiZX0wctPDnLaaYKXG9bms4CXjJaads3Gl6y9eFBGfemrS91J25QqDEA6m6owUpEum4q5+u6Vpu7OohO++f6B3fbcQr88mbQMyrFxXyq83p+Ykr0jBV8+Yhf751KRaiHec9qfqZa82aClahFSrmwp5CtIaUd8KDqw26bVe+IlMTGRdZlcgoXKnxkI/SdY73bAb7XJuhtbFA7OkWwREcWkYqa1oLiIrDYz353MevWcV+BW4qnJ9o6mtNskO6RqixjkAec0nLbLbXwnvnp1ofaf2Hxo9YWz7mk3n+KfGvv4x0v4PbXXkRw5rMEXAL8Imm6yxcYW1YT6GvCohxXk+SzgBDj2sEo14/sWv62iXWI+VOX2qHe+j9QKd1fErC9lx9XfmPkMnBgsBUJD84lIVY73vAngkZI4tttm9WJ5+jNdRmfvbKGwt+wzOQNmRbdzgPvXQX66UWZLm5IE2WQr7Fkew0B5K6rxT1J9oyy8hbYfW5W4h6LuaA57cJcz5jcPDPmnBcP0521lvziU8WL/3ujSMX9hIprdOxiMTazmCrOeoi0sC9WZaHrjcvwD3uMdOTEXTh49lpkt+SPzr50/cvtKNOoDrFKk3gTrYEJyK7vQZKXgS2byFTsgVzRlct7UqNFLn9A6k+PZ2hfx/f3TCat25cC7br9pbFqeOX3nXxxW3nOMA9++BnwzIR+aRKtNuc6t+UYJYJSuu1plG2XZiu6sIoMpZicBQnwqkQRoS08zN8LKoEZcrmD1VBXd7KgnsphgNKrja+3hvK9v0purBiPy3wrSQDE94IwX/9MuJnmp5M2NRRNjWZfZFTDZYwGrvzSXSc+XhY95UqKlkg+kRI/RPXmrb0TKjfZlfXHB9eVAMSl2aV3Tcngo6XSlq2FXJiZ0arsC0bwQqmTc7uwI4lAM+PL5Jnli4qNyRRUwKk+IaJAqT7omebJSJ8p8hV5RC6WizghXo8qfBkCQZob5OAjVAohTUBGn2NJRvzwZye4bFIk4yfNuXIpmNo6QQxXVmW/yXu/olfOR1HZxivgIVqObvwQKXgPW0olanADUkAWqezXAYFRyOSXJ6ZLyPknyeSSJjJHd3I+ehDHMdIymV7mI2oNzhGXVSY2nt7KJnEbvlEIuITw1fJ21Esc+rydQSI8ecZM9DcULf4+LottgipNIBy2B07m2wW+nOr4C93+LOwNzmpEbtTiq0q6wOipDZVLfQNFJbonr7A/U6eJGNjZqt71DpU4ZewZ9C7/QcmzD1tgBLEFNrmzbovNxNjaMS8Y/sLGBr1OH5lAE5OgN3P9Lz3XHmk5AUrExq2/IhLcOXkQq2mbvquUbKfqgzuyPBvnenDjIZ4XF0sJhpxQVLdrDQRfBwxV8MlotZtIZT6w/4EmGsiuzYimXzfdLKmrUZg7R8w6tcDMQ3DoAN3MTbthh1NrVsr8iv9poI0dwe1sPQS7XKw7wuTpygkW7UWcSFih2WXesz+9JSrkW2BGZ24v/Bj3AfRLWpBOREpwWAly6vCYsLWnWND15bv+hQ39FZQbuxWtwbwTdRfbHcP04HaMfP4jOcRD0giafQ21s7wujGLGdsZPlRM9FIhqdTfTzfrFU4O519SWx2+X2ZVJDa34Fnyrg8wXuHBurEyBtylh/JPjVSFjTZhf8vC9YLuB/cvWlsNvp9mdTQ6t+RPEFvPApwCuC7iH4wvXjFA5z4Fdz5wB+L4HDlcE3X4W+gG7gCARd/EIdhl8N+8Qog9G7oVXeXRhD/4YfRTrGOx3lXblY5vW8Pn1H6JprQt+v3O9/3xmvQlsJ7n+m4f425f5ombfrow9eI915p/Ti+/z3V+73UvzJ+PhauD+C3kXwhOsZCi8B/I0U/m4Ch+sZOn4cfQ+bsBtpKO8wwoirVw6k+IUL2F1VxmX3cTr0Xvbcv+PHsLP5OfrGcxzfVXsNdvaz+7jf7HQf952LebiP2hysAZvzJk5PbYuJzhFDd2MDfiOyUto7AEZpb3qv8T18LGCxBGK8I07a+N1mIeXxpEk/7fGkBDM7E/Dn+EebZxkeBIIpHqToi39UC02Xy+xc5l7czj2NjHROHZuTVw4o8/QkMq//58rUVEUe6O8fkD9+/AenT3//CueR7732td874qRjRDb3ou+rY+hhRjoG8a40C0MkMRqhj9OB3sMevuL7p0//4LgyxsbmldjMfRme5pl+6ADf5hUyw2fjU69//ae4L0MZKUvfb74S/Z36HGK6cA5pVYp52I7I8Hn76z75yZNcLfviR8lzQTZfL0qq/CHPeaH10V4P9Ew0PxWDlrzTQbjHHGVjCA+cgu0GDSMc6u8nkD9kbE4PDmg0GoMVzGHQamgfGEx74h6dVxC8Ok/8ZLZv3tObybhMHoexYLR7jfJaeb4vm+jNLMu9cru2vZCTlzO9CYVO3A34yiq+vRRLB/SstCcSqii+HdCKFF+eFsDUs+ZKNazINhtkae00EKhXzIr9cU+bRxDgEnen/CmfsT2Ud4k2A9Bg9AHgJEUtVwDU5F6KWrbbL0fLa7IviwtGh8fkymR6PVHZ351V9Llj8wjexz0Jcuhsde6XA3+NZSxh+1DtV0Pcky++F545tKlD7+J+gTREnuj6U3sLi3hoYID7xYtOgBlh3OWXHtcK4xoHcecgHRZxDePq2cht6sggJWYYnYxfK87Xcvi8gv8A/ii+EnJ6RmRrijcstFRCLFNZ5qk5i+qP9l9uell7b/sJ0+V90elp/IWNWMZ11SlnJrbh27+f4JDe/Dzsex8EiZqmEpulek/Wb4KtJMT8FcwiVJrhrB/S0xLh8uNWeedSQ3ECj/gCsaHpgE+IDU/POWHvF45XpivxbMTY40wOSqW1QN41EIul48Mzw/FcxGS2RvsiuXnfF4wRMZgSvZaeWEBMCj58yuYOuz0hvz85nMkNOsxRN2T27REx7fTGfN6I35cYzqRG7D0xiFVD9iBP6Itvfhyfp79X0IPCf+r9aACQ5TdLGmbjoMSOb1heXi0dHo+ERtcKKXgRX8B5V3Gxf3Cf7Kgt0r1C7+avuMe4t9Gzw5NIBi4OkOiScjENvQINFdI09q/vSEZJxfpp9VDmnzg7zDcdkGw8XdhbWLtmdPT164XC+utHR69ZK2RTs8f7+4/PpurtywLyeCgx6JRscXd5fLQk5Bx+byUdGct78fjcdfszmf3Xzc1fv5rNrl4/P3ZqPpGYPzU2emouHp87NZqcyrtFZ9Rmnx0cm3dZE07RI+8isqhFPK7ie7ivszPTw3SXVMiSwhe0QJ/t6fr+MU7yik/X67l5VKC1L1IUPotKFJaFNk9g9Mne80odbFDhzp94k17DogteikqkyK4cAfiH6AR5YXIiGp0k7WQ0UI7xfKwcEEpRno+WYnNCxmbOZKR4fDxp4b246ivMpFIzBV+9DfOJwZA0mHQSMQ0NJnhs8IUsAd3S1NTLgibLOJEvefMKjZe7DdZ9GVVh5QfVlY9Cz0R7u6DnoD0/id5or41ZzCdQECDsLEvTjoedxtw6bazXyU2kUzlQCl9UKKpY2UzJ46999/LcG1Yy4YHpYLy/jXONiPKkJz+dcseMXGc56l92ZqNOaerKifGbjg6m9149klkMOlZO7n7jZWVsiPntabfP6Lf5s6L5m8c/+rpqbvW63WPHxoORiC86MxlbGk/wrtjXXu8tzvcOndidKB+/a23XaxbiLluvnDl422ou4Q18gLcK+SHgEdko/QPsGT0ogO5vqkM0Z/QA0HwYpOWvBmhYBugc0kNPSzVMz1L3pDVCq2U5IA20FpYLsrHWyeBedp+PZArBMdOPrKcfu0Q/Uhk+Vri8du+qZeUQX+RvdRad+0jfVXSedpWespy2fKv/HQOPwb+Bd/R/61vfwm3veAdSzsngx9HPNBVOjx5DCPoEFsOP4xHNboA9rsIkiHurFHZWhSXgvr302ScUGOEl7Jvm8X3gL3i6N9FlXzI9a2mRni3VkwyOxyf27ZsYX1wcj+RykWgu13XV8WOnTh07fpW8Z2Fhfn5hYY/yTjxcf4Zvg9UKshmbzgEqcyFFWVmyVXpwRUzZCiO4kNBZTbXrqF9BeARwd6MUopWH5gooO/J0KSlVyOUW0lMZZ0xyJcwRfjw2OO2IFgP/T+2X4fhoKeuIpESnFEyMVLyFYp9AY0vKuyrMn0AVmJ+vzx8l80fJ/CRUaafVgig7kMJeZmTHUNpRx7a0aUQqQtOUNG30iQuWfEDIjPTL5ajgiIHDGo670kFbxC/Kjojsze3mL/cEQsGgFI/5vUGPLeiTjJ6oS8oaDb1RTzLQkxQRB/givBf43waIzCHirVr8ogTa9osSnPoDP23nlWNSevbuYhdTANN5uu2RTJgnCVGpSOo3T8OZ/+FnBwY++dGZgXdNXb3vaObKK8WL/w2yGMYcTnDLnA59hsqmH/5eoH9/lsol+x7WMQg2kIgjSZKeQzoWv8JfW6fnWTUVAcTNcqUIuZWj1cqmM4O31tuH6cH0aL5cpPH/HenSZVdit+Qu23jXZHSonNdrbfnuaHJ+GLAqJIvlDiFh7070eKZGvhy2ePKxlfaI1+2keDK8kRWW04PIWTKCa+Ovt7hb+5wGtXk0UE44nYlyoN4mMplEIpvFHJ+sxmLVJF9v5dFiaWSkVBylc8fxQ1jiHiJz01jLptqwDui5offS8/6lKy3ZbFLaVW/FREKED37IAjliP+TbLcGs35cVLeDVwplMWMrQed3oBJyXSSMRFVGs6Z3Djnrp2w/TK6/p9lIGVDGN0Ek8WVLOmIBcUy+rpwWuqL45SHGPmK1W80hqIlf1DfATqVGzzWYeTU3wA76R7MSQVUiBIxWsVgEcakqwvmOUH5CkQX4sPRW2WqbSY/ygJA3wo+lJqzUy+Yy/N2S1hnr9rFV+r+s+/Aj3M5RGh9F23TW2eK9pS4lhHwmgqPLuO/MJVlX2fPC8lX4Xhe8SoBZ4W4QLypHBTdFtUbb7MY9v8BbjmVK8sqsSX5oZyh0QIsJ8pq9MAXvGhnLjHq7L4y+kQhlJzE6WZg51XHGsLenr90b7srFiNJiq9o6vdx4/1pag+cgiQrQOl0LrTe/ZGFuciwKDq75Tw1H8Q1nF/4WANkktandmt97AgVge9tvKRphofAY3v3qjvvoLhOOv1F4tl7Wa4+OWDjM5DVOadeesuWCp39T+gfs4jeCfns57L+s3eowFrTAsJCZy7qSQ5f2jBWFQ6DN6TMWpgFbZt/BweQpy9z50Gm0/fGFsfq/F2PKdXC3ysfcwXKxHPAeIsHrMqPs8tMy1O1nq3dlTfw/STL9Vf6Kt4RcvgGA7fKgRlMlHc/WCydejNftMC8sfvesjH/nI4oc//OG3fYS76WPiuCSNix+rfT47mjtzJjeaxSOENvArP4V1K6Krm85IGVtU4pU9QYaqW4ZVrdLQdrA2/LQSrYTVt+2sDHVBfQk531iPJysZKRSHcBk6Fc3WahInqbPZvVhZUxIr4p+Ky+X0/FC8fWQQ48GR9kR1qTe94O6154WxfV21Wg3jp/6pe74SLrv3pXrFkYODSW27pd3QY0i1DR4eDabFrFs8utqrNxo6rO25lSMhidqY4c1/x0/CHkiHwqgMti3I3NE5omVIZPl/xbUHodURt8QWzQ0tT/SOmvcANFTp6GthTNWAoKgZ4gqsK49U+j/vm4kWyBvkgdLuRCE1wsdsabdQsOHvd6+Nja91t7uDmhNXtoWHlouFxaGw7uUnuKCjYHFwtQc4j4X+fso65Mz/wL2vxX6dmg4M+3U9lvFf4Zf31d7VxT1wcY3Gc/XnwO89Sf4mMYXGAeNE0aPkFD+0T+JX0wzfo/jrCNF7RuAcw8/pPY9t/pjdc5De8wR+f8M9v2u6Z4re8zgmWLG5uBfpPWfVud7IxnmQ3EMzLlfi73IpWsvfipKVWFfHXL+BnCgnsi4rcv+vjz46Cv9xsvKlL1W+xMbBSfxdfLJl3jCIH6ot4ZMVcl8e3YQ/x9lRJ81gtNHwiTiPaLmeg8jff3/q/vem4Pre+z/2gTPpD3wgfeYD6Q/S9241gO8RwFem8awHiTCXCUa1U8nRqf7R1eAft0JZTUPiPwp4/Y4XBJ4PBBboleeFD1JEv+Z38n4/7/TXvqO0/p9WKD+D6AtA4z9yEfRtdArp4PouxsNbACc70GKnWRkD0AS0QYuVGCVsN2F7ANvDwIl/uO22f4BJrn5geiGvzS9MP4BajAFPsjEwe1MdtxWruJjFxTb83doSGQSCg2TjIByMgdhamlAf+qMzTTsvL9zFDpFuX2g9MIkt9qOUNbnKZz9b+cRoRcF5CHvBv54BnN3NmZLGV+CtEv19W+kfd71nZrRXmx+d4c7Ubvr4rbd+HMYIYYy/jd8OqxlF5yjdJoqgieUDNKpfMjGttyqBc0NGg5iukGdM8GfjEbsjYs409HF/LuEW3D4hmIvTlsk8kdU2VVa1TFYhzUqkdaz2Tnyytvklml/uw/+ILwB+IdT89qRawrUhM30duPXrkVHeLi2St8n7yRuSnG8y4cuY2xa3vSQZTZvdB+4FGfsgN4z3ct+on0amcvcdgN1EYc+osLsAdg+FPafCTnPj+L3cQwC7oMLu4Obwu7iHAfZjFfYObh2/mbsPYD9SYZ/k/Pgq7l8A9lMV9vlNQA5p6Elyu2IvNu/lhtEozKvaQdVVddXtoEbmpVfuP1DlvvE25ZkbuS8iWZNX86EakG3ltls21iY1+VfBPZ+DcV+5fVwmS3RcjowrlWWuemA/N8zG/TaMe4s6Lsdy/1a4zTK5tsF9kYx7M4z7JsqnctNJJRi3BQG6HgA+rbxqqqNLSvAkP2QUnLhsaFw7PgRkHTxI5/8zmP8OTZ7m7iFmpDqrjKB/GkYFWdWUlYfvOLL3yDg8DDi9amKCPPt6bhy9BdbKgUZb0duougA4h6xk9G0/AKgocSeZpa4RZbmuEHrpLdpY2Cp4nF3e9gzpBjzObuhy4+nhHmuPrFwVPr6S+yq6VyMDFRLQgZCZ2Z4uaLdMRBe0NqJ/TOMkVQ/JtEM2weM2Gh3aTFsq1NDnvtpj64H86jA0HlemQua7lVtH13P30XUj3K+LjV26/te/5u7rR4jbfAM3h24HuaV5i1Yc0rFDPF3s1WoLYUSDsZcb+vsdHo/D7vW+Gz526HNzXjt0Gz5ImfOb6HZNWvEtajXp/2Kab9KGgFiL6Dwf4fzoMOgamccK83TRech8FirB5h3SL/FCIU4+Nq/XBh/OLycS+XwiIYf9bpff73L76+OfQ4c1p/5X48eKxZjV67WS8c9l47FcLhbPhjwu3uvlXR5FXj62mUQTsHpmOjrH5KUTddDRozJLaOijysbBWoRStTNW9BpivaJfUzQIAYfg6NbE23ibzW5oMSZcOTpmB+r8344J9ov9dg7EXhGE/n/1rmepAAEAAAABDMx3DJjcXw889QAJA+gAAAAAzZeApQAAAADNl+MW/0D+vQSIA7gAAAAJAAIAAAAAAAB4AWNgZGBgvvHvPQMDy6r/Dv/FWDqAIqjgIwClcgdKAHgBbdMDjF9BGATw+fZs27Zt27UR1Ahq23ZY241RhLXtBkVQm6+TzfkuyS/zPe7OH2o1+gMALhGpQJioWgxQOUhVI5kOGCDvmKN5vIqzBb1AgCphvuL5+7SUtvF6YXOWMd0RrTzhqxajrzIBTJzhoQQO8sQwlA/iFJCtYtBDLiCemSAfEC+h8JcfvBaIHviGEnw3Hsh7zgZ6mOShhwohf/B+ZjifGcjnd8NXesCTz9TIbdiqa/CUM7AXznIQAbpPN0zSYak7sl977AdqFAvjN/1p7dZVskkDk/3a0/3mIEhV8Dl27I6JE7KFvdm5A3a2pWx8M37RF86tnbtjUsEs5HV2b093JxWENP1Z5cNf9aDhCJSPiKUGCiVvciIvSmy+FkWR+IkiSiCdEoTe2kGEUyLPNTIbmVlUwDmLCqBnckAcpctR5lFmBqK0F4gyCUQU+F5JRKR8ga0koAeF0xAKoGCdQ2CNjxjEtNe5C9nyEXEqEVHqDZJkPzz4HaTwXIhMRF+KpBCKpq7n+iNMxsO/NWt0V29KpTTyaJaHPkZ/kEkmQkzsUGgSSb4I4bXgZsk6u8qlUApW4fq720HXaA0tomW0iU7ROcBYT7PpLF2luTSNJtNEWkzTm+1udli7BID/k/8nHMTsAAAAeAEtwQPAtDAAANCz1artWLvOa9dv27Zt27Zt27Zt27ZtG997KpWqbsrVaqhuqJ6rvq2Bmvaa1ZrTWo22qLa19rwurOus261Pqx+of2/AhoaGuUa3sb1xrfGxqabpqpmY51p0lrJWjbW0tb/1uI2zFbQdtPvtre3PHXUdS53QWdu50fnW1di1l3EzJZmqTEOmLdOTGcpMZL4CAwBAAARkBHlBSVAVNARtQU8wFEwEc8FKNis7mp3OLmbXs7vZ41yYa8y153pzw7nJ3HxuNd+Qb8v35IfyE/m5/Ep+K3+QP8u/5X9CC4RQggrMCheinGg7OozOo9voOfrs1rkHup976nseeqd67/r8vsa+q37iH+x/GQgHVgtQyCzUFVoKXYXFwnlREtuKg8Xx4lbxtvhc/Iw12IXjODcujivj5rg9Hoyn44X4OL4d1ATDwdzB6sH1kksiUnapotRamhwyhbKGNoYzh49GCkfKRqpHGkZaR5ZGTkYd0abR5dH10e0xV6xkbGjsZbxm/Hgia2I74UhqkpkUJdVJS9KbzCebyWUZypKcX24oz5R3y1+pixamZWl12pC2pl1pfzqSTqZr6X56k75OCsnsyb7JrcnjikHJqjRVpir7lZd/fvwNsPeMZgAAAAEAAADxAFoABwBxAAUAAQAAAAAACgAAAgABcwADAAF4AY2Og07GYRjFf9ndQW7I9pBt1/TZxh/zuuymzp5l79V5jvYCTdxTR019CzV0wzOupVOTYd2HjD7j+neeBh7wnnEjwzw+4052awa5IE+VEmGi9GoKkqMsdEpJyrnYuPSM+BJTjDPJnM4Ka2yyzpbQ+4bn/HN67GP+OXViuWspJTmT8uXo/dB9qp7eHxxaciTEVQib4vCs6SyYmiVImqh5YoyTUUeIaaE5O0vW8v9/J6hoFVhmQsu1NS4lIm+IqHBYKCutgi9fVChjDUG945bPknkC7V1CCAB4AWzBRQHCAAAAwNtWBXeG2w93d2ISjARQgDsh8H2L/fMhCINIJCEpJS0jKyevoKikrKIqVlPX0NTS1tHV0zcwNDI2MTUzt7C0sraxtbN3cHRydnF1c/fw9PoRBA/dQihgAADne+/atm/9jLzMWmbb9sl2q2zbtnky19m1bsY+t9ywX0+9LNPbHX3cdNsDd91zX1+PPfTIAf0sd9AzTzzV3ycffDHQAIMMMdhQGw0zwnAjjTLGaGON89x4E00wyRSTbXLIfNNMNd0Mn5320ddIiESp8hUoVKRYiVJlylXIla1SlQRb1NksTaLjTjjqmKuuOe+CXXbLiSSHHXHFXM2ly9BYU111l6ebJt7LNMc8Cy1QLymSIyVSIy3SIyMyIyuyrbM+ciLXTOckm21N5JnlbORL8cdvf2X55ruTalSrtVIzPWyNAi0s9sJLS7zy1rsojKIojpIojbIoj4qojKqojpqojbqojwbRMKFN51at/mvR4f++3f41NRZGDMMwACyTZgnjcXGPMrdhXD5gmRLb/ye8dzedZ/+Xrm/2i+B/U5MsTBZpERBZpc/4hrHlPchitPoWB+r1H/z6BpS+oRdsIIoNvc1OpG+GMyeHLA76laZpOLh6i7R0HSjVyymkMQMopZgJlFLMAkopZlP6QCnlNkApYqYHlDymb/TpS03jTwtov73SAAAAsAArALIBAQIrAbICAgIrAbcCRDYqIRQACCu3A0A2KiEUAAgrALcBUUM0JBcACCsAsgQIByuwACBFfWkYREuwYFJYsAEbsABZsAGOAA==) format('woff')}@font-face{font-family:'Source Sans Pro';font-weight:600;src:url(data:application/font-woff;base64,d09GRgABAAAAAEPkABEAAAAAh/gAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABgAAAADYAAABAA0QDckdQT1MAAAG4AAAHKwAAGWjKVuYJR1NVQgAACOQAAACJAAAA4PFn1ldPUy8yAAAJcAAAAFMAAABgW3yVKGNtYXAAAAnEAAABRQAAAebzMPm1Y3Z0IAAACwwAAAAoAAAAKA27ATZmcGdtAAALNAAAAQIAAAFzBlmcN2dhc3AAAAw4AAAACAAAAAj//wADZ2x5ZgAADEAAADAwAABc4JLzHaJoZWFkAAA8cAAAADYAAAA2/iCz02hoZWEAADyoAAAAHwAAACQHrwOeaG10eAAAPMgAAAHjAAADxLPOIg9sb2NhAAA+rAAAAeEAAAHkDsgmwm1heHAAAECQAAAAIAAAACADCwIzbmFtZQAAQLAAAADgAAAB3CQ5P0Rwb3N0AABBkAAAAgUAAAM5bFBIb3ByZXAAAEOYAAAASwAAAEte3My2eAENwbEBQDAQAMADgH10RmAMHSUyaqbJ36mwAmqj1qayx8rhVLli5faofLHyS2o5VgWBWga3AAB4AbSOM8BXcRiFn3s//G1jybZbclNbmLNda1rD1pptt2S3ZNu23dvJNYbv5/ec5z334gB+GtIRt2PnLt0ID+01djh5SgHM5PPzrekO6Td6OOGvJ0jRFuGCtt9pq6q6auR2ZiDjmc1adnOW27xVgN/JOpWdujjyfSTIUaAJzejBSKYwjZksYgk78VKwmzSxGzSzK/QjQn8qM9Mes8j2sooQZYiQ+1jqO6k7pWbwqueV1Ef0IUw/atKf1iLeiDgnoimlIu6JeCn1NMt0l5Kzt/Lvq3onpiplUp4rX4xUMVKzlIg5zpIvXQe/51Aq5Qqr5DjqOomrc6N694vPySloN1OKS3+CPCb6hTisbmlMsPdM0naTUymF5MV0XVpRH+yxrbe7OjdTpKMILxU47OR/zrtne4Ei/3nYVtC58TftpuaF/5A9345oTgea/tDW22Gd2+y4Xtf/wzdeEv3xPv39/7Vv22qidv+vMk/bs0/cmDWYKzcQx8PMzMxUp0pfpi9Spg1U6cLMTbgJMzMz02Omtc+89ls7+8nJPn/fL7uTsU5aK2e/C2dM0mo085f0nxndFXyRdiaMEWEbRjyKMOJJ1lDnaTsvIeIT3iEhCVtWbJnaNiR2ZmY1evJdyT8WvXrVOWSYoO2+jKXEnu06hs3atvjz6N7TzsssCsFBNImfeF7Pn+cLjVI/mYnbCYaOoO5QY5O0emNrpL99U8u/v2fgrDcttIlkpMNqTNknN7GZL3hR1rWB5TwyRkkkK/5Jd/UiXuIBHrDzXmEVt/MITZZL/0eqwT0Z+RzDuOfG0uJbZ+7C0OpYoe20U9+24HN1DQOxLx/6zonuGdpv4cpGKmopKdlOAyfltekXnJ/h/H6ZeCIxQ7WIcO5mDVv4kTrv0edtrsufrdDccwivTHB4jpgtvMR3rOBz3mHNhG3jMKePmeci33mazcB+G1Kfvb7ws21tKE5XXsI4vnLUTpAoDYjGpjC5jJjPJBqT/PsBEno0SXgr7y3hIfGXC9/vcAxfeozQs2MDb/MRa1jCd3xBxVnRiKTQZKBIR1QwxAyI2Spn0FP/h9hZI4tN5lClGs5V1G2/E14zP07Ec3sqZ+JxbpL70C4TMdT02LhvgKUyh9rveohEZwFh3SJqw9002SB54niqxFzrDdt44eLS8xfyz13cS4MlC9YGQ4TN9W60MrLtKjUyh6kdlCfKmV1kzJdz5HMOqVo/1mGZ1Avr8by8fwFDzSnn73BEfuc7Nm/tkTMzw1ZFGT1LepH1fhaZH1f0dETzINtyu6fZ0dWhWFR98a9RMcBg1NeeHk9ayuJhMUZazqwW74eKMsXqUFdWTxGNDD+HTBG+0GgW72T2VPUsfZ46mLoFboY6q60sH4UqPnO2dZ+9o6wjCd2rWMJVXn/KCjByQ96HlDfp6l7X5PlHqrIHg5wr3t4IT/bRyBZWYXx+e9WxN1tcYkqae/nj3jlbfw5P+pbfI0VwrDMa+VmDvmfBIqYy9cy7i7j71G2UKHI+IXP4nVFx8soRNr9kvh+f9X6dd2wv8e9zXm09hWHAduZj9WsrGSNfy18NRrEfEKqUGn1/v2hVo1N45wXW0aSCkYqfcEWYmeW7BDExTclg/zuhXuqvXbSlrkZwyz65gR8ki0ic0+bSmewkmk/iv20PzJifDCUfv0PCgC5G6thW3pstJ9Ahpk/L2YHYYVHiVm/amKKvt/owc1M62t5CojZC2fzCMKLF7yCJ+H97ey0SOegjkul+6PDWZK4IS6HLZ/QW0FgjO/vFrOsM/20WXiXfezVRhQr9+XYYtbfe+kx45v5Svl/HR9ubRQMMCOLlct7R1gYGs6yDNPSftHBNk17m17RZ7VMv46XpV1Wy8T3m18LrWU8OKACj+Bnbtj2zRpw2aWLbtlWuJrZt59XyCPnKeO//d9rL6k7/q/npPX++FyMezOSBIhUy1KVAk20U2c0p1nOGc+xnglsc4i5PucALGeeVTPBGJnknU3yQLgbMvMaKDTtuqtRwUKdBEycu6uKmpTx48eGnQ4AegoQIkyECxEmQJA1ACkhLRorkyFOgyCglZlCmgvUnVexKpCGCVvtJFb8IQdy0Cf8kTkSJAAAU0co/idNPEZGyQIYEBf4+YhIVA0aGMDMsWUZQAAwwiKIPERMAvT9lwcgCzOLGKj6c4sMjIYLiIyw+ouIhJkbiYiIhWZJiJiVl0mIhI16y4idPkSAlCVAW3aGEqYmNJi3qtCVIR+r0SJBeqdMnDfqlyIBEGJQSQ2JnWCqMiINRaTKDubiZLz4WshgfS1iOiRViYaUYWM1G/GxiG0G2s4swu9lDnb0cJshRThDhJKdwcJZzBJhgEhdTdGlzUVJckjSXJccVyXCVa+S5LgVuiJObUuCWOLkrNe5Ji/tS44G0eMgTqjyVOC+lw2vp8FY6vJMkHyTGJ4nyTRJ8l/APlZVMdQB4AWNgZGBg4GJIYJjHwOTi5hPCwJeTWJLHIMXAwgAE//+D5cUYmB2jXBUY5JyDQhQYNEKCvBUYjKCyjGCaCQhZoSxmBjYoi4WBnYEjJzM9kcEIJ5mfnMNggUwyMEBMBWMmoFlSUB4HELMxnALLqjGwwMWkQDRYJQ9DJZBWZqgCkoZgdcoAVmYjbAAAAHgBJcWxYQFQEADQd3dJApUN/gwmACoLqAykZTOgNIENAF7zVOxz6Zdcv2bxPVaG0cj+X9VPZf40/hEA49lkDvdHXh9/1DkGTWwgLrlFU08ZkAyKAHgBXcsDjB0LAADAeYg/amNzsWrbtm3btm3bdhvVtn3e2HXjXu2JMYgihmxiIsgmHnmICgJx+fCvwsqqop0uxpnknHvSvfQxUjTSPHo1mhRNCbIFeYICQUJQLqgRjAgOJGRJyP7xI/hXoIhyqmqvq/G2OO++0KvPu9n3nTXIFeT7vMv+sSMf36MBHy/x8Qx8HMPH0XwsRUa2jPiHj+FcCCeGHcNpYdUwV9qptGMioAxaoBN6GYA9uIwQQJI0f4uZZ7GjLrhuvEW22ma/22ZbZ5bd5pjrgfseWuCI7fba4bEtnttpj12eeGm+Ky4675J9DkhxyClrPLXeYanOeeS4ExI9c9Am6Y45ba0bbrploZbqqKue+hpo6K5GmmiqmeZaaGWI0froq5/+BhjonkGGGma4EUYaY5TGBltiqRWWWW7lJ/Pae2MAAAAAFABeAHQAeAAAAAz/MwAMAeYADAIGAAwCPgAMAn4ADAKQAAwCyAAMeAFdUEVaAzEUTr09AVLL+3Co7NlgmcHd8nC3M+BO7SyPXdjNwbBXl5H8+sWMEIlZI0JL+tPlKqBx/b0YoaJfIiQ8+3tJI1wJKa1LRa4DFu4EG4PAzJOQNnl67BXdhTIjM9MnGWnLi8MT8vaUkIPTDKYliVV9yeOaBhrFcI2eIg7zPN7iPN7SPBnkGa4qMzCyxRP8cMmXmJXk6V3Sy5ruVZhGFYYBpEXOkiZHhQGRW/7aThlvLtsrew4kOBpkEizPsqppNEwCM5my6gK6z2TCmS6saiOcFsMlWo3RimFE0eGbsIzrfokjhi4IF40u6AIMAypeO5SYXdWWCgNg8h8k6mvkAAAAAAAB//8AAngBpXwHeBvH0ejuAQTABhBEOfR2AA69HQEQbGBvIiVKoihRxZYpOZLjKom2XCT3luZUx46S+OVP783/79hMeXHs9Kb0/C9Kc9yV3osFvtm9xRGgwFR9OtxicLc7MzttZ3aJmtA8QtwM9zqkQhrUjAzIgg6hhxEa2wUfaecjCKGOgUUG4AiAQ+qBxRXUBA1k7CytII3S0iotHWs9gloQ/v0KamffoaOOR1AnwDJZk9/oN4lGv1FrSmPBgE1lLHjxPP5zRffolXj7FcnulDpduiyVeTSb4l537jI8/43KfvyWb1T+dPjgwcN4/ILFxQsQAvRXK/CR4e4B/I1AD0NWT5DV12DfRgBtSK0AjARgZOTAL3qKeqvcAlRb04+gZooqFlNYFPTYgHlBK5gkk4Av4TvcerXe1cE/eka65jt3f+d63DbXdUWheHnX3J6PcPecO0pwU6HtDDcTcqEQ2rMhb00EYJKREeAXE0VBSK+gFvhuVnjnBIQADi0XbZmgFZL5KeU8nMWs0Vo82GLWc3FsyhXyXeE4tiiN7d+QdpSFYmliU3hkb+GJz4/1Dw7uWJrcPDO9xN1jDA0k03MGddvscHYyw+Obcqlk5H8qPyh0S9lvIoRRdvX3XJg7hbyoVKWjiaDdVEOHkwCcMpebALc2iqUTWmYZy64BTspZeW0KCwE94GuVcoUir4dvKS6b3nZFue9C75C9FI6OZJ0XbU8MuYYiLxooXzaXwmNzN+1K5fyTdre/f1fxqpsC3olUNrFwI5EByudB4HMLMDG/IZdbCKCFYPcIagV8ZJ4aq/yzynwrMnYZt3/o9e+Obz029T/vOnjpZQck7p63vnXrie2xC45dd+0VlRT0jMm4+BkYtxXNbjhqKwG0klHl8VRUJ7TwTCudV20aeEixaYVfdUw7LH6LZBEscN+OX1p5/uxZ7Ktcxt2z9K6lR5eQMvaPqNyPbDh2MwE0Nxq7uW7stRGNwnb8sspzv/zlMgz2yFLlV6g69wWY+xCa3XDu/QTgX5t7PaXJDy0rHdUNT/vpqG4Y1VkjDx5unUDkhbxkZEKRmT86HOjJRAyD9t5UYkJyXXzAles48WTgitGhq7YxudDzno5Jm0eWDIP+FRdWfuJPgnAwPnGjdI6m/ok5WkHNysw0A54qJicPI/VpeA7u2tNEWsBySUYBPoXty7h5ebnyJ2BY5S9Ye+4oDlR+JI+LPg/jqlBow3FVBKAi49I+oa9lZjwwsq3+Hj8A7xtRvPq+ljyuVd5XzByVaBXlaNGDgYN5QA6EWQiExcPmLk8qtWzNTEtjF3ofCPq7x/HHKuXklt5Ad7HKnwUYp+3v2Cc1Aaj/njnVEYBO5iAH77ZRcwqih3TQotRJJklrElRaC5Cpeu6WL569+SMXEKZ9D0dXK0dw19X/V5Hr7wE+TSi+IT5M+CjdHNNfI9GVZdxROQEs/PYSqtL2eujLja7csC8NAWhqAG4CcCN1Ay1akT0lpa1DaZngaQ1tWaFnN6PXxEtFMqdGkPGioFUJKlHwqCwwyTdfE1fHlm++XKXWcOr8kcKRLjXX1KS6FHjxqRMn8DCIUJwHLeDvJby5l88XC3zlO1V6+ok/2Xiu2Ew0dC4NpF0NDR3FXc/8DsM9hYkQUaQxQfr4V0bU6tYr2Q1QfdXLpKuL2AXIXvWyrqPFyhOIo3biKNgJExJQHtWZXQUhBWUYX/YL8AjzaJlsE+h8nlkFYo/F3AB8DwsBjcXs4fDWwRfPxGIzLx4cfPFsLDb74kFxIGHbs3nzHltioC29eOPclpMLqdTCyS1zNy6mLw6UF4tXnDx5RXGxHFD8RAT4p0fWDaxmvVatgIwjFhC0gTUwKF7DUut1iduwCKYcc7NnCnuGw+GRPfnDy9961eTkhdw95sxsqTCXt/8azx4fSD+JEOXVDu6lwCsDyqBNVVx8ZGhfDS5WArASXD6OkgiTeYwDtKPqM3zQaqctK9imGLFNwCdJ4RrWY2IFUqpapmqpkSi+G5gmVhmInzAWAku3+/ILA/4qg709W3P81owVgp/p0t7BQGBwbwn4ee5Dupa9k83FheWJKrtHjyx0t4oBTRNS+NwDfNYh2z8fkymKRonFCKMWSqyeucUVxBMYE1HmpU0qAVSsGvyYPr9vZGJiZN9vXnFw+YfSrsHQRO+b8N7u7lLi3NcxWkUwE8np4qa9ZubXwmBjz8EcSIpvYAyvwTJDAJk1G0uCRCvKABryPLgpLASPWAGmWOCuFJfGMttZxMNk2IvBCjDj/GJhd7K7x+AMWXfb0mHe37M1U3pxuNc3Gkt3G91h86I7I5h9/TuLwzek3+aPpsJml6WzWWzxxLqD4kjOlQxPOzzRoMlpMepCbb5YMRgZz3tzKUKbAyFOB3OgBerqQ3hGm+JQ5BnggBZEJamJSJGQ95No4HufxL/85DKXWFo692057pgDHS9Dv3bUt6EGtRNAew3ATABmmYnaqsFmgqo4K9D8uWWTkPF4ckHzlYGhPaXSniEBTM3d+amEyZSYyuNrK6/eemTS7588shVwWX0BaHQBLu1o8B9EfecjxyJ/hNrptKkkkwfz0gAGo60S1Dfs7bAb1Ua7Yc/xL78EUPhY98FS6WA3ngLvjFEnQqp2GNePrvnnx2U8aBwxUQBPAPwaZn5qmA3wEk9bLuRX5F/Btg5vlbxUAW/Tebyk69CpWy2tcxfPtVpb1TqDrnTlyVdt1Rk0aq1Bt+UVQNVD+cNdXZd04enKQ9IlefiCp8Ge7wpvCsP/yrvpfLcjEA0634rPwQRPfH78tAawE4C9hjQDARhk0jTwrl32nABqlQlSSbxMRxlLKpOgrLmEz3/wlsk2HhYmltaxGz64cstcu92gbufbNt+At+L+V/MZtzvDv7ryqcp/n3JIHo/kOIUQkw38F8DbgAb/ebzXo2mQVboOOYvw6cdPTMDiT93h0U9c99nl3+E3vis0Fg6Phd5Vufh3hGcR+Pgjjc37UX2ot95Pyyqhkf3KaXgMtBArWljEflBDDJFNBPdVfoo/VHkMlyu5cfz00njFsURtbRfYsC/gPyEnEkEa6+nQbTwhTBpZUGOgE0KwhHcVe2uBBwwUbgeYl6JlgVYAWjKqghx0dhWKYO/AwGnFAUxWUkZBFKjnKbKIXq/qeul9LRa/PTTmi0Uv7t23+4JmlX9TYmjrVaLXk3PmRqOdt29ZMJoEp8FlnnWELz9U+WHBFd4y7fe4HFdozHYf0DoFtqeZ+wyyAgq3rdO8zvUR3RrxHQTQgdTrvaoS0XUwaeygcZy8ruYpkR3QctCWBlo+hfAAM2IWLSHYw1UtOmQImAMg7lc7dUzlnQqnJnMOoWcmGhjzqlr2HHNI1uxwrrurm/vMQ9ucAWFob6l8YDTo8+RO8J075qe30bg8CvP6MZhX9/r1doNZrV9v25UsS7HO+9Qvt6OR0cVsZsJeNEf4vumpHo9kzgjzsdzOweAf+vcPC37blNG0uGl60W6c8voCw/uB/3HA6Szw3wycuG1DrJjlW5uQTgLoROqGJlgHvzSxCLQJdVL+dwIFNkqLXuG/juU/ZEo9dfSJeale4op5SnFc6NkUifW41ccu1Kmdk5F9lx8vFYoZR9aeGf5D3/7hoNu+5aPnBvLO8F03zG0vPsV3zlOdIrz/LtWpANq6EZ115roNfuFZnAi/AIYdMq6K4rQpUlNUQjGt34NrpiREQzZtNFzeFs/NSI6myhd0s32+olMUFtLSruEQnspKiWEdRT0wsCAFzCOTJuMk7xGGL7ykd7g8gDDgjPBu7muA2MKGmDPl72yYKDADopyCvI62Wk+TbzAdYJQwIOnGNFdgFPLFaoSDR53Ogf37jw0Pd0b6E460kTvUtj2/p3IVftVCkxQeSjtatCQnFlkV8V+Bt+1oDHh7Oap3hms4DhHAENKsxV3d8PsQ5XE3S87Ba4BU6DRcAAlTSDdA0gBJA2RcyUHMUM7zIBsscAyIGm1XgchJnt0YHUCYB1fVWQ8tay8Gi8YiFDmZJivT3SO9fVPJ2EjEGd4fDJjsfotd9OXHRXPI8OHdbR3u3KjYETC2tgvJnYt7QqMX9SU2FbzfFHLZYDCXm/SH0j63WevOe/MWdVvE404b1KaRRGw4yWubdujjjuhAzKrRtpiNFltpMDiYcnb4MniflElJuXRGor55Cj5+TvVxF9rI+CnmsbHl17D0IswwNXEGxcR11pk4YtONsnwahaljau8macvksWDUl/UeAzt2oTt58QWVr+BQf87rqbwHcagPBv0S9zhqhS6TqD7k1zVca2lIzjQOQwbCMCcwJAnpQbFv80YiXrgWjx3jhoJudyjkdgcXKhY5//vN1Rz6CoxjRG5UXOcTdOsjXQ0Zx4Y6YRyAsSwVjKikS5jxrkVherkj1Bd3inpri6XVz/d4olGPNxrFZ8+9IA5nnbqmBVVTMs7trWImzwvHwbwoORUlhbLxvPzzORVUk1MpkpyKCDmVqSOqh2589yMnb5w9wn2mMvHEFyo//M6ukwgTn4n+CLjoz8undDbICICdUDhilXpwkdcIwIupYx3Naq3WwIcHctxnzj1kMXILLdODTAZVOtBnP7r/H8mgbuPQl3myxvl5TomFm1mLyCpw7LS8EAf7pThsDVuSyQsaLyOGl6gM83WSrF27T4F/8E6lCkNGYSa9eXo5FE6VjoVE+MBnR/ypVCycq4p3f+U97IbWdBDoN6OL/3n6FR2snwGFXKaUTCEZMZ0KMfUKyfD3zyoaCVgP+1J1Cqn4tRcAVxMwJt/QO9QHwcBUxRG4630Xs4fU2zKTGA0PzadSOwZFcXBHKjU/FL6mv1AYGCgU+mlAIQzv74d7IAD36cXF6U2Li4jGdANcM+DEYro6f6RbnynVrS9HaRrWUJrgXRPLyJlYrtuqiIhJiSmalZiuSfHOPDg2YHF9WFcb033iGBGWmWhqQo7qfIMelTPHQrrIKe4DXY6wMExiupGgy7FNCelo/DSAz1JZ8aGbNuQ/kwXd+nBJ0yCggviJRU2wUGSRVDV2MvxTsROYEOLGN4idyjRiVTvHw+fHTp70ue3/tdURUGInRR+OA41GtO2f14f12m5c0/EzsvAb6oSfRbRM8h3TomAxt1s73IM8iP7OjNRyUK1Odle+zfLorwZ80oBZvaisDZ8ggER9jidBJUiRFoAF4J3EuhwPzawV6jM8/FqCZ7N/KBwJd3rCJltOtO/fLm3z5B15d0gwukJmey5qt2cmU9KB4K0Ot99h4DsNumZbIOkd2hbkx028lzdYOvTaZj6QDfiKUZvoBnrMQM9u7gjM8Vhj/tbpgmxDmtn6UmapCu4mtlqgmZ4ylmi1yVwTi+Z3j8/o9588OdbubDWbJXe8N2Qkkdydd+55fkHTtEPb6kgPRQCfUcDnZ/isYv/qp7exOivz3dD5GVl2k1SOm6nXAxVgoSjzTARR41rMnx89Foz4sq5jsKj0kcAEn638oD/nC+K5imUyDJZQttVtCBFcG+SL1q9eGGDjfJGpNiUgtF29ucXSom41t26+8l134LOrwelweDq4WrHItRyEuG/DuJAv+ufHZT5C98/mi/ys/G4AkJnlixDyrce2Du+1NIttuVfXqoXMUPPgJUPNkDnStml7Lzv5kpJOD229rvt2oEoYDwbHhVV2B+qeEsZEcVx4EmFSP8Dfwmcb5IkaW7Z/Jk8EsSMlxMwyRkCIWIO/lq8h4L2vOVFs5VvVzebm9PFXv+lEX5u9naS/uq7F6Ow+c9wM//f9+XcXWRMWS5y/SJaH1QKVBzcaR/VBl269rV2Pc+3CpB3pGHo8NQxlXJcw0quoCylrTc1iuq350at3tFoAT1Pz9JUfNQTT/RGzR295EdeUDnnxU7/xTQiBCf9vzq3GhtP2pqbdBM8AfLwVn22QT9L9a/kkmtZVSRZ/4Nmv4Que+VXl5gG8bU9/5YN7EVpdlfdSqLq4MFIjhLTop9jSEP6zDeBP1cH7FfgzdfBeBf58HXxZgf8c4ABaPYeQysDdA/RY0YkN6whMQRsX7Dh4tJ0KUQtrEVUDFikBgFZJNHSwrJeehQwWeM4k7wExUiFrsvjzkOHB1fI8fqnaluIlzl654dPbr8cT+NHKX374Q9yEM9LBzCVvW3rve+WSPdAI9R6R7J8AGjWUxl/i3Qjg8h4CykMd47m3IfxnG8CfqoP3KvDnZfjqX+HbV+jzbXL/6BX0+c3wvK8G/jMZvvoHgIdq4E/VwXsV+PMyXK5J0rk2srmepnBaC6T9dDK6tjaE/2wD+FN18H4F/kwdvFeBP0/goBHJ1TPczdxHUTPAE+iOqtREiExEGu5VWUEIGhEaOXFUBJrhSacsNKwFHTCPuYI62dOQMVSecyswn9KKsV9B76xyRS6NWeqAx9Ucg6gVi9XkA1/kfxHtJxU5Uzw+uHlPayu+1u0h1bgDkwP9Y14/qXgemhwcuDVKSpujXZPxk1eEh534wtgVo6TiWdyS2ZU4RqtyPdsySK4zAlNIzdMMOO7YYPdGfYq0SYkbtaflzFaHUgFtUbKidlltTrNdHDU7VuRlQS82Aq3Z9MJ10/gdlb+eOQOl+U2Tx+fTx4++KBi8+Bie3nrd1ijd0BLbfmLr9UfxNy+HeZXrS1QObExuJlAj+M82gD9VB+9V4M/LcLkuQJ93MbknGsIhD3zsovsgnGgR1QtJo3pstVrm7JCNCZEVJ7IqBqaZVbYfRjaww81wtwOvQiz9RHllqu45IgG4EXceuOzyJeklL+m5D/YdLU/99+ws3Wv0qd7/5TLy3iNcLKqxC9XVsjuAvlJ9VNFwXj+ObNRLq+MfRxy0eGjR9E5tDbtYrWEvHz6viH3Ykp4tFee67Hc/dbw/8yTwUq49cF8HXnooL3+CTagR/IkN4E/WwVcU+NMbwJ+rgz+pwM8SOFIROEfW4m2oD42gN6J6JujWz+IaIEUAKcKmenNBASll51QXSjFd6GK6QDTfiiJsX5NsJfyKjpRoywqtQdpyQ2uEOhRWF0nhNNZjlZKcOD8rHBbpJ0lSplRkydMkr/nfLnrdg06dFnMXbN4bXSpIYybPpkx6ImNvqnxbs6ng63JGhR3pnnm3dGFmaOtL7sXNZr9dGPNxfCYT69e+wu/xOo0JK9+ew2pHOB2LpYWBnZBRHhy3GGd5T1ZIRLYMbdnZIVdkNvWNDPdT3kdXRVKXAN77KO9/ga0MDjlzOrcBJgutDeFPbAB/sg6+osCfY3AM375Knz8k949uo/ASQripBv4EgTN/ZayBP8ngZwHu5FYU+HMAZ+tW9Al8tsF+rfOSDsp+raljsNys0BiG5WJJv3EmwwmFnhcoHgnGl86G8Cc2gD9ZB19R4E9vAH+OwJEK4M8CXA0QA4qtrX3DhIRwDU02ArAxPwiNcI0f1MGTNirb4Q5Z3mVfVlhzZWueLLzOlX3ekyyM6v3eRGFEjz0ON8kKhcbdqVTO4SaZotnedOoyz+JkXsy7d092iZjz7SyQBJHV2h3s9y2QtpAoBSktv1c5QK9bQQN70RFUv4rUbbhXQ15LNrMWLJhOwzN0n8wK8iIexSht3jRAWDFTrt1005YXWr012pqXNQ/oq9VWTQN1VUo5D/TlpL4+Kdd3dWhgW6qw0+leyKansnYVV/mGZqbgy4OSLmSkXW7PjlJqW3+Qe0U2E+/X/oHlxzYRFhRy2SJRTaOrmelmMSvlIYt2SNZJOcfLDXFfAbwP1nv4mhxvNwDgAXXDAMgADTl15KcTH2TfV1CGteSwQKrbtgnzDpdHXZ1/Jhopdb5L2bUFSuIqbC0kx9K8w+dwRbNRF58YisbGAzEp5w+6opmoy+62h4K5/FH8h8yWktcR7/aGpGjYanbZ7KGC6C/FbHbrFjHssPkcNn8y4E1G455AYnKs4gL5p/kHqi85pl88Qg3gT2wAf7IOvqLAn2NwWCOy5wvMvqzWwFcU+HMEjjiUgX5+y/0YdaA8WkT1gtnZYOH7cSJ3ZLUSXwENjtLKhF3ZixMFaTWC1Cbgnjkj3yUljhCVnJOJaaGyfc7Kp7CskCR5CeqIO4t9GDtLnuKgyZ840jvj653vSs4NwI7QaQgneX0mcI/3ej7nkoLjvld6t+3Z3m5vC9imh/z5kJlr7nvnsb6LpyJki23XfK/X7pY2+yfzA0aPYcIUd0wVEMvlKvX5nRuuqJkVVZjB8gQN85u1GUyfnJ8VjOdV3fl8XX72GOceC6enJSdJzwr9bk5Jz35GdftDszZfNT3rsG1+u5KflWun3J1QO02iE2ij5X+MAGKos0EJC35UaqedMFFBknCDqxOgBqWOqoVLD7+2kF/Y9MZomMjqFfUV1roqK8sywo9VAHeV01GGsusjUHg1RweTrlynVe8ye4OPkDKsM1VThpVLsfHwYMrRolvQaqKxuFKXBdrBhhwD2hPoJf8obtqoZlW3oxleV/YmNp8m3yC7A3dDGvph+z1lqA+gQRYjh1jOIkoZEZKNaT9ezwWZQcWatCV3zCx5YpCkfNtaBdoKFen36C+88cZxSGJq5Izl83UF6QXgDctoEp/pAh5cwZ1CrYhHY+jl9dpbE1V7CMCD1A2daC9yUiX2KGGjBiC9rH7lQRm2uRxYcQb6ZptJV1AENbN0V7fSGmZv1gSM4GBMZuJ+2AZPCqnqveJ/rWyfJzxXlPSYtF46mHXm7YktPpcwpm+LdtvcXhHfGp6Likt92dmC2ywk7dBxoi88MmuLSE7PbMwZ07+4szvs63M8mOq2dIa2J/ksFvm0w1EIO4Kpc99tFhy+sDs7IgYKyZAxPBkI9CYcM2VfIRkxRWd9wZ3SwFHJ7CTnhSJgF7/P3Y1aae2pPnOtb7g3uE0p6FnZ3uC1oizdF0z3HFEXxGF0cHn5ILl40dXR4RJ5dm9731vf+o53/tdb37fb3rN3dGRPN8937xkZ3dtjRxhNI4S/yJ2AkUbW7bHXNxBxlkmjdWM1tKqqjjuIhGeytJBC+A0Gyj99z9WZMc/CycotF+C3TjbbO889cQG1MX74+DbwwYv61+3W0jeQr6qe2E/L+TyDkgcvDqhYQCKuGUOwi73Y4g8PX7opUpBC/Y7BxL7h1HSXO1SeT9t7+DdVto0MXrZ3Lij1xtyDacklTSZiOxcvyHFNS6xW8nXALY22r1sB6/9urcSJEsqhqfozPxBKAKR1XdVEyMP01WGvnF1iFn2LezrTNdQZnyyky6Ixlwj3u3uj+0qpmYKnt6u0Sejfmtp8OHC/Oyol7RGXYd7gDFsjSb9zIpiwxsuRwqRZbZgrJ0bT9q6wvF7uJPkxmOtmwGvnOuuub5BL/DjZG0BEMS6baTUoKyIzzcx3K7NYHadluIl4ZbJMkzOFggVSrXlSosdxFaduc+iX8YG5yyrPvPCSuXOvcs247BnHRxbf/vbNcLznFsR4/w3AL4Am6jOdNdh5CcCr8J4lK70sWQn4nFlLZCr8Zmw2+S3COrc5657MFMYG896Yu2jC879uN4v26FTBFxrYmpi9RHi9MzrYUxowmgTcc9GpllZbdpPUta3bnRcIPxOA7w+oPqfWtCdOEIwjfYOaxcfp9m0N5adfiW/iSmkvpKDqwSb5OJCCbn1gg42RYlfG2edsLn/YWHR78hE759okRMeyTl9hKpqeL4diY7uzPdvdMwO9pW5/ZMcM/qDFaSvuHhYitux0LjNb8IRmrpnbd/t8JOYBrBL0zMkJwERqbAfqLBP8yCpUst5LA6oi2aepV2mFd6ut0cFU5Vv4fd1jEaP66h2vuPm6wYvGx07e9rp9iK3LZDvYDlozokgiCytqxisTQLlez8ooxvSMZN7gJaZl8skk/xlZMrvOyO6CVSj1uEidBYsUyfpF2e/EuAuap6py+Z7RbKrXFu32ecQ3O5zZZChnFlLPmHxRPiS5pYlofCzrMNi8BmvE0+ntnsukt5T8H8qUcklHyG1rt/a91lHwxIrJmC3o5L/myUZ8bWrHpESSmvZkWbTFQ95WdZs3kvOFyuBHUoNElkLAk28pslQvOgpHmHDJskQ0hldkSbNOlkzUU8qpVnrYAUhec5K1KvB7Z7+jeSgldud3dlZFaUaIjsqilAJRio7tyZbm3Xg8EJmfGegplR5noiQyUXKHZq4GUdoBokSwKsN+m2+jVwEtNtRgd52KTE7ViwEG5bjPF4v5fPHhQCIRgIv0kV6dRk9CH5Dbq98/JPfRAn2QAECo3SiVzuSb3cmkL5rcuum4aTAcCvjEge7pyxyrqzJO+GlORHdC9wtIQ+4Ap+PUwe+Sc4vw/LPcAyDjRuRADbaFNMtsFiWtStDWUHNItazumA1WaeISR49W3nevQpnc9yx6Fv+J9t2OPo4MbPWTyXqxUDRJRdIRo+vb6mUVdEe6Il0OHT2Kd7xeSCQEP/TGYopXce9FbhRFkSqeIsFTlMXEqJwiEVliDrgfHlCf5znVfC0Zb24yuEN+azzszVkHfbuKkdGckw+IXpP6RBWX7wX7cvFohA8mXLbecNYSG4j6pFQs1eVTyOWgFjGE7wVaCX480NoC6BhpBlaElp9Sja16df3OXJBRtVjLBuxr6qD4iJ6cdci/qxAdIfiEvZ0qgk+MXNgc6gWEolYh6ebrEVI4RrVmM74bPc59GPjfikh5TA3BKp1KAxa2qa9R8cNc+ciRj1H5gGfxa+HZMLqHrHHh8+u0j258IzrNfQR10BlsAkirnE3WY4vCRUC+eDoWVWmcYsgdjpzgbrSVktjv9oV68rcN0X5G8F3oDLfC+mkFSJPcDwi3pka6tSOkH4cYdIejJ/FKfT+AJ+CD/w/gE0avInjC59cpHPrH7+NWAP5qAodPBl89iM6g+zgCQed+WIXh96H7kMhg9Gm4U31E/aAnH0caxjMN5VkxX+S1vDZ9V+rYsdQ35+8NnXpjUOZxAZ4/W/N8k/y8WOQtWvHdR5N335184VTo3vl7g4iODf3j18DzYfR6gid8fpDCCwB/A4XfR+H3ETj0H0PfxDYcRirKN4ww4qrZfCH2y1/i8LzcL3uO06BT7L3v4odw4Pz3oGoTwy+vLOPADHuO++NGz3FfOddFniN2Bf0O7MorgY93kgHpGBF0B3bjlyATpb0FIZn2dbb3rfak32TyJ+32ZMBkCiTvMAlptysTMJsDGZc7LZhYff4G/OzqpxkeBIIpHqQOi5+tBAfHxih0dnUztnM/QnoyprKPlJf39vJ0Ey+v/UZ5bKw8WcrnS5MfPXTmjjvOHPJd9P0rr/z+RT7aR3x1M/qT0ocWtct9UNdJUijExopx+jrt6DXsZdYV7WP36mHs4R5HWqrtTSzeWT9DRrh2P3755Y9zj5fPBcr0HNph9LzyHpL1AN5TKxTzsLSQ4Hrt5Y8/vsj9qPzC28l7PjZeEcUV/pD3AnAXaKuTWZwVlII72SdGuFf1hDWxuAyAxYPsNZVwnHyRsKfgS3kMKpVKB0UEu9+sg6bBk/IV+KC9yScIviZ7cLHc7stHnMlY3KZ3WvTjeotDb4/Hks5I3tdeFpLxrZlCrlndLOUzW+NJAcl042Ad/gWKtQVaZtoKkaco/i1wD1H8eRopKtu25YJaviYhyUmWAG1QAPzWv4YlX/Cmz6ck7S0sUgzzEmCYK1AMFXrsegelx6m3KfQw+dyPR7jPgXzaGu2t5c6QCVcJf9j83Gbucy/IOri4qkHv5H6JVFTOMOJk+wuTuzg7y/3yBRvALNDv7D/q1yRYNj9f7Zer6VfLem5SegbpMULvpP/K4L5KHn9ZtlMZfDO+k/sCyLz5/LwxvEosVlHiqZkTtXt7XmS6vLXceoXpUI+0aQa/YynZZz9yxN6XXArt3CnvL30IYtsPIQMap5KcJvaAzuMIO7sqn08FrWKhKQvTaBKZRP7rN1ym1GSew9XQ/+XB4rCH5wXyaXuWL3oEb7Rvsi+aENv0ttRwtLDkG7V3C0Io0jfWF0mGPZZYfyS/w4MXXCGP09gWcLhEj8OoD6yYnS4r73V7E+VkaoDvEO0w19ZIuM/u9NlsPpcnPpCODlv9WU9X2BrmaazxYfwnep6/Y4MT/fWbU8CzA+dVgapiGfFty8t7C/vGRKiqdiXgkHoAB53d86X+7RJfGadrg/TqH7lP0PPdIhpDEnCwB2E0RDmYhFYXDWWSEOMLZ6orkCH4Zj+jRL7r9+cSbajJlfB/50RQOr94fHj4msV8fvGa4eHji/me2NRSd/fSZCw2Se5TsT3e/Ggo2WdLmoK81FPKuGNm0VWORcZybjw6e/1CMrlw/ezsDTvT6Z03zJYPT0ej04fL5UNTkcjUocHERNYZsOU6TaNdPaNWYzfvdUqk6IbUyIV78QPclyjdeTRAV0Vdach+p+EO9JnP1ObDPWeqtdcccCRKV0Q5yDwUKCwN9xzA5Dezp8k7j6A+mTvrvND6LJKKxdK8IAqkGM5q9T+EU7UeT348Ur17C6LVKha8vjy555Pb/WlTZyLvD+SmjGajyYZ73XBkNTGdd7vz0wk4uOqO89GSIPTEeD7WIwilKI85j2D06LYMDV3coW839lEdzq3uVfVwr0GjaBsqs9kfpLMfhpaBtiahZaUtD4nmoCV7mgy1kn6AeORsy7qVDtspSaE8OTegkdZRT0VBQ6tXVC7KOEx+yo1f88CuzTctZsSBmVC8T8vxw77ihLtrMtHl4Nqn4vZNfCpo9Y8cnhg+sb83NXdZf2yb37nrkuk7DvbioN/VGeYdVr8j6jZ85dC7jg/l9tw0M3LxaCAc9oRnx+I7RuPZ6LcuduYmkqWDmxKFAy9bmL5qU5jvHMyl9r1kXzRkc71W8KTk/dMC2b/JvRQ5kRddh87bksEADgJwyAsAFcvhfBxpoaWm+qNlyXdy18NdzbI4Krh3smyOmd1tDO5iz7lJng9cMb0kLb0sAr2EIlwm+LhqdkfH1j3WMcvV/KhlbrFjxwX8OH+1bfxMxzUdj83cMvNB+Ae3xx57DBtuuYXEUdvww2hVNcBpEcTd0CYwP34YT6imAfawAvPi9+NJCntEgYXhuf303RUKo3tZYF20gF9D89gRRNKsG6dXlYMFfF16de2Y08rYzp1j5PJFoz642i4/cODSSw8cuHxkZHp6aGh6eoT6km0I0MF3Ih0KKCPWbUuUx0KyIrJEqfDuhVCCH9uJwwWNyVi5nuVHJwB3G0ogko87b69r9cwHOi1vC9awdCiuWdSx/DfkYTPzFwhea9iYc0zGBjZBpOl4ReW5zORS1B9286lQdLjsyGRzzilKgxf4NgljR9EAIoJVXzNnfx5GR7P7YaXSY2VphxXkgd+alQ1flIN5uNWtNevd3I5g0SUWu+JZn9McMuTcZShbBC2wtknykbw7t9l2Wd7jcnsln4t32UwJX8DgjNj9SUNrOuRMeI2xoLymDgPy+4H3TagDzSDihRr8RQVU9xcVOOUP2zSdlrcpadkZvzYm+obTdIkjGDBPkplCnpRb/rdJnZSe7pl98P7DPaeKR7Ys9l54YeJvP6fyinCeW+A06JNULp3wfQf9/imKI/sd5tWPehERRVJgImskjNwUrQ6lVq9oINtnJv9lFbYDmG2GwQpT3ZjuAxfpOWNg7R2pwp4XcTYPn+U97gnx5IhObc5wRp+xnMWoGMuXmp0+U3ux03HPp4VO+6CxY1zr5yMER4YzPR/jRGQfF8Gz9q+UOBr7kRp1edjXnbDbE90+b4ncS14hFBLgwohPlCPRwQTPJwajkXKCHymlMqVSJlWi/IngN+MC9wEyNo2dzIrlaoGWA1p/f9wPubIhiyWUdVXvPlH0wYXfbA7m6J9UqN574wEhHhcCcTouj/bh57kCzEoSRdad02shchNiEyQfYY1TBpQxjbxJfFigmz2oTGtImCjvgtDWbpngB4ydncaB3tFMT09mrLdMvpV7x/iiy1Xgxyb7YvHe3nis7w1lvuQPlPhy37jXO94H3wJ++q3DYOgY/3lXudyVL5eJGCdXX4M/yz0PGO9bl/XRNzwDIa6V4gAksnMdxOrXnu9wI5Gd+RDZ31LBddEpKEEKr4tM85LFg3l8s6srGE1F+sf7I3Ob5lL7Al3CZDSbkQGjc5kxF/6DwxkVPaLfnxrJTV6kP3SwSfRtdgWSUX8i6E+W0+P72l50UBN3E/ok+DgD+dD42g4CCyEHsv0N9iKBUWWVFgviKP7BtOzhgsqultbTctbYovylnRCsoeXFLdFsmM71kSE5CSuvA/GZys5MD8cdHGnTtZOIprjFM2iKudNdLdpTL1E5nEtLE7adxVZT61iTQ7InJ3POWLBstRcle9o2bbEk+9xqgiOtwfwK8uxudDuq/2tL+vWbt/UNz6uqkZudbbCzFtlKsXbwHTzCabgzx20jJobcO6o+w0h/Vf78WM0mfVJIhosaOolcqpcM6h0Gtd6pH1x+4+0PPPDA8pvf/ObbTnEn7vOTM2z++ypf6+3tu//+vt5enCMxGfiOv8CcSehyVO+z9A2q4x8nioQSVK0SzPjG4a5jd+GMjLoAyAYU1Hn45lEO6GbWauR0EsNd+T5chMaAam0iSWiiMVtcmEynvFEJ/0XYWdx7m7lcwrhUbkkOz+eS29xD5qS7MN72/K8x/sKj+lHJl+O3JjPXH+smfzlEp9d2q3t2D/hhXnnX3GS/pk3bbNT1j8/6vIhD/as/oH+zSwO4FsF2eZmrkfeceiiVng7ZZXvhrmEup51ODkgl0TWapvHCjSoalUemXtqwaIRwAbencrnUl9wTweELjvqKM8nhWJkvCKIj0Yk/b5jp7p7Riw5v09KBthsv69oxIGovXuJ8/IRXXfkKZzMghFd3whxx3KkG62pqJjCsq7VYwu/BF49X3tTGvfPcbppHq74Hvuxz5DuJr1Rx6EdED0G3Orh/Dh+nGbqH8HcQknOS8IyWPvOx1afZM/vpMyv4XTXP6Nc9s5k+8zD+PELKWCb6zCOrq+yZk6yfB+Eunw85gH/MZVErcqO1mBekhPEaUcmiYZeUl2Q5/86DD+6E/9g3/8lPzn+S9YN9+Md4uWHeL4A/WRnBy/PkuSw6jk9zLhivFdHTljTTUAbjWM0VZF/3ut7X3dt7L/x/3YOvv4/oyX2v77vvPrkGFkDzgG+ZxqdOlFu3G6BtYFEuC1ioCNmZ+6uNH3FNrKqqSdyLgChGblF0w3UR/YTrFEX9E+RwNjmkXfl/8j30t3ngbwB9BGgme4m/iq5CGvh8NePp1YCjALRZgEZMwz6gFe5YjkNCFgO2eLElBJx59K67HoUhrvnARaMZdWb0og+gBn3Am6wPzE5w46Z8GefTON+Ef1wZIZ3gT+JAbScc9IHY3BpQNzpvz9HG0w1PsSML9ROvBR6xyX+QMia+/cEHt7+pb7uMcxHz+EvcA4CzY32Wo/Z4uEmgf7dVOL3/1EXFjDpdvIh7oHLdfXfddR/JS6I/45/gNwHVIsiRju07l6uTpjOyL5Y9lJ5pv0mOjqVcNRkhBLQ+34TgSUdClpiv3zdebeLerqTdY+9NS/SmyD+R2yZFbtVMbiFlSiR3Z+UUXq786pN0r2Q3fg6fRTwqo7pjaQ0Pn7OTiIC/kW2INLNW45OIIm+BXcGhiC/jJseE+XLYFtZrjtWfR4y0mXffTXT7VVw/3st9Xd4VzPT9CwB7HYU9ocBuB9hbKOxJBXYdN0RrG7ATWIFdz43g91LYcwxGxpiH/t4CsGcV2Ps4P76OexJgZxXYw6siPoBUbEc3lYXVG7l+NA3jKvZScWFtVXupknjh1qO7F7ivv0Z+53Lu0wjOYyj5TRXIvPzYa2+c36fqWoZnHoJ+b63vl8lYm5LfFIpS28Luo1w/6/dR6Pe1Sr8cy/GTxzz75m/kPk36hafRPZRPxfNtSgMCNB0APCMf69TAnMp4kj/ik9q7f65P3TcHZO3YQcc/DOPfD+OTHD3EkVSX5R60Z8iZWni3KL98//6r9u+GlwGn5cFB8u4hbgi9gSOrmKEG9NapNDV+JtJ73R+/k5W7lYyiaIqkaIrwBk02bPK7bG2Oln5omn1OO2lyQ6VBQ6chI3/KfFziHkMPqPqBCgGtAMTIbFIb3NdMRxtdgckrDhJFCNVRC2TYAbPfbW/Xm5v6NV3hmjb3mMFksNl7yHC8vTQky9A8zPVb6LwR7jOxASN0K1ZXXuDeMkPXsMDfEXQ/t8JyFg24pGEbcNrYMeZOwowaPyDVtPc5/H6HIxB4C1wOaHMjAYc9ELDDV3ZnYz4Oc9or+yGlcvQfDPM4fJWb7C7T9gbOjw5xT9JxTDBOGx2HjNdJpdi4QeolVSqlyGWy201wcf58IpnPJxP5lIvn3W6ed7H+gW9Qev+3+k/29CQ7HY5O0v9KLh6TpFg8l7LxVrvdyttkmXnrqohmYQaNtHeOyUwraqG9ixLLaGhFeUHhHuftAUeyx9cSzrh9qmyz32sL2CCT2GQxd1h1jfo0Io722YJa/90+wYaxvznDaUhG5P8DKD5xZwABAAAAAQzMTspD0l8PPPUACQPoAAAAAM2XgKcAAAAAzZfjGP8s/p8EqAPDAAAACQACAAAAAAAAeAFjYGRgYL7x7z0DA8ux/zr/vrOsAIqggo8AuyEIUgB4AW3TA4xeQRTF8TN3atu2bdu2bdt2o25UREVcRrWjmjGLqLb1+n+zVvLLucO8O5vPDmmY4r97gDWI/tlwDbVOamyLyTIa6v6Ryxjvoy6DdyoV73HfmH+JBBxhfVBSjiQrq4ZVURnbo0FWKvrjK6mIlVd+9zX6Y+VUy/KqmTVTb3dftcm67rdqu/Yqa7lYa8B8DrVzOaPHloe6oHr7vuptjVA93k825sxy1k6plJutYlZP3dx75bM3jJ+pQFy7i6oS+smC7xL9DT3ST1quTPSLHgeQX2MpvWVW348l6S+t0N9uVbBRnKPHrPjCahb3Tc/p5ZBDa5cj+orX1Ck9Z8V3IMexTu9phd5htdQovNUs3nULNqqcOdVCH1RGKRRJynpJaqO6K6qOaJiSbTU0uKRaqM/cAPccRdUKXalboWtijTJqiBbuJnmT7KI6wVPV8ZXI+N7u3PWFb+2m/qiLMaiBSiFHKZ8ijSaLhzyjRuH7RtLDa+69qHJ+rpowV9Ut0xBURkXUQIVMc5OoF6tsSo4Lb1ES9dEURVAC7TU2GhnzvVXBl1Ur3xY1VYG1Kkkah8ysJSqFukn43yXgJrZhNdYkzR3DOSnagPk4g+tYglmYiSlJ67OSHEhyJLgnxb+T/+MKsQQAeAEtwQMUIzkAANDaSmZSt0kxSnK3to2HtW3bts2HtW3btm3btv7X6XT1/12p9+sb6mfqrxtEQ3vDcsNxo8FY2tjWeNGkmrqatpu5ua/5uSVsqWuZaoXW1tal1tu2qrbj9qB9pP2xI7fjpvN/Z33nUudrV9I12/XRXdi90UM97T2XvUlvV+9an+zr6zsLbEAEBHCQExQFZcF6sBscB5fBffAafIcO6IdJ+D/MDYvD8rA2bA5fC6WFykJ9obXQXRgsXBeJyMWcYlGxrFhTbIrCSEaZUX5UGlVG9VFr1B0NRvPRarQdHUbn0W303F/RfzdQP9A60D0wODA+MD2wOKgGh4YsoaHh7OG+4c0RW6RzZH80Hh0Y3RwrGlscOxp7G9fFXfHc8brxlfHP2IejuCBuiNvinngwnoiX4r34JL6KX+KPxEUIoaQsaUgGk4VkL7mZKJiYmdicOJq4n9Ql1eTk5MFU0dTBdNX03vTx9MX07fRTSZZKS0Olu7Is/y9nl4fK+xWDUlM5quZU52tpras2UZuqLdd2aqe1u9RGozQv7U9H05X0NHMwzhqy7mwnO8zOsuvsIXvNvnIL93HK8/LKvCmfzNfyz/9l/q/4f33/W/vf3f/F//P/3/TXtj8B7WaMQAAAAAABAAAA8QBVAAcAaQAFAAEAAAAAAAoAAAIAAXMAAwABeAGVkLN2RGEUhb/YeYbbx7bZxE41tueyTZ2Hzl5nxZ7162wOgB7u6aCts482JuF5bmdYyGbdx6w/z53vPF088Pg8dzPaNvw8D3PYtswFVTwaJEnjCMWp0NR0Kq5qTJoyeRJCJVKcC2fxNMflmGKcSeZ0Vthil232NL3vfG587hv73PecO7HktbSGvHlpFZwP7adKOj84tOTIiXNJmuLzrOksmFomTpG0eTJiS/YdpjXN2Vmylla+eQ5Xq8YyE1qBrXHiUqrypDUnNZWluUTypTWV7N+L6x23fJnSE374RZx4AWzBRQHCAAAAwNtWBXeG2w93d2ISjARQgDsh8H2L/fMhCINIJCEpJS0jKyevoKikrKIqVlPX0NTS1tHV0zcwNDI2MTUzt7C0sraxtbN3cHRydnF1c/fw9PoRBA/dQihgAADne+/atm/9jLzMWmbb9sl2q2zbtnky19m1bsY+t9ywX0+9LNPbHX3cdNsDd91zX1+PPfTIAf0sd9AzTzzV3ycffDHQAIMMMdhQGw0zwnAjjTLGaGON89x4E00wyRSTbXLIfNNMNd0Mn5320ddIiESp8hUoVKRYiVJlylXIla1SlQRb1NksTaLjTjjqmKuuOe+CXXbLiSSHHXHFXM2ly9BYU111l6ebJt7LNMc8Cy1QLymSIyVSIy3SIyMyIyuyrbM+ciLXTOckm21N5JnlbORL8cdvf2X55ruTalSrtVIzPWyNAi0s9sJLS7zy1rsojKIojpIojbIoj4qojKqojpqojbqojwbRMKFN51at/mvR4f++3f41NRZGDMMwACyTZgnjcXGPMrdhXD5gmRLb/ye8dzedZ/+Xrm/2i+B/U5MsTBZpERBZpc/4hrHlPchitPoWB+r1H/z6BpS+oRdsIIoNvc1OpG+GMyeHLA76laZpOLh6i7R0HSjVyymkMQMopZgJlFLMAkopZlP6QCnlNkApYqYHlDymb/TpS03jTwtov73SAAAAsAArALIBAQIrAbICAgIrAbcCMCYeFw4ACCu3Ay4mHhcOAAgrALcBOzAmHBIACCsAsgQIByuwACBFfWkYREuwYFJYsAEbsABZsAGOAA==) format('woff')}</style><link rel="preload" href="integrator.js"><script type="text/javascript" src="integrator.js"></script><link rel="preload" href="integrator_002.js"><script type="text/javascript" src="integrator_002.js"></script><script src="pubads_impl_240.js" async=""></script></head>
    <body><div id="popup-references" class="u-composite-layer popup-base-theme" style="top:-60px;left:-60px;visibility:hidden;opacity:0;" aria-hidden="true" aria-label="popup"><svg class="popup-arrow" xmlns="http://www.w3.org/2000/svg" width="33.7" height="18.4" viewBox="0 0 33.7 18.4"><path class="fill" fill="#F7FBFE" d="M1.4 18.4h30.9l-15.5-16.9z"></path><path class="stroke" fill="#98BED7" d="M0 18.4h1.4l15.4-16.9 15.5 16.9h1.4l-16.9-18.4z"></path></svg><div class="popup-arrow popup-arrow-shadow icon--popup-arrow-shadow"></div><div class="popup__references"><div tabindex="-1" class="popup-base-theme__inner" data-component="SpringerLink-Popup-inner"></div></div><button tabindex="-1" class="popup-close" data-component="SpringerLink-Popup-close"><img alt="Close" src="data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2214%22%20height%3D%2214%22%3E%3Cpath%20fill%3D%22%23666%22%20d%3D%22M1.706.292A.996.996%200%200%200%20.294.294a1.002%201.002%200%200%200-.002%201.412L5.527%206.94.292%2012.177a.996.996%200%200%200%20.002%201.412c.39.39%201.026.385%201.412%200L6.94%208.354l5.236%205.235a.998.998%200%200%200%201.414-1.413L8.355%206.94l5.235-5.234a.996.996%200%200%200-.002-1.412%201.002%201.002%200%200%200-1.412-.002L6.94%205.527%201.707.292z%22%2F%3E%3C%2Fsvg%3E"></button></div><div id="popup-article-dates" class="u-composite-layer popup-base-theme" style="top:-60px;left:-60px;visibility:hidden;opacity:0;" aria-hidden="true" aria-label="popup"><svg class="popup-arrow" xmlns="http://www.w3.org/2000/svg" width="33.7" height="18.4" viewBox="0 0 33.7 18.4"><path class="fill" fill="#F7FBFE" d="M1.4 18.4h30.9l-15.5-16.9z"></path><path class="stroke" fill="#98BED7" d="M0 18.4h1.4l15.4-16.9 15.5 16.9h1.4l-16.9-18.4z"></path></svg><div class="popup-arrow popup-arrow-shadow icon--popup-arrow-shadow"></div><div class="popup__article-dates"><div tabindex="-1" class="popup-base-theme__inner" data-component="SpringerLink-Popup-inner"></div></div><button tabindex="-1" class="popup-close" data-component="SpringerLink-Popup-close"><img alt="Close" src="data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2214%22%20height%3D%2214%22%3E%3Cpath%20fill%3D%22%23666%22%20d%3D%22M1.706.292A.996.996%200%200%200%20.294.294a1.002%201.002%200%200%200-.002%201.412L5.527%206.94.292%2012.177a.996.996%200%200%200%20.002%201.412c.39.39%201.026.385%201.412%200L6.94%208.354l5.236%205.235a.998.998%200%200%200%201.414-1.413L8.355%206.94l5.235-5.234a.996.996%200%200%200-.002-1.412%201.002%201.002%200%200%200-1.412-.002L6.94%205.527%201.707.292z%22%2F%3E%3C%2Fsvg%3E"></button></div><div id="popup-search" class="u-composite-layer popup-search-theme" style="top:-60px;left:-60px;visibility:hidden;opacity:0;" aria-hidden="true" aria-label="popup"><svg class="popup-arrow" xmlns="http://www.w3.org/2000/svg" width="33.7" height="18.4" viewBox="0 0 33.7 18.4"><path class="fill" fill="#F7FBFE" d="M1.4 18.4h30.9l-15.5-16.9z"></path><path class="stroke" fill="#98BED7" d="M0 18.4h1.4l15.4-16.9 15.5 16.9h1.4l-16.9-18.4z"></path></svg><div class="popup-arrow popup-arrow-shadow icon--popup-arrow-shadow"></div><div class="popup__search"><div tabindex="-1" class="popup-base-theme__inner" data-component="SpringerLink-Popup-inner"></div></div><button tabindex="-1" class="popup-close" data-component="SpringerLink-Popup-close"><img alt="Close" src="data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2214%22%20height%3D%2214%22%3E%3Cpath%20fill%3D%22%23666%22%20d%3D%22M1.706.292A.996.996%200%200%200%20.294.294a1.002%201.002%200%200%200-.002%201.412L5.527%206.94.292%2012.177a.996.996%200%200%200%20.002%201.412c.39.39%201.026.385%201.412%200L6.94%208.354l5.236%205.235a.998.998%200%200%200%201.414-1.413L8.355%206.94l5.235-5.234a.996.996%200%200%200-.002-1.412%201.002%201.002%200%200%200-1.412-.002L6.94%205.527%201.707.292z%22%2F%3E%3C%2Fsvg%3E"></button></div>
        <noscript><iframe src="//www.googletagmanager.com/ns.html?id=GTM-WCF9Z9"
                      height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>

    <div class="skip-to">
    <a class="skip-to__link pseudo-focus" href="#main-content">Skip to main content</a>
        <a class="skip-to__link skip-to__link--contents pseudo-focus" href="#article-contents">Skip to sections</a>
</div>
        <div class="page-wrapper">
            <noscript>
    <div class="nojs-banner u-interface">
        <p>This service is more advanced with JavaScript available, learn more at <a
                href="http://activatejavascript.org" target="_blank" rel="noopener">http://activatejavascript.org</a>
        </p>
    </div>
</noscript>
                    <div id="leaderboard" class="leaderboard u-hide" data-component="SpringerLink.GoogleAds" data-namespace="leaderboard">
            <div class="leaderboard__wrapper">
                <p class="leaderboard__label">Advertisement</p>
                <button class="leaderboard__hide" title="Hide this advertisement">Hide</button>
                <div id="doubleclick-leaderboard-ad" class="leaderboard__ad"><div id="google_ads_iframe_270604982/springerlink/10618/article_0__container__" style="border: 0pt none;"><iframe id="google_ads_iframe_270604982/springerlink/10618/article_0" title="3rd party ad content" name="google_ads_iframe_270604982/springerlink/10618/article_0" scrolling="no" marginwidth="0" marginheight="0" style="border: 0px none; vertical-align: bottom;" srcdoc="" width="728" frameborder="0" height="90"></iframe></div></div>
            </div>
        </div>

                <header id="header" class="header u-interface">
        <div class="header__content">
            <div class="header__menu-container">
                    <a id="logo" class="site-logo" href="https://link.springer.com/" title="Go to homepage">
                <div class="u-screenreader-only">SpringerLink</div>
    <svg class="site-logo__springer" width="148" height="30" role="img" focusable="false" aria-hidden="true">
        <image width="148" height="30" alt="" src="/springerlink-static/439917616/images/png/springerlink.png" xlink="http://www.w3.org/1999/xlink" xlink:href="/springerlink-static/439917616/images/svg/springerlink.svg"></image>
    </svg>

    </a>


                    <nav id="search-container" class="u-inline-block">
                        <div class="search">
                            <div class="search__content">
                                <form class="u-form-single-input" action="/search" method="get" role="search">
    <label for="search-springerlinks">Search SpringerLink</label>
    <div class="u-relative">
        <input aria-label="Search SpringerLink" name="query" autocomplete="off" type="text">
        <input id="search-springerlink" class="u-hide-text" value="Submit" title="Submit" type="submit">
        <svg class="u-vertical-align-absolute" width="13" height="13" viewBox="222 151 13 13" version="1.1" xmlns="http://www.w3.org/2000/svg" focusable="false" aria-hidden="true" role="presentation">
            <path d="M227 159C228.7 159 230 157.7 230 156 230 154.3 228.7 153 227 153 225.3 153 224 154.3 224 156 224 157.7 225.3 159 227 159L227 159 227 159 227 159ZM230 160.1L231.1 159 233.9 161.7C234.2 162.1 234.2 162.6 233.9 162.9 233.6 163.2 233.1 163.2 232.7 162.9L230 160.1 230 160.1 230 160.1 230 160.1ZM227 161L227 161C224.2 161 222 158.8 222 156 222 153.2 224.2 151 227 151 229.8 151 232 153.2 232 156 232 158.8 229.8 161 227 161L227 161 227 161 227 161 227 161Z" stroke="none" fill-rule="evenodd"></path>
        </svg>
    </div>
</form>
                            </div>
                        </div>
                    </nav>

                    <nav class="nav-container u-interface">
    <div class="global-nav__wrapper">
        <div class="search-button">
            <a class="search-button__label" href="#search-container" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-search">
                <span class="search-button__title">Search</span><svg width="12" height="12" viewBox="222 151 12 12" version="1.1" xmlns="http://www.w3.org/2000/svg" focusable="false" aria-hidden="true" role="presentation">
                    <path d="M227 159C228.7 159 230 157.7 230 156 230 154.3 228.7 153 227 153 225.3 153 224 154.3 224 156 224 157.7 225.3 159 227 159L227 159 227 159 227 159ZM230 160.1L231.1 159 233.9 161.7C234.2 162.1 234.2 162.6 233.9 162.9 233.6 163.2 233.1 163.2 232.7 162.9L230 160.1 230 160.1 230 160.1 230 160.1ZM227 161L227 161C224.2 161 222 158.8 222 156 222 153.2 224.2 151 227 151 229.8 151 232 153.2 232 156 232 158.8 229.8 161 227 161L227 161 227 161 227 161 227 161Z" stroke="none" fill-rule="evenodd"></path>
                </svg>
            </a>
        </div>

        <ul class="global-nav" data-component="SpringerLink.Menu" data-title="Navigation menu" data-text="Menu">
            <li>
                <a href="https://link.springer.com/">
                    <span class="u-overflow-ellipsis">Home</span>
                </a>
            </li>
            <li>
                <a href="https://link.springer.com/contactus">
                    <span class="u-overflow-ellipsis">Contact us</span>
                </a>
            </li>

                <li class="global-nav__logged-out">
                    <a class="test-login-link" href="https://link.springer.com/signup-login?previousUrl=https%3A%2F%2Flink.springer.com%2Farticle%2F10.1007%2Fs10618-014-0367-9">
                        <span class="u-overflow-ellipsis">Log in</span>
                    </a>
                </li>

        </ul><div class="main-menu u-composite-layer c-button-dropdown c-button-dropdown--ghost" data-component="SV.Dropdown" data-namespace="Menu" aria-label="button with dropdown options" style=""><button id="button-Dropdown-Menu-dropdown" type="button" title="Navigation menu" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown-Menu-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">Menu</span><span class="c-button-dropdown__icon"></span></button><div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown-Menu-dropdown"><ul class="main-menu__content" data-role="button-dropdown__content"><li>
                <a href="https://link.springer.com/">
                    <span class="u-overflow-ellipsis">Home</span>
                </a>
            </li><li>
                <a href="https://link.springer.com/contactus">
                    <span class="u-overflow-ellipsis">Contact us</span>
                </a>
            </li><li class="global-nav__logged-out">
                    <a class="test-login-link" href="https://link.springer.com/signup-login?previousUrl=https%3A%2F%2Flink.springer.com%2Farticle%2F10.1007%2Fs10618-014-0367-9">
                        <span class="u-overflow-ellipsis">Log in</span>
                    </a>
                </li></ul></div></div>
    </div> 
</nav> 
            </div>

        </div>
    </header>

            

            <main id="main-content" class="main-wrapper" tabindex="-1">
                <div class="main-container uptodate-recommendations-on">
                    <aside class="main-sidebar-left">
                        <div class="main-sidebar-left__content">
                            <div class="cover-image test-cover" itemscope="">
                                    <a class="test-cover-link" href="https://link.springer.com/journal/10618">
        <span class="u-screenreader-only">Data Mining and Knowledge Discovery</span>
        <img class="test-cover-image" src="5.jpg" itemprop="image" alt="">
    </a>


                            </div>
                        </div>
                    </aside>
                    <div class="main-body" data-role="NavigationContainer">
                                <div class="cta-button-container cta-button-container--top cta-button-container--stacked u-mb-16 u-hide-two-col">
                    <div>
            <a href="https://link.springer.com/content/pdf/10.1007%2Fs10618-014-0367-9.pdf" target="_blank" class="c-button c-button--blue c-button__icon-right gtm-pdf-link" title="Download this article in PDF format" rel="noopener">
                <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" version="1.1"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g fill="#fff"><g transform="translate(12.000000, 5.000000)"><path d="M7 7.3L7 1C7 0.4 6.6 0 6 0 5.4 0 5 0.4 5 1L5 7.3 3.5 5.7C3.1 5.3 2.5 5.3 2.1 5.7L2.1 5.7C1.7 6.1 1.7 6.7 2.1 7.1L5.3 10.3C5.7 10.7 6.3 10.7 6.7 10.3L9.9 7.1C10.3 6.7 10.3 6.1 9.9 5.7L9.9 5.7C9.5 5.3 8.9 5.3 8.5 5.7L7 7.3 7 7.3ZM0 13C0 12.4 0.5 12 1 12L11 12C11.6 12 12 12.4 12 13 12 13.6 11.5 14 11 14L1 14C0.4 14 0 13.6 0 13L0 13Z"></path></g></g></g></svg>
                <span class="hide-text-small">Download</span>
                <span>PDF</span>
            </a>
        </div>

        </div>



                        <article class="main-body__content">
                            <div xmlns="http://www.w3.org/1999/xhtml" class="FulltextWrapper"><div class="ArticleHeader main-context"><div id="enumeration" class="enumeration"><p><a href="https://link.springer.com/journal/10618" title="Data Mining and Knowledge Discovery"><span class="JournalTitle">Data Mining and Knowledge Discovery</span></a></p><p class="icon--meta-keyline-before"><span class="ArticleCitation_Year"><time datetime="2014-09">September 2014</time>, </span><span class="ArticleCitation_Volume">Volume 28, </span><a class="ArticleCitation_Issue" href="https://link.springer.com/journal/10618/28/5/page/1">Issue&nbsp;56</a>,
                       <span class="ArticleCitation_Pages"> pp 15541585</span><span class="u-inline-block u-ml-4"> | <a class="gtm-cite-link" href="#citeas">Cite as</a></span></p></div><div class="MainTitleSection"><h1 class="ArticleTitle" lang="en">Classy: fast clustering streams of call-graphs</h1></div><div class="authors u-clearfix authors--enhanced" data-component="SpringerLink.Authors"><ul class="u-interface u-inline-list authors__title" data-role="AuthorsNavigation"><li><a href="#authors" class="gtm-tab-authors selected">Authors</a></li><li><a href="#authorsandaffiliations" class="gtm-tab-authorsandaffiliations">Authors and affiliations</a></li></ul><span class="marker" style="width: 160px;"></span><div class="authors__list" data-role="AuthorsList" id="authors" tabindex="-1"><ul class="test-contributor-names"><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors__name">Orestis&nbsp;Kostakis</span><span class="author-information"><span class="authors__contact"><a href="mailto:orestis.kostakis@f-secure.com" title="orestis.kostakis@f-secure.com" itemprop="email" class="gtm-email-author"><img src="email.svg" alt="Email author" width="24" height="24"></a></span></span></li></ul></div><div class="authors__affiliations" id="authorsandaffiliations" tabindex="-1"><div class="authors-affiliations u-interface"><ul class="test-contributor-names"><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors-affiliations__name">Orestis&nbsp;Kostakis</span><ul class="authors-affiliations__indexes u-inline-list" data-role="AuthorsIndexes"><li data-affiliation="affiliation-1">1</li><li data-affiliation="affiliation-2">2</li></ul><span class="author-information"><span class="author-information__contact u-icon-before icon--email-before"><a href="mailto:orestis.kostakis@f-secure.com" title="orestis.kostakis@f-secure.com" itemprop="email" class="gtm-email-author" tabindex="-1">Email author</a></span></span></li></ul><ol class="test-affiliations"><li class="affiliation" data-test="affiliation-1" data-affiliation-highlight="affiliation-1" itemscope="" itemtype="http://schema.org/Organization"><span class="affiliation__count">1.</span><span class="affiliation__item"><span itemprop="name" class="affiliation__name">Labs, F-Secure</span><span itemprop="address" itemscope="" itemtype="http://schema.org/PostalAddress" class="affiliation__address"><span itemprop="addressRegion" class="affiliation__city">Helsinki</span><span itemprop="addressCountry" class="affiliation__country">Finland</span></span></span></li><li class="affiliation" data-test="affiliation-2" data-affiliation-highlight="affiliation-2" itemscope="" itemtype="http://schema.org/Organization"><span class="affiliation__count">2.</span><span class="affiliation__item"><span itemprop="name" class="affiliation__name">Aalto University</span><span itemprop="address" itemscope="" itemtype="http://schema.org/PostalAddress" class="affiliation__address"><span itemprop="addressRegion" class="affiliation__city">Espoo</span><span itemprop="addressCountry" class="affiliation__country">Finland</span></span></span></li></ol></div></div></div><div class="main-context__container" data-component="SpringerLink.ArticleMetrics"><div class="main-context__column"><span><span class="test-render-category">Article</span></span><div class="article-dates article-dates--enhanced" data-component="SpringerLink.ArticleDates"><div class="article-dates__entry"><span class="article-dates__label">First Online: </span><span class="article-dates__first-online"><a href="#article-dates-history" class="gtm-first-online" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-article-dates">10 July 2014</a></span></div><div class="article-dates__history" id="article-dates-history"><div class="article-dates__entry"><span class="article-dates__label">Received: </span><span><time datetime="2014-02-28">28 February 2014</time></span></div><div class="article-dates__entry"><span class="article-dates__label">Accepted: </span><span><time datetime="2014-06-18">18 June 2014</time></span></div></div></div></div><div class="main-context__column">    <ul id="book-metrics" class="article-metrics u-sansSerif">
            <li class="article-metrics__item">
                     <span class="article-metrics__views">607</span>
                     <span class="article-metrics__label">Downloads</span>
            </li>
            <li class="article-metrics__item">
                    <a class="article-metrics__link gtm-citations-count" href="https://citations.springer.com/item?doi=10.1007/s10618-014-0367-9" target="_blank" rel="noopener" title="Visit Springer Citations for full citation details" id="citations-link">
                            <span id="citations-count-number" class="test-metric-count c-button-circle gtm-citations-count">10</span>
                       <span class="test-metric-name article-metrics__label gtm-citations-count">Citations</span>
                    </a>
            </li>
    </ul>
</div></div></div><section class="Abstract" id="Abs1" tabindex="-1" lang="en"><h2 class="Heading">Abstract</h2><p class="Para">An
 abstraction resilient to common malware obfuscation techniques is the 
call-graph. A call-graph is the representation of an executable file as a
 directed graph with labeled vertices, where the vertices correspond to 
functions and the edges to function calls. Unfortunately, most of the 
interesting graph comparison problems, including full-graph comparison 
and computing the largest common subgraph, belong to the <span class="InlineEquation" id="IEq1">\(NP\)</span>-hard
 class. This makes the study and use of graphs in large scale systems 
difficult. Existing work has focused only on offline clustering and has 
not addressed the issue of clustering streams of graphs. In this paper 
we present Classy, a scalable distributed system that clusters streams 
of large call-graphs for purposes including automated malware 
classification and facilitating malware analysts. Since algorithms aimed
 at clustering sets are not suitable for clustering streams of objects, 
we propose the use of a clustering algorithm that relies on the notion 
of candidate clusters and reference samples therein. We demonstrate via 
thorough experimentation that this approach yields results very close to
 the offline optimal. Graph similarity is determined by computing a 
graph edit distance (GED) of pairs of graphs using an adapted version of
 simulated annealing. Furthermore, we present a novel lower bound for 
the GED. We also study the problem of approximating statistics of 
clusters of graphs when the distances of only a fraction of all possible
 pairs have been computed. Finally, we present results and statistics 
from a real production-side system that has clustered and contains more 
than 0.8 million graphs.</p></section><div class="KeywordGroup" lang="en"><h2 class="Heading">Keywords</h2><span class="Keyword">Clustering&nbsp;</span><span class="Keyword">Streams&nbsp;</span><span class="Keyword">Call-graphs&nbsp;</span><span class="Keyword">Malware&nbsp;</span><span class="Keyword">Graph edit distance&nbsp;</span></div><div class="HeaderArticleNotes"><aside class="ArticleNote ArticleNoteCommunicatedBy"><p class="SimplePara">Responsible editors: Toon Calders, Floriana Esposito, Eyke Hllermeier, Rosa Meo.</p></aside><aside class="ArticleNote ArticleNoteMisc"><p class="SimplePara">The work was done while the author was with F-Secure.</p></aside></div><div class="article-actions--inline" id="article-actions--inline" data-component="article-actions--inline"><div class="citations u-interface c-button-dropdown" data-component="SV.Dropdown" data-namespace="citations--inline" aria-label="button with dropdown options">
        <button id="button-Dropdown-citations--inline-dropdown" type="button" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown-citations--inline-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">
    <span>Cite</span>
    <span class="hide-text-small">article</span>
</span><span class="c-button-dropdown__icon"></span></button>
<div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown-citations--inline-dropdown"><ul class="citations__content" data-role="button-dropdown__content">
    <li>
        <a href="#citeas" class="gtm-cite-dropdown">How to cite?</a>
    </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-014-0367-9?format=refman&amp;flavour=citation" title="Download this article's citation as a .RIS file" class="gtm-export-citation" data-gtmlabel="RIS">
                <span class="citations__extension" data-gtmlabel="RIS">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .RIS
                </span>
                <span class="citations__types">
                        <span>
                            Papers
                        </span>
                        <span>
                            Reference Manager
                        </span>
                        <span>
                            RefWorks
                        </span>
                        <span>
                            Zotero
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-014-0367-9?format=endnote&amp;flavour=citation" title="Download this article's citation as a .ENW file" class="gtm-export-citation" data-gtmlabel="ENW">
                <span class="citations__extension" data-gtmlabel="ENW">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .ENW
                </span>
                <span class="citations__types">
                        <span>
                            EndNote
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-014-0367-9?format=bibtex&amp;flavour=citation" title="Download this article's citation as a .BIB file" class="gtm-export-citation" data-gtmlabel="BIB">
                <span class="citations__extension" data-gtmlabel="BIB">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .BIB
                </span>
                <span class="citations__types">
                        <span>
                            BibTeX
                        </span>
                        <span>
                            JabRef
                        </span>
                        <span>
                            Mendeley
                        </span>
                </span>
            </a>
        </li>
</ul></div>
    </div></div><div id="body"><section id="Sec1" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">1 </span>Introduction<span class="section-icon"></span></h2><div class="content"><p class="Para">The
 number of uniques files produced daily has long surpassed the 
capabilities of manual analysis. Meanwhile, malware (malicious software)
 remains a highly prevalent threat for computer systems, their users and
 ICT infrastructure. In recent years, we have witnessed cases of malware
 that steals credit card information, users accounts and 
Internet-banking credentials, extorts users into paying ransom and even 
targets uranium enrichment centrifuges in nuclear reactors. The total 
cost incurred on the global economy by the presence of malware has 
established the detection of malicious files as a major task for data 
security organizations. Furthermore, a continuous race is underway 
between malware authors and security researchers, in which the first 
invent new attack vectors and the latter reactively create defense 
mechanisms. To facilitate the work of security researchers and analysts,
 enhanced automation is imperative.</p><div class="Para">Over the past 
years, researchers have focused on executable file analysis techniques. 
These techniques can be divided into two categories: behavioral 
(runtime) analysis and static analysis, with each one having its own 
advantages and limitations. The most widely used automated methods apply
 runtime analysis by executing files in controlled (sandbox) 
environments (Rieck et al. <span class="CitationRef"><a href="#CR49" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2008</a></span>; Bayer et al. <span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>; Moser et al. <span class="CitationRef"><a href="#CR45" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007a</a></span>; Willems et al. <span class="CitationRef"><a href="#CR63" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>; Kolbitsch et al. <span class="CitationRef"><a href="#CR37" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>).
 However, malware authors have developed a plethora of 
anti-virtualization techniques to detect whether the file is executed on
 a computer in the wild or in a labs isolated environment. In addition,
 instances of malware might require special conditions, temporal or 
environmental, to hold in order to express any behavior. This is not the
 case with static analysis, which attempts to analyze a file by 
statically examining its code (Briones and Gomez <span class="CitationRef"><a href="#CR7" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2008</a></span>; Carrera and Erdlyi <span class="CitationRef"><a href="#CR11" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>; Dullien and Rolles <span class="CitationRef"><a href="#CR18" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2005</a></span>; Flake <span class="CitationRef"><a href="#CR21" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>; Hu et al. <span class="CitationRef"><a href="#CR31" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>; Kinable and Kostakis <span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>; Kostakis et al. <span class="CitationRef"><a href="#CR39" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>).
 On the other hand, code re-ordering, routine re-ordering, self-mutation
 and code-obfuscation techniques are used to evade pattern-based 
detections (Christodorescu and Jha <span class="CitationRef"><a href="#CR14" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>).
 While pattern-based signatures could be modified to take into account 
such modifications by searching over the whole file, this solution would
 make the scanning process slower and more prone to false alarms. An 
abstraction resilient to such attempts against static analysis is the <em class="EmphasisTypeItalic ">call-graph</em> (Ryder <span class="CitationRef"><a href="#CR51" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1979</a></span>).
 A call-graph is the representation of an executable file as a directed 
graph with labeled vertices, where the vertices correspond to functions 
and the edges to function calls. An example of a very small call-graph 
is depicted in Fig.&nbsp;<span class="InternalRef"><a href="#Fig1">1</a></span>. For modern software it is not unusual to encounter call-graphs of several thousands of vertices.<figure class="Figure" id="Fig1"><div class="MediaObject" id="MO17"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Fig1_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Fig1_HTML.gif" alt="Fig. 1"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 1</span> <p class="SimplePara">A very small call-graph. Call-graphs are directed graphs with labeled vertices. Each <em class="EmphasisTypeItalic ">vertex</em> corresponds to a function and <em class="EmphasisTypeItalic ">edges</em> represent function calls. Call-graphs may contain cycles and self-loops </p> </div></figcaption></figure> </div><p class="Para">On
 a daily basis, data security companies currently receive hundreds of 
thousand of unique files. These files are submitted by clients, partners
 or other entities and form a constant stream of data. For each incoming
 sample, it has to be decided whether it is of malicious nature or not. 
At the same time, the number of unique new files is increasing 
constantly. Due to that, automated means of analysis are essential. A 
step toward achieving that is possessing the ability to identify similar
 files and group them together. For this purpose, we propose Classy, a 
system that clusters streams of call-graphs.</p><p class="Para">The 
benefits of a system like Classy is two-fold. First, it can be turned 
into an automated malware classification service by applying supervised 
learning through the clustering results. Second, it facilitates the 
whole stack of the Security Response Unit of an AV company; from the 
customer-care team, to the analysts dissecting malware and writing 
detections and the data scientists that strive to comprehend every 
aspect of malware and create new prototype solutions. For example, in 
the case of analysts writing pattern-based signatures, by providing them
 with real-time clustering information, Classy facilitates their work 
since they become aware of as many relevant samples as possible. Thus, 
the end result provides better and more accurate detection coverage, and
 the fact that information is provided in real time allows to save 
critical time for protecting the end-users. Alternatively, by examining a
 group of related malware samples, the analyst might decide that a 
different approach must be taken to better detect them in the future. 
Finally, the added knowledge gained from the clustering results may also
 be used for prioritizing samples in the queues of other automation 
systems.</p><p class="Para">Due to the fact that most interesting graph 
comparison-related problems are computationally hard, the use of 
call-graphs in large-scale systems becomes difficult. In particular, 
going from a restricted graph clustering application or offline 
experiment to a full-fledged online stream-clustering scalable 
production-side service is a very difficult and ambitious task. This is 
the fact that there has not yet been developed, to the best of our 
knowledge, a system such as the one that we describe.</p><p class="Para">In
 this work, we focus on clustering streams of data where each data 
object is a whole graph. This differs from the literature that shares 
the same title and deals with data that are part of one big graph, as in
 (Schaeffer <span class="CitationRef"><a href="#CR52" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>; Kulis et al. <span class="CitationRef"><a href="#CR41" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>; Zhou et al. <span class="CitationRef"><a href="#CR70" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>) and in social network analysis work (Mishra et al. <span class="CitationRef"><a href="#CR44" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>; Kollios et al. <span class="CitationRef"><a href="#CR38" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2013</a></span>).
 Consequently, our work differs from the existing publications that 
refer to clustering streams of graphs where the edges or vertices of a 
single large graph arrive in a streaming fashion (Aggarwal et al. <span class="CitationRef"><a href="#CR3" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2010</a></span>).
 In our application, one of the ultimate goals of clustering streams of 
graphs is to identify similarities between executables so the 
information can be used for facilitating the work of analysts in 
real-time and for automated malware classification. Due to that, the 
elapsed time from the moment a sample enters the system until it has 
been clustered should be as short as possible. Given the rapid pace at 
which malware outbreaks occur, obtaining a result within an hour or even
 minutes is essential; most of the top anti-virus software vendors 
provide in-the-cloud services for real-time propagation of malware 
detections. An additional constraint is that since the system is 
operation-critical, there should not be any downtime for reasons such as
 re-computing the indexing structure. In our model of the stream, we can
 access data elements (in our case graphs) that have arrived in earlier 
time instances but we must handle every incoming data object 
independently and at the moment it arrives. Finally, and in addition to 
the reasons mentioned above, algorithms aimed at clustering sets of 
objects are not suitable for clustering streams. So, we face an 
additional challenge when designing such a system.</p><p class="Para">Our
 work focuses on call-graphs extracted from PE32 (Windows executable) 
files. However, the same file extraction techniques and clustering 
approach can be directly applied on Apple Disk Image (.dmg) files when 
the amount of malware for MacOS deems such system necessary. In general,
 our graph comparison and clustering methods are suitable for any 
application where graphs are data objects that arrive as a stream and 
must be clustered.</p><div class="Para">This paper mainly addresses the 
problem of transitioning from offline clustering to clustering streams 
of graphs in the context of a live production system. As an application 
of our technique, we use it to cluster call-graphs of binary executable 
files. The main contributions in this paper include:<div class="UnorderedList"><ul class="UnorderedListMarkBullet"><li> <p class="Para">Classy,
 a scalable distributed system for fast clustering streams of 
call-graphs with the aim of facilitating the work of security analysts 
and automated malware classification. The graphs can consist of up to 
several thousands of vertices. The similarity of the graphs is 
determined on the basis of their graph edit distance (GED), a problem 
that is known to belong in the <span class="InlineEquation" id="IEq2">\(NP\)</span>-hard
 complexity class. We are able to approximate the edit distance of 
graphs by using an adapted version of simulated annealing (SA). To the 
best of our knowledge, we are the first to present a system for 
clustering streams of graphs and especially graphs of such size.</p> </li><li> <p class="Para">For
 the clustering phase, due to the fact that algorithms aimed at 
clustering sets of objects are not suitable for clustering streams, we 
propose and demonstrate the efficiency of an algorithm which relies on 
the notion of candidate clusters and reference samples in each cluster. 
We demonstrate, via thorough experimentation, that this approach yields 
results very close to the offline optimal; as those were presented in 
Kinable and Kostakis (<span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>).</p> </li><li> <p class="Para">We present a novel lower bound of the GED with <span class="InlineEquation" id="IEq3">\(O(n)\)</span>
 time complexity that provides very high tightness and pruning power. 
The computation of the lower bound is used to prune the set of candidate
 clusters of the incoming sample.</p> </li><li> <p class="Para">We study
 the problem of approximating statistics of clusters of graphs when the 
distances of only a fraction of all possible pairs have been computed.</p> </li><li> <p class="Para">Finally,
 we present experimental results for those algorithms and provide 
statistics from a real active production-side system that has clustered 
and contains more than 0.8 million graphs. The results prove that such a
 system is highly valuable and under circumstances reliable enough to 
provide automatic classifications.</p> </li></ul></div> </div></div></section><section id="Sec2" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">2 </span>Related work<span class="section-icon"></span></h2><div class="content"><p class="Para">The study of call-graphs in the context of malware analysis was introduced by Flake in his seminal paper (Flake <span class="CitationRef"><a href="#CR21" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>). Similarly in his work jointly with Rolles (Dullien and Rolles <span class="CitationRef"><a href="#CR18" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2005</a></span>),
 the goal is to find differences between modified versions of software. 
By using the graph representation of executables and automatically 
finding commonalities, the work of a malware analyst is facilitated and 
reduced to having only to analyze the non-common sections.</p><p class="Para">Carrera and Erdlyi (<span class="CitationRef"><a href="#CR11" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>)
 used call-graphs to create taxonomies for rapid analysis of multiple 
variants belonging to the same malware family. They use call-vector 
comparisons to heuristically match functions in an iterative 
match-and-filter manner. Improving on their work, Briones and Gomez (<span class="CitationRef"><a href="#CR7" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2008</a></span>)
 incorporate control flow graphs of functions in the attempt to create a
 system which finds similar variants of malware from an existing DB. 
Recently, the first large-scale work on indexing and retrieving 
call-graphs, called SMIT, was presented (Hu et al. <span class="CitationRef"><a href="#CR31" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>).
 SMIT contains only malware, is intended to be used only by malware 
analysts and uses an improved version of bipartite matching for the 
graph comparison and the combination of <span class="InlineEquation" id="IEq4">\(B+\)</span>
 and VP trees for the indexing. However, despite its usefulness, the 
purpose of SMIT is to index call-graphs in order to provide <span class="InlineEquation" id="IEq5">\(k\)</span>-NN search capability to the analysts and it is not intended for online clustering. Due to that, its complex combination of <span class="InlineEquation" id="IEq6">\(B+\)</span>-trees
 and VP trees results into having to perform an excess amount of 
expensive graph comparisons per query; this becomes clear in Sect.&nbsp;<span class="InternalRef"><a href="#Sec18">8.1.2</a></span>. Finally, Kinable and Kostakis (<span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>)
 studied the feasibility of using call-graph clustering to identify 
groups of similar executables with the ultimate intention of 
automatically classifying malware.</p><p class="Para">All the 
aforementioned results have inspired multiple recent works that focus on
 variants of the call-graph or the trade-offs between accuracy and 
complexity requirements. For example, (Xu et al. <span class="CitationRef"><a href="#CR66" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2013</a></span>; Veeramani and Rai <span class="CitationRef"><a href="#CR59" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2012</a></span>; Bourquin et al. <span class="CitationRef"><a href="#CR6" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2013</a></span>; Elhadi et al. <span class="CitationRef"><a href="#CR19" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2013</a></span>) have focused on Windows PE32 files and (Gascon et al. <span class="CitationRef"><a href="#CR23" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2013</a></span>) have focused on Android <span class="EmphasisFontCategoryNonProportional ">APK</span> files. The two works closer to Classy are SMIT and (Kinable and Kostakis <span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>).
 Both of them use a similar graph edit distance as the distance function
 between graphs. The basic edit operations of the edit distance resemble
 very well the edit operations performed on executable files. Due to the
 fact that computing the graph edit distance is NP-complete, for 
approximating it in polynomial time, SMIT uses bipartite-matching that 
takes <span class="InlineEquation" id="IEq7">\(O(n^3)\)</span> time, while (Kinable and Kostakis <span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>)
 employ a Simulated Annealing approach that has been demonstrated to be 
faster and more accurate than bipartite matching (Kostakis et al. <span class="CitationRef"><a href="#CR39" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>). Furthermore, in (Kinable and Kostakis <span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>; Bunke <span class="CitationRef"><a href="#CR8" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1997</a></span>) the connection between GED and the Maximum Common Subgraph (MCS) is established; both problems are <span class="InlineEquation" id="IEq8">\(NP\)</span>-hard.</p><p class="Para">Reducing
 the graph comparison problem to string or tree comparison creates 
ambiguities and violates basic distance function requirements, such as <span class="InlineEquation" id="IEq9">\(\delta (G,H)&gt;0\)</span> for <span class="InlineEquation" id="IEq10">\(G\ne H\)</span>,
 that in turn may cause severe false positives. Furthermore, for the 
case of call-graphs, no assumption can be made regarding the statistical
 properties of the graphs. For instance, we cannot assume that the 
graphs have a bounded degree, nor can we assume any restriction on the 
(edge) density of the call-graphs. This renders methods designed for 
molecular graphs, such as (Schietgat et al. <span class="CitationRef"><a href="#CR53" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2013</a></span>), or other types of graphs, such as almost-trees (Akutsu <span class="CitationRef"><a href="#CR4" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1993</a></span>),
 unsuitable for our application. Similarly, call-graphs may contain 
cycles and self-loops since those correspond to valid programming 
paradigms such as recursion. Consequently, methods designed exclusively 
for directed acyclic graphs (DAGs) (Lin and Kung <span class="CitationRef"><a href="#CR42" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1997</a></span>)
 are inapplicable, too. The interested reader may find an extensive 
review of matching algorithms for different types of graphs in (Conte et
 al. <span class="CitationRef"><a href="#CR15" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>).
 In general, any approximation approach would need to be extensively 
tested and its suitability must be benchmarked in the context of 
comparing call-graphs.</p><p class="Para">In recent years, graph kernels
 have been introduced for the task of learning structured data and have 
found success in the field of bioinformatics. They can be divided into 
three main categories: graph kernels based on walks or paths, based on 
subgraphs and based on subtrees. The work by Vishwanathan et al. (<span class="CitationRef"><a href="#CR61" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2010</a></span>) improves the time complexity of the random-walk kernel computation between unlabeled graphs from <span class="InlineEquation" id="IEq11">\(O(n^6)\)</span> to <span class="InlineEquation" id="IEq12">\(O(n^3)\)</span>. However, call-graphs are labeled and in that case their method takes <span class="InlineEquation" id="IEq13">\(O(dn^3)\)</span> time per iteration, where <span class="InlineEquation" id="IEq14">\(d\)</span> is the size of the label set. Recently, Kriege and Mutzel (<span class="CitationRef"><a href="#CR40" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2012</a></span>) have proposed subgraph kernels for attributed graphs. However, there method taken <span class="InlineEquation" id="IEq15">\(O(kn^{k+1}_p)\)</span>, where <span class="InlineEquation" id="IEq16">\(n_p\)</span> = <span class="InlineEquation" id="IEq17">\(|V_{G}|\cdot |V_H|\)</span> and <span class="InlineEquation" id="IEq18">\(k\)</span> is the allowed recursion depth. Another option is the classic subtree kernel (Ramon and Grtner <span class="CitationRef"><a href="#CR48" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2003</a></span>) that takes <span class="InlineEquation" id="IEq19">\(O(n^24^dh)\)</span> time, where <span class="InlineEquation" id="IEq20">\(d\)</span> is the maximum degree, and is clearly inefficient for call-graphs since the maximum degree can be equal to <span class="InlineEquation" id="IEq21">\(n\)</span>.</p><p class="Para">Recently, the family of WeisfeilerLehman (WL) graph kernels have been proposed (Shervashidze et al. <span class="CitationRef"><a href="#CR55" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>) as a framework that promises fast comparisons. Still, the WL shortest path kernel requires <span class="InlineEquation" id="IEq22">\(O(n^4)\)</span> time for comparing a pair of graphs, while the WL edge kernel takes <span class="InlineEquation" id="IEq23">\(O(m^2)\)</span> per iteration (where <span class="InlineEquation" id="IEq24">\(m\)</span> is the number of edges, in call-graphs <span class="InlineEquation" id="IEq25">\(m\)</span> can be close to <span class="InlineEquation" id="IEq26">\(n^2\)</span>).
 While the WL subtree kernel appears promising in terms of complexity, 
in the experiments of the original work, it performs worst than the 
other two benchmarked WL graph kernels for most test cases in terms of 
classification accuracy. In addition, WL kernels have been proposed for 
undirected graphs, while directivity in call-graphs is a very important 
aspect. Transforming a directed graph to an undirected simply by 
ignoring the direction of the arcs would make different call-graphs to 
appear as similar. This is highly undesired for the application of 
malware clustering since pairing different files could have detrimental 
results. Other potentially more complicated methods for transforming 
directed graphs to undirected would need to take into account 
additionally the time and space efficiency requirements related to 
clustering streams of call-graphs. Nevertheless, studying the 
applicability of graph kernels for classifying call-graphs would benefit
 the field of malware detection. Before choosing them as the graph 
comparison method in the stream setting, the performance and behavior of
 graph kernels needs to be thoroughly examined offline. However, the 
clustering algorithm that we propose is capable of handling the results 
of any distance function.</p><p class="Para">Furthermore, there are 
important limitations to applying graph kernels to the stream clustering
 setting. In the inductive setting where the kernel is computed based on
 a training set, and an unknown graph needs to be classified, the 
unknown graph must be mapped to the feature space of compressed labels 
occurring in the training set. This may additionally require to store, 
for each known graph, information related to the data structures 
encountered while computing the kernel. In particular for the 
Weisfeiler-Lehman kernels, this takes <span class="InlineEquation" id="IEq27">\(O(Nmh)\)</span> space, where <span class="InlineEquation" id="IEq28">\(N\)</span> is the number of graphs, <span class="InlineEquation" id="IEq29">\(m\)</span> is the number of edges and <span class="InlineEquation" id="IEq30">\(h\)</span> is the number of iterations; in our application <span class="InlineEquation" id="IEq31">\(m\)</span>
 is often in the order of thousands. Moreover, this introduces the 
problem of selecting and maintaining a training set that offers good 
coverage of malware and benign software, and is up to date with current 
and past threat landscapes; in general, this is a serious problem for 
large-scale malware classifiers. Furthermore, this would require that 
the kernel is computed periodically and that would result to down-time 
of the system, or additional infrastructure. If graph kernels were to be
 applied in a transductive setting, the test set would need to be known.
 While information about past data structures would not need to be 
stored, the majority of the computations would need to be repeated every
 time. Applying a trade-off between the number of repetitions of the 
computations and handling several files every time, would convert the 
stream setting into a batch setting. The incurred delay might not be 
tolerable in certain time-critical applications.</p><p class="Para">Since
 graphs are a very prevalent method to depict structure and relations, 
work on indexing and retrieving them exceeds the boundaries of the 
malware analysis domain. In the past years, significant work has been 
performed on this subject in the database field. Most of the work is 
focused on finding occurrences of small query subgraphs in databases of 
graphs (Cheng et al. <span class="CitationRef"><a href="#CR13" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>; Giugno and Shasha <span class="CitationRef"><a href="#CR26" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2002</a></span>; He and Singh <span class="CitationRef"><a href="#CR28" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>; Jiang et al. <span class="CitationRef"><a href="#CR34" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>; Tian and Patel <span class="CitationRef"><a href="#CR58" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2008</a></span>; Williams et al. <span class="CitationRef"><a href="#CR64" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>; Yan et al. <span class="CitationRef"><a href="#CR67" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2005</a></span>; Zhao et al. <span class="CitationRef"><a href="#CR69" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>).
 However, these cannot capture the typical modifications applied to 
software. Instead, we need to identify those graphs in the database 
which are similar to the query graph using full-graph comparison 
techniques. Closure-Tree (He and Singh <span class="CitationRef"><a href="#CR28" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>)
 was recently proposed as an indexing solution for this purpose but 
focuses only on very small graphs. Only SMIT, as mentioned earlier, 
focuses on large-scale graph indexing. In Sect.&nbsp;<span class="InternalRef"><a href="#Sec18">8.1.2</a></span> we demonstrate how our approach results to fewer graph comparisons than SMIT and other embedding-based approaches.</p><p class="Para">Significant work has also been published on clustering streams of data. Aggarwal et al. (<span class="CitationRef"><a href="#CR1" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2003</a></span>) proposed CluStream, a framework for clustering evolving data streams. Furthermore, they propose HPStream (Aggarwal et al. <span class="CitationRef"><a href="#CR2" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>)
 for projected clustering of high dimensional data. By defining clusters
 for groups of dimensions, they attempt to tackle the sparsity problem 
in high-dimensional spaces. HPStream was shown to be more effective than
 CluStreams full dimensional clustering. Guha et al. (<span class="CitationRef"><a href="#CR27" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2003</a></span>) presented a single-pass variation of <span class="InlineEquation" id="IEq32">\(k\)</span>-median which has a small memory footprint. Charikar et al. (<span class="CitationRef"><a href="#CR12" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2003</a></span>) improve on (Guha et al. <span class="CitationRef"><a href="#CR27" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2003</a></span>) with their own version of the <span class="InlineEquation" id="IEq33">\(k\)</span>-median technique. All the the above focus on the <span class="InlineEquation" id="IEq34">\(k\)</span>-means or <span class="InlineEquation" id="IEq35">\(k\)</span>-median approach which require an exact or maximum value for <span class="InlineEquation" id="IEq36">\(k\)</span>, the number of clusters. In our problem, the value of <span class="InlineEquation" id="IEq37">\(k\)</span> cannot be known in advance and can change over time due to the nature of streams. If one selects a small value for <span class="InlineEquation" id="IEq38">\(k\)</span>, highly different files would be clustered together, while a big value for <span class="InlineEquation" id="IEq39">\(k\)</span> would force the algorithm to divide the data, thus resulting in low grouping power. Furthermore, <span class="InlineEquation" id="IEq40">\(k\)</span>-means/median
 approaches tend to form spherical clusters, which is not always the 
case with clusters of call-graphs. Intuitively, the life cycle of modern
 software would suggest clusters of arbitrary shape.</p><p class="Para">In our method, instead of guessing the value for <span class="InlineEquation" id="IEq41">\(k\)</span>,
 it is enough to define a similarity threshold value; this is the 
limit on the distance value for two objects to be considered close. Such
 a threshold is a basic component of density-based methods. In (Kinable 
and Kostakis <span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>) it is shown that density-based clustering approaches should be preferred over <span class="InlineEquation" id="IEq42">\(k\)</span>-means or <span class="InlineEquation" id="IEq43">\(k\)</span>-median techniques for clustering call-graphs. Recently, Cao et al. (<span class="CitationRef"><a href="#CR10" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>)
 presented DenStream, an approach for density-based clustering for data 
streams suitable for identifying clusters of arbitrary shape. The method
 relies on the presence of potential micro-clusters and outlier 
micro-clusters and manages to handle large amounts of data by 
effectively pruning those two sets. DenStream, despite being efficient, 
cannot be applied out-of-the-box to our problem for the reason that it 
requires being able to identify the micro-clusters that are closest to 
the arriving point. This cannot be done in our case due to the fact 
that, first, our data are graphs and not points in a multi-dimensional 
space, and secondly, that computing the distances between graphs is very
 expensive. Thus it would be practically impossible to compute all the 
required distance values for an incoming graph. We must note that the 
methods proposed in this work can be combined with other methods to 
provide greater efficiency. Examining approaches that combine elements 
of other methods with our method would be a very interesting direction 
for future work.</p><p class="Para">Finally, well-established indexing methods such as Locality Sensitive Hashing (Datar et al. <span class="CitationRef"><a href="#CR16" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>; Gionis et al. <span class="CitationRef"><a href="#CR24" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1999</a></span>; Indyk and Motwani <span class="CitationRef"><a href="#CR33" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1998</a></span>)
 cannot be applied to our setting. For LSH in particular, it remains an 
open problem to derive a scheme for GED. Even for even easier problems 
such as the string edit distance there have been no solutions that 
satisfy any performance guarantees.</p></div></section><section id="Sec3" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">3 </span>Background<span class="section-icon"></span></h2><div class="content"><p class="Para">In this section we provide the reader with the necessary background. In Sect.&nbsp;<span class="InternalRef"><a href="#Sec4">3.1</a></span> we describe our data and the means used for pre-processing the binary executable files. In Sect.&nbsp;<span class="InternalRef"><a href="#Sec5">3.2</a></span> we introduce the used terminology and notation while in Sect.&nbsp;<span class="InternalRef"><a href="#Sec6">3.3</a></span> we provide a formal definition of our graph comparison problem.</p><section id="Sec4" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">3.1 </span>Data description and extraction</h3><p class="Para">Call-graphs
 are extracted using static analysis. In other words, the file is not 
executed but instead it is analyzed by examining its contents. To 
statically extract the call-graph from a binary file we use a very 
popular disassembler, IDA Pro (Hex-Rays <span class="CitationRef"><a href="#CR30" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2008</a></span>), in conjunction with IDA Python (Carrera and Erdlyi <span class="CitationRef"><a href="#CR11" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>)
 and our in-house unpacking suite. The product is an instruction-level 
disassembly of the binary file from which we are able to get a list of 
functions and inter-function calls. Basic block information and 
control-flow graphs can also be extracted. The functions are labeled and
 belong to two categories: (file-specific) local functions and imported 
(external) functions. For the latter, their names can be retrieved (e.g.
 GetSystemDirectoryA) and serve as the labels of those vertices, while 
for the first the returned names are based on the address of their 
starting point. The names of imported functions are common across 
executables and can be retrieved from the Import Address Table (IAT), 
for those external functions that are imported dynamically. 
Alternatively, statically linked library functions and their canonical 
names can be retrieved with tools such as IDA Pros FLIRT. Similarity of
 imported functions can be determined simply by comparing the function 
names. The same does not hold for the rest of the functions. No valid 
function calls are discarded. Thus, cycles and loops are present in our 
graph model.</p><p class="Para">Before a call-graph is extracted, it is 
first examined to determine whether it is packed or protected; in the 
remainder of this paper we will refer to both conditions simply as 
packed. Packing refers to techniques that are applied, in the usual 
case, during the post-compilation phase and aim to protect the 
intellectual property of the software author from reverse engineering 
efforts; typical examples include <span class="EmphasisFontCategoryNonProportional ">UPX</span> and <span class="EmphasisFontCategoryNonProportional ">ASPack</span>.
 Packers do not affect the behavior of an executable, but they alter its
 contents by compressing and/or obfuscating them. When a packed file is 
executed, an extraction component, also known as the unpackers stub, 
recovers (decompresses and/or decrypts) the original contents of the 
file, restores the execution context, and the execution resumes from the
 original starting point.</p><p class="Para">To detect whether a file is packed, it is common to use pattern-matching techniques. Tools such as PEiD (Snaker et al. <span class="CitationRef"><a href="#CR56" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>)
 contain signature databases for a series of known packers. Once a 
packer has been identified, one needs to apply the appropriate unpacker.
 This approach is fast and works well for the vast majority of known 
packers, but for each custom/unknown packer a new signature and the 
corresponding unpacking mechanism need to be added. An alternative 
solution is to use heuristics. Most of them can be considered as 
run-time unpacking tools (Kang et al. <span class="CitationRef"><a href="#CR35" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>; Martignoni et al. <span class="CitationRef"><a href="#CR43" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>)
 and may require an isolated environment. In general, unpacking 
heuristics are of questionable reliability and can be evaded. For 
example, they may fail under the presence of anti-virtualization and 
anti-emulator techniques. Finally, more than one packing technique may 
be applied simultaneously.</p><p class="Para">In our system, if the file
 is packed, it undergoes the appropriate unpacking procedure(s) to 
remove any such layer. This is necessary, otherwise the graph extraction
 would return the call-graph representation of the unpackers module. 
Creating the call-graphs of the unpackers stub would result in 
clustering binary executable files based on the used packer and not the 
actual executable. In the case of heuristic unpacking, the process may 
fail to produce a valid executable or the product may contain severe 
inconsistencies. In such case the file must be discarded, since the 
output may be arbitrary.</p><p class="Para">Finally, static analysis is 
preferred over run-time analysis due to the fact that the latter allows 
to analyze only the part of the code that managed to get executed. This 
is an important factor, since modern malware may apply a wide series of 
anti-debugging and anti-virtualization techniques and do not exhibit 
important behavior if they detect that they are being executed in a 
controlled (sandbox) environment. As a result, call-graphs produced via 
static analysis are more accurate and provide greater code coverage. In 
turn, static analysis suffers from several limitations of its own. Most 
importantly, some values, such as register values, are resolved at 
run-time; for example <span class="EmphasisFontCategoryNonProportional ">call eax;</span>. Recovering those values statically can be NP-hard (Moser et al. <span class="CitationRef"><a href="#CR46" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007b</a></span>).
 Similarly, static analysis retrieves all function calls regardless of 
whether they are performed during execution. Determining which function 
calls (control flow transfers) may actually be performed is undecidable 
in the general case.</p><p class="Para">However, one could combine 
static and run-time analysis to extract the call-graph from the memory 
image; limitations regarding the need for parts of the code to be 
executed still apply. Furthermore, the most important challenge is 
determining the appropriate moment to dump the memory to disk, and 
identifying the appropriate memory regions that contain the interesting 
code, in a manner that allows the whole procedure to remain effective, 
efficient and accurate. Still, any combination would require 
infrastructure for both techniques, but our system can work with such 
graphs too. Techniques and methods on improving the accuracy of the 
call-graph extraction would be highly beneficial for our system, but 
this is beyond the scope of the current work.</p><p class="Para">After a
 preliminary study of call-graphs and in conjunction with the malware 
threat landscape, we have decided to handle only graphs of up to 3000 
vertices. In addition, extensive sanity checks are performed on the 
call-graphs to determine the validity of the output. This includes 
simple checks on the number of edges and vertices and more complex 
checks such as verifying the validity of certain structures created by 
the compiler. We prefer to discard samples for which we are not 
confident about the validity of their call-graph rather than erroneously
 clustering them, which could ultimately result in a severe false 
detection.</p></section><section id="Sec5" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">3.2 </span>Graphs</h3><p class="Para">A <em class="EmphasisTypeItalic ">call-graph</em> is a directed graph <span class="InlineEquation" id="IEq44">\(G=(V,E)\)</span> with labeled vertices, where each vertex (or node) <span class="InlineEquation" id="IEq45">\(u\in V\)</span> represents a program function and each edge, or <em class="EmphasisTypeItalic ">arc</em>, <span class="InlineEquation" id="IEq46">\((u,v)\in E\)</span> a call of function <span class="InlineEquation" id="IEq47">\(v\in V\)</span> by function <span class="InlineEquation" id="IEq48">\(u\in V\)</span>. Between any ordered pair of vertices, there can be at most one arc. <span class="InlineEquation" id="IEq49">\(u_N\)</span> denotes the label, or function name, of <span class="InlineEquation" id="IEq50">\(u\in V\)</span>. In directed graphs, the direction of the edges matters; <span class="InlineEquation" id="IEq51">\((u,v)\in E\)</span> is different than <span class="InlineEquation" id="IEq52">\((v,u)\in E\)</span>. The outdegree <span class="InlineEquation" id="IEq53">\(d^+(u)\)</span> of a vertex <span class="InlineEquation" id="IEq54">\(u\)</span> is the number of its outgoing edges <span class="InlineEquation" id="IEq55">\(|\{(u,v)\in E | v\in V\}|\)</span>, while its indegree <span class="InlineEquation" id="IEq56">\(d^-(u)\)</span> is the number of its incoming edges. The sum of the outdegree and indegree of a vertex is simply its degree. For a graph <span class="InlineEquation" id="IEq57">\(G\)</span>, we denote the outdegree sequence of its nodes in ascending order by <span class="InlineEquation" id="IEq58">\(D^{+}_{G}\)</span>. <span class="InlineEquation" id="IEq59">\(u^{G+}_i\)</span> denotes the vertex whose outdegree is in the <span class="InlineEquation" id="IEq60">\(i\)</span>-th position of <span class="InlineEquation" id="IEq61">\(D^{+}_{G}\)</span>. Symbols <span class="InlineEquation" id="IEq62">\(D^{-}_{G}\)</span> and <span class="InlineEquation" id="IEq63">\(u^{G-}_i\)</span> denote the same for the ordered indegree sequence.</p><p class="Para">The
 vertices can be split into two main categories: local (file-specific) 
functions and external (imported) functions. The number of vertices in 
graph <span class="InlineEquation" id="IEq64">\(G\)</span>, <span class="InlineEquation" id="IEq65">\(|V_G|\)</span>, is the <em class="EmphasisTypeItalic ">order</em> of <span class="InlineEquation" id="IEq66">\(G\)</span>, while the number of edges, <span class="InlineEquation" id="IEq67">\(|E_G|\)</span>, is the <em class="EmphasisTypeItalic ">size</em> of <span class="InlineEquation" id="IEq68">\(G\)</span>. The set of external functions of <span class="InlineEquation" id="IEq69">\(G\)</span> is denoted by <span class="InlineEquation" id="IEq70">\(S_G\)</span> </p></section><section id="Sec6" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">3.3 </span>Graph edit distance</h3><div class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading"> <strong class="EmphasisTypeBold ">Definition 1</strong> </span></h3><p class="Para">(<em class="EmphasisTypeItalic ">Graph edit distance</em>) The Graph Edit Distance between two graphs <span class="InlineEquation" id="IEq71">\(G,H\)</span> is the minimum cost induced by transforming graph <span class="InlineEquation" id="IEq72">\(G\)</span> into <span class="InlineEquation" id="IEq73">\(H\)</span> using elementary edit operations. Each edit operation is assigned a cost value.</p></div><p class="Para">In
 our case, these elementary edit operations are: adding/ removing a 
vertex, adding/ removing an edge and relabeling a vertex. We assign unit
 costs to all operations. Computing the GED of a pair of graphs is <span class="InlineEquation" id="IEq74">\(NP\)</span>-hard (Riesen and Bunke <span class="CitationRef"><a href="#CR50" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>) and makes use of graph matchings; these are bijective functions mapping the vertices of one graph to those of the other.</p><div class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading"> <strong class="EmphasisTypeBold ">Definition 2</strong> </span></h3><p class="Para">(<em class="EmphasisTypeItalic ">Graph matching</em>) For two graphs, <span class="InlineEquation" id="IEq75">\(G\)</span> and <span class="InlineEquation" id="IEq76">\(H\)</span>, of equal order, the graph matching problem consists of finding a bijective mapping <span class="InlineEquation" id="IEq77">\(\phi \)</span> that maps the vertices in <span class="InlineEquation" id="IEq78">\(V_G\)</span> to those in <span class="InlineEquation" id="IEq79">\(V_H\)</span> such that <span class="InlineEquation" id="IEq80">\(\forall i,j\in V_G, (i,j)\in E_G\)</span>, if and only if <span class="InlineEquation" id="IEq81">\((\phi (i),\phi (j))\in E_H\)</span>. If such a matching exists, <span class="InlineEquation" id="IEq82">\(\phi \)</span> is an isomorphism between <span class="InlineEquation" id="IEq83">\(G\)</span> and <span class="InlineEquation" id="IEq84">\(H\)</span> and the two graphs are isomorphic.</p></div><p class="Para">In
 the case of graphs with labeled or attributed vertices, the above 
definition is extended so that the mapping preserves the vertex labels 
or attributes.</p><div class="Para">The above definition requires that <span class="InlineEquation" id="IEq85">\(G\)</span> and <span class="InlineEquation" id="IEq86">\(H\)</span>
 have the same number of vertices. This is hardly the case in general. 
This can be circumvented by adding extra (dummy) zero-degree vertices,
 denoted by <span class="InlineEquation" id="IEq87">\(\epsilon \)</span>, to the one with the fewer vertices. This results to the augmented versions <span class="InlineEquation" id="IEq88">\(G'=(V_{G'},E_{G'})\)</span> and <span class="InlineEquation" id="IEq89">\(H'=(V_{H'},E_{H'})\)</span>, with <span class="InlineEquation" id="IEq90">\(|V_{G'}|=|V_{H'}|\)</span>. Matching a vertex <span class="InlineEquation" id="IEq91">\(u\in V_{G'}\)</span> to an added vertex <span class="InlineEquation" id="IEq92">\(\epsilon \)</span> corresponds to deleting <span class="InlineEquation" id="IEq93">\(u\)</span> from <span class="InlineEquation" id="IEq94">\(G\)</span>. The reverse corresponds to adding a vertex in <span class="InlineEquation" id="IEq95">\(H\)</span>. Given the above, we can define the three partial costs of the GED, which are VertexCost, EdgeCost and RelabelCost. <div class="DefinitionList"><dl><dt class="Term"> <em class="EmphasisTypeItalic ">VertexCost</em> </dt><dd class="Description"> <p class="Para">The number of deleted/inserted vertices: <span class="InlineEquation" id="IEq96">\(| \{ u : u \in [V_{G'}\cup V_{H'}] \wedge [ \phi (u)\in \epsilon \vee \phi ^{-1}(u)\in \epsilon ]\}|.\)</span> </p> </dd><dt class="Term"> <em class="EmphasisTypeItalic ">EdgeCost</em> </dt><dd class="Description"> <p class="Para">The number of deleted/inserted edges: <span class="InlineEquation" id="IEq97">\( |E_{G'}| + |E_{H'}| - 2 \times |\{ (i,j) : [(i,j)\in E_{G'} \wedge (\phi (i),\phi (j))\in E_{H'}]\}|.\)</span> </p> </dd><dt class="Term"> <em class="EmphasisTypeItalic ">RelabelCost</em> </dt><dd class="Description"> <p class="Para">The
 number of mismatched functions. A function is mismatched if it is 
either a local function and matched to any external function, or the 
reverse, or an external function matched to an external function with a 
different name.</p> </dd></dl></div> </div><div class="Para">Finally, the edit cost <span class="InlineEquation" id="IEq98">\(\lambda _{\phi }(G,H)\)</span>, incurred by a matching <span class="InlineEquation" id="IEq99">\(\phi \)</span>, is the sum of these value:<div id="Equ1" class="Equation EquationMathjax"><div class="EquationContent">$$\begin{aligned} \lambda _{\phi }(G,H) = VertexCost + EdgeCost + RelabelCost \end{aligned}$$</div> <div class="EquationNumber">(1)</div></div>Hence, the GED problem breaks down to computing the matching that yields the minimum cost, i.e. <span class="InlineEquation" id="IEq100">\(\min _{\phi }\lambda _{\phi }\)</span>.
 The edit cost incurred by matching two graphs is an arbitrary 
non-negative integer value. There is a need to acquire a normalized 
value. For this, we use the <em class="EmphasisTypeItalic ">graph dissimilarity</em>.</div><div class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading"> <strong class="EmphasisTypeBold ">Definition 3</strong> </span></h3><p class="Para">(<em class="EmphasisTypeItalic ">Graph dissimilarity</em>) The dissimilarity <span class="InlineEquation" id="IEq101">\(\delta \)</span> <span class="InlineEquation" id="IEq102">\((G,H)\)</span> between two graphs, <span class="InlineEquation" id="IEq103">\(G\)</span> and <span class="InlineEquation" id="IEq104">\(H\)</span>
 is a real value in [0,1], where 0 indicates that the graphs are 
identical while a value near 1 indicates that the graphs are highly 
dissimilar. Furthermore, it holds that: <span class="InlineEquation" id="IEq105">\(\delta (G,H)=\delta (H,G)\)</span>, <span class="InlineEquation" id="IEq106">\(\delta (G,G)=0\)</span> and <span class="InlineEquation" id="IEq107">\(\delta (G,K_0)=1\)</span>, where <span class="InlineEquation" id="IEq108">\(K_0\)</span> is the order-zero graph.</p></div><div class="Para">Finally, the graph dissimilarity <span class="InlineEquation" id="IEq109">\(\delta (G,H)\)</span> can be obtained by the graph edit distance <span class="InlineEquation" id="IEq110">\(\lambda _{\phi }(G,H)\)</span>:<div id="Equ2" class="Equation EquationMathjax"><div class="EquationContent">$$\begin{aligned} \delta (G,H) = \frac{\lambda _{\phi }(G,H)}{|V_G|+|V_H|+|E_G|+|E_H|} \end{aligned}$$</div> <div class="EquationNumber">(2)</div></div>We will use the term GED to denote both the GED and the normalized GED interchangeably depending on the context.</div></section></div></section><section id="Sec7" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">4 </span>Classy: architectural overview<span class="section-icon"></span></h2><div class="content"><p class="Para">In this section we provide an overview of the architecture of our system and the sample processing pipeline.</p><section id="Sec8" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">4.1 </span>Architectural model</h3><p class="Para">Our
 system is distributed and highly scalable. External dependencies are 
limited to a filesystem for storing and retrieving the graphs and a 
relational database (RDB) for maintaining information about the samples 
and the clusters. Any highly-available distributed filesystem (DFS) is 
sufficient since the use pattern is write-once-read-many; the graphs are
 never modified.</p><div class="Para">The architectural model relies on 
the presence of two types of processes: the workers and the masters. 
Communication between different types of processes happens through 
queues. There are two queues: the masters queue, from which the masters
 fetch messages put by the workers and the workers queue which is used 
in the opposite way. The workers are responsible for handling the most 
computationally intensive tasks (graph extraction, graph comparisons, 
etc). The masters are responsible for retrieving the results of the 
tasks completed by the workers and, based on the result, crafting a new 
worker task. They are the ones accessing and modifying the RDB. All 
actors are stateless so there can be multiple instances of them in 
parallel. Essentially, the masters push the information between the 
different stages of the sample-processing pipeline while the workers are
 the ones performing each stage. Finally, tasks are inserted to the 
system via a dispatcher who places the relevant messages in the masters
 queue. The model can be seen in Fig.&nbsp;<span class="InternalRef"><a href="#Fig2">2</a></span>.<figure class="Figure" id="Fig2"><div class="MediaObject" id="MO18"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Fig2_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Fig2_HTML.gif" alt="Fig. 2"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 2</span> <p class="SimplePara">High-level architectural view of Classy. The workers (<em class="EmphasisTypeItalic ">W</em>) perform the computationally intensive tasks, while the masters (<em class="EmphasisTypeItalic ">M</em>) craft the tasks. Communication happens using only two queues. The dispatcher (<em class="EmphasisTypeItalic ">D</em>) introduces tasks into the system. External dependencies are the graph-storage file-system (<em class="EmphasisTypeItalic ">GS</em>) and a relational database (<em class="EmphasisTypeItalic ">RDB</em>) </p> </div></figcaption></figure> </div><p class="Para">The
 architectural model can be seen as a double producer-consumer pattern; 
the masters produce tasks that are consumed by the workers, that in turn
 place the results in another queue to be processed by the masters. 
Consequently, it is possible to add more resources of each type on 
demand. The obvious requirement is that the DFS should scale with the 
amount of workers, and the RDB with the amount of masters. In 
Sect.&nbsp;<span class="InternalRef"><a href="#Sec20">8.1.4</a></span> 
we provide more details and insights regarding our running instance of 
Classy, as well as a discussion about specific technologies that may be 
used and their limitations.</p></section><section id="Sec9" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">4.2 </span>Process pipeline</h3><div class="Para">Once our call-graph clustering system is presented with an incoming sample, the sample undergoes 5 phases as shown in Fig.&nbsp;<span class="InternalRef"><a href="#Fig3">3</a></span>.<figure class="Figure" id="Fig3"><div class="MediaObject" id="MO19"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Fig3_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Fig3_HTML.gif" alt="Fig. 3"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 3</span> <p class="SimplePara">The high-level sample flow pipeline. Each stage corresponds to a separate worker task </p> </div></figcaption></figure> </div><p class="Para">First
 is the call-graph extraction phase. This includes unpacking the sample.
 Many binary files, even benign ones, are packed and/or protected 
against debugging, dumping and disassembling. Removing the protective 
layer(s) before running the call-graph extraction tools is imperative, 
otherwise the call-graph would correspond to the protective layer and 
not the actual application. Additionally, at this phase samples can be 
discarded by the system if the unpacking is unsuccessful or if the 
unpacked binary turns out to be a file which should not be clustered 
(for example a RAR self-extractor). The second phase is the 
fast-filtering (or pruning) of the candidate-clusters list for that 
sample using the lower bounding method described in Sect.&nbsp;<span class="InternalRef"><a href="#Sec11">6</a></span>.
 This is the most I/O intensive part of the whole process. The third 
phase consists of comparing the incoming graph to those in the pruned 
candidate-cluster list. This phase is the most CPU-intensive. The fourth
 phase is the clustering phase. Given a list of graph distances the 
worker determines to which cluster(s) it should assign the incoming 
graph and if any more complex cluster operations, such as cluster 
merging, should be performed. Finally, if the incoming graph has not 
been identified as being isomorphic to any of the existing graphs, it 
gets compared to those graphs that belong to the same cluster and have 
the same order and size.</p></section></div></section><section id="Sec10" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">5 </span>Graph comparison<span class="section-icon"></span></h2><div class="content"><p class="Para">The
 graph comparison is the most essential part of the whole system. Its 
accuracy and time requirements are in direct relation to the performance
 of the system from the scope of quality and throughput respectively.</p><div class="Para">Our work is aligned to that of (Hu et al. <span class="CitationRef"><a href="#CR31" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>) and (Kinable and Kostakis <span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>)
 in using the same graph edit distance (GED) as the distance measure 
between pairs of graphs. In preliminary work (Kostakis et al. <span class="CitationRef"><a href="#CR39" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>)
 we have demonstrated that a Simulated Annealing (SA) comparison 
algorithm is faster and yields better results than the bipartite 
matching-based approaches for approximating GED; an example of the 
superior performance is depicted in Fig.&nbsp;<span class="InternalRef"><a href="#Fig4">4</a></span>,
 taken from the same work. While GED is a computationally hard problem, 
SA runs in polynomial time with respect to the input size.<figure class="Figure" id="Fig4"><div class="MediaObject" id="MO20"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Fig4_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Fig4_HTML.gif" alt="Fig. 4"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 4</span> <p class="SimplePara">Performance
 of simulated annealing versus bipartite matching when computing the 
graph edit distance for 1000 randomly chosen pairs of call-graphs; less 
is better. Taken from (Kostakis et al. <span class="CitationRef"><a href="#CR39" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>) </p> </div></figcaption></figure> </div><div class="Para">Simulated
 annealing is a local search method. By starting from a random solution,
 it traverses the solution space in the attempt to find the global 
minimum (if it is used for minimization problems such as GED, global 
maximum otherwise). At every step of the process, a random neighboring 
solution is selected. Neighboring solutions are produced by choosing two
 vertices in <span class="InlineEquation" id="IEq111">\(G'\)</span> and swapping the vertices in <span class="InlineEquation" id="IEq112">\(H'\)</span> to which they are mapped by <span class="InlineEquation" id="IEq113">\(\phi _i\)</span>.
 If the new solution yields a better (lower) score for the objective 
function, that solution is always accepted and at the next step the 
process resumes with the new solution as its current. If the new 
solution does not yield a better score, it might be accepted with 
certain probability; this probability is controlled by a parameter 
called <em class="EmphasisTypeItalic ">temperature</em>. For computing 
the score of the neighboring solution it is sufficient to simply compute
 the change to the score induced locally by swapping the matching 
counterparts. This alleviates the need to re-compute the total score at 
every step. The rationale for accepting worse solutions is to avoid 
getting trapped in local minima. Since the solutions returned by SA are 
of greater score than the ground truth, SA is an over-approximating 
algorithm. Algorithm&nbsp;1 depicts the SA process in pseudocode.<figure class="Figure" id="Figa"><div class="MediaObject" id="MO7"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Figa_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Figa_HTML.gif" alt=""></a> </div></figure> </div><p class="Para">The value returned is the lowest score witnessed for <span class="InlineEquation" id="IEq133">\(\lambda _{\phi _{i}}\)</span>. The time complexity of SA for comparing graphs <span class="InlineEquation" id="IEq134">\(G,H\)</span> is equal to <span class="InlineEquation" id="IEq135">\(O(k\cdot |V_{\max }|^2\cdot d_{\max })\)</span>, where <span class="InlineEquation" id="IEq136">\(|V_{\max }|=\max \{|V_G|,|V_H|\}\)</span>, <span class="InlineEquation" id="IEq137">\(k\)</span> is the number of relaxation iterations and <span class="InlineEquation" id="IEq138">\(d_{\max }\)</span> is the maximum degree of the nodes in <span class="InlineEquation" id="IEq139">\(V_G\cup V_H\)</span>. The value <span class="InlineEquation" id="IEq140">\(k\)</span>
 can either be a fixed input or SA can be allowed to progress until its 
terminating conditions are met. Experiments have shown that when SA is 
allowed to reach either of its terminating conditions, the value of <span class="InlineEquation" id="IEq141">\(k\)</span> (the number of iterations) is close to <span class="InlineEquation" id="IEq142">\(\log n\)</span>.
 There are two terminating conditions for SA. The first is achieving the
 latest possible score. But since this is the problem SA is called to 
solve, a lower bound is computed and used instead. The second 
terminating condition comprises of terminating the SA process when no 
better solution has been identified within a certain number of the most 
recent neighboring solutions; this is the no_progress() function in 
Algorithm&nbsp;1. For more details about the SA comparison method we 
refer readers to (Kostakis et al. <span class="CitationRef"><a href="#CR39" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>).</p><p class="Para">One
 of the advantages of SA is that it attempts to solve the problem 
directly instead of applying some sub-optimal heuristic like this is 
done by the bipartite-matching approaches. In addition, since it is a 
local search method, it has the advantages of an <em class="EmphasisTypeItalic ">anytime algorithm</em> (Dean and Boddy <span class="CitationRef"><a href="#CR17" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1988</a></span>).
 At any given point during the execution of a problem instance, a valid 
solution is maintained. So, in the case that the execution is terminated
 even before the halting conditions have been satisfied, a result can 
still be obtained. If given more time, SA continues to improve the 
solution. This provides the flexibility to impose timeout thresholds so 
to adhere to various Service Level Agreements (SLAs).</p></div></section><section id="Sec11" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">6 </span>Lower bound of the graph edit distance<span class="section-icon"></span></h2><div class="content"><div class="Para">The
 number of candidate clusters that a sample has to be compared against 
might be very high. While the SA algorithm might be fast enough for the 
problem that it solves, it is necessary to keep the number of 
comparisons to a minimum. Essentially, we only need to know which 
comparisons might provide a score below the threshold for which the 
clustering algorithm would consider the samples to be similar. Since SA 
is an over-approximating algorithm, a lower bound on the value of GED 
for a pair of graphs is a lower value on the score returned by SA. We 
have devised a <span class="InlineEquation" id="IEq143">\(O(n)\)</span> lower bound for GED, which we denote as <span class="InlineEquation" id="IEq144">\(GED_{LB}\)</span>. The lower bound presented in (Zeng et al. <span class="CitationRef"><a href="#CR68" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>) is not suitable since the <span class="InlineEquation" id="IEq145">\(\Theta (n^3)\)</span> time complexity is comparable to the complexity of approximating GED via bipartite matching or SA. <span class="InlineEquation" id="IEq146">\(GED_{LB}\)</span> is based on lower bounding individually each of the coefficients of Eq.&nbsp;<span class="InternalRef"><a href="#Equ2">2</a></span>:<div class="OrderedList"><ol><li class="ListItem"><span class="ItemNumber">(i)</span><div class="ItemContent"> <p class="Para">For the cost of adding or removing vertices, we know that <span class="InlineEquation" id="IEq147">\(VertexCost\)</span> <span class="InlineEquation" id="IEq148">\(\ge \)</span> <span class="InlineEquation" id="IEq149">\(||V_G|-|V_H||\)</span>.
 By editing one graph to acquire the other, one would need to perform at
 least as many vertex deletion (or addition) operations as the 
difference of the two graph orders.</p> </div><div class="ClearBoth">&nbsp;</div></li><li class="ListItem"><span class="ItemNumber">(ii)</span><div class="ItemContent"> <p class="Para">For the cost induced by relabeling nodes, we compute the intersection of the sets of system calls in each graph. We know that <span class="InlineEquation" id="IEq150">\(\max \{|S_G|,|S_H|\} - |S_G\cap S_H|\)</span>
 vertices do not have an equivalent counterpart in the other graph. 
However, these vertices might be eventually deleted, thus contributing 
to <span class="InlineEquation" id="IEq151">\(VertexCost\)</span>. So, to avoid any double counting, <span class="InlineEquation" id="IEq152">\(RelabelCost \ge \max \{ |S_G|,|S_H|\} - | S_G \cap S_H | - V_c\)</span>. Of course, this value might be negative, so we take <span class="InlineEquation" id="IEq153">\(\max \{0, \max \{|S_G|,|S_H|\} - |S_G \cap S_H| - V_c\}\)</span>.</p> </div><div class="ClearBoth">&nbsp;</div></li><li class="ListItem"><span class="ItemNumber">(iii)</span><div class="ItemContent"> <p class="Para">For the cost of adding or removing edges, a trivial bound would be <span class="InlineEquation" id="IEq154">\(EdgeCost\)</span> <span class="InlineEquation" id="IEq155">\(\ge \)</span> <span class="InlineEquation" id="IEq156">\(||E_G|-|E_H||\)</span>.
 However, a tighter bound can be achieved by examining the ordered 
outdegree or indegree sequences. The idea is that identical degree 
sequences are a requirement, but not sufficient, for two graphs to be 
isomorphic. Thus, the cost required to transform one sequence into the 
other is a lower bound for <span class="InlineEquation" id="IEq157">\(EdgeCost\)</span>. With the distance between vertex degrees being <span class="InlineEquation" id="IEq158">\(|d^{+}(u_i)-d^{+}(v_j)|\)</span>, <span class="InlineEquation" id="IEq159">\({u_i}\in V_G'\)</span>, <span class="InlineEquation" id="IEq160">\({v_j}\in V_H'\)</span> (if we consider the outdegrees), the optimal solution can be retrieved using maximum weight bipartite matching which takes <span class="InlineEquation" id="IEq161">\(O(n^3)\)</span> time. Surprisingly, however, it can be done in linear time by computing <span class="InlineEquation" id="IEq162">\(\sum ^{|V_G'|}_{i} |d^{+}(u^{G'+}_{i}) - d^{+}(u^{H'+}_{i})|\)</span>, given <span class="InlineEquation" id="IEq163">\(D^{+}_{G'}\)</span> and <span class="InlineEquation" id="IEq164">\(D^{+}_{H'}\)</span>. The correctness of the latter is proved below.</p> </div><div class="ClearBoth">&nbsp;</div></li></ol></div> </div><div class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading"> <strong class="EmphasisTypeBold ">Lemma 1</strong> </span></h3><p class="Para">Given two sorted degree sequences, the cost of transforming one into the other can be computed in linear time, by <span class="InlineEquation" id="IEq165">\(\sum ^{|V_G'|}_{i} |d^{+}(u^{G'+}_{i}) - d^{+}(u^{H'+}_{i})|\)</span>.</p></div><div class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading"> <em class="EmphasisTypeItalic ">Proof</em> </span></h3><p class="Para">It holds that <span class="InlineEquation" id="IEq166">\(|a_1-b_1| + |a_2 - b_2 | \le |a_1 - b_2| + |a_2 - b_1 |\)</span> with <span class="InlineEquation" id="IEq167">\(0\le a_1\le a_2\)</span>, <span class="InlineEquation" id="IEq168">\(0 \le b_1 \le b_2\)</span> (trivial, the proof is omitted). So, if the optimal matching <span class="InlineEquation" id="IEq169">\(h:D^{+}_{G'} \rightarrow D^{+}_{H'}\)</span> would have matched <span class="InlineEquation" id="IEq170">\(u^{G'+}_{i}\)</span> to <span class="InlineEquation" id="IEq171">\(u^{H'+}_{k}\)</span> and <span class="InlineEquation" id="IEq172">\(u^{G'+}_{j}\)</span> to <span class="InlineEquation" id="IEq173">\(u^{H'+}_{l}\)</span>, with <span class="InlineEquation" id="IEq174">\(i&lt;j\)</span> and <span class="InlineEquation" id="IEq175">\(k&gt;l\)</span>,
 it would be possible to swap their matching counterparts and acquire a 
less or equal total score, because of the aforementioned inequality. 
However, since the matching is the optimal, the score should remain the 
same. This swapping operation can be repeated until <span class="InlineEquation" id="IEq176">\(\forall i\)</span>, <span class="InlineEquation" id="IEq177">\(u^{G'+}_{i}\)</span> is matched to <span class="InlineEquation" id="IEq178">\(u^{H'+}_{i}\)</span>.
 It remains to be proved that such a matching is indeed reachable by 
swaps from any possible optimal matching. A constructive proof exists: 
Given the optimal matching <span class="InlineEquation" id="IEq179">\(h\)</span>, for each <span class="InlineEquation" id="IEq180">\(i\)</span> in <span class="InlineEquation" id="IEq181">\(1,\ldots , |D^{+}_{G'}|\)</span>, we iteratively swap the matching counterparts of <span class="InlineEquation" id="IEq182">\(u^{G'+}_{i}\)</span> and <span class="InlineEquation" id="IEq183">\(h^{-1}(|u^{H'+}_{i}|)\)</span>. At step <span class="InlineEquation" id="IEq184">\(i\)</span>, for each <span class="InlineEquation" id="IEq185">\(1\le j\le i\)</span>, the <span class="InlineEquation" id="IEq186">\(j\)</span>-th element of the ordered degree sequence, <span class="InlineEquation" id="IEq187">\({ u^{G'+}_{j}}\)</span> is matched to <span class="InlineEquation" id="IEq188">\(u^{H'+}_{j}\)</span>. <span class="InlineEquation" id="IEq189">\(\square \)</span> </p></div><p class="Para">The
 above hold for the indegree sequences, too. In practice, we have the 
capability to compute both, if necessary, and retain the one providing 
the highest score.</p><div class="Para">We combine all of the above inequalities to acquire a lower bound for Eq.&nbsp;<span class="InternalRef"><a href="#Equ2">2</a></span>, denote as <span class="InlineEquation" id="IEq190">\(GED_{LB}\)</span>, in the formula depicted in Eq.&nbsp;<span class="InternalRef"><a href="#Equ3">3</a></span>. The total time complexity for computing the <span class="InlineEquation" id="IEq191">\(GED_{LB}\)</span> of graphs <span class="InlineEquation" id="IEq192">\(G,H\)</span> is <span class="InlineEquation" id="IEq193">\(O(n)\)</span>, with <span class="InlineEquation" id="IEq194">\(n=|V_G'|\)</span>. This is achieved by using Counting sort to sort them in linear time (Seward <span class="CitationRef"><a href="#CR54" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1954</a></span>). Sorting <span class="InlineEquation" id="IEq195">\(n\)</span> non-negative integers whose maximum value is at most <span class="InlineEquation" id="IEq196">\(k\)</span>, via Counting sort takes <span class="InlineEquation" id="IEq197">\(O(n+k)\)</span> time. Since <span class="InlineEquation" id="IEq198">\(k\)</span> is a degree value, and <span class="InlineEquation" id="IEq199">\(k\le n\)</span>, given the degree sequences the whole process takes linear time.<div id="Equ3" class="Equation"><div class="EquationContent"><div class="MediaObject" id="MO8"> <a href="https://static-content.springer.com/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Equ3_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Equ3_HTML.gif" alt=""></a> </div></div> <div class="EquationNumber">(3)</div></div> </div></div></section><section id="Sec12" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">7 </span>Clustering<span class="section-icon"></span></h2><div class="content"><p class="Para">In
 this Section we introduce the clustering algorithm that is in use by 
Classy. The proposed clustering algorithm is introduced in Sect.&nbsp;<span class="InternalRef"><a href="#Sec13">7.1</a></span>, while in Sect.&nbsp;<span class="InternalRef"><a href="#Sec14">7.2</a></span>
 we describe how we can approximate the statistics of a cluster when 
there are available only a part of all possible pair-wised distances.</p><section id="Sec13" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">7.1 </span>Clustering algorithm</h3><p class="Para">Graph
 comparisons are expensive. We cannot afford to compute full distances 
matrices, nor compare an incoming sample with all the existing samples 
in the DB. Thus, it is imperative to keep the number of graph 
comparisons to a minimum while still acquiring the important 
information. Due to the cost of graph comparisons we cannot use 
NN-retrieval methods that would require comparing the query graph to a 
set of database reference samples (Burkhard and Keller <span class="CitationRef"><a href="#CR9" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1973</a></span>; Venkateswaran et al. <span class="CitationRef"><a href="#CR60" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>; Papapetrou et al. <span class="CitationRef"><a href="#CR47" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>).
 At this point we must make clear that improperly grouping two 
executables together is highly undesirable since it could result in a 
significant false alarm that would be propagated to the customers. So, 
while we aim for clustering together as many graphs as possible, we can 
tolerate improperly separated executables hoping that they will be 
detected by other systems.</p><p class="Para">As a first step, a 
procedure is needed to determine the possible clusters to which an 
incoming sample would be added. Graph invariants prove very useful for 
this problem. We can expect that pairs of graphs with a very large 
difference in the number of edges and vertices would give a high edit 
distance. Performing such a comparison would just verify that they 
should not belong to the same cluster. Conversely, it is graphs with 
similar numbers of edges and graphs that are useful in this context. So,
 for each incoming sample <span class="InlineEquation" id="IEq200">\(q\)</span>, we query the database for clusters that contain graphs with similar graph order. The results of such query are the <em class="EmphasisTypeItalic ">candidate clusters</em>, <span class="InlineEquation" id="IEq201">\(\mathcal {C}^q\)</span>, for that sample.</p><p class="Para">Given
 those clusters, the second step involves deciding which of the 
candidate clusters is the appropriate. Again, comparing the incoming 
sample to every sample of each cluster is not affordable, nor could the 
result justify such a cost. In most of the cases, comparing to a few 
samples provides all the necessary information. Based on that, we 
identify <em class="EmphasisTypeItalic ">reference samples</em>, <span class="InlineEquation" id="IEq202">\(r^c_{i}\)</span>, for each cluster <span class="InlineEquation" id="IEq203">\(C\)</span>
 in the database. These are the samples of the candidate cluster that 
would be compared to the incoming sample. There can be several 
strategies for selecting the reference samples: the first graphs seen by
 the system, the ones that are in the geometric center of the cluster 
(or their sub-clusters), the outliers of that cluster, the major malware
 variants for that family, etc.</p><p class="Para">In practice, we have 
found that selecting the first sample of the cluster to be its reference
 graph works rather well. The main reason is that given the first 
version of a piece of software, either benign or malicious, subsequent 
versions share with it a great amount of code and structure. So, we can 
expect the first graph of a cluster to be a central element. 
Furthermore, over the lifetime of the system, we can assume that the 
first version of a software would be the one to arrive first into our 
system. This is a valid assumption given the great number of sample 
exchange-feeds shared between AV companies and the file upstream 
functionality of modern AV clients; consequently the time it takes for a
 file that first appears in the wild to reach the AV companys backend 
is very short.</p><p class="Para">Our approach can be seen as belonging 
to the family of clustering algorithms where at first, one selects an 
arbitrary sample, takes all near-by elements to form a cluster and 
removes that cluster. Several such algorithms have been presented in the
 past and some of those have approximation guarantees under different 
objective functions. One example is <span class="EmphasisFontCategoryNonProportional ">BALLS</span> studied in (Gionis et al. <span class="CitationRef"><a href="#CR25" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2005</a></span>).
 However, we leave it as future work to investigate other methods for 
selecting the candidate samples and study the trade-offs in each case.</p><p class="Para">Based on the above, upon the arrival of a new sample, its candidate clusters are defined and then it is compared against <span class="InlineEquation" id="IEq204">\(\mathcal {S}\)</span>,
 the set denoting the union of their reference samples. This stage is 
divided into two parts. The first comprises of using the aforementioned 
lower bound <span class="InlineEquation" id="IEq205">\(GED_{LB}\)</span> on <span class="InlineEquation" id="IEq206">\(S\)</span> to quickly prune many of the comparisons, which yields <span class="InlineEquation" id="IEq207">\(\mathcal {P}^q\)</span>. SA is then used for comparing <span class="InlineEquation" id="IEq208">\(q\)</span> to the graphs in <span class="InlineEquation" id="IEq209">\(\mathcal {P}^q\)</span> in order to identify the suitable candidate clusters for <span class="InlineEquation" id="IEq210">\(q\)</span>. Suitable clusters are determined based on whether the incoming sample <span class="InlineEquation" id="IEq211">\(q\)</span> and any of their reference samples are within <span class="InlineEquation" id="IEq212">\(threshold\)</span> distance. This is a common notion for density-based clustering algorithms (Ester et al. <span class="CitationRef"><a href="#CR20" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1996</a></span>).
 So instead of defining the number of clusters, which is often very hard
 to know when clustering streams, it is sufficient to define the 
distance threshold for two data elements to be considered similar. 
Clearly, the choice for the value of the threshold is often application 
dependent and may be determined empirically.</p><div class="Para">Several
 actions are taken depending on the outcome of determining suitable 
candidate clusters. If no suitable candidate cluster is found, the 
incoming sample forms a new one of its own. If only one suitable cluster
 is found, the incoming sample is assigned to that (the cluster is 
augmented with the new sample). If several suitable clusters have been 
identified, then those are merged into a single one and the set of 
reference samples is recalculated based on the used strategy. The 
rationale of merging those clusters together is the fact that the 
appearance of sample <span class="InlineEquation" id="IEq213">\(q\)</span>
 has indicated the existence of an edit path involving their samples. 
This piece of information will be used by the analysts and other systems
 aimed at providing added-value. There is no concern that the samples 
will all be added to just a few clusters; the authors of (Kinable and 
Kostakis <span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>)
 demonstrated that for reasonable parameter values, density based 
algorithms perform very well in terms of not lumping unrelated samples. 
Variations of the clustering algorithm itself include breaking on the 
comparisons as soon as a suitable candidate has been found. The way the 
lower bound is combined with the aforementioned clustering approach in 
Classy is described in Algorithm&nbsp;2. The exact parameter values used
 for Classy are discussed in Sect.&nbsp;<span class="InternalRef"><a href="#Sec16">8.1</a></span>.<figure class="Figure" id="Figb"><div class="MediaObject" id="MO9"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Figb_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Figb_HTML.gif" alt=""></a> </div></figure> </div><p class="Para">To facilitate the work of our human analysts, a final step involves finding isomorphisms between the incoming sample <span class="InlineEquation" id="IEq249">\(q\)</span>
 and the samples in the cluster to which it was assigned. For that, it 
is enough to identify which ones have the same number of edges and 
vertices and determine how they are grouped in sets of isomorphic based 
on the comparisons of their own classifications. Then, it is simply an 
issue of comparing <span class="InlineEquation" id="IEq250">\(q\)</span> to only one of the graphs of each existing set of isomorphic graphs.</p></section><section id="Sec14" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">7.2 </span>Numerical approximations of cluster statistics</h3><p class="Para">Both
 the automation systems and malware analysts require more information 
than simply the fact that several files are clustered together. In order
 to evaluate and make use of the clustering results, they require 
numerical information regarding each cluster. However, due to the 
aforementioned clustering algorithm an incoming sample is only compared 
against the clusters reference graphs and those that might be 
isomorphic to it. As a result, for most of the possible pairs of graphs 
in a cluster the system has not executed any comparison. Computing the 
distance between all pairs of samples in a cluster can be performed on 
special occasions but it is not feasible for every cluster. Thus, we 
resort to approximating the distances.</p><p class="Para">Our approach 
relies on using the computed distances to infer a score for those 
distances that should be approximated. We assume that each call-graph is
 a point in a metric space, and the distances between pairs of 
call-graphs are either a known distance if that has been computed or 1.0
 otherwise. An alternative view of this is a complete graph where the 
call-graphs are its nodes and their distances are its weighted edges. 
Then it is a matter of using an all-pairs shortest path algorithm to 
approximate the unknown distances. Such an algorithm is the 
FloydWarshall algorithm (<span class="CitationRef"><a href="#CR22" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1962</a></span>, <span class="CitationRef"><a href="#CR62" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1962</a></span>).</p><p class="Para">The time complexity of the FloydWarshall algorithm is <span class="InlineEquation" id="IEq251">\(\Theta (|V|^3)\)</span>.
 For large clusters with several thousand samples, such complexity 
renders the approach nonpractical. For that reason, we devised a 
speed-up pre-filtering step. The samples are bagged into sets of 
isomorphic graphs. Each set contains those samples for which there is a 
chain of isomorphic relations. For example, given that we know that 
graph <span class="InlineEquation" id="IEq252">\(A\)</span> is isomorphic to graph <span class="InlineEquation" id="IEq253">\(B\)</span>, <span class="InlineEquation" id="IEq254">\(B\)</span> to <span class="InlineEquation" id="IEq255">\(C\)</span> and <span class="InlineEquation" id="IEq256">\(C\)</span> to <span class="InlineEquation" id="IEq257">\(D\)</span>, then <span class="InlineEquation" id="IEq258">\(A\)</span>, <span class="InlineEquation" id="IEq259">\(B\)</span>, <span class="InlineEquation" id="IEq260">\(C\)</span>, and <span class="InlineEquation" id="IEq261">\(D\)</span> would be assigned to the same set. Given <span class="InlineEquation" id="IEq262">\(n\)</span> samples, <span class="InlineEquation" id="IEq263">\(m\)</span> known pair isomorphisms (distances between samples with value equal to 0) and <span class="InlineEquation" id="IEq264">\(k\)</span> the number of returned sets, the total complexity of approximating the distances becomes <span class="InlineEquation" id="IEq265">\(O(n + \log ^{*}m + k^3)\)</span>, when using union-find algorithms (Tarjan and Leeuwen <span class="CitationRef"><a href="#CR57" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1984</a></span>) to compute the sets.</p><div class="Para">Finally,
 the speed-up setting allows to efficiently compute the cluster 
statistics immediately instead of computing them over all <span class="InlineEquation" id="IEq266">\(n^2\)</span> distances. The cluster diameter is the maximum value of the <span class="InlineEquation" id="IEq267">\(k^2\)</span> distances, while the mean distance for that cluster can be computed by:<div id="Equ4" class="Equation EquationMathjax"><div class="EquationContent">$$\begin{aligned}
 \frac{\sum \nolimits _{i=1}^{k}\sum \nolimits _{j=i+1}^{k}{ |k_i|\cdot 
|k_j|\cdot d(k_i,k_j)} }{n\cdot (n-1)/2}, \end{aligned}$$</div></div>where <span class="InlineEquation" id="IEq268">\(k_i\)</span>,<span class="InlineEquation" id="IEq269">\(k_j\)</span> are the <span class="InlineEquation" id="IEq270">\(i\)</span>-th and <span class="InlineEquation" id="IEq271">\(j\)</span>-th sets respectively and <span class="InlineEquation" id="IEq272">\(d(k_i\)</span>, <span class="InlineEquation" id="IEq273">\(k_j)\)</span> is the distance between them.</div></section></div></section><section id="Sec15" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">8 </span>Experiments<span class="section-icon"></span></h2><div class="content"><p class="Para">In Sect.&nbsp;<span class="InternalRef"><a href="#Sec16">8.1</a></span> we evaluate the performance of our system and provide relevant statistics, while in Sect.&nbsp; <span class="InternalRef"><a href="#Sec22">8.2</a></span>
 we benchmark our proposed on-line clustering algorithm against the 
competitor which in our case is the offline optimal, that uses DBSCAN, 
described in Kinable and Kostakis (<span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>).</p><section id="Sec16" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">8.1 </span>Live system performance and statistics</h3><p class="Para">Results
 have been extracted from a live production-side system that has 
classified more than 0.8 million graphs. These were part of the raw 
stream of PE32 files that have arrived in the backend of F-Secure, and 
were not discarded by the simple filtering rules described in 
Sect.&nbsp;<span class="InternalRef"><a href="#Sec9">4.2</a></span>. We 
choose to limit the valid graphs to those that contain at least 20 
vertices, but not more than 3000; no limit is imposed on the edges. 
Malware components tend to be small in size and compact, for a series of
 reasons. Flame/Flamer/Skywipper was a notable exception. We make the 
trade-off of focusing on graphs of such order range. We decided on the 
upper limit of 3000 after looking into tens of thousands of samples. For
 software larger in size, other approaches, such as md5 checksum 
verification, certificates and software signing are more appropriate for
 making sure they are not trojanized. The value of 3000 is a 
configuration value that can be changed at any time. Furthermore, as the
 order of call-graphs increases the witnessed call-graph space should 
become sparser, so we can anticipate that the task of grouping becomes
 easier since <span class="InlineEquation" id="IEq274">\(GED_{LB}\)</span> would be more effective.</p><p class="Para">The
 following results and statistics correspond to an instance of Classy 
with at most 48 workers. The majority of the system has been implemented
 using Python, apart from the SA method that was implemented in C. The 
relational database is PostgreSQL and the distributed file-system 
comprises of GlusterFS. More information such as parameter values may be
 found in the following sections.</p><section id="Sec17" tabindex="-1" class="Section3 RenderAsSection3"><h4 class="Heading"><span class="HeadingNumber">8.1.1 </span>Preprocessing and call-graph extraction.</h4><div class="Para">First,
 we provide numbers related to the preprocessing phase. There is little 
room for improving the performance of the binary unpacking and graph 
extraction phase. However, from the system perspective it is an 
unavoidable cost. To gain an intuition about the required resources we 
investigate the time required for this phase. We chose to study the 
unpacking and extraction steps separately, since the extraction time 
depends to a great extent on the size of the graph and the size of the 
binary file, contrary to the unpacking which depends primarily on the 
technique used to pack the file. Figure <span class="InternalRef"><a href="#Fig5">5</a></span>
 depicts the normalized histogram of the call-graph extraction phase. 
The original data values correspond to successfully extracted graphs and
 the chosen set of graphs is the 10454 graphs clustered in the last 24 h
 interval of the system. For the unpacking phase, we witness that the 
mean value is 0.6&nbsp; s (graph omitted). The mean extraction time is 9
 s while we notice in some cases values over a minute.<figure class="Figure" id="Fig5"><div class="MediaObject" id="MO21"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Fig5_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Fig5_HTML.gif" alt="Fig. 5"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 5</span> <p class="SimplePara">Normalized histogram of the time required for extracting the call-graphs. The mean extraction time is 9&nbsp;s </p> </div></figcaption></figure> </div><p class="Para">The
 percentage of the samples that are successfully preprocessed depend to a
 great extent on the distribution of each file type and the quality of 
the pre-filtering. In particular, cases such as Zip and RAR 
self-extracting executables should not be given as input to Classy, and 
their contents should be extracted separately. The same applies for 
known installers such as <span class="EmphasisFontCategoryNonProportional ">InstallShield</span> and <span class="EmphasisFontCategoryNonProportional ">Inno Setup</span>.
 Without counting such files, an average ratio of 7075&nbsp;% files are
 successfuly preprocessed. This includes potential unpacking and 
producing the call-graph.</p></section><section id="Sec18" tabindex="-1" class="Section3 RenderAsSection3"><h4 class="Heading"><span class="HeadingNumber">8.1.2 </span>Lower bounding the GED</h4><p class="Para">The performance of the lower bounding method is crucial to the overall throughput of the system. This is due to the fact that <span class="InlineEquation" id="IEq275">\(GED_{LB}\)</span> is responsible for reducing the number of graph comparisons the system will need to perform.
</p><p class="Para">For that reason, we conduct a deeper investigation into its performance. We studied its <em class="EmphasisTypeItalic ">tightness</em> and its <em class="EmphasisTypeItalic ">pruning power</em> for 1-NN searches in offline experiments. The tightness of <span class="InlineEquation" id="IEq276">\(GED_{LB}\)</span> is defined as the average ratio of <span class="InlineEquation" id="IEq277">\(\frac{GED_{LB}(A,B)}{GED(A,B)}\)</span> of a dataset, for graphs <span class="InlineEquation" id="IEq278">\(A\)</span> and <span class="InlineEquation" id="IEq279">\(B\)</span>. The tightness value is in <span class="InlineEquation" id="IEq280">\([0,1]\)</span>.
 Values closer to 1 denote a tighter lower bound which in turn is more 
useful. The pruning power is defined as the number of comparisons using a
 distance function <span class="InlineEquation" id="IEq281">\(d\)</span> that are pruned from a brute-force 1-NN search when its lower bound <span class="InlineEquation" id="IEq282">\(d_{LB}\)</span> is first computed.</p><p class="Para">In
 practice, we have no exact algorithm, that is fast enough, to compute 
the true value of function GED in each instance. Instead, we compute the
 ratio of the lower bound over the distance returned by the Simulated 
Annealing algorithm. Thus, not only is the true tightness of the LB 
higher than the one witnessed, we also acquire an insight on how tight 
both SA and the lower bound turn out to be, since SA over-approximates 
GED.</p><p class="Para">For the offline experiments, we chose 2 rather 
small data-sets. The first, is a selection of 194 samples hand-picked by
 human analysts and manually pre-classified into malware families. The 
second, is a collection of 2,867 samples that arrived in F-Secures 
backend systems in two consecutive days. For those days, we kept only 
the samples whose graphs contain 50500 vertices. In Table <span class="InternalRef"><a href="#Tab1">1</a></span>
 we present the dataset descriptions and the related results. As 
expected, we observe that the presence of clusters of similar files in 
the pre-classified set result into a lower tightness and weaker pruning 
power, compared to the randomly selected graphs. It is interesting to 
note the tightness of <span class="InlineEquation" id="IEq283">\(GED_{LB}\)</span> over SA. For the random set of graphs we witnessed a <span class="InlineEquation" id="IEq284">\(0.885\)</span> tightness value resulting in a 98.6&nbsp;% pruning power. Such a high tightness value shows that both <span class="InlineEquation" id="IEq285">\(GED_{LB}\)</span>
 and SA are tight. For the dataset containing hand-picked samples 
belonging to a restricted set of malware families, we notice that the 
tightness and prunning power are lower. This provides the insight that <span class="InlineEquation" id="IEq286">\(GED_{LB}\)</span>
 suffices for the case where the compared samples are very different, 
but for more difficult cases it is imperative to use a more expensive 
comparison algorithm. Finally, the run-time for computing the <span class="InlineEquation" id="IEq287">\(GED_{LB}\)</span> we had witnessed in our offline experiments was approximately 0.15&nbsp;s for graphs of order <span class="InlineEquation" id="IEq288">\(500\)</span> while for graphs of order 3000 it was less than 1 s.</p><div class="Para">In the live system, the performance of the <span class="InlineEquation" id="IEq289">\(GED_{LB}\)</span>
 is measured by the ratio of comparisons that are pruned from the 
combination of the incoming sample and the reference samples of its 
candidate clusters. Given an incoming sample <span class="InlineEquation" id="IEq290">\(q\)</span> with graph <span class="InlineEquation" id="IEq291">\(G_q=(V_q,E_q)\)</span>, The candidate clusters <span class="InlineEquation" id="IEq292">\(\mathcal {C}^q\)</span> are selected by querying for clusters whose reference samples contain <span class="InlineEquation" id="IEq293">\(a\cdot |V_q|\)</span> vertices, with <span class="InlineEquation" id="IEq294">\(a\in [0.95,1.05]\)</span>. The threshold for which candidate comparisons are pruned was set to <span class="InlineEquation" id="IEq295">\(0.1\)</span>
 (there is no ground truth for determining this value, it was chosen 
empirically after consulting with the malware analysts). Naturally, this
 is also the similarity threshold for our clustering algorithm. For the 
last 98806 successfully extracted graphs, the total sum of candidate 
comparisons was 105374736, while after the pruning phase remained <span class="InlineEquation" id="IEq296">\(2383025\)</span>,
 resulting in a 97.7&nbsp;% pruning power. Such a speedup proves 
significant especially in the case of samples that are not similar to 
any existing sample and form a cluster of their own. The <span class="InlineEquation" id="IEq297">\(GED_{LB}\)</span>
 is the component that makes the whole task of clustering streams of 
call-graphs feasible, since in average a set of 1066 comparisons are 
reduced to 24 by performing very cheap operations.<div id="Tab1" class="Table"><div class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Table 1</span> <p class="SimplePara">Tightness and pruning power of the lower bound on the offline experiments</p> </div></div><div class="u-scroll-horizontal"><table><colgroup><col class="tcol1 align-left"><col class="tcol2 align-left"><col class="tcol3 align-left"><col class="tcol4 align-left"><col class="tcol5 align-left"></colgroup><thead><tr><th> <p class="SimplePara">Dataset</p> </th><th> <p class="SimplePara">Graphs</p> </th><th> <p class="SimplePara">Graph order</p> </th><th> <p class="SimplePara">Tightness</p> </th><th> <p class="SimplePara">Pruning power</p> </th></tr></thead><tbody><tr><td> <p class="SimplePara">Pre-classified</p> </td><td> <p class="SimplePara">194</p> </td><td> <p class="SimplePara">201800</p> </td><td> <p class="SimplePara">0.749</p> </td><td> <p class="SimplePara">79.6</p> </td></tr><tr><td> <p class="SimplePara">Random</p> </td><td> <p class="SimplePara">2867</p> </td><td> <p class="SimplePara">50500</p> </td><td> <p class="SimplePara">0.885</p> </td><td> <p class="SimplePara">98.6</p> </td></tr></tbody></table></div></div> </div><p class="Para">Suppose we had selected indexing schemes that rely on embeddings such as (Venkateswaran et al. <span class="CitationRef"><a href="#CR60" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>; Papapetrou et al. <span class="CitationRef"><a href="#CR47" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>)
 and had selected an embedding set with cardinality of 24; a low number 
for the graph cardinality range that we study. Then, immediately at the 
first stage of the process (comparing the query call-graph with the 
reference objects to determine the suitable sub-space) the number of 
expensive graph comparisons would exceed that required by the average 
case of our approach. Similarly, SMIT (Hu et al. <span class="CitationRef"><a href="#CR31" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>)
 requires, based on the authors claims, on average 112 graph 
comparisons (median is 78, maximum is 918) for 5-NN queries over a 
dataset that contains only 102391 samples; significantly fewer indexed 
samples than in our database.</p></section><section id="Sec19" tabindex="-1" class="Section3 RenderAsSection3"><h4 class="Heading"><span class="HeadingNumber">8.1.3 </span>Graph comparisons using simulated annealing</h4><div class="Para">The
 most important part of the system is the SA graph comparison algorithm.
 We omit its comparison with existing bipartite matching-based methods 
in the literature, since this aspect has been the main subject of 
Kostakis et al. (<span class="CitationRef"><a href="#CR39" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>). In our Classy instance, SA is used with parameter values <span class="InlineEquation" id="IEq298">\(\beta _0=4.0\)</span> and <span class="InlineEquation" id="IEq299">\(\epsilon =0.9\)</span>. In Fig.&nbsp;<span class="InternalRef"><a href="#Fig6">6</a></span>
 we present a normalized histogram for the time required in the 
comparison phase for the samples clustered during the last 24 h. The 
mean comparison time is 23&nbsp;s, however we do witness 0.6&nbsp;% of 
samples requiring more than 200 s. We must make clear that each sample 
is not necessarily compared against every sample in its pruned 
candidates list. Instead, we take advantage of the fact that the lower 
bounds have already been computed. So, the list is sorted in ascending 
order of the LB scores and the comparisons stop as soon as a suitable 
candidate has been found. This trade-off offers a significant speedup 
and allows to maximize the systems throughput. Despite this trade-off, 
two isomorphic samples would still be assigned to the same cluster since
 their candidate list would be the same and thus the comparisons would 
happen in the same order with (approximately) the same results. The 
average ratio of comparisons avoided only in this stage is 84.42&nbsp;%.<figure class="Figure" id="Fig6"><div class="MediaObject" id="MO22"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Fig6_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Fig6_HTML.gif" alt="Fig. 6"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 6</span> <p class="SimplePara">Normalized
 histogram of the total time consumed in the comparison stage for each 
graph. The mean comparison time is 23&nbsp;s. We have witnessed 
0.6&nbsp;% of the graphs requiring more than 200&nbsp;s</p> </div></figcaption></figure> </div><p class="Para">The
 last phase of the whole process is to identify isomorphic graphs within
 the cluster(s) which the incoming sample has been assigned to. This 
step is performed if and only if the incoming samples has not been found
 to be isomorphic to any other sample it had been compared against, 
during the clustering phase. For the last 81064 graphs reaching this 
phase, 34047 graphs had already been identified as isomorphic to an 
existing database graph, while for the rest 47017, 184710 comparisons 
were performed. This is an average 3.9 additional comparisons for the 
samples that had not been identified as isomorphic to any database 
graphs in the comparison stage, and a system average of 2.2 comparisons 
per clustered sample.</p><p class="Para">Finally, we compute the count 
of identified sets of isomorphic graphs in each cluster. This provides 
an intuition about a combination of things: the power of the SA 
comparison algorithm, the clustering quality cost induced by applying 
the aforementioned trade-off in the comparison stage and more directly, 
the actual speedup in computing the cluster statistics (presented in 
Sect.&nbsp;<span class="InternalRef"><a href="#Sec14">7.2</a></span>). For all clusters containing more than 20 graphs the mean ratio of the number of sets of isomorphic graphs (<span class="InlineEquation" id="IEq300">\(k\)</span>) over the total number of samples in a cluster is <span class="InlineEquation" id="IEq301">\(0.132\)</span>.
 There is no clear indication, by only looking at the number of 
isomorphic sets and/or samples in a cluster, whether the samples are 
benign or malicious.</p></section><section id="Sec20" tabindex="-1" class="Section3 RenderAsSection3"><h4 class="Heading"><span class="HeadingNumber">8.1.4 </span>Scalability</h4><p class="Para">The
 fact that our system follows a double producer-consumer design pattern 
allows us to add more resources of each type. In particular, we are able
 to freely add more masters or slaves if that becomes necessary. 
Clearly, more workers allow for greater system throughput. On a typical 
day, we observe a per-worker average of 240 samples clustered in 24 h, 
each one using its own dedicated CPU core. The statistics presented so 
far are drawn from our instance of Classy that has up to 48 workers, 
hence running on 48 CPU cores. Hence, Classy is able to cluster more 
than 11000 call-graphs per day.</p><p class="Para">The throughput of the
 system appears to scale linearly with the number of available workers. 
Since the workers only access the distributed file-system for storing 
the call-graphs that should scale to handle the added workers. Regarding
 the file-system, one must make a trade-off between storage space and 
latency. In particular, Network File System (NFS) can be used for small 
to medium clusters and for data that fit in a RAID configuration. For 
larger clusters and graph repositories, more scalable filesystems are 
needed; Classy uses GlusterFS. In practice, we have observed that a 
dual-layer (in-memory and local disk) caching under an LRU scheme works 
very well to bridge the latency gap. Since the call-graphs are never 
modified, there is no concern for maintaining consistency of data among 
different caches and across different servers.</p><p class="Para">Regarding
 the masters, the number of instances does not significantly affect the 
system performance. A single master process can support up to 48 workers
 without a humanly observable delay. However, we would suggest the 
presence of at least two master processes. This is to avoid having a 
single point of failure, to minimize the wait time of tasks in the 
masters queue, and to minimize the delay incurred by potentially slow 
queries to the relational database. The database that is accessed by the
 masters is used only for maintaining the information about the samples 
and the clusters. Hence, Classy is far from saturating the scalability 
capabilities of that.</p></section><section id="Sec21" tabindex="-1" class="Section3 RenderAsSection3"><h4 class="Heading"><span class="HeadingNumber">8.1.5 </span>Application: malware clustering and detection</h4><p class="Para">Classy
 can be used for clustering graphs related to any application. The 
original motivation was that of clustering executable programs for 
facilitating the work of security response teams in data security 
companies. Below we provide information related to the application of 
large-scale malware clustering and provide some insights.</p><p class="Para">The
 feature that clearly separates Classy from other automation systems is 
the ability to process and group DLL (dynamic-link library) files. In 
the typical case, DLL files are imported by a host executable file and 
they cannot be executed in stand-alone mode. Hence behavior-based 
analysis systems, such as Hegedus et al. (<span class="CitationRef"><a href="#CR29" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>), Rieck et al. (<span class="CitationRef"><a href="#CR49" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2008</a></span>), Bayer et al. (<span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>),
 are incapable or inefficient, since this imposes the requirement to 
have acquired both the host executable and the DLL files during the time
 of execution. Instead, Classy can process and cluster together similar 
DLL files independently of their launching executable. Modern malware 
make extensive use of DLL files; reasons for this include software 
modularity and code injection capabilities. Furthermore, DLL files are 
susceptible to most of the aforementioned obfuscation techniques. The 
amount of DLL files clustered by Classy corresponds to 13.2&nbsp;% of 
the total files, and 18.8&nbsp;% of those are malicious. The number of 
malicious DLLs in the wild is in the order of tens of millions. So, this
 is a clear indicator of part of the added value Classy may provide for 
the application of malware clustering.</p><p class="Para">Apart from 
knowledge extraction capabilities aimed at human analysts, clustering 
data objects allows the task of supervised classification. While the 
classification results of offline clustering of call-graphs (Kinable and
 Kostakis <span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>) carry over to the stream setting (as we demonstrate in Sect.&nbsp;<span class="InternalRef"><a href="#Sec22">8.2</a></span>),
 we are able to provide results for a larger set of files. The first 
important insight is that, in the general case, automatically 
classifying files as malware should not be performed by default by 
taking into account all the samples labels in a cluster. This is due to
 the fact that malware is often masqueraded to appear as legitimate 
software (by <em class="EmphasisTypeItalic ">trojanizing</em> them); a 
legitimate and a trojanized version of the same software would get 
clustered together by SA but they would not be isomorphic. Similarly, 
benign and malware could share great portions of common functionality. 
Hence, this could lead to severe false alarms. Instead, unless it is 
possible to use additional information or an analysts intervention, 
automated classifications should be restricted to only groups of 
isomorphic graphs. This does not mean that the clustering task should be
 restricted only to identifying isomorphisms, since, as discussed above,
 grouping similar but not necessarily isomorphic call-graphs is a highly
 valuable tool for the whole stack of a security response team.</p><div class="Para">We
 proceed to evaluate the clustering results. To be sure that a cluster 
has enough samples to make a confident decision, we investigate those 
clusters that contain more than 20 graphs. In total, they consist of 
74.3&nbsp;% of all clustered files. Table <span class="InternalRef"><a href="#Tab2">2</a></span>
 depicts the percentage of clusters that contain only malware, only 
benign files or both. The percentage of respecitive samples are in 
brackets. We witness that only 5.4&nbsp;% of clusters contain both clean
 and malware samples, affecting 12.42&nbsp;% of the samples (larger 
clusters are expected to contain misclassifications produced by human 
analysts or imprecise pattern based detections). Clearly, by decreasing 
the distance threshold used in clustering, the value of clusters 
containing both malware and clean files should decrease. Table <span class="InternalRef"><a href="#Tab3">3</a></span>
 depicts the percentage of samples that belong to groups of isomorphic 
call-graphs that contain only malware, only benign files or both. So, 
the groups containing only malware correspond to 16.3&nbsp;% of the 
total samples. The number of samples present in sets containing both 
benign and malware is equal to 2.9&nbsp;% of the total samples. This is 
the source of the false-positives. The strategy by which the labels are 
assigned to new files would determine the amount of experienced 
false-positives or missed files. At this stage, assigning the label 
malware preferentially, the ratio of false-positive files is 
1.2&nbsp;% over all samples. With information from other means and 
systems, we can identify which of those call-graphs correspond to 
installers, custom packers etc, dropping the actual false-positive rate 
to 0.5&nbsp;%. In practice, the results of any automated malware 
classifier are passed onto a rule-based system. Through that, we provide
 all involved parties with the capability to write sophisticated rules 
on top of the clustering results and to combine file metadata and 
information from other sources. As a result, this allows to reduce the 
false positives that affect the end-users while also being able to 
increase the coverage of automated malware classification beyond 
isomorphism relations on a per-case basis. Overall, the combination of 
Classy and a rule-based decision system enables not only a more 
powerfull automated classification mechanism but also the basis for 
other knowledge extraction and recommender systems.<div id="Tab2" class="Table"><div class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Table 2</span> <p class="SimplePara">Percentage
 of clusters that contain only malware, clean files or both; in brackets
 the percentage of corresponding samples. The numbers correspond to the 
samples in the clusters with more than 20 graphs</p> </div></div><div class="u-scroll-horizontal"><table><colgroup><col class="tcol1 align-left"><col class="tcol2 align-left"><col class="tcol3 align-left"><col class="tcol4 align-left"></colgroup><thead><tr><th>&nbsp;</th><th>&nbsp;</th><th colspan="2"> <p class="SimplePara">Contains clean files</p> </th></tr><tr><th>&nbsp;</th><th>&nbsp;</th><th> <p class="SimplePara">Yes</p> </th><th> <p class="SimplePara">No</p> </th></tr></thead><tbody><tr><td rowspan="2"> <p class="SimplePara">Cluster contains malware</p> </td><td> <p class="SimplePara">Yes</p> </td><td> <p class="SimplePara">5.4 (12.42)&nbsp;%</p> </td><td> <p class="SimplePara">12.5 (10.2)&nbsp;%</p> </td></tr><tr><td> <p class="SimplePara">No</p> </td><td> <p class="SimplePara">82.1 (77.38)&nbsp;%</p> </td><td> <p class="SimplePara"></p> </td></tr></tbody></table></div></div> <div id="Tab3" class="Table"><div class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Table 3</span> <p class="SimplePara">Percentage
 of samples belonging to groups of isomorphic graphs that contain only 
malware, clean files or both. The numbers correspond to the samples in 
the clusters with more than 20 graphs</p> </div></div><div class="u-scroll-horizontal"><table><colgroup><col class="tcol1 align-left"><col class="tcol2 align-left"><col class="tcol3 align-left"><col class="tcol4 align-left"></colgroup><thead><tr><th>&nbsp;</th><th>&nbsp;</th><th colspan="2"> <p class="SimplePara">Contains clean files</p> </th></tr><tr><th>&nbsp;</th><th>&nbsp;</th><th> <p class="SimplePara">Yes</p> </th><th> <p class="SimplePara">No</p> </th></tr></thead><tbody><tr><td rowspan="2"> <p class="SimplePara">Group contains malware</p> </td><td> <p class="SimplePara">Yes</p> </td><td> <p class="SimplePara">2.9&nbsp;%</p> </td><td> <p class="SimplePara">16.3&nbsp;%</p> </td></tr><tr><td> <p class="SimplePara">No</p> </td><td> <p class="SimplePara">80.8&nbsp;%</p> </td><td> <p class="SimplePara"></p> </td></tr></tbody></table></div></div> </div><p class="Para">The
 structural similarity aspect is a big advantage for Classy over backend
 run-time behavior analysis systems in the cases of malware that rely on
 <em class="EmphasisTypeItalic ">downloader</em> components to fetch the
 actual malicious payload from the Internet. In an isolated execution 
environment, it is impossible for the payload to be downloaded and 
consequently any relevant behavior to be expressed. This problem is 
usually circumvented by implementing a host-based intrusion prevention 
system on the users machine. We witness that with Classy, we were able 
to detect different instances of similar downloading components in the 
backend, by clustering them together; including variants of downloader 
components of <span class="EmphasisFontCategoryNonProportional ">ZeroAccess</span>. This is also beneficial since it allows to handle right away the downloading component instead of the downloaded files.</p><p class="Para">Additional
 interesting cases where Classy proved superior to other systems 
comprise of instances of files that, upon execution, write to disk a 
modified version of themselves (with some element of randomness). In a 
sandboxed execution setting, the newly dropped file would have to be 
analyzed too via execution that would in turn produce yet another 
modified version and so on, thus resulting in an endless inheritance 
path. With Classy, we are able to cluster the original file immediately,
 and also verify that a potential new file is similar to the original. 
We have observed such cases for both benign and malicious software.</p><p class="Para">As
 discussed previously, one of the limitations of Classy concerns unknown
 packers and the case where the extracted call-graphs correspond to the 
common unpacking module. This is the source of some of the false alarms.
 However, we witness that in certain occasions this can be an advantage,
 since certain packers are used exclusively by malware authors. Hence, 
while new malware might exhibit behavior never seen before, and thus go 
undetected by run-time analysis systems, the produced call-graphs would 
be similar to previously known. By using the clustering results by 
Classy, the malware analysts are enabled to write rules that capture and
 assign guilty-by-association verdicts to previously unknown malware. 
Example cases of malware that share common packers include variants of 
the <span class="EmphasisFontCategoryNonProportional ">W32/Carberp</span> bootkit and <span class="EmphasisFontCategoryNonProportional ">W32/Lockscreen</span>
 ransomware. Furthermore, in some of those cases, to avoid pattern-based
 detection of the common unpacking module that remains visible, the 
authors modify them manually, or the packer self-mutates; the latter is 
known widely as <em class="EmphasisTypeItalic ">polymorphic malware</em> (Xu et al. <span class="CitationRef"><a href="#CR65" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>)
 and pose an important challenge in malware detection. Even in such 
cases, Classy is expected to be able to group them together, provided 
that the call-graph representation is not severely altered.</p><p class="Para">Similarly
 to packers, other problematic cases that we have observed include files
 that enclose interpreters. Such examples include <span class="EmphasisFontCategoryNonProportional ">AutoIt</span>
 files, where custom bytecode is produced for execution by the 
interpreter. As a result, any call-graphs extracted via static analysis 
fail to encapsulate the actual functionality that is performed through 
the interpreter. Under these circumstances, two options are available. 
The first option is to extend the file-filtering rules and avoid 
clustering them altogether, or to gray-list the file type and avoid 
making automated decisions for those. The second option is to improve 
the call-graph extraction procedure in a way that can trace the 
functionality of the interpreter. As discussed in Sect.&nbsp;<span class="InternalRef"><a href="#Sec4">3.1</a></span>,
 this entails additional technical challenges and requirements, but our 
clustering framework would be capable to handle the data.</p></section></section><section id="Sec22" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">8.2 </span>Comparison to offline optimal clustering</h3><p class="Para">The
 statistics from the live system prove that the used techniques provide 
very good speedups. Still, we need to gain an insight on the quality of 
the clustering. At this point we investigate the performance of our 
stream clustering algorithm versus the optimal results, or the ground 
truth, as that is provided by an offline algorithm. The baseline method 
in this case is the DBSCAN method (Ester et al. <span class="CitationRef"><a href="#CR20" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1996</a></span>), used by the authors in Kinable and Kostakis (<span class="CitationRef"><a href="#CR36" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2011</a></span>),
 and is applied on whole distance matrices. So, for this experiment we 
produce the distance matrix for the Random dataset of Table <span class="InternalRef"><a href="#Tab1">1</a></span>,
 and we apply DBSCAN to obtain the optimal clustering; optimal in the 
sense that it is computed offline with no performance trade-offs. We 
then proceed to compute the clustering produced by our stream clustering
 algorithm. Comparing the two algorithms is meaningful only if the same 
parameters are used. In particular, the threshold parameter of our 
algorithm is equivalent to the <span class="EmphasisFontCategoryNonProportional ">eps</span>
 parameter of DBSCAN. Since in our clustering algorithm a single sample 
can form a cluster of its own, for DBSCAN this translates to setting <span class="EmphasisFontCategoryNonProportional ">minpts</span>
 to 1. The algorithm we propose will never produce less clusters than 
DBSCAN. It is trivial to prove that if two samples would be clustered 
together in the stream setting, the conditions for doing so would 
suffice for DBSCAN to cluster them together, too.</p><p class="Para">We 
benchmark the stream clustering algorithm in terms of the number of 
clusters produced, compared to the optimal, and we also compute the 
Adjusted Rand Index (Hubert and Arabie <span class="CitationRef"><a href="#CR32" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1985</a></span>).
 The number of clusters can be misleading since it is not clear how the 
samples are distributed among the clusters. It might be that the extra 
clusters consist of individual samples or that a real cluster is split 
into two equal parts. For this reason we believe the Adjusted Rand Index
 is a more accurate indication of clustering quality. The Adjusted Rand 
Index compares two clusterings and provides a measure of agreement. It 
takes values between <span class="InlineEquation" id="IEq302">\(-\)</span>1
 and 1, with 1 indicating that the two clusterings are identical. Due to
 the fact that in the stream algorithm there is a strong element of 
randomness, i.e. the order in which the samples arrive through the 
stream, we perform a significant number of simulations. In particular, 
we perform <span class="InlineEquation" id="IEq303">\(10000\)</span> simulations with a random permutation of the stream in each case. We benchmark the algorithms with values <span class="InlineEquation" id="IEq304">\(0.1\)</span>, <span class="InlineEquation" id="IEq305">\(0.2\)</span> and <span class="InlineEquation" id="IEq306">\(0.3\)</span> for the threshold parameter and <span class="EmphasisFontCategoryNonProportional ">eps</span>. While <span class="InlineEquation" id="IEq307">\(0.1\)</span>
 is the one used in Classy, we investigate the behavior under other 
parameter values in order to get a strong insight on the algorithm. We 
also simulate the case where the streaming algorithm breaks as soon as a
 first candidate sample has been found.</p><div class="Para">Figure&nbsp;<span class="InternalRef"><a href="#Fig7">7</a></span>ac depict the sorted distribution of the increase in the number of produced clusters. Figure&nbsp;<span class="InternalRef"><a href="#Fig8">8</a></span>ac depict the histograms of the those distributions. We notice, in all cases of Fig.&nbsp;<span class="InternalRef"><a href="#Fig8">8</a></span>,
 that the distributions resemble a Gaussian-like bell curve. As 
expected, their curve centers deviate from the optimal solution as the 
value of <span class="EmphasisFontCategoryNonProportional ">eps</span> increases. The deviation remains limited, even for <span class="EmphasisFontCategoryNonProportional ">eps</span> <span class="InlineEquation" id="IEq308">\(=0.3\)</span>,
 for the case of computing the distance with the reference samples of 
all candidate clusters. Essentially, when the comparison breaks on the 
first suitable candidate, no cluster merges happen, so the number of 
produced clusters is greater. Figure <span class="InternalRef"><a href="#Fig9">9</a></span>ac depict the sorted distributions of the Adjusted Rand index for all <span class="EmphasisFontCategoryNonProportional ">eps</span> values. We witness that for the vast majority of simulations,the value of the Adjusted Rand index remains above <span class="InlineEquation" id="IEq309">\(0.975\)</span> for both scenarios and <span class="EmphasisFontCategoryNonProportional ">eps</span> <span class="InlineEquation" id="IEq310">\(=0.1,0.2\)</span>. For <span class="EmphasisFontCategoryNonProportional ">eps</span> <span class="InlineEquation" id="IEq311">\(=0.3\)</span>, when the clustering does not break once a suitable candidate has been found, the Index remains above <span class="InlineEquation" id="IEq312">\(0.95\)</span> for almost all permutations. Finally, the value of the Adjusted Rand Index deteriorates for <span class="EmphasisFontCategoryNonProportional ">eps</span> <span class="InlineEquation" id="IEq313">\(=0.3\)</span> when the algorithm does not merge clusters.
<figure class="Figure" id="Fig7"><div class="MediaObject" id="MO23"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Fig7_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Fig7_HTML.gif" alt="Fig. 7"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 7</span> <p class="SimplePara">Sorted distribution of the increase in the number of clusters </p> </div></figcaption></figure> <figure class="Figure" id="Fig8"><div class="MediaObject" id="MO24"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Fig8_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Fig8_HTML.gif" alt="Fig. 8"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 8</span> <p class="SimplePara">Histogram of the increase in the number of clusters produced by our stream clustering algorithm versus offline DBSCAN </p> </div></figcaption></figure> <figure class="Figure" id="Fig9"><div class="MediaObject" id="MO25"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-014-0367-9/MediaObjects/10618_2014_367_Fig9_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2014_367_Fig9_HTML.gif" alt="Fig. 9"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 9</span> <p class="SimplePara">Sorted distribution of the values of the adjusted rand index between our stream clustering algorithm and offline DBSCAN </p> </div></figcaption></figure> </div></section></div></section><section id="Sec23" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">9 </span>Conclusions<span class="section-icon"></span></h2><div class="content"><p class="Para">To
 handle the ever increasing number of binary executable files received 
by anti-virus companies every day, the design and use of system 
automation is required. Furthermore, instruction-level obfuscation 
techniques are used by malware authors in an attempt to prevent the 
detection of their malware by pattern-based techniques. An abstraction 
resilient to many obfuscation methods is the call-graph; the 
representation of a binary executable file as a directed graph. 
Unfortunately, interesting graph comparison problems such as full-graph 
comparison belong to the class of <span class="InlineEquation" id="IEq314">\(NP\)</span>-hard problems.</p><p class="Para">We
 presented Classy, a distributed system that clusters streams of large 
graphs. The graph comparison measure is the GED and it is approximated 
using an adapted version of Simulated Annealing. In addition, we 
presented a linear-time lower bound for the GED that proved very tight 
and with great pruning power. We proposed a suitable clustering 
algorithm that demonstrates a performance not far from the offline 
optimal. Finally, we studied the problem of approximating the statistics
 of clusters for extracting cluster summaries when only part of all 
distances are available. Experiments were conducted on the proposed 
algorithms and statistics were provided of a system instance that has 
clustered and contains over 0.8 million graphs. The results were highly 
promising. Overall, we demonstrated how a set of seemingly simple 
algorithms can be assembled together in order to build a system that 
adequately approximates instances of NP-hard problems tens of thousands 
of times per day for delicate tasks such as automated malware 
classification.</p></div></section></div><section id="Notes" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1">Notes<span class="section-icon"></span></h2><div class="content"><section><h3 class="Heading">Acknowledgments</h3><p class="SimplePara">This
 work was supported by TEKES as part of the Future Internet Programme of
 TIVIT (Finnish Strategic Centre for Science, Technology and Innovation 
in the field of ICT). Special thanks to Paolo Palumbo for providing the 
file filtering rules, Gergely Erdlyi for his support on IDA Python and 
the call-graph unpacking code, and Stefan Lundstrm for the early 
integration of the system with the backend APIs.</p></section></div></section><section class="Section1 RenderAsSection1" id="Bib1" tabindex="-1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1">References<span class="section-icon"></span></h2><div class="content"><ol class="BibliographyWrapper"><li class="Citation"><div class="CitationContent" id="CR1">Aggarwal
 CC, Han J, Wang J, Yu PS (2003) A framework for clustering evolving 
data streams. In: Proceedings of the 29th international conference on 
very large data bases-volume 29, VLDB Endowment, pp 8192<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Aggarwal%20CC%2C%20Han%20J%2C%20Wang%20J%2C%20Yu%20PS%20%282003%29%20A%20framework%20for%20clustering%20evolving%20data%20streams.%20In%3A%20Proceedings%20of%20the%2029th%20international%20conference%20on%20very%20large%20data%20bases-volume%2029%2C%20VLDB%20Endowment%2C%20pp%2081%E2%80%9392"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR2">Aggarwal
 CC, Han J, Wang J, Yu PS (2004) A framework for projected clustering of
 high dimensional data streams. In: Proceedings of the thirtieth 
international conference on very large data bases-volume 30, VLDB 
Endowment, pp 852863<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Aggarwal%20CC%2C%20Han%20J%2C%20Wang%20J%2C%20Yu%20PS%20%282004%29%20A%20framework%20for%20projected%20clustering%20of%20high%20dimensional%20data%20streams.%20In%3A%20Proceedings%20of%20the%20thirtieth%20international%20conference%20on%20very%20large%20data%20bases-volume%2030%2C%20VLDB%20Endowment%2C%20pp%20852%E2%80%93863"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR3">Aggarwal
 C, Zhao Y, Yu P (2010) On clustering graph streams. In: Proceedings of 
the SIAM international conference on data mining, pp 478489<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Aggarwal%20C%2C%20Zhao%20Y%2C%20Yu%20P%20%282010%29%20On%20clustering%20graph%20streams.%20In%3A%20Proceedings%20of%20the%20SIAM%20international%20conference%20on%20data%20mining%2C%20pp%20478%E2%80%93489"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR4">Akutsu
 T (1993) A polynomial time algorithm for finding a largest common 
subgraph of almost trees of bounded degree. IEICE Trans Fundam Electron 
Commun Comput Sci 76(9):14881493<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20polynomial%20time%20algorithm%20for%20finding%20a%20largest%20common%20subgraph%20of%20almost%20trees%20of%20bounded%20degree&amp;author=T.%20Akutsu&amp;journal=IEICE%20Trans%20Fundam%20Electron%20Commun%20Comput%20Sci&amp;volume=76&amp;issue=9&amp;pages=1488-1493&amp;publication_year=1993"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR5">Bayer
 U, Comparetti PM, Hlauschek C, Kruegel C, Kirda E (2009) Scalable, 
behavior-based malware clustering. In: 16th Network &amp; distributed 
system security conference, vol 9, pp 811<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Bayer%20U%2C%20Comparetti%20PM%2C%20Hlauschek%20C%2C%20Kruegel%20C%2C%20Kirda%20E%20%282009%29%20Scalable%2C%20behavior-based%20malware%20clustering.%20In%3A%2016th%20Network%20%26%20distributed%20system%20security%20conference%2C%20vol%209%2C%20pp%208%E2%80%9311"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR6">Bourquin
 M, King A, Robbins E (2013) Binslayer: accurate comparison of binary 
executables. In: Proceedings of the 2nd ACM SIGPLAN program protection 
and reverse engineering workshop, ACM, p 4<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Bourquin%20M%2C%20King%20A%2C%20Robbins%20E%20%282013%29%20Binslayer%3A%20accurate%20comparison%20of%20binary%20executables.%20In%3A%20Proceedings%20of%20the%202nd%20ACM%20SIGPLAN%20program%20protection%20and%20reverse%20engineering%20workshop%2C%20ACM%2C%20p%204"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR7">Briones
 I, Gomez A (2008) Graphs, entropy and grid computing: automatic 
comparison of malware. Proceedings of the virus bulletin conference, pp 
112<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Briones%20I%2C%20Gomez%20A%20%282008%29%20Graphs%2C%20entropy%20and%20grid%20computing%3A%20automatic%20comparison%20of%20malware.%20Proceedings%20of%20the%20virus%20bulletin%20conference%2C%20pp%201%E2%80%9312"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR8">Bunke H (1997) On a relation between graph edit distance and maximum common subgraph. Pattern Recognit Lett 18(8):689694<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/S0167-8655%2897%2900060-3"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=1441173"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=On%20a%20relation%20between%20graph%20edit%20distance%20and%20maximum%20common%20subgraph&amp;author=H.%20Bunke&amp;journal=Pattern%20Recognit%20Lett&amp;volume=18&amp;issue=8&amp;pages=689-694&amp;publication_year=1997"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR9">Burkhard W, Keller R (1973) Some approaches to best-match file searching. Commun ACM 16(4):230236<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/362003.362025"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceZLBID"><a class="gtm-reference" data-reference-type="MATH" target="_blank" rel="noopener" href="http://www.emis.de/MATH-item?0269.68062"><span><span>MATH</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Some%20approaches%20to%20best-match%20file%20searching&amp;author=W.%20Burkhard&amp;author=R.%20Keller&amp;journal=Commun%20ACM&amp;volume=16&amp;issue=4&amp;pages=230-236&amp;publication_year=1973"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR10">Cao
 F, Ester M, Qian W, Zhou A (2006) Density-based clustering over an 
evolving data stream with noise. In: Proceedings of the SIAM 
international conference on data mining, pp 328339<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Cao%20F%2C%20Ester%20M%2C%20Qian%20W%2C%20Zhou%20A%20%282006%29%20Density-based%20clustering%20over%20an%20evolving%20data%20stream%20with%20noise.%20In%3A%20Proceedings%20of%20the%20SIAM%20international%20conference%20on%20data%20mining%2C%20pp%20328%E2%80%93339"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR11">Carrera
 E, Erdlyi G (2004) Digital genome mapping-advanced binary malware 
analysis. In: Proceedings of the virus bulletin conference, pp 187197<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Carrera%20E%2C%20Erd%C3%A9lyi%20G%20%282004%29%20Digital%20genome%20mapping-advanced%20binary%20malware%20analysis.%20In%3A%20Proceedings%20of%20the%20virus%20bulletin%20conference%2C%20pp%20187%E2%80%93197"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR12">Charikar
 M, OCallaghan L, Panigrahy R (2003) Better streaming algorithms for 
clustering problems. In: Proceedings of the ACM symposium on theory of 
computing, ACM, pp 3039<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Charikar%20M%2C%20O%E2%80%99Callaghan%20L%2C%20Panigrahy%20R%20%282003%29%20Better%20streaming%20algorithms%20for%20clustering%20problems.%20In%3A%20Proceedings%20of%20the%20ACM%20symposium%20on%20theory%20of%20computing%2C%20ACM%2C%20pp%2030%E2%80%9339"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR13">Cheng J, Ke Y, Ng W (2009) Efficient query processing on graph databases. ACM Trans Database Syst (TODS) 34(1):2<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/1508857.1508859"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Efficient%20query%20processing%20on%20graph%20databases&amp;author=J.%20Cheng&amp;author=Y.%20Ke&amp;author=W.%20Ng&amp;journal=ACM%20Trans%20Database%20Syst%20%28TODS%29&amp;volume=34&amp;issue=1&amp;pages=2&amp;publication_year=2009"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR14">Christodorescu M, Jha S (2004) Testing malware detectors. ACM SIGSOFT Softw Eng Notes 29(4):3444<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/1013886.1007518"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Testing%20malware%20detectors&amp;author=M.%20Christodorescu&amp;author=S.%20Jha&amp;journal=ACM%20SIGSOFT%20Softw%20Eng%20Notes&amp;volume=29&amp;issue=4&amp;pages=34-44&amp;publication_year=2004"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR15">Conte
 D, Foggia P, Sansone C, Vento M (2004) Thirty years of graph matching 
in pattern recognition. Int J Pattern Recognit Artif Intell 
18(03):265298<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1142/S0218001404003228"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Thirty%20years%20of%20graph%20matching%20in%20pattern%20recognition&amp;author=D.%20Conte&amp;author=P.%20Foggia&amp;author=C.%20Sansone&amp;author=M.%20Vento&amp;journal=Int%20J%20Pattern%20Recognit%20Artif%20Intell&amp;volume=18&amp;issue=03&amp;pages=265-298&amp;publication_year=2004"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR16">Datar
 M, Immorlica N, Indyk P, Mirrokni VS (2004) Locality-sensitive hashing 
scheme based on p-stable distributions. In: Proceedings of the twentieth
 annual symposium on computational geometry, ACM, pp 253262<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Datar%20M%2C%20Immorlica%20N%2C%20Indyk%20P%2C%20Mirrokni%20VS%20%282004%29%20Locality-sensitive%20hashing%20scheme%20based%20on%20p-stable%20distributions.%20In%3A%20Proceedings%20of%20the%20twentieth%20annual%20symposium%20on%20computational%20geometry%2C%20ACM%2C%20pp%20253%E2%80%93262"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR17">Dean
 T, Boddy M (1988) An analysis of time-dependent planning. In: 
Proceedings of the 17th national conference on artificial intelligence, 
pp 4954<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Dean%20T%2C%20Boddy%20M%20%281988%29%20An%20analysis%20of%20time-dependent%20planning.%20In%3A%20Proceedings%20of%20the%2017th%20national%20conference%20on%20artificial%20intelligence%2C%20pp%2049%E2%80%9354"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR18">Dullien T, Rolles R (2005) Graph-based comparison of executable objects. SSTIC 5:13<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Graph-based%20comparison%20of%20executable%20objects&amp;author=T.%20Dullien&amp;author=R.%20Rolles&amp;journal=SSTIC&amp;volume=5&amp;pages=1-3&amp;publication_year=2005"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR19">Elhadi
 AAE, Maarof MA, Barry BI (2013) Improving the detection of malware 
behaviour using simplified data dependent api call graph. Int J Secur 
Appl 7(5):2942<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Improving%20the%20detection%20of%20malware%20behaviour%20using%20simplified%20data%20dependent%20api%20call%20graph&amp;author=AAE.%20Elhadi&amp;author=MA.%20Maarof&amp;author=BI.%20Barry&amp;journal=Int%20J%20Secur%20Appl&amp;volume=7&amp;issue=5&amp;pages=29-42&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR20">Ester
 M, Kriegel HP, Sander J, Xu X (1996) A density-based algorithm for 
discovering clusters in large spatial databases with noise. KDD 
96:226231<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20density-based%20algorithm%20for%20discovering%20clusters%20in%20large%20spatial%20databases%20with%20noise&amp;author=M.%20Ester&amp;author=HP.%20Kriegel&amp;author=J.%20Sander&amp;author=X.%20Xu&amp;journal=KDD&amp;volume=96&amp;pages=226-231&amp;publication_year=1996"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR21">Flake
 H (2004) Structural comparison of executable objects. In: Proceedings 
of the international GI workshop on detection of intrusions and malware 
&amp; vulnerability assessment, pp 161174<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Flake%20H%20%282004%29%20Structural%20comparison%20of%20executable%20objects.%20In%3A%20Proceedings%20of%20the%20international%20GI%20workshop%20on%20detection%20of%20intrusions%20and%20malware%20%26%20vulnerability%20assessment%2C%20pp%20161%E2%80%93174"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR22">Floyd R (1962) Algorithm 97: shortest path. Commun ACM 5(6):345<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/367766.368168"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Algorithm%2097%3A%20shortest%20path&amp;author=R.%20Floyd&amp;journal=Commun%20ACM&amp;volume=5&amp;issue=6&amp;pages=345&amp;publication_year=1962"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR23">Gascon
 H, Yamaguchi F, Arp D, Rieck K (2013) Structural detection of android 
malware using embedded call graphs. In: Proceedings of the 2013 ACM 
workshop on artificial intelligence and security, ACM, pp 4554<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Gascon%20H%2C%20Yamaguchi%20F%2C%20Arp%20D%2C%20Rieck%20K%20%282013%29%20Structural%20detection%20of%20android%20malware%20using%20embedded%20call%20graphs.%20In%3A%20Proceedings%20of%20the%202013%20ACM%20workshop%20on%20artificial%20intelligence%20and%20security%2C%20ACM%2C%20pp%2045%E2%80%9354"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR24">Gionis A, Indyk P, Motwani R et al (1999) Similarity search in high dimensions via hashing. VLDB 99:518529<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Similarity%20search%20in%20high%20dimensions%20via%20hashing&amp;author=A.%20Gionis&amp;author=P.%20Indyk&amp;author=R.%20Motwani&amp;journal=VLDB&amp;volume=99&amp;pages=518-529&amp;publication_year=1999"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR25">Gionis
 A, Mannila H, Tsaparas P (2005) Clustering aggregation. In: Proceedings
 of the 21st international conference on data engineering (ICDE), IEEE, 
pp 341352<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Gionis%20A%2C%20Mannila%20H%2C%20Tsaparas%20P%20%282005%29%20Clustering%20aggregation.%20In%3A%20Proceedings%20of%20the%2021st%20international%20conference%20on%20data%20engineering%20%28ICDE%29%2C%20IEEE%2C%20pp%20341%E2%80%93352"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR26">Giugno
 R, Shasha D (2002) Graphgrep: a fast and universal method for querying 
graphs. In: Proceedings of the 16th international conference on pattern 
recognition, IEEE, vol 2, pp 112115<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Giugno%20R%2C%20Shasha%20D%20%282002%29%20Graphgrep%3A%20a%20fast%20and%20universal%20method%20for%20querying%20graphs.%20In%3A%20Proceedings%20of%20the%2016th%20international%20conference%20on%20pattern%20recognition%2C%20IEEE%2C%20vol%202%2C%20pp%20112%E2%80%93115"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR27">Guha
 S, Meyerson A, Mishra N, Motwani R, OCallaghan L (2003) Clustering 
data streams: theory and practice. IEEE Trans Knowl Data Eng 
15(3):515528<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/TKDE.2003.1198387"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Clustering%20data%20streams%3A%20theory%20and%20practice&amp;author=S.%20Guha&amp;author=A.%20Meyerson&amp;author=N.%20Mishra&amp;author=R.%20Motwani&amp;author=L.%20O%E2%80%99Callaghan&amp;journal=IEEE%20Trans%20Knowl%20Data%20Eng&amp;volume=15&amp;issue=3&amp;pages=515-528&amp;publication_year=2003"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR28">He
 H, Singh A (2006) Closure-tree: an index structure for graph queries. 
In: Proceedings of the 22nd international conference on data 
engineering, IEEE, pp 3838<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=He%20H%2C%20Singh%20A%20%282006%29%20Closure-tree%3A%20an%20index%20structure%20for%20graph%20queries.%20In%3A%20Proceedings%20of%20the%2022nd%20international%20conference%20on%20data%20engineering%2C%20IEEE%2C%20pp%2038%E2%80%9338"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR29">Hegedus
 J, Miche Y, Ilin A, Lendasse A (2011) Methodology for behavioral-based 
malware analysis and detection using random projections and k-nearest 
neighbors classifiers. In: Seventh international conference on 
computational intelligence and security (CIS), IEEE, pp 10161023<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Hegedus%20J%2C%20Miche%20Y%2C%20Ilin%20A%2C%20Lendasse%20A%20%282011%29%20Methodology%20for%20behavioral-based%20malware%20analysis%20and%20detection%20using%20random%20projections%20and%20k-nearest%20neighbors%20classifiers.%20In%3A%20Seventh%20international%20conference%20on%20computational%20intelligence%20and%20security%20%28CIS%29%2C%20IEEE%2C%20pp%201016%E2%80%931023"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR30">Hex-Rays (2008) Ida pro. <span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.hex-rays.com/"><span class="RefSource">http://www.hex-rays.com/</span></a></span>
                <span class="Occurrences"></span></div></li><li class="Citation"><div class="CitationContent" id="CR31">Hu
 X, Chiueh T, Shin K (2009) Large-scale malware indexing using 
function-call graphs. In: Proceedings of the 16th ACM conference on 
computer and communications security, ACM, pp 611620<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Hu%20X%2C%20Chiueh%20T%2C%20Shin%20K%20%282009%29%20Large-scale%20malware%20indexing%20using%20function-call%20graphs.%20In%3A%20Proceedings%20of%20the%2016th%20ACM%20conference%20on%20computer%20and%20communications%20security%2C%20ACM%2C%20pp%20611%E2%80%93620"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR32">Hubert L, Arabie P (1985) Comparing partitions. J Classif 2(1):193218<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1007/BF01908075"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Comparing%20partitions&amp;author=L.%20Hubert&amp;author=P.%20Arabie&amp;journal=J%20Classif&amp;volume=2&amp;issue=1&amp;pages=193-218&amp;publication_year=1985"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR33">Indyk
 P, Motwani R (1998) Approximate nearest neighbors: towards removing the
 curse of dimensionality. In: Proceedings of the thirtieth annual ACM 
symposium on theory of computing, ACM, pp 604613<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Indyk%20P%2C%20Motwani%20R%20%281998%29%20Approximate%20nearest%20neighbors%3A%20towards%20removing%20the%20curse%20of%20dimensionality.%20In%3A%20Proceedings%20of%20the%20thirtieth%20annual%20ACM%20symposium%20on%20theory%20of%20computing%2C%20ACM%2C%20pp%20604%E2%80%93613"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR34">Jiang
 H, Wang H, Yu P, Zhou S (2007) Gstring: a novel approach for efficient 
search in graph databases. In: Proceedings of the IEEE 23rd 
international conference on data engineering, IEEE, pp 566575<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Jiang%20H%2C%20Wang%20H%2C%20Yu%20P%2C%20Zhou%20S%20%282007%29%20Gstring%3A%20a%20novel%20approach%20for%20efficient%20search%20in%20graph%20databases.%20In%3A%20Proceedings%20of%20the%20IEEE%2023rd%20international%20conference%20on%20data%20engineering%2C%20IEEE%2C%20pp%20566%E2%80%93575"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR35">Kang
 MG, Poosankam P, Yin H (2007) Renovo: a hidden code extractor for 
packed executables. In: Proceedings of the 2007 ACM workshop on 
recurring malcode, ACM, pp 4653<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Kang%20MG%2C%20Poosankam%20P%2C%20Yin%20H%20%282007%29%20Renovo%3A%20a%20hidden%20code%20extractor%20for%20packed%20executables.%20In%3A%20Proceedings%20of%20the%202007%20ACM%20workshop%20on%20recurring%20malcode%2C%20ACM%2C%20pp%2046%E2%80%9353"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR36">Kinable J, Kostakis O (2011) Malware classification based on call graph clustering. J Comput Virol 7(4):233245<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1007/s11416-011-0151-y"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Malware%20classification%20based%20on%20call%20graph%20clustering&amp;author=J.%20Kinable&amp;author=O.%20Kostakis&amp;journal=J%20Comput%20Virol&amp;volume=7&amp;issue=4&amp;pages=233-245&amp;publication_year=2011"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR37">Kolbitsch
 C, Comparetti PM, Kruegel C, Kirda E, Zhou Xy, Wang X (2009) Effective 
and efficient malware detection at the end host. In: USENIX security 
symposium, pp 351366<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Kolbitsch%20C%2C%20Comparetti%20PM%2C%20Kruegel%20C%2C%20Kirda%20E%2C%20Zhou%20Xy%2C%20Wang%20X%20%282009%29%20Effective%20and%20efficient%20malware%20detection%20at%20the%20end%20host.%20In%3A%20USENIX%20security%20symposium%2C%20pp%20351%E2%80%93366"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR38">Kollios G, Potamias M, Terzi E (2013) Clustering large probabilistic graphs. IEEE Trans Knowl Data Eng 25(2):325336<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/TKDE.2011.243"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Clustering%20large%20probabilistic%20graphs&amp;author=G.%20Kollios&amp;author=M.%20Potamias&amp;author=E.%20Terzi&amp;journal=IEEE%20Trans%20Knowl%20Data%20Eng&amp;volume=25&amp;issue=2&amp;pages=325-336&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR39">Kostakis
 O, Kinable J, Mahmoudi H, Mustonen K (2011) Improved call graph 
comparison using simulated annealing. In: Proceedings of the 2011 ACM 
symposium on applied computing, ACM, pp 15161523<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Kostakis%20O%2C%20Kinable%20J%2C%20Mahmoudi%20H%2C%20Mustonen%20K%20%282011%29%20Improved%20call%20graph%20comparison%20using%20simulated%20annealing.%20In%3A%20Proceedings%20of%20the%202011%20ACM%20symposium%20on%20applied%20computing%2C%20ACM%2C%20pp%201516%E2%80%931523"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR40">Kriege N, Mutzel P (2012) Subgraph matching kernels for attributed graphs. arXiv preprint <span class="ExternalRef"><a target="_blank" rel="noopener" href="http://arxiv.org/abs/1206.6483"><span class="RefSource">arXiv:1206.6483</span></a></span>
                <span class="Occurrences"></span></div></li><li class="Citation"><div class="CitationContent" id="CR41">Kulis B, Basu S, Dhillon I, Mooney R (2009) Semi-supervised graph clustering: a kernel approach. Mach Learn 74(1):122<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1007/s10994-008-5084-4"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Semi-supervised%20graph%20clustering%3A%20a%20kernel%20approach&amp;author=B.%20Kulis&amp;author=S.%20Basu&amp;author=I.%20Dhillon&amp;author=R.%20Mooney&amp;journal=Mach%20Learn&amp;volume=74&amp;issue=1&amp;pages=1-22&amp;publication_year=2009"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR42">Lin
 IJ, Kung SY (1997) Coding and comparison of dags as a novel neural 
structure with applications to on-line handwriting recognition. IEEE 
Trans Signal Process 45(11):27012708<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/78.650096"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Coding%20and%20comparison%20of%20dag%E2%80%99s%20as%20a%20novel%20neural%20structure%20with%20applications%20to%20on-line%20handwriting%20recognition&amp;author=IJ.%20Lin&amp;author=SY.%20Kung&amp;journal=IEEE%20Trans%20Signal%20Process&amp;volume=45&amp;issue=11&amp;pages=2701-2708&amp;publication_year=1997"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR43">Martignoni
 L, Christodorescu M, Jha S (2007) Omniunpack: fast, generic, and safe 
unpacking of malware. In: Twenty-third annual computer security 
applications conference (ACSAC) 2007, IEEE, pp 431441<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Martignoni%20L%2C%20Christodorescu%20M%2C%20Jha%20S%20%282007%29%20Omniunpack%3A%20fast%2C%20generic%2C%20and%20safe%20unpacking%20of%20malware.%20In%3A%20Twenty-third%20annual%20computer%20security%20applications%20conference%20%28ACSAC%29%202007%2C%20IEEE%2C%20pp%20431%E2%80%93441"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR44">Mishra
 N, Schreiber R, Stanton I, Tarjan RE (2007) Clustering social networks.
 In: Algorithms and models for the web-graph. Springer, Berlin, pp 5667<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Mishra%20N%2C%20Schreiber%20R%2C%20Stanton%20I%2C%20Tarjan%20RE%20%282007%29%20Clustering%20social%20networks.%20In%3A%20Algorithms%20and%20models%20for%20the%20web-graph.%20Springer%2C%20Berlin%2C%20pp%2056%E2%80%9367"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR45">Moser
 A, Kruegel C, Kirda E (2007a) Exploring multiple execution paths for 
malware analysis. In: IEEE symposium on security and privacy, IEEE, pp 
231245<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Moser%20A%2C%20Kruegel%20C%2C%20Kirda%20E%20%282007a%29%20Exploring%20multiple%20execution%20paths%20for%20malware%20analysis.%20In%3A%20IEEE%20symposium%20on%20security%20and%20privacy%2C%20IEEE%2C%20pp%20231%E2%80%93245"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR46">Moser
 A, Kruegel C, Kirda E, (2007b) Limits of static analysis for malware 
detection. In: Computer security applications conference, 2007. ACSAC 
2007. Twenty-third annual, IEEE, pp 421430<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Moser%20A%2C%20Kruegel%20C%2C%20Kirda%20E%2C%20%282007b%29%20Limits%20of%20static%20analysis%20for%20malware%20detection.%20In%3A%20Computer%20security%20applications%20conference%2C%202007.%20ACSAC%202007.%20Twenty-third%20annual%2C%20IEEE%2C%20pp%20421%E2%80%93430"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR47">Papapetrou
 P, Athitsos V, Kollios G, Gunopulos D (2009) Reference-based alignment 
in large sequence databases. Proc VLDB Endow 2(1):205216<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.14778/1687627.1687651"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Reference-based%20alignment%20in%20large%20sequence%20databases&amp;author=P.%20Papapetrou&amp;author=V.%20Athitsos&amp;author=G.%20Kollios&amp;author=D.%20Gunopulos&amp;journal=Proc%20VLDB%20Endow&amp;volume=2&amp;issue=1&amp;pages=205-216&amp;publication_year=2009"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR48">Ramon
 J, Grtner T (2003) Expressivity versus efficiency of graph kernels. 
First international workshop on mining graphs, trees and sequences, pp 
6574<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Ramon%20J%2C%20G%C3%A4rtner%20T%20%282003%29%20Expressivity%20versus%20efficiency%20of%20graph%20kernels.%20First%20international%20workshop%20on%20mining%20graphs%2C%20trees%20and%20sequences%2C%20pp%2065%E2%80%9374"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR49">Rieck
 K, Holz T, Willems C, Dssel P, Laskov P (2008) Learning and 
classification of malware behavior. In: Detection of intrusions and 
malware, and vulnerability assessment. Springer, Berlin, pp 108125<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Rieck%20K%2C%20Holz%20T%2C%20Willems%20C%2C%20D%C3%BCssel%20P%2C%20Laskov%20P%20%282008%29%20Learning%20and%20classification%20of%20malware%20behavior.%20In%3A%20Detection%20of%20intrusions%20and%20malware%2C%20and%20vulnerability%20assessment.%20Springer%2C%20Berlin%2C%20pp%20108%E2%80%93125"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR50">Riesen
 K, Bunke H (2009) Approximate graph edit distance computation by means 
of bipartite graph matching. Image Vis Comput 27(7):950959<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/j.imavis.2008.04.004"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Approximate%20graph%20edit%20distance%20computation%20by%20means%20of%20bipartite%20graph%20matching&amp;author=K.%20Riesen&amp;author=H.%20Bunke&amp;journal=Image%20Vis%20Comput&amp;volume=27&amp;issue=7&amp;pages=950-959&amp;publication_year=2009"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR51">Ryder BG (1979) Constructing the call graph of a program. IEEE Trans Softw Eng 3:216226<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/TSE.1979.234183"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=531940"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Constructing%20the%20call%20graph%20of%20a%20program&amp;author=BG.%20Ryder&amp;journal=IEEE%20Trans%20Softw%20Eng&amp;volume=3&amp;pages=216-226&amp;publication_year=1979"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR52">Schaeffer S (2007) Graph clustering. Comput Sci Rev 1(1):2764<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1016/j.cosrev.2007.05.001"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=2302548"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Graph%20clustering&amp;author=S.%20Schaeffer&amp;journal=Comput%20Sci%20Rev&amp;volume=1&amp;issue=1&amp;pages=27-64&amp;publication_year=2007"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR53">Schietgat
 L, Ramon J, Bruynooghe M (2013) A polynomial-time maximum common 
subgraph algorithm for outerplanar graphs and its application to 
chemoinformatics. Ann Math Artif Intell 69(4):343376<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1007/s10472-013-9335-0"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceZLBID"><a class="gtm-reference" data-reference-type="MATH" target="_blank" rel="noopener" href="http://www.emis.de/MATH-item?06308972"><span><span>MATH</span></span></a></span><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=3171418"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20polynomial-time%20maximum%20common%20subgraph%20algorithm%20for%20outerplanar%20graphs%20and%20its%20application%20to%20chemoinformatics&amp;author=L.%20Schietgat&amp;author=J.%20Ramon&amp;author=M.%20Bruynooghe&amp;journal=Ann%20Math%20Artif%20Intell&amp;volume=69&amp;issue=4&amp;pages=343-376&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR54">Seward
 HH (1954) Information sorting in the application of electronic digital 
computers to business operations. PhD thesis, Department of Electrical 
Engineering, Massachusetts Institute of Technology<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Seward%20HH%20%281954%29%20Information%20sorting%20in%20the%20application%20of%20electronic%20digital%20computers%20to%20business%20operations.%20PhD%20thesis%2C%20Department%20of%20Electrical%20Engineering%2C%20Massachusetts%20Institute%20of%20Technology"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR55">Shervashidze
 N, Schweitzer P, Van Leeuwen EJ, Mehlhorn K, Borgwardt KM (2011) 
WeisfeilerLehman graph kernels. J Mach Learn Res 12:25392561<span class="Occurrences"><span class="Occurrence OccurrenceZLBID"><a class="gtm-reference" data-reference-type="MATH" target="_blank" rel="noopener" href="http://www.emis.de/MATH-item?1280.68194"><span><span>MATH</span></span></a></span><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=2845672"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Weisfeiler%E2%80%93Lehman%20graph%20kernels&amp;author=N.%20Shervashidze&amp;author=P.%20Schweitzer&amp;author=EJ.%20Leeuwen&amp;author=K.%20Mehlhorn&amp;author=KM.%20Borgwardt&amp;journal=J%20Mach%20Learn%20Res&amp;volume=12&amp;pages=2539-2561&amp;publication_year=2011"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR56">Snaker, Qwerton, Jibz (2006) Peid. <span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.aldeid.com/wiki/PEiD"><span class="RefSource">http://www.aldeid.com/wiki/PEiD</span></a></span>
                <span class="Occurrences"></span></div></li><li class="Citation"><div class="CitationContent" id="CR57">Tarjan R, Van Leeuwen J (1984) Worst-case analysis of set union algorithms. J ACM 31(2):245281<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/62.2160"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceZLBID"><a class="gtm-reference" data-reference-type="MATH" target="_blank" rel="noopener" href="http://www.emis.de/MATH-item?0632.68043"><span><span>MATH</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Worst-case%20analysis%20of%20set%20union%20algorithms&amp;author=R.%20Tarjan&amp;author=J.%20Leeuwen&amp;journal=J%20ACM&amp;volume=31&amp;issue=2&amp;pages=245-281&amp;publication_year=1984"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR58">Tian
 Y, Patel J (2008) Tale: A tool for approximate large graph matching. 
In: Proceedings of the IEEE 24th international conference on data 
engineering, IEEE, pp 963972<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Tian%20Y%2C%20Patel%20J%20%282008%29%20Tale%3A%20A%20tool%20for%20approximate%20large%20graph%20matching.%20In%3A%20Proceedings%20of%20the%20IEEE%2024th%20international%20conference%20on%20data%20engineering%2C%20IEEE%2C%20pp%20963%E2%80%93972"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR59">Veeramani
 R, Rai N (2012) Windows api based malware detection and framework 
analysis. In: International conference on networks and cyber security, p
 25<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Veeramani%20R%2C%20Rai%20N%20%282012%29%20Windows%20api%20based%20malware%20detection%20and%20framework%20analysis.%20In%3A%20International%20conference%20on%20networks%20and%20cyber%20security%2C%20p%2025"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR60">Venkateswaran
 J, Lachwani D, Kahveci T, Jermaine C (2006) Reference-based indexing of
 sequence databases. In: Proceedings of the 32nd international 
conference on very large data bases, VLDB Endowment, pp 906917<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Venkateswaran%20J%2C%20Lachwani%20D%2C%20Kahveci%20T%2C%20Jermaine%20C%20%282006%29%20Reference-based%20indexing%20of%20sequence%20databases.%20In%3A%20Proceedings%20of%20the%2032nd%20international%20conference%20on%20very%20large%20data%20bases%2C%20VLDB%20Endowment%2C%20pp%20906%E2%80%93917"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR61">Vishwanathan S, Schraudolph NN, Kondor R, Borgwardt KM (2010) Graph kernels. J Mach Learn Res 11:12011242<span class="Occurrences"><span class="Occurrence OccurrenceZLBID"><a class="gtm-reference" data-reference-type="MATH" target="_blank" rel="noopener" href="http://www.emis.de/MATH-item?1242.05112"><span><span>MATH</span></span></a></span><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=2645450"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Graph%20kernels&amp;author=S.%20Vishwanathan&amp;author=NN.%20Schraudolph&amp;author=R.%20Kondor&amp;author=KM.%20Borgwardt&amp;journal=J%20Mach%20Learn%20Res&amp;volume=11&amp;pages=1201-1242&amp;publication_year=2010"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR62">Warshall S (1962) A theorem on Boolean matrices. J ACM 9(1):1112<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/321105.321107"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceZLBID"><a class="gtm-reference" data-reference-type="MATH" target="_blank" rel="noopener" href="http://www.emis.de/MATH-item?0118.33104"><span><span>MATH</span></span></a></span><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=149688"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20theorem%20on%20Boolean%20matrices&amp;author=S.%20Warshall&amp;journal=J%20ACM&amp;volume=9&amp;issue=1&amp;pages=11-12&amp;publication_year=1962"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR63">Willems
 C, Holz T, Freiling F (2007) Toward automated dynamic malware analysis 
using cwsandbox. Proceedings of the 28th IEEE symposium on security and 
privacy, vol 5(2), pp 3239<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Willems%20C%2C%20Holz%20T%2C%20Freiling%20F%20%282007%29%20Toward%20automated%20dynamic%20malware%20analysis%20using%20cwsandbox.%20Proceedings%20of%20the%2028th%20IEEE%20symposium%20on%20security%20and%20privacy%2C%20vol%205%282%29%2C%20pp%2032%E2%80%9339"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR64">Williams
 D, Huan J, Wang W (2007) Graph database indexing using structured graph
 decomposition. In: Proceedings of the IEEE 23rd international 
conference on data engineering, IEEE, pp 976985<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Williams%20D%2C%20Huan%20J%2C%20Wang%20W%20%282007%29%20Graph%20database%20indexing%20using%20structured%20graph%20decomposition.%20In%3A%20Proceedings%20of%20the%20IEEE%2023rd%20international%20conference%20on%20data%20engineering%2C%20IEEE%2C%20pp%20976%E2%80%93985"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR65">Xu
 JY, Sung AH, Chavez P, Mukkamala S (2004) Polymorphic malicious 
executable scanner by api sequence analysis. In: Fourth international 
conference on hybrid intelligent systems, HIS04., IEEE, pp 378383<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Xu%20JY%2C%20Sung%20AH%2C%20Chavez%20P%2C%20Mukkamala%20S%20%282004%29%20Polymorphic%20malicious%20executable%20scanner%20by%20api%20sequence%20analysis.%20In%3A%20Fourth%20international%20conference%20on%20hybrid%20intelligent%20systems%2C%20HIS%E2%80%9904.%2C%20IEEE%2C%20pp%20378%E2%80%93383"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR66">Xu
 M, Wu L, Qi S, Xu J, Zhang H, Ren Y, Zheng N (2013) A similarity metric
 method of obfuscated malware using function-call graph. J Comput Virol 
Hacking Tech 9(1):3547<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1007/s11416-012-0175-y"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20similarity%20metric%20method%20of%20obfuscated%20malware%20using%20function-call%20graph&amp;author=M.%20Xu&amp;author=L.%20Wu&amp;author=S.%20Qi&amp;author=J.%20Xu&amp;author=H.%20Zhang&amp;author=Y.%20Ren&amp;author=N.%20Zheng&amp;journal=J%20Comput%20Virol%20Hacking%20Tech&amp;volume=9&amp;issue=1&amp;pages=35-47&amp;publication_year=2013"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR67">Yan
 X, Yu P, Han J (2005) Substructure similarity search in graph 
databases. In: Proceedings of the 2005 ACM SIGMOD international 
conference on management of data, ACM, pp 766777<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Yan%20X%2C%20Yu%20P%2C%20Han%20J%20%282005%29%20Substructure%20similarity%20search%20in%20graph%20databases.%20In%3A%20Proceedings%20of%20the%202005%20ACM%20SIGMOD%20international%20conference%20on%20management%20of%20data%2C%20ACM%2C%20pp%20766%E2%80%93777"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR68">Zeng Z, Tung A, Wang J, Feng J, Zhou L (2009) Comparing stars: on approximating graph edit distance. Proc VLDB Endow 2(1):2536<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.14778/1687627.1687631"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Comparing%20stars%3A%20on%20approximating%20graph%20edit%20distance&amp;author=Z.%20Zeng&amp;author=A.%20Tung&amp;author=J.%20Wang&amp;author=J.%20Feng&amp;author=L.%20Zhou&amp;journal=Proc%20VLDB%20Endow&amp;volume=2&amp;issue=1&amp;pages=25-36&amp;publication_year=2009"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR69">Zhao P, Yu J, Yu P (2007) Graph indexing: tree+ delta<span class="InlineEquation" id="IEq316">\(\le \)</span> graph. In: Proceedings of the 33rd international conference on very large data bases, VLDB Endowment, pp 938949<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Zhao%20P%2C%20Yu%20J%2C%20Yu%20P%20%282007%29%20Graph%20indexing%3A%20tree%2B%20delta%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%24%24%5Cle%20%24%24%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%E2%89%A4%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20graph.%20In%3A%20Proceedings%20of%20the%2033rd%20international%20conference%20on%20very%20large%20data%20bases%2C%20VLDB%20Endowment%2C%20pp%20938%E2%80%93949"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR70">Zhou Y, Cheng H, Yu JX (2009) Graph clustering based on structural/attribute similarities. Proc VLDB Endow 2(1):718729<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.14778/1687627.1687709"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Graph%20clustering%20based%20on%20structural%2Fattribute%20similarities&amp;author=Y.%20Zhou&amp;author=H.%20Cheng&amp;author=JX.%20Yu&amp;journal=Proc%20VLDB%20Endow&amp;volume=2&amp;issue=1&amp;pages=718-729&amp;publication_year=2009"><span><span>Google Scholar</span></span></a></span></span></div></li></ol></div></section><section class="Section1 RenderAsSection1"><h2 class="Heading" id="copyrightInformation" data-role="collapsible-handle" tabindex="-1">Copyright information<span class="section-icon"></span></h2><div class="ArticleCopyright content"><div class="ArticleCopyright">&nbsp;The Author(s)&nbsp;2014</div></div></section></div>
                        </article>
                        <aside class="section section--collapsible" id="AboutThisContent">
    <h2 class="section__heading" id="aboutcontent" data-role="collapsible-handle" tabindex="-1">About this article<span class="section-icon"></span></h2>
    <div class="section__content bibliographic-information">
                <div id="crossMark" class="crossmark">
            <a data-crossmark="10.1007%2Fs10618-014-0367-9" class="gtm-crossmark" target="_blank" rel="noopener" href="https://crossmark.crossref.org/dialog/?doi=10.1007%2Fs10618-014-0367-9" title="Verify currency and authenticity via CrossMark">
                <span class="u-screenreader-only">CrossMark</span>
                <svg class="CrossMark" id="crossmark-icon" width="57" height="81">
                    <image width="57" height="81" alt="Verify currency and authenticity via CrossMark" src="/springerlink-static/439917616/images/png/crossmark.png" xlink="http://www.w3.org/1999/xlink" xlink:href="/springerlink-static/439917616/images/svg/crossmark.svg"></image>
                </svg>
            </a>
        </div>

        <div class="crossmark__adjacent">
            <dl class="citation-info u-highlight-target u-mb-16" id="citeas" tabindex="-1">
    <dt class="test-cite-heading">
        Cite this article as:
    </dt>
    <dd id="citethis-text">Kostakis, O. Data Min Knowl Disc (2014) 28: 1554. https://doi.org/10.1007/s10618-014-0367-9</dd>
</dl>
                <ul class="bibliographic-information__list bibliographic-information__list--inline">
        <li class="bibliographic-information__item">
            <span class="bibliographic-information__title"><abbr title="Digital Object Identifier">DOI</abbr></span>
            <span class="bibliographic-information__value u-overflow-wrap" id="doi-url">https://doi.org/10.1007/s10618-014-0367-9</span>
        </li>
            <li class="bibliographic-information__item">
                <span class="bibliographic-information__title">Publisher Name</span>
                <span class="bibliographic-information__value" id="publisher-name">Springer US</span>
            </li>
            <li class="bibliographic-information__item">
                <span class="bibliographic-information__title">Print ISSN</span>
                <span class="bibliographic-information__value" id="print-issn">1384-5810</span>
            </li>
            <li class="bibliographic-information__item ">
                <span class="bibliographic-information__title">Online ISSN</span>
                <span class="bibliographic-information__value" id="electronic-issn">1573-756X</span>
            </li>

        
    </ul>

            <ul class="bibliographic-information__list">
        <li class="bibliographic-information__item">
            <a id="about-journal" class="bibliographic-information__misc-links gtm-about-this" title="Visit Springer.com for information about this article's journal" href="https://www.springer.com/journal/10618/about">About this journal</a>
        </li>
        <li class="bibliographic-information__item">
            <a id="reprintsandpermissions-link" target="_blank" rel="noopener" href="https://s100.copyright.com/AppDispatchServlet?publisherName=SpringerNature&amp;orderBeanReset=true&amp;orderSource=SpringerLink&amp;copyright=The+Author%28s%29&amp;author=Orestis+Kostakis&amp;issueNum=5&amp;contentID=10.1007%2Fs10618-014-0367-9&amp;openAccess=false&amp;endPage=1585&amp;publicationDate=2014&amp;startPage=1554&amp;volumeNum=28&amp;title=Classy%3A+fast+clustering+streams+of+call-graphs&amp;imprint=The+Author%28s%29&amp;publication=1384-5810" title="Visit RightsLink for information about reusing this article">Reprints and Permissions</a>
        </li>
</ul>



        </div>
      
      
          
    </div>
</aside>

                        <div class="section section--collapsible uptodate-recommendations gtm-recommendations">
    <h2 class="uptodate-recommendations__title section__heading gtm-recommendations__title" id="uptodaterecommendations" data-role="collapsible-handle" tabindex="-1">Personalised recommendations<span class="section-icon"></span></h2>
    <div class="section__content">
        <div class="uptodate-recommendations__container">
             <iframe style="width: 100%; height: 411px; bottom: 0px; right: 0px; z-index: 100000;" scrolling="no" src="https://recommended.springernature.com/v2197/generated/client.html?source=https://link.springer.com" name="uptodate-client" id="uptodate-client" title="Personalised recommendations" allowtransparency="true" frameborder="0"></iframe>
        </div>
    </div>
</div>
                                <div id="doubleclick-native-ad" data-component="SpringerLink.GoogleAds" data-namespace="native"></div>

                                    <div class="sticky-banner u-interface u-js-screenreader-only" aria-hidden="true" data-component="SpringerLink.StickyBanner" data-namespace="hasButton">
                <div class="sticky-banner__container"><span class="sticky-banner__title sticky-banner__title--short u-overflow-ellipsis" title="Classy: fast clustering streams of call-graphs">Classy: fast clustering streams of call-graphs</span>
                        <div class="citations c-button-dropdown" data-component="SV.Dropdown" data-namespace="citationsSticky" aria-label="button with dropdown options">
        <button id="button-Dropdown-citationsSticky-dropdown" type="button" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown-citationsSticky-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">
    <span>Cite</span>
    <span class="hide-text-small">article</span>
</span><span class="c-button-dropdown__icon"></span></button>
<div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown-citationsSticky-dropdown"><ul class="citations__content" data-role="button-dropdown__content">
    <li>
        <a href="#citeas" class="gtm-cite-dropdown" tabindex="-1">How to cite?</a>
    </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-014-0367-9?format=refman&amp;flavour=citation" title="Download this article's citation as a .RIS file" class="gtm-export-citation" data-gtmlabel="RIS" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="RIS">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .RIS
                </span>
                <span class="citations__types">
                        <span>
                            Papers
                        </span>
                        <span>
                            Reference Manager
                        </span>
                        <span>
                            RefWorks
                        </span>
                        <span>
                            Zotero
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-014-0367-9?format=endnote&amp;flavour=citation" title="Download this article's citation as a .ENW file" class="gtm-export-citation" data-gtmlabel="ENW" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="ENW">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .ENW
                </span>
                <span class="citations__types">
                        <span>
                            EndNote
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-014-0367-9?format=bibtex&amp;flavour=citation" title="Download this article's citation as a .BIB file" class="gtm-export-citation" data-gtmlabel="BIB" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="BIB">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .BIB
                </span>
                <span class="citations__types">
                        <span>
                            BibTeX
                        </span>
                        <span>
                            JabRef
                        </span>
                        <span>
                            Mendeley
                        </span>
                </span>
            </a>
        </li>
</ul></div>
    </div>

                            <div>
        <a class="c-button share-this gtm-shareby-sharelink-link test-shareby-sharelink-link" data-test="shareable-link" target="_blank" rel="noopener" href="https://link.springer.com/sharelink/10.1007/s10618-014-0367-9">
            <span>Share</span>
            <span class="hide-text-small">article</span>
        </a>
    </div>




                                    <div>
            <a href="https://link.springer.com/content/pdf/10.1007%2Fs10618-014-0367-9.pdf" target="_blank" class="c-button c-button--blue c-button__icon-right gtm-pdf-link" title="Download this article in PDF format" rel="noopener">
                <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" version="1.1"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g fill="#fff"><g transform="translate(12.000000, 5.000000)"><path d="M7 7.3L7 1C7 0.4 6.6 0 6 0 5.4 0 5 0.4 5 1L5 7.3 3.5 5.7C3.1 5.3 2.5 5.3 2.1 5.7L2.1 5.7C1.7 6.1 1.7 6.7 2.1 7.1L5.3 10.3C5.7 10.7 6.3 10.7 6.7 10.3L9.9 7.1C10.3 6.7 10.3 6.1 9.9 5.7L9.9 5.7C9.5 5.3 8.9 5.3 8.5 5.7L7 7.3 7 7.3ZM0 13C0 12.4 0.5 12 1 12L11 12C11.6 12 12 12.4 12 13 12 13.6 11.5 14 11 14L1 14C0.4 14 0 13.6 0 13L0 13Z"></path></g></g></g></svg>
                <span class="hide-text-small">Download</span>
                <span>PDF</span>
            </a>
        </div>

                </div>
            </div>




                    </div>
                    <aside class="main-sidebar-right u-interface">
                        <div data-role="sticky-wrapper">
                            <div class="main-sidebar-right__content u-composite-layer" data-component="SpringerLink.StickySidebar">
                                <div class="article-actions" id="article-actions">
                                    <h2 class="u-screenreader-only">Actions</h2>


                                    <div class="u-js-hide u-js-show-two-col">
                                        

                                                <div class="download-article test-pdf-link">
                                                            <div>
            <a href="https://link.springer.com/content/pdf/10.1007%2Fs10618-014-0367-9.pdf" target="_blank" class="c-button c-button--blue c-button__icon-right gtm-pdf-link" title="Download this article in PDF format" rel="noopener">
                <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" version="1.1"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g fill="#fff"><g transform="translate(12.000000, 5.000000)"><path d="M7 7.3L7 1C7 0.4 6.6 0 6 0 5.4 0 5 0.4 5 1L5 7.3 3.5 5.7C3.1 5.3 2.5 5.3 2.1 5.7L2.1 5.7C1.7 6.1 1.7 6.7 2.1 7.1L5.3 10.3C5.7 10.7 6.3 10.7 6.7 10.3L9.9 7.1C10.3 6.7 10.3 6.1 9.9 5.7L9.9 5.7C9.5 5.3 8.9 5.3 8.5 5.7L7 7.3 7 7.3ZM0 13C0 12.4 0.5 12 1 12L11 12C11.6 12 12 12.4 12 13 12 13.6 11.5 14 11 14L1 14C0.4 14 0 13.6 0 13L0 13Z"></path></g></g></g></svg>
                <span class="hide-text-small">Download</span>
                <span>PDF</span>
            </a>
        </div>

                                                </div>


                                            <div class="citations c-button-dropdown" data-component="SV.Dropdown" data-namespace="citations" aria-label="button with dropdown options">
        <button id="button-Dropdown-citations-dropdown" type="button" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown-citations-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">
    <span>Cite</span>
    <span class="hide-text-small">article</span>
</span><span class="c-button-dropdown__icon"></span></button>
<div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown-citations-dropdown"><ul class="citations__content" data-role="button-dropdown__content">
    <li>
        <a href="#citeas" class="gtm-cite-dropdown" tabindex="-1">How to cite?</a>
    </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-014-0367-9?format=refman&amp;flavour=citation" title="Download this article's citation as a .RIS file" class="gtm-export-citation" data-gtmlabel="RIS" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="RIS">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .RIS
                </span>
                <span class="citations__types">
                        <span>
                            Papers
                        </span>
                        <span>
                            Reference Manager
                        </span>
                        <span>
                            RefWorks
                        </span>
                        <span>
                            Zotero
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-014-0367-9?format=endnote&amp;flavour=citation" title="Download this article's citation as a .ENW file" class="gtm-export-citation" data-gtmlabel="ENW" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="ENW">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .ENW
                </span>
                <span class="citations__types">
                        <span>
                            EndNote
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-014-0367-9?format=bibtex&amp;flavour=citation" title="Download this article's citation as a .BIB file" class="gtm-export-citation" data-gtmlabel="BIB" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="BIB">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .BIB
                </span>
                <span class="citations__types">
                        <span>
                            BibTeX
                        </span>
                        <span>
                            JabRef
                        </span>
                        <span>
                            Mendeley
                        </span>
                </span>
            </a>
        </li>
</ul></div>
    </div>

                                                <div>
        <a class="c-button share-this gtm-shareby-sharelink-link test-shareby-sharelink-link" data-test="shareable-link" target="_blank" rel="noopener" href="https://link.springer.com/sharelink/10.1007/s10618-014-0367-9">
            <span>Share</span>
            <span class="hide-text-small">article</span>
        </a>
    </div>




                                    </div>
                                </div>
                                <nav class="toc" aria-label="article contents">
    <h2 class="u-h4 u-screenreader-only">Table of contents</h2>
    <ul id="article-contents" class="article-contents" role="menu" tabindex="-1">
            <li role="menuitem">
                <a title="Article" href="#enumeration"><span class="u-overflow-ellipsis">Article</span></a>
            </li>
            <li role="menuitem">
                <a title="Abstract" href="#Abs1"><span class="u-overflow-ellipsis">Abstract</span></a>
            </li>
            <li role="menuitem">
                <a title="1 Introduction" href="#Sec1"><span class="u-overflow-ellipsis">1 Introduction</span></a>
            </li>
            <li role="menuitem">
                <a title="2 Related work" href="#Sec2"><span class="u-overflow-ellipsis">2 Related work</span></a>
            </li>
            <li role="menuitem">
                <a title="3 Background" href="#Sec3"><span class="u-overflow-ellipsis">3 Background</span></a>
            </li>
            <li role="menuitem">
                <a title="4 Classy: architectural overview" href="#Sec7"><span class="u-overflow-ellipsis">4 Classy: architectural overview</span></a>
            </li>
            <li role="menuitem">
                <a title="5 Graph comparison" href="#Sec10"><span class="u-overflow-ellipsis">5 Graph comparison</span></a>
            </li>
            <li role="menuitem">
                <a title="6 Lower bound of the graph edit distance" href="#Sec11"><span class="u-overflow-ellipsis">6 Lower bound of the graph edit distance</span></a>
            </li>
            <li role="menuitem">
                <a title="7 Clustering" href="#Sec12"><span class="u-overflow-ellipsis">7 Clustering</span></a>
            </li>
            <li role="menuitem">
                <a title="8 Experiments" href="#Sec15"><span class="u-overflow-ellipsis">8 Experiments</span></a>
            </li>
            <li role="menuitem">
                <a title="9 Conclusions" href="#Sec23"><span class="u-overflow-ellipsis">9 Conclusions</span></a>
            </li>
            <li role="menuitem">
                <a title="Notes" href="#Notes"><span class="u-overflow-ellipsis">Notes</span></a>
            </li>
            <li role="menuitem">
                <a title="References" href="#Bib1"><span class="u-overflow-ellipsis">References</span></a>
            </li>
            <li role="menuitem">
                <a title="Copyright information" href="#copyrightInformation"><span class="u-overflow-ellipsis">Copyright information</span></a>
            </li>
            
            <li role="menuitem">
                <a title="About this article" href="#aboutcontent"><span class="u-overflow-ellipsis">About this article</span></a>
            </li>
    </ul>
</nav>

                            </div>
                                <div class="skyscraper-ad u-hide" data-component="SpringerLink.GoogleAds" data-namespace="skyscraper">
        <div class="skyscraper-ad__wrapper">
            <p class="skyscraper-ad__label">Advertisement</p>
            <button class="skyscraper-ad__hide" title="Hide this advertisement">Hide</button>
            <div id="doubleclick-ad" class="skyscraper-ad__ad"></div>
        </div>
    </div>

                        </div>
                    </aside>
                </div>
            </main>
                <footer class="footer u-interface">
        <div class="footer__aside-wrapper">
            <div class="footer__content">
                <div class="footer__aside">
                    <p class="footer__strapline">Over 10 million scientific documents at your fingertips</p>
                                <div class="footer__edition c-button-dropdown c-button-dropdown--ghost" data-component="SpringerLink.EditionSwitcher" aria-label="button with dropdown options">
                                    <button id="button-EditionSwitcher-1101011910-dropdown" type="button" title="Switch between Academic &amp; Corporate Edition" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="EditionSwitcher-1101011910-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">Academic Edition</span><span class="c-button-dropdown__icon"></span></button>
                                    <div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="EditionSwitcher-1101011910-dropdown"><ul data-role="button-dropdown__content">
                                        <li class="selected"><a href="https://link.springer.com/siteEdition/link?previousUrl=/article/10.1007/s10618-014-0367-9&amp;id=siteedition-academic-link" id="siteedition-academic-link">Academic Edition</a></li>
                                        <li><a href="https://link.springer.com/siteEdition/rd?previousUrl=/article/10.1007/s10618-014-0367-9&amp;id=siteedition-corporate-link" id="siteedition-corporate-link" tabindex="-1">Corporate Edition</a></li>
                                    </ul></div>
                                </div>
                </div>
            </div>
        </div>
        <div class="footer__content">
            <ul class="footer__nav">
                <li>
                    <a href="https://link.springer.com/">Home</a>
                </li>
                <li>
                    <a href="https://link.springer.com/impressum">Impressum</a>
                </li>
                <li>
                    <a href="https://link.springer.com/termsandconditions">Legal information</a>
                </li>
                <li>
                    <a href="https://link.springer.com/privacystatement">Privacy statement</a>
                </li>
                <li>
                    <a href="https://link.springer.com/cookiepolicy">How we use cookies</a>
                </li>
                <li>
                    <a href="https://link.springer.com/accessibility" class="gtm-footer-accessibility">Accessibility</a>
                </li>
                <li>
                    <a id="contactus-footer-link" href="https://link.springer.com/contactus">Contact us</a>
                </li>
            </ul>
            <a class="parent-logo" target="_blank" rel="noopener" href="https://www.springernature.com/" title="Go to Springer Nature">
                <span class="u-screenreader-only">Springer Nature</span>
                <svg width="125" height="12" focusable="false" aria-hidden="true">
                    <image width="125" height="12" src="/springerlink-static/439917616/images/png/springernature.png" xlink="http://www.w3.org/1999/xlink" xlink:href="/springerlink-static/439917616/images/svg/springernature.svg">
                    </image>
                </svg>
            </a>

            <p class="footer__copyright"> 2017 Springer Nature Switzerland AG. Part of <a target="_blank" rel="noopener" href="https://www.springernature.com/">Springer Nature</a>.</p>
                <p class="footer__user-access-info">
                    <span>Not logged in</span>
                    <span>CAPES MEC (3000197460) - Universidade Tecnologica Federal do Parana (3000201946)</span>
                    <span>200.17.97.45</span>
                </p>
        </div>
    </footer>

        </div>
          <script type="text/javascript">
    (function() {
        document.addEventListener('readystatechange', function () {
            if (document.readyState === 'complete') {
                var linkEl = document.querySelector('.js-ctm');
                if (window.matchMedia && window.matchMedia(linkEl.media).matches) {
                    var scriptMathJax = document.createElement('script');
                    scriptMathJax.async = true;
                    scriptMathJax.src = '/springerlink-static/439917616/js/mathJax.js';
                    var s0 = document.getElementsByTagName('script')[0];
                    s0.parentNode.insertBefore(scriptMathJax, s0);
                }
            }
        });
    })();
</script>

<script type="text/javascript">

    (function() {
        var linkEl = document.querySelector('.js-ctm');
        var scriptsList = [];
        var polyfillFeatures = '';

        window.SpringerLink = window.SpringerLink || {};
        window.SpringerLink.staticLocation = '/springerlink-static/439917616';

        if (window.matchMedia && window.matchMedia(linkEl.media).matches) {
            (function(h){h.className = h.className.replace('no-js', 'js')})(document.documentElement);
            
            polyfillFeatures = 'default,fetch,Promise,Object.setPrototypeOf,Object.entries,Number.isInteger,MutationObserver,startsWith,Array.prototype.includes,Array.from,IntersectionObserver';
            
            scriptsList = [
                'https://cdn.polyfill.io/v2/polyfill.js?features=' + polyfillFeatures,
                window.SpringerLink.staticLocation + '/js/main.js'
            ];

            scriptsList.forEach(function(script) {
                var tag = document.createElement('script');
                
                tag.async = false;
                tag.src = script;

                document.body.appendChild(tag);
            });
        }
    })();
</script><script src="polyfill.js"></script><script src="main.js"></script>

<script type="text/javascript" class="optanon-category-2">

   function viewport() {
        if (document.documentElement.clientWidth < 620) {
            size = 'small';
        }
        else if(document.documentElement.clientWidth < 1075 ) {
            size = 'medium';
        }
        else {
            size = 'wide';
        }
        return size;
    };

    function reportForMouseEvents(linkCssSelectors, nolardUrl, experiment, abgroup) {
        var counter = 0;
        linkCssSelectors.forEach(function(cssSelector) {
            $('body').delegate(cssSelector, 'click', function() {
                if(counter == 0) {
                    reportConversion(nolardUrl, experiment, abgroup);
                    counter++;
                }
            });
        });
    };

    function reportConversion(nolardUrl, experiment, abgroup) {
        $.ajax({ url: nolardUrl + '/convert/' + experiment + '/' + abgroup });
    };

    function reportForMouseEvent(linkCssSelector, nolardUrl, experiment, abgroup) {
        var counter = 0;
        var elem = document.querySelector(linkCssSelector)

        if (elem.addEventListener) {
            elem.addEventListener("click", function(e) {
                if(counter == 0) {
                    reportConversion(nolardUrl, experiment, abgroup);
                    counter++;
                }
            });
        } else  {
            elem.attachEvent("click", function(e) {
                if(counter == 0) {
                    reportConversion(nolardUrl, experiment, abgroup);
                    counter++;
                }
            });
        }
    };

    function reportParticipation(nolardUrl, experiment, abgroup, participations) {
        if (participations)
            participations.push(experiment + '/' + abgroup)

        var xhr = new XMLHttpRequest()
        xhr.open('GET', nolardUrl + '/participate/' + experiment + '/' + abgroup);
        xhr.send();
    };

</script>
    <script type="text/javascript" id="googletag-push">
        
            var adSlot = '270604982/springerlink/10618/article';
        

        var definedSlots = [
                {slot: [728, 90], containerName: 'doubleclick-leaderboard-ad'},
                {slot: [160, 600], containerName: 'doubleclick-ad'},
            {slot: [2, 2], containerName: 'doubleclick-native-ad'}
        ];
    </script>


        
        <span id="chat-widget" class="u-hide"></span>
                    <noscript>
                <img aria-hidden="true" role="presentation" src="https://ssl-springer.met.vgwort.de/na/pw-vgzm.415900-10.1007-s10618-014-0367-9" width='1' height='1' alt='' />
            </noscript>

    

<div style="display: none; visibility: hidden;"><script id="gpt-control-script" src="gpt.js"></script>

<script id="gpt-control-setup">var googletag=googletag||{};googletag.cmd=googletag.cmd||[];</script></div><script type="text/javascript" id="gtm-in-page-section-view">(function(d,z){function n(){for(var a in dataLayer)if(dataLayer.hasOwnProperty(a)&&dataLayer[a]["Event Category"])return dataLayer[a]["Event Category"];return"warning: no event category"}function v(a,b,c,h){"undefined"!==typeof dataLayer?dataLayer.push({event:"Scroll To Section",eventCategory:n(),eventAction:a,eventLabel:b,eventValue:1,eventNonInteraction:h}):("undefined"!==typeof ga&&ga("send","event",n(),a,b,1,{nonInteraction:1}),"undefined"!==typeof _gaq&&_gaq.push(["_trackEvent",n(),a,b,1,!0]))}
function k(b,g,c,h){if(-1===a[c].cache.indexOf(c+"-"+h)&&document.querySelectorAll(b).length){var f="viewed"===h?document.querySelectorAll(b)[0].getBoundingClientRect().height:0,x=document.querySelectorAll(b)[0].getBoundingClientRect().top+d.pageYOffset;g>=x+f&&(v(p(c)+" "+p(h),b,c,!1),a[c].cache.push(c+"-"+h),q++)}}function r(a){a=document.querySelectorAll(a);return a.length?a[0].offsetHeight||0:0}function m(b,g,c){var f=d.pageYOffset+d.innerHeight;if(q>=w)"handleClick"===g&&d.removeEventListener(b,
t),"throttle"===g&&d.removeEventListener(b,u);else if(c)setTimeout(function(){0<r(a[c].content)&&(a[c].reached&&k(a[c].content,f,c,"reached"),a[c].viewed&&k(a[c].content,f,c,"viewed"))},10);else for(var e in a)a.hasOwnProperty(e)&&0<r(a[e].content)&&(a[e].reached&&k(a[e].content,f,e,"reached"),a[e].viewed&&k(a[e].content,f,e,"viewed"))}function y(b){b=document.querySelectorAll(a[b].content).length;return 0<b?!0:!1}function p(a){return a.charAt(0).toUpperCase()+a.slice(1)}function u(a,b){var c,f,e,
g=null,d=0,k=function(){d=new Date;g=null;e=a.apply(c,f)};return function(){var h=new Date;d||(d=h);var l=b-(h-d);c=this;f=arguments;0>=l?(clearTimeout(g),g=null,d=h,e=a.apply(c,f)):g||(g=setTimeout(k,l));return e}}function t(a,b){return function(c){a(c,b)}}var w=0,q=0,a={recommendations:{content:".gtm-recommendations iframe",clickable:".gtm-recommendations .gtm-recommendations__title",exists:!1,reached:!1,viewed:!0,size:0,cache:[]},"abstract":{content:"Section.Abstract",clickable:null,exists:!0,
reached:!1,viewed:!0,size:0,cache:[]},references:{content:".Bibliography \x3e .content, [id^\x3dBib1] \x3e .content",clickable:".Bibliography \x3e .Heading, [id^\x3dBib1] \x3e .Heading",exists:!0,reached:!0,viewed:!0,size:0,cache:[]},about:{content:"#AboutThisContent",clickable:"#AboutThisContent \x3e #aboutcontent",exists:!1,reached:!1,viewed:!0,size:0,cache:[]}};d.addEventListener("scroll",u(function(){m("scroll","throttle")},500));d.addEventListener("orientationchange",u(function(){m("orientationchange",
"throttle")},500));for(var b in a)if(a.hasOwnProperty(b)&&(a[b].exists&&a[b].size++,a[b].reached&&a[b].size++,a[b].viewed&&a[b].size++,w+=a[b].size,a[b].exists&&y(b)||!a[b].exists)){a[b].exists&&-1===a[b].cache.indexOf(b+"-exists")&&(v(p(b)+" Exists",a[b].content,b,!0),a[b].cache.push(b+"-exists"),q++);if(0<r(a[b].content)){var l=d.pageYOffset+d.innerHeight;a[b].reached&&k(a[b].content,l,b,"reached");a[b].viewed&&k(a[b].content,l,b,"viewed")}a[b].clickable&&(l=document.querySelectorAll(a[b].clickable)[0])&&
(l.addEventListener("click",t(function(a,b){m("click","handleClick",b)},b)),l.addEventListener("click",t(function(a,b){13==a.keyCode&&m("click","handleClick",b)},b)))}})(window);</script><script type="text/javascript" id="gtm-recommendations-script" src="entry-point"></script><script type="text/javascript" id="gtm-polar" src="polar.js"></script><script type="text/javascript" id="gtm-krux-control">window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);(function(){var a=document.createElement("script");a.type="text/javascript";a.async=!0;a.src=("https:"===location.protocol?"https:":"http:")+"//cdn.krxd.net/controltag/KDqyaFZ_.js";var b=document.getElementsByTagName("script")[0];b.parentNode.insertBefore(a,b)})();</script><script type="text/javascript" id="">window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);(function(){function c(b){b="kxmacmillan_"+b;try{var a=window.localStorage}catch(d){a=null}return a?a[b]||"":navigator.cookieEnabled?(a=document.cookie.match(b+"\x3d([^;]*)"))&&unescape(a[1])||"":""}Krux.user=c("user");Krux.segments=c("segs")?c("segs").split(","):[]})();</script><script type="text/javascript" id="gtm-krux-consent">var allowed=google_tag_manager["GTM-WCF9Z9"].macro(29);allowed=!0===allowed?1:0;Krux("consent:set",{dc:allowed,al:allowed,tg:allowed,cd:!1,sh:!1,re:!1},function(a,b){a?console.error(a):console.log("consent flags set ",b)});</script><script type="text/javascript" id="gtm-gpt-setup">window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);
googletag.cmd.push(function(){googletag.pubads().setTargeting("doi",google_tag_manager["GTM-WCF9Z9"].macro(30));googletag.pubads().setTargeting("kwrd",google_tag_manager["GTM-WCF9Z9"].macro(31));googletag.pubads().setTargeting("pmc",google_tag_manager["GTM-WCF9Z9"].macro(32));googletag.pubads().setTargeting("BPID",google_tag_manager["GTM-WCF9Z9"].macro(33));googletag.pubads().setTargeting("edition",google_tag_manager["GTM-WCF9Z9"].macro(34));googletag.pubads().setTargeting("sucode",google_tag_manager["GTM-WCF9Z9"].macro(35));googletag.pubads().setTargeting("eissn",google_tag_manager["GTM-WCF9Z9"].macro(36));googletag.pubads().setTargeting("pissn",
google_tag_manager["GTM-WCF9Z9"].macro(37));googletag.pubads().setTargeting("eisbn",google_tag_manager["GTM-WCF9Z9"].macro(38));googletag.pubads().setTargeting("pisbn",google_tag_manager["GTM-WCF9Z9"].macro(39));googletag.pubads().setTargeting("logged",google_tag_manager["GTM-WCF9Z9"].macro(40));googletag.pubads().setTargeting("EventCategory",google_tag_manager["GTM-WCF9Z9"].macro(41));googletag.pubads().setTargeting("ksg",Krux.segments);googletag.pubads().setTargeting("kuid",Krux.uid);googletag.pubads().setRequestNonPersonalizedAds(google_tag_manager["GTM-WCF9Z9"].macro(45)?0:1);googletag.pubads().disableInitialLoad();
googletag.enableServices()});</script><div id="optanon" class="modern"><div id="optanon-popup-bg"></div><div id="optanon-popup-wrapper" role="dialog" aria-modal="true" tabindex="-1" lang="en-GB"><div id="optanon-popup-top"><a href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Close Button');" class="optanon-close-link optanon-close optanon-close-ui" title="Close Preference Centre"><div id="optanon-close" style="background: url(https://optanon.blob.core.windows.net/skins/default_flat_bottom_two_button_white/v2/images/optanon-pop-up-close.png);width:34px;height:34px;"></div></a></div><div id="optanon-popup-body"><div id="optanon-popup-body-left"><div id="optanon-popup-body-left-shading"></div><div id="optanon-branding-top-logo" style="background-image: url(https://optanon.blob.core.windows.net/logos/5138/5138:link.springer.com/springer.png) !important;"></div><ul id="optanon-menu"><li class="menu-item-on menu-item-about" title="Your Privacy"><p><a href="#">Your Privacy</a></p></li><li class="menu-item-necessary menu-item-on" title="Strictly Necessary Cookies"><p><a href="#">Strictly Necessary Cookies</a></p></li><li class="menu-item-on menu-item-performance" title="Performance Cookies"><p><a href="#">Performance Cookies</a></p></li><li class="menu-item-on menu-item-functional" title="Functional Cookies"><p><a href="#">Functional Cookies</a></p></li><li class="menu-item-on menu-item-advertising" title="Targeting Cookies"><p><a href="#">Targeting Cookies</a></p></li><li class="menu-item-moreinfo menu-item-off" title="More Information"><p><a target="_blank" href="https://link.springer.com/cookiepolicy" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Cookie Policy');">More Information</a></p></li></ul></div><div id="optanon-popup-body-right"><h2 aria-label="true">Privacy Preference Centre</h2><h3></h3><div id="optanon-popup-more-info-bar"><div class="optanon-status"><div class="optanon-status-editable"><form><fieldset><p><input value="check" id="chkMain" checked="checked" class="optanon-status-checkbox" type="checkbox"><label for="chkMain">Active</label></p></fieldset></form></div><div class="optanon-status-always-active optanon-status-on"><p>Always Active</p></div></div></div><div id="optanon-main-info-text"></div></div><div class="optanon-bottom-spacer"></div></div><div id="optanon-popup-bottom"> <a href="https://onetrust.com/poweredbyonetrust" target="_blank"><div id="optanon-popup-bottom-logo" style="background: url(https://optanon.blob.core.windows.net/skins/default_flat_bottom_two_button_white/v2/images/cookie-collective-top-bottom.png);width:155px;height:35px;" title="powered by OneTrust"></div></a><div class="optanon-button-wrapper optanon-save-settings-button optanon-close optanon-close-consent"><div class="optanon-white-button-left"></div><div class="optanon-white-button-middle"><a href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Save Settings');">Save Settings</a></div><div class="optanon-white-button-right"></div></div><div class="optanon-button-wrapper optanon-allow-all-button optanon-allow-all" style="display: none;"><div class="optanon-white-button-left"></div><div class="optanon-white-button-middle"><a href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Allow All');">Allow All</a></div><div class="optanon-white-button-right"></div></div></div></div></div><div class="optanon-alert-box-wrapper  " style="bottom: 0px;"><div class="optanon-alert-box-bottom-top"><div class="optanon-alert-box-corner-close"><a class="optanon-alert-box-close" href="#" title="Close Banner" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Banner Close Button');"></a></div></div><div class="optanon-alert-box-bg"><div class="optanon-alert-box-logo"> </div><div class="optanon-alert-box-body"><p>We
 use cookies to personalise content and ads, to provide social media 
features and to analyse our traffic. We also share information about 
your use of our site with our social media, advertising and analytics 
partners in accordance with our <a href="https://link.springer.com/privacystatement">Privacy Statement</a>. You can manage your preferences in Manage Cookies.</p></div><div class="optanon-clearfix"></div><div class="optanon-alert-box-button-container"><div class="optanon-alert-box-button optanon-button-close"><div class="optanon-alert-box-button-middle"><a class="optanon-alert-box-close" href="#">Close</a></div></div><div class="optanon-alert-box-button optanon-button-allow"><div class="optanon-alert-box-button-middle"><a class="optanon-allow-all" href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Banner Accept Cookies');">OK</a></div></div><div class="optanon-alert-box-button optanon-button-more"><div class="optanon-alert-box-button-middle"><a class="optanon-toggle-display" href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Banner Open Preferences');">Manage Cookies</a></div></div></div><div class="optanon-clearfix optanon-alert-box-bottom-padding"></div></div></div><img class="pub_300x250 pub_300x250m pub_728x90 text-ad textAd text_ad text_ads text-ads text-ad-links" style="width: 1px !important; height: 1px !important; position: absolute !important; left: -10000px !important; top: -1000px !important;"><div class="kxhead" data-id="KDqyaFZ_" style="display:none !important;"></div></body><iframe id="kx-proxy-KDqyaFZ_" src="https://cdn.krxd.net/partnerjs/xdi/proxy.3d2100fd7107262ecb55ce6847f01fa5.html#%21kxcid=KDqyaFZ_&amp;kxt=https%3A%2F%2Flink.springer.com&amp;kxcl=cdn&amp;kxp=" style="display: none; visibility: hidden; height: 0; width: 0;"></iframe></html>