<!DOCTYPE html>
<html class="js" lang="en-gb"><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
        <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1">
<meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=2.5,user-scalable=yes">
    <meta name="citation_publisher" content="Springer US">
    <meta name="citation_title" content="A distributed approach for graph mining in massive networks">
    <meta name="citation_doi" content="10.1007/s10618-016-0466-x">
    <meta name="citation_language" content="en">
    <meta name="citation_abstract_html_url" content="https://link.springer.com/article/10.1007/s10618-016-0466-x">
    <meta name="citation_fulltext_html_url" content="https://link.springer.com/article/10.1007/s10618-016-0466-x">
    <meta name="citation_pdf_url" content="https://link.springer.com/content/pdf/10.1007%2Fs10618-016-0466-x.pdf">
    <meta name="citation_springer_api_url" content="http://api.springer.com/metadata/pam?q=doi:10.1007/s10618-016-0466-x&amp;api_key=">
    <meta name="citation_firstpage" content="1024">
    <meta name="citation_lastpage" content="1052">
    <meta name="citation_author" content="N. Talukder">
    <meta name="citation_author_institution" content="Rensselaer Polytechnic Institute">
    <meta name="citation_author" content="M. J. Zaki">
    <meta name="citation_author_institution" content="Rensselaer Polytechnic Institute">
    <meta name="citation_author_email" content="zaki@cs.rpi.edu">
    <meta name="dc.identifier" content="10.1007/s10618-016-0466-x">
    <meta name="format-detection" content="telephone=no">
    <meta name="description" content="We propose a novel distributed algorithm for mining frequent subgraphs from a single, very large, labeled network. Our approach is the first distributed method to mine a massive input graph that is...">
    <meta name="twitter:card" content="summary">
    <meta name="twitter:title" content="A distributed approach for graph mining in massive networks">
    <meta name="twitter:image" content="https://static-content.springer.com/cover/journal/10618/30/5.jpg">
    <meta name="twitter:image:alt" content="Content cover image">
    <meta name="twitter:site" content="SpringerLink">
    <meta name="twitter:description" content="We propose a novel distributed algorithm for mining frequent subgraphs from a single, very large, labeled network. Our approach is the first distributed method to mine a massive input graph that is...">
    <meta name="citation_journal_title" content="Data Mining and Knowledge Discovery">
    <meta name="citation_journal_abbrev" content="Data Min Knowl Disc">
    <meta name="citation_volume" content="30">
    <meta name="citation_issue" content="5">
    <meta name="citation_issn" content="1384-5810">
    <meta name="citation_issn" content="1573-756X">
    <meta name="citation_online_date" content="2016/06/09">
    <meta name="citation_cover_date" content="2016/09/01">
    <meta name="citation_article_type" content="Article">
    <meta property="og:title" content="A distributed approach for graph mining in massive networks">
    <meta property="og:type" content="Article">
    <meta property="og:url" content="https://link.springer.com/article/10.1007/s10618-016-0466-x">
    <meta property="og:image" content="https://static-content.springer.com/cover/journal/10618/30/5.jpg">
    <meta property="og:site_name" content="SpringerLink">
    <meta property="og:description" content="We propose a novel distributed algorithm for mining frequent subgraphs from a single, very large, labeled network. Our approach is the first distributed method to mine a massive input graph that is...">

        <title>A distributed approach for graph mining in massive networks | SpringerLink</title>
        <link rel="canonical" href="https://link.springer.com/article/10.1007/s10618-016-0466-x">
        <link rel="shortcut icon" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon.ico">
<link rel="icon" sizes="16x16 32x32 48x48" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon.ico">
<link rel="icon" sizes="16x16" type="image/png" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon-16x16.png">
<link rel="icon" sizes="32x32" type="image/png" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon-32x32.png">
<link rel="icon" sizes="48x48" type="image/png" href="https://link.springer.com/springerlink-static/439917616/images/favicon/favicon-48x48.png">
<link rel="apple-touch-icon" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-iphone@3x.png">
<link rel="apple-touch-icon" sizes="72x72" href="https://link.springer.com/springerlink-static/439917616/images/favicon/ic_launcher_hdpi.png">
<link rel="apple-touch-icon" sizes="76x76" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-ipad.png">
<link rel="apple-touch-icon" sizes="114x114" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-114x114.png">
<link rel="apple-touch-icon" sizes="120x120" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-iphone@2x.png">
<link rel="apple-touch-icon" sizes="144x144" href="https://link.springer.com/springerlink-static/439917616/images/favicon/ic_launcher_xxhdpi.png">
<link rel="apple-touch-icon" sizes="152x152" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-ipad@2x.png">
<link rel="apple-touch-icon" sizes="180x180" href="https://link.springer.com/springerlink-static/439917616/images/favicon/app-icon-iphone@3x.png">
<meta name="msapplication-TileColor" content="#ffffff">
<meta name="msapplication-TileImage" content="/springerlink-static/439917616/images/favicon/ic_launcher_xxhdpi.png">
        <link rel="dns-prefetch" href="https://fonts.gstatic.com/">
<link rel="dns-prefetch" href="https://fonts.googleapis.com/">
<link rel="dns-prefetch" href="https://google-analytics.com/">
<link rel="dns-prefetch" href="https://www.google-analytics.com/">
<link rel="dns-prefetch" href="https://www.googletagservices.com/">
<link rel="dns-prefetch" href="https://www.googletagmanager.com/">
<link rel="dns-prefetch" href="https://static-content.springer.com/">
        <link rel="stylesheet" href="basic.css" media="screen">
<link rel="stylesheet" href="styles.css" class="js-ctm" media="only screen and (-webkit-min-device-pixel-ratio:0) and (min-color-index:0), (-ms-high-contrast: none), only all and (min--moz-device-pixel-ratio:0) and (min-resolution: 3e1dpcm)">
<link rel="stylesheet" href="print.css" media="print">


            <script type="text/javascript" async="" src="//cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.4/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><script type="text/javascript" src="get"></script><script type="text/javascript" src="optout_check"></script><script async="" src="mathJax.js"></script><script type="text/javascript" src="bd339c69-af54-4a21-b4f1-654bcfcd83ca_002"></script><script type="text/javascript" src="bd339c69-af54-4a21-b4f1-654bcfcd83ca"></script><script async="" src="controltag.10b38d7fdd1a9bfe7b2b04b85118d005"></script><script id="nativeads-plugin" type="text/javascript" async="" src="plugin.js"></script><script type="text/javascript" async="" src="KDqyaFZ_.js"></script><script type="text/javascript" async="" src="analytics.js"></script><script async="" src="gtm.js"></script><script type="text/javascript">
        window.Krux||((Krux=function(){Krux.q.push(arguments);}).q=[]);
        var dataLayer = [{
                    'GA Key':'UA-26408784-1',
                    'Features':["leaderboardadverts","abtesting"],
                    'Event Category':'Article',
                    'Open Access':'N',
                    'Labs':'Y',
                    'DOI':'10.1007\/s10618-016-0466-x',
                    'VG Wort Identifier':'pw-vgzm.415900-10.1007-s10618-016-0466-x',
                    'HasAccess':'Y',
                    'Full HTML':'Y',
                    'Has Body':'Y',
                    'Static Hash':'439917616',
                    'Has Preview':'N',
                    'user':{'license': {'businessPartnerID': ['3000197460', '3000201946'], 'businessPartnerIDString': '3000197460|3000201946'}},
                    'content':{'type': 'article', 'serial': {'eissn': '1573-756X', 'pissn': '1384-5810'}, 'category': {'pmc': {'primarySubject': 'Computer Science', 'primarySubjectCode': 'I', 'secondarySubjects': {'1': 'Data Mining and Knowledge Discovery', '2': 'Artificial Intelligence (incl. Robotics)', '3': 'Information Storage and Retrieval', '4': 'Statistics for Engineering, Physics, Computer Science, Chemistry and Earth Sciences'}, 'secondarySubjectCodes': {'1': 'I18030', '2': 'I21017', '3': 'I18032', '4': 'S14000'}}, 'sucode': 'Computer Science'}},
                    'Access Type':'subscription',
                    'Page':'article',
                    'Bpids':'3000197460, 3000201946',
                    'Bpnames':'CAPES MEC, Universidade Tecnologica Federal do Parana',
                    'SubjectCodes':'SCI, SCI18030, SCI21017, SCI18032, SCS14000',
                    'session':{'authentication': {'loginStatus': 'N'}, 'attributes': {'edition': 'academic'}},
                    'Keywords':'Parallel graph mining, Distributed graph mining, Single large graph, Frequent subgraph mining, High performance computing',
                    'Country':'BR',
                    'Journal Id':'10618',
                    'Journal Title':'Data Mining and Knowledge Discovery',

                    'doi': "10.1007-s10618-016-0466-x",
                    'kwrd': ["Parallel_graph_mining","Distributed_graph_mining","Single_large_graph","Frequent_subgraph_mining","High_performance_computing"],
                    'pmc': ["I","I18030","I21017","I18032","S14000"],
                    'BPID': ["3000197460","3000201946"],
                    'ksg': Krux.segments,
                    'kuid': Krux.uid,

        }];
    </script>

<script type="text/javascript" src="jquery-3.js"></script>

        <script type="text/javascript" src="6b2ec9cd-5ace-4387-96d2-963e596401c6.js" charset="UTF-8"></script><link type="text/css" href="optanon.css" rel="stylesheet"><style>#optanon ul#optanon-menu li { background-color: #F7FBFE !important }#optanon ul#optanon-menu li.menu-item-selected { background-color: #FFFFFF !important }#optanon #optanon-popup-wrapper .optanon-white-button-middle { background-color: #3365A4 !important }.optanon-alert-box-wrapper .optanon-alert-box-button-middle { background-color: #3365A4 !important; border-color: #3365A4 !important; }#optanon #optanon-popup-wrapper .optanon-white-button-middle a { color: #ffffff !important }.optanon-alert-box-wrapper .optanon-alert-box-button-middle a { color: #ffffff !important }#optanon #optanon-popup-bottom { background-color: #F7FBFE !important }#optanon.modern #optanon-popup-top, #optanon.modern #optanon-popup-body-left-shading { background-color: #F7FBFE !important }.optanon-alert-box-wrapper { background-color:#F7FBFE !important }.optanon-alert-box-wrapper .optanon-alert-box-bg p { color:#333333 !important }</style>

<script type="text/javascript">
    function OptanonWrapper() {
        dataLayer.push({
            'event' : 'onetrustActive'
        });
    }
</script>

    <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
                new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
            j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
            'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
    })(window,document,'script','dataLayer','GTM-WCF9Z9');</script>

    <style>@font-face{font-family:'Source Sans Pro';font-weight:400;src:url(data:application/font-woff;base64,d09GRgABAAAAAERcABEAAAAAiPgAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABgAAAADYAAABAA0QDckdQT1MAAAG4AAAHPAAAGXDUveN5R1NVQgAACPQAAACJAAAA4PFn1ldPUy8yAAAJgAAAAFcAAABgWrSUW2NtYXAAAAnYAAABRQAAAebzMPm1Y3Z0IAAACyAAAAAoAAAAKA2ZAPpmcGdtAAALSAAAAQIAAAFzBlmcN2dhc3AAAAxMAAAACAAAAAj//wADZ2x5ZgAADFQAADCcAABd7CMplIhoZWFkAAA88AAAADYAAAA2/hSz4mhoZWEAAD0oAAAAHwAAACQHowOhaG10eAAAPUgAAAHlAAADxKaoJjpsb2NhAAA/MAAAAd4AAAHkOIhQ1G1heHAAAEEQAAAAIAAAACADCwJAbmFtZQAAQTAAAADXAAABxiK6PDBwb3N0AABCCAAAAgUAAAM5bFBIb3ByZXAAAEQQAAAASwAAAEuWBPrreAENwbEBQDAQAMADgH10RmAMHSUyaqbJ36mwAmqj1qayx8rhVLli5faofLHyS2o5VgWBWga3AAB4AdxWVXDkRhB9ZqjyYZiZmfknzJzvMDMz/YThJ8x4zMzMzGz2eu8sr9ebs867UeqlqzM1JU2dDOHkaaHV06Se7mkhD0A5TsAFyL/goituQI/7b3n8QeyFQghIXbe0XPn33fHog+jx2y+g63kikQ/ItzzvErk7TDV64CLcj1fxPSZiMWqRzsvPK83rk3dA3lF5p4hUD5ShL/bE3jgZp+ImPIy38A6+RD8MwExZ21Os7M0kTqaHU5nAHeiDO3EIvmQG/TgJw7ALilTiVGaE2y7cMcLdD6Wis0O4rbgN5aJ3hOidLBJtIrFCJE4VvZOZFglfuOsxiL5wS8XfngxExhNOu3COFbk9mRUfIqfcVuEehgKRmYsBqjlTtNQWioWTwDBuQZ5oVSNffueK7lqR35MJoRLi0RP+nShBGnuqxCKcrLyXmMMr8s3f5W2xiF0qdzsGZ+I4gI0cxDr5HYcDcalIVOAvhMT9p4JprgIk8j8ZnK6/cyO8pFwb/wTbsrNy9QNwtuXN5BaAq7lFqKY/wYePPS29HiZ+gDUcLitl9NBtcD1bAWaUDpjhAsNvp68eFcLP8VuRTYonheoUMsHFnKvaDmxsgbHtM2NtBVYiDfXBWiAcPXMMrLSPnYApih1m6UVsJ4Wus9my2lyMHlYzoLUNMCG/DtRm4PD03gXr9a80lJPOa1tioj630B5TMPFzm65nAWhUGsVKWHA7wCxFQ75p1tJ3PfJ11nE+RzMAWMnNHAmggGKXCc3FdpXy6PNuDuf3/N5qjgZYzX4cyc0AF7JypznJRSuMvvKdijS6xXZPfLYALpiN2nZ3mj71nvplCgrD0b10dYTeyFpDZRzbWUSg6w6tO9GOQnQTzHCc5rsK4Idczw1cyCTHMcWpfFUrdaU5fXZhv2hcAKtEfjhXcCXncIzm0MJ5vjRT9GFrkSsc2RqbOd/Jclz1/mSpSoA5vdJ6Pw1hHM66neoH6kkr2Y2Ys7UbMyjjD8JLa0VPAbiOn6tUJaD1faBwo1Whq5zKBVwv0iu4io2hqshRd5dpE2mOtfS1Izw2sUq1t5oIdrF6uciJU8nKmLMqGZKLARe6Ha29GQtbg0nNTLFcbg81ObW5i1ubqlcaf74zAXRWuVyGboIf0mOlnhJ7cZvQHzvr7TBnIJ90Vsbr79cczC0dzAafCc1JEYroh08CQyvFSnoMnHxrnYhW1mYz/PTH6fc42z0HcWl4LwwXbAVwnnAuZ8acKafhQOyFI3AQKuzkMv840WYayjlUJ4lbCZlIT7WjB06ABTdHu9HK9wDsBGijT9/46hGpkxRUw3RAInKq2RpkwNUmyiytBJMmwk7h9oZ6znaoMc/0spmuuqsxE5GJsDU2M6DErbO8BcW2W12UscrqDLHvKGuYsfwIuIjPWT+dgr6+IfdllvPYbJ64UfnqSVHBNM5zc4OD0Fc721NpP1rfkenY1tW+pB+RLgdc2Lq0/kJ1kgrVd87EcBB6RHbAUkxFLdgOrEWnYNKNqmsabAtFuI5B6KkD1oZPFhym/7YHXbAFcOd8yPaW6PucM1uPZcbaDjp+OiaMnOYzKs2tIY++iX1PuFAtpvH3o1CfxcxLjucmejoxK7XrnrTTu0PQ06sW/x/E1/GGP2aJPmst530upc+RANfoOfEgugA2mvPEw98CW7lNAHdwB8Bp+s7WTJ/bALZyHH/p/FQQ+SZ63M5GNlueF3lDT+uZXKDzq4U+5Z4/a3X6MTYD1QAbmDE29N7BNXF1i98JNqrvqd23yAQT7qkVD83JfPekiJ3VyucULkcMuF4zuwrdhGplO35KrnRmooK1dorEVgibmInrtzgN/IXgO1zQRcmgaxXAKks9z7mGqmRa+TFgNu5dIH6m6fQMYmZaJ36YdONla3R2q2WffmS2dla/yUhe8lHxa1vzrLaJAQBh9Py2jc/WGnHapIlt21a5jG3bubVcQqaLds7z1mNUGVUNBe2o6LpD1b2ecaPnvOBBJ7zjEe/73Eu+iuO+iRO+i5N+iFN+itOGjPrWuAmTZjW1TGnr6Jo2ox2zemnOvAWLBpbssWzFqoI1bNqybRfsYDcKUVVSVlF1VM1Z6hrG/6ZpMkV0IuTtb5oWIyyb1bf6N5vWUgRAVZ7/ZtN+1Qj1oGBLxZm3EesxZNghow5H0REJHHBQsi/CCNj7d8YMu8hozBqPBdOxYC5WLMeC1ViwHnM2YthmjNiKku0YtRNluzGmEPOKsaisalktltRjXDNWtWJCV09bP5YNom1PLNsbbfuiY3+sORDrDkbdoZh0OBqOxJSj0XWW8826MBZc7FILLnOlEVfFmKtjyLVutugWd1h2p3usutd92u73qGWPe8q6pz1jyvNesOSEk2acclrfy7HrlSh4NSpei6LXvaHqzah5K6a9HTXvxLT3o+WD6PkwWj6Kno99punz2PJ1DHwbA9/HwA+x46fY9Ets+CO2/RmrfwE9c0UteAFjYGRgYOBiSGCYx8Dk4uYTwsCXk1iSxyDFwMIABP//g+XFGJgdo1wVGOScg0IUGDRCgrwVGIygsoxgmgkIWaEsZgY2KIuFgZ2BIyczPZHBCCeZn5zDYIFMMjBATAVjJqBZUlAeBxCzMZwCy6oxsMDFpEA0WCUPQyWQVmaoApKGYHXKAFZmI2wAAAB4AS2HNwGDYBSEv3svPVMc/BpiIH2KgUwRxEL1hAVGFOCADtcPV6mELVhkP+C7tP5cFcDOW/eNm23GswfEisf7+eEOXW9NvwOvdAkoB1BtBRDwATBlDPMAeAFdywOMHQsAAMB5iD9qY3Oxatu2bdu2bdt2G9W2fd7YdeNe7YkxiCKGbGIiyCYeeYgKAnH58K/CyqqinS7GmeSce9K99DFSNNI8ejWaFE0JsgV5ggJBQlAuqBGMCA4kZEnI/vEj+FegiHKqaq+r8bY4777Qq8+72fedNcgV5Pu8y/6xIx/fowEfL/HxDHwcw8fRfCxFRraM+IeP4VwIJ4Ydw2lh1TBX2qm0YyKgDFqgE3oZgD24jBBAkjR/i5lnsaMuuG68RbbaZr/bZltnlt3mmOuB+x5a4Ijt9trhsS2e22mPXZ54ab4rLjrvkn0OSHHIKWs8td5hqc555LgTEj1z0CbpjjltrRtuumWhluqoq576GmjorkaaaKqZ5lpoZYjR+uirn/4GGOieQYYaZrgRRhpjlMYGW2KpFZZZbuUn89p7YwAAAAAUAEQAUgBWAAAADP8zAAwB5gAMAgYADAI+AAwCfgAMApAADALIAAx4AV1QRVoDMRROvT0BUsv7cKjs2WCZwd3ycLcz4E7tLI9d2M3BsFeXkfz6xYwQiVkjQkv60+UqoHH9vRihol8iJDz7e0kjXAkprUtFrgMW7gQbg8DMk5A2eXrsFd2FMiMz0ycZacuLwxPy9pSQg9MMpiWJVX3J45oGGsVwjZ4iDvM83uI83tI8GeQZriozMLLFE/xwyZeYleTpXdLLmu5VmEYVhgGkRc6SJkeFAZFb/tpOGW8u2yt7DiQ4GmQSLM+yqmk0TAIzmbLqArrPZMKZLqxqI5wWwyVajdGKYUTR4ZuwjOt+iSOGLggXjS7oAgwDKl47lJhd1ZYKA2DyHyTqa+QAAAAAAAH//wACeAGlvAd8G8eVODyzAAGwAURddOxi0TuxKKwAOymxSGIRSRValuTIiuz02BfX5Cy3FLfUixOlfne+L7Yv5VxkpRenJ7p+5zi5Oxel5+78pccCvzezgyWIgI7+99dP2Bk+7M689+a1eW8WqA2tI8QVuXcgDdKhdmRCdnQCPY7Q5Bpcsp6zCKGeyjoDcATAIW1l/QnUBh1ktvQ/gXRqT6/2DKx3FnUg/MsnUDf7GwbqOYssAMv1WkWzaI2aRbPemsWSCVurWArgdfx0LfT5m/HaTYu7CtrS7tO9Yz8dy3PvuHgVXv5RLY+/86PahdMnTpzGuzauvXYD0IORNXDxcfcA/ma0XMfeSJA1NmDfRQBdSKsCzARgZuTAN0aKeifpUVQ7s2dRO0UVRzM4KhmxCfOSXrLKVgm/OmARerQmwRJ46r+qKz/Y+ME67jxevrq//+ry8fGvcfdcfBXBTYMOAG4hwM2KvCiMDu7IWysBWBVkJPjGSlGQsk8AByVkU3nnAYQADj0v7VmhF1b4Kef9nN2m09v92G4zcklszZeKhUgS29XOgS/llwbFvoGZvaGxQ+W/+9uVXfPzV7xmZePQ6mu4eyyR4WTvHpO2c2EiN5Nz4j8bKJey/1z7RnV8uP+3CGFU2Pwl5+Heh0TUX6ejjaDd1kCHjwB8CpfbALcuiqUPeg4Fy0KFk/MOXp/BUtAI+Prhz1KZN8KfGa6QWjhVHTrgy7qriexcye8rzGZz00I2erRaOTmfxDN7blrL9gpFrxiePDpcOTYViUhyOp9ZvRnYTHldAF53ACOLO3K6gwA6CIZnUSfBjvLVXOehQ+FdmbHMfOCj7zkTW3jN7gfOvBb+zXL3PPD+hetX0hNvvvHG0xeBLQiTefEPYN5ONL/jrJ0E0En5QufTUL3Qwz2ddG31WeAjxaYTvjUwDbGLdtku2aE9gK+r/dvvfod7azdw90x9Z/oX00id+x+p7I/vOHc7AbS3mrt929xbM5olOt/vf78Ek/1gqvYUqq9/BNY/jOZ3XH+RAMTm9Rfp+pNZfXC3SGf1wayel5KJolSUzUwu0vteMTZx0Jd1jebze/sCvvK+olS1veVfhJNjo1ctpBTBkP1Ft1QXDLvlY5O1n4lJkA3GJy5P12jXJazRE6hdXZl2wFPD5ORxpD0P90GrP0+kBayXbJbgKh1Ywr3Ly7XvAsNqP8fWi6/Cxdo3lXnRQzCvBoV3nFdDABoyLx0TxlpiBgSjwOYv8Z3wfA9K1p/Xk9v16vOqqaMSraEcLfsxcLAIyIEwS8HoQUefVCws25OTuesHbo9Ko7P4q7XexGyfcJLxpgpzdL2EfdISgPalzKmBAAwK9zh4touaUxA7ZIAepUzGst4qafR2IFGDzRvf/vnhL7+acOwxvOv3tZfjlVvP12X6W4BPG0ruiA8TPEozx3TXTPRkCQdqdwH7HptG9XW/EcZyoVfuOJaOAHQNABcBuJC2hQY9oXhKSptR7Znhbh3t2WBkF6PXystlsp7mQqks6TWSJiqBaYb1PXpzTpu56egevVZbvCl7c0GrMcwDFz542234CEjOydTLU/fV/hqv3ged2rtVOtLUj+y8RmwFWjmVFhKuhY6B0UH8jYpzBhPB2cJ2/8+ntRr9nv2/mNZq9XsAz+N35E8V8BJg+v7b8ycKtb9GHLUNV4BtsKKoanuZqVURUlGmq2ajOtUBlsFP168N9LxILYGfIzY4mq/A3xmubhLwXOXE7nh894lK5cRsPD57opKalr1eeTrF2q7c+s179ty4msmu3riw5+b13K7I9LHK8NGJcHiC2ITpCPMRduCjEfGtLWaTRj0BMo5YQNAFlsDELIHqz+rogquQrKrT+HrpQFWSRtZL1yx9bv/S8spV3D229O5SYS7vqr2IZ0anpsu1PyDKt0PcaynfCmi2jk+ITB9qwMdNAG6CzznUizDgrk0CtKfuM0I9DDOAPY4yxDZRI7rFQUwMaSSa0TSyWE/NRPnNKhNZi39syvsOXucrrVWDdYaLw/sL/HLJIRrxbPWKmUhk5ooq4+3FZ9s6Vif0pbVrZ+rsn379ark9ldBp635ZAJ4bkPPS4zNV6SjRGGHUQYk2Mvf4BOIJjIktYzzWSKBu9UDI+tAr5hcX51+B7WeuXfp2fmU4uHf8L/F6dWS0cPFT2F77KXePNTUlzx8y154hOpYAW/tfsBbylo9wEBwcDVjmCCC3ZWtJwOhAOSofDkDNR2FhuMUBMNUSwypkscJ81dPpiFQHsF9DloIsziHpMnl0xuyL8Rt8LuIMlGfTg8eltH9XpjxuFSK2w958yB4YWCxMXpO8O5wezLkkl60z3uGNlUPhkaw7HSr5RDntFF1WQ6wrEC+HohOyr5AmtHnh8iKsgR6o2xbOq7SpjoWuAAe0ICpRbUSapKJIooJnnsT/8eQyt3tq6uKjSvyxCnqfofZ1aEdt6iaA7gaAjQBsChP1dePNBFZ1WmANVpec8T5B7Es494aq+/P51WoIzM+Ng3tzdntu7yC+pXbznqvGfb7xq/YALnQzoAFcutHIn4r+mpGr7wIQ6qbLZpatfszLFQwGXCOtn7R4rFqL13xy5fzlgMGZweMDA8cH8VFw0lQbEfd7mFdE11z6vIwHrSMnCuAJgN/CTKTG2gQP8bTnRaIq/yq22/DWKNsWcD38/sl2W7u209l5cO5gl7NL227pmNx7x/Fj7SaD1mA2XH4MyLq/dGWpdLKIr6jdX3i50gMb/5bIrgj8r72errcJLi/Q9Vb9ECZ44j+Ko17Kn5oIwKSQpoNnXZQgK4A6FYI0Mq/QUcWyxiqp+y/pO588tN7Nm7RGe9fq2qc+d+hIt6dHa3R3XbaGl/DwA3zS50vyD9Q+V/vEJ5yy3y87P0HldHMTLt8HvE1o5NLxbkbTpKj0NuTs0pe/sr7S4zdrzUHT8upXlmv4vrOhqXB4KnS29ooa4RlRwP+hMfow2h7yNftuRSV0io85D7f1AIaqFpaxCGqIIcpJ4/Xak/jDtb/Eh2tDec451XvxR9N0LzwANuwx/ALyoChI43Y6DDsvCJNGFuCY6IIQLOFZ1d7a4QYThbsAFqBo2aEXpD2CqqQEnxDuUK+j00crmOyozFJUov6nbHPQyN6oGbj7fe18xJ/c5RODRwfX907oOWlfurJvX29IqkrDy7Ljpr0rZj4VtAu2fkf86oO1fx7wxWdmEiFRuN3g8EeB1kWg9Q/c15ADRdAtTZpnaY7utojvIYAelXjVyKvRXQ+Txh4a0yl7bJ4S2QM9N+3poBdSCY8wI2bXE4L9XN2iQ7aAOADVD+sXFzlxIZ6d6XUFB+YSwV2CRj+2KFSdfjlsNwZ6g4PT3Ne+uOaDCP1gX+XohBTw5a7y8p7emXRsLOseU2xuBuj+OH6hxT7ccMn78PJL7sMzkZGVbG7SFXdkfWJfjHdE+0Sp3xEW96Vyy8PSC8OXjQUlvmh1evKTicREr8dnK/mCwdHLqAxmAb9/g3WxIQndsiOGzCJuLZSFACxI29I0G+CbNhattiELXRcLUOOkdBnVzIhBzYwQqgPbaI0W5T+SRIX8rASrkagI2qUJvcY3mzx09crUSHDIH+j3++SI44XBy8ZCfvfSuYt9/d746evXjorBGlmWCUIvXY9vUJ0LoX3N9LY0513wDc9iSvgGMO3ZhjPAmXwx7KlM6UU/blikcCFCcNdnwsMLicyuXrem9njb6pQ06PEJC/nelWoIL1f73dmIS/vC0GVjkjC0XPBY9hyym0s2vzS6cfnsHnu0P6zIlB8uVe4fQcn370gBWyBLq8QCXQxOJcJAe53nyV+wPGC8MCDrwzS3YJaK5WKhRJYEpz3eiRMnFsfGJnfxcTN3a9fG/HjtMP5IpW1+dn+7nubQ0psC/hnw1wxx2Tq6uo5fmMwebsBvkgAmkW4rNhuC7ycpn4ey9WxPGBCKn4cPQBIUMgQQGSAyQHaze86iRcp9nshJhaPRJSg0ONMClZoia5hoAV1+XNd6I/QcgxgMHwtkrMoILOR7/eyY1B9z5HPLRrHH7ZyJiBZX0OqM+vLjYZtk/OvjZt6bGw1bIuZuSyi/sbISHDnYn95d9H3dHvb1DCWTQz2+8GI0Y/aErEG3ps2aEAJ9vLYz6gtkurW28VRsJO3Ut1XNhUC0mnS2Gzo9Nt5fHg4OZzzmYA4v9PiTntLgYMmT9PeQtV+Ey79QfV1DOxlN1ay29hg6lqKEFaem0aSaRotqGuu+wEw5oTdLi4saaU6en15M5SID4UWwfMfF7JGN2rdwfHI4Eq59EHFoHCZ9hPtb1AlDptH2rYKh5X5NR2QmCVMGI4qQka0AKP4NTlF0wmcMbPBuP+/0+Zy8/3jNTcPGzX/djNJ5TKDF5SZfYmjOOuhgHrIrgHmYdVWDWB0RABb0N2KwtDi0iw91m9ud3SFngqDCiyK+cPHZ3WuGtoqmLZ/m3lRHqr4mF2BN1JyMmoLZeU0uPSeDGnIyVpKTiUJOZnGf5rsHP3L2wF0H93Ffq/kx+lLt+z95+RupfSC+Fv0r4GNCSfSnpwcmqRxxyANYMVmR4uKisV2r13caHOZsjPvaxXfYeriKVttXQoxm7n+oX3vPn5JDw85hM4ttWuf5OTWObmc9Iq+A+vmzyE5ltl119jq2nVM2QwFGEC9TOea3SbN+q10EDyLMpvqqpsh8enbXYgo2b4upLFzwhSkp25tKyHURn619kDUIqXoI9NvQFZdOP9PD1jkeIJcpJlNKRoxFJaZZKSn+0ryqlRTr7Uqp+rz/BlxpvqfZY7QIoGFJVOcgbfdrzEiWmzIWmXBlXzq9rxKutyu2cCEQKIRt9ZaGIcGxy4aHLhslwccQiUTAKScSk3kPRCaIxogy/gPg6QA8b0Hb/ZahOQNraC516VrWZ9rgWSsVIgeA2tW9P69WZNyqKIVUqqOKPwEHCGzfHiZuixF1+o+SNRD3JLMzOSVMHA9yYkWNEQPVUPyj3Kf6vDGIEvsrl09IfvfIVpAYFJQYTMb/RmVJQjfvuD5MVgzNIZeuyeezGIxFXrAJZdFYPf4yXVL8JZdl4vp3jL9GRM4wtqT1zSZaxV8+X+bi/EP7PdHt8RfTmw0aHyxeut40WwXzli14WlES0zYloZGyqiGu6YSP7+mymQLjLlCRtUypY0arzVdqXyO5FtCNWwCf3q08kp3MZm+YPkMAme15pAyyUyTshIcUFoJnMk15pGirNBK/lUWaFMZTuZwtELO5C3E3n6xEM/PeqKskZJIWf9TuLiY8fHo8WdiQXuEWUoLVa+8xtPNiKuAD+RL4gtWZ8JpdVpOhwyHmgv5CjI/4FPvkBLomuVeAJk828bmlOaI2p13dyxLWaqC1AbQboDSrVMUyrTfZHA2BbfHIzFzH5C23iIluf1ePLQuBu2gkIeEdd4zXXkj3gsPUdzoTQyHAaRZw+g6+wGxm66VW1b157Vs6TjPJqrKqdTv1mKAWWzU5meFq3tpLFGeXkmArqWAEZ+WFaXyh9i+Tw9Ek7NLdc7HckQ2Ff11w+Sq+0CI/1bwrYoCd81Pb8zxdK4c6+U5tp6Pz0L4Hj8DsPwvNSNJMCNtqbrJuCHGPwryQn7r0edlCGi41PyWy0r8JQDaWn0JIuMT8lHN5xmDUaw097bN759rNkJMyGqYX/vz4VLupHaAdE4SqH0pg7seD2KX0xiTouXGbBEnoSan2IuWxES4fxxda5KdaWL1LzE9B7MlqO5jVdsrRhhSQnm/ITz309vXRTmc3WYvB/W8/sz7d7TZqu51dY/trPzpli9vg/6n/+dWrHSm7Pcm/msnFZpbKhQ9NoW2B2x/vvZpxbtzsdCMDQ4+nxqKKNY3cNmqoq+k12NoTRVPnl1aOdbpAZmwdq3sf6xZSfWGzt6vnNq5tMB3CP6z9f8J0UJwRcPfFF0KVtLutbVTBNQCXu/GFFrksw/9hLktiuawARs/huU2EDbV7Uvja8VTtThL0b24q5zo0eS6CtADQo//EppbwZ3eAP78NLqvwH26Dp1X4T7bBT6nwnwEcQEAS0rTRuqQd3YC26eelF0a71SAcemqBuU11ngrDupn/IUIH60rvs2XrgRs2SzoTbrOLRUgw4fopAXyvtlga5oTaW79zbA2v4ftrP/nRjzCH506duubc9Oc+p5wbAPoKm0lyjgPo01H6foFXKd30HAPln4Hx29sS/uwO8Oe3wdMq/CcMTv49Su/vUsZHt1D4Xrhf1wB/VoFv/gHgnQ3w5xn8RYB3aNIq/CcKXKmR0nU2s3WeonBam6TjWBhdcy3hz+4Af34bXFbhP9wGT6vwnxA4SFF+8xnuFPcw6gR5yaFb6xKTIvKQapAYPwH4FQFB0EnBcqtRAdzppwLTxXowAAiHkaVkyd00ElXvE1VYSO1l2LcsjqDZiCyOKJk/Hm+lLvTRcj2pwZf5866RCKkHmnuC8lxqRvYG8Z8FglB6HS6sS9XMlED6lSunq5nXuX2k2JrbK+RLpNp6ZAK/OfnaGVKFFYWl0MHUNbQ0WN0nkZonMMcPemRBHrTSZEVa1WAV2+FktoOciNKwimxHtp5qMqrxdxepV8G2u+HwDN1dQEbGDNFS4YG/wB+gmiHUlmZfv5jcf921udy11+HZ9z9Az9WkV65fuOlOznCari2tb1FZcDLZGWsJf3YH+PPb4GkV/hMFrtQl6P1eJvu/oLZGhMtueibDg9brPPIQlnhQT4vMdb1a52EmhHDJgxwql9pZtf1x5ARb3A6tC/gUZnVTyihr/ewTCdLN2EjPO73znfvY+ac9e+iZp6ezF7k55QwUHk1qsQATN9TVe5AL9beqgjWl3s/BfZgsb/Ic4qDnhB5NE23V0+3S1iGspWv+uKD+CltmV6kwn3e+66mR6elS7UXCZ1r/4L4B/PRTfv4H7kat4M/sAH9uG/whFX5hG/xhFf7jbfB/UeE/JXCkIXDup/gFoHQETaP70U7u3EkAzgZAngDyhFXbzQYF5NVTXH0oz7Lhfdn6acsADJSisEAWnmMnBJUU7TDtOaE3QXsB6E1Tp8J2ZBDJQHlGU09yqJlnjRqgR6L0Wtwq87QpyYO39IaCe0VDBx7YJ/NDwSOl0pTJty+fnsw6NbWvYFesFJD6PH5hLj+86M1fka3su+s9BigCxWZ8nDzYJ+/Sn06EoiFr0esiFR7sdMTLuVxZGFopeC2RSsrJm/tt/nIonZkZU4tDx2f3Ls0o8UJmM0nqIrAGAl2Dn+MeBof8PF3jIJOJ9pbwZ3aAP7cN/rAK/zGDEy/0aXr/Cejp0H9cjxSZgOYnDfBnFPjm76F5sQH+HIP/EiHMkfEZ/MfXq3kh9FF8ocUZsuaExdYZssVF2JbW3BQPmleispxkshxS6flvikeK8cXUEv7MDvDntsEfUuEXtsEfVuE/JnCkAfiPIXd0kWRRUGZrjxwnJMQbaPISgJf5RejEqV/kmHWLIy+V8biafy2zVH0Lz9bo2Bzg2T4ZLQWKUbvTE+4LFCMOHodcPpJn2jPUGy4p3ciMLx86HIUkQ3w8G415eifj4xnsFlZLJN3UO+QbFVZKJAPlcg37GF2/1Ig0JzaExtGr0fZIfYuuPgLo28qTKgVVE+uBTp6He6CVwddFQJtlSmckCxC4L6Xq7QjtRaA3TulvzqDpKbk7arFOLSLp9H9uC+X9/nzIVm+XQ0N7U30bbu96KT2Vc2lqX8CueDkgDYD+LuT7N7z916X2DEncxyp98m79C02ZtwThS2Wof7RJd0f63wB5uiNUZWlOmZO4v0dldLwpCtjKKZcJoNwyvQs7HuhoqFAIVCgk9vcTKMt6SuiQZ8dMS8y5wMevrYc8amZFWyyoZRpQoJdtRIcTdq/k9aXklM+RGAwnJgKiW/YlEhTiEb3xRGlwkdNsvJqPFf2RYirudgoud6gUE0ox3s0POOw5wRfxuoS0FMylMkKkd36h1kV1g+YtqC7lme71tIQ/swP8uW3wh1X4jxmc7inp/SVlHPRCA/xhFf5jAkccKsI4z3NPIRvq24o3ZMJrGVlaVJXPkXwmSTMnn0A5lKaVEJ96ZigN0msHKc5BKz+ttCUab1AOR7eKzta6hma4unZmMNFWujCgqdgUzAfNGItjQbEQsjqi+VNDc4H+Pb2JhaGQUJ5NRabcxnLkbc6bC0OR0JRwl7PPV9iVMnq7U35najAIh9qdXHj4gWuHjk1HheHVkrw0KAjByqHIrvhoTCjYUp5dcYRYrpidJ4ii1dY7cdXSqkxR8wutcqb1TLBBrebykvmPTgnwxeb87yIWIG+0K0/Tv9KYiLenf7+teecXl9zhevrX61565/b8L63ncgehnhtDN+wYc7ACqqVlyB1W67kWWDyRJPDgYwGoSa3t6uFjhG87zit3maENQ+s4ryQwm6q+2yq/eqJn8BX7k9vtc09CIfgbUAqenXNmLWa/FIp9o6kqrFSGR+bWOgyVbjlXYTViQi/YkkmgN4nuRNuDzeaUVut6WfNJbOWsR7eSxz9P/oKsELSmLIyjZvgJ1A/QIIupJZbniFHiwyyd2YJyILzcmPnkJlma8+F6KdwChfFPdkycPh0kWVCdkvDE3Q218SpwRc2IEv8jAA8WufcB1h7wqm+t86FAiCo0ROFBAggibUtHW0UBqsxBNQ1qAEiVvRsTRAX2bgysNih2Ae5y0btS8K2Z3jWk9qbYk8w1KdGlEVtt4JM0iqXFFFJX/y0fzY6ngu8qy0ZMeq+ySVlPsOLLrUtSdt5sdMQDFp8YTuGbpblY5PBAelfRZw9lXTbeEesLDk+4E0WPsJr1Z03ztr5osOL5G09aMHtdieNF7zBOewb8poAYdviSpYtf7wh7g3FPbjwmljJRc3w+RM4czgwIpWzcJm9E40f6R68p8IJBsRMpsJVf4d4MZPpQsSkTbmx51rlLLSg6FCvQUBnefs7Z8bvLr7rq8iNXXXXEmxYtFjHt9WZEy2TXgx/60AMPfOhDD44FRq+Ymjpe9furx6feejOV/T1weYy7CWYZb3pHwNhC1OtZOD1go4VeXc1xD5H0XC8t0hC2g6ES99z3utSou3q69p4J/M9FA99z8Uml5hKCy5eBBz403HTKzNhCrOr6wp9XcoEmNZ9eZoKgnhUj617Bg9guxmZeNRcbGo5NeLOxA9X1l0UnL+t397nO1o73zrzu2FK0PJUWsqniiZX0wctPDnLaaYKXG9bms4CXjJaads3Gl6y9eFBGfemrS91J25QqDEA6m6owUpEum4q5+u6Vpu7OohO++f6B3fbcQr88mbQMyrFxXyq83p+Ykr0jBV8+Yhf751KRaiHec9qfqZa82aClahFSrmwp5CtIaUd8KDqw26bVe+IlMTGRdZlcgoXKnxkI/SdY73bAb7XJuhtbFA7OkWwREcWkYqa1oLiIrDYz353MevWcV+BW4qnJ9o6mtNskO6RqixjkAec0nLbLbXwnvnp1ofaf2Hxo9YWz7mk3n+KfGvv4x0v4PbXXkRw5rMEXAL8Imm6yxcYW1YT6GvCohxXk+SzgBDj2sEo14/sWv62iXWI+VOX2qHe+j9QKd1fErC9lx9XfmPkMnBgsBUJD84lIVY73vAngkZI4tttm9WJ5+jNdRmfvbKGwt+wzOQNmRbdzgPvXQX66UWZLm5IE2WQr7Fkew0B5K6rxT1J9oyy8hbYfW5W4h6LuaA57cJcz5jcPDPmnBcP0521lvziU8WL/3ujSMX9hIprdOxiMTazmCrOeoi0sC9WZaHrjcvwD3uMdOTEXTh49lpkt+SPzr50/cvtKNOoDrFKk3gTrYEJyK7vQZKXgS2byFTsgVzRlct7UqNFLn9A6k+PZ2hfx/f3TCat25cC7br9pbFqeOX3nXxxW3nOMA9++BnwzIR+aRKtNuc6t+UYJYJSuu1plG2XZiu6sIoMpZicBQnwqkQRoS08zN8LKoEZcrmD1VBXd7KgnsphgNKrja+3hvK9v0purBiPy3wrSQDE94IwX/9MuJnmp5M2NRRNjWZfZFTDZYwGrvzSXSc+XhY95UqKlkg+kRI/RPXmrb0TKjfZlfXHB9eVAMSl2aV3Tcngo6XSlq2FXJiZ0arsC0bwQqmTc7uwI4lAM+PL5Jnli4qNyRRUwKk+IaJAqT7omebJSJ8p8hV5RC6WizghXo8qfBkCQZob5OAjVAohTUBGn2NJRvzwZye4bFIk4yfNuXIpmNo6QQxXVmW/yXu/olfOR1HZxivgIVqObvwQKXgPW0olanADUkAWqezXAYFRyOSXJ6ZLyPknyeSSJjJHd3I+ehDHMdIymV7mI2oNzhGXVSY2nt7KJnEbvlEIuITw1fJ21Esc+rydQSI8ecZM9DcULf4+LottgipNIBy2B07m2wW+nOr4C93+LOwNzmpEbtTiq0q6wOipDZVLfQNFJbonr7A/U6eJGNjZqt71DpU4ZewZ9C7/QcmzD1tgBLEFNrmzbovNxNjaMS8Y/sLGBr1OH5lAE5OgN3P9Lz3XHmk5AUrExq2/IhLcOXkQq2mbvquUbKfqgzuyPBvnenDjIZ4XF0sJhpxQVLdrDQRfBwxV8MlotZtIZT6w/4EmGsiuzYimXzfdLKmrUZg7R8w6tcDMQ3DoAN3MTbthh1NrVsr8iv9poI0dwe1sPQS7XKw7wuTpygkW7UWcSFih2WXesz+9JSrkW2BGZ24v/Bj3AfRLWpBOREpwWAly6vCYsLWnWND15bv+hQ39FZQbuxWtwbwTdRfbHcP04HaMfP4jOcRD0giafQ21s7wujGLGdsZPlRM9FIhqdTfTzfrFU4O519SWx2+X2ZVJDa34Fnyrg8wXuHBurEyBtylh/JPjVSFjTZhf8vC9YLuB/cvWlsNvp9mdTQ6t+RPEFvPApwCuC7iH4wvXjFA5z4Fdz5wB+L4HDlcE3X4W+gG7gCARd/EIdhl8N+8Qog9G7oVXeXRhD/4YfRTrGOx3lXblY5vW8Pn1H6JprQt+v3O9/3xmvQlsJ7n+m4f425f5ombfrow9eI915p/Ti+/z3V+73UvzJ+PhauD+C3kXwhOsZCi8B/I0U/m4Ch+sZOn4cfQ+bsBtpKO8wwoirVw6k+IUL2F1VxmX3cTr0Xvbcv+PHsLP5OfrGcxzfVXsNdvaz+7jf7HQf952LebiP2hysAZvzJk5PbYuJzhFDd2MDfiOyUto7AEZpb3qv8T18LGCxBGK8I07a+N1mIeXxpEk/7fGkBDM7E/Dn+EebZxkeBIIpHqToi39UC02Xy+xc5l7czj2NjHROHZuTVw4o8/QkMq//58rUVEUe6O8fkD9+/AenT3//CueR7732td874qRjRDb3ou+rY+hhRjoG8a40C0MkMRqhj9OB3sMevuL7p0//4LgyxsbmldjMfRme5pl+6ADf5hUyw2fjU69//ae4L0MZKUvfb74S/Z36HGK6cA5pVYp52I7I8Hn76z75yZNcLfviR8lzQTZfL0qq/CHPeaH10V4P9Ew0PxWDlrzTQbjHHGVjCA+cgu0GDSMc6u8nkD9kbE4PDmg0GoMVzGHQamgfGEx74h6dVxC8Ok/8ZLZv3tObybhMHoexYLR7jfJaeb4vm+jNLMu9cru2vZCTlzO9CYVO3A34yiq+vRRLB/SstCcSqii+HdCKFF+eFsDUs+ZKNazINhtkae00EKhXzIr9cU+bRxDgEnen/CmfsT2Ud4k2A9Bg9AHgJEUtVwDU5F6KWrbbL0fLa7IviwtGh8fkymR6PVHZ351V9Llj8wjexz0Jcuhsde6XA3+NZSxh+1DtV0Pcky++F545tKlD7+J+gTREnuj6U3sLi3hoYID7xYtOgBlh3OWXHtcK4xoHcecgHRZxDePq2cht6sggJWYYnYxfK87Xcvi8gv8A/ii+EnJ6RmRrijcstFRCLFNZ5qk5i+qP9l9uell7b/sJ0+V90elp/IWNWMZ11SlnJrbh27+f4JDe/Dzsex8EiZqmEpulek/Wb4KtJMT8FcwiVJrhrB/S0xLh8uNWeedSQ3ECj/gCsaHpgE+IDU/POWHvF45XpivxbMTY40wOSqW1QN41EIul48Mzw/FcxGS2RvsiuXnfF4wRMZgSvZaeWEBMCj58yuYOuz0hvz85nMkNOsxRN2T27REx7fTGfN6I35cYzqRG7D0xiFVD9iBP6Itvfhyfp79X0IPCf+r9aACQ5TdLGmbjoMSOb1heXi0dHo+ERtcKKXgRX8B5V3Gxf3Cf7Kgt0r1C7+avuMe4t9Gzw5NIBi4OkOiScjENvQINFdI09q/vSEZJxfpp9VDmnzg7zDcdkGw8XdhbWLtmdPT164XC+utHR69ZK2RTs8f7+4/PpurtywLyeCgx6JRscXd5fLQk5Bx+byUdGct78fjcdfszmf3Xzc1fv5rNrl4/P3ZqPpGYPzU2emouHp87NZqcyrtFZ9Rmnx0cm3dZE07RI+8isqhFPK7ie7ivszPTw3SXVMiSwhe0QJ/t6fr+MU7yik/X67l5VKC1L1IUPotKFJaFNk9g9Mne80odbFDhzp94k17DogteikqkyK4cAfiH6AR5YXIiGp0k7WQ0UI7xfKwcEEpRno+WYnNCxmbOZKR4fDxp4b246ivMpFIzBV+9DfOJwZA0mHQSMQ0NJnhs8IUsAd3S1NTLgibLOJEvefMKjZe7DdZ9GVVh5QfVlY9Cz0R7u6DnoD0/id5or41ZzCdQECDsLEvTjoedxtw6bazXyU2kUzlQCl9UKKpY2UzJ46999/LcG1Yy4YHpYLy/jXONiPKkJz+dcseMXGc56l92ZqNOaerKifGbjg6m9149klkMOlZO7n7jZWVsiPntabfP6Lf5s6L5m8c/+rpqbvW63WPHxoORiC86MxlbGk/wrtjXXu8tzvcOndidKB+/a23XaxbiLluvnDl422ou4Q18gLcK+SHgEdko/QPsGT0ogO5vqkM0Z/QA0HwYpOWvBmhYBugc0kNPSzVMz1L3pDVCq2U5IA20FpYLsrHWyeBedp+PZArBMdOPrKcfu0Q/Uhk+Vri8du+qZeUQX+RvdRad+0jfVXSedpWespy2fKv/HQOPwb+Bd/R/61vfwm3veAdSzsngx9HPNBVOjx5DCPoEFsOP4xHNboA9rsIkiHurFHZWhSXgvr302ScUGOEl7Jvm8X3gL3i6N9FlXzI9a2mRni3VkwyOxyf27ZsYX1wcj+RykWgu13XV8WOnTh07fpW8Z2Fhfn5hYY/yTjxcf4Zvg9UKshmbzgEqcyFFWVmyVXpwRUzZCiO4kNBZTbXrqF9BeARwd6MUopWH5gooO/J0KSlVyOUW0lMZZ0xyJcwRfjw2OO2IFgP/T+2X4fhoKeuIpESnFEyMVLyFYp9AY0vKuyrMn0AVmJ+vzx8l80fJ/CRUaafVgig7kMJeZmTHUNpRx7a0aUQqQtOUNG30iQuWfEDIjPTL5ajgiIHDGo670kFbxC/Kjojsze3mL/cEQsGgFI/5vUGPLeiTjJ6oS8oaDb1RTzLQkxQRB/givBf43waIzCHirVr8ogTa9osSnPoDP23nlWNSevbuYhdTANN5uu2RTJgnCVGpSOo3T8OZ/+FnBwY++dGZgXdNXb3vaObKK8WL/w2yGMYcTnDLnA59hsqmH/5eoH9/lsol+x7WMQg2kIgjSZKeQzoWv8JfW6fnWTUVAcTNcqUIuZWj1cqmM4O31tuH6cH0aL5cpPH/HenSZVdit+Qu23jXZHSonNdrbfnuaHJ+GLAqJIvlDiFh7070eKZGvhy2ePKxlfaI1+2keDK8kRWW04PIWTKCa+Ovt7hb+5wGtXk0UE44nYlyoN4mMplEIpvFHJ+sxmLVJF9v5dFiaWSkVBylc8fxQ1jiHiJz01jLptqwDui5offS8/6lKy3ZbFLaVW/FREKED37IAjliP+TbLcGs35cVLeDVwplMWMrQed3oBJyXSSMRFVGs6Z3Djnrp2w/TK6/p9lIGVDGN0Ek8WVLOmIBcUy+rpwWuqL45SHGPmK1W80hqIlf1DfATqVGzzWYeTU3wA76R7MSQVUiBIxWsVgEcakqwvmOUH5CkQX4sPRW2WqbSY/ygJA3wo+lJqzUy+Yy/N2S1hnr9rFV+r+s+/Aj3M5RGh9F23TW2eK9pS4lhHwmgqPLuO/MJVlX2fPC8lX4Xhe8SoBZ4W4QLypHBTdFtUbb7MY9v8BbjmVK8sqsSX5oZyh0QIsJ8pq9MAXvGhnLjHq7L4y+kQhlJzE6WZg51XHGsLenr90b7srFiNJiq9o6vdx4/1pag+cgiQrQOl0LrTe/ZGFuciwKDq75Tw1H8Q1nF/4WANkktandmt97AgVge9tvKRphofAY3v3qjvvoLhOOv1F4tl7Wa4+OWDjM5DVOadeesuWCp39T+gfs4jeCfns57L+s3eowFrTAsJCZy7qSQ5f2jBWFQ6DN6TMWpgFbZt/BweQpy9z50Gm0/fGFsfq/F2PKdXC3ysfcwXKxHPAeIsHrMqPs8tMy1O1nq3dlTfw/STL9Vf6Kt4RcvgGA7fKgRlMlHc/WCydejNftMC8sfvesjH/nI4oc//OG3fYS76WPiuCSNix+rfT47mjtzJjeaxSOENvArP4V1K6Krm85IGVtU4pU9QYaqW4ZVrdLQdrA2/LQSrYTVt+2sDHVBfQk531iPJysZKRSHcBk6Fc3WahInqbPZvVhZUxIr4p+Ky+X0/FC8fWQQ48GR9kR1qTe94O6154WxfV21Wg3jp/6pe74SLrv3pXrFkYODSW27pd3QY0i1DR4eDabFrFs8utqrNxo6rO25lSMhidqY4c1/x0/CHkiHwqgMti3I3NE5omVIZPl/xbUHodURt8QWzQ0tT/SOmvcANFTp6GthTNWAoKgZ4gqsK49U+j/vm4kWyBvkgdLuRCE1wsdsabdQsOHvd6+Nja91t7uDmhNXtoWHlouFxaGw7uUnuKCjYHFwtQc4j4X+fso65Mz/wL2vxX6dmg4M+3U9lvFf4Zf31d7VxT1wcY3Gc/XnwO89Sf4mMYXGAeNE0aPkFD+0T+JX0wzfo/jrCNF7RuAcw8/pPY9t/pjdc5De8wR+f8M9v2u6Z4re8zgmWLG5uBfpPWfVud7IxnmQ3EMzLlfi73IpWsvfipKVWFfHXL+BnCgnsi4rcv+vjz46Cv9xsvKlL1W+xMbBSfxdfLJl3jCIH6ot4ZMVcl8e3YQ/x9lRJ81gtNHwiTiPaLmeg8jff3/q/vem4Pre+z/2gTPpD3wgfeYD6Q/S9241gO8RwFem8awHiTCXCUa1U8nRqf7R1eAft0JZTUPiPwp4/Y4XBJ4PBBboleeFD1JEv+Z38n4/7/TXvqO0/p9WKD+D6AtA4z9yEfRtdArp4PouxsNbACc70GKnWRkD0AS0QYuVGCVsN2F7ANvDwIl/uO22f4BJrn5geiGvzS9MP4BajAFPsjEwe1MdtxWruJjFxTb83doSGQSCg2TjIByMgdhamlAf+qMzTTsvL9zFDpFuX2g9MIkt9qOUNbnKZz9b+cRoRcF5CHvBv54BnN3NmZLGV+CtEv19W+kfd71nZrRXmx+d4c7Ubvr4rbd+HMYIYYy/jd8OqxlF5yjdJoqgieUDNKpfMjGttyqBc0NGg5iukGdM8GfjEbsjYs409HF/LuEW3D4hmIvTlsk8kdU2VVa1TFYhzUqkdaz2Tnyytvklml/uw/+ILwB+IdT89qRawrUhM30duPXrkVHeLi2St8n7yRuSnG8y4cuY2xa3vSQZTZvdB+4FGfsgN4z3ct+on0amcvcdgN1EYc+osLsAdg+FPafCTnPj+L3cQwC7oMLu4Obwu7iHAfZjFfYObh2/mbsPYD9SYZ/k/Pgq7l8A9lMV9vlNQA5p6Elyu2IvNu/lhtEozKvaQdVVddXtoEbmpVfuP1DlvvE25ZkbuS8iWZNX86EakG3ltls21iY1+VfBPZ+DcV+5fVwmS3RcjowrlWWuemA/N8zG/TaMe4s6Lsdy/1a4zTK5tsF9kYx7M4z7JsqnctNJJRi3BQG6HgA+rbxqqqNLSvAkP2QUnLhsaFw7PgRkHTxI5/8zmP8OTZ7m7iFmpDqrjKB/GkYFWdWUlYfvOLL3yDg8DDi9amKCPPt6bhy9BdbKgUZb0duougA4h6xk9G0/AKgocSeZpa4RZbmuEHrpLdpY2Cp4nF3e9gzpBjzObuhy4+nhHmuPrFwVPr6S+yq6VyMDFRLQgZCZ2Z4uaLdMRBe0NqJ/TOMkVQ/JtEM2weM2Gh3aTFsq1NDnvtpj64H86jA0HlemQua7lVtH13P30XUj3K+LjV26/te/5u7rR4jbfAM3h24HuaV5i1Yc0rFDPF3s1WoLYUSDsZcb+vsdHo/D7vW+Gz526HNzXjt0Gz5ImfOb6HZNWvEtajXp/2Kab9KGgFiL6Dwf4fzoMOgamccK83TRech8FirB5h3SL/FCIU4+Nq/XBh/OLycS+XwiIYf9bpff73L76+OfQ4c1p/5X48eKxZjV67WS8c9l47FcLhbPhjwu3uvlXR5FXj62mUQTsHpmOjrH5KUTddDRozJLaOijysbBWoRStTNW9BpivaJfUzQIAYfg6NbE23ibzW5oMSZcOTpmB+r8344J9ov9dg7EXhGE/n/1rmepAAEAAAABDMx3DJjcXw889QAJA+gAAAAAzZeApQAAAADNl+MW/0D+vQSIA7gAAAAJAAIAAAAAAAB4AWNgZGBgvvHvPQMDy6r/Dv/FWDqAIqjgIwClcgdKAHgBbdMDjF9BGATw+fZs27Zt27UR1Ahq23ZY241RhLXtBkVQm6+TzfkuyS/zPe7OH2o1+gMALhGpQJioWgxQOUhVI5kOGCDvmKN5vIqzBb1AgCphvuL5+7SUtvF6YXOWMd0RrTzhqxajrzIBTJzhoQQO8sQwlA/iFJCtYtBDLiCemSAfEC+h8JcfvBaIHviGEnw3Hsh7zgZ6mOShhwohf/B+ZjifGcjnd8NXesCTz9TIbdiqa/CUM7AXznIQAbpPN0zSYak7sl977AdqFAvjN/1p7dZVskkDk/3a0/3mIEhV8Dl27I6JE7KFvdm5A3a2pWx8M37RF86tnbtjUsEs5HV2b093JxWENP1Z5cNf9aDhCJSPiKUGCiVvciIvSmy+FkWR+IkiSiCdEoTe2kGEUyLPNTIbmVlUwDmLCqBnckAcpctR5lFmBqK0F4gyCUQU+F5JRKR8ga0koAeF0xAKoGCdQ2CNjxjEtNe5C9nyEXEqEVHqDZJkPzz4HaTwXIhMRF+KpBCKpq7n+iNMxsO/NWt0V29KpTTyaJaHPkZ/kEkmQkzsUGgSSb4I4bXgZsk6u8qlUApW4fq720HXaA0tomW0iU7ROcBYT7PpLF2luTSNJtNEWkzTm+1udli7BID/k/8nHMTsAAAAeAEtwQPAtDAAANCz1artWLvOa9dv27Zt27Zt27Zt27ZtG997KpWqbsrVaqhuqJ6rvq2Bmvaa1ZrTWo22qLa19rwurOus261Pqx+of2/AhoaGuUa3sb1xrfGxqabpqpmY51p0lrJWjbW0tb/1uI2zFbQdtPvtre3PHXUdS53QWdu50fnW1di1l3EzJZmqTEOmLdOTGcpMZL4CAwBAAARkBHlBSVAVNARtQU8wFEwEc8FKNis7mp3OLmbXs7vZ41yYa8y153pzw7nJ3HxuNd+Qb8v35IfyE/m5/Ep+K3+QP8u/5X9CC4RQggrMCheinGg7OozOo9voOfrs1rkHup976nseeqd67/r8vsa+q37iH+x/GQgHVgtQyCzUFVoKXYXFwnlREtuKg8Xx4lbxtvhc/Iw12IXjODcujivj5rg9Hoyn44X4OL4d1ATDwdzB6sH1kksiUnapotRamhwyhbKGNoYzh49GCkfKRqpHGkZaR5ZGTkYd0abR5dH10e0xV6xkbGjsZbxm/Hgia2I74UhqkpkUJdVJS9KbzCebyWUZypKcX24oz5R3y1+pixamZWl12pC2pl1pfzqSTqZr6X56k75OCsnsyb7JrcnjikHJqjRVpir7lZd/fvwNsPeMZgAAAAEAAADxAFoABwBxAAUAAQAAAAAACgAAAgABcwADAAF4AY2Og07GYRjFf9ndQW7I9pBt1/TZxh/zuuymzp5l79V5jvYCTdxTR019CzV0wzOupVOTYd2HjD7j+neeBh7wnnEjwzw+4052awa5IE+VEmGi9GoKkqMsdEpJyrnYuPSM+BJTjDPJnM4Ka2yyzpbQ+4bn/HN67GP+OXViuWspJTmT8uXo/dB9qp7eHxxaciTEVQib4vCs6SyYmiVImqh5YoyTUUeIaaE5O0vW8v9/J6hoFVhmQsu1NS4lIm+IqHBYKCutgi9fVChjDUG945bPknkC7V1CCAB4AWzBRQHCAAAAwNtWBXeG2w93d2ISjARQgDsh8H2L/fMhCINIJCEpJS0jKyevoKikrKIqVlPX0NTS1tHV0zcwNDI2MTUzt7C0sraxtbN3cHRydnF1c/fw9PoRBA/dQihgAADne+/atm/9jLzMWmbb9sl2q2zbtnky19m1bsY+t9ywX0+9LNPbHX3cdNsDd91zX1+PPfTIAf0sd9AzTzzV3ycffDHQAIMMMdhQGw0zwnAjjTLGaGON89x4E00wyRSTbXLIfNNMNd0Mn5320ddIiESp8hUoVKRYiVJlylXIla1SlQRb1NksTaLjTjjqmKuuOe+CXXbLiSSHHXHFXM2ly9BYU111l6ebJt7LNMc8Cy1QLymSIyVSIy3SIyMyIyuyrbM+ciLXTOckm21N5JnlbORL8cdvf2X55ruTalSrtVIzPWyNAi0s9sJLS7zy1rsojKIojpIojbIoj4qojKqojpqojbqojwbRMKFN51at/mvR4f++3f41NRZGDMMwACyTZgnjcXGPMrdhXD5gmRLb/ye8dzedZ/+Xrm/2i+B/U5MsTBZpERBZpc/4hrHlPchitPoWB+r1H/z6BpS+oRdsIIoNvc1OpG+GMyeHLA76laZpOLh6i7R0HSjVyymkMQMopZgJlFLMAkopZlP6QCnlNkApYqYHlDymb/TpS03jTwtov73SAAAAsAArALIBAQIrAbICAgIrAbcCRDYqIRQACCu3A0A2KiEUAAgrALcBUUM0JBcACCsAsgQIByuwACBFfWkYREuwYFJYsAEbsABZsAGOAA==) format('woff')}@font-face{font-family:'Source Sans Pro';font-weight:600;src:url(data:application/font-woff;base64,d09GRgABAAAAAEPkABEAAAAAh/gAAQABAAAAAAAAAAAAAAAAAAAAAAAAAABHREVGAAABgAAAADYAAABAA0QDckdQT1MAAAG4AAAHKwAAGWjKVuYJR1NVQgAACOQAAACJAAAA4PFn1ldPUy8yAAAJcAAAAFMAAABgW3yVKGNtYXAAAAnEAAABRQAAAebzMPm1Y3Z0IAAACwwAAAAoAAAAKA27ATZmcGdtAAALNAAAAQIAAAFzBlmcN2dhc3AAAAw4AAAACAAAAAj//wADZ2x5ZgAADEAAADAwAABc4JLzHaJoZWFkAAA8cAAAADYAAAA2/iCz02hoZWEAADyoAAAAHwAAACQHrwOeaG10eAAAPMgAAAHjAAADxLPOIg9sb2NhAAA+rAAAAeEAAAHkDsgmwm1heHAAAECQAAAAIAAAACADCwIzbmFtZQAAQLAAAADgAAAB3CQ5P0Rwb3N0AABBkAAAAgUAAAM5bFBIb3ByZXAAAEOYAAAASwAAAEte3My2eAENwbEBQDAQAMADgH10RmAMHSUyaqbJ36mwAmqj1qayx8rhVLli5faofLHyS2o5VgWBWga3AAB4AbSOM8BXcRiFn3s//G1jybZbclNbmLNda1rD1pptt2S3ZNu23dvJNYbv5/ec5z334gB+GtIRt2PnLt0ID+01djh5SgHM5PPzrekO6Td6OOGvJ0jRFuGCtt9pq6q6auR2ZiDjmc1adnOW27xVgN/JOpWdujjyfSTIUaAJzejBSKYwjZksYgk78VKwmzSxGzSzK/QjQn8qM9Mes8j2sooQZYiQ+1jqO6k7pWbwqueV1Ef0IUw/atKf1iLeiDgnoimlIu6JeCn1NMt0l5Kzt/Lvq3onpiplUp4rX4xUMVKzlIg5zpIvXQe/51Aq5Qqr5DjqOomrc6N694vPySloN1OKS3+CPCb6hTisbmlMsPdM0naTUymF5MV0XVpRH+yxrbe7OjdTpKMILxU47OR/zrtne4Ei/3nYVtC58TftpuaF/5A9345oTgea/tDW22Gd2+y4Xtf/wzdeEv3xPv39/7Vv22qidv+vMk/bs0/cmDWYKzcQx8PMzMxUp0pfpi9Spg1U6cLMTbgJMzMz02Omtc+89ls7+8nJPn/fL7uTsU5aK2e/C2dM0mo085f0nxndFXyRdiaMEWEbRjyKMOJJ1lDnaTsvIeIT3iEhCVtWbJnaNiR2ZmY1evJdyT8WvXrVOWSYoO2+jKXEnu06hs3atvjz6N7TzsssCsFBNImfeF7Pn+cLjVI/mYnbCYaOoO5QY5O0emNrpL99U8u/v2fgrDcttIlkpMNqTNknN7GZL3hR1rWB5TwyRkkkK/5Jd/UiXuIBHrDzXmEVt/MITZZL/0eqwT0Z+RzDuOfG0uJbZ+7C0OpYoe20U9+24HN1DQOxLx/6zonuGdpv4cpGKmopKdlOAyfltekXnJ/h/H6ZeCIxQ7WIcO5mDVv4kTrv0edtrsufrdDccwivTHB4jpgtvMR3rOBz3mHNhG3jMKePmeci33mazcB+G1Kfvb7ws21tKE5XXsI4vnLUTpAoDYjGpjC5jJjPJBqT/PsBEno0SXgr7y3hIfGXC9/vcAxfeozQs2MDb/MRa1jCd3xBxVnRiKTQZKBIR1QwxAyI2Spn0FP/h9hZI4tN5lClGs5V1G2/E14zP07Ec3sqZ+JxbpL70C4TMdT02LhvgKUyh9rveohEZwFh3SJqw9002SB54niqxFzrDdt44eLS8xfyz13cS4MlC9YGQ4TN9W60MrLtKjUyh6kdlCfKmV1kzJdz5HMOqVo/1mGZ1Avr8by8fwFDzSnn73BEfuc7Nm/tkTMzw1ZFGT1LepH1fhaZH1f0dETzINtyu6fZ0dWhWFR98a9RMcBg1NeeHk9ayuJhMUZazqwW74eKMsXqUFdWTxGNDD+HTBG+0GgW72T2VPUsfZ46mLoFboY6q60sH4UqPnO2dZ+9o6wjCd2rWMJVXn/KCjByQ96HlDfp6l7X5PlHqrIHg5wr3t4IT/bRyBZWYXx+e9WxN1tcYkqae/nj3jlbfw5P+pbfI0VwrDMa+VmDvmfBIqYy9cy7i7j71G2UKHI+IXP4nVFx8soRNr9kvh+f9X6dd2wv8e9zXm09hWHAduZj9WsrGSNfy18NRrEfEKqUGn1/v2hVo1N45wXW0aSCkYqfcEWYmeW7BDExTclg/zuhXuqvXbSlrkZwyz65gR8ki0ic0+bSmewkmk/iv20PzJifDCUfv0PCgC5G6thW3pstJ9Ahpk/L2YHYYVHiVm/amKKvt/owc1M62t5CojZC2fzCMKLF7yCJ+H97ey0SOegjkul+6PDWZK4IS6HLZ/QW0FgjO/vFrOsM/20WXiXfezVRhQr9+XYYtbfe+kx45v5Svl/HR9ubRQMMCOLlct7R1gYGs6yDNPSftHBNk17m17RZ7VMv46XpV1Wy8T3m18LrWU8OKACj+Bnbtj2zRpw2aWLbtlWuJrZt59XyCPnKeO//d9rL6k7/q/npPX++FyMezOSBIhUy1KVAk20U2c0p1nOGc+xnglsc4i5PucALGeeVTPBGJnknU3yQLgbMvMaKDTtuqtRwUKdBEycu6uKmpTx48eGnQ4AegoQIkyECxEmQJA1ACkhLRorkyFOgyCglZlCmgvUnVexKpCGCVvtJFb8IQdy0Cf8kTkSJAAAU0co/idNPEZGyQIYEBf4+YhIVA0aGMDMsWUZQAAwwiKIPERMAvT9lwcgCzOLGKj6c4sMjIYLiIyw+ouIhJkbiYiIhWZJiJiVl0mIhI16y4idPkSAlCVAW3aGEqYmNJi3qtCVIR+r0SJBeqdMnDfqlyIBEGJQSQ2JnWCqMiINRaTKDubiZLz4WshgfS1iOiRViYaUYWM1G/GxiG0G2s4swu9lDnb0cJshRThDhJKdwcJZzBJhgEhdTdGlzUVJckjSXJccVyXCVa+S5LgVuiJObUuCWOLkrNe5Ji/tS44G0eMgTqjyVOC+lw2vp8FY6vJMkHyTGJ4nyTRJ8l/APlZVMdQB4AWNgZGBg4GJIYJjHwOTi5hPCwJeTWJLHIMXAwgAE//+D5cUYmB2jXBUY5JyDQhQYNEKCvBUYjKCyjGCaCQhZoSxmBjYoi4WBnYEjJzM9kcEIJ5mfnMNggUwyMEBMBWMmoFlSUB4HELMxnALLqjGwwMWkQDRYJQ9DJZBWZqgCkoZgdcoAVmYjbAAAAHgBJcWxYQFQEADQd3dJApUN/gwmACoLqAykZTOgNIENAF7zVOxz6Zdcv2bxPVaG0cj+X9VPZf40/hEA49lkDvdHXh9/1DkGTWwgLrlFU08ZkAyKAHgBXcsDjB0LAADAeYg/amNzsWrbtm3btm3bdhvVtn3e2HXjXu2JMYgihmxiIsgmHnmICgJx+fCvwsqqop0uxpnknHvSvfQxUjTSPHo1mhRNCbIFeYICQUJQLqgRjAgOJGRJyP7xI/hXoIhyqmqvq/G2OO++0KvPu9n3nTXIFeT7vMv+sSMf36MBHy/x8Qx8HMPH0XwsRUa2jPiHj+FcCCeGHcNpYdUwV9qptGMioAxaoBN6GYA9uIwQQJI0f4uZZ7GjLrhuvEW22ma/22ZbZ5bd5pjrgfseWuCI7fba4bEtnttpj12eeGm+Ky4675J9DkhxyClrPLXeYanOeeS4ExI9c9Am6Y45ba0bbrploZbqqKue+hpo6K5GmmiqmeZaaGWI0froq5/+BhjonkGGGma4EUYaY5TGBltiqRWWWW7lJ/Pae2MAAAAAFABeAHQAeAAAAAz/MwAMAeYADAIGAAwCPgAMAn4ADAKQAAwCyAAMeAFdUEVaAzEUTr09AVLL+3Co7NlgmcHd8nC3M+BO7SyPXdjNwbBXl5H8+sWMEIlZI0JL+tPlKqBx/b0YoaJfIiQ8+3tJI1wJKa1LRa4DFu4EG4PAzJOQNnl67BXdhTIjM9MnGWnLi8MT8vaUkIPTDKYliVV9yeOaBhrFcI2eIg7zPN7iPN7SPBnkGa4qMzCyxRP8cMmXmJXk6V3Sy5ruVZhGFYYBpEXOkiZHhQGRW/7aThlvLtsrew4kOBpkEizPsqppNEwCM5my6gK6z2TCmS6saiOcFsMlWo3RimFE0eGbsIzrfokjhi4IF40u6AIMAypeO5SYXdWWCgNg8h8k6mvkAAAAAAAB//8AAngBpXwHeBvH0ejuAQTABhBEOfR2AA69HQEQbGBvIiVKoihRxZYpOZLjKom2XCT3luZUx46S+OVP783/79hMeXHs9Kb0/C9Kc9yV3osFvtm9xRGgwFR9OtxicLc7MzttZ3aJmtA8QtwM9zqkQhrUjAzIgg6hhxEa2wUfaecjCKGOgUUG4AiAQ+qBxRXUBA1k7CytII3S0iotHWs9gloQ/v0KamffoaOOR1AnwDJZk9/oN4lGv1FrSmPBgE1lLHjxPP5zRffolXj7FcnulDpduiyVeTSb4l537jI8/43KfvyWb1T+dPjgwcN4/ILFxQsQAvRXK/CR4e4B/I1AD0NWT5DV12DfRgBtSK0AjARgZOTAL3qKeqvcAlRb04+gZooqFlNYFPTYgHlBK5gkk4Av4TvcerXe1cE/eka65jt3f+d63DbXdUWheHnX3J6PcPecO0pwU6HtDDcTcqEQ2rMhb00EYJKREeAXE0VBSK+gFvhuVnjnBIQADi0XbZmgFZL5KeU8nMWs0Vo82GLWc3FsyhXyXeE4tiiN7d+QdpSFYmliU3hkb+GJz4/1Dw7uWJrcPDO9xN1jDA0k03MGddvscHYyw+Obcqlk5H8qPyh0S9lvIoRRdvX3XJg7hbyoVKWjiaDdVEOHkwCcMpebALc2iqUTWmYZy64BTspZeW0KCwE94GuVcoUir4dvKS6b3nZFue9C75C9FI6OZJ0XbU8MuYYiLxooXzaXwmNzN+1K5fyTdre/f1fxqpsC3olUNrFwI5EByudB4HMLMDG/IZdbCKCFYPcIagV8ZJ4aq/yzynwrMnYZt3/o9e+Obz029T/vOnjpZQck7p63vnXrie2xC45dd+0VlRT0jMm4+BkYtxXNbjhqKwG0klHl8VRUJ7TwTCudV20aeEixaYVfdUw7LH6LZBEscN+OX1p5/uxZ7Ktcxt2z9K6lR5eQMvaPqNyPbDh2MwE0Nxq7uW7stRGNwnb8sspzv/zlMgz2yFLlV6g69wWY+xCa3XDu/QTgX5t7PaXJDy0rHdUNT/vpqG4Y1VkjDx5unUDkhbxkZEKRmT86HOjJRAyD9t5UYkJyXXzAles48WTgitGhq7YxudDzno5Jm0eWDIP+FRdWfuJPgnAwPnGjdI6m/ok5WkHNysw0A54qJicPI/VpeA7u2tNEWsBySUYBPoXty7h5ebnyJ2BY5S9Ye+4oDlR+JI+LPg/jqlBow3FVBKAi49I+oa9lZjwwsq3+Hj8A7xtRvPq+ljyuVd5XzByVaBXlaNGDgYN5QA6EWQiExcPmLk8qtWzNTEtjF3ofCPq7x/HHKuXklt5Ad7HKnwUYp+3v2Cc1Aaj/njnVEYBO5iAH77ZRcwqih3TQotRJJklrElRaC5Cpeu6WL569+SMXEKZ9D0dXK0dw19X/V5Hr7wE+TSi+IT5M+CjdHNNfI9GVZdxROQEs/PYSqtL2eujLja7csC8NAWhqAG4CcCN1Ay1akT0lpa1DaZngaQ1tWaFnN6PXxEtFMqdGkPGioFUJKlHwqCwwyTdfE1fHlm++XKXWcOr8kcKRLjXX1KS6FHjxqRMn8DCIUJwHLeDvJby5l88XC3zlO1V6+ok/2Xiu2Ew0dC4NpF0NDR3FXc/8DsM9hYkQUaQxQfr4V0bU6tYr2Q1QfdXLpKuL2AXIXvWyrqPFyhOIo3biKNgJExJQHtWZXQUhBWUYX/YL8AjzaJlsE+h8nlkFYo/F3AB8DwsBjcXs4fDWwRfPxGIzLx4cfPFsLDb74kFxIGHbs3nzHltioC29eOPclpMLqdTCyS1zNy6mLw6UF4tXnDx5RXGxHFD8RAT4p0fWDaxmvVatgIwjFhC0gTUwKF7DUut1iduwCKYcc7NnCnuGw+GRPfnDy9961eTkhdw95sxsqTCXt/8azx4fSD+JEOXVDu6lwCsDyqBNVVx8ZGhfDS5WArASXD6OkgiTeYwDtKPqM3zQaqctK9imGLFNwCdJ4RrWY2IFUqpapmqpkSi+G5gmVhmInzAWAku3+/ILA/4qg709W3P81owVgp/p0t7BQGBwbwn4ee5Dupa9k83FheWJKrtHjyx0t4oBTRNS+NwDfNYh2z8fkymKRonFCKMWSqyeucUVxBMYE1HmpU0qAVSsGvyYPr9vZGJiZN9vXnFw+YfSrsHQRO+b8N7u7lLi3NcxWkUwE8np4qa9ZubXwmBjz8EcSIpvYAyvwTJDAJk1G0uCRCvKABryPLgpLASPWAGmWOCuFJfGMttZxMNk2IvBCjDj/GJhd7K7x+AMWXfb0mHe37M1U3pxuNc3Gkt3G91h86I7I5h9/TuLwzek3+aPpsJml6WzWWzxxLqD4kjOlQxPOzzRoMlpMepCbb5YMRgZz3tzKUKbAyFOB3OgBerqQ3hGm+JQ5BnggBZEJamJSJGQ95No4HufxL/85DKXWFo692057pgDHS9Dv3bUt6EGtRNAew3ATABmmYnaqsFmgqo4K9D8uWWTkPF4ckHzlYGhPaXSniEBTM3d+amEyZSYyuNrK6/eemTS7588shVwWX0BaHQBLu1o8B9EfecjxyJ/hNrptKkkkwfz0gAGo60S1Dfs7bAb1Ua7Yc/xL78EUPhY98FS6WA3ngLvjFEnQqp2GNePrvnnx2U8aBwxUQBPAPwaZn5qmA3wEk9bLuRX5F/Btg5vlbxUAW/Tebyk69CpWy2tcxfPtVpb1TqDrnTlyVdt1Rk0aq1Bt+UVQNVD+cNdXZd04enKQ9IlefiCp8Ge7wpvCsP/yrvpfLcjEA0634rPwQRPfH78tAawE4C9hjQDARhk0jTwrl32nABqlQlSSbxMRxlLKpOgrLmEz3/wlsk2HhYmltaxGz64cstcu92gbufbNt+At+L+V/MZtzvDv7ryqcp/n3JIHo/kOIUQkw38F8DbgAb/ebzXo2mQVboOOYvw6cdPTMDiT93h0U9c99nl3+E3vis0Fg6Phd5Vufh3hGcR+Pgjjc37UX2ot95Pyyqhkf3KaXgMtBArWljEflBDDJFNBPdVfoo/VHkMlyu5cfz00njFsURtbRfYsC/gPyEnEkEa6+nQbTwhTBpZUGOgE0KwhHcVe2uBBwwUbgeYl6JlgVYAWjKqghx0dhWKYO/AwGnFAUxWUkZBFKjnKbKIXq/qeul9LRa/PTTmi0Uv7t23+4JmlX9TYmjrVaLXk3PmRqOdt29ZMJoEp8FlnnWELz9U+WHBFd4y7fe4HFdozHYf0DoFtqeZ+wyyAgq3rdO8zvUR3RrxHQTQgdTrvaoS0XUwaeygcZy8ruYpkR3QctCWBlo+hfAAM2IWLSHYw1UtOmQImAMg7lc7dUzlnQqnJnMOoWcmGhjzqlr2HHNI1uxwrrurm/vMQ9ucAWFob6l8YDTo8+RO8J075qe30bg8CvP6MZhX9/r1doNZrV9v25UsS7HO+9Qvt6OR0cVsZsJeNEf4vumpHo9kzgjzsdzOweAf+vcPC37blNG0uGl60W6c8voCw/uB/3HA6Szw3wycuG1DrJjlW5uQTgLoROqGJlgHvzSxCLQJdVL+dwIFNkqLXuG/juU/ZEo9dfSJeale4op5SnFc6NkUifW41ccu1Kmdk5F9lx8vFYoZR9aeGf5D3/7hoNu+5aPnBvLO8F03zG0vPsV3zlOdIrz/LtWpANq6EZ115roNfuFZnAi/AIYdMq6K4rQpUlNUQjGt34NrpiREQzZtNFzeFs/NSI6myhd0s32+olMUFtLSruEQnspKiWEdRT0wsCAFzCOTJuMk7xGGL7ykd7g8gDDgjPBu7muA2MKGmDPl72yYKDADopyCvI62Wk+TbzAdYJQwIOnGNFdgFPLFaoSDR53Ogf37jw0Pd0b6E460kTvUtj2/p3IVftVCkxQeSjtatCQnFlkV8V+Bt+1oDHh7Oap3hms4DhHAENKsxV3d8PsQ5XE3S87Ba4BU6DRcAAlTSDdA0gBJA2RcyUHMUM7zIBsscAyIGm1XgchJnt0YHUCYB1fVWQ8tay8Gi8YiFDmZJivT3SO9fVPJ2EjEGd4fDJjsfotd9OXHRXPI8OHdbR3u3KjYETC2tgvJnYt7QqMX9SU2FbzfFHLZYDCXm/SH0j63WevOe/MWdVvE404b1KaRRGw4yWubdujjjuhAzKrRtpiNFltpMDiYcnb4MniflElJuXRGor55Cj5+TvVxF9rI+CnmsbHl17D0IswwNXEGxcR11pk4YtONsnwahaljau8macvksWDUl/UeAzt2oTt58QWVr+BQf87rqbwHcagPBv0S9zhqhS6TqD7k1zVca2lIzjQOQwbCMCcwJAnpQbFv80YiXrgWjx3jhoJudyjkdgcXKhY5//vN1Rz6CoxjRG5UXOcTdOsjXQ0Zx4Y6YRyAsSwVjKikS5jxrkVherkj1Bd3inpri6XVz/d4olGPNxrFZ8+9IA5nnbqmBVVTMs7trWImzwvHwbwoORUlhbLxvPzzORVUk1MpkpyKCDmVqSOqh2589yMnb5w9wn2mMvHEFyo//M6ukwgTn4n+CLjoz8undDbICICdUDhilXpwkdcIwIupYx3Naq3WwIcHctxnzj1kMXILLdODTAZVOtBnP7r/H8mgbuPQl3myxvl5TomFm1mLyCpw7LS8EAf7pThsDVuSyQsaLyOGl6gM83WSrF27T4F/8E6lCkNGYSa9eXo5FE6VjoVE+MBnR/ypVCycq4p3f+U97IbWdBDoN6OL/3n6FR2snwGFXKaUTCEZMZ0KMfUKyfD3zyoaCVgP+1J1Cqn4tRcAVxMwJt/QO9QHwcBUxRG4630Xs4fU2zKTGA0PzadSOwZFcXBHKjU/FL6mv1AYGCgU+mlAIQzv74d7IAD36cXF6U2Li4jGdANcM+DEYro6f6RbnynVrS9HaRrWUJrgXRPLyJlYrtuqiIhJiSmalZiuSfHOPDg2YHF9WFcb033iGBGWmWhqQo7qfIMelTPHQrrIKe4DXY6wMExiupGgy7FNCelo/DSAz1JZ8aGbNuQ/kwXd+nBJ0yCggviJRU2wUGSRVDV2MvxTsROYEOLGN4idyjRiVTvHw+fHTp70ue3/tdURUGInRR+OA41GtO2f14f12m5c0/EzsvAb6oSfRbRM8h3TomAxt1s73IM8iP7OjNRyUK1Odle+zfLorwZ80oBZvaisDZ8ggER9jidBJUiRFoAF4J3EuhwPzawV6jM8/FqCZ7N/KBwJd3rCJltOtO/fLm3z5B15d0gwukJmey5qt2cmU9KB4K0Ot99h4DsNumZbIOkd2hbkx028lzdYOvTaZj6QDfiKUZvoBnrMQM9u7gjM8Vhj/tbpgmxDmtn6UmapCu4mtlqgmZ4ylmi1yVwTi+Z3j8/o9588OdbubDWbJXe8N2Qkkdydd+55fkHTtEPb6kgPRQCfUcDnZ/isYv/qp7exOivz3dD5GVl2k1SOm6nXAxVgoSjzTARR41rMnx89Foz4sq5jsKj0kcAEn638oD/nC+K5imUyDJZQttVtCBFcG+SL1q9eGGDjfJGpNiUgtF29ucXSom41t26+8l134LOrwelweDq4WrHItRyEuG/DuJAv+ufHZT5C98/mi/ys/G4AkJnlixDyrce2Du+1NIttuVfXqoXMUPPgJUPNkDnStml7Lzv5kpJOD229rvt2oEoYDwbHhVV2B+qeEsZEcVx4EmFSP8Dfwmcb5IkaW7Z/Jk8EsSMlxMwyRkCIWIO/lq8h4L2vOVFs5VvVzebm9PFXv+lEX5u9naS/uq7F6Ow+c9wM//f9+XcXWRMWS5y/SJaH1QKVBzcaR/VBl269rV2Pc+3CpB3pGHo8NQxlXJcw0quoCylrTc1iuq350at3tFoAT1Pz9JUfNQTT/RGzR295EdeUDnnxU7/xTQiBCf9vzq3GhtP2pqbdBM8AfLwVn22QT9L9a/kkmtZVSRZ/4Nmv4Que+VXl5gG8bU9/5YN7EVpdlfdSqLq4MFIjhLTop9jSEP6zDeBP1cH7FfgzdfBeBf58HXxZgf8c4ABaPYeQysDdA/RY0YkN6whMQRsX7Dh4tJ0KUQtrEVUDFikBgFZJNHSwrJeehQwWeM4k7wExUiFrsvjzkOHB1fI8fqnaluIlzl654dPbr8cT+NHKX374Q9yEM9LBzCVvW3rve+WSPdAI9R6R7J8AGjWUxl/i3Qjg8h4CykMd47m3IfxnG8CfqoP3KvDnZfjqX+HbV+jzbXL/6BX0+c3wvK8G/jMZvvoHgIdq4E/VwXsV+PMyXK5J0rk2srmepnBaC6T9dDK6tjaE/2wD+FN18H4F/kwdvFeBP0/goBHJ1TPczdxHUTPAE+iOqtREiExEGu5VWUEIGhEaOXFUBJrhSacsNKwFHTCPuYI62dOQMVSecyswn9KKsV9B76xyRS6NWeqAx9Ucg6gVi9XkA1/kfxHtJxU5Uzw+uHlPayu+1u0h1bgDkwP9Y14/qXgemhwcuDVKSpujXZPxk1eEh534wtgVo6TiWdyS2ZU4RqtyPdsySK4zAlNIzdMMOO7YYPdGfYq0SYkbtaflzFaHUgFtUbKidlltTrNdHDU7VuRlQS82Aq3Z9MJ10/gdlb+eOQOl+U2Tx+fTx4++KBi8+Bie3nrd1ijd0BLbfmLr9UfxNy+HeZXrS1QObExuJlAj+M82gD9VB+9V4M/LcLkuQJ93MbknGsIhD3zsovsgnGgR1QtJo3pstVrm7JCNCZEVJ7IqBqaZVbYfRjaww81wtwOvQiz9RHllqu45IgG4EXceuOzyJeklL+m5D/YdLU/99+ws3Wv0qd7/5TLy3iNcLKqxC9XVsjuAvlJ9VNFwXj+ObNRLq+MfRxy0eGjR9E5tDbtYrWEvHz6viH3Ykp4tFee67Hc/dbw/8yTwUq49cF8HXnooL3+CTagR/IkN4E/WwVcU+NMbwJ+rgz+pwM8SOFIROEfW4m2oD42gN6J6JujWz+IaIEUAKcKmenNBASll51QXSjFd6GK6QDTfiiJsX5NsJfyKjpRoywqtQdpyQ2uEOhRWF0nhNNZjlZKcOD8rHBbpJ0lSplRkydMkr/nfLnrdg06dFnMXbN4bXSpIYybPpkx6ImNvqnxbs6ng63JGhR3pnnm3dGFmaOtL7sXNZr9dGPNxfCYT69e+wu/xOo0JK9+ew2pHOB2LpYWBnZBRHhy3GGd5T1ZIRLYMbdnZIVdkNvWNDPdT3kdXRVKXAN77KO9/ga0MDjlzOrcBJgutDeFPbAB/sg6+osCfY3AM375Knz8k949uo/ASQripBv4EgTN/ZayBP8ngZwHu5FYU+HMAZ+tW9Al8tsF+rfOSDsp+raljsNys0BiG5WJJv3EmwwmFnhcoHgnGl86G8Cc2gD9ZB19R4E9vAH+OwJEK4M8CXA0QA4qtrX3DhIRwDU02ArAxPwiNcI0f1MGTNirb4Q5Z3mVfVlhzZWueLLzOlX3ekyyM6v3eRGFEjz0ON8kKhcbdqVTO4SaZotnedOoyz+JkXsy7d092iZjz7SyQBJHV2h3s9y2QtpAoBSktv1c5QK9bQQN70RFUv4rUbbhXQ15LNrMWLJhOwzN0n8wK8iIexSht3jRAWDFTrt1005YXWr012pqXNQ/oq9VWTQN1VUo5D/TlpL4+Kdd3dWhgW6qw0+leyKansnYVV/mGZqbgy4OSLmSkXW7PjlJqW3+Qe0U2E+/X/oHlxzYRFhRy2SJRTaOrmelmMSvlIYt2SNZJOcfLDXFfAbwP1nv4mhxvNwDgAXXDAMgADTl15KcTH2TfV1CGteSwQKrbtgnzDpdHXZ1/Jhopdb5L2bUFSuIqbC0kx9K8w+dwRbNRF58YisbGAzEp5w+6opmoy+62h4K5/FH8h8yWktcR7/aGpGjYanbZ7KGC6C/FbHbrFjHssPkcNn8y4E1G455AYnKs4gL5p/kHqi85pl88Qg3gT2wAf7IOvqLAn2NwWCOy5wvMvqzWwFcU+HMEjjiUgX5+y/0YdaA8WkT1gtnZYOH7cSJ3ZLUSXwENjtLKhF3ZixMFaTWC1Cbgnjkj3yUljhCVnJOJaaGyfc7Kp7CskCR5CeqIO4t9GDtLnuKgyZ840jvj653vSs4NwI7QaQgneX0mcI/3ej7nkoLjvld6t+3Z3m5vC9imh/z5kJlr7nvnsb6LpyJki23XfK/X7pY2+yfzA0aPYcIUd0wVEMvlKvX5nRuuqJkVVZjB8gQN85u1GUyfnJ8VjOdV3fl8XX72GOceC6enJSdJzwr9bk5Jz35GdftDszZfNT3rsG1+u5KflWun3J1QO02iE2ij5X+MAGKos0EJC35UaqedMFFBknCDqxOgBqWOqoVLD7+2kF/Y9MZomMjqFfUV1roqK8sywo9VAHeV01GGsusjUHg1RweTrlynVe8ye4OPkDKsM1VThpVLsfHwYMrRolvQaqKxuFKXBdrBhhwD2hPoJf8obtqoZlW3oxleV/YmNp8m3yC7A3dDGvph+z1lqA+gQRYjh1jOIkoZEZKNaT9ezwWZQcWatCV3zCx5YpCkfNtaBdoKFen36C+88cZxSGJq5Izl83UF6QXgDctoEp/pAh5cwZ1CrYhHY+jl9dpbE1V7CMCD1A2daC9yUiX2KGGjBiC9rH7lQRm2uRxYcQb6ZptJV1AENbN0V7fSGmZv1gSM4GBMZuJ+2AZPCqnqveJ/rWyfJzxXlPSYtF46mHXm7YktPpcwpm+LdtvcXhHfGp6Likt92dmC2ywk7dBxoi88MmuLSE7PbMwZ07+4szvs63M8mOq2dIa2J/ksFvm0w1EIO4Kpc99tFhy+sDs7IgYKyZAxPBkI9CYcM2VfIRkxRWd9wZ3SwFHJ7CTnhSJgF7/P3Y1aae2pPnOtb7g3uE0p6FnZ3uC1oizdF0z3HFEXxGF0cHn5ILl40dXR4RJ5dm9731vf+o53/tdb37fb3rN3dGRPN8937xkZ3dtjRxhNI4S/yJ2AkUbW7bHXNxBxlkmjdWM1tKqqjjuIhGeytJBC+A0Gyj99z9WZMc/CycotF+C3TjbbO889cQG1MX74+DbwwYv61+3W0jeQr6qe2E/L+TyDkgcvDqhYQCKuGUOwi73Y4g8PX7opUpBC/Y7BxL7h1HSXO1SeT9t7+DdVto0MXrZ3Lij1xtyDacklTSZiOxcvyHFNS6xW8nXALY22r1sB6/9urcSJEsqhqfozPxBKAKR1XdVEyMP01WGvnF1iFn2LezrTNdQZnyyky6Ixlwj3u3uj+0qpmYKnt6u0Sejfmtp8OHC/Oyol7RGXYd7gDFsjSb9zIpiwxsuRwqRZbZgrJ0bT9q6wvF7uJPkxmOtmwGvnOuuub5BL/DjZG0BEMS6baTUoKyIzzcx3K7NYHadluIl4ZbJMkzOFggVSrXlSosdxFaduc+iX8YG5yyrPvPCSuXOvcs247BnHRxbf/vbNcLznFsR4/w3AL4Am6jOdNdh5CcCr8J4lK70sWQn4nFlLZCr8Zmw2+S3COrc5657MFMYG896Yu2jC879uN4v26FTBFxrYmpi9RHi9MzrYUxowmgTcc9GpllZbdpPUta3bnRcIPxOA7w+oPqfWtCdOEIwjfYOaxcfp9m0N5adfiW/iSmkvpKDqwSb5OJCCbn1gg42RYlfG2edsLn/YWHR78hE759okRMeyTl9hKpqeL4diY7uzPdvdMwO9pW5/ZMcM/qDFaSvuHhYitux0LjNb8IRmrpnbd/t8JOYBrBL0zMkJwERqbAfqLBP8yCpUst5LA6oi2aepV2mFd6ut0cFU5Vv4fd1jEaP66h2vuPm6wYvGx07e9rp9iK3LZDvYDlozokgiCytqxisTQLlez8ooxvSMZN7gJaZl8skk/xlZMrvOyO6CVSj1uEidBYsUyfpF2e/EuAuap6py+Z7RbKrXFu32ecQ3O5zZZChnFlLPmHxRPiS5pYlofCzrMNi8BmvE0+ntnsukt5T8H8qUcklHyG1rt/a91lHwxIrJmC3o5L/myUZ8bWrHpESSmvZkWbTFQ95WdZs3kvOFyuBHUoNElkLAk28pslQvOgpHmHDJskQ0hldkSbNOlkzUU8qpVnrYAUhec5K1KvB7Z7+jeSgldud3dlZFaUaIjsqilAJRio7tyZbm3Xg8EJmfGegplR5noiQyUXKHZq4GUdoBokSwKsN+m2+jVwEtNtRgd52KTE7ViwEG5bjPF4v5fPHhQCIRgIv0kV6dRk9CH5Dbq98/JPfRAn2QAECo3SiVzuSb3cmkL5rcuum4aTAcCvjEge7pyxyrqzJO+GlORHdC9wtIQ+4Ap+PUwe+Sc4vw/LPcAyDjRuRADbaFNMtsFiWtStDWUHNItazumA1WaeISR49W3nevQpnc9yx6Fv+J9t2OPo4MbPWTyXqxUDRJRdIRo+vb6mUVdEe6Il0OHT2Kd7xeSCQEP/TGYopXce9FbhRFkSqeIsFTlMXEqJwiEVliDrgfHlCf5znVfC0Zb24yuEN+azzszVkHfbuKkdGckw+IXpP6RBWX7wX7cvFohA8mXLbecNYSG4j6pFQs1eVTyOWgFjGE7wVaCX480NoC6BhpBlaElp9Sja16df3OXJBRtVjLBuxr6qD4iJ6cdci/qxAdIfiEvZ0qgk+MXNgc6gWEolYh6ebrEVI4RrVmM74bPc59GPjfikh5TA3BKp1KAxa2qa9R8cNc+ciRj1H5gGfxa+HZMLqHrHHh8+u0j258IzrNfQR10BlsAkirnE3WY4vCRUC+eDoWVWmcYsgdjpzgbrSVktjv9oV68rcN0X5G8F3oDLfC+mkFSJPcDwi3pka6tSOkH4cYdIejJ/FKfT+AJ+CD/w/gE0avInjC59cpHPrH7+NWAP5qAodPBl89iM6g+zgCQed+WIXh96H7kMhg9Gm4U31E/aAnH0caxjMN5VkxX+S1vDZ9V+rYsdQ35+8NnXpjUOZxAZ4/W/N8k/y8WOQtWvHdR5N335184VTo3vl7g4iODf3j18DzYfR6gid8fpDCCwB/A4XfR+H3ETj0H0PfxDYcRirKN4ww4qrZfCH2y1/i8LzcL3uO06BT7L3v4odw4Pz3oGoTwy+vLOPADHuO++NGz3FfOddFniN2Bf0O7MorgY93kgHpGBF0B3bjlyATpb0FIZn2dbb3rfak32TyJ+32ZMBkCiTvMAlptysTMJsDGZc7LZhYff4G/OzqpxkeBIIpHqQOi5+tBAfHxih0dnUztnM/QnoyprKPlJf39vJ0Ey+v/UZ5bKw8WcrnS5MfPXTmjjvOHPJd9P0rr/z+RT7aR3x1M/qT0ocWtct9UNdJUijExopx+jrt6DXsZdYV7WP36mHs4R5HWqrtTSzeWT9DRrh2P3755Y9zj5fPBcr0HNph9LzyHpL1AN5TKxTzsLSQ4Hrt5Y8/vsj9qPzC28l7PjZeEcUV/pD3AnAXaKuTWZwVlII72SdGuFf1hDWxuAyAxYPsNZVwnHyRsKfgS3kMKpVKB0UEu9+sg6bBk/IV+KC9yScIviZ7cLHc7stHnMlY3KZ3WvTjeotDb4/Hks5I3tdeFpLxrZlCrlndLOUzW+NJAcl042Ad/gWKtQVaZtoKkaco/i1wD1H8eRopKtu25YJaviYhyUmWAG1QAPzWv4YlX/Cmz6ck7S0sUgzzEmCYK1AMFXrsegelx6m3KfQw+dyPR7jPgXzaGu2t5c6QCVcJf9j83Gbucy/IOri4qkHv5H6JVFTOMOJk+wuTuzg7y/3yBRvALNDv7D/q1yRYNj9f7Zer6VfLem5SegbpMULvpP/K4L5KHn9ZtlMZfDO+k/sCyLz5/LwxvEosVlHiqZkTtXt7XmS6vLXceoXpUI+0aQa/YynZZz9yxN6XXArt3CnvL30IYtsPIQMap5KcJvaAzuMIO7sqn08FrWKhKQvTaBKZRP7rN1ym1GSew9XQ/+XB4rCH5wXyaXuWL3oEb7Rvsi+aENv0ttRwtLDkG7V3C0Io0jfWF0mGPZZYfyS/w4MXXCGP09gWcLhEj8OoD6yYnS4r73V7E+VkaoDvEO0w19ZIuM/u9NlsPpcnPpCODlv9WU9X2BrmaazxYfwnep6/Y4MT/fWbU8CzA+dVgapiGfFty8t7C/vGRKiqdiXgkHoAB53d86X+7RJfGadrg/TqH7lP0PPdIhpDEnCwB2E0RDmYhFYXDWWSEOMLZ6orkCH4Zj+jRL7r9+cSbajJlfB/50RQOr94fHj4msV8fvGa4eHji/me2NRSd/fSZCw2Se5TsT3e/Ggo2WdLmoK81FPKuGNm0VWORcZybjw6e/1CMrlw/ezsDTvT6Z03zJYPT0ej04fL5UNTkcjUocHERNYZsOU6TaNdPaNWYzfvdUqk6IbUyIV78QPclyjdeTRAV0Vdach+p+EO9JnP1ObDPWeqtdcccCRKV0Q5yDwUKCwN9xzA5Dezp8k7j6A+mTvrvND6LJKKxdK8IAqkGM5q9T+EU7UeT348Ur17C6LVKha8vjy555Pb/WlTZyLvD+SmjGajyYZ73XBkNTGdd7vz0wk4uOqO89GSIPTEeD7WIwilKI85j2D06LYMDV3coW839lEdzq3uVfVwr0GjaBsqs9kfpLMfhpaBtiahZaUtD4nmoCV7mgy1kn6AeORsy7qVDtspSaE8OTegkdZRT0VBQ6tXVC7KOEx+yo1f88CuzTctZsSBmVC8T8vxw77ihLtrMtHl4Nqn4vZNfCpo9Y8cnhg+sb83NXdZf2yb37nrkuk7DvbioN/VGeYdVr8j6jZ85dC7jg/l9tw0M3LxaCAc9oRnx+I7RuPZ6LcuduYmkqWDmxKFAy9bmL5qU5jvHMyl9r1kXzRkc71W8KTk/dMC2b/JvRQ5kRddh87bksEADgJwyAsAFcvhfBxpoaWm+qNlyXdy18NdzbI4Krh3smyOmd1tDO5iz7lJng9cMb0kLb0sAr2EIlwm+LhqdkfH1j3WMcvV/KhlbrFjxwX8OH+1bfxMxzUdj83cMvNB+Ae3xx57DBtuuYXEUdvww2hVNcBpEcTd0CYwP34YT6imAfawAvPi9+NJCntEgYXhuf303RUKo3tZYF20gF9D89gRRNKsG6dXlYMFfF16de2Y08rYzp1j5PJFoz642i4/cODSSw8cuHxkZHp6aGh6eoT6km0I0MF3Ih0KKCPWbUuUx0KyIrJEqfDuhVCCH9uJwwWNyVi5nuVHJwB3G0ogko87b69r9cwHOi1vC9awdCiuWdSx/DfkYTPzFwhea9iYc0zGBjZBpOl4ReW5zORS1B9286lQdLjsyGRzzilKgxf4NgljR9EAIoJVXzNnfx5GR7P7YaXSY2VphxXkgd+alQ1flIN5uNWtNevd3I5g0SUWu+JZn9McMuTcZShbBC2wtknykbw7t9l2Wd7jcnsln4t32UwJX8DgjNj9SUNrOuRMeI2xoLymDgPy+4H3TagDzSDihRr8RQVU9xcVOOUP2zSdlrcpadkZvzYm+obTdIkjGDBPkplCnpRb/rdJnZSe7pl98P7DPaeKR7Ys9l54YeJvP6fyinCeW+A06JNULp3wfQf9/imKI/sd5tWPehERRVJgImskjNwUrQ6lVq9oINtnJv9lFbYDmG2GwQpT3ZjuAxfpOWNg7R2pwp4XcTYPn+U97gnx5IhObc5wRp+xnMWoGMuXmp0+U3ux03HPp4VO+6CxY1zr5yMER4YzPR/jRGQfF8Gz9q+UOBr7kRp1edjXnbDbE90+b4ncS14hFBLgwohPlCPRwQTPJwajkXKCHymlMqVSJlWi/IngN+MC9wEyNo2dzIrlaoGWA1p/f9wPubIhiyWUdVXvPlH0wYXfbA7m6J9UqN574wEhHhcCcTouj/bh57kCzEoSRdad02shchNiEyQfYY1TBpQxjbxJfFigmz2oTGtImCjvgtDWbpngB4ydncaB3tFMT09mrLdMvpV7x/iiy1Xgxyb7YvHe3nis7w1lvuQPlPhy37jXO94H3wJ++q3DYOgY/3lXudyVL5eJGCdXX4M/yz0PGO9bl/XRNzwDIa6V4gAksnMdxOrXnu9wI5Gd+RDZ31LBddEpKEEKr4tM85LFg3l8s6srGE1F+sf7I3Ob5lL7Al3CZDSbkQGjc5kxF/6DwxkVPaLfnxrJTV6kP3SwSfRtdgWSUX8i6E+W0+P72l50UBN3E/ok+DgD+dD42g4CCyEHsv0N9iKBUWWVFgviKP7BtOzhgsqultbTctbYovylnRCsoeXFLdFsmM71kSE5CSuvA/GZys5MD8cdHGnTtZOIprjFM2iKudNdLdpTL1E5nEtLE7adxVZT61iTQ7InJ3POWLBstRcle9o2bbEk+9xqgiOtwfwK8uxudDuq/2tL+vWbt/UNz6uqkZudbbCzFtlKsXbwHTzCabgzx20jJobcO6o+w0h/Vf78WM0mfVJIhosaOolcqpcM6h0Gtd6pH1x+4+0PPPDA8pvf/ObbTnEn7vOTM2z++ypf6+3tu//+vt5enCMxGfiOv8CcSehyVO+z9A2q4x8nioQSVK0SzPjG4a5jd+GMjLoAyAYU1Hn45lEO6GbWauR0EsNd+T5chMaAam0iSWiiMVtcmEynvFEJ/0XYWdx7m7lcwrhUbkkOz+eS29xD5qS7MN72/K8x/sKj+lHJl+O3JjPXH+smfzlEp9d2q3t2D/hhXnnX3GS/pk3bbNT1j8/6vIhD/as/oH+zSwO4FsF2eZmrkfeceiiVng7ZZXvhrmEup51ODkgl0TWapvHCjSoalUemXtqwaIRwAbencrnUl9wTweELjvqKM8nhWJkvCKIj0Yk/b5jp7p7Riw5v09KBthsv69oxIGovXuJ8/IRXXfkKZzMghFd3whxx3KkG62pqJjCsq7VYwu/BF49X3tTGvfPcbppHq74Hvuxz5DuJr1Rx6EdED0G3Orh/Dh+nGbqH8HcQknOS8IyWPvOx1afZM/vpMyv4XTXP6Nc9s5k+8zD+PELKWCb6zCOrq+yZk6yfB+Eunw85gH/MZVErcqO1mBekhPEaUcmiYZeUl2Q5/86DD+6E/9g3/8lPzn+S9YN9+Md4uWHeL4A/WRnBy/PkuSw6jk9zLhivFdHTljTTUAbjWM0VZF/3ut7X3dt7L/x/3YOvv4/oyX2v77vvPrkGFkDzgG+ZxqdOlFu3G6BtYFEuC1ioCNmZ+6uNH3FNrKqqSdyLgChGblF0w3UR/YTrFEX9E+RwNjmkXfl/8j30t3ngbwB9BGgme4m/iq5CGvh8NePp1YCjALRZgEZMwz6gFe5YjkNCFgO2eLElBJx59K67HoUhrvnARaMZdWb0og+gBn3Am6wPzE5w46Z8GefTON+Ef1wZIZ3gT+JAbScc9IHY3BpQNzpvz9HG0w1PsSML9ROvBR6xyX+QMia+/cEHt7+pb7uMcxHz+EvcA4CzY32Wo/Z4uEmgf7dVOL3/1EXFjDpdvIh7oHLdfXfddR/JS6I/45/gNwHVIsiRju07l6uTpjOyL5Y9lJ5pv0mOjqVcNRkhBLQ+34TgSUdClpiv3zdebeLerqTdY+9NS/SmyD+R2yZFbtVMbiFlSiR3Z+UUXq786pN0r2Q3fg6fRTwqo7pjaQ0Pn7OTiIC/kW2INLNW45OIIm+BXcGhiC/jJseE+XLYFtZrjtWfR4y0mXffTXT7VVw/3st9Xd4VzPT9CwB7HYU9ocBuB9hbKOxJBXYdN0RrG7ATWIFdz43g91LYcwxGxpiH/t4CsGcV2Ps4P76OexJgZxXYw6siPoBUbEc3lYXVG7l+NA3jKvZScWFtVXupknjh1qO7F7ivv0Z+53Lu0wjOYyj5TRXIvPzYa2+c36fqWoZnHoJ+b63vl8lYm5LfFIpS28Luo1w/6/dR6Pe1Sr8cy/GTxzz75m/kPk36hafRPZRPxfNtSgMCNB0APCMf69TAnMp4kj/ik9q7f65P3TcHZO3YQcc/DOPfD+OTHD3EkVSX5R60Z8iZWni3KL98//6r9u+GlwGn5cFB8u4hbgi9gSOrmKEG9NapNDV+JtJ73R+/k5W7lYyiaIqkaIrwBk02bPK7bG2Oln5omn1OO2lyQ6VBQ6chI3/KfFziHkMPqPqBCgGtAMTIbFIb3NdMRxtdgckrDhJFCNVRC2TYAbPfbW/Xm5v6NV3hmjb3mMFksNl7yHC8vTQky9A8zPVb6LwR7jOxASN0K1ZXXuDeMkPXsMDfEXQ/t8JyFg24pGEbcNrYMeZOwowaPyDVtPc5/H6HIxB4C1wOaHMjAYc9ELDDV3ZnYz4Oc9or+yGlcvQfDPM4fJWb7C7T9gbOjw5xT9JxTDBOGx2HjNdJpdi4QeolVSqlyGWy201wcf58IpnPJxP5lIvn3W6ed7H+gW9Qev+3+k/29CQ7HY5O0v9KLh6TpFg8l7LxVrvdyttkmXnrqohmYQaNtHeOyUwraqG9ixLLaGhFeUHhHuftAUeyx9cSzrh9qmyz32sL2CCT2GQxd1h1jfo0Io722YJa/90+wYaxvznDaUhG5P8DKD5xZwABAAAAAQzMTspD0l8PPPUACQPoAAAAAM2XgKcAAAAAzZfjGP8s/p8EqAPDAAAACQACAAAAAAAAeAFjYGRgYL7x7z0DA8ux/zr/vrOsAIqggo8AuyEIUgB4AW3TA4xeQRTF8TN3atu2bdu2bdt2o25UREVcRrWjmjGLqLb1+n+zVvLLucO8O5vPDmmY4r97gDWI/tlwDbVOamyLyTIa6v6Ryxjvoy6DdyoV73HfmH+JBBxhfVBSjiQrq4ZVURnbo0FWKvrjK6mIlVd+9zX6Y+VUy/KqmTVTb3dftcm67rdqu/Yqa7lYa8B8DrVzOaPHloe6oHr7vuptjVA93k825sxy1k6plJutYlZP3dx75bM3jJ+pQFy7i6oS+smC7xL9DT3ST1quTPSLHgeQX2MpvWVW348l6S+t0N9uVbBRnKPHrPjCahb3Tc/p5ZBDa5cj+orX1Ck9Z8V3IMexTu9phd5htdQovNUs3nULNqqcOdVCH1RGKRRJynpJaqO6K6qOaJiSbTU0uKRaqM/cAPccRdUKXalboWtijTJqiBbuJnmT7KI6wVPV8ZXI+N7u3PWFb+2m/qiLMaiBSiFHKZ8ijSaLhzyjRuH7RtLDa+69qHJ+rpowV9Ut0xBURkXUQIVMc5OoF6tsSo4Lb1ES9dEURVAC7TU2GhnzvVXBl1Ur3xY1VYG1Kkkah8ysJSqFukn43yXgJrZhNdYkzR3DOSnagPk4g+tYglmYiSlJ67OSHEhyJLgnxb+T/+MKsQQAeAEtwQMUIzkAANDaSmZSt0kxSnK3to2HtW3bts2HtW3btm3btv7X6XT1/12p9+sb6mfqrxtEQ3vDcsNxo8FY2tjWeNGkmrqatpu5ua/5uSVsqWuZaoXW1tal1tu2qrbj9qB9pP2xI7fjpvN/Z33nUudrV9I12/XRXdi90UM97T2XvUlvV+9an+zr6zsLbEAEBHCQExQFZcF6sBscB5fBffAafIcO6IdJ+D/MDYvD8rA2bA5fC6WFykJ9obXQXRgsXBeJyMWcYlGxrFhTbIrCSEaZUX5UGlVG9VFr1B0NRvPRarQdHUbn0W303F/RfzdQP9A60D0wODA+MD2wOKgGh4YsoaHh7OG+4c0RW6RzZH80Hh0Y3RwrGlscOxp7G9fFXfHc8brxlfHP2IejuCBuiNvinngwnoiX4r34JL6KX+KPxEUIoaQsaUgGk4VkL7mZKJiYmdicOJq4n9Ql1eTk5MFU0dTBdNX03vTx9MX07fRTSZZKS0Olu7Is/y9nl4fK+xWDUlM5quZU52tpras2UZuqLdd2aqe1u9RGozQv7U9H05X0NHMwzhqy7mwnO8zOsuvsIXvNvnIL93HK8/LKvCmfzNfyz/9l/q/4f33/W/vf3f/F//P/3/TXtj8B7WaMQAAAAAABAAAA8QBVAAcAaQAFAAEAAAAAAAoAAAIAAXMAAwABeAGVkLN2RGEUhb/YeYbbx7bZxE41tueyTZ2Hzl5nxZ7162wOgB7u6aCts482JuF5bmdYyGbdx6w/z53vPF088Pg8dzPaNvw8D3PYtswFVTwaJEnjCMWp0NR0Kq5qTJoyeRJCJVKcC2fxNMflmGKcSeZ0Vthil232NL3vfG587hv73PecO7HktbSGvHlpFZwP7adKOj84tOTIiXNJmuLzrOksmFomTpG0eTJiS/YdpjXN2Vmylla+eQ5Xq8YyE1qBrXHiUqrypDUnNZWluUTypTWV7N+L6x23fJnSE374RZx4AWzBRQHCAAAAwNtWBXeG2w93d2ISjARQgDsh8H2L/fMhCINIJCEpJS0jKyevoKikrKIqVlPX0NTS1tHV0zcwNDI2MTUzt7C0sraxtbN3cHRydnF1c/fw9PoRBA/dQihgAADne+/atm/9jLzMWmbb9sl2q2zbtnky19m1bsY+t9ywX0+9LNPbHX3cdNsDd91zX1+PPfTIAf0sd9AzTzzV3ycffDHQAIMMMdhQGw0zwnAjjTLGaGON89x4E00wyRSTbXLIfNNMNd0Mn5320ddIiESp8hUoVKRYiVJlylXIla1SlQRb1NksTaLjTjjqmKuuOe+CXXbLiSSHHXHFXM2ly9BYU111l6ebJt7LNMc8Cy1QLymSIyVSIy3SIyMyIyuyrbM+ciLXTOckm21N5JnlbORL8cdvf2X55ruTalSrtVIzPWyNAi0s9sJLS7zy1rsojKIojpIojbIoj4qojKqojpqojbqojwbRMKFN51at/mvR4f++3f41NRZGDMMwACyTZgnjcXGPMrdhXD5gmRLb/ye8dzedZ/+Xrm/2i+B/U5MsTBZpERBZpc/4hrHlPchitPoWB+r1H/z6BpS+oRdsIIoNvc1OpG+GMyeHLA76laZpOLh6i7R0HSjVyymkMQMopZgJlFLMAkopZlP6QCnlNkApYqYHlDymb/TpS03jTwtov73SAAAAsAArALIBAQIrAbICAgIrAbcCMCYeFw4ACCu3Ay4mHhcOAAgrALcBOzAmHBIACCsAsgQIByuwACBFfWkYREuwYFJYsAEbsABZsAGOAA==) format('woff')}</style><link rel="preload" href="integrator_002.js"><script type="text/javascript" src="integrator_002.js"></script><link rel="preload" href="integrator.js"><script type="text/javascript" src="integrator.js"></script><script src="pubads_impl_240.js" async=""></script><style type="text/css">.MathJax_Hover_Frame {border-radius: .25em; -webkit-border-radius: .25em; -moz-border-radius: .25em; -khtml-border-radius: .25em; box-shadow: 0px 0px 15px #83A; -webkit-box-shadow: 0px 0px 15px #83A; -moz-box-shadow: 0px 0px 15px #83A; -khtml-box-shadow: 0px 0px 15px #83A; border: 1px solid #A6D ! important; display: inline-block; position: absolute}
.MathJax_Menu_Button .MathJax_Hover_Arrow {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 4px; -webkit-border-radius: 4px; -moz-border-radius: 4px; -khtml-border-radius: 4px; font-family: 'Courier New',Courier; font-size: 9px; color: #F0F0F0}
.MathJax_Menu_Button .MathJax_Hover_Arrow span {display: block; background-color: #AAA; border: 1px solid; border-radius: 3px; line-height: 0; padding: 4px}
.MathJax_Hover_Arrow:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_Hover_Arrow:hover span {background-color: #CCC!important}
</style><style type="text/css">#MathJax_About {position: fixed; left: 50%; width: auto; text-align: center; border: 3px outset; padding: 1em 2em; background-color: #DDDDDD; color: black; cursor: default; font-family: message-box; font-size: 120%; font-style: normal; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 15px; -webkit-border-radius: 15px; -moz-border-radius: 15px; -khtml-border-radius: 15px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_About.MathJax_MousePost {outline: none}
.MathJax_Menu {position: absolute; background-color: white; color: black; width: auto; padding: 2px; border: 1px solid #CCCCCC; margin: 0; cursor: default; font: menu; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
.MathJax_MenuItem {padding: 2px 2em; background: transparent}
.MathJax_MenuArrow {position: absolute; right: .5em; padding-top: .25em; color: #666666; font-size: .75em}
.MathJax_MenuActive .MathJax_MenuArrow {color: white}
.MathJax_MenuArrow.RTL {left: .5em; right: auto}
.MathJax_MenuCheck {position: absolute; left: .7em}
.MathJax_MenuCheck.RTL {right: .7em; left: auto}
.MathJax_MenuRadioCheck {position: absolute; left: 1em}
.MathJax_MenuRadioCheck.RTL {right: 1em; left: auto}
.MathJax_MenuLabel {padding: 2px 2em 4px 1.33em; font-style: italic}
.MathJax_MenuRule {border-top: 1px solid #CCCCCC; margin: 4px 1px 0px}
.MathJax_MenuDisabled {color: GrayText}
.MathJax_MenuActive {background-color: Highlight; color: HighlightText}
.MathJax_MenuDisabled:focus, .MathJax_MenuLabel:focus {background-color: #E8E8E8}
.MathJax_ContextMenu:focus {outline: none}
.MathJax_ContextMenu .MathJax_MenuItem:focus {outline: none}
#MathJax_AboutClose {top: .2em; right: .2em}
.MathJax_Menu .MathJax_MenuClose {top: -10px; left: -10px}
.MathJax_MenuClose {position: absolute; cursor: pointer; display: inline-block; border: 2px solid #AAA; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; font-family: 'Courier New',Courier; font-size: 24px; color: #F0F0F0}
.MathJax_MenuClose span {display: block; background-color: #AAA; border: 1.5px solid; border-radius: 18px; -webkit-border-radius: 18px; -moz-border-radius: 18px; -khtml-border-radius: 18px; line-height: 0; padding: 8px 0 6px}
.MathJax_MenuClose:hover {color: white!important; border: 2px solid #CCC!important}
.MathJax_MenuClose:hover span {background-color: #CCC!important}
.MathJax_MenuClose:hover:focus {outline: none}
</style><style type="text/css">.MJX_Assistive_MathML {position: absolute!important; top: 0; left: 0; clip: rect(1px, 1px, 1px, 1px); padding: 1px 0 0 0!important; border: 0!important; height: 1px!important; width: 1px!important; overflow: hidden!important; display: block!important; -webkit-touch-callout: none; -webkit-user-select: none; -khtml-user-select: none; -moz-user-select: none; -ms-user-select: none; user-select: none}
.MJX_Assistive_MathML.MJX_Assistive_MathML_Block {width: 100%!important}
</style><style type="text/css">#MathJax_Zoom {position: absolute; background-color: #F0F0F0; overflow: auto; display: block; z-index: 301; padding: .5em; border: 1px solid black; margin: 0; font-weight: normal; font-style: normal; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; -webkit-box-sizing: content-box; -moz-box-sizing: content-box; box-sizing: content-box; box-shadow: 5px 5px 15px #AAAAAA; -webkit-box-shadow: 5px 5px 15px #AAAAAA; -moz-box-shadow: 5px 5px 15px #AAAAAA; -khtml-box-shadow: 5px 5px 15px #AAAAAA; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_ZoomOverlay {position: absolute; left: 0; top: 0; z-index: 300; display: inline-block; width: 100%; height: 100%; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
#MathJax_ZoomFrame {position: relative; display: inline-block; height: 0; width: 0}
#MathJax_ZoomEventTrap {position: absolute; left: 0; top: 0; z-index: 302; display: inline-block; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
</style><style type="text/css">.MathJax_Preview {color: #888}
#MathJax_Message {position: fixed; left: 1px; bottom: 2px; background-color: #E6E6E6; border: 1px solid #959595; margin: 0px; padding: 2px 8px; z-index: 102; color: black; font-size: 80%; width: auto; white-space: nowrap}
#MathJax_MSIE_Frame {position: absolute; top: 0; left: 0; width: 0px; z-index: 101; border: 0px; margin: 0px; padding: 0px}
.MathJax_Error {color: #CC0000; font-style: italic}
</style></head>
    <body data-mediavoice-version="1.99.1"><div id="MathJax_Message" style="">Processing math: 19%</div><div id="popup-references" class="u-composite-layer popup-base-theme" style="top:-60px;left:-60px;visibility:hidden;opacity:0;" aria-hidden="true" aria-label="popup"><svg class="popup-arrow" xmlns="http://www.w3.org/2000/svg" width="33.7" height="18.4" viewBox="0 0 33.7 18.4"><path class="fill" fill="#F7FBFE" d="M1.4 18.4h30.9l-15.5-16.9z"></path><path class="stroke" fill="#98BED7" d="M0 18.4h1.4l15.4-16.9 15.5 16.9h1.4l-16.9-18.4z"></path></svg><div class="popup-arrow popup-arrow-shadow icon--popup-arrow-shadow"></div><div class="popup__references"><div tabindex="-1" class="popup-base-theme__inner" data-component="SpringerLink-Popup-inner"></div></div><button tabindex="-1" class="popup-close" data-component="SpringerLink-Popup-close"><img alt="Close" src="data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2214%22%20height%3D%2214%22%3E%3Cpath%20fill%3D%22%23666%22%20d%3D%22M1.706.292A.996.996%200%200%200%20.294.294a1.002%201.002%200%200%200-.002%201.412L5.527%206.94.292%2012.177a.996.996%200%200%200%20.002%201.412c.39.39%201.026.385%201.412%200L6.94%208.354l5.236%205.235a.998.998%200%200%200%201.414-1.413L8.355%206.94l5.235-5.234a.996.996%200%200%200-.002-1.412%201.002%201.002%200%200%200-1.412-.002L6.94%205.527%201.707.292z%22%2F%3E%3C%2Fsvg%3E"></button></div><div id="popup-article-dates" class="u-composite-layer popup-base-theme" style="top:-60px;left:-60px;visibility:hidden;opacity:0;" aria-hidden="true" aria-label="popup"><svg class="popup-arrow" xmlns="http://www.w3.org/2000/svg" width="33.7" height="18.4" viewBox="0 0 33.7 18.4"><path class="fill" fill="#F7FBFE" d="M1.4 18.4h30.9l-15.5-16.9z"></path><path class="stroke" fill="#98BED7" d="M0 18.4h1.4l15.4-16.9 15.5 16.9h1.4l-16.9-18.4z"></path></svg><div class="popup-arrow popup-arrow-shadow icon--popup-arrow-shadow"></div><div class="popup__article-dates"><div tabindex="-1" class="popup-base-theme__inner" data-component="SpringerLink-Popup-inner"></div></div><button tabindex="-1" class="popup-close" data-component="SpringerLink-Popup-close"><img alt="Close" src="data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2214%22%20height%3D%2214%22%3E%3Cpath%20fill%3D%22%23666%22%20d%3D%22M1.706.292A.996.996%200%200%200%20.294.294a1.002%201.002%200%200%200-.002%201.412L5.527%206.94.292%2012.177a.996.996%200%200%200%20.002%201.412c.39.39%201.026.385%201.412%200L6.94%208.354l5.236%205.235a.998.998%200%200%200%201.414-1.413L8.355%206.94l5.235-5.234a.996.996%200%200%200-.002-1.412%201.002%201.002%200%200%200-1.412-.002L6.94%205.527%201.707.292z%22%2F%3E%3C%2Fsvg%3E"></button></div><div id="popup-search" class="u-composite-layer popup-search-theme" style="top:-60px;left:-60px;visibility:hidden;opacity:0;" aria-hidden="true" aria-label="popup"><svg class="popup-arrow" xmlns="http://www.w3.org/2000/svg" width="33.7" height="18.4" viewBox="0 0 33.7 18.4"><path class="fill" fill="#F7FBFE" d="M1.4 18.4h30.9l-15.5-16.9z"></path><path class="stroke" fill="#98BED7" d="M0 18.4h1.4l15.4-16.9 15.5 16.9h1.4l-16.9-18.4z"></path></svg><div class="popup-arrow popup-arrow-shadow icon--popup-arrow-shadow"></div><div class="popup__search"><div tabindex="-1" class="popup-base-theme__inner" data-component="SpringerLink-Popup-inner"></div></div><button tabindex="-1" class="popup-close" data-component="SpringerLink-Popup-close"><img alt="Close" src="data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2214%22%20height%3D%2214%22%3E%3Cpath%20fill%3D%22%23666%22%20d%3D%22M1.706.292A.996.996%200%200%200%20.294.294a1.002%201.002%200%200%200-.002%201.412L5.527%206.94.292%2012.177a.996.996%200%200%200%20.002%201.412c.39.39%201.026.385%201.412%200L6.94%208.354l5.236%205.235a.998.998%200%200%200%201.414-1.413L8.355%206.94l5.235-5.234a.996.996%200%200%200-.002-1.412%201.002%201.002%200%200%200-1.412-.002L6.94%205.527%201.707.292z%22%2F%3E%3C%2Fsvg%3E"></button></div>
        <noscript><iframe src="//www.googletagmanager.com/ns.html?id=GTM-WCF9Z9"
                      height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>

    <div class="skip-to">
    <a class="skip-to__link pseudo-focus" href="#main-content">Skip to main content</a>
        <a class="skip-to__link skip-to__link--contents pseudo-focus" href="#article-contents">Skip to sections</a>
</div>
        <div class="page-wrapper">
            <noscript>
    <div class="nojs-banner u-interface">
        <p>This service is more advanced with JavaScript available, learn more at <a
                href="http://activatejavascript.org" target="_blank" rel="noopener">http://activatejavascript.org</a>
        </p>
    </div>
</noscript>
                    <div id="leaderboard" class="leaderboard u-hide" data-component="SpringerLink.GoogleAds" data-namespace="leaderboard">
            <div class="leaderboard__wrapper">
                <p class="leaderboard__label">Advertisement</p>
                <button class="leaderboard__hide" title="Hide this advertisement">Hide</button>
                <div id="doubleclick-leaderboard-ad" class="leaderboard__ad"><div id="google_ads_iframe_270604982/springerlink/10618/article_0__container__" style="border: 0pt none;"><iframe id="google_ads_iframe_270604982/springerlink/10618/article_0" title="3rd party ad content" name="google_ads_iframe_270604982/springerlink/10618/article_0" scrolling="no" marginwidth="0" marginheight="0" style="border: 0px none; vertical-align: bottom;" srcdoc="" width="728" frameborder="0" height="90"></iframe></div></div>
            </div>
        </div>

                <header id="header" class="header u-interface">
        <div class="header__content">
            <div class="header__menu-container">
                    <a id="logo" class="site-logo" href="https://link.springer.com/" title="Go to homepage">
                <div class="u-screenreader-only">SpringerLink</div>
    <svg class="site-logo__springer" width="148" height="30" role="img" focusable="false" aria-hidden="true">
        <image width="148" height="30" alt="" src="/springerlink-static/439917616/images/png/springerlink.png" xlink="http://www.w3.org/1999/xlink" xlink:href="/springerlink-static/439917616/images/svg/springerlink.svg"></image>
    </svg>

    </a>


                    <nav id="search-container" class="u-inline-block">
                        <div class="search">
                            <div class="search__content">
                                <form class="u-form-single-input" action="/search" method="get" role="search">
    <label for="search-springerlinks">Search SpringerLink</label>
    <div class="u-relative">
        <input aria-label="Search SpringerLink" name="query" autocomplete="off" type="text">
        <input id="search-springerlink" class="u-hide-text" value="Submit" title="Submit" type="submit">
        <svg class="u-vertical-align-absolute" width="13" height="13" viewBox="222 151 13 13" version="1.1" xmlns="http://www.w3.org/2000/svg" focusable="false" aria-hidden="true" role="presentation">
            <path d="M227 159C228.7 159 230 157.7 230 156 230 154.3 228.7 153 227 153 225.3 153 224 154.3 224 156 224 157.7 225.3 159 227 159L227 159 227 159 227 159ZM230 160.1L231.1 159 233.9 161.7C234.2 162.1 234.2 162.6 233.9 162.9 233.6 163.2 233.1 163.2 232.7 162.9L230 160.1 230 160.1 230 160.1 230 160.1ZM227 161L227 161C224.2 161 222 158.8 222 156 222 153.2 224.2 151 227 151 229.8 151 232 153.2 232 156 232 158.8 229.8 161 227 161L227 161 227 161 227 161 227 161Z" stroke="none" fill-rule="evenodd"></path>
        </svg>
    </div>
</form>
                            </div>
                        </div>
                    </nav>

                    <nav class="nav-container u-interface">
    <div class="global-nav__wrapper">
        <div class="search-button">
            <a class="search-button__label" href="#search-container" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-search">
                <span class="search-button__title">Search</span><svg width="12" height="12" viewBox="222 151 12 12" version="1.1" xmlns="http://www.w3.org/2000/svg" focusable="false" aria-hidden="true" role="presentation">
                    <path d="M227 159C228.7 159 230 157.7 230 156 230 154.3 228.7 153 227 153 225.3 153 224 154.3 224 156 224 157.7 225.3 159 227 159L227 159 227 159 227 159ZM230 160.1L231.1 159 233.9 161.7C234.2 162.1 234.2 162.6 233.9 162.9 233.6 163.2 233.1 163.2 232.7 162.9L230 160.1 230 160.1 230 160.1 230 160.1ZM227 161L227 161C224.2 161 222 158.8 222 156 222 153.2 224.2 151 227 151 229.8 151 232 153.2 232 156 232 158.8 229.8 161 227 161L227 161 227 161 227 161 227 161Z" stroke="none" fill-rule="evenodd"></path>
                </svg>
            </a>
        </div>

        <ul class="global-nav" data-component="SpringerLink.Menu" data-title="Navigation menu" data-text="Menu">
            <li>
                <a href="https://link.springer.com/">
                    <span class="u-overflow-ellipsis">Home</span>
                </a>
            </li>
            <li>
                <a href="https://link.springer.com/contactus">
                    <span class="u-overflow-ellipsis">Contact us</span>
                </a>
            </li>

                <li class="global-nav__logged-out">
                    <a class="test-login-link" href="https://link.springer.com/signup-login?previousUrl=https%3A%2F%2Flink.springer.com%2Farticle%2F10.1007%2Fs10618-016-0466-x">
                        <span class="u-overflow-ellipsis">Log in</span>
                    </a>
                </li>

        </ul><div class="main-menu u-composite-layer c-button-dropdown c-button-dropdown--ghost" data-component="SV.Dropdown" data-namespace="Menu" aria-label="button with dropdown options" style=""><button id="button-Dropdown-Menu-dropdown" type="button" title="Navigation menu" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown-Menu-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">Menu</span><span class="c-button-dropdown__icon"></span></button><div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown-Menu-dropdown"><ul class="main-menu__content" data-role="button-dropdown__content"><li>
                <a href="https://link.springer.com/">
                    <span class="u-overflow-ellipsis">Home</span>
                </a>
            </li><li>
                <a href="https://link.springer.com/contactus">
                    <span class="u-overflow-ellipsis">Contact us</span>
                </a>
            </li><li class="global-nav__logged-out">
                    <a class="test-login-link" href="https://link.springer.com/signup-login?previousUrl=https%3A%2F%2Flink.springer.com%2Farticle%2F10.1007%2Fs10618-016-0466-x">
                        <span class="u-overflow-ellipsis">Log in</span>
                    </a>
                </li></ul></div></div>
    </div> 
</nav> 
            </div>

        </div>
    </header>

            

            <main id="main-content" class="main-wrapper" tabindex="-1">
                <div class="main-container uptodate-recommendations-on">
                    <aside class="main-sidebar-left">
                        <div class="main-sidebar-left__content">
                            <div class="cover-image test-cover" itemscope="">
                                    <a class="test-cover-link" href="https://link.springer.com/journal/10618">
        <span class="u-screenreader-only">Data Mining and Knowledge Discovery</span>
        <img class="test-cover-image" src="5.jpg" itemprop="image" alt="">
    </a>


                            </div>
                        </div>
                    </aside>
                    <div class="main-body" data-role="NavigationContainer">
                                <div class="cta-button-container cta-button-container--top cta-button-container--stacked u-mb-16 u-hide-two-col">
                    <div>
            <a href="https://link.springer.com/content/pdf/10.1007%2Fs10618-016-0466-x.pdf" target="_blank" class="c-button c-button--blue c-button__icon-right gtm-pdf-link" title="Download this article in PDF format" rel="noopener">
                <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" version="1.1"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g fill="#fff"><g transform="translate(12.000000, 5.000000)"><path d="M7 7.3L7 1C7 0.4 6.6 0 6 0 5.4 0 5 0.4 5 1L5 7.3 3.5 5.7C3.1 5.3 2.5 5.3 2.1 5.7L2.1 5.7C1.7 6.1 1.7 6.7 2.1 7.1L5.3 10.3C5.7 10.7 6.3 10.7 6.7 10.3L9.9 7.1C10.3 6.7 10.3 6.1 9.9 5.7L9.9 5.7C9.5 5.3 8.9 5.3 8.5 5.7L7 7.3 7 7.3ZM0 13C0 12.4 0.5 12 1 12L11 12C11.6 12 12 12.4 12 13 12 13.6 11.5 14 11 14L1 14C0.4 14 0 13.6 0 13L0 13Z"></path></g></g></g></svg>
                <span class="hide-text-small">Download</span>
                <span>PDF</span>
            </a>
        </div>

        </div>



                        <article class="main-body__content">
                            <div xmlns="http://www.w3.org/1999/xhtml" class="FulltextWrapper"><div class="ArticleHeader main-context"><div id="enumeration" class="enumeration"><p><a href="https://link.springer.com/journal/10618" title="Data Mining and Knowledge Discovery"><span class="JournalTitle">Data Mining and Knowledge Discovery</span></a></p><p class="icon--meta-keyline-before"><span class="ArticleCitation_Year"><time datetime="2016-09">September 2016</time>, </span><span class="ArticleCitation_Volume">Volume 30, </span><a class="ArticleCitation_Issue" href="https://link.springer.com/journal/10618/30/5/page/1">Issue&nbsp;5</a>,
                       <span class="ArticleCitation_Pages"> pp 1024–1052</span><span class="u-inline-block u-ml-4"> | <a class="gtm-cite-link" href="#citeas">Cite as</a></span></p></div><div class="MainTitleSection"><h1 class="ArticleTitle" lang="en">A distributed approach for graph mining in massive networks</h1></div><div class="authors u-clearfix authors--enhanced" data-component="SpringerLink.Authors"><ul class="u-interface u-inline-list authors__title" data-role="AuthorsNavigation"><li><a href="#authors" class="gtm-tab-authors selected">Authors</a></li><li><a href="#authorsandaffiliations" class="gtm-tab-authorsandaffiliations">Authors and affiliations</a></li></ul><span class="marker" style="width: 160px;"></span><div class="authors__list" data-role="AuthorsList" id="authors" tabindex="-1"><ul class="test-contributor-names"><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors__name">N.&nbsp;Talukder</span></li><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors__name">M.&nbsp;J.&nbsp;Zaki</span><span class="author-information"><span class="authors__contact"><a href="mailto:zaki@cs.rpi.edu" title="zaki@cs.rpi.edu" itemprop="email" class="gtm-email-author"><img src="email.svg" alt="Email author" width="24" height="24"></a></span></span></li></ul></div><div class="authors__affiliations" id="authorsandaffiliations" tabindex="-1"><div class="authors-affiliations u-interface"><ul class="test-contributor-names"><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors-affiliations__name">N.&nbsp;Talukder</span><ul class="authors-affiliations__indexes u-inline-list" data-role="AuthorsIndexes"><li data-affiliation="affiliation-1">1</li></ul></li><li itemscope="" itemtype="http://schema.org/Person" class="u-mb-2 u-pt-4 u-pb-4"><span itemprop="name" class="authors-affiliations__name">M.&nbsp;J.&nbsp;Zaki</span><ul class="authors-affiliations__indexes u-inline-list" data-role="AuthorsIndexes"><li data-affiliation="affiliation-1">1</li></ul><span class="author-information"><span class="author-information__contact u-icon-before icon--email-before"><a href="mailto:zaki@cs.rpi.edu" title="zaki@cs.rpi.edu" itemprop="email" class="gtm-email-author" tabindex="-1">Email author</a></span><span class="author-information__orcid u-icon-before icon--ORCID-before"><a class="gtm-orcid-link" href="http://orcid.org/0000-0003-4711-0234" itemprop="url" title="View OrcID profile" target="_blank" rel="noopener" tabindex="-1">View author's OrcID profile</a></span></span></li></ul><ol class="test-affiliations"><li class="affiliation" data-test="affiliation-1" data-affiliation-highlight="affiliation-1" itemscope="" itemtype="http://schema.org/Organization"><span class="affiliation__count">1.</span><span class="affiliation__item"><span itemprop="name" class="affiliation__name">Rensselaer Polytechnic Institute</span><span itemprop="address" itemscope="" itemtype="http://schema.org/PostalAddress" class="affiliation__address"><span itemprop="addressRegion" class="affiliation__city">Troy</span><span itemprop="addressCountry" class="affiliation__country">USA</span></span></span></li></ol></div></div></div><div class="main-context__container" data-component="SpringerLink.ArticleMetrics"><div class="main-context__column"><span><span class="test-render-category">Article</span></span><div class="article-dates article-dates--enhanced" data-component="SpringerLink.ArticleDates"><div class="article-dates__entry"><span class="article-dates__label">First Online: </span><span class="article-dates__first-online"><a href="#article-dates-history" class="gtm-first-online" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-article-dates">09 June 2016</a></span></div><div class="article-dates__history" id="article-dates-history"><div class="article-dates__entry"><span class="article-dates__label">Received: </span><span><time datetime="2015-11-14">14 November 2015</time></span></div><div class="article-dates__entry"><span class="article-dates__label">Accepted: </span><span><time datetime="2016-05-23">23 May 2016</time></span></div></div></div></div><div class="main-context__column">    <ul id="book-metrics" class="article-metrics u-sansSerif">
            <li class="article-metrics__item">
                     <span class="article-metrics__views">1.1k</span>
                     <span class="article-metrics__label">Downloads</span>
            </li>
            <li class="article-metrics__item">
                    <a class="article-metrics__link gtm-citations-count" href="https://citations.springer.com/item?doi=10.1007/s10618-016-0466-x" target="_blank" rel="noopener" title="Visit Springer Citations for full citation details" id="citations-link">
                            <span id="citations-count-number" class="test-metric-count c-button-circle gtm-citations-count">7</span>
                       <span class="test-metric-name article-metrics__label gtm-citations-count">Citations</span>
                    </a>
            </li>
    </ul>
</div></div></div><section class="Abstract" id="Abs1" tabindex="-1" lang="en"><h2 class="Heading">Abstract</h2><p id="Par1" class="Para">We
 propose a novel distributed algorithm for mining frequent subgraphs 
from a single, very large, labeled network. Our approach is the first 
distributed method to mine a massive input graph that is too large to 
fit in the memory of any individual compute node. The input graph thus 
has to be partitioned among the nodes, which can lead to potential false
 negatives. Furthermore, for scalable performance it is crucial to 
minimize the communication among the compute nodes. Our algorithm, <span class="EmphasisTypeSmallCaps ">DistGraph</span>,
 ensures that there are no false negatives, and uses a set of 
optimizations and efficient collective communication operations to 
minimize information exchange. To our knowledge <span class="EmphasisTypeSmallCaps ">DistGraph</span>
 is the first approach demonstrated to scale to graphs with over a 
billion vertices and edges. Scalability results on up to 2048 IBM Blue 
Gene/Q compute nodes, with 16 cores each, show very good speedup.</p></section><div class="KeywordGroup" lang="en"><h2 class="Heading">Keywords</h2><span class="Keyword">Parallel graph mining&nbsp;</span><span class="Keyword">Distributed graph mining&nbsp;</span><span class="Keyword">Single large graph&nbsp;</span><span class="Keyword">Frequent subgraph mining&nbsp;</span><span class="Keyword">High performance computing&nbsp;</span></div><div class="HeaderArticleNotes"><aside class="ArticleNote ArticleNoteCommunicatedBy"><p class="SimplePara">Responsible editor: Thomas Gärtner, Mirco Nanni, Andrea Passerini and Celine Robardet.</p></aside></div><div class="article-actions--inline" id="article-actions--inline" data-component="article-actions--inline"><div class="citations u-interface c-button-dropdown" data-component="SV.Dropdown" data-namespace="citations--inline" aria-label="button with dropdown options">
        <button id="button-Dropdown-citations--inline-dropdown" type="button" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown-citations--inline-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">
    <span>Cite</span>
    <span class="hide-text-small">article</span>
</span><span class="c-button-dropdown__icon"></span></button>
<div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown-citations--inline-dropdown"><ul class="citations__content" data-role="button-dropdown__content">
    <li>
        <a href="#citeas" class="gtm-cite-dropdown">How to cite?</a>
    </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-016-0466-x?format=refman&amp;flavour=citation" title="Download this article's citation as a .RIS file" class="gtm-export-citation" data-gtmlabel="RIS">
                <span class="citations__extension" data-gtmlabel="RIS">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .RIS
                </span>
                <span class="citations__types">
                        <span>
                            Papers
                        </span>
                        <span>
                            Reference Manager
                        </span>
                        <span>
                            RefWorks
                        </span>
                        <span>
                            Zotero
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-016-0466-x?format=endnote&amp;flavour=citation" title="Download this article's citation as a .ENW file" class="gtm-export-citation" data-gtmlabel="ENW">
                <span class="citations__extension" data-gtmlabel="ENW">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .ENW
                </span>
                <span class="citations__types">
                        <span>
                            EndNote
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-016-0466-x?format=bibtex&amp;flavour=citation" title="Download this article's citation as a .BIB file" class="gtm-export-citation" data-gtmlabel="BIB">
                <span class="citations__extension" data-gtmlabel="BIB">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .BIB
                </span>
                <span class="citations__types">
                        <span>
                            BibTeX
                        </span>
                        <span>
                            JabRef
                        </span>
                        <span>
                            Mendeley
                        </span>
                </span>
            </a>
        </li>
</ul></div>
    </div></div><div id="body"><section id="Sec1" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">1 </span>Introduction<span class="section-icon"></span></h2><div class="content"><p id="Par2" class="Para">Frequent
 graph mining is a well-studied problem with numerous applications in 
areas such as computational chemistry, bioinformatics, and social 
networks. For instance, protein structures or protein-protein 
interactions can be modeled by labeled graphs, and a bioinformatics 
researcher might be interested in finding common sub-structures within 
these graphs. Likewise, mining frequent subgraphs from a massive social 
graph or citation network can help find communities which may be of 
interest to social scientists. Other applications include graph 
classification, clustering, and indexing.</p><p id="Par3" class="Para">The
 frequent subgraph mining (FSM) problem is divided into two broad 
categories, namely, finding frequent patterns in either (i) a <em class="EmphasisTypeItalic ">graph database</em> comprising multiple input graphs (e.g., a set of chemical compounds), or (ii) a <em class="EmphasisTypeItalic ">single large input graph</em>
 setting (e.g., a social network or citation graph). The FSM task is to 
enumerate all subgraphs with frequency (or support) above some <em class="EmphasisTypeItalic ">minimum support</em>
 threshold. In the multiple graphs case, frequency is simply the number 
of graphs that contain the pattern. However, defining the notion of 
support in a single graph is more challenging, since it is not enough to
 simply state whether a pattern exists or not. Instead, we have to find 
all the distinct isomorphisms from the pattern to the input graph. 
However, this violates the <em class="EmphasisTypeItalic ">anti-monotonicity</em>
 principle that is required for effective pruning of the output pattern 
space. That is, a subgraph can have fewer isomorphisms (i.e., lower 
support) than its supergraph (e.g., assume that the input graph has only
 one vertex labeled <em class="EmphasisTypeItalic ">A</em>, which is connected to 100 vertices labeled <em class="EmphasisTypeItalic ">B</em>, then the vertex <em class="EmphasisTypeItalic ">A</em> occurs only once, but the edge <span class="InlineEquation" id="IEq1"><span class="MathJax_Preview">A-B</span><script type="math/tex" id="MathJax-Element-1">A-B</script></span>
 occurs 100 times). Several anti-monotonic support measures have been 
proposed for use in the single graph setting (Bringmann and Nijssen <span class="CitationRef"><a href="#CR3" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2008</a></span>).</p><p id="Par4" class="Para">In
 the graph database case there are usually many, moderately sized input 
graphs, which can be horizontally partitioned among the compute nodes 
and mined in a distributed manner. Most existing parallel&nbsp;(Buehrer 
et&nbsp;al. <span class="CitationRef"><a href="#CR4" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>; Meinl et&nbsp;al. <span class="CitationRef"><a href="#CR19" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>; Kessl et&nbsp;al. <span class="CitationRef"><a href="#CR12" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>; Fatta and Berthold <span class="CitationRef"><a href="#CR6" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>) and mapreduce-based approaches&nbsp;(Wu and Bai <span class="CitationRef"><a href="#CR26" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2010</a></span>; Lin et&nbsp;al. <span class="CitationRef"><a href="#CR16" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>)
 adopt this setting. In contrast, our focus is on the more challenging 
task of mining patterns from a massive and sparse input graph. Existing 
sequential solutions for this problem include SiGraM&nbsp;(Kuramochi and
 Karypis <span class="CitationRef"><a href="#CR15" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2005</a></span>) and GraMi&nbsp;(Elseidy et&nbsp;al. <span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>), whereas existing parallel approaches target only shared-memory (SMP) systems&nbsp;(Reinhardt and Karypis <span class="CitationRef"><a href="#CR20" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>).
 To our knowledge, there is no current distributed FSM algorithm, that 
can handle a massive input graph that is too large to fit in the memory 
of a single compute node. The input graph thus has to be partitioned 
among the nodes, which poses several additional mining challenges.</p><div id="Par5" class="Para">One challenge is the problem of <em class="EmphasisTypeItalic ">false negatives</em>,
 i.e., frequent patterns that may be missed because some of their 
occurrences span across partitions. Another challenge is that the global
 pattern support determination now requires a reduction operation over 
the partitions, using information from the potentially exponential 
number of occurrences of a pattern in each partition. Our distributed 
solution, called <span class="EmphasisTypeSmallCaps ">DistGraph</span> (for <strong class="EmphasisTypeBold ">Dist</strong>ributed <strong class="EmphasisTypeBold ">Graph</strong>
 Mining), guarantees that all frequent patterns that appear in a 
sequential approach are found. That means it allows no false negatives 
(frequent patterns that are not discovered) and no false positives 
(discovered patterns that are actually not frequent). To maintain 
scalability and efficiency, our approach minimizes the amount of 
communication to determine the global support via a set of 
optimizations. Our main contributions are as follows:<div class="UnorderedList"><ul class="UnorderedListMarkDash"><li> <p id="Par6" class="Para">We
 develop a distributed solution for mining a single massive graph that 
leverages effective collective communication primitives for scalable 
performance.</p> </li><li> <p id="Par7" class="Para">Unlike previous 
approaches that assume the input graph fits in the memory, we assume the
 network is so large that it will not fit in the local memory at each 
compute node. Our approach is the first one to partition the massive 
input graph into different segments, which are mined in a distributed 
manner.</p> </li><li> <p id="Par8" class="Para">We propose a hybrid 
solution for subgraph mining that utilizes both thread-based parallelism
 within each compute node and distributed computation across multiple 
compute nodes. Our method is therefore suitable for a variety of 
environments, such as distributed systems, shared memory systems, and 
cluster environments over the cloud.</p> </li><li> <p id="Par9" class="Para">We show that <span class="EmphasisTypeSmallCaps ">DistGraph</span>
 can scale to a massive network with over a billion vertices and four 
billion edges. To the best of our knowledge this is the largest network 
considered, to date, for frequent subgraph mining.</p> </li></ul></div> </div></div></section><section id="Sec2" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">2 </span>Background<span class="section-icon"></span></h2><div class="content"><p id="Par10" class="Para">In this section, we review some definitions. A <em class="EmphasisTypeItalic ">graph</em> is defined as <span class="InlineEquation" id="IEq2"><span class="MathJax_Preview">G=(V,E)</span><script type="math/tex" id="MathJax-Element-2">G=(V,E)</script></span>, where <em class="EmphasisTypeItalic ">V</em> is a set of <em class="EmphasisTypeItalic ">vertices</em> and <span class="InlineEquation" id="IEq3"><span class="MathJax_Preview">E \subseteq V \times V</span><script type="math/tex" id="MathJax-Element-3">E \subseteq V \times V</script></span> is a set of <em class="EmphasisTypeItalic ">edges</em>. In addition, we assume that <em class="EmphasisTypeItalic ">L</em> is a labeling function for vertices and edges; we denote by <em class="EmphasisTypeItalic ">L</em>(<em class="EmphasisTypeItalic ">v</em>) the label for vertex <span class="InlineEquation" id="IEq4"><span class="MathJax_Preview">v \in V</span><script type="math/tex" id="MathJax-Element-4">v \in V</script></span>, and by <span class="InlineEquation" id="IEq5"><span class="MathJax_Preview">L(v_1,v_2)</span><script type="math/tex" id="MathJax-Element-5">L(v_1,v_2)</script></span> the label for the edge <span class="InlineEquation" id="IEq6"><span class="MathJax_Preview">(v_1,v_2) \in E</span><script type="math/tex" id="MathJax-Element-6">(v_1,v_2) \in E</script></span>. In the following, we use <span class="InlineEquation" id="IEq7"><span class="MathJax_Preview">G=(V,E)</span><script type="math/tex" id="MathJax-Element-7">G=(V,E)</script></span> for the <em class="EmphasisTypeItalic ">single large input graph</em>, and we use <span class="InlineEquation" id="IEq8"><span class="MathJax_Preview">P=(V_P, E_P)</span><script type="math/tex" id="MathJax-Element-8">P=(V_P, E_P)</script></span> for a <em class="EmphasisTypeItalic ">pattern graph</em>, i.e., one of the subgraphs we wish to mine in <em class="EmphasisTypeItalic ">G</em>.</p><p id="Par11" class="Para"> <em class="EmphasisTypeItalic ">Subgraph isomorphism and embedding</em>: We say that the pattern <em class="EmphasisTypeItalic ">P</em> is <em class="EmphasisTypeItalic ">subgraph isomorphic</em> to <span class="InlineEquation" id="IEq9"><span class="MathJax_Preview">G = (V, E)</span><script type="math/tex" id="MathJax-Element-9">G = (V, E)</script></span>, denoted as <span class="InlineEquation" id="IEq10"><span class="MathJax_Preview">P \subseteq G</span><script type="math/tex" id="MathJax-Element-10">P \subseteq G</script></span>, if there exists an injective function, <span class="InlineEquation" id="IEq11"><span class="MathJax_Preview">\phi \!: V_P \rightarrow V</span><script type="math/tex" id="MathJax-Element-11">\phi \!: V_P \rightarrow V</script></span> such that: 1) <span class="InlineEquation" id="IEq12"><span class="MathJax_Preview">\forall v \in V_P</span><script type="math/tex" id="MathJax-Element-12">\forall v \in V_P</script></span>, <span class="InlineEquation" id="IEq13"><span class="MathJax_Preview">L(v) = L(\phi (v))</span><script type="math/tex" id="MathJax-Element-13">L(v) = L(\phi (v))</script></span>, and 2) <span class="InlineEquation" id="IEq14"><span class="MathJax_Preview">\forall (v_i,v_j) \in E_P</span><script type="math/tex" id="MathJax-Element-14">\forall (v_i,v_j) \in E_P</script></span>, <span class="InlineEquation" id="IEq15"><span class="MathJax_Preview">(\phi (v_i),\phi (v_j)) \in E</span><script type="math/tex" id="MathJax-Element-15">(\phi (v_i),\phi (v_j)) \in E</script></span> and <span class="InlineEquation" id="IEq16"><span class="MathJax_Preview">L(v_i,v_j)= L(\phi (v_i), \phi (v_j))</span><script type="math/tex" id="MathJax-Element-16">L(v_i,v_j)= L(\phi (v_i), \phi (v_j))</script></span>. In this case, the isomorphic subgraph in <em class="EmphasisTypeItalic ">G</em> comprising the vertices <span class="InlineEquation" id="IEq17"><span class="MathJax_Preview">\phi (v_1), \phi (v_2), \ldots , \phi (v_p)</span><script type="math/tex" id="MathJax-Element-17">\phi (v_1), \phi (v_2), \ldots , \phi (v_p)</script></span> (where <span class="InlineEquation" id="IEq18"><span class="MathJax_Preview">p = |V_P|</span><script type="math/tex" id="MathJax-Element-18">p = |V_P|</script></span>) is also called an <em class="EmphasisTypeItalic ">embedding</em> of the pattern <em class="EmphasisTypeItalic ">P</em> in the input graph <em class="EmphasisTypeItalic ">G</em>. We use the terms isomorphism and embedding interchangeably, since given the isomorphism function <span class="InlineEquation" id="IEq19"><span class="MathJax_Preview">\phi </span><script type="math/tex" id="MathJax-Element-19">\phi </script></span>, we can uniquely identify the corresponding embedding (which is a subgraph of <em class="EmphasisTypeItalic ">G</em>).</p><div id="Par12" class="Para"> <em class="EmphasisTypeItalic ">Support:</em> For a pattern <em class="EmphasisTypeItalic ">P</em> and input graph <em class="EmphasisTypeItalic ">G</em>, let <span class="InlineEquation" id="IEq20"><span class="MathJax_Preview">\varSigma (P) = \{\phi _1, \phi _2, \ldots \}</span><script type="math/tex" id="MathJax-Element-20">\varSigma (P) = \{\phi _1, \phi _2, \ldots \}</script></span> denote the set of all isomorphisms/embeddings of <em class="EmphasisTypeItalic ">P</em> in <em class="EmphasisTypeItalic ">G</em>. The support of a pattern <em class="EmphasisTypeItalic ">P</em> is defined based on some function of <span class="InlineEquation" id="IEq21"><span class="MathJax_Preview">\varSigma (P)</span><script type="math/tex" id="MathJax-Element-21">\varSigma (P)</script></span>. For instance, support can be defined as the cardinality of <span class="InlineEquation" id="IEq22"><span class="MathJax_Preview">\varSigma (P)</span><script type="math/tex" id="MathJax-Element-22">\varSigma (P)</script></span>, i.e., the number of embeddings of <em class="EmphasisTypeItalic ">P</em> in <em class="EmphasisTypeItalic ">G</em>. However, this definition violates the <em class="EmphasisTypeItalic ">anti-monotonicity</em>
 principle, which requires that the support of a pattern should not be 
greater than the support of its subgraphs. To address this issue a 
support measure using the maximum independent set of the overlap graph 
between embeddings was proposed by &nbsp;Kuramochi and Karypis (<span class="CitationRef"><a href="#CR15" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2005</a></span>). However, this is known to be a NP-hard problem.<figure class="Figure" id="Fig1"><div class="MediaObject" id="MO1"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig1_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig1_HTML.gif" alt="Fig. 1"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 1</span> <p class="SimplePara">Support in a single graph</p> </div></figcaption></figure> </div><div id="Par13" class="Para">We use the most restricted node (MRN) support&nbsp;(Bringmann and Nijssen <span class="CitationRef"><a href="#CR3" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2008</a></span>),
 also called the minimum image based support, defined as the minimum 
number of unique vertex mappings over any of the vertices in <em class="EmphasisTypeItalic ">P</em>. More formally, given the set of embeddings <span class="InlineEquation" id="IEq23"><span class="MathJax_Preview">\varSigma (P) = \{ \phi _1, \phi _2, \ldots \}</span><script type="math/tex" id="MathJax-Element-23">\varSigma (P) = \{ \phi _1, \phi _2, \ldots \}</script></span>, the MRN support of <em class="EmphasisTypeItalic ">P</em> is defined as:<div id="Equ1" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned} \sigma (P) = \min _{v \in V_P} \bigl \{ |\varPhi (v)|\bigr \} \end{aligned}</span><script type="math/tex; mode=display" id="MathJax-Element-24">\begin{aligned} \sigma (P) = \min _{v \in V_P} \bigl \{ |\varPhi (v)|\bigr \} \end{aligned}</script></div> <div class="EquationNumber">(1)</div></div>where <span class="InlineEquation" id="IEq24"><span class="MathJax_Preview">\varPhi (v)</span><script type="math/tex" id="MathJax-Element-25">\varPhi (v)</script></span> denotes the set of unique mappings for a vertex <span class="InlineEquation" id="IEq25"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex" id="MathJax-Element-26">v \in V_P</script></span>, given as<div id="Equ5" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned} \varPhi (v) = \bigcup _{i=1}^{|\varSigma (P)|} \phi _i(v) \end{aligned}</span><script type="math/tex; mode=display" id="MathJax-Element-27">\begin{aligned} \varPhi (v) = \bigcup _{i=1}^{|\varSigma (P)|} \phi _i(v) \end{aligned}</script></div></div>The <em class="EmphasisTypeItalic ">relative support</em> of a pattern, denoted <span class="InlineEquation" id="IEq26"><span class="MathJax_Preview">\sigma _r(P)</span><script type="math/tex" id="MathJax-Element-28">\sigma _r(P)</script></span>, is expressed as a fraction of the number of vertices in the input graph <em class="EmphasisTypeItalic ">G</em>, i.e., <span class="InlineEquation" id="IEq27"><span class="MathJax_Preview">\sigma _r(P) = \sigma (P)/|V|</span><script type="math/tex" id="MathJax-Element-29">\sigma _r(P) = \sigma (P)/|V|</script></span>.</div><p id="Par14" class="Para">Figure&nbsp;<span class="InternalRef"><a href="#Fig1">1</a></span> shows the embeddings of the pattern <span class="InlineEquation" id="IEq28"><span class="MathJax_Preview">P \equiv A\text {--}B\text {--}C\text {--}A</span><script type="math/tex" id="MathJax-Element-30">P \equiv A\text {--}B\text {--}C\text {--}A</script></span> in the input graph. For instance, one of the embeddings is <span class="InlineEquation" id="IEq29"><span class="MathJax_Preview">\phi _1 = \{0, 2, 5, 7\}</span><script type="math/tex" id="MathJax-Element-31">\phi _1 = \{0, 2, 5, 7\}</script></span>, shown in the first column to the right of the pattern. In total there are four embeddings for <em class="EmphasisTypeItalic ">P</em>. Therefore, the set of distinct mappings for the first vertex is <span class="InlineEquation" id="IEq30"><span class="MathJax_Preview">\{0,1,7\}</span><script type="math/tex" id="MathJax-Element-32">\{0,1,7\}</script></span>. The number of unique mappings over all the pattern vertices are 3, 3, 2 and 2, respectively. Thus, the MRN support of <em class="EmphasisTypeItalic ">P</em> is <span class="InlineEquation" id="IEq31"><span class="MathJax_Preview">\sigma (P) = \min \{3,3,2,2\} = 2</span><script type="math/tex" id="MathJax-Element-33">\sigma (P) = \min \{3,3,2,2\} = 2</script></span>, or equivalently, its relative support is <span class="InlineEquation" id="IEq32"><span class="MathJax_Preview">\sigma _r(P) = 2/8 = 0.25</span><script type="math/tex" id="MathJax-Element-34">\sigma _r(P) = 2/8 = 0.25</script></span> or <span class="InlineEquation" id="IEq33"><span class="MathJax_Preview">25\,\%</span><script type="math/tex" id="MathJax-Element-35">25\,\%</script></span>.</p><p id="Par15" class="Para">The
 MRN support also makes intuitive sense, since it avoids over-counting 
embeddings in the same “locations” in the input graph. For example, 
consider the pattern <span class="InlineEquation" id="IEq34"><span class="MathJax_Preview">P \equiv A\text {--}B\text {--}A</span><script type="math/tex" id="MathJax-Element-36">P \equiv A\text {--}B\text {--}A</script></span>. This has two embeddings in the input graph in Fig.&nbsp;<span class="InternalRef"><a href="#Fig1">1</a></span>, namely, <span class="InlineEquation" id="IEq35"><span class="MathJax_Preview">\phi _1 = \{0,3,1\}</span><script type="math/tex" id="MathJax-Element-37">\phi _1 = \{0,3,1\}</script></span> and <span class="InlineEquation" id="IEq36"><span class="MathJax_Preview">\phi _2 = \{1,3,0\}</span><script type="math/tex" id="MathJax-Element-38">\phi _2 = \{1,3,0\}</script></span>. The set of distinct mappings for the first pattern vertex (labeled <em class="EmphasisTypeItalic ">A</em>) is <span class="InlineEquation" id="IEq37"><span class="MathJax_Preview">\{0,1\}</span><script type="math/tex" id="MathJax-Element-39">\{0,1\}</script></span>, which is also the same for the third vertex (labeled <em class="EmphasisTypeItalic ">A</em>). On the other hand, the set of distinct mappings for the second vertex (labeled <em class="EmphasisTypeItalic ">B</em>) is <span class="InlineEquation" id="IEq38"><span class="MathJax_Preview">\{3\}</span><script type="math/tex" id="MathJax-Element-40">\{3\}</script></span>, resulting in a support value of <span class="InlineEquation" id="IEq39"><span class="MathJax_Preview">\sigma (P) = 1</span><script type="math/tex" id="MathJax-Element-41">\sigma (P) = 1</script></span>.
 Finally, one of the major advantages of using the MRN support is that 
even if there are an exponential number of embeddings, storing the 
mappings <span class="InlineEquation" id="IEq40"><span class="MathJax_Preview">\varPhi (v)</span><script type="math/tex" id="MathJax-Element-42">\varPhi (v)</script></span> across all vertices <span class="InlineEquation" id="IEq41"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex" id="MathJax-Element-43">v \in V_P</script></span> takes at most polynomial space, namely <span class="InlineEquation" id="IEq42"><span class="MathJax_Preview">O(|V_P| \times |V|)</span><script type="math/tex" id="MathJax-Element-44">O(|V_P| \times |V|)</script></span>.</p><p id="Par16" class="Para"> <em class="EmphasisTypeItalic ">Frequent single graph mining:</em> Given a single input graph <em class="EmphasisTypeItalic ">G</em> and a user specified parameter called <em class="EmphasisTypeItalic ">minimum support threshold</em>, i.e., <span class="InlineEquation" id="IEq43"><span class="MathJax_Preview">minsup\in \mathbf {Z}</span><script type="math/tex" id="MathJax-Element-45">minsup\in \mathbf {Z}</script></span>, where <span class="InlineEquation" id="IEq44"><span class="MathJax_Preview">\mathbf {Z}</span><script type="math/tex" id="MathJax-Element-46">\mathbf {Z}</script></span> denotes the set of positive integers, the task is to discover all subgraph patterns, such that for each pattern <em class="EmphasisTypeItalic ">P</em>, we have <span class="InlineEquation" id="IEq45"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex" id="MathJax-Element-47">\sigma (P) \ge minsup</script></span>.</p></div></section><section id="Sec3" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">3 </span>Related work<span class="section-icon"></span></h2><div class="content"><p id="Par17" class="Para">FSM is a well studied problem, in both the graph database&nbsp;(Inokuchi et&nbsp;al. <span class="CitationRef"><a href="#CR10" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2000</a></span>; Kuramochi and Karypis <span class="CitationRef"><a href="#CR14" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2001</a></span>; Yan and Han <span class="CitationRef"><a href="#CR27" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2002</a></span>) and the single graph&nbsp;(Kuramochi and Karypis <span class="CitationRef"><a href="#CR15" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2005</a></span>; Elseidy et&nbsp;al. <span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>) setting.</p><p id="Par18" class="Para"> <em class="EmphasisTypeItalic ">Graph database (multiple graphs):</em> Initial FSM methods used systematic generation of candidate subgraph patterns&nbsp;(Inokuchi et&nbsp;al. <span class="CitationRef"><a href="#CR10" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2000</a></span>; Kuramochi and Karypis <span class="CitationRef"><a href="#CR14" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2001</a></span>) and their level-wise growth via breadth-first search (BFS). Later methods like gSpan&nbsp;(Yan and Han <span class="CitationRef"><a href="#CR27" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2002</a></span>) and FFSM&nbsp;(Huan et&nbsp;al. <span class="CitationRef"><a href="#CR9" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2003</a></span>)
 use canonical ordering of the patterns, and a depth-first (DFS) 
exploration. Parallel graph mining algorithms targeting SMP systems were
 proposed in Buehrer et&nbsp;al. (<span class="CitationRef"><a href="#CR4" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>) and Meinl et&nbsp;al. (<span class="CitationRef"><a href="#CR19" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>),
 where they explore different schemes for load balancing among the 
multiple processors. One of the first distributed methods was (Fatta and
 Berthold <span class="CitationRef"><a href="#CR6" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>), which uses a message passing approach for mining molecular graphs. Recently, map-reduce based approaches&nbsp;(Wu and Bai <span class="CitationRef"><a href="#CR26" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2010</a></span>; Lu et&nbsp;al. <span class="CitationRef"><a href="#CR18" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2013</a></span>; Lin et&nbsp;al. <span class="CitationRef"><a href="#CR16" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>; Bhuiyan and Al <span class="CitationRef"><a href="#CR2" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2015</a></span>; Hill et&nbsp;al. <span class="CitationRef"><a href="#CR7" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2012</a></span>) have attracted attention; GPUs have also been exploited&nbsp;(Kessl et&nbsp;al. <span class="CitationRef"><a href="#CR12" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>).
 It is important to note that all of these parallel/distributed methods 
are designed for multiple input graphs, and cannot be adapted for the 
single graph case due to the underlying algorithmic assumptions.</p><p id="Par19" class="Para"> <em class="EmphasisTypeItalic ">Single input graph:</em> Sequential single input graph mining methods include SUBDUE&nbsp;(Holder and Cook <span class="CitationRef"><a href="#CR8" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1993</a></span>), SiGraM&nbsp;(Kuramochi and Karypis <span class="CitationRef"><a href="#CR15" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2005</a></span>), and more recently GraMi&nbsp;(Elseidy et&nbsp;al. <span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>).
 One of the earliest algorithms, SUBDUE searches for the potentially 
frequent substructures in the graph that can also compress it well. 
SiGraM uses the maximum independent set based approach for computing 
support which can be very expensive. GraMi takes a constraint 
satisfaction based approach for support computation and optimizes the 
storage of isomorphisms in memory by detecting automorphism groups. A 
parallel single graph mining method for SMP systems was proposed in 
Reinhardt and Karypis (<span class="CitationRef"><a href="#CR20" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2007</a></span>). Distributed map-reduce based graphlet and motif discovery methods have also been proposed&nbsp;(Shahrivari and Jalili <span class="CitationRef"><a href="#CR21" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2015</a></span>; Liu et&nbsp;al. <span class="CitationRef"><a href="#CR17" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2009</a></span>, where the task is to mine the frequency distribution of all vertex induced subgraphs with up to <em class="EmphasisTypeItalic ">k</em> vertices, for some small <em class="EmphasisTypeItalic ">k</em>.
 In contrast, for FSM, there is no restriction on the size of the 
patterns to be mined, but the minimum support can be used to prune the 
search space. Also related is the work on finding all matchings of a 
query subgraph in a larger graph (Sun et&nbsp;al. <span class="CitationRef"><a href="#CR23" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2012</a></span>; Afrati et&nbsp;al. <span class="CitationRef"><a href="#CR1" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2013</a></span>; Shao et&nbsp;al. <span class="CitationRef"><a href="#CR22" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>).
 However, the key difference is that here the query pattern is fixed 
(i.e., user provided), and one typically has to report one or all 
matches of the given query in the input graph. On the other hand, FSM is
 a much harder problem, since it has to first enumerate patterns from a 
exponentially large search space, and then it has to perform subgraph 
isomorphism checks to compute the support for each pattern.</p><p id="Par20" class="Para">Finally, large distributed graph processing platforms (e.g., <span class="ExternalRef"><a target="_blank" rel="noopener" href="http://spark.apache.org/GraphX"><span class="RefSource">spark.apache.org/GraphX</span></a></span>)
 have become very popular. However, these systems are mainly optimized 
for vertex centric graph algorithms (e.g., PageRank). The recent 
Arabesque framework (Teixeira et&nbsp;al. <span class="CitationRef"><a href="#CR24" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2015</a></span>)
 proposes a general embeddings centric or “think like an embedding” 
paradigm for graph problems like FSM, motif discovery, and clique 
finding. However, it assumes that the input graph fits in memory of each
 compute node. We propose the first distributed subgraph mining method 
for a partitioned input graph, and one that can handle web-scale 
networks spanning over a billion vertices and edges.</p></div></section><section id="Sec4" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">4 </span> <span class="EmphasisTypeSmallCaps ">DistGraph</span>: distributed mining on a partitioned input graph<span class="section-icon"></span></h2><div class="content"><div id="Par21" class="Para">Our
 distributed method allows for a massive input graph, that does not fit 
in the memory of any individual compute node. We adopt a partitioning 
approach, where we split the input graph into multiple partitions and 
perform the mining task in a distributed fashion. Furthermore, we design
 a hybrid algorithm that leverages multi-core CPUs within each compute 
node and leverages distributed computation over multiple compute nodes.<figure class="Figure" id="Fig2"><div class="MediaObject" id="MO4"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig2_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig2_HTML.gif" alt="Fig. 2"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 2</span> <p class="SimplePara">Distributed mining approach</p> </div></figcaption></figure> </div><div id="Par22" class="Para">Figure&nbsp;<span class="InternalRef"><a href="#Fig2">2</a></span> illustrates our generic distributed mining system framework. Given the massive input graph <span class="InlineEquation" id="IEq46"><span class="MathJax_Preview">G = (V,E)</span><script type="math/tex" id="MathJax-Element-48">G = (V,E)</script></span>, we first split it into <em class="EmphasisTypeItalic ">K</em> partitions, <span class="InlineEquation" id="IEq47"><span class="MathJax_Preview">G_i = (V_i, E_i), 1 \le i \le K</span><script type="math/tex" id="MathJax-Element-49">G_i = (V_i, E_i), 1 \le i \le K</script></span>. We assume that we have <em class="EmphasisTypeItalic ">p</em> compute nodes, with each one running a single process with up to <em class="EmphasisTypeItalic ">k</em> threads. For greater flexibility over the degree of partitioning and parallelism, a group of compute nodes, denoted <span class="InlineEquation" id="IEq48"><span class="MathJax_Preview">R_i</span><script type="math/tex" id="MathJax-Element-50">R_i</script></span>, handles the same input graph partition <span class="InlineEquation" id="IEq49"><span class="MathJax_Preview">G_i</span><script type="math/tex" id="MathJax-Element-51">G_i</script></span>. Thus, partition <span class="InlineEquation" id="IEq50"><span class="MathJax_Preview">G_i</span><script type="math/tex" id="MathJax-Element-52">G_i</script></span> is processed by group <span class="InlineEquation" id="IEq51"><span class="MathJax_Preview">R_i</span><script type="math/tex" id="MathJax-Element-53">R_i</script></span>. For load balancing, we assume that the number of compute nodes <em class="EmphasisTypeItalic ">p</em> is a multiple of <em class="EmphasisTypeItalic ">K</em>, so that each group has an equal number of nodes <span class="InlineEquation" id="IEq52"><span class="MathJax_Preview">r=p/K</span><script type="math/tex" id="MathJax-Element-54">r=p/K</script></span>.
 All communication between nodes, either within a group or across 
groups, is done via message passing, whereas all coordination within a 
process or compute node is done using shared-memory threads.<div id="Tab1" class="Table"><div class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Table 1</span> <p class="SimplePara"> <span class="EmphasisTypeSmallCaps ">AllGather</span> and <span class="EmphasisTypeSmallCaps ">AlltoAll</span> examples</p> </div></div><div class="u-scroll-horizontal"><table><colgroup><col class="tcol1 align-left"><col class="tcol2 align-left"><col class="tcol3 align-left"><col class="tcol4 align-left"><col class="tcol5 align-left"><col class="tcol6 align-left"><col class="tcol7 align-left"></colgroup><thead><tr><th rowspan="2"> <p class="SimplePara">Proc</p> </th><th> <p class="SimplePara"> <span class="EmphasisTypeSmallCaps ">AllGather</span> </p> </th><th>&nbsp;</th><th>&nbsp;</th><th> <p class="SimplePara"> <span class="EmphasisTypeSmallCaps ">AlltoAll</span> </p> </th><th>&nbsp;</th><th>&nbsp;</th></tr><tr><th> <p class="SimplePara">sent</p> </th><th>&nbsp;</th><th> <p class="SimplePara">Received</p> </th><th> <p class="SimplePara">sent</p> </th><th>&nbsp;</th><th> <p class="SimplePara">Received</p> </th></tr></thead><tbody><tr><td> <p class="SimplePara">0</p> </td><td> <p class="SimplePara">ABC</p> </td><td>&nbsp;</td><td> <p class="SimplePara">ABCDEFGHI</p> </td><td> <p class="SimplePara">ABC</p> </td><td>&nbsp;</td><td> <p class="SimplePara">ADG</p> </td></tr><tr><td> <p class="SimplePara">1</p> </td><td> <p class="SimplePara">DEF</p> </td><td> <p class="SimplePara"> <span class="InlineEquation" id="IEq53"><span class="MathJax_Preview">\implies </span><script type="math/tex" id="MathJax-Element-55">\implies </script></span> </p> </td><td> <p class="SimplePara">ABCDEFGHI</p> </td><td> <p class="SimplePara">DEF</p> </td><td> <p class="SimplePara"> <span class="InlineEquation" id="IEq54"><span class="MathJax_Preview">\implies </span><script type="math/tex" id="MathJax-Element-56">\implies </script></span> </p> </td><td> <p class="SimplePara">BEH</p> </td></tr><tr><td> <p class="SimplePara">2</p> </td><td> <p class="SimplePara">GHI</p> </td><td>&nbsp;</td><td> <p class="SimplePara">ABCDEFGHI</p> </td><td> <p class="SimplePara">GHI</p> </td><td>&nbsp;</td><td> <p class="SimplePara">CFI</p> </td></tr></tbody></table></div></div> </div><div id="Par23" class="Para"> <em class="EmphasisTypeItalic ">Collective communication:</em>
 Our distributed approach is generic in that it can be instantiated on a
 tightly-coupled HPC system (e.g., via the message passing interface 
(MPI) on an IBM Blue Gene supercomputer), or a loosely-coupled platform 
(e.g., via Hadoop or Spark on a cluster or cloud-based system). We rely 
only on collective communication operations, described below, that are 
easy to instantiate on most of the available distributed frameworks 
(e.g., they are natively available in MPI):<div class="UnorderedList"><ul class="UnorderedListMarkDash"><li> <p id="Par24" class="Para"> <span class="EmphasisTypeSmallCaps ">AllToAll</span>: This is a personalized broadcast operation which rearranges <em class="EmphasisTypeItalic ">n</em> items of data so that node <em class="EmphasisTypeItalic ">n</em> gets the <em class="EmphasisTypeItalic ">n</em>th
 item. We use a hash function to assign specific data to relevant 
compute nodes. Also, the sent data can have variable lengths. This is 
essentially a scatter (or map) operation followed by a gather (or 
reduce) operation performed by all nodes simultaneously.</p> </li><li> <p id="Par25" class="Para"> <span class="EmphasisTypeSmallCaps ">AllGather</span>:
 Through this operation a compute node can obtain or gather a piece of 
data from every other node, with all nodes doing this simultaneously. 
Again, the data can be of variable lengths. The difference between <span class="EmphasisTypeSmallCaps ">AlltoAll</span> and <span class="EmphasisTypeSmallCaps ">AllGather</span> is shown in Table&nbsp;<span class="InternalRef"><a href="#Tab1">1</a></span>.</p> </li><li> <p id="Par26" class="Para"> <span class="EmphasisTypeSmallCaps ">AllReduce</span>:
 Finally, this is a reduction operation performed simultaneously by all 
nodes on a specific operator, such as, SUM. After the operation all 
nodes will have the same result.</p> </li></ul></div>We now outline the 
challenges posed in distributed mining over a partitioned input graph, 
and our corresponding solutions, namely: (i) eliminating false negative 
patterns, (ii) allowing local pruning, and (iii) minimizing 
communication.</div><section id="Sec5" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">4.1 </span>Eliminating false negative patterns via external neighbors</h3><p id="Par27" class="Para">Recall that the support of a pattern <em class="EmphasisTypeItalic ">P</em> is given as <span class="InlineEquation" id="IEq55"><span class="MathJax_Preview">\sigma (P) = \min _{v \in V_P} \{|\varPhi (v)|\}</span><script type="math/tex" id="MathJax-Element-57">\sigma (P) = \min _{v \in V_P} \{|\varPhi (v)|\}</script></span>, where <span class="InlineEquation" id="IEq56"><span class="MathJax_Preview">\varPhi (v)</span><script type="math/tex" id="MathJax-Element-58">\varPhi (v)</script></span> denotes the set of unique vertex mappings of <span class="InlineEquation" id="IEq57"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex" id="MathJax-Element-59">v \in V_P</script></span> in the set of embeddings <span class="InlineEquation" id="IEq58"><span class="MathJax_Preview">\varSigma (P)</span><script type="math/tex" id="MathJax-Element-60">\varSigma (P)</script></span> in the input graph <em class="EmphasisTypeItalic ">G</em>. We also call <span class="InlineEquation" id="IEq59"><span class="MathJax_Preview">\sigma (P)</span><script type="math/tex" id="MathJax-Element-61">\sigma (P)</script></span> the <em class="EmphasisTypeItalic ">global support</em> of <em class="EmphasisTypeItalic ">P</em>, and say that <em class="EmphasisTypeItalic ">P</em> is <em class="EmphasisTypeItalic ">globally frequent</em> if <span class="InlineEquation" id="IEq60"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex" id="MathJax-Element-62">\sigma (P) \ge minsup</script></span>.</p><div id="Par28" class="Para">The first challenge in mining a partitioned graph is that there can be false negative patterns, i.e., a pattern <em class="EmphasisTypeItalic ">P</em> that is globally frequent can be missed due to the fact that certain edges involved in subgraph isomorphisms for <em class="EmphasisTypeItalic ">P</em> span different partitions. For example in Fig.&nbsp;<span class="InternalRef"><a href="#Fig3">3</a></span>, the pattern <span class="InlineEquation" id="IEq61"><span class="MathJax_Preview">A\text {--}B\text {--}C</span><script type="math/tex" id="MathJax-Element-63">A\text {--}B\text {--}C</script></span> has three embeddings, namely, (1,&nbsp;2,&nbsp;3), (5,&nbsp;2,&nbsp;3) and (10,&nbsp;11,&nbsp;7), and has global support <span class="InlineEquation" id="IEq62"><span class="MathJax_Preview">\sigma (P) = 2</span><script type="math/tex" id="MathJax-Element-64">\sigma (P) = 2</script></span> (since <em class="EmphasisTypeItalic ">B</em> and <em class="EmphasisTypeItalic ">C</em> both have two distinct mappings). However, after splitting <em class="EmphasisTypeItalic ">G</em> into two partitions <span class="InlineEquation" id="IEq63"><span class="MathJax_Preview">G_1</span><script type="math/tex" id="MathJax-Element-65">G_1</script></span> and <span class="InlineEquation" id="IEq64"><span class="MathJax_Preview">G_2</span><script type="math/tex" id="MathJax-Element-66">G_2</script></span> as shown, the embedding (10,&nbsp;11,&nbsp;7) will not be found in <span class="InlineEquation" id="IEq65"><span class="MathJax_Preview">G_2</span><script type="math/tex" id="MathJax-Element-67">G_2</script></span> since the edge (11,&nbsp;7) spans the partitions; it belongs to neither <span class="InlineEquation" id="IEq66"><span class="MathJax_Preview">G_1</span><script type="math/tex" id="MathJax-Element-68">G_1</script></span> nor <span class="InlineEquation" id="IEq67"><span class="MathJax_Preview">G_2</span><script type="math/tex" id="MathJax-Element-69">G_2</script></span>. If we expand <span class="InlineEquation" id="IEq68"><span class="MathJax_Preview">G_2</span><script type="math/tex" id="MathJax-Element-70">G_2</script></span>
 with its 1-hop neighborhood, we can find the embedding. Nevertheless, 
even with 1-hop overlap, subsequent patterns can still be missed. For 
example, <span class="InlineEquation" id="IEq69"><span class="MathJax_Preview">A\text {--}B\text {--}C\text {--}D</span><script type="math/tex" id="MathJax-Element-71">A\text {--}B\text {--}C\text {--}D</script></span> has support <span class="InlineEquation" id="IEq70"><span class="MathJax_Preview">\sigma (P)=2</span><script type="math/tex" id="MathJax-Element-72">\sigma (P)=2</script></span>,
 but we will not be able to find the embedding 
(10,&nbsp;11,&nbsp;7,&nbsp;6) or (10,&nbsp;11,&nbsp;7,&nbsp;4), since 
edges (7,&nbsp;6) and (7,&nbsp;4) do not belong to the 1-hop extension 
of <span class="InlineEquation" id="IEq71"><span class="MathJax_Preview">G_2</span><script type="math/tex" id="MathJax-Element-73">G_2</script></span>.<figure class="Figure" id="Fig3"><div class="MediaObject" id="MO5"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig3_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig3_HTML.gif" alt="Fig. 3"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 3</span> <p class="SimplePara">Partitioning graph <em class="EmphasisTypeItalic ">G</em> into <span class="InlineEquation" id="IEq72"><span class="MathJax_Preview">G_1</span><script type="math/tex" id="MathJax-Element-74">G_1</script></span> and <span class="InlineEquation" id="IEq73"><span class="MathJax_Preview">G_2</span><script type="math/tex" id="MathJax-Element-75">G_2</script></span>. Edges that are cut by the partitioning are shown <em class="EmphasisTypeItalic ">dashed</em> </p> </div></figcaption></figure> </div><p id="Par29" class="Para">To eliminate false negatives, we propose a technique we call <em class="EmphasisTypeItalic ">external neighbor expansion</em> as explained below. Our distributed approach relies on a level-wise or breadth-first approach, where at level <em class="EmphasisTypeItalic ">l</em> we compute the support of candidate subgraph patterns comprising <em class="EmphasisTypeItalic ">l</em> edges. The main idea in external neighbor expansion is that before computing patterns at level <em class="EmphasisTypeItalic ">l</em>, we expand the partitions <span class="InlineEquation" id="IEq74"><span class="MathJax_Preview">G_i</span><script type="math/tex" id="MathJax-Element-76">G_i</script></span>
 by requesting from other partitions the neighboring edges (and their 
end-points) for vertices in the embeddings for each frequent pattern <em class="EmphasisTypeItalic ">P</em> from the previous level <span class="InlineEquation" id="IEq75"><span class="MathJax_Preview">l-1</span><script type="math/tex" id="MathJax-Element-77">l-1</script></span>.</p><div id="Par30" class="Para">More formally, given the input graph <em class="EmphasisTypeItalic ">G</em>, we first split it into <em class="EmphasisTypeItalic ">K</em> <em class="EmphasisTypeItalic ">initial</em> partitions, <span class="InlineEquation" id="IEq76"><span class="MathJax_Preview">G^0_i = (V^0_i, E^0_i)</span><script type="math/tex" id="MathJax-Element-78">G^0_i = (V^0_i, E^0_i)</script></span> for <span class="InlineEquation" id="IEq77"><span class="MathJax_Preview">1 \le i \le K</span><script type="math/tex" id="MathJax-Element-79">1 \le i \le K</script></span>. Vertices <span class="InlineEquation" id="IEq78"><span class="MathJax_Preview">v \in V_i^0</span><script type="math/tex" id="MathJax-Element-80">v \in V_i^0</script></span> are called <em class="EmphasisTypeItalic ">local vertices</em>. The graph partition that is available in local memory is always denoted <span class="InlineEquation" id="IEq79"><span class="MathJax_Preview">G_i</span><script type="math/tex" id="MathJax-Element-81">G_i</script></span>, and initially <span class="InlineEquation" id="IEq80"><span class="MathJax_Preview">G_i \equiv G_i^0</span><script type="math/tex" id="MathJax-Element-82">G_i \equiv G_i^0</script></span>. Now consider level <span class="InlineEquation" id="IEq81"><span class="MathJax_Preview">l=1</span><script type="math/tex" id="MathJax-Element-83">l=1</script></span>, where we have to compute the support for all the single edge patterns. To avoid false negatives, we have to first extend <span class="InlineEquation" id="IEq82"><span class="MathJax_Preview">E_i^0</span><script type="math/tex" id="MathJax-Element-84">E_i^0</script></span> by adding edges that span across different partitions, such as (11,&nbsp;4) and (11,&nbsp;7) in Fig.&nbsp;<span class="InternalRef"><a href="#Fig3">3</a></span>. Thus, we create the extended local partition <span class="InlineEquation" id="IEq83"><span class="MathJax_Preview">G_i^1 = (V_i^1, E_i^1)</span><script type="math/tex" id="MathJax-Element-85">G_i^1 = (V_i^1, E_i^1)</script></span>, where<div id="Equ6" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned}
 V_i^1&amp;= V_i^0 \cup \big \{ v |\; (u, v) \in G, \; u \in V_i^0, 
\text { and } v \not \in V_i^0\big \}\\ E_i^1&amp;= E_i^0 \cup \big 
\{(u,v) |\; u \in V_i^0, v \in V_i^1\setminus V_i^0\big \} \end{aligned}</span><script type="math/tex; mode=display" id="MathJax-Element-86">\begin{aligned} V_i^1&= V_i^0 \cup \big \{ v |\; (u, v) \in G, \; u \in V_i^0, \text { and } v \not \in V_i^0\big \}\\ E_i^1&= E_i^0 \cup \big \{(u,v) |\; u \in V_i^0, v \in V_i^1\setminus V_i^0\big \} \end{aligned}</script></div></div>The extended graph partition that is now available in local memory is <span class="InlineEquation" id="IEq84"><span class="MathJax_Preview">G_i \equiv G_i^1</span><script type="math/tex">G_i \equiv G_i^1</script></span>.
 Due to external neighbor expansion all relevant edges that span 
partitions are available in local memory, which ensures that there will 
be no false negatives when determining the frequent single edge 
patterns.</div><div id="Par31" class="Para">Let <span class="InlineEquation" id="IEq85"><span class="MathJax_Preview">\mathcal {G}^1</span><script type="math/tex">\mathcal {G}^1</script></span> denote the set of all frequent single edge patterns. For any frequent pattern <span class="InlineEquation" id="IEq86"><span class="MathJax_Preview">P \in \mathcal {G}^1</span><script type="math/tex">P \in \mathcal {G}^1</script></span>, the set of mappings in partition <span class="InlineEquation" id="IEq87"><span class="MathJax_Preview">G_i^1</span><script type="math/tex">G_i^1</script></span> for each vertex <span class="InlineEquation" id="IEq88"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span> is given as: <span class="InlineEquation" id="IEq89"><span class="MathJax_Preview">\varPhi _i(v) = \displaystyle \cup _{\phi \in \varSigma _i(P)} \phi (v)</span><script type="math/tex">\varPhi _i(v) = \displaystyle \cup _{\phi \in \varSigma _i(P)} \phi (v)</script></span>. Before we begin level <span class="InlineEquation" id="IEq90"><span class="MathJax_Preview">l=2</span><script type="math/tex">l=2</script></span>, we perform external neighbor expansion, by creating expanded local partitions as follows:<div id="Equ7" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned}
 V_i^2&amp;= V_i^1 \cup \big \{ x |\; (x,y) \in G, \text { such that } y
 \in \varPhi _i(v), v \in V_P, P \in \mathcal {G}^1 \text { and } x \not
 \in G_i^1 \big \}\\ E_i^2&amp;= E_i^1 \cup \big \{(x,y) |\; y \in V_i^1
 \text { and } x \in V_i^2 \setminus V_i^1 \big \} \end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} V_i^2&= V_i^1 \cup \big \{ x |\; (x,y) \in G, \text { such that } y \in \varPhi _i(v), v \in V_P, P \in \mathcal {G}^1 \text { and } x \not \in G_i^1 \big \}\\ E_i^2&= E_i^1 \cup \big \{(x,y) |\; y \in V_i^1 \text { and } x \in V_i^2 \setminus V_i^1 \big \} \end{aligned}</script></div></div>In other words, we make sure that for any input graph vertex <em class="EmphasisTypeItalic ">x</em> that does not belong to <span class="InlineEquation" id="IEq91"><span class="MathJax_Preview">G_i^1</span><script type="math/tex">G_i^1</script></span>, but is a neighbor of some vertex <em class="EmphasisTypeItalic ">y</em> that belongs to the set of mappings for a frequent single edge pattern <em class="EmphasisTypeItalic ">P</em>, we request the external edge (<em class="EmphasisTypeItalic ">x</em>,&nbsp;<em class="EmphasisTypeItalic ">y</em>) via neighbor expansion. This way, the expanded local partition <span class="InlineEquation" id="IEq92"><span class="MathJax_Preview">G_i^2 = (V_i^2, E_i^2)</span><script type="math/tex">G_i^2 = (V_i^2, E_i^2)</script></span> will have all relevant edges for <span class="InlineEquation" id="IEq93"><span class="MathJax_Preview">l=2</span><script type="math/tex">l=2</script></span>, and the support of possible frequent patterns with two edges can be computed without false negatives.</div><div id="Par32" class="Para">In general, before we begin mining patterns at any level <span class="InlineEquation" id="IEq94"><span class="MathJax_Preview">l \ge 2</span><script type="math/tex">l \ge 2</script></span>,
 we make sure that all relevant external neighbors are available 
locally, where “relevant” means that these pertain to extensions of some
 frequent pattern from the previous level. Given the frequent patterns <span class="InlineEquation" id="IEq95"><span class="MathJax_Preview">\mathcal {G}^{l-1}</span><script type="math/tex">\mathcal {G}^{l-1}</script></span> from level <span class="InlineEquation" id="IEq96"><span class="MathJax_Preview">l-1</span><script type="math/tex">l-1</script></span>, and the set of mappings <span class="InlineEquation" id="IEq97"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> in <span class="InlineEquation" id="IEq98"><span class="MathJax_Preview">G_i^{l-1}</span><script type="math/tex">G_i^{l-1}</script></span> for vertices <em class="EmphasisTypeItalic ">v</em> in some frequent pattern <em class="EmphasisTypeItalic ">P</em>, the expanded partitions at level <em class="EmphasisTypeItalic ">l</em> are obtained as follows:<div id="Equ8" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned}
 V_i^l&amp;= V_i^{l-1} \cup \big \{ x |\; (x,y) \in G, \text { such that
 } y \in \varPhi _i(v), v \in V_P, P \in \mathcal {G}^{l-1} \text { and }
 x \not \in G_i^{l-1} \big \}\\ E_i^l&amp;= E_i^1 \cup \big \{(x,y) |\; y
 \in V_i^{l-1} \text { and } x \in V_i^l \setminus V_i^{l-1}\big \} 
\end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} V_i^l&= V_i^{l-1} \cup \big \{ x |\; (x,y) \in G, \text { such that } y \in \varPhi _i(v), v \in V_P, P \in \mathcal {G}^{l-1} \text { and } x \not \in G_i^{l-1} \big \}\\ E_i^l&= E_i^1 \cup \big \{(x,y) |\; y \in V_i^{l-1} \text { and } x \in V_i^l \setminus V_i^{l-1}\big \} \end{aligned}</script></div></div>The expanded partition is denoted as <span class="InlineEquation" id="IEq99"><span class="MathJax_Preview">G_i \equiv G_i^l = (V_i^l, E_i^l)</span><script type="math/tex">G_i \equiv G_i^l = (V_i^l, E_i^l)</script></span>. The vertices in the set <span class="InlineEquation" id="IEq100"><span class="MathJax_Preview">V_i^0</span><script type="math/tex">V_i^0</script></span> (corresponding to the initial partitioning), are called <em class="EmphasisTypeItalic ">local vertices</em>. The vertices in the set <span class="InlineEquation" id="IEq101"><span class="MathJax_Preview">V_i^l\setminus V_i^0</span><script type="math/tex">V_i^l\setminus V_i^0</script></span> are called <em class="EmphasisTypeItalic ">external vertices</em>, and likewise, the edges in the set <span class="InlineEquation" id="IEq102"><span class="MathJax_Preview">E_i^1\setminus E_i^0</span><script type="math/tex">E_i^1\setminus E_i^0</script></span> are called <em class="EmphasisTypeItalic ">external edges</em>. As before, for any pattern <em class="EmphasisTypeItalic ">P</em>, its embeddings in the expanded partition <span class="InlineEquation" id="IEq103"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span> are denoted as <span class="InlineEquation" id="IEq104"><span class="MathJax_Preview">\varSigma _i(P)</span><script type="math/tex">\varSigma _i(P)</script></span>, and the set of mappings in partition <span class="InlineEquation" id="IEq105"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span> for each vertex <span class="InlineEquation" id="IEq106"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span> is then given as <span class="InlineEquation" id="IEq107"><span class="MathJax_Preview">\varPhi _i(v) = \displaystyle \cup _{\phi \in \varSigma _i(P)} \phi (v)</span><script type="math/tex">\varPhi _i(v) = \displaystyle \cup _{\phi \in \varSigma _i(P)} \phi (v)</script></span>. Since the expanded partition comprises both local and external vertices, we can correspondingly divide the set of mappings <span class="InlineEquation" id="IEq108"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> into <em class="EmphasisTypeItalic ">local mappings</em>, denoted <span class="InlineEquation" id="IEq109"><span class="MathJax_Preview">\varPhi _i^l(v)</span><script type="math/tex">\varPhi _i^l(v)</script></span> and <em class="EmphasisTypeItalic ">external mappings</em>, denoted <span class="InlineEquation" id="IEq110"><span class="MathJax_Preview">\varPhi _i^e(v)</span><script type="math/tex">\varPhi _i^e(v)</script></span>, as follows:<div id="Equ9" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned}
 \text {Local Mappings: } \varPhi _i^l(v)&amp;= \big \{u \in G_i | \; 
\exists \phi \in \varSigma _i(P),\; u = \phi (v), u \in V_i^0\big \}\\ 
\text {External Mappings: } \varPhi _i^e(v)&amp;= \big \{u \in G_i | \; 
\exists \phi \in \varSigma _i(P),\; u = \phi (v), u \in V_i^l \setminus 
V_i^0\big \} \end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} \text {Local Mappings: } \varPhi _i^l(v)&= \big \{u \in G_i | \; \exists \phi \in \varSigma _i(P),\; u = \phi (v), u \in V_i^0\big \}\\ \text {External Mappings: } \varPhi _i^e(v)&= \big \{u \in G_i | \; \exists \phi \in \varSigma _i(P),\; u = \phi (v), u \in V_i^l \setminus V_i^0\big \} \end{aligned}</script></div></div> </div><div id="FPar1" class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading">Theorem 1</span></h3><p id="Par33" class="Para">At any level <em class="EmphasisTypeItalic ">l</em>, let <span class="InlineEquation" id="IEq111"><span class="MathJax_Preview">G_i \equiv G_i^l</span><script type="math/tex">G_i \equiv G_i^l</script></span> for <span class="InlineEquation" id="IEq112"><span class="MathJax_Preview">1\le i\le K</span><script type="math/tex">1\le i\le K</script></span> denote the set of expanded partitions. Then <span class="InlineEquation" id="IEq113"><span class="MathJax_Preview">\varSigma (P) = \bigcup _{i=1}^K \varSigma _i(P)</span><script type="math/tex">\varSigma (P) = \bigcup _{i=1}^K \varSigma _i(P)</script></span> for any pattern <em class="EmphasisTypeItalic ">P</em>.</p></div><div id="FPar2" class="FormalPara FormalParaRenderingStyle2"><h3 class="u-regular"><span class="Heading">Proof</span></h3><p id="Par34" class="Para">Let <em class="EmphasisTypeItalic ">P</em> be a pattern with <em class="EmphasisTypeItalic ">l</em> edges, and let <span class="InlineEquation" id="IEq114"><span class="MathJax_Preview">\phi \in \varSigma (P)</span><script type="math/tex">\phi \in \varSigma (P)</script></span>. We prove by induction that <span class="InlineEquation" id="IEq115"><span class="MathJax_Preview">\phi \in \varSigma _i(P)</span><script type="math/tex">\phi \in \varSigma _i(P)</script></span> for some <em class="EmphasisTypeItalic ">i</em>. Let <em class="EmphasisTypeItalic ">P</em> be obtained by adding the pattern edge (<em class="EmphasisTypeItalic ">u</em>,&nbsp;<em class="EmphasisTypeItalic ">v</em>) to a globally frequent <span class="InlineEquation" id="IEq116"><span class="MathJax_Preview">P'</span><script type="math/tex">P'</script></span> from the previous level, i.e., with <span class="InlineEquation" id="IEq117"><span class="MathJax_Preview">l-1</span><script type="math/tex">l-1</script></span> edges. Let <span class="InlineEquation" id="IEq118"><span class="MathJax_Preview">(x,y) = (\phi (u), \phi (v))</span><script type="math/tex">(x,y) = (\phi (u), \phi (v))</script></span>, where <span class="InlineEquation" id="IEq119"><span class="MathJax_Preview">x,y \in G</span><script type="math/tex">x,y \in G</script></span>. Either (<em class="EmphasisTypeItalic ">x</em>,&nbsp;<em class="EmphasisTypeItalic ">y</em>) was already in <span class="InlineEquation" id="IEq120"><span class="MathJax_Preview">G_i^{l-1}</span><script type="math/tex">G_i^{l-1}</script></span> or it was an external edge that was requested by external neighbor expansion before mining level <em class="EmphasisTypeItalic ">l</em>. In either case, the edge (<em class="EmphasisTypeItalic ">x</em>,&nbsp;<em class="EmphasisTypeItalic ">y</em>) must belong to some expanded partition <span class="InlineEquation" id="IEq121"><span class="MathJax_Preview">G_i^l</span><script type="math/tex">G_i^l</script></span>. Thus <span class="InlineEquation" id="IEq122"><span class="MathJax_Preview">\phi \in \varSigma _i(P)</span><script type="math/tex">\phi \in \varSigma _i(P)</script></span> for some <em class="EmphasisTypeItalic ">i</em>. <span class="InlineEquation" id="IEq123"><span class="MathJax_Preview">\square </span><script type="math/tex">\square </script></span> </p></div><p id="Par35" class="Para">Thus, external neighbor expansion ensures the set of embeddings for any pattern <em class="EmphasisTypeItalic ">P</em> at level <em class="EmphasisTypeItalic ">l</em> can be found from the embeddings of <em class="EmphasisTypeItalic ">P</em> in each expanded partition <span class="InlineEquation" id="IEq124"><span class="MathJax_Preview">G_i \equiv G_i^l</span><script type="math/tex">G_i \equiv G_i^l</script></span>. This will guarantee that there are no false negatives.</p></section><section id="Sec6" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">4.2 </span>Enabling local pruning via local support</h3><p id="Par36" class="Para">The second challenge we now face is that since the MRN support of a pattern <em class="EmphasisTypeItalic ">P</em>
 depends on the minimum number of unique mappings of its vertices, at 
first glance, it appears that we will not be able to prune any pattern 
locally. To see this, for a globally frequent pattern <em class="EmphasisTypeItalic ">P</em>, let <span class="InlineEquation" id="IEq125"><span class="MathJax_Preview">v^* \in V_P</span><script type="math/tex">v^* \in V_P</script></span> be the vertex with the minimum number of mappings, i.e., <span class="InlineEquation" id="IEq126"><span class="MathJax_Preview">v^* = \arg \min _{v \in V_P} \{ | \varPhi (v)| \}</span><script type="math/tex">v^* = \arg \min _{v \in V_P} \{ | \varPhi (v)| \}</script></span>. That is <span class="InlineEquation" id="IEq127"><span class="MathJax_Preview">\sigma (P) = |\varPhi (v^*)|</span><script type="math/tex">\sigma (P) = |\varPhi (v^*)|</script></span>. Now, for each partition <span class="InlineEquation" id="IEq128"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>, let <span class="InlineEquation" id="IEq129"><span class="MathJax_Preview">\varSigma _i(P)</span><script type="math/tex">\varSigma _i(P)</script></span> be the set of embeddings, and let <span class="InlineEquation" id="IEq130"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> be the corresponding set of mappings for any <span class="InlineEquation" id="IEq131"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span>. Just like the global support in Eq. (<span class="InternalRef"><a href="#Equ1">1</a></span>) we could define the local support of <em class="EmphasisTypeItalic ">P</em> in partition <span class="InlineEquation" id="IEq132"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span> as <span class="InlineEquation" id="IEq133"><span class="MathJax_Preview">\min _{v \in V_P} \{|\varPhi _i(v)|\}</span><script type="math/tex">\min _{v \in V_P} \{|\varPhi _i(v)|\}</script></span>. Further, let the minimum cardinality vertex in partition <span class="InlineEquation" id="IEq134"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span> be <span class="InlineEquation" id="IEq135"><span class="MathJax_Preview">v_i^* = \arg \min _{v \in V_P} \{ | \varPhi _i(v)| \}</span><script type="math/tex">v_i^* = \arg \min _{v \in V_P} \{ | \varPhi _i(v)| \}</script></span>. The problem is that different partitions may have different minimum cardinality vertices <span class="InlineEquation" id="IEq136"><span class="MathJax_Preview">v_i^*</span><script type="math/tex">v_i^*</script></span>, and they need not match <span class="InlineEquation" id="IEq137"><span class="MathJax_Preview">v^*</span><script type="math/tex">v^*</script></span>, and we thus we cannot make any guarantees about the global support of <em class="EmphasisTypeItalic ">P</em> using information from <span class="InlineEquation" id="IEq138"><span class="MathJax_Preview">\varPhi _i(v_i^*)</span><script type="math/tex">\varPhi _i(v_i^*)</script></span>. For example, in Fig.&nbsp;<span class="InternalRef"><a href="#Fig3">3</a></span>, consider the pattern <span class="InlineEquation" id="IEq139"><span class="MathJax_Preview">P\equiv A\text {--}B</span><script type="math/tex">P\equiv A\text {--}B</script></span>. The set of embeddings for <em class="EmphasisTypeItalic ">P</em> is given as <span class="InlineEquation" id="IEq140"><span class="MathJax_Preview">\varSigma (P) = \{(1,2), (5,2), (10,8), (10,9), (10,11)\}</span><script type="math/tex">\varSigma (P) = \{(1,2), (5,2), (10,8), (10,9), (10,11)\}</script></span>, and thus the set of mappings for the pattern vertex labeled <em class="EmphasisTypeItalic ">A</em> is <span class="InlineEquation" id="IEq141"><span class="MathJax_Preview">\{1,5,10\}</span><script type="math/tex">\{1,5,10\}</script></span>, and for <em class="EmphasisTypeItalic ">B</em> it is <span class="InlineEquation" id="IEq142"><span class="MathJax_Preview">\{ 2,8,9,11\}</span><script type="math/tex">\{ 2,8,9,11\}</script></span>. Thus, the minimum cardinality vertex is <em class="EmphasisTypeItalic ">A</em> and the global support is <span class="InlineEquation" id="IEq143"><span class="MathJax_Preview">\sigma (P) = 3</span><script type="math/tex">\sigma (P) = 3</script></span>. However, in partition <span class="InlineEquation" id="IEq144"><span class="MathJax_Preview">G^0_1</span><script type="math/tex">G^0_1</script></span> the minimum cardinality vertex is <em class="EmphasisTypeItalic ">B</em> with mappings <span class="InlineEquation" id="IEq145"><span class="MathJax_Preview">\{2\}</span><script type="math/tex">\{2\}</script></span>, and in <span class="InlineEquation" id="IEq146"><span class="MathJax_Preview">G^0_2</span><script type="math/tex">G^0_2</script></span> the minimum cardinality vertex is <em class="EmphasisTypeItalic ">A</em> with mappings <span class="InlineEquation" id="IEq147"><span class="MathJax_Preview">\{10\}</span><script type="math/tex">\{10\}</script></span>. As such there is no relationship between the cardinalities of <span class="InlineEquation" id="IEq148"><span class="MathJax_Preview">|\varPhi (v^*)|</span><script type="math/tex">|\varPhi (v^*)|</script></span> and <span class="InlineEquation" id="IEq149"><span class="MathJax_Preview">|\varPhi _i(v_i^*)|</span><script type="math/tex">|\varPhi _i(v_i^*)|</script></span>.
 So at first glance, it appears that our only recourse is to compute the
 embeddings for all possible patterns in each partition <span class="InlineEquation" id="IEq150"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>, without any local pruning, followed by a reduction step to compute the global support. We now show that we can do better.</p><div id="Par37" class="Para">Given the set of embeddings <span class="InlineEquation" id="IEq151"><span class="MathJax_Preview">\varSigma _i(P)</span><script type="math/tex">\varSigma _i(P)</script></span> for (expanded) partition <span class="InlineEquation" id="IEq152"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>, let <span class="InlineEquation" id="IEq153"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> be the corresponding set of mappings for any <span class="InlineEquation" id="IEq154"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span>. We define the <em class="EmphasisTypeItalic ">local support</em> of a pattern <em class="EmphasisTypeItalic ">P</em> as:<div id="Equ10" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned} \sigma _i(P) = \max _{v \in V_P} \{|\varPhi _i(v)|\} \end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} \sigma _i(P) = \max _{v \in V_P} \{|\varPhi _i(v)|\} \end{aligned}</script></div></div>and we say that pattern <em class="EmphasisTypeItalic ">P</em> is <em class="EmphasisTypeItalic ">locally frequent</em> if<div id="Equ2" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned} \sigma _i(P) \ge \frac{minsup}{K} \end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} \sigma _i(P) \ge \frac{minsup}{K} \end{aligned}</script></div> <div class="EquationNumber">(2)</div></div> </div><div id="FPar3" class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading">Theorem 2</span></h3><p id="Par38" class="Para">Let <span class="InlineEquation" id="IEq155"><span class="MathJax_Preview">\varSigma (P) = \cup _{i=1}^K \varSigma _i(P)</span><script type="math/tex">\varSigma (P) = \cup _{i=1}^K \varSigma _i(P)</script></span>, and let <span class="InlineEquation" id="IEq156"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>. Then there exists some partition such that <span class="InlineEquation" id="IEq157"><span class="MathJax_Preview">\sigma _i(P) \ge \frac{minsup}{K}</span><script type="math/tex">\sigma _i(P) \ge \frac{minsup}{K}</script></span>.</p></div><div id="FPar4" class="FormalPara FormalParaRenderingStyle2"><h3 class="u-regular"><span class="Heading">Proof</span></h3><div id="Par39" class="Para">Assume that for all partitions we have <span class="InlineEquation" id="IEq158"><span class="MathJax_Preview">\sigma _i(P) &lt; \frac{minsup}{K}</span><script type="math/tex">\sigma _i(P) < \frac{minsup}{K}</script></span>. Summing over all partitions we have<div id="Equ15" class="Equation"><div class="EquationContent"><div class="MediaObject" id="MO12"> <a href="https://static-content.springer.com/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Equ15_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Equ15_HTML.gif" alt=""></a> </div></div></div>However, note that<div id="Equ11" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned}
 \sum _{i=1}^K \max _{v \in V_P} \left\{ |\varPhi _i(v)|\right\} \ge 
\max _{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi _i(v)| \right\} \ge 
\min _{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi _i(v)| \right\} \ge 
\min _{v \in V_P} \left\{ |\varPhi (v)| \right\} \end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} \sum _{i=1}^K \max _{v \in V_P} \left\{ |\varPhi _i(v)|\right\} \ge \max _{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi _i(v)| \right\} \ge \min _{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi _i(v)| \right\} \ge \min _{v \in V_P} \left\{ |\varPhi (v)| \right\} \end{aligned}</script></div></div>Thus, we conclude that <span class="InlineEquation" id="IEq159"><span class="MathJax_Preview">minsup&gt; \min _{v \in V_P} \left\{ |\varPhi (v)| \right\} = \sigma (P)</span><script type="math/tex">minsup> \min _{v \in V_P} \left\{ |\varPhi (v)| \right\} = \sigma (P)</script></span>, which contradicts the fact that <span class="InlineEquation" id="IEq160"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>. This means our assumption was false, and there must exist at least one partition such that <span class="InlineEquation" id="IEq161"><span class="MathJax_Preview">\sigma _i(P) \ge \frac{minsup}{K}</span><script type="math/tex">\sigma _i(P) \ge \frac{minsup}{K}</script></span>. <span class="InlineEquation" id="IEq162"><span class="MathJax_Preview">\square </span><script type="math/tex">\square </script></span> </div></div><p id="Par40" class="Para">Thus, we need to compute the global support only for those patterns <em class="EmphasisTypeItalic ">P</em>
 that are locally frequent in at least one partition. That is, each 
compute node can find the embeddings in its local (expanded) partition <span class="InlineEquation" id="IEq163"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>,
 and we can prune patterns that are not locally frequent in at least one
 partition. This can cut down on a lot of unnecessary computation to 
determine the global support.</p></section><section id="Sec7" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">4.3 </span>Minimizing communication via support bounding</h3><p id="Par41" class="Para">The third challenge we face is how to cut down on the cost of communicating the embeddings <span class="InlineEquation" id="IEq164"><span class="MathJax_Preview">\varSigma _i(P)</span><script type="math/tex">\varSigma _i(P)</script></span> across the distributed system so as to estimate the global set of embeddings <span class="InlineEquation" id="IEq165"><span class="MathJax_Preview">\varSigma (P)</span><script type="math/tex">\varSigma (P)</script></span> for each pattern <em class="EmphasisTypeItalic ">P</em>.
 As such, a pattern can have an exponential number of embeddings, and 
likewise there can be an exponential number of embeddings. However, 
observe that we do not need to communicate <span class="InlineEquation" id="IEq166"><span class="MathJax_Preview">\varSigma _i(P)</span><script type="math/tex">\varSigma _i(P)</script></span> to all other nodes to correctly determine the global support of a pattern <em class="EmphasisTypeItalic ">P</em>; all we need is to send over the mappings <span class="InlineEquation" id="IEq167"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> for each vertex in the pattern, since <span class="InlineEquation" id="IEq168"><span class="MathJax_Preview">\varPhi (v) = \cup _{i=1}^K \varPhi _i(v)</span><script type="math/tex">\varPhi (v) = \cup _{i=1}^K \varPhi _i(v)</script></span>, and thus we can determine the global support <span class="InlineEquation" id="IEq169"><span class="MathJax_Preview">\sigma (P) = \min _v {|\varPhi (v)|}</span><script type="math/tex">\sigma (P) = \min _v {|\varPhi (v)|}</script></span>. Nevertheless, note that the size of communicating <span class="InlineEquation" id="IEq170"><span class="MathJax_Preview">\varPhi _i</span><script type="math/tex">\varPhi _i</script></span> is <span class="InlineEquation" id="IEq171"><span class="MathJax_Preview">O(|V_P| \times |V|)</span><script type="math/tex">O(|V_P| \times |V|)</script></span>,
 which is still too expensive for graphs with billions of vertices. 
Furthermore, we have to communicate this information not just for one 
pattern but for very many candidate patterns during the mining process.</p><p id="Par42" class="Para">Instead of communicating the entire set of mappings for each partition and each vertex, i.e., <span class="InlineEquation" id="IEq172"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span>, we communicate only the cardinality values, i.e., we exchange <span class="InlineEquation" id="IEq173"><span class="MathJax_Preview">\{ |\varPhi _i(v)|\}</span><script type="math/tex">\{ |\varPhi _i(v)|\}</script></span>, for all <span class="InlineEquation" id="IEq174"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span>. This drastically cuts down on the communication cost to only <span class="InlineEquation" id="IEq175"><span class="MathJax_Preview">O(|V_P|)</span><script type="math/tex">O(|V_P|)</script></span> per pattern, which is independent of the input graph size, and is orders of magnitude lower than communicating <span class="InlineEquation" id="IEq176"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span>,
 especially in very large graphs. Unfortunately, in this case we can no 
longer compute the exact global support based on only the cardinality 
information. We solve this problem via a two-step support bounding 
technique, described below.</p><div id="Par43" class="Para">As mentioned earlier, we divide the mappings <span class="InlineEquation" id="IEq177"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> for each vertex <span class="InlineEquation" id="IEq178"><span class="MathJax_Preview">v \in P</span><script type="math/tex">v \in P</script></span>, into local mappings, denoted <span class="InlineEquation" id="IEq179"><span class="MathJax_Preview">\varPhi _i^l(v)</span><script type="math/tex">\varPhi _i^l(v)</script></span> and external mappings, denoted <span class="InlineEquation" id="IEq180"><span class="MathJax_Preview">\varPhi _i^e(v)</span><script type="math/tex">\varPhi _i^e(v)</script></span>. The local mappings only involve vertices in the initial vertex partition <span class="InlineEquation" id="IEq181"><span class="MathJax_Preview">V_i^0</span><script type="math/tex">V_i^0</script></span> (i.e., without the neighborhood expansion step). The external mappings for <em class="EmphasisTypeItalic ">v</em> involve a vertex in some other partition <span class="InlineEquation" id="IEq182"><span class="MathJax_Preview">V_j^0</span><script type="math/tex">V_j^0</script></span> such that <span class="InlineEquation" id="IEq183"><span class="MathJax_Preview">j \ne i</span><script type="math/tex">j \ne i</script></span>. Note that the following property holds:<div id="Equ12" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned} |\varPhi _i(v)| = \big |\varPhi ^l_i(v)\big | + \big |\varPhi ^e_i(v)\big | \end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} |\varPhi _i(v)| = \big |\varPhi ^l_i(v)\big | + \big |\varPhi ^e_i(v)\big | \end{aligned}</script></div></div>Now, instead of communicating <span class="InlineEquation" id="IEq184"><span class="MathJax_Preview">|\varPhi _i(v)|</span><script type="math/tex">|\varPhi _i(v)|</script></span>, we communicate <span class="InlineEquation" id="IEq185"><span class="MathJax_Preview">|\varPhi ^l_i(v)|</span><script type="math/tex">|\varPhi ^l_i(v)|</script></span> and <span class="InlineEquation" id="IEq186"><span class="MathJax_Preview">|\varPhi ^e_i(v)|</span><script type="math/tex">|\varPhi ^e_i(v)|</script></span>.</div><div id="Par44" class="Para">Next, we derive a lower-bound and upper-bound on the support of <em class="EmphasisTypeItalic ">P</em>. Define the <em class="EmphasisTypeItalic ">lower-support estimate</em> of <em class="EmphasisTypeItalic ">P</em> as follows:<div id="Equ3" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned} \widehat{\sigma }_l(P) = min_{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi ^l_i(v)| \right\} \end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} \widehat{\sigma }_l(P) = min_{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi ^l_i(v)| \right\} \end{aligned}</script></div> <div class="EquationNumber">(3)</div></div>The theorem below shows that <span class="InlineEquation" id="IEq187"><span class="MathJax_Preview">\widehat{\sigma }_l(P)</span><script type="math/tex">\widehat{\sigma }_l(P)</script></span> is indeed a lower-bound for the true global support <span class="InlineEquation" id="IEq188"><span class="MathJax_Preview">\sigma (P)</span><script type="math/tex">\sigma (P)</script></span>.</div><div id="FPar5" class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading">Theorem 3</span></h3><p id="Par45" class="Para">It holds that <span class="InlineEquation" id="IEq189"><span class="MathJax_Preview">\widehat{\sigma }_l(P) \le \sigma (P)</span><script type="math/tex">\widehat{\sigma }_l(P) \le \sigma (P)</script></span>. Thus, if <span class="InlineEquation" id="IEq190"><span class="MathJax_Preview">\widehat{\sigma }_l(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_l(P) \ge minsup</script></span>, then <span class="InlineEquation" id="IEq191"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>.</p></div><div id="FPar6" class="FormalPara FormalParaRenderingStyle2"><h3 class="u-regular"><span class="Heading">Proof</span></h3><div id="Par46" class="Para">For any vertex <span class="InlineEquation" id="IEq192"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span>, we have<div id="Equ13" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned} \bigcup _{i=1}^K \varPhi ^l_i(v) \subseteq \bigcup _{i=1}^K \varPhi _i(v) = \varPhi (v)\end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} \bigcup _{i=1}^K \varPhi ^l_i(v) \subseteq \bigcup _{i=1}^K \varPhi _i(v) = \varPhi (v)\end{aligned}</script></div></div>However, since <span class="InlineEquation" id="IEq193"><span class="MathJax_Preview">\varPhi ^l(v)</span><script type="math/tex">\varPhi ^l(v)</script></span> are all disjoint by definition, this implies <span class="InlineEquation" id="IEq194"><span class="MathJax_Preview">\sum _{i=1}^K \left| \varPhi ^l_i(v)\right| \le |\varPhi (v)|</span><script type="math/tex">\sum _{i=1}^K \left| \varPhi ^l_i(v)\right| \le |\varPhi (v)|</script></span>. Therefore, <span class="InlineEquation" id="IEq195"><span class="MathJax_Preview">\widehat{\sigma
 }_l(P) = min_{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi 
^l_i(v)|\right\} \le min_{v \in V_P} \{|\varPhi (v)|\} = \sigma (P)</span><script type="math/tex">\widehat{\sigma }_l(P) = min_{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi ^l_i(v)|\right\} \le min_{v \in V_P} \{|\varPhi (v)|\} = \sigma (P)</script></span>. Hence, <span class="InlineEquation" id="IEq196"><span class="MathJax_Preview">\widehat{\sigma }_l(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_l(P) \ge minsup</script></span> implies <span class="InlineEquation" id="IEq197"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>. <span class="InlineEquation" id="IEq198"><span class="MathJax_Preview">\square </span><script type="math/tex">\square </script></span> </div></div><div id="Par47" class="Para">Finally, we derive an upper-bound on the support of <em class="EmphasisTypeItalic ">P</em>. Define the <em class="EmphasisTypeItalic ">upper-support estimate</em> of <em class="EmphasisTypeItalic ">P</em> as follows:<div id="Equ4" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned}
 \widehat{\sigma }_u(P)&amp;= min_{v \in V_P} \left\{ \sum _{i=1}^K 
|\varPhi _i(v)| \right\} = min_{v \in V_P} \left\{ \sum _{i=1}^K \left( 
|\varPhi ^l_i(v)| + |\varPhi ^e_i(v)| \right) \right\} \end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} \widehat{\sigma }_u(P)&= min_{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi _i(v)| \right\} = min_{v \in V_P} \left\{ \sum _{i=1}^K \left( |\varPhi ^l_i(v)| + |\varPhi ^e_i(v)| \right) \right\} \end{aligned}</script></div> <div class="EquationNumber">(4)</div></div>The theorem below shows that <span class="InlineEquation" id="IEq199"><span class="MathJax_Preview">\widehat{\sigma }_u(P)</span><script type="math/tex">\widehat{\sigma }_u(P)</script></span> is an upper-bound for the true global support <span class="InlineEquation" id="IEq200"><span class="MathJax_Preview">\sigma (P)</span><script type="math/tex">\sigma (P)</script></span>.</div><div id="FPar7" class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading">Theorem 4</span></h3><p id="Par48" class="Para">It holds that <span class="InlineEquation" id="IEq201"><span class="MathJax_Preview">\widehat{\sigma }_u(P) \ge \sigma (P)</span><script type="math/tex">\widehat{\sigma }_u(P) \ge \sigma (P)</script></span>. Thus, if <span class="InlineEquation" id="IEq202"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>, then <span class="InlineEquation" id="IEq203"><span class="MathJax_Preview">\widehat{\sigma }_u(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_u(P) \ge minsup</script></span>.</p></div><div id="FPar8" class="FormalPara FormalParaRenderingStyle2"><h3 class="u-regular"><span class="Heading">Proof</span></h3><p id="Par49" class="Para">For any vertex <span class="InlineEquation" id="IEq204"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span>, we have <span class="InlineEquation" id="IEq205"><span class="MathJax_Preview">\bigcup _{i=1}^K \varPhi _i(v) = \varPhi (v)</span><script type="math/tex">\bigcup _{i=1}^K \varPhi _i(v) = \varPhi (v)</script></span>, which in turn implies <span class="InlineEquation" id="IEq206"><span class="MathJax_Preview">\sum _{i=1}^K \left| \varPhi _i(v)\right| \ge |\varPhi (v)|</span><script type="math/tex">\sum _{i=1}^K \left| \varPhi _i(v)\right| \ge |\varPhi (v)|</script></span>. Therefore, <span class="InlineEquation" id="IEq207"><span class="MathJax_Preview">min_{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi _i(v)|\right\} \ge min_{v \in V_P} \{|\varPhi (v)|\} = \sigma (P)</span><script type="math/tex">min_{v \in V_P} \left\{ \sum _{i=1}^K |\varPhi _i(v)|\right\} \ge min_{v \in V_P} \{|\varPhi (v)|\} = \sigma (P)</script></span>. Thus, <span class="InlineEquation" id="IEq208"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span> implies <span class="InlineEquation" id="IEq209"><span class="MathJax_Preview">\widehat{\sigma }_u(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_u(P) \ge minsup</script></span>. <span class="InlineEquation" id="IEq210"><span class="MathJax_Preview">\square </span><script type="math/tex">\square </script></span> </p></div><p id="Par50" class="Para">Then, by the counter-positive of Thm.&nbsp;<span class="InternalRef"><a href="#FPar7">4</a></span>, it must be true that if <span class="InlineEquation" id="IEq211"><span class="MathJax_Preview">\widehat{\sigma }_u(P) &lt; minsup</span><script type="math/tex">\widehat{\sigma }_u(P) < minsup</script></span>, then <span class="InlineEquation" id="IEq212"><span class="MathJax_Preview">\sigma (P) &lt; minsup</span><script type="math/tex">\sigma (P) < minsup</script></span>.</p><p id="Par51" class="Para">The two-step support bounding technique is then used to prune patterns as follows. First, we check whether for a pattern <em class="EmphasisTypeItalic ">P</em> it holds that <span class="InlineEquation" id="IEq213"><span class="MathJax_Preview">\widehat{\sigma }_l(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_l(P) \ge minsup</script></span>. In this case, by Thm.&nbsp;<span class="InternalRef"><a href="#FPar5">3</a></span>, <em class="EmphasisTypeItalic ">P</em> must be globally frequent, and thus we can simply output the pattern. On the other hand, if <span class="InlineEquation" id="IEq214"><span class="MathJax_Preview">\widehat{\sigma }_l(P) &lt; minsup</span><script type="math/tex">\widehat{\sigma }_l(P) < minsup</script></span>, then we cannot be sure whether <em class="EmphasisTypeItalic ">P</em> is globally frequent or not. Second, we check whether <span class="InlineEquation" id="IEq215"><span class="MathJax_Preview">\widehat{\sigma }_u(P) &lt; minsup</span><script type="math/tex">\widehat{\sigma }_u(P) < minsup</script></span>. In this case, by the counter-positive of Thm.&nbsp;<span class="InternalRef"><a href="#FPar7">4</a></span>, <em class="EmphasisTypeItalic ">P</em> cannot possibly be globally frequent, and we can thus prune it from further consideration. On the other hand, if <span class="InlineEquation" id="IEq216"><span class="MathJax_Preview">\widehat{\sigma }_u(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_u(P) \ge minsup</script></span>, we call such patterns <em class="EmphasisTypeItalic ">potentially frequent patterns</em>, since <span class="InlineEquation" id="IEq217"><span class="MathJax_Preview">\widehat{\sigma }_u(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_u(P) \ge minsup</script></span>, does not imply that <span class="InlineEquation" id="IEq218"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>.
 For correctness, we have to determine their exact global support and 
eliminate the infrequent ones. To summarize, the basic idea behind 
two-step support bounding approach to minimize communication is to 
determine patterns that are definitely globally frequent, and to 
eliminate patterns that cannot possibly be globally frequent. For the 
remaining patterns we determine their exact global support and retain 
only the frequent ones. This approach is very effective in reducing the 
amount of communication in the distributed system.</p></section></div></section><section id="Sec8" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">5 </span> <span class="EmphasisTypeSmallCaps ">DistGraph</span>: algorithm details<span class="section-icon"></span></h2><div class="content"><div id="Par52" class="Para">Algorithm&nbsp;1 outlines the <span class="EmphasisTypeSmallCaps ">DistGraph</span> approach, which is executed by a group of compute nodes <span class="InlineEquation" id="IEq219"><span class="MathJax_Preview">R_i</span><script type="math/tex">R_i</script></span> working on the graph partition <span class="InlineEquation" id="IEq220"><span class="MathJax_Preview">G_i = (V_i, E_i)</span><script type="math/tex">G_i = (V_i, E_i)</script></span> as illustrated in Fig.&nbsp;<span class="InternalRef"><a href="#Fig2">2</a></span>.
 Starting from the single edge patterns, the candidates are extended in a
 level-by-level or breadth-first (BFS) fashion. This choice is dictated 
by the fact that in the partitioned approach, each compute node can only
 obtain the local support of a pattern. If we perform a DFS exploration,
 there will be an overwhelming number of potentially frequent patterns. 
Instead, a BFS exploration allows one to eliminate the globally 
infrequent patterns at the current level before proceeding to the next 
level.<figure class="Figure" id="Figa"><div class="MediaObject" id="MO18"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Figa_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Figa_HTML.gif" alt=""></a> </div></figure> </div><p id="Par53" class="Para">Let <span class="InlineEquation" id="IEq241"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span> be the set of globally frequent patterns from the previous level; initially <span class="InlineEquation" id="IEq242"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span> contains the empty graph. In Step&nbsp;4, the patterns <span class="InlineEquation" id="IEq243"><span class="MathJax_Preview">P \in \mathcal {P}_i</span><script type="math/tex">P \in \mathcal {P}_i</script></span> are extended by one edge and each valid candidate pattern is checked to see whether it is locally frequent (w.r.t. partition <span class="InlineEquation" id="IEq244"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>) or not, using Eq.&nbsp;(<span class="InternalRef"><a href="#Equ2">2</a></span>). Locally frequent candidates are added to the set <span class="InlineEquation" id="IEq245"><span class="MathJax_Preview">\mathcal {P}_{lf}</span><script type="math/tex">\mathcal {P}_{lf}</script></span>, and those not locally frequent are added to the set <span class="InlineEquation" id="IEq246"><span class="MathJax_Preview">\mathcal {P}_{nlf}</span><script type="math/tex">\mathcal {P}_{nlf}</script></span>. This step is performed across all compute nodes, using <em class="EmphasisTypeItalic ">k</em>
 threads each. Next, in Step&nbsp;5, we use two-step support-bounding to
 prune patterns that are definitely globally frequent and infrequent. 
The patterns that are guaranteed to be globally frequent based on the 
local-only support [via Eq.&nbsp;(<span class="InternalRef"><a href="#Equ3">3</a></span>)] are stored in a new set <span class="InlineEquation" id="IEq247"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span>. On the other hand, the patterns that are potentially globally frequent [via Eq.&nbsp;(<span class="InternalRef"><a href="#Equ4">4</a></span>)] are stored in <span class="InlineEquation" id="IEq248"><span class="MathJax_Preview">\mathcal {C}_i</span><script type="math/tex">\mathcal {C}_i</script></span>. For patterns in <span class="InlineEquation" id="IEq249"><span class="MathJax_Preview">\mathcal {C}_i</span><script type="math/tex">\mathcal {C}_i</script></span>, we determine their exact global support in Step&nbsp;6, and then remove the globally infrequent ones from <span class="InlineEquation" id="IEq250"><span class="MathJax_Preview">\mathcal {C}_i</span><script type="math/tex">\mathcal {C}_i</script></span>. After taking a union with <span class="InlineEquation" id="IEq251"><span class="MathJax_Preview">\mathcal {C}_i</span><script type="math/tex">\mathcal {C}_i</script></span>, the set <span class="InlineEquation" id="IEq252"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span> contains patterns that are globally frequent (i.e., have <span class="InlineEquation" id="IEq253"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>),
 and these are now exchanged with all groups in Step&nbsp;7, so that 
each group has the full set of globally frequent patterns (across the 
entire system) before the next level begins. However, each group retains
 only those patterns that are relevant to its partition <span class="InlineEquation" id="IEq254"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>.
 Finally, to eliminate false negatives, in Step&nbsp;8 we fetch relevant
 external neighbors from other partitions to create the expanded local 
partitions <span class="InlineEquation" id="IEq255"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>. The level-wise approach is then repeated until no more globally frequent patterns are found.</p><p id="Par54" class="Para">We
 now give detailed description of our approach for partitioning the 
input graph, computing local support, estimating the global support, 
support-bound based pruning to minimize communication, determining exact
 global support, and fetching external neighbors.</p><section id="Sec9" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">5.1 </span>Input graph partitioning</h3><p id="Par55" class="Para">Let <span class="InlineEquation" id="IEq256"><span class="MathJax_Preview">G= (V,E)</span><script type="math/tex">G= (V,E)</script></span> denote the massive input graph, and let <span class="InlineEquation" id="IEq257"><span class="MathJax_Preview">\{V_1^0, V_2^0, \ldots V_K^0\}</span><script type="math/tex">\{V_1^0, V_2^0, \ldots V_K^0\}</script></span> denote a <em class="EmphasisTypeItalic ">K</em>-way partitioning of the vertex set, such that partitions are pair-wise disjoint (i.e., <span class="InlineEquation" id="IEq258"><span class="MathJax_Preview">V^0_i \cap V^0_j = \emptyset </span><script type="math/tex">V^0_i \cap V^0_j = \emptyset </script></span>) and together encompass all the vertices (i.e., <span class="InlineEquation" id="IEq259"><span class="MathJax_Preview">\cup _i V^0_i = V</span><script type="math/tex">\cup _i V^0_i = V</script></span>). We denote the <em class="EmphasisTypeItalic ">i</em>-th initial partition as <span class="InlineEquation" id="IEq260"><span class="MathJax_Preview">G^0_i = (V^0_i, E^0_i)</span><script type="math/tex">G^0_i = (V^0_i, E^0_i)</script></span>, where the edge set <span class="InlineEquation" id="IEq261"><span class="MathJax_Preview">E^0_i</span><script type="math/tex">E^0_i</script></span> includes all edges in <em class="EmphasisTypeItalic ">E</em> both of whose end-points are in <span class="InlineEquation" id="IEq262"><span class="MathJax_Preview">V^0_i</span><script type="math/tex">V^0_i</script></span>, i.e., <span class="InlineEquation" id="IEq263"><span class="MathJax_Preview">E^0_i = \{ (u,v) | (u,v) \in E \text { and } u,v \in V^0_i \}</span><script type="math/tex">E^0_i = \{ (u,v) | (u,v) \in E \text { and } u,v \in V^0_i \}</script></span>.
 While any partitioning scheme, even a random one, can be used for graph
 partitioning, for good performance we need the partitions to be 
balanced and to minimize the edge cut (the number of vertices with an 
endpoint in another partition). Graph partitioning is a well-studied 
problem and solutions are generally derived using heuristics and 
approximation algorithms, since the balanced partitioning problem is 
NP-complete. We use the efficient PARMETIS&nbsp;(Karypis and Kumar <span class="CitationRef"><a href="#CR11" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">1998</a></span>) tool, which results in a good quality partitioning.</p><div id="Par56" class="Para">Furthermore,
 before the mining begins, we include the 1-hop overlap of the bordering
 edges between two neighboring partitions. That is, for any two 
partitions <span class="InlineEquation" id="IEq264"><span class="MathJax_Preview">G^0_i= (V^0_i, E^0_i)</span><script type="math/tex">G^0_i= (V^0_i, E^0_i)</script></span> and <span class="InlineEquation" id="IEq265"><span class="MathJax_Preview">G^0_j= (V^0_j, E^0_j)</span><script type="math/tex">G^0_j= (V^0_j, E^0_j)</script></span>, if there exists an edge (<em class="EmphasisTypeItalic ">u</em>,&nbsp;<em class="EmphasisTypeItalic ">v</em>) with <span class="InlineEquation" id="IEq266"><span class="MathJax_Preview">u \in V^0_i</span><script type="math/tex">u \in V^0_i</script></span> and <span class="InlineEquation" id="IEq267"><span class="MathJax_Preview">v \in V^0_j</span><script type="math/tex">v \in V^0_j</script></span>, then we add the edge (<em class="EmphasisTypeItalic ">u</em>,&nbsp;<em class="EmphasisTypeItalic ">v</em>) to both <span class="InlineEquation" id="IEq268"><span class="MathJax_Preview">E^0_i</span><script type="math/tex">E^0_i</script></span> and <span class="InlineEquation" id="IEq269"><span class="MathJax_Preview">E^0_j</span><script type="math/tex">E^0_j</script></span>. Thus, given the initial <em class="EmphasisTypeItalic ">K</em>-way partitioning of the vertices <span class="InlineEquation" id="IEq270"><span class="MathJax_Preview">\{V^0_1, \ldots , V^0_k\}</span><script type="math/tex">\{V^0_1, \ldots , V^0_k\}</script></span>, the expended graph partitions <span class="InlineEquation" id="IEq271"><span class="MathJax_Preview">G^1_i = (V^1_i, E^1_i)</span><script type="math/tex">G^1_i = (V^1_i, E^1_i)</script></span> are defined as follows:<div id="Equ14" class="Equation EquationMathjax"><div class="EquationContent"><span class="MathJax_Preview">\begin{aligned}
 E^1_i&amp;= \big \{(u,v) \in E\; |\; u \in V^0_i \text { and } v \in 
V\big \}\\ V^1_i&amp;= V_i^0 \cup \{v\; |\; (u,v) \in E, u \in V_i \text
 { and } v \not \in V_i \} \end{aligned}</span><script type="math/tex; mode=display">\begin{aligned} E^1_i&= \big \{(u,v) \in E\; |\; u \in V^0_i \text { and } v \in V\big \}\\ V^1_i&= V_i^0 \cup \{v\; |\; (u,v) \in E, u \in V_i \text { and } v \not \in V_i \} \end{aligned}</script></div></div>After 1-hop expansion, the expended partitions are given as <span class="InlineEquation" id="IEq272"><span class="MathJax_Preview">G_i \equiv G_i^1</span><script type="math/tex">G_i \equiv G_i^1</script></span>.
 Note that in the discussion below we avoid using the superscripts for 
the expanded partitions to reduce clutter in the notation; <span class="InlineEquation" id="IEq273"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span> always refers to the expanded local partitions before the next level begins.</div></section><section id="Sec10" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">5.2 </span>Computing local support</h3><div id="Par57" class="Para">Given the current (expanded) local partition <span class="InlineEquation" id="IEq274"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>, and the set of globally frequent patterns <span class="InlineEquation" id="IEq275"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span> from the previous level, Alg.&nbsp;<span class="InternalRef"><a href="#Sec10">2</a></span> shows the steps to extend each pattern <span class="InlineEquation" id="IEq276"><span class="MathJax_Preview">P \in \mathcal {P}_i</span><script type="math/tex">P \in \mathcal {P}_i</script></span> and compute its local support. This function is carried out in parallel by all the <em class="EmphasisTypeItalic ">k</em> threads within each compute node, as well as collectively by all the compute nodes in each group <span class="InlineEquation" id="IEq277"><span class="MathJax_Preview">R_i</span><script type="math/tex">R_i</script></span>. That is, the patterns in <span class="InlineEquation" id="IEq278"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span> are divided up equally among all available nodes/threads.<figure class="Figure" id="Figb"><div class="MediaObject" id="MO20"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Figb_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Figb_HTML.gif" alt=""></a> </div></figure> </div><div id="Par58" class="Para">For a specific pattern <em class="EmphasisTypeItalic ">P</em>, each processing element first computes its embeddings <span class="InlineEquation" id="IEq300"><span class="MathJax_Preview">\varSigma _i(P)</span><script type="math/tex">\varSigma _i(P)</script></span> in <span class="InlineEquation" id="IEq301"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>. Next, from each embedding, the set of all edge extensions are collected in <span class="InlineEquation" id="IEq302"><span class="MathJax_Preview">\mathcal{E}_i(P)</span><script type="math/tex">\mathcal{E}_i(P)</script></span>.
 For systematic and non-redundant graph enumeration, we use the DFScode 
for canonicality checking and rightmost path edge extensions (Yan and 
Han <span class="CitationRef"><a href="#CR27" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2002</a></span>), as detailed below in Alg.&nbsp;<span class="InternalRef"><a href="#Sec10">3</a></span>. Given the set of all extensions from each embedding of <em class="EmphasisTypeItalic ">P</em>, we can compute the local support of each extension. The pattern <em class="EmphasisTypeItalic ">P</em> is extended by each of the new edges <span class="InlineEquation" id="IEq303"><span class="MathJax_Preview">e \in \mathcal{E}_i(P)</span><script type="math/tex">e \in \mathcal{E}_i(P)</script></span> to obtain the pattern <span class="InlineEquation" id="IEq304"><span class="MathJax_Preview">P'</span><script type="math/tex">P'</script></span>, which is retained only if it is canonical. Finally, <span class="InlineEquation" id="IEq305"><span class="MathJax_Preview">P'</span><script type="math/tex">P'</script></span> is either added to <span class="InlineEquation" id="IEq306"><span class="MathJax_Preview">\mathcal {P}_{lf}</span><script type="math/tex">\mathcal {P}_{lf}</script></span> if it is locally frequent, or else it is added to <span class="InlineEquation" id="IEq307"><span class="MathJax_Preview">\mathcal {P}_{nlf}</span><script type="math/tex">\mathcal {P}_{nlf}</script></span>. It is important to note that we discard all of the embeddings <span class="InlineEquation" id="IEq308"><span class="MathJax_Preview">\varSigma _i(P)</span><script type="math/tex">\varSigma _i(P)</script></span> after computing the cardinalities of the vertex mappings <span class="InlineEquation" id="IEq309"><span class="MathJax_Preview">|\varPhi ^l_i(v)|</span><script type="math/tex">|\varPhi ^l_i(v)|</script></span> and <span class="InlineEquation" id="IEq310"><span class="MathJax_Preview">|\varPhi ^e_i(v)|</span><script type="math/tex">|\varPhi ^e_i(v)|</script></span> for each vertex <span class="InlineEquation" id="IEq311"><span class="MathJax_Preview">v\in V_P</span><script type="math/tex">v\in V_P</script></span>, otherwise the amount of memory required to retain all embeddings can easily exceed available resources.<figure class="Figure" id="Figc"><div class="MediaObject" id="MO21"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Figc_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Figc_HTML.gif" alt=""></a> </div></figure> </div><p id="Par59" class="Para">We now discuss Alg.&nbsp;<span class="InternalRef"><a href="#Sec10">3</a></span>
 which identifies the valid extensions, and also determines which 
external vertices are be fetched later to create the expanded partitions
 <span class="InlineEquation" id="IEq324"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>. For a pattern <span class="InlineEquation" id="IEq325"><span class="MathJax_Preview">P</span><script type="math/tex">P</script></span>, Alg.&nbsp;<span class="InternalRef"><a href="#Sec10">3</a></span> finds all valid backward and forward extensions of its embeddings in <span class="InlineEquation" id="IEq326"><span class="MathJax_Preview">\varSigma _i</span><script type="math/tex">\varSigma _i</script></span>.
 Backward extensions are those that introduce cycles in an existing 
patterns, whereas forward extensions introduce a new vertex with the 
corresponding edge. Note that extensions happen only along the 
right-most path; see Yan and Han (<span class="CitationRef"><a href="#CR27" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2002</a></span>) for more details. Note that only forward extensions can introduce an external vertex. That is, if (<em class="EmphasisTypeItalic ">u</em>,&nbsp;<em class="EmphasisTypeItalic ">v</em>) is a forward extension where <span class="InlineEquation" id="IEq327"><span class="MathJax_Preview">u \in V_i</span><script type="math/tex">u \in V_i</script></span> but <span class="InlineEquation" id="IEq328"><span class="MathJax_Preview">v \notin V_i</span><script type="math/tex">v \notin V_i</script></span> we would need to fetch this vertex for the next extension if <em class="EmphasisTypeItalic ">P</em> turns out to be globally frequent. So in steps 4–6, we collect such external neighbors for <em class="EmphasisTypeItalic ">P</em> and store them in the set <span class="InlineEquation" id="IEq329"><span class="MathJax_Preview">Ext_i[P]</span><script type="math/tex">Ext_i[P]</script></span> for use later in the external neighbor expansion step. So we add to <span class="InlineEquation" id="IEq330"><span class="MathJax_Preview">Ext_i[P]</span><script type="math/tex">Ext_i[P]</script></span> the pair (<em class="EmphasisTypeItalic ">v</em>,&nbsp;<em class="EmphasisTypeItalic ">j</em>) where <em class="EmphasisTypeItalic ">v</em> is the external vertex id, and <em class="EmphasisTypeItalic ">j</em> is the partition id for <em class="EmphasisTypeItalic ">v</em>. Then, in Alg.&nbsp;<span class="InternalRef"><a href="#Sec13">6</a></span> the actual requests are made to the corresponding partitions only if <em class="EmphasisTypeItalic ">P</em> is found to be globally frequent.</p></section><section id="Sec11" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">5.3 </span>Support-bound based pruning</h3><div id="Par60" class="Para">Alg.&nbsp;<span class="InternalRef"><a href="#Sec11">4</a></span> outlines the steps in computing the lower and upper support estimates for all locally frequent patterns <em class="EmphasisTypeItalic ">P</em>,
 so that we can determine the globally frequent patterns. We distribute 
the task of support-bound estimation among all compute nodes.<figure class="Figure" id="Figd"><div class="MediaObject" id="MO22"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Figd_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Figd_HTML.gif" alt=""></a> </div></figure> </div><p id="Par61" class="Para">First,
 only those patterns that are locally frequent in some partition are 
considered. To achieve this, in Step&nbsp;1, each pattern in <span class="InlineEquation" id="IEq365"><span class="MathJax_Preview">\mathcal {P}_{lf}</span><script type="math/tex">\mathcal {P}_{lf}</script></span>
 is assigned to a designated compute node via hashing on its unique 
DFScode. Each compute node distributes the patterns and their vertex 
mapping cardinalities <span class="InlineEquation" id="IEq366"><span class="MathJax_Preview">|\varPhi ^l_i(v)|</span><script type="math/tex">|\varPhi ^l_i(v)|</script></span> and <span class="InlineEquation" id="IEq367"><span class="MathJax_Preview">|\varPhi ^e_i(v)|, \forall v \in V_P</span><script type="math/tex">|\varPhi ^e_i(v)|, \forall v \in V_P</script></span>, to the designated compute node via <span class="EmphasisTypeSmallCaps ">AllToAll</span>.
 All compute nodes accumulate the received vertex mapping cardinalities 
for their assigned patterns and determine their lower-bound and 
upper-bound supports, <span class="InlineEquation" id="IEq368"><span class="MathJax_Preview">\widehat{\sigma }_l(P)</span><script type="math/tex">\widehat{\sigma }_l(P)</script></span> and <span class="InlineEquation" id="IEq369"><span class="MathJax_Preview">\widehat{\sigma }_u(P)</span><script type="math/tex">\widehat{\sigma }_u(P)</script></span>, respectively. If pattern <em class="EmphasisTypeItalic ">P</em> satisfies <span class="InlineEquation" id="IEq370"><span class="MathJax_Preview">\widehat{\sigma }_l(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_l(P) \ge minsup</script></span>, then it is definitely globally frequent, and is added to the set <span class="InlineEquation" id="IEq371"><span class="MathJax_Preview">\mathcal {F}</span><script type="math/tex">\mathcal {F}</script></span>. Otherwise, if <em class="EmphasisTypeItalic ">P</em> satisfies <span class="InlineEquation" id="IEq372"><span class="MathJax_Preview">\widehat{\sigma }_u(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_u(P) \ge minsup</script></span>, then we add it to the set of potentially frequent patterns <span class="InlineEquation" id="IEq373"><span class="MathJax_Preview">\mathcal {G}</span><script type="math/tex">\mathcal {G}</script></span>. Now, the patterns with <span class="InlineEquation" id="IEq374"><span class="MathJax_Preview">\widehat{\sigma }_u(P) &lt; minsup</span><script type="math/tex">\widehat{\sigma }_u(P) < minsup</script></span> comprise the set <span class="InlineEquation" id="IEq375"><span class="MathJax_Preview">\mathcal {Q}</span><script type="math/tex">\mathcal {Q}</script></span>;
 such patterns can be infrequent either because they are globally 
infrequent, or it may be because they were locally infrequent in one or 
more partitions, and thus their vertex mapping cardinalities were not 
communicated in Step&nbsp;1. To eliminate this possibility, we first 
communicate the patterns in <span class="InlineEquation" id="IEq376"><span class="MathJax_Preview">\mathcal {Q}</span><script type="math/tex">\mathcal {Q}</script></span> to all compute nodes via <span class="EmphasisTypeSmallCaps ">AllGather</span> (Step&nbsp;12). The unified set of such patterns is denoted <span class="InlineEquation" id="IEq377"><span class="MathJax_Preview">\mathcal {Q}^G</span><script type="math/tex">\mathcal {Q}^G</script></span>. Next, each compute node looks up the vertex mapping cardinalities for these patterns in <span class="InlineEquation" id="IEq378"><span class="MathJax_Preview">\mathcal {P}_{nlf}</span><script type="math/tex">\mathcal {P}_{nlf}</script></span>, and sends them over to the designated compute node via <span class="EmphasisTypeSmallCaps ">AllToAll</span> in Step&nbsp;13. Note that we do not consider patterns in <span class="InlineEquation" id="IEq379"><span class="MathJax_Preview">\mathcal {P}_{lf}</span><script type="math/tex">\mathcal {P}_{lf}</script></span>, since they were sent earlier. Given the mapping cardinalities, each compute node can again compute the estimates <span class="InlineEquation" id="IEq380"><span class="MathJax_Preview">\widehat{\sigma }_l(P)</span><script type="math/tex">\widehat{\sigma }_l(P)</script></span> and <span class="InlineEquation" id="IEq381"><span class="MathJax_Preview">\widehat{\sigma }_u(P)</span><script type="math/tex">\widehat{\sigma }_u(P)</script></span> for patterns in <span class="InlineEquation" id="IEq382"><span class="MathJax_Preview">\mathcal {Q}^G</span><script type="math/tex">\mathcal {Q}^G</script></span>, and add them to either <span class="InlineEquation" id="IEq383"><span class="MathJax_Preview">\mathcal {F}</span><script type="math/tex">\mathcal {F}</script></span> or <span class="InlineEquation" id="IEq384"><span class="MathJax_Preview">\mathcal {G}</span><script type="math/tex">\mathcal {G}</script></span>, accordingly.</p><p id="Par62" class="Para">It is important to stress that <span class="EmphasisTypeSmallCaps ">DistGraph</span> does not communicate the vertex mapping cardinalities for all patterns in <span class="InlineEquation" id="IEq385"><span class="MathJax_Preview">\mathcal {P}_{lf}</span><script type="math/tex">\mathcal {P}_{lf}</script></span> and <span class="InlineEquation" id="IEq386"><span class="MathJax_Preview">\mathcal {P}_{nlf}</span><script type="math/tex">\mathcal {P}_{nlf}</script></span>.
 The global support is estimated only for those patterns that are 
locally frequent in at least one partition. Next, only for these 
patterns, we communicate the mapping counts in <span class="InlineEquation" id="IEq387"><span class="MathJax_Preview">\mathcal {P}_{nlf}</span><script type="math/tex">\mathcal {P}_{nlf}</script></span>, if we determine that the estimated global support is below <span class="InlineEquation" id="IEq388"><span class="MathJax_Preview">minsup</span><script type="math/tex">minsup</script></span>.
 Our approach thus performs only the necessary amount of communication. 
Also, whereas compute nodes can be selective in not requesting the 
mapping counts from the nodes where a pattern <em class="EmphasisTypeItalic ">P</em>
 is already locally frequent (and thus we will have access to its 
mapping counts from Step&nbsp;1), such cases are typically too few to 
outperform the efficient collective communication done in Step&nbsp;12.</p></section><section id="Sec12" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">5.4 </span>Computing exact global support</h3><div id="Par63" class="Para">To determine the exact global support for a pattern <em class="EmphasisTypeItalic ">P</em> we need to have access to all its vertex mappings <span class="InlineEquation" id="IEq389"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> in each partition <span class="InlineEquation" id="IEq390"><span class="MathJax_Preview">G_i = (V_i, E_i)</span><script type="math/tex">G_i = (V_i, E_i)</script></span> for each <span class="InlineEquation" id="IEq391"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span>. However, after the local support computation phase, we have access to only the mapping cardinalities <span class="InlineEquation" id="IEq392"><span class="MathJax_Preview">|\varPhi ^l_i(v)|</span><script type="math/tex">|\varPhi ^l_i(v)|</script></span> and <span class="InlineEquation" id="IEq393"><span class="MathJax_Preview">|\varPhi ^e_i(v)|</span><script type="math/tex">|\varPhi ^e_i(v)|</script></span> for each vertex, but not the mappings <span class="InlineEquation" id="IEq394"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> or the embeddings <span class="InlineEquation" id="IEq395"><span class="MathJax_Preview">\varSigma _i(P)</span><script type="math/tex">\varSigma _i(P)</script></span>. The vertex mappings <span class="InlineEquation" id="IEq396"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> can be obtained in several ways: (i) <em class="EmphasisTypeItalic ">memory-based:</em> keep the mappings <span class="InlineEquation" id="IEq397"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> in memory for all possible patterns, but communicate the external mappings only if <span class="InlineEquation" id="IEq398"><span class="MathJax_Preview">P \in \mathcal {C}_i</span><script type="math/tex">P \in \mathcal {C}_i</script></span>.
 This approach consumes too much memory, since it is not possible to 
retain the mappings for all patterns (both locally infrequent or 
frequent) for all except the smallest of graphs with few embeddings per 
pattern; (ii) <em class="EmphasisTypeItalic ">file-based:</em> we can store the set of mappings <span class="InlineEquation" id="IEq399"><span class="MathJax_Preview">\varPhi _i</span><script type="math/tex">\varPhi _i</script></span> to local disk for all (infrequent and frequent) patterns. For each pattern <span class="InlineEquation" id="IEq400"><span class="MathJax_Preview">P \in \mathcal {C}_i</span><script type="math/tex">P \in \mathcal {C}_i</script></span>, we can then read the mappings from disk; iii) <em class="EmphasisTypeItalic ">regenerate embeddings:</em>
 This is our default approach, where we regenerate or recompute the 
embeddings for each pattern in a distributed manner. We experimentally 
evaluate these approaches in Sect.&nbsp;<span class="InternalRef"><a href="#Sec15">6</a></span>.<figure class="Figure" id="Fige"><div class="MediaObject" id="MO23"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fige_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fige_HTML.gif" alt=""></a> </div></figure> </div><p id="Par64" class="Para">Alg.&nbsp;<span class="InternalRef"><a href="#Sec12">5</a></span> describes how we determine the exact global support for all patterns <span class="InlineEquation" id="IEq419"><span class="MathJax_Preview">P \in \mathcal {C}_i</span><script type="math/tex">P \in \mathcal {C}_i</script></span>. We divide up the set <span class="InlineEquation" id="IEq420"><span class="MathJax_Preview">\mathcal {C}_i</span><script type="math/tex">\mathcal {C}_i</script></span> among the compute nodes in the group <span class="InlineEquation" id="IEq421"><span class="MathJax_Preview">R_i</span><script type="math/tex">R_i</script></span> to carry out the exact support determination in a distributed manner. Further, all the <em class="EmphasisTypeItalic ">k</em> threads are also involved in regenerating the mappings <span class="InlineEquation" id="IEq422"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> in parallel for all <span class="InlineEquation" id="IEq423"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span> in Step&nbsp;3. The external mappings for <em class="EmphasisTypeItalic ">v</em> involve a vertex in some other partition <span class="InlineEquation" id="IEq424"><span class="MathJax_Preview">V_j</span><script type="math/tex">V_j</script></span> (<span class="InlineEquation" id="IEq425"><span class="MathJax_Preview">j \ne i</span><script type="math/tex">j \ne i</script></span>). Therefore, <span class="InlineEquation" id="IEq426"><span class="MathJax_Preview">\varPhi ^e_i(v) = \bigcup _{j \ne i}^K \varPhi ^e_{ij}(v) </span><script type="math/tex">\varPhi ^e_i(v) = \bigcup _{j \ne i}^K \varPhi ^e_{ij}(v) </script></span>, where <span class="InlineEquation" id="IEq427"><span class="MathJax_Preview">\varPhi ^e_{ij}(v)</span><script type="math/tex">\varPhi ^e_{ij}(v)</script></span> denotes the set of external mappings in partition <span class="InlineEquation" id="IEq428"><span class="MathJax_Preview">V_i</span><script type="math/tex">V_i</script></span> that come from partition <span class="InlineEquation" id="IEq429"><span class="MathJax_Preview">V_j</span><script type="math/tex">V_j</script></span>. Also, we ignore <span class="InlineEquation" id="IEq430"><span class="MathJax_Preview">\varPhi ^e_{ii}(v)</span><script type="math/tex">\varPhi ^e_{ii}(v)</script></span>, since it is essentially <span class="InlineEquation" id="IEq431"><span class="MathJax_Preview">\varPhi ^l_i(v)</span><script type="math/tex">\varPhi ^l_i(v)</script></span>. The local mappings do not have to be communicated to any other group. However, we use the <span class="EmphasisTypeSmallCaps ">AllToAll</span> operation to communicate to group <span class="InlineEquation" id="IEq432"><span class="MathJax_Preview">R_j</span><script type="math/tex">R_j</script></span> the external mappings <span class="InlineEquation" id="IEq433"><span class="MathJax_Preview">\varPhi ^e_{ij}(v)</span><script type="math/tex">\varPhi ^e_{ij}(v)</script></span> involving initial partition <span class="InlineEquation" id="IEq434"><span class="MathJax_Preview">V_j</span><script type="math/tex">V_j</script></span> (Step&nbsp;4). At the end of this exchange, each group <span class="InlineEquation" id="IEq435"><span class="MathJax_Preview">R_j</span><script type="math/tex">R_j</script></span> has all the local vertex mappings for each vertex <em class="EmphasisTypeItalic ">v</em> in pattern <em class="EmphasisTypeItalic ">P</em>, which it uses to update <span class="InlineEquation" id="IEq436"><span class="MathJax_Preview">\varPhi ^l_i(v)</span><script type="math/tex">\varPhi ^l_i(v)</script></span> (Step&nbsp;5). Formally, we can write <span class="InlineEquation" id="IEq437"><span class="MathJax_Preview">\varPhi ^l_i(v) = \varPhi ^l_i(v) \cup \{ \cup _{j \ne i}^K \varPhi ^e_{ji}(v) \}</span><script type="math/tex">\varPhi ^l_i(v) = \varPhi ^l_i(v) \cup \{ \cup _{j \ne i}^K \varPhi ^e_{ji}(v) \}</script></span>. Since a graph vertex can belong to at most one partition, we can then compute the true global support of <em class="EmphasisTypeItalic ">P</em> as <span class="InlineEquation" id="IEq438"><span class="MathJax_Preview">\sigma (P) = \min _{v \in V_P}\sum _{i=1}^{K}{|\varPhi ^l_i(v)|}</span><script type="math/tex">\sigma (P) = \min _{v \in V_P}\sum _{i=1}^{K}{|\varPhi ^l_i(v)|}</script></span>, which is accomplished by performing an <span class="EmphasisTypeSmallCaps ">AllReduce</span>
 operation on the local vertex mappings and considering the minimum 
value (Step&nbsp;6). Finally, the pattern is marked as infrequent if the
 support is below <span class="InlineEquation" id="IEq439"><span class="MathJax_Preview">minsup</span><script type="math/tex">minsup</script></span>, and each group removes such patterns from <span class="InlineEquation" id="IEq440"><span class="MathJax_Preview">\mathcal {C}_i</span><script type="math/tex">\mathcal {C}_i</script></span>. Finally, the remaining patterns in <span class="InlineEquation" id="IEq441"><span class="MathJax_Preview">\mathcal {C}_i</span><script type="math/tex">\mathcal {C}_i</script></span> are globally frequent, and are thus added to the set <span class="InlineEquation" id="IEq442"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span>.</p></section><section id="Sec13" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">5.5 </span>External neighbor expansion</h3><div id="Par65" class="Para">To avoid false negatives, the external neighbor expansion in Alg.&nbsp;<span class="InternalRef"><a href="#Sec13">6</a></span> ensures that all edges needed for computing the embeddings are available locally. For each pattern <em class="EmphasisTypeItalic ">P</em> that is globally frequent, we have to collect the external neighbor requests for any vertex in <span class="InlineEquation" id="IEq443"><span class="MathJax_Preview">\varPhi _i(v)</span><script type="math/tex">\varPhi _i(v)</script></span> for <span class="InlineEquation" id="IEq444"><span class="MathJax_Preview">v \in V_P</span><script type="math/tex">v \in V_P</script></span>. We collect this list for each pattern, denoted <span class="InlineEquation" id="IEq445"><span class="MathJax_Preview">Ext_i[P]</span><script type="math/tex">Ext_i[P]</script></span>, when we compute the local support, as detailed in Alg.&nbsp;<span class="InternalRef"><a href="#Sec10">3</a></span>. Recall that <span class="InlineEquation" id="IEq446"><span class="MathJax_Preview">Ext_i[P]</span><script type="math/tex">Ext_i[P]</script></span> comprises pairs of the form (<em class="EmphasisTypeItalic ">v</em>,&nbsp;<em class="EmphasisTypeItalic ">j</em>), where <span class="InlineEquation" id="IEq447"><span class="MathJax_Preview">v \in V_j</span><script type="math/tex">v \in V_j</script></span>, with <span class="InlineEquation" id="IEq448"><span class="MathJax_Preview">j \ne i</span><script type="math/tex">j \ne i</script></span>. The first step is to do an <span class="EmphasisTypeSmallCaps ">AllToAll</span> to request vertex <em class="EmphasisTypeItalic ">v</em> from partition <span class="InlineEquation" id="IEq449"><span class="MathJax_Preview">G_j</span><script type="math/tex">G_j</script></span>. Next, another <span class="EmphasisTypeSmallCaps ">AllToAll</span> is used to respond back with the degree and neighbor list. Finally, each partition <span class="InlineEquation" id="IEq450"><span class="MathJax_Preview">G_i</span><script type="math/tex">G_i</script></span>
 extends its edge and vertex sets. The external neighbor requests are 
carried out simultaneously by dividing up the requests among nodes in a 
group <span class="InlineEquation" id="IEq451"><span class="MathJax_Preview">R_i</span><script type="math/tex">R_i</script></span>.
 It is worth remarking that the neighborhood extension does not 
significantly expand the memory footprint. This is because we request 
the external neighbors only for the embeddings of frequent patterns, and
 as computation progresses, fewer and fewer of these are found. It is 
possible to further reduce the local memory by actually deleting from <span class="InlineEquation" id="IEq452"><span class="MathJax_Preview">G_j</span><script type="math/tex">G_j</script></span> vertices and edges that do not participate in the mappings for any frequent pattern.<figure class="Figure" id="Figf"><div class="MediaObject" id="MO24"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Figf_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Figf_HTML.gif" alt=""></a> </div></figure> </div></section><section id="Sec14" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">5.6 </span> <span class="EmphasisTypeSmallCaps ">DistGraph</span> correctness</h3><p id="Par66" class="Para">Our
 method guarantees correctness, i.e., it produces no false positive or 
false negative patterns, as proved in the theorem below.</p><div id="FPar9" class="FormalPara FormalParaRenderingStyle1"><h3 class="u-regular"><span class="Heading">Theorem 5</span></h3><p id="Par67" class="Para">Algorithm&nbsp;1 is correct, i.e., it outputs a pattern <em class="EmphasisTypeItalic ">P</em> if and only if <span class="InlineEquation" id="IEq458"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>.</p></div><div id="FPar10" class="FormalPara FormalParaRenderingStyle2"><h3 class="u-regular"><span class="Heading">Proof</span></h3><p id="Par68" class="Para">We prove the correctness by induction. Let <span class="InlineEquation" id="IEq459"><span class="MathJax_Preview">\mathcal {P}'_i</span><script type="math/tex">\mathcal {P}'_i</script></span> be the set of globally frequent patterns from the previous level comprising patterns with <span class="InlineEquation" id="IEq460"><span class="MathJax_Preview">r-1</span><script type="math/tex">r-1</script></span> edges.</p><p id="Par69" class="Para">Let <em class="EmphasisTypeItalic ">P</em> be a pattern with <em class="EmphasisTypeItalic ">r</em> edges, and let it be globally frequent, i.e., <span class="InlineEquation" id="IEq461"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>. Monotonicity of <span class="InlineEquation" id="IEq462"><span class="MathJax_Preview">\sigma (P)</span><script type="math/tex">\sigma (P)</script></span> guarantees that there exists a pattern in <span class="InlineEquation" id="IEq463"><span class="MathJax_Preview">\mathcal {P}_i'</span><script type="math/tex">\mathcal {P}_i'</script></span> from which <em class="EmphasisTypeItalic ">P</em> can be obtained by an edge extension. In the <span class="EmphasisTypeSmallCaps ">Compute-Local-Support</span> step, we try all extensions of all patterns <span class="InlineEquation" id="IEq464"><span class="MathJax_Preview">P' \in \mathcal {P}'_i</span><script type="math/tex">P' \in \mathcal {P}'_i</script></span>, and therefore <em class="EmphasisTypeItalic ">P</em> will be one of the candidate extensions. Given the invariant <span class="InlineEquation" id="IEq465"><span class="MathJax_Preview">\varSigma (P) = \cup \varSigma _i(P)</span><script type="math/tex">\varSigma (P) = \cup \varSigma _i(P)</script></span>, Thm.&nbsp;<span class="InternalRef"><a href="#FPar3">2</a></span> guarantees that <em class="EmphasisTypeItalic ">P</em> will be locally frequent in at least one partition, so that <span class="InlineEquation" id="IEq466"><span class="MathJax_Preview">P \in \mathcal {P}_{lf}</span><script type="math/tex">P \in \mathcal {P}_{lf}</script></span> for some partition. In <span class="EmphasisTypeSmallCaps ">Support-bound-Pruning</span> either <span class="InlineEquation" id="IEq467"><span class="MathJax_Preview">\widehat{\sigma }_l(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_l(P) \ge minsup</script></span> and thus it will be added to the new set of globally frequent patterns <span class="InlineEquation" id="IEq468"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span>, or <span class="InlineEquation" id="IEq469"><span class="MathJax_Preview">\widehat{\sigma }_u(P) \ge minsup</span><script type="math/tex">\widehat{\sigma }_u(P) \ge minsup</script></span> and thus it will be added to the set of potentially frequent patterns <span class="InlineEquation" id="IEq470"><span class="MathJax_Preview">\mathcal {C}_i</span><script type="math/tex">\mathcal {C}_i</script></span>. In the latter case, after exact global support computation in <span class="EmphasisTypeSmallCaps ">Exact-Global-Support</span>, <em class="EmphasisTypeItalic ">P</em> will be found to be globally frequent and thus will be added to <span class="InlineEquation" id="IEq471"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span>, the set of globally frequent patterns at the current level <em class="EmphasisTypeItalic ">r</em>.</p><p id="Par70" class="Para">On the other hand, assume that <em class="EmphasisTypeItalic ">P</em> is not globally frequent, i.e., <span class="InlineEquation" id="IEq472"><span class="MathJax_Preview">\sigma (P) &lt; minsup</span><script type="math/tex">\sigma (P) < minsup</script></span>. In <span class="EmphasisTypeSmallCaps ">Compute-Local-Support</span> if <em class="EmphasisTypeItalic ">P</em> is not locally frequent in at least one partition, we are done. Otherwise, if <em class="EmphasisTypeItalic ">P</em> is locally frequent in some partition, it will be added to the corresponding set <span class="InlineEquation" id="IEq473"><span class="MathJax_Preview">\mathcal {P}_{lf}</span><script type="math/tex">\mathcal {P}_{lf}</script></span>. Next in <span class="EmphasisTypeSmallCaps ">Support-bound-Pruning</span> if <span class="InlineEquation" id="IEq474"><span class="MathJax_Preview">\widehat{\sigma }_u(P) &lt; minsup</span><script type="math/tex">\widehat{\sigma }_u(P) < minsup</script></span>, then <em class="EmphasisTypeItalic ">P</em> will be pruned and we are done. Otherwise, <em class="EmphasisTypeItalic ">P</em> will be added to the set <span class="InlineEquation" id="IEq475"><span class="MathJax_Preview">\mathcal {C}_i</span><script type="math/tex">\mathcal {C}_i</script></span>, but will be eliminated in <span class="EmphasisTypeSmallCaps ">Exact-Global-Support</span> when we determine its exact global support.</p><p id="Par71" class="Para">Thus, a pattern <em class="EmphasisTypeItalic ">P</em> is added to <span class="InlineEquation" id="IEq476"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span> if and only if <span class="InlineEquation" id="IEq477"><span class="MathJax_Preview">\sigma (P) \ge minsup</span><script type="math/tex">\sigma (P) \ge minsup</script></span>. To complete the proof, we have to show that the base case holds. Note that initially, when <span class="InlineEquation" id="IEq478"><span class="MathJax_Preview">r=0</span><script type="math/tex">r=0</script></span>, we start with the empty graph <span class="InlineEquation" id="IEq479"><span class="MathJax_Preview">\emptyset </span><script type="math/tex">\emptyset </script></span> which is trivially frequent. For <span class="InlineEquation" id="IEq480"><span class="MathJax_Preview">r=1</span><script type="math/tex">r=1</script></span>,
 this empty graph will be extended by all possible edges in each 
partition, and thus the set of all frequent patterns with a single edge 
will be correctly found. Finally, the algorithm will terminate when each
 <span class="InlineEquation" id="IEq481"><span class="MathJax_Preview">\mathcal {P}_i</span><script type="math/tex">\mathcal {P}_i</script></span> is empty, which implies there are no new frequent patterns to be found. <span class="InlineEquation" id="IEq482"><span class="MathJax_Preview">\square </span><script type="math/tex">\square </script></span> </p></div><p id="Par72" class="Para">The worse-case time complexity for mining all frequent subgraphs is clearly exponential. To see this, consider an input graph <em class="EmphasisTypeItalic ">G</em> with <span class="InlineEquation" id="IEq483"><span class="MathJax_Preview">|V|=n</span><script type="math/tex">|V|=n</script></span> vertices and <span class="InlineEquation" id="IEq484"><span class="MathJax_Preview">|E|=m</span><script type="math/tex">|E|=m</script></span> edges. Assume that the largest frequent pattern has <em class="EmphasisTypeItalic ">l</em> edges, then the number of possible frequent patterns are: <span class="InlineEquation" id="IEq485"><span class="MathJax_Preview">\sum _{i=0}^l {m \atopwithdelims ()l} = \tfrac{m^{l+1}-1}{m-1} = O(m^l)</span><script type="math/tex">\sum _{i=0}^l {m \atopwithdelims ()l} = \tfrac{m^{l+1}-1}{m-1} = O(m^l)</script></span>. Next, for any pattern <em class="EmphasisTypeItalic ">P</em> with <em class="EmphasisTypeItalic ">l</em> edges, we have to compute its support, which can take <span class="InlineEquation" id="IEq486"><span class="MathJax_Preview">O\left( {m \atopwithdelims ()l}\right) = O(m^l)</span><script type="math/tex">O\left( {m \atopwithdelims ()l}\right) = O(m^l)</script></span> time. Thus, the worst case time complexity for frequent subgraph mining is <span class="InlineEquation" id="IEq487"><span class="MathJax_Preview">O(m^l \times m^l) = O(m^{2l})</span><script type="math/tex">O(m^l \times m^l) = O(m^{2l})</script></span>. Our distributed approach reduces the computational time to <span class="InlineEquation" id="IEq488"><span class="MathJax_Preview">O(m^{2l} / p)</span><script type="math/tex">O(m^{2l} / p)</script></span> in the ideal case, where <em class="EmphasisTypeItalic ">p</em> is the number of compute elements. For further results on the complexity of subgraph mining problems see&nbsp;Yang (<span class="CitationRef"><a href="#CR28" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2004</a></span>) and Kimelfeld and Kolaitis (<span class="CitationRef"><a href="#CR13" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>).</p></section></div></section><section id="Sec15" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">6 </span>Experiments<span class="section-icon"></span></h2><div class="content"><div id="Par73" class="Para">All
 our distributed experiments are performed on up to 2048 IBM Blue Gene/Q
 (BG/Q) nodes. Each node consists of a 16-core 1.6 GHz A2 processor, 
with 16 GB of DDR3 memory. <span class="EmphasisTypeSmallCaps ">DistGraph</span>
 is implemented in C++, compiled using g++ (v. 4.4.7) and –O3 
optimization flag. We use the OpenMP (v. 3.0) library for thread-based 
parallelism within compute nodes, and the portable, open-source and 
freely available MPI implementation MPICH2 (v. 1.5) for distributed 
computation across nodes. We compared our sequential implementation with
 GraMi&nbsp;(Elseidy et&nbsp;al. <span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>), which is written in Java, and is compiled using JDK v. 1.7. Our <span class="EmphasisTypeSmallCaps ">DistGraph</span> code is available for download at <span class="ExternalRef"><a target="_blank" rel="noopener" href="https://github.com/zakimjz/DistGraph"><span class="RefSource">https://github.com/zakimjz/DistGraph</span></a></span>.<div id="Tab2" class="Table"><div class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Table 2</span> <p class="SimplePara">Datasets and their properties: number of vertices |<em class="EmphasisTypeItalic ">V</em>|, edges |<em class="EmphasisTypeItalic ">E</em>|, labels |<em class="EmphasisTypeItalic ">L</em>|, avg. degree <em class="EmphasisTypeItalic ">AD</em>, clustering coefficient <em class="EmphasisTypeItalic ">CC</em> </p> </div></div><div class="u-scroll-horizontal"><table><colgroup><col class="tcol1 align-left"><col class="tcol2 align-left"><col class="tcol3 align-left"><col class="tcol4 align-left"><col class="tcol5 align-left"><col class="tcol6 align-left"></colgroup><thead><tr><th> <p class="SimplePara">Dataset</p> </th><th> <p class="SimplePara">|<em class="EmphasisTypeItalic ">V</em>|</p> </th><th> <p class="SimplePara">|<em class="EmphasisTypeItalic ">E</em>|</p> </th><th> <p class="SimplePara">|<em class="EmphasisTypeItalic ">L</em>|</p> </th><th> <p class="SimplePara"> <em class="EmphasisTypeItalic ">AD</em> </p> </th><th> <p class="SimplePara"> <em class="EmphasisTypeItalic ">CC</em> </p> </th></tr></thead><tbody><tr><td> <p class="SimplePara">PDB1 (small)</p> </td><td> <p class="SimplePara">20,226</p> </td><td> <p class="SimplePara">83,356</p> </td><td> <p class="SimplePara">22</p> </td><td> <p class="SimplePara">8.2</p> </td><td> <p class="SimplePara">0.57</p> </td></tr><tr><td> <p class="SimplePara">PDB2 (medium)</p> </td><td> <p class="SimplePara">2,020,188</p> </td><td> <p class="SimplePara">7,905,260</p> </td><td> <p class="SimplePara">22</p> </td><td> <p class="SimplePara">3.9</p> </td><td> <p class="SimplePara">0.55</p> </td></tr><tr><td> <p class="SimplePara">PDB3 (large)</p> </td><td> <p class="SimplePara">17,400,398</p> </td><td> <p class="SimplePara">68,599,675</p> </td><td> <p class="SimplePara">22</p> </td><td> <p class="SimplePara">7.9</p> </td><td> <p class="SimplePara">0.55</p> </td></tr><tr><td> <p class="SimplePara">PDB4 (16<span class="InlineEquation" id="IEq489"><span class="MathJax_Preview">\times </span><script type="math/tex">\times </script></span>)</p> </td><td> <p class="SimplePara">278,406,368</p> </td><td> <p class="SimplePara">1,097,594,800</p> </td><td> <p class="SimplePara">22</p> </td><td> <p class="SimplePara">7.9</p> </td><td> <p class="SimplePara">0.55</p> </td></tr><tr><td> <p class="SimplePara">PDB5 (32<span class="InlineEquation" id="IEq490"><span class="MathJax_Preview">\times </span><script type="math/tex">\times </script></span>)</p> </td><td> <p class="SimplePara">556,812,736</p> </td><td> <p class="SimplePara">2,195,189,600</p> </td><td> <p class="SimplePara">22</p> </td><td> <p class="SimplePara">7.9</p> </td><td> <p class="SimplePara">0.55</p> </td></tr><tr><td> <p class="SimplePara">PDB6 (64<span class="InlineEquation" id="IEq491"><span class="MathJax_Preview">\times </span><script type="math/tex">\times </script></span>)</p> </td><td> <p class="SimplePara">1,113,625,472</p> </td><td> <p class="SimplePara">4,390,379,200</p> </td><td> <p class="SimplePara">22</p> </td><td> <p class="SimplePara">7.9</p> </td><td> <p class="SimplePara">0.55</p> </td></tr><tr><td> <p class="SimplePara">Patent</p> </td><td> <p class="SimplePara">2,942,159</p> </td><td> <p class="SimplePara">14,275,931</p> </td><td> <p class="SimplePara">37</p> </td><td> <p class="SimplePara">9.7</p> </td><td> <p class="SimplePara">0.07</p> </td></tr><tr><td> <p class="SimplePara">Youtube</p> </td><td> <p class="SimplePara">4,584,572</p> </td><td> <p class="SimplePara">23,236,009</p> </td><td> <p class="SimplePara">80</p> </td><td> <p class="SimplePara">5.8</p> </td><td> <p class="SimplePara">0.19</p> </td></tr></tbody></table></div></div> </div><section id="Sec16" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">6.1 </span>Datasets</h3><p id="Par74" class="Para">We used three main datasets in our experiments, namely <em class="EmphasisTypeItalic ">PDB</em>, <em class="EmphasisTypeItalic ">Patent</em> and <em class="EmphasisTypeItalic ">Youtube</em>. The properties of the datasets are shown in Table&nbsp;<span class="InternalRef"><a href="#Tab2">2</a></span>.
 The PDB datasets (PDB1–6) were created from the 3D distance matrices 
using different number of protein 3D structures from the RCSB Protein 
Data Bank (<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.rcsb.org/pdb"><span class="RefSource">www.rcsb.org/pdb</span></a></span>). The datasets PDB4, PDB5 and PDB6 are actually 16<span class="InlineEquation" id="IEq492"><span class="MathJax_Preview">\times </span><script type="math/tex">\times </script></span>, 32<span class="InlineEquation" id="IEq493"><span class="MathJax_Preview">\times </span><script type="math/tex">\times </script></span> and 64<span class="InlineEquation" id="IEq494"><span class="MathJax_Preview">\times </span><script type="math/tex">\times </script></span>
 replications of PDB3, respectively, used for scalability experiments. 
Our largest graph PDB6 consists of more than one billion vertices and 4 
billion edges. The effective diameter for the PDB graphs (for 90&nbsp;% 
of the node pairs) is between 24 and 32 hops.</p><p id="Par75" class="Para">The Patent dataset (<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://www.nber.org/patents"><span class="RefSource">www.nber.org/patents</span></a></span>)
 consists of pairwise citation data of the US patents granted between 
Jan 1963 and Dec 1999. We considered the “grant year” of the patents as 
the vertex labels. The effective diameter (90&nbsp;%) for Patent is 9.5.
 The Patent dataset has some hub vertices with very high degrees. We use
 the hub-duplication technique&nbsp;Ucar et&nbsp;al. (<span class="CitationRef"><a href="#CR25" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2006</a></span>) to deal with such high degree nodes. If any node <em class="EmphasisTypeItalic ">v</em> has degree more than 30, we make several duplicates of that node, say <span class="InlineEquation" id="IEq495"><span class="MathJax_Preview">v_1, v_2, \ldots </span><script type="math/tex">v_1, v_2, \ldots </script></span>, and redistribute the neighbors of <em class="EmphasisTypeItalic ">v</em> among the duplicates <span class="InlineEquation" id="IEq496"><span class="MathJax_Preview">v_i</span><script type="math/tex">v_i</script></span>, so that no node has degree more than 30. Finally, we connect <em class="EmphasisTypeItalic ">v</em> and all <span class="InlineEquation" id="IEq497"><span class="MathJax_Preview">v_i</span><script type="math/tex">v_i</script></span>
 in a clique. Note that no interactions are lost, and this technique 
allows us to avoid the potential explosion in the number of embeddings 
for hub nodes.</p><p id="Par76" class="Para">The Youtube dataset (<span class="ExternalRef"><a target="_blank" rel="noopener" href="http://netsg.cs.sfu.ca/youtubedata"><span class="RefSource">netsg.cs.sfu.ca/youtubedata</span></a></span>)
 lists the crawled video ids and related videos for each video posted 
from Feb 2007 to Jul 2008. We restricted the maximum vertex degree to 
15. We created 10 buckets for rating (scale 0 to 5) and 8 buckets for 
length of the video. Combining these two we generated 80 different 
vertex labels. For all the datasets the edges are unlabeled. The 
effective diameter (90&nbsp;%) for Youtube it is 7.7.</p><div id="Par77" class="Para">Both
 Patent and Youtube have a giant component that comprises over 
99.6&nbsp;% of the vertices, and thus can easily be split into <em class="EmphasisTypeItalic ">K</em> parts. However, the PDB datasets comprise many components (each being a separate protein structure), and straightforward <em class="EmphasisTypeItalic ">K</em>-way split results in partitions with no external edges. Since we desire external neighbors, we split each protein into <em class="EmphasisTypeItalic ">K</em> parts provided it has at least 64 vertices, which results in many external edges, and the <em class="EmphasisTypeItalic ">K</em> parts of a single protein are then assigned to different partitions. Table&nbsp;<span class="InternalRef"><a href="#Tab3">3</a></span>
 shows the average number of external edges for different number of 
partitions, respv., for the different datasets. For instance, Youtube 
has one giant component comprising 4,583,987 vertices, and it has on 
avg. 71,606 external neighbors with 128 partitions.<div id="Tab3" class="Table"><div class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Table 3</span> <p class="SimplePara">Partitions and external neighbors</p> </div></div><div class="u-scroll-horizontal"><table><colgroup><col class="tcol1 align-left"><col class="tcol2 align-left"><col class="tcol3 align-left"></colgroup><thead><tr><th> <p class="SimplePara">Dataset</p> </th><th> <p class="SimplePara">Max comp.</p> </th><th> <p class="SimplePara">Num partitions (avg. ext neighbors)</p> </th></tr></thead><tbody><tr><td> <p class="SimplePara">PDB1</p> </td><td> <p class="SimplePara">4,477</p> </td><td> <p class="SimplePara">4 (688), 8 (764)</p> </td></tr><tr><td> <p class="SimplePara">PDB2</p> </td><td> <p class="SimplePara">7,856</p> </td><td> <p class="SimplePara">16 (57,246), 32 (38,625), 64 (23,801)</p> </td></tr><tr><td> <p class="SimplePara">PDB3</p> </td><td> <p class="SimplePara">11,568</p> </td><td> <p class="SimplePara">32 (350,144), 64 (213,725), 128 (109,570)</p> </td></tr><tr><td> <p class="SimplePara">Patent</p> </td><td> <p class="SimplePara">2,934,409</p> </td><td> <p class="SimplePara">32 (128,431), 64 (75,694), 128 (43,597)</p> </td></tr><tr><td> <p class="SimplePara">Youtube</p> </td><td> <p class="SimplePara">4,583,987</p> </td><td> <p class="SimplePara">128 (71,606), 256 (38,628)</p> </td></tr></tbody></table></div></div> </div><div id="Par78" class="Para"> <figure class="Figure" id="Fig4"><div class="MediaObject" id="MO25"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig4_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig4_HTML.gif" alt="Fig. 4"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 4</span> <p class="SimplePara">PDB1: sequential comparison of <span class="EmphasisTypeSmallCaps ">DistGraph</span> with GraMi</p> </div></figcaption></figure> </div></section><section id="Sec17" tabindex="-1" class="Section2 RenderAsSection2"><h3 class="Heading"><span class="HeadingNumber">6.2 </span>Performance results</h3><p id="Par79" class="Para">We
 study the performance of our distributed algorithm by varying one of 
the three different parameters, namely the number of compute nodes (<em class="EmphasisTypeItalic ">p</em>), threads (<em class="EmphasisTypeItalic ">k</em>), and partitions (<em class="EmphasisTypeItalic ">K</em>), while keeping the other two fixed. All our performance results are shown as barplots. In the plots the left <em class="EmphasisTypeItalic ">Y</em> axis shows the total time, and the bars show the time for different methods with different minimum support values. <em class="EmphasisTypeItalic ">A value</em> <em class="EmphasisTypeItalic ">s</em> <em class="EmphasisTypeItalic ">on the</em> <em class="EmphasisTypeItalic ">X</em> <em class="EmphasisTypeItalic ">axis indicates</em> <span class="InlineEquation" id="IEq498"><span class="MathJax_Preview">minsup</span><script type="math/tex">minsup</script></span> <em class="EmphasisTypeItalic ">of</em> <span class="InlineEquation" id="IEq499"><span class="MathJax_Preview">s\%</span><script type="math/tex">s\%</script></span> <em class="EmphasisTypeItalic ">with respect to</em> |<em class="EmphasisTypeItalic ">V</em>|, i.e., we are using relative support for each pattern. So for a pattern to be frequent it must satisfy the condition <span class="InlineEquation" id="IEq500"><span class="MathJax_Preview">\sigma (P) \ge |V| \times s/100</span><script type="math/tex">\sigma (P) \ge |V| \times s/100</script></span>. In most cases the right <em class="EmphasisTypeItalic ">Y</em> axis records the speedup with respect to a baseline method in the same plot.</p><div id="Par80" class="Para"> <em class="EmphasisTypeItalic ">Sequential comparison:</em> We compared a sequential run of <span class="EmphasisTypeSmallCaps ">DistGraph</span> with the state-of-art sequential single graph mining method GraMi&nbsp;(Elseidy et&nbsp;al. <span class="CitationRef"><a href="#CR5" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2014</a></span>)
 on the small PDB1 dataset. Since the Java code cannot be run on the 
Blue Gene/Q system, we test on a 16-core, 2.9&nbsp;Ghz AMD Opteron 6272 
processor with 256GB memory. Both methods were run using a single core. 
The results are shown in Fig.&nbsp;<span class="InternalRef"><a href="#Fig4">4</a></span>. We can see that sequential <span class="EmphasisTypeSmallCaps ">DistGraph</span> is an order of magnitude faster (the <em class="EmphasisTypeItalic ">y</em>-axis is in log-scale). For <span class="InlineEquation" id="IEq501"><span class="MathJax_Preview">minsup= 0.59\,\%</span><script type="math/tex">minsup= 0.59\,\%</script></span>, GraMi took 22787s versus 311s for <span class="EmphasisTypeSmallCaps ">DistGraph</span>
 (i.e., 73 times faster). Although the comparison is not entirely fair, 
since GraMi is in Java and our code is in C++, we report these timings 
to show that we have a very efficient implementation. That is, when 
reporting the distributed performance below, we are not using a slow 
sequential method that can artificially boost speedup numbers.<figure class="Figure" id="Fig5"><div class="MediaObject" id="MO26"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig5_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig5_HTML.gif" alt="Fig. 5"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 5</span> <p class="SimplePara"> <span class="EmphasisTypeSmallCaps ">DistGraph</span> on PDB1: sequential (<span class="InlineEquation" id="IEq502"><span class="MathJax_Preview">p=1, k=1</span><script type="math/tex">p=1, k=1</script></span>) and distributed (<span class="InlineEquation" id="IEq503"><span class="MathJax_Preview">p=4, 8, 16</span><script type="math/tex">p=4, 8, 16</script></span>)</p> </div></figcaption></figure> <figure class="Figure" id="Fig6"><div class="MediaObject" id="MO27"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig6_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig6_HTML.gif" alt="Fig. 6"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 6</span> <p class="SimplePara">PDB3 and PDB4: varying compute nodes; effect of external neighbors</p> </div></figcaption></figure> </div><p id="Par81" class="Para"> <em class="EmphasisTypeItalic ">Varying compute nodes:</em> We show the effect of varying the number of compute nodes. Figure&nbsp;<span class="InternalRef"><a href="#Fig5">5</a></span> compares the sequential run of <span class="EmphasisTypeSmallCaps ">DistGraph</span> (with <span class="InlineEquation" id="IEq504"><span class="MathJax_Preview">p=1,k=1</span><script type="math/tex">p=1,k=1</script></span>), against using 4, 8, and 16 compute nodes on BG/Q for PDB1. We obtain speedups of 10 with <span class="InlineEquation" id="IEq505"><span class="MathJax_Preview">p=4</span><script type="math/tex">p=4</script></span>, speedup of 16 with <span class="InlineEquation" id="IEq506"><span class="MathJax_Preview">p=8</span><script type="math/tex">p=8</script></span>, and a speedup of 25 with <span class="InlineEquation" id="IEq507"><span class="MathJax_Preview">p=16</span><script type="math/tex">p=16</script></span>
 compute nodes. While the speedups are good, they are less than ideal, 
since PDB1 is a relatively small graph with limited work load.</p><p id="Par82" class="Para">Figure&nbsp;<span class="InternalRef"><a href="#Fig6">6</a></span>
 shows the effect of varying the compute nodes, and the effect of 
external neighbors on PDB3 and PDB4. For these datasets, we were unable 
to run the sequential version of <span class="EmphasisTypeSmallCaps ">DistGraph</span> (with <span class="InlineEquation" id="IEq508"><span class="MathJax_Preview">p=1</span><script type="math/tex">p=1</script></span>)
 for the lower support value, and therefore the speedups are reported 
with respect to the baseline number of compute nodes shown on the right <em class="EmphasisTypeItalic ">y</em>-axis.</p><div id="Par83" class="Para">We
 can see that the speedup is generally good, indicating relatively good 
load balancing across compute nodes. For instance, for PDB3 we get a 
speedup close to 3 for the lowest support, when comparing <span class="InlineEquation" id="IEq509"><span class="MathJax_Preview">p=128</span><script type="math/tex">p=128</script></span> with <span class="InlineEquation" id="IEq510"><span class="MathJax_Preview">p=32</span><script type="math/tex">p=32</script></span>,
 with the ideal speedup being 4. Likewise, for PDB4 (which contains more
 than 1 billion edges), we get a speedup of 1.6 (ideal being 2), when we
 go from <span class="InlineEquation" id="IEq511"><span class="MathJax_Preview">p=512</span><script type="math/tex">p=512</script></span> to <span class="InlineEquation" id="IEq512"><span class="MathJax_Preview">p=1024</span><script type="math/tex">p=1024</script></span>
 compute nodes. Finally, whereas we get similar speedups without and 
with external neighbors, the run times with external neighbor requests 
are about twice as high, at the lowest <span class="InlineEquation" id="IEq513"><span class="MathJax_Preview">minsup</span><script type="math/tex">minsup</script></span>
 value. This is mainly due to the cost of determining the exact global 
support, which is not required when there are no external neighbors.<figure class="Figure" id="Fig7"><div class="MediaObject" id="MO28"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig7_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig7_HTML.gif" alt="Fig. 7"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 7</span> <p class="SimplePara">Patent and Youtube: varying number of compute nodes</p> </div></figcaption></figure> </div><div id="Par84" class="Para">Figure&nbsp;<span class="InternalRef"><a href="#Fig7">7</a></span>
 shows results of varying compute nodes on the Patent and Youtube 
datasets. Even though these graphs are relatively large, they do not 
have many frequent patterns. Therefore, the speedups are relatively low 
for Youtube, whereas Patent has more patterns and has better speedups 
(close to 2 with <span class="InlineEquation" id="IEq514"><span class="MathJax_Preview">p=128</span><script type="math/tex">p=128</script></span> wrt <span class="InlineEquation" id="IEq515"><span class="MathJax_Preview">p=32</span><script type="math/tex">p=32</script></span>).
 On the other hand, the PDB datasets have many more “structural motifs” 
or frequent patterns and thus there is enough work to achieve good 
distributed performance.<figure class="Figure" id="Fig8"><div class="MediaObject" id="MO29"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig8_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig8_HTML.gif" alt="Fig. 8"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 8</span> <p class="SimplePara">Effect of varying partitions</p> </div></figcaption></figure> </div><p id="Par85" class="Para"> <em class="EmphasisTypeItalic ">Varying partitions:</em> In Fig.&nbsp;<span class="InternalRef"><a href="#Fig8">8</a></span> we show the effect of varying the number of partitions <em class="EmphasisTypeItalic ">K</em>. All results on PDB are <em class="EmphasisTypeItalic ">with</em> external neighbors. Consider the PDB2 dataset. With <span class="InlineEquation" id="IEq516"><span class="MathJax_Preview">K=16</span><script type="math/tex">K=16</script></span> partitions of the input graph, since we have <span class="InlineEquation" id="IEq517"><span class="MathJax_Preview">p=64</span><script type="math/tex">p=64</script></span>
 compute nodes, this means that there are 4 compute nodes in each group,
 which can work in a distributed manner to process the same partition. 
On the other hand, with <span class="InlineEquation" id="IEq518"><span class="MathJax_Preview">K=64</span><script type="math/tex">K=64</script></span>
 partitions of the input graph, each partition is processed by only one 
compute node. Although there are fewer external neighbors with <span class="InlineEquation" id="IEq519"><span class="MathJax_Preview">K=64</span><script type="math/tex">K=64</script></span> partitions compared to <span class="InlineEquation" id="IEq520"><span class="MathJax_Preview">K=16</span><script type="math/tex">K=16</script></span>,
 the latter has the advantage of distributing the work load for a 
partition over 4 nodes. Across the datasets, the results show that for a
 fixed number of compute nodes, generally speaking, having fewer input 
partitions (with multiple compute nodes per group) results in better 
performance. However, Youtube is an exception. With fewer frequent 
patterns, having more compute nodes in a group does not help.</p><div id="Par86" class="Para"> <em class="EmphasisTypeItalic ">Varying threads:</em> Figure&nbsp;<span class="InternalRef"><a href="#Fig9">9</a></span> shows the effect of varying the number of threads <em class="EmphasisTypeItalic ">k</em> per compute node. The PDB datasets are <em class="EmphasisTypeItalic ">with</em> external neighbors. For both PDB2 and PDB3 we get speedup around 10 wrt <span class="InlineEquation" id="IEq521"><span class="MathJax_Preview">k=1</span><script type="math/tex">k=1</script></span> at the lowest support considered when using <span class="InlineEquation" id="IEq522"><span class="MathJax_Preview">k=16</span><script type="math/tex">k=16</script></span> threads (ideal speedup is 16).<figure class="Figure" id="Fig9"><div class="MediaObject" id="MO30"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig9_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig9_HTML.gif" alt="Fig. 9"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 9</span> <p class="SimplePara">PDB: varying threads</p> </div></figcaption></figure> <figure class="Figure" id="Fig10"><div class="MediaObject" id="MO31"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig10_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig10_HTML.gif" alt="Fig. 10"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 10</span> <p class="SimplePara">PDB3: effect of two-step support bounding</p> </div></figcaption></figure> </div><p id="Par87" class="Para"> <em class="EmphasisTypeItalic ">Effect of support bounding optimization:</em> In Fig.&nbsp;<span class="InternalRef"><a href="#Fig10">10</a></span>
 we compare the effect of using the lower and upper support estimates to
 prune patterns. We also analyze the effect of regenerating the 
embeddings versus using the file-based approach to compute the global 
support. Here, “No opt” refers to using no optimization, i.e., 
considering all locally frequent patterns as the candidates for exact 
support computation. We can immediately see that the pruning based on 
support bounds is extremely effective in cutting down the communication 
cost, and thus the run times. The upper-bound effectively prunes 
patterns that cannot possibly be globally frequent. The lower-bound 
further determines the patterns that are truly globally frequent and 
prevents costly exact support computation.</p><p id="Par88" class="Para">For
 the remaining patterns, we have to determine the global support by 
either keeping the embeddings in memory, in a file on disk, or via 
regeneration. We show the time for the regeneration and file-based case 
in Fig.&nbsp;<span class="InternalRef"><a href="#Fig10">10</a></span>a, 
b, respv., on PDB3. As expected, we were not able to run the 
memory-based version for this graph. We can see that the file-based time
 is about 2 times faster than regeneration. However, with both lower and
 upper support bounding optimizations turned on, the times are more or 
less comparable. For instance, for the lowest support, the regeneration 
time is 4721.3s, whereas the file-based time is 3831.8s.</p><div id="Par89" class="Para">We also profiled the time spent in the various phases of <span class="EmphasisTypeSmallCaps ">DistGraph</span>, as shown in Fig.&nbsp;<span class="InternalRef"><a href="#Fig11">11</a></span>a for PDB3 for both the regeneration and file-based approaches using <span class="InlineEquation" id="IEq523"><span class="MathJax_Preview">minsup=0.75\%</span><script type="math/tex">minsup=0.75\%</script></span>. Both approaches show that the time for support-bound based pruning (<span class="InlineEquation" id="IEq524"><span class="MathJax_Preview">{&lt;}0.5\,\%</span><script type="math/tex">{<}0.5\,\%</script></span> of total time), and expanding the partitions via external neighbor requests (<span class="InlineEquation" id="IEq525"><span class="MathJax_Preview">{&lt;}2\,\%</span><script type="math/tex">{<}2\,\%</script></span> of total time) is negligible compared to the time it takes for computing the local support (<span class="InlineEquation" id="IEq526"><span class="MathJax_Preview">{&gt;}58\,\%</span><script type="math/tex">{>}58\,\%</script></span> of total time) and the time to determine the exact global support (41 and <span class="InlineEquation" id="IEq527"><span class="MathJax_Preview">28\,\%</span><script type="math/tex">28\,\%</script></span>,
 respectively, for regeneration and file-based approaches). Obviously in
 regeneration, we have to recompute the set of embeddings to determine 
actual support, which can be avoided by storing the embeddings on disk. 
Nevertheless, the regeneration approach is more general, since it can be
 applied to massive networks and in situations where the I/O cost for 
storing and retrieving potentially millions of embeddings can be a 
bottleneck.<figure class="Figure" id="Fig11"><div class="MediaObject" id="MO32"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig11_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig11_HTML.gif" alt="Fig. 11"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 11</span> <p class="SimplePara">Timing profile</p> </div></figcaption></figure> <figure class="Figure" id="Fig12"><div class="MediaObject" id="MO33"> <a href="https://media.springernature.com/original/springer-static/image/art%3A10.1007%2Fs10618-016-0466-x/MediaObjects/10618_2016_466_Fig12_HTML.gif" target="_blank" rel="noopener"><span class="u-screenreader-only">Open image in new window</span><img src="10618_2016_466_Fig12_HTML.gif" alt="Fig. 12"></a> </div><figcaption class="Caption" lang="en"><div class="CaptionContent"><span class="CaptionNumber">Fig. 12</span> <p class="SimplePara"> <span class="EmphasisTypeSmallCaps ">DistGraph</span>: scalability on billion vertex graph</p> </div></figcaption></figure> </div><p id="Par90" class="Para"> <em class="EmphasisTypeItalic ">Scalability:</em> Finally, in Fig.&nbsp;<span class="InternalRef"><a href="#Fig12">12</a></span> we show the scalability of our distributed mining algorithm. We show the performance on PDB3, PDB4, PDB5, and PDB6 (all <em class="EmphasisTypeItalic ">with</em>
 external neighbors). PDB6 has over a billion vertices and four billion 
edges, and is the largest graph to be used for frequent subgraph mining 
experiments. We use the following number of compute nodes/partitions (<span class="InlineEquation" id="IEq528"><span class="MathJax_Preview">p=K</span><script type="math/tex">p=K</script></span>): PDB3 (32), PDB4 (512), PDB5 (1024), PDB6 (2048). That is, to mine PDB6, we use <span class="InlineEquation" id="IEq529"><span class="MathJax_Preview">p=2048</span><script type="math/tex">p=2048</script></span> compute nodes and <span class="InlineEquation" id="IEq530"><span class="MathJax_Preview">K=2048</span><script type="math/tex">K=2048</script></span> partitions. We can see that <span class="EmphasisTypeSmallCaps ">DistGraph</span> can scale to billion vertex graphs. For instance, for <span class="InlineEquation" id="IEq531"><span class="MathJax_Preview">minsup=0.69</span><script type="math/tex">minsup=0.69</script></span>,
 the time on PDB6 (64x) is 7313s compared to 5491s for PDB3 (1x), which 
is only a factor of 1.33 higher (ideally the factor should be 1, since 
we are using 2048 = 64 x 32 compute nodes), which shows very good 
distributed performance, considering that PDB6 has many more partitions 
and thus many more external neighbors (64x in absolute terms) compared 
to PDB3.</p></section></div></section><section id="Sec18" tabindex="-1" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1"><span class="HeadingNumber">7 </span>Conclusions<span class="section-icon"></span></h2><div class="content"><p id="Par91" class="Para">We
 presented a novel distributed approach for mining frequent subgraph 
patterns from a single large graph. We partition the input graph to 
distribute the workload across the system. We minimize the amount of 
communication by sending compact mapping counts instead of the pattern 
embeddings/mappings. Our solution is also very flexible, since it relies
 on collective communication primitives that are available on almost all
 distributed frameworks. Our hybrid approach also leverages both shared 
memory and distributed systems. Further, based on the resource 
availability, one can consider different number of partitions, compute 
nodes and threads, for scalability.</p><p id="Par92" class="Para">Our 
work is scalable, however, there are also some limitations that can be 
improved. One limitation of our current implementation of external 
neighbors expansion is that we do not prune any of the irrelevant 
vertices. While the increase is usually gradual, in pathological cases, 
the partitions may eventually become too large, especially if there are 
very large patterns. In such cases, we can <em class="EmphasisTypeItalic ">prune</em>
 from each partition any vertex that does not belong to the mappings for
 any frequent pattern. We can also use a fixed size memory-based cache 
for the local graph partition, swapping out least recently used 
vertices/edges to disk. Another approach is to move the computation 
instead of the data, i.e., migrate patterns to the partitions that own 
the external vertex, as opposed to requesting the neighbors. This is 
likely incur more communication overhead, but we plan to explore these 
alternatives in the future. Another limitation is dealing with hub nodes
 with very high degrees, which can lead to exponentially many 
embeddings. We outlined one approach where we duplicate the hubs nodes, 
creating a clique, and redistributing the neighbors among them, but more
 effective solutions are desired, especially since this is a problem for
 any graph mining algorithm. Finally, the techniques proposed in this 
paper can be used within the Arabesque&nbsp;(Teixeira et&nbsp;al. <span class="CitationRef"><a href="#CR24" title="View reference" role="button" aria-pressed="false" aria-haspopup="true" aria-controls="popup-references">2015</a></span>)
 distributed graph mining framework for FSM over massive partitioned 
graphs. We plan to explore this in the future, and also plan to compare 
with a Spark implementation of <span class="EmphasisTypeSmallCaps ">DistGraph</span>.</p></div></section></div><section id="Notes" class="Section1 RenderAsSection1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1">Notes<span class="section-icon"></span></h2><div class="content"><section><h3 class="Heading">Acknowledgments</h3><p class="SimplePara">This
 work was supported by NSF Award IIS-1302231. We thank Chris Carothers 
and Bulent Yener for several discussions on the practical and 
theoretical aspects of our distributed algorithm.</p></section></div></section><section class="Section1 RenderAsSection1" id="Bib1" tabindex="-1"><h2 class="Heading" data-role="collapsible-handle" tabindex="-1">References<span class="section-icon"></span></h2><div class="content"><ol class="BibliographyWrapper"><li class="Citation"><div class="CitationContent" id="CR1">Afrati
 FN, Fotakis D, Ullman JD (2013) Enumerating subgraph instances using 
map-reduce. In: IEEE international conference on data engineering<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Afrati%20FN%2C%20Fotakis%20D%2C%20Ullman%20JD%20%282013%29%20Enumerating%20subgraph%20instances%20using%20map-reduce.%20In%3A%20IEEE%20international%20conference%20on%20data%20engineering"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR2">Bhuiyan
 M, Al Hasan M (2015) An iterative mapreduce based frequent subgraph 
mining algorithm. IEEE Trans Knowl Data Eng 27(3):608–620<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/TKDE.2014.2345408"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=An%20iterative%20mapreduce%20based%20frequent%20subgraph%20mining%20algorithm&amp;author=M.%20Bhuiyan&amp;author=Hasan%20M.%20Al&amp;journal=IEEE%20Trans%20Knowl%20Data%20Eng&amp;volume=27&amp;issue=3&amp;pages=608-620&amp;publication_year=2015"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR3">Bringmann
 B, Nijssen S (2008) What is frequent in a single graph? In: 
Pacific-Asia conference on advances in knowledge discovery and data 
mining<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Bringmann%20B%2C%20Nijssen%20S%20%282008%29%20What%20is%20frequent%20in%20a%20single%20graph%3F%20In%3A%20Pacific-Asia%20conference%20on%20advances%20in%20knowledge%20discovery%20and%20data%20mining"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR4">Buehrer
 G, Parthasarathy S, Chen Y-K (2006) Adaptive parallel graph mining for 
cmp architectures. In: IEEE international conference on data mining<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Buehrer%20G%2C%20Parthasarathy%20S%2C%20Chen%20Y-K%20%282006%29%20Adaptive%20parallel%20graph%20mining%20for%20cmp%20architectures.%20In%3A%20IEEE%20international%20conference%20on%20data%20mining"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR5">Elseidy
 M, Abdelhamid E, Skiadopoulos S, Kalnis P (2014) Grami: frequent 
subgraph and pattern mining in a single large graph. Proc VLDB Endow 
7:517–528<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.14778/2732286.2732289"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Grami%3A%20frequent%20subgraph%20and%20pattern%20mining%20in%20a%20single%20large%20graph&amp;author=M.%20Elseidy&amp;author=E.%20Abdelhamid&amp;author=S.%20Skiadopoulos&amp;author=P.%20Kalnis&amp;journal=Proc%20VLDB%20Endow&amp;volume=7&amp;pages=517-528&amp;publication_year=2014"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR6">Fatta
 GD, Berthold MR (2006) Dynamic load balancing for the distributed 
mining of molecular structures. IEEE Trans Parallel Distrib Syst 
17(8):773–785<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/TPDS.2006.101"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Dynamic%20load%20balancing%20for%20the%20distributed%20mining%20of%20molecular%20structures&amp;author=GD.%20Fatta&amp;author=MR.%20Berthold&amp;journal=IEEE%20Trans%20Parallel%20Distrib%20Syst&amp;volume=17&amp;issue=8&amp;pages=773-785&amp;publication_year=2006"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR7">Hill
 S, Srichandan B, Sunderraman R (2012) An iterative mapreduce approach 
to frequent subgraph mining in biological datasets. In: ACM conference 
on bioinformatics, computational biology and biomedicine<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Hill%20S%2C%20Srichandan%20B%2C%20Sunderraman%20R%20%282012%29%20An%20iterative%20mapreduce%20approach%20to%20frequent%20subgraph%20mining%20in%20biological%20datasets.%20In%3A%20ACM%20conference%20on%20bioinformatics%2C%20computational%20biology%20and%20biomedicine"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR8">Holder LB, Cook DJ (1993) Discovery of inexact concepts from structural data. IEEE Trans Knowl Data Eng 5(6):992–994<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1109/69.250085"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Discovery%20of%20inexact%20concepts%20from%20structural%20data&amp;author=LB.%20Holder&amp;author=DJ.%20Cook&amp;journal=IEEE%20Trans%20Knowl%20Data%20Eng&amp;volume=5&amp;issue=6&amp;pages=992-994&amp;publication_year=1993"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR9">Huan
 J, Wang W, Prins J(2003) Efficient mining of frequent subgraphs in the 
presence of isomorphism. In: IEEE international conference on data 
mining<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Huan%20J%2C%20Wang%20W%2C%20Prins%20J%282003%29%20Efficient%20mining%20of%20frequent%20subgraphs%20in%20the%20presence%20of%20isomorphism.%20In%3A%20IEEE%20international%20conference%20on%20data%20mining"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR10">Inokuchi
 A, Washio T, Motoda H (2000) An apriori-based algorithm for mining 
frequent substructures from graph data. In: Principles of data mining 
and knowledge discovery. LNCS vol. 1910. Springer, pp 13–23<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Inokuchi%20A%2C%20Washio%20T%2C%20Motoda%20H%20%282000%29%20An%20apriori-based%20algorithm%20for%20mining%20frequent%20substructures%20from%20graph%20data.%20In%3A%20Principles%20of%20data%20mining%20and%20knowledge%20discovery.%20LNCS%20vol.%201910.%20Springer%2C%20pp%2013%E2%80%9323"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR11">Karypis
 G, Kumar V (1998) A fast and high quality multilevel scheme for 
partitioning irregular graphs. SIAM J Sci Comput 20(1):359–392<span class="Occurrences"><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=1639073"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1137/S1064827595287997"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceZLBID"><a class="gtm-reference" data-reference-type="MATH" target="_blank" rel="noopener" href="http://www.emis.de/MATH-item?0915.68129"><span><span>MATH</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=A%20fast%20and%20high%20quality%20multilevel%20scheme%20for%20partitioning%20irregular%20graphs&amp;author=G.%20Karypis&amp;author=V.%20Kumar&amp;journal=SIAM%20J%20Sci%20Comput&amp;volume=20&amp;issue=1&amp;pages=359-392&amp;publication_year=1998"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR12">Kessl
 R, Talukder N, Anchuri P, Zaki MJ (2014) Parallel graph mining with 
GPUs. Proceedings of the BigMine workshop (ACM SIGKDD), Journal of 
Machine Learning Research: conference and workshop proceedings, pp 
36:1–16<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Kessl%20R%2C%20Talukder%20N%2C%20Anchuri%20P%2C%20Zaki%20MJ%20%282014%29%20Parallel%20graph%20mining%20with%20GPUs.%20Proceedings%20of%20the%20BigMine%20workshop%20%28ACM%20SIGKDD%29%2C%20Journal%20of%20Machine%20Learning%20Research%3A%20conference%20and%20workshop%20proceedings%2C%20pp%2036%3A1%E2%80%9316"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR13">Kimelfeld B, Kolaitis PG (2014) The complexity of mining maximal frequent subgraphs. ACM Trans Database Syst (TODS) 39(4):32<span class="Occurrences"><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=3295729"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1145/2629550"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=The%20complexity%20of%20mining%20maximal%20frequent%20subgraphs&amp;author=B.%20Kimelfeld&amp;author=PG.%20Kolaitis&amp;journal=ACM%20Trans%20Database%20Syst%20%28TODS%29&amp;volume=39&amp;issue=4&amp;pages=32&amp;publication_year=2014"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR14">Kuramochi M, Karypis G (2001) Frequent subgraph discovery. In: IEEE international conference on data mining<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Kuramochi%20M%2C%20Karypis%20G%20%282001%29%20Frequent%20subgraph%20discovery.%20In%3A%20IEEE%20international%20conference%20on%20data%20mining"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR15">Kuramochi M, Karypis G (2005) Finding frequent patterns in a large sparse graph. Data Min Knowl Discov 11(3):243–271<span class="Occurrences"><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=2221928"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1007/s10618-005-0003-9"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Finding%20frequent%20patterns%20in%20a%20large%20sparse%20graph&amp;author=M.%20Kuramochi&amp;author=G.%20Karypis&amp;journal=Data%20Min%20Knowl%20Discov&amp;volume=11&amp;issue=3&amp;pages=243-271&amp;publication_year=2005"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR16">Lin
 W, Xiao X, Ghinita G (2014) Large-scale frequent subgraph mining in 
mapreduce. In: IEEE international conference on data engineering<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Lin%20W%2C%20Xiao%20X%2C%20Ghinita%20G%20%282014%29%20Large-scale%20frequent%20subgraph%20mining%20in%20mapreduce.%20In%3A%20IEEE%20international%20conference%20on%20data%20engineering"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR17">Liu
 Y, Jiang X, Chen H, Ma J, Zhang X (2009) Mapreduce-based pattern 
finding algorithm applied in motif detection for prescription 
compatibility network. In: Advanced parallel processing technologies, 
LNCS vol. 5737. Springer, pp 341–355<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Liu%20Y%2C%20Jiang%20X%2C%20Chen%20H%2C%20Ma%20J%2C%20Zhang%20X%20%282009%29%20Mapreduce-based%20pattern%20finding%20algorithm%20applied%20in%20motif%20detection%20for%20prescription%20compatibility%20network.%20In%3A%20Advanced%20parallel%20processing%20technologies%2C%20LNCS%20vol.%205737.%20Springer%2C%20pp%20341%E2%80%93355"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR18">Lu
 W, Chen G, Tung A, Zhao F(2013) Efficiently extracting frequent 
subgraphs using mapreduce. In: IEEE international conference on big data<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Lu%20W%2C%20Chen%20G%2C%20Tung%20A%2C%20Zhao%20F%282013%29%20Efficiently%20extracting%20frequent%20subgraphs%20using%20mapreduce.%20In%3A%20IEEE%20international%20conference%20on%20big%20data"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR19">Meinl
 T, Wörlein M, Fischer I, Philippsen M (2006) Mining molecular datasets 
on symmetric multiprocessor systems. In: IEEE international conference 
on systems, man and cybernetics, vol 2<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Meinl%20T%2C%20W%C3%B6rlein%20M%2C%20Fischer%20I%2C%20Philippsen%20M%20%282006%29%20Mining%20molecular%20datasets%20on%20symmetric%20multiprocessor%20systems.%20In%3A%20IEEE%20international%20conference%20on%20systems%2C%20man%20and%20cybernetics%2C%20vol%202"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR20">Reinhardt
 S, Karypis G (2007) A multi-level parallel implementation of a program 
for finding frequent patterns in a large sparse graph. In: IEEE 
international parallel and distributed processing symposium<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Reinhardt%20S%2C%20Karypis%20G%20%282007%29%20A%20multi-level%20parallel%20implementation%20of%20a%20program%20for%20finding%20frequent%20patterns%20in%20a%20large%20sparse%20graph.%20In%3A%20IEEE%20international%20parallel%20and%20distributed%20processing%20symposium"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR21">Shahrivari
 S, Jalili S (2015) Distributed discovery of frequent subgraphs of a 
network using MapReduce. Computing 97(11):1101–1120<span class="Occurrences"><span class="Occurrence OccurrenceAMSID"><a class="gtm-reference" data-reference-type="MathSciNet" target="_blank" rel="noopener" href="http://www.ams.org/mathscinet-getitem?mr=3397476"><span><span>MathSciNet</span></span></a></span><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1007/s00607-015-0446-9"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceZLBID"><a class="gtm-reference" data-reference-type="MATH" target="_blank" rel="noopener" href="http://www.emis.de/MATH-item?06506032"><span><span>MATH</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Distributed%20discovery%20of%20frequent%20subgraphs%20of%20a%20network%20using%20MapReduce&amp;author=S.%20Shahrivari&amp;author=S.%20Jalili&amp;journal=Computing&amp;volume=97&amp;issue=11&amp;pages=1101-1120&amp;publication_year=2015"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR22">Shao
 Y, Cui B, Chen L, Ma L, Yao J, Xu N (2014) Parallel subgraph listing in
 a large-scale graph. In: ACM SIGMOD international conference on 
management of data<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Shao%20Y%2C%20Cui%20B%2C%20Chen%20L%2C%20Ma%20L%2C%20Yao%20J%2C%20Xu%20N%20%282014%29%20Parallel%20subgraph%20listing%20in%20a%20large-scale%20graph.%20In%3A%20ACM%20SIGMOD%20international%20conference%20on%20management%20of%20data"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR23">Sun Z, Wang H, Wang H, Shao B, Li J (2012) Efficient subgraph matching on billion node graphs. Proc VLDB Endow 5(9):788–799<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.14778/2311906.2311907"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Efficient%20subgraph%20matching%20on%20billion%20node%20graphs&amp;author=Z.%20Sun&amp;author=H.%20Wang&amp;author=H.%20Wang&amp;author=B.%20Shao&amp;author=J.%20Li&amp;journal=Proc%20VLDB%20Endow&amp;volume=5&amp;issue=9&amp;pages=788-799&amp;publication_year=2012"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR24">Teixeira
 CHC, Fonseca AJ, Serafini M, Siganos G, Zaki MJ, Aboulnaga A (2015) 
Arabesque: a system for distributed graph pattern mining. In: 25th ACM 
symposium on operating systems principles<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Teixeira%20CHC%2C%20Fonseca%20AJ%2C%20Serafini%20M%2C%20Siganos%20G%2C%20Zaki%20MJ%2C%20Aboulnaga%20A%20%282015%29%20Arabesque%3A%20a%20system%20for%20distributed%20graph%20pattern%20mining.%20In%3A%2025th%20ACM%20symposium%20on%20operating%20systems%20principles"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR25">Ucar
 D, Asur S, Catalyurek U, Parthasarathy S (2006) Improving functional 
modularity in protein–protein interactions graphs using hub-induced 
subgraphs. In: Fürnkranz J, Scheffer T, Spiliopoulou M (eds) Knowledge 
discovery in databases: PKDD 2006. Springer, Berlin, pp 371–382<span class="Occurrences"><span class="Occurrence OccurrenceDOI"><a class="gtm-reference" data-reference-type="CrossRef" target="_blank" rel="noopener" href="https://doi.org/10.1007/11871637_36"><span><span>CrossRef</span></span></a></span><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="http://scholar.google.com/scholar_lookup?title=Improving%20functional%20modularity%20in%20protein%E2%80%93protein%20interactions%20graphs%20using%20hub-induced%20subgraphs&amp;author=D.%20Ucar&amp;author=S.%20Asur&amp;author=U.%20Catalyurek&amp;author=S.%20Parthasarathy&amp;pages=371-382&amp;publication_year=2006"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR26">Wu
 B, Bai Y (2010) An efficient distributed subgraph mining algorithm in 
extreme large graphs. In: International conference on artificial 
intelligence and computational intelligence: part I<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Wu%20B%2C%20Bai%20Y%20%282010%29%20An%20efficient%20distributed%20subgraph%20mining%20algorithm%20in%20extreme%20large%20graphs.%20In%3A%20International%20conference%20on%20artificial%20intelligence%20and%20computational%20intelligence%3A%20part%20I"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR27">Yan X, Han J (2002) gspan: Graph-based substructure pattern mining. In: IEEE international conference on data mining<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Yan%20X%2C%20Han%20J%20%282002%29%20gspan%3A%20Graph-based%20substructure%20pattern%20mining.%20In%3A%20IEEE%20international%20conference%20on%20data%20mining"><span><span>Google Scholar</span></span></a></span></span></div></li><li class="Citation"><div class="CitationContent" id="CR28">Yang
 G (2004) The complexity of mining maximal frequent itemsets and maximal
 frequent patterns. In: Proceedings of the tenth ACM SIGKDD 
international conference on knowledge discovery and data mining. ACM, pp
 344–353<span class="Occurrences"><span class="Occurrence OccurrenceGS"><a target="_blank" rel="noopener" class="google-scholar-link gtm-reference" data-reference-type="Google Scholar" href="https://scholar.google.com/scholar?q=Yang%20G%20%282004%29%20The%20complexity%20of%20mining%20maximal%20frequent%20itemsets%20and%20maximal%20frequent%20patterns.%20In%3A%20Proceedings%20of%20the%20tenth%20ACM%20SIGKDD%20international%20conference%20on%20knowledge%20discovery%20and%20data%20mining.%20ACM%2C%20pp%20344%E2%80%93353"><span><span>Google Scholar</span></span></a></span></span></div></li></ol></div></section><section class="Section1 RenderAsSection1"><h2 class="Heading" id="copyrightInformation" data-role="collapsible-handle" tabindex="-1">Copyright information<span class="section-icon"></span></h2><div class="ArticleCopyright content"><div class="ArticleCopyright">©&nbsp;The Author(s)&nbsp;2016</div></div></section></div>
                        </article>
                        <aside class="section section--collapsible" id="AboutThisContent">
    <h2 class="section__heading" id="aboutcontent" data-role="collapsible-handle" tabindex="-1">About this article<span class="section-icon"></span></h2>
    <div class="section__content bibliographic-information">
                <div id="crossMark" class="crossmark">
            <a data-crossmark="10.1007%2Fs10618-016-0466-x" class="gtm-crossmark" target="_blank" rel="noopener" href="https://crossmark.crossref.org/dialog/?doi=10.1007%2Fs10618-016-0466-x" title="Verify currency and authenticity via CrossMark">
                <span class="u-screenreader-only">CrossMark</span>
                <svg class="CrossMark" id="crossmark-icon" width="57" height="81">
                    <image width="57" height="81" alt="Verify currency and authenticity via CrossMark" src="/springerlink-static/439917616/images/png/crossmark.png" xlink="http://www.w3.org/1999/xlink" xlink:href="/springerlink-static/439917616/images/svg/crossmark.svg"></image>
                </svg>
            </a>
        </div>

        <div class="crossmark__adjacent">
            <dl class="citation-info u-highlight-target u-mb-16" id="citeas" tabindex="-1">
    <dt class="test-cite-heading">
        Cite this article as:
    </dt>
    <dd id="citethis-text">Talukder, N. &amp; Zaki, M.J. Data Min Knowl Disc (2016) 30: 1024. https://doi.org/10.1007/s10618-016-0466-x</dd>
</dl>
                <ul class="bibliographic-information__list bibliographic-information__list--inline">
        <li class="bibliographic-information__item">
            <span class="bibliographic-information__title"><abbr title="Digital Object Identifier">DOI</abbr></span>
            <span class="bibliographic-information__value u-overflow-wrap" id="doi-url">https://doi.org/10.1007/s10618-016-0466-x</span>
        </li>
            <li class="bibliographic-information__item">
                <span class="bibliographic-information__title">Publisher Name</span>
                <span class="bibliographic-information__value" id="publisher-name">Springer US</span>
            </li>
            <li class="bibliographic-information__item">
                <span class="bibliographic-information__title">Print ISSN</span>
                <span class="bibliographic-information__value" id="print-issn">1384-5810</span>
            </li>
            <li class="bibliographic-information__item ">
                <span class="bibliographic-information__title">Online ISSN</span>
                <span class="bibliographic-information__value" id="electronic-issn">1573-756X</span>
            </li>

        
    </ul>

            <ul class="bibliographic-information__list">
        <li class="bibliographic-information__item">
            <a id="about-journal" class="bibliographic-information__misc-links gtm-about-this" title="Visit Springer.com for information about this article's journal" href="https://www.springer.com/journal/10618/about">About this journal</a>
        </li>
        <li class="bibliographic-information__item">
            <a id="reprintsandpermissions-link" target="_blank" rel="noopener" href="https://s100.copyright.com/AppDispatchServlet?publisherName=SpringerNature&amp;orderBeanReset=true&amp;orderSource=SpringerLink&amp;copyright=The+Author%28s%29&amp;author=N.+Talukder%2C+M.+J.+Zaki&amp;issueNum=5&amp;contentID=10.1007%2Fs10618-016-0466-x&amp;openAccess=false&amp;endPage=1052&amp;publicationDate=2016&amp;startPage=1024&amp;volumeNum=30&amp;title=A+distributed+approach+for+graph+mining+in+massive+networks&amp;imprint=The+Author%28s%29&amp;publication=1384-5810" title="Visit RightsLink for information about reusing this article">Reprints and Permissions</a>
        </li>
</ul>



        </div>
      
      
          
    </div>
</aside>

                        <div class="section section--collapsible uptodate-recommendations gtm-recommendations">
    <h2 class="uptodate-recommendations__title section__heading gtm-recommendations__title" id="uptodaterecommendations" data-role="collapsible-handle" tabindex="-1">Personalised recommendations<span class="section-icon"></span></h2>
    <div class="section__content">
        <div class="uptodate-recommendations__container">
             <iframe style="width: 100%; height: 411px; bottom: 0px; right: 0px; z-index: 100000;" scrolling="no" src="https://recommended.springernature.com/v2197/generated/client.html?source=https://link.springer.com" name="uptodate-client" id="uptodate-client" title="Personalised recommendations" allowtransparency="true" frameborder="0"></iframe>
        </div>
    </div>
</div>
                                <div id="doubleclick-native-ad" data-component="SpringerLink.GoogleAds" data-namespace="native"><div id="google_ads_iframe_270604982/springerlink/10618/article_1__container__" style="border: 0pt none;"><iframe id="google_ads_iframe_270604982/springerlink/10618/article_1" title="3rd party ad content" name="google_ads_iframe_270604982/springerlink/10618/article_1" scrolling="no" marginwidth="0" marginheight="0" style="border: 0px none; vertical-align: bottom;" srcdoc="" width="2" frameborder="0" height="2"></iframe></div></div>

                                    <div class="sticky-banner u-interface u-js-screenreader-only" aria-hidden="true" data-component="SpringerLink.StickyBanner" data-namespace="hasButton">
                <div class="sticky-banner__container"><span class="sticky-banner__title sticky-banner__title--short u-overflow-ellipsis" title="A distributed approach for graph mining in massive networks">A distributed approach for graph mining in massive networks</span>
                        <div class="citations c-button-dropdown" data-component="SV.Dropdown" data-namespace="citationsSticky" aria-label="button with dropdown options">
        <button id="button-Dropdown-citationsSticky-dropdown" type="button" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown-citationsSticky-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">
    <span>Cite</span>
    <span class="hide-text-small">article</span>
</span><span class="c-button-dropdown__icon"></span></button>
<div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown-citationsSticky-dropdown"><ul class="citations__content" data-role="button-dropdown__content">
    <li>
        <a href="#citeas" class="gtm-cite-dropdown" tabindex="-1">How to cite?</a>
    </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-016-0466-x?format=refman&amp;flavour=citation" title="Download this article's citation as a .RIS file" class="gtm-export-citation" data-gtmlabel="RIS" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="RIS">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .RIS
                </span>
                <span class="citations__types">
                        <span>
                            Papers
                        </span>
                        <span>
                            Reference Manager
                        </span>
                        <span>
                            RefWorks
                        </span>
                        <span>
                            Zotero
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-016-0466-x?format=endnote&amp;flavour=citation" title="Download this article's citation as a .ENW file" class="gtm-export-citation" data-gtmlabel="ENW" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="ENW">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .ENW
                </span>
                <span class="citations__types">
                        <span>
                            EndNote
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-016-0466-x?format=bibtex&amp;flavour=citation" title="Download this article's citation as a .BIB file" class="gtm-export-citation" data-gtmlabel="BIB" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="BIB">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .BIB
                </span>
                <span class="citations__types">
                        <span>
                            BibTeX
                        </span>
                        <span>
                            JabRef
                        </span>
                        <span>
                            Mendeley
                        </span>
                </span>
            </a>
        </li>
</ul></div>
    </div>

                            <div>
        <a class="c-button share-this gtm-shareby-sharelink-link test-shareby-sharelink-link" data-test="shareable-link" target="_blank" rel="noopener" href="https://link.springer.com/sharelink/10.1007/s10618-016-0466-x">
            <span>Share</span>
            <span class="hide-text-small">article</span>
        </a>
    </div>




                                    <div>
            <a href="https://link.springer.com/content/pdf/10.1007%2Fs10618-016-0466-x.pdf" target="_blank" class="c-button c-button--blue c-button__icon-right gtm-pdf-link" title="Download this article in PDF format" rel="noopener">
                <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" version="1.1"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g fill="#fff"><g transform="translate(12.000000, 5.000000)"><path d="M7 7.3L7 1C7 0.4 6.6 0 6 0 5.4 0 5 0.4 5 1L5 7.3 3.5 5.7C3.1 5.3 2.5 5.3 2.1 5.7L2.1 5.7C1.7 6.1 1.7 6.7 2.1 7.1L5.3 10.3C5.7 10.7 6.3 10.7 6.7 10.3L9.9 7.1C10.3 6.7 10.3 6.1 9.9 5.7L9.9 5.7C9.5 5.3 8.9 5.3 8.5 5.7L7 7.3 7 7.3ZM0 13C0 12.4 0.5 12 1 12L11 12C11.6 12 12 12.4 12 13 12 13.6 11.5 14 11 14L1 14C0.4 14 0 13.6 0 13L0 13Z"></path></g></g></g></svg>
                <span class="hide-text-small">Download</span>
                <span>PDF</span>
            </a>
        </div>

                </div>
            </div>




                    </div>
                    <aside class="main-sidebar-right u-interface">
                        <div data-role="sticky-wrapper">
                            <div class="main-sidebar-right__content u-composite-layer" data-component="SpringerLink.StickySidebar">
                                <div class="article-actions" id="article-actions">
                                    <h2 class="u-screenreader-only">Actions</h2>


                                    <div class="u-js-hide u-js-show-two-col">
                                        

                                                <div class="download-article test-pdf-link">
                                                            <div>
            <a href="https://link.springer.com/content/pdf/10.1007%2Fs10618-016-0466-x.pdf" target="_blank" class="c-button c-button--blue c-button__icon-right gtm-pdf-link" title="Download this article in PDF format" rel="noopener">
                <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" version="1.1"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g fill="#fff"><g transform="translate(12.000000, 5.000000)"><path d="M7 7.3L7 1C7 0.4 6.6 0 6 0 5.4 0 5 0.4 5 1L5 7.3 3.5 5.7C3.1 5.3 2.5 5.3 2.1 5.7L2.1 5.7C1.7 6.1 1.7 6.7 2.1 7.1L5.3 10.3C5.7 10.7 6.3 10.7 6.7 10.3L9.9 7.1C10.3 6.7 10.3 6.1 9.9 5.7L9.9 5.7C9.5 5.3 8.9 5.3 8.5 5.7L7 7.3 7 7.3ZM0 13C0 12.4 0.5 12 1 12L11 12C11.6 12 12 12.4 12 13 12 13.6 11.5 14 11 14L1 14C0.4 14 0 13.6 0 13L0 13Z"></path></g></g></g></svg>
                <span class="hide-text-small">Download</span>
                <span>PDF</span>
            </a>
        </div>

                                                </div>


                                            <div class="citations c-button-dropdown" data-component="SV.Dropdown" data-namespace="citations" aria-label="button with dropdown options">
        <button id="button-Dropdown-citations-dropdown" type="button" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="Dropdown-citations-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">
    <span>Cite</span>
    <span class="hide-text-small">article</span>
</span><span class="c-button-dropdown__icon"></span></button>
<div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="Dropdown-citations-dropdown"><ul class="citations__content" data-role="button-dropdown__content">
    <li>
        <a href="#citeas" class="gtm-cite-dropdown" tabindex="-1">How to cite?</a>
    </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-016-0466-x?format=refman&amp;flavour=citation" title="Download this article's citation as a .RIS file" class="gtm-export-citation" data-gtmlabel="RIS" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="RIS">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .RIS
                </span>
                <span class="citations__types">
                        <span>
                            Papers
                        </span>
                        <span>
                            Reference Manager
                        </span>
                        <span>
                            RefWorks
                        </span>
                        <span>
                            Zotero
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-016-0466-x?format=endnote&amp;flavour=citation" title="Download this article's citation as a .ENW file" class="gtm-export-citation" data-gtmlabel="ENW" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="ENW">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .ENW
                </span>
                <span class="citations__types">
                        <span>
                            EndNote
                        </span>
                </span>
            </a>
        </li>
        <li>
            <a href="https://citation-needed.springer.com/v2/references/10.1007/s10618-016-0466-x?format=bibtex&amp;flavour=citation" title="Download this article's citation as a .BIB file" class="gtm-export-citation" data-gtmlabel="BIB" tabindex="-1">
                <span class="citations__extension" data-gtmlabel="BIB">
                    <svg class="u-vertical-align-absolute" width="12" height="14" viewBox="0 0 12 14" xmlns="http://www.w3.org/2000/svg"><path d="M7 7.269v-6.271c0-.551-.448-.998-1-.998-.556 0-1 .447-1 .998v6.271l-1.5-1.547c-.375-.387-1.01-.397-1.401-.006l.016-.016c-.397.397-.391 1.025-.001 1.416l3.178 3.178c.392.392 1.024.391 1.415 0l3.178-3.178c.392-.392.391-1.025-.001-1.416l.016.016c-.397-.397-1.018-.388-1.401.006l-1.5 1.547zm-7 5.731c0-.552.456-1 1.002-1h9.995c.554 0 1.002.444 1.002 1 0 .552-.456 1-1.002 1h-9.995c-.554 0-1.002-.444-1.002-1z" fill="#0176C3"></path></svg>
                    .BIB
                </span>
                <span class="citations__types">
                        <span>
                            BibTeX
                        </span>
                        <span>
                            JabRef
                        </span>
                        <span>
                            Mendeley
                        </span>
                </span>
            </a>
        </li>
</ul></div>
    </div>

                                                <div>
        <a class="c-button share-this gtm-shareby-sharelink-link test-shareby-sharelink-link" data-test="shareable-link" target="_blank" rel="noopener" href="https://link.springer.com/sharelink/10.1007/s10618-016-0466-x">
            <span>Share</span>
            <span class="hide-text-small">article</span>
        </a>
    </div>




                                    </div>
                                </div>
                                <nav class="toc" aria-label="article contents">
    <h2 class="u-h4 u-screenreader-only">Table of contents</h2>
    <ul id="article-contents" class="article-contents" role="menu" tabindex="-1">
            <li role="menuitem">
                <a title="Article" href="#enumeration"><span class="u-overflow-ellipsis">Article</span></a>
            </li>
            <li role="menuitem">
                <a title="Abstract" href="#Abs1"><span class="u-overflow-ellipsis">Abstract</span></a>
            </li>
            <li role="menuitem">
                <a title="1 Introduction" href="#Sec1"><span class="u-overflow-ellipsis">1 Introduction</span></a>
            </li>
            <li role="menuitem">
                <a title="2 Background" href="#Sec2"><span class="u-overflow-ellipsis">2 Background</span></a>
            </li>
            <li role="menuitem">
                <a title="3 Related work" href="#Sec3"><span class="u-overflow-ellipsis">3 Related work</span></a>
            </li>
            <li role="menuitem">
                <a title="4 DistGraph: distributed mining on a partitioned input graph" href="#Sec4"><span class="u-overflow-ellipsis">4 DistGraph: distributed mining on a partitioned input graph</span></a>
            </li>
            <li role="menuitem">
                <a title="5 DistGraph: algorithm details" href="#Sec8"><span class="u-overflow-ellipsis">5 DistGraph: algorithm details</span></a>
            </li>
            <li role="menuitem">
                <a title="6 Experiments" href="#Sec15"><span class="u-overflow-ellipsis">6 Experiments</span></a>
            </li>
            <li role="menuitem">
                <a title="7 Conclusions" href="#Sec18"><span class="u-overflow-ellipsis">7 Conclusions</span></a>
            </li>
            <li role="menuitem">
                <a title="Notes" href="#Notes"><span class="u-overflow-ellipsis">Notes</span></a>
            </li>
            <li role="menuitem">
                <a title="References" href="#Bib1"><span class="u-overflow-ellipsis">References</span></a>
            </li>
            <li role="menuitem">
                <a title="Copyright information" href="#copyrightInformation"><span class="u-overflow-ellipsis">Copyright information</span></a>
            </li>
            
            <li role="menuitem">
                <a title="About this article" href="#aboutcontent"><span class="u-overflow-ellipsis">About this article</span></a>
            </li>
    </ul>
</nav>

                            </div>
                                <div class="skyscraper-ad u-hide" data-component="SpringerLink.GoogleAds" data-namespace="skyscraper">
        <div class="skyscraper-ad__wrapper">
            <p class="skyscraper-ad__label">Advertisement</p>
            <button class="skyscraper-ad__hide" title="Hide this advertisement">Hide</button>
            <div id="doubleclick-ad" class="skyscraper-ad__ad"></div>
        </div>
    </div>

                        </div>
                    </aside>
                </div>
            </main>
                <footer class="footer u-interface">
        <div class="footer__aside-wrapper">
            <div class="footer__content">
                <div class="footer__aside">
                    <p class="footer__strapline">Over 10 million scientific documents at your fingertips</p>
                                <div class="footer__edition c-button-dropdown c-button-dropdown--ghost" data-component="SpringerLink.EditionSwitcher" aria-label="button with dropdown options">
                                    <button id="button-EditionSwitcher-0501111430-dropdown" type="button" title="Switch between Academic &amp; Corporate Edition" class="c-button-dropdown__button" data-role="button-dropdown__control" aria-pressed="false" aria-expanded="false" aria-controls="EditionSwitcher-0501111430-dropdown"><span class="u-overflow-ellipsis c-button-dropdown__button-title">Academic Edition</span><span class="c-button-dropdown__icon"></span></button>
                                    <div class="u-composite-layer c-button-dropdown__container" aria-hidden="true" aria-label="dropdown" id="EditionSwitcher-0501111430-dropdown"><ul data-role="button-dropdown__content">
                                        <li class="selected"><a href="https://link.springer.com/siteEdition/link?previousUrl=/article/10.1007/s10618-016-0466-x&amp;id=siteedition-academic-link" id="siteedition-academic-link">Academic Edition</a></li>
                                        <li><a href="https://link.springer.com/siteEdition/rd?previousUrl=/article/10.1007/s10618-016-0466-x&amp;id=siteedition-corporate-link" id="siteedition-corporate-link" tabindex="-1">Corporate Edition</a></li>
                                    </ul></div>
                                </div>
                </div>
            </div>
        </div>
        <div class="footer__content">
            <ul class="footer__nav">
                <li>
                    <a href="https://link.springer.com/">Home</a>
                </li>
                <li>
                    <a href="https://link.springer.com/impressum">Impressum</a>
                </li>
                <li>
                    <a href="https://link.springer.com/termsandconditions">Legal information</a>
                </li>
                <li>
                    <a href="https://link.springer.com/privacystatement">Privacy statement</a>
                </li>
                <li>
                    <a href="https://link.springer.com/cookiepolicy">How we use cookies</a>
                </li>
                <li>
                    <a href="https://link.springer.com/accessibility" class="gtm-footer-accessibility">Accessibility</a>
                </li>
                <li>
                    <a id="contactus-footer-link" href="https://link.springer.com/contactus">Contact us</a>
                </li>
            </ul>
            <a class="parent-logo" target="_blank" rel="noopener" href="https://www.springernature.com/" title="Go to Springer Nature">
                <span class="u-screenreader-only">Springer Nature</span>
                <svg width="125" height="12" focusable="false" aria-hidden="true">
                    <image width="125" height="12" src="/springerlink-static/439917616/images/png/springernature.png" xlink="http://www.w3.org/1999/xlink" xlink:href="/springerlink-static/439917616/images/svg/springernature.svg">
                    </image>
                </svg>
            </a>

            <p class="footer__copyright">© 2017 Springer Nature Switzerland AG. Part of <a target="_blank" rel="noopener" href="https://www.springernature.com/">Springer Nature</a>.</p>
                <p class="footer__user-access-info">
                    <span>Not logged in</span>
                    <span>CAPES MEC (3000197460) - Universidade Tecnologica Federal do Parana (3000201946)</span>
                    <span>200.17.97.45</span>
                </p>
        </div>
    </footer>

        </div>
          <script type="text/javascript">
    (function() {
        document.addEventListener('readystatechange', function () {
            if (document.readyState === 'complete') {
                var linkEl = document.querySelector('.js-ctm');
                if (window.matchMedia && window.matchMedia(linkEl.media).matches) {
                    var scriptMathJax = document.createElement('script');
                    scriptMathJax.async = true;
                    scriptMathJax.src = '/springerlink-static/439917616/js/mathJax.js';
                    var s0 = document.getElementsByTagName('script')[0];
                    s0.parentNode.insertBefore(scriptMathJax, s0);
                }
            }
        });
    })();
</script>

<script type="text/javascript">

    (function() {
        var linkEl = document.querySelector('.js-ctm');
        var scriptsList = [];
        var polyfillFeatures = '';

        window.SpringerLink = window.SpringerLink || {};
        window.SpringerLink.staticLocation = '/springerlink-static/439917616';

        if (window.matchMedia && window.matchMedia(linkEl.media).matches) {
            (function(h){h.className = h.className.replace('no-js', 'js')})(document.documentElement);
            
            polyfillFeatures = 'default,fetch,Promise,Object.setPrototypeOf,Object.entries,Number.isInteger,MutationObserver,startsWith,Array.prototype.includes,Array.from,IntersectionObserver';
            
            scriptsList = [
                'https://cdn.polyfill.io/v2/polyfill.js?features=' + polyfillFeatures,
                window.SpringerLink.staticLocation + '/js/main.js'
            ];

            scriptsList.forEach(function(script) {
                var tag = document.createElement('script');
                
                tag.async = false;
                tag.src = script;

                document.body.appendChild(tag);
            });
        }
    })();
</script><script src="polyfill.js"></script><script src="main.js"></script>

<script type="text/javascript" class="optanon-category-2">

   function viewport() {
        if (document.documentElement.clientWidth < 620) {
            size = 'small';
        }
        else if(document.documentElement.clientWidth < 1075 ) {
            size = 'medium';
        }
        else {
            size = 'wide';
        }
        return size;
    };

    function reportForMouseEvents(linkCssSelectors, nolardUrl, experiment, abgroup) {
        var counter = 0;
        linkCssSelectors.forEach(function(cssSelector) {
            $('body').delegate(cssSelector, 'click', function() {
                if(counter == 0) {
                    reportConversion(nolardUrl, experiment, abgroup);
                    counter++;
                }
            });
        });
    };

    function reportConversion(nolardUrl, experiment, abgroup) {
        $.ajax({ url: nolardUrl + '/convert/' + experiment + '/' + abgroup });
    };

    function reportForMouseEvent(linkCssSelector, nolardUrl, experiment, abgroup) {
        var counter = 0;
        var elem = document.querySelector(linkCssSelector)

        if (elem.addEventListener) {
            elem.addEventListener("click", function(e) {
                if(counter == 0) {
                    reportConversion(nolardUrl, experiment, abgroup);
                    counter++;
                }
            });
        } else  {
            elem.attachEvent("click", function(e) {
                if(counter == 0) {
                    reportConversion(nolardUrl, experiment, abgroup);
                    counter++;
                }
            });
        }
    };

    function reportParticipation(nolardUrl, experiment, abgroup, participations) {
        if (participations)
            participations.push(experiment + '/' + abgroup)

        var xhr = new XMLHttpRequest()
        xhr.open('GET', nolardUrl + '/participate/' + experiment + '/' + abgroup);
        xhr.send();
    };

</script>
    <script type="text/javascript" id="googletag-push">
        
            var adSlot = '270604982/springerlink/10618/article';
        

        var definedSlots = [
                {slot: [728, 90], containerName: 'doubleclick-leaderboard-ad'},
                {slot: [160, 600], containerName: 'doubleclick-ad'},
            {slot: [2, 2], containerName: 'doubleclick-native-ad'}
        ];
    </script>


        
        <span id="chat-widget" class="u-hide"></span>
                    <noscript>
                <img aria-hidden="true" role="presentation" src="https://ssl-springer.met.vgwort.de/na/pw-vgzm.415900-10.1007-s10618-016-0466-x" width='1' height='1' alt='' />
            </noscript>

    

<div style="display: none; visibility: hidden;"><script id="gpt-control-script" src="gpt.js"></script>

<script id="gpt-control-setup">var googletag=googletag||{};googletag.cmd=googletag.cmd||[];</script></div><script type="text/javascript" id="gtm-in-page-section-view">(function(d,z){function n(){for(var a in dataLayer)if(dataLayer.hasOwnProperty(a)&&dataLayer[a]["Event Category"])return dataLayer[a]["Event Category"];return"warning: no event category"}function v(a,b,c,h){"undefined"!==typeof dataLayer?dataLayer.push({event:"Scroll To Section",eventCategory:n(),eventAction:a,eventLabel:b,eventValue:1,eventNonInteraction:h}):("undefined"!==typeof ga&&ga("send","event",n(),a,b,1,{nonInteraction:1}),"undefined"!==typeof _gaq&&_gaq.push(["_trackEvent",n(),a,b,1,!0]))}
function k(b,g,c,h){if(-1===a[c].cache.indexOf(c+"-"+h)&&document.querySelectorAll(b).length){var f="viewed"===h?document.querySelectorAll(b)[0].getBoundingClientRect().height:0,x=document.querySelectorAll(b)[0].getBoundingClientRect().top+d.pageYOffset;g>=x+f&&(v(p(c)+" "+p(h),b,c,!1),a[c].cache.push(c+"-"+h),q++)}}function r(a){a=document.querySelectorAll(a);return a.length?a[0].offsetHeight||0:0}function m(b,g,c){var f=d.pageYOffset+d.innerHeight;if(q>=w)"handleClick"===g&&d.removeEventListener(b,
t),"throttle"===g&&d.removeEventListener(b,u);else if(c)setTimeout(function(){0<r(a[c].content)&&(a[c].reached&&k(a[c].content,f,c,"reached"),a[c].viewed&&k(a[c].content,f,c,"viewed"))},10);else for(var e in a)a.hasOwnProperty(e)&&0<r(a[e].content)&&(a[e].reached&&k(a[e].content,f,e,"reached"),a[e].viewed&&k(a[e].content,f,e,"viewed"))}function y(b){b=document.querySelectorAll(a[b].content).length;return 0<b?!0:!1}function p(a){return a.charAt(0).toUpperCase()+a.slice(1)}function u(a,b){var c,f,e,
g=null,d=0,k=function(){d=new Date;g=null;e=a.apply(c,f)};return function(){var h=new Date;d||(d=h);var l=b-(h-d);c=this;f=arguments;0>=l?(clearTimeout(g),g=null,d=h,e=a.apply(c,f)):g||(g=setTimeout(k,l));return e}}function t(a,b){return function(c){a(c,b)}}var w=0,q=0,a={recommendations:{content:".gtm-recommendations iframe",clickable:".gtm-recommendations .gtm-recommendations__title",exists:!1,reached:!1,viewed:!0,size:0,cache:[]},"abstract":{content:"Section.Abstract",clickable:null,exists:!0,
reached:!1,viewed:!0,size:0,cache:[]},references:{content:".Bibliography \x3e .content, [id^\x3dBib1] \x3e .content",clickable:".Bibliography \x3e .Heading, [id^\x3dBib1] \x3e .Heading",exists:!0,reached:!0,viewed:!0,size:0,cache:[]},about:{content:"#AboutThisContent",clickable:"#AboutThisContent \x3e #aboutcontent",exists:!1,reached:!1,viewed:!0,size:0,cache:[]}};d.addEventListener("scroll",u(function(){m("scroll","throttle")},500));d.addEventListener("orientationchange",u(function(){m("orientationchange",
"throttle")},500));for(var b in a)if(a.hasOwnProperty(b)&&(a[b].exists&&a[b].size++,a[b].reached&&a[b].size++,a[b].viewed&&a[b].size++,w+=a[b].size,a[b].exists&&y(b)||!a[b].exists)){a[b].exists&&-1===a[b].cache.indexOf(b+"-exists")&&(v(p(b)+" Exists",a[b].content,b,!0),a[b].cache.push(b+"-exists"),q++);if(0<r(a[b].content)){var l=d.pageYOffset+d.innerHeight;a[b].reached&&k(a[b].content,l,b,"reached");a[b].viewed&&k(a[b].content,l,b,"viewed")}a[b].clickable&&(l=document.querySelectorAll(a[b].clickable)[0])&&
(l.addEventListener("click",t(function(a,b){m("click","handleClick",b)},b)),l.addEventListener("click",t(function(a,b){13==a.keyCode&&m("click","handleClick",b)},b)))}})(window);</script><script type="text/javascript" id="gtm-recommendations-script" src="entry-point"></script><script type="text/javascript" id="gtm-polar" src="polar.js"></script><script type="text/javascript" id="gtm-krux-control">window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);(function(){var a=document.createElement("script");a.type="text/javascript";a.async=!0;a.src=("https:"===location.protocol?"https:":"http:")+"//cdn.krxd.net/controltag/KDqyaFZ_.js";var b=document.getElementsByTagName("script")[0];b.parentNode.insertBefore(a,b)})();</script><script type="text/javascript" id="">window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);(function(){function c(b){b="kxmacmillan_"+b;try{var a=window.localStorage}catch(d){a=null}return a?a[b]||"":navigator.cookieEnabled?(a=document.cookie.match(b+"\x3d([^;]*)"))&&unescape(a[1])||"":""}Krux.user=c("user");Krux.segments=c("segs")?c("segs").split(","):[]})();</script><script type="text/javascript" id="gtm-krux-consent">var allowed=google_tag_manager["GTM-WCF9Z9"].macro(29);allowed=!0===allowed?1:0;Krux("consent:set",{dc:allowed,al:allowed,tg:allowed,cd:!1,sh:!1,re:!1},function(a,b){a?console.error(a):console.log("consent flags set ",b)});</script><script type="text/javascript" id="gtm-gpt-setup">window.Krux||((Krux=function(){Krux.q.push(arguments)}).q=[]);
googletag.cmd.push(function(){googletag.pubads().setTargeting("doi",google_tag_manager["GTM-WCF9Z9"].macro(30));googletag.pubads().setTargeting("kwrd",google_tag_manager["GTM-WCF9Z9"].macro(31));googletag.pubads().setTargeting("pmc",google_tag_manager["GTM-WCF9Z9"].macro(32));googletag.pubads().setTargeting("BPID",google_tag_manager["GTM-WCF9Z9"].macro(33));googletag.pubads().setTargeting("edition",google_tag_manager["GTM-WCF9Z9"].macro(34));googletag.pubads().setTargeting("sucode",google_tag_manager["GTM-WCF9Z9"].macro(35));googletag.pubads().setTargeting("eissn",google_tag_manager["GTM-WCF9Z9"].macro(36));googletag.pubads().setTargeting("pissn",
google_tag_manager["GTM-WCF9Z9"].macro(37));googletag.pubads().setTargeting("eisbn",google_tag_manager["GTM-WCF9Z9"].macro(38));googletag.pubads().setTargeting("pisbn",google_tag_manager["GTM-WCF9Z9"].macro(39));googletag.pubads().setTargeting("logged",google_tag_manager["GTM-WCF9Z9"].macro(40));googletag.pubads().setTargeting("EventCategory",google_tag_manager["GTM-WCF9Z9"].macro(41));googletag.pubads().setTargeting("ksg",Krux.segments);googletag.pubads().setTargeting("kuid",Krux.uid);googletag.pubads().setRequestNonPersonalizedAds(google_tag_manager["GTM-WCF9Z9"].macro(45)?0:1);googletag.pubads().disableInitialLoad();
googletag.enableServices()});</script><div id="optanon" class="modern"><div id="optanon-popup-bg"></div><div id="optanon-popup-wrapper" role="dialog" aria-modal="true" tabindex="-1" lang="en-GB"><div id="optanon-popup-top"><a href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Close Button');" class="optanon-close-link optanon-close optanon-close-ui" title="Close Preference Centre"><div id="optanon-close" style="background: url(https://optanon.blob.core.windows.net/skins/default_flat_bottom_two_button_white/v2/images/optanon-pop-up-close.png);width:34px;height:34px;"></div></a></div><div id="optanon-popup-body"><div id="optanon-popup-body-left"><div id="optanon-popup-body-left-shading"></div><div id="optanon-branding-top-logo" style="background-image: url(https://optanon.blob.core.windows.net/logos/5138/5138:link.springer.com/springer.png) !important;"></div><ul id="optanon-menu"><li class="menu-item-on menu-item-about" title="Your Privacy"><p><a href="#">Your Privacy</a></p></li><li class="menu-item-necessary menu-item-on" title="Strictly Necessary Cookies"><p><a href="#">Strictly Necessary Cookies</a></p></li><li class="menu-item-on menu-item-performance" title="Performance Cookies"><p><a href="#">Performance Cookies</a></p></li><li class="menu-item-on menu-item-functional" title="Functional Cookies"><p><a href="#">Functional Cookies</a></p></li><li class="menu-item-on menu-item-advertising" title="Targeting Cookies"><p><a href="#">Targeting Cookies</a></p></li><li class="menu-item-moreinfo menu-item-off" title="More Information"><p><a target="_blank" href="https://link.springer.com/cookiepolicy" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Cookie Policy');">More Information</a></p></li></ul></div><div id="optanon-popup-body-right"><h2 aria-label="true">Privacy Preference Centre</h2><h3></h3><div id="optanon-popup-more-info-bar"><div class="optanon-status"><div class="optanon-status-editable"><form><fieldset><p><input value="check" id="chkMain" checked="checked" class="optanon-status-checkbox" type="checkbox"><label for="chkMain">Active</label></p></fieldset></form></div><div class="optanon-status-always-active optanon-status-on"><p>Always Active</p></div></div></div><div id="optanon-main-info-text"></div></div><div class="optanon-bottom-spacer"></div></div><div id="optanon-popup-bottom"> <a href="https://onetrust.com/poweredbyonetrust" target="_blank"><div id="optanon-popup-bottom-logo" style="background: url(https://optanon.blob.core.windows.net/skins/default_flat_bottom_two_button_white/v2/images/cookie-collective-top-bottom.png);width:155px;height:35px;" title="powered by OneTrust"></div></a><div class="optanon-button-wrapper optanon-save-settings-button optanon-close optanon-close-consent"><div class="optanon-white-button-left"></div><div class="optanon-white-button-middle"><a href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Save Settings');">Save Settings</a></div><div class="optanon-white-button-right"></div></div><div class="optanon-button-wrapper optanon-allow-all-button optanon-allow-all" style="display: none;"><div class="optanon-white-button-left"></div><div class="optanon-white-button-middle"><a href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Preferences Allow All');">Allow All</a></div><div class="optanon-white-button-right"></div></div></div></div></div><div class="optanon-alert-box-wrapper  " style="bottom: 0px;"><div class="optanon-alert-box-bottom-top"><div class="optanon-alert-box-corner-close"><a class="optanon-alert-box-close" href="#" title="Close Banner" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Banner Close Button');"></a></div></div><div class="optanon-alert-box-bg"><div class="optanon-alert-box-logo"> </div><div class="optanon-alert-box-body"><p>We
 use cookies to personalise content and ads, to provide social media 
features and to analyse our traffic. We also share information about 
your use of our site with our social media, advertising and analytics 
partners in accordance with our <a href="https://link.springer.com/privacystatement">Privacy Statement</a>. You can manage your preferences in Manage Cookies.</p></div><div class="optanon-clearfix"></div><div class="optanon-alert-box-button-container"><div class="optanon-alert-box-button optanon-button-close"><div class="optanon-alert-box-button-middle"><a class="optanon-alert-box-close" href="#">Close</a></div></div><div class="optanon-alert-box-button optanon-button-allow"><div class="optanon-alert-box-button-middle"><a class="optanon-allow-all" href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Banner Accept Cookies');">OK</a></div></div><div class="optanon-alert-box-button optanon-button-more"><div class="optanon-alert-box-button-middle"><a class="optanon-toggle-display" href="#" onclick="Optanon.TriggerGoogleAnalyticsEvent('OneTrust Cookie Consent', 'Banner Open Preferences');">Manage Cookies</a></div></div></div><div class="optanon-clearfix optanon-alert-box-bottom-padding"></div></div></div><div class="kxhead" data-id="KDqyaFZ_" style="display:none !important;"><span class="kxtag kxinvisible" data-id="28134"><script>
(function() {

    // cient only wants to collect first two paths
    Krux('scrape', {
        'page_attr_Springer_Link_url_path_1': {
            'url_path': '1'
        }
    });
    Krux('scrape', {
        'page_attr_Springer_Link_url_path_2': {
            'url_path': '2'
        }
    });
    // Krux('scrape',{'page_attr_url_path_3':{'url_path':'3'}});
    // Krux('scrape',{'page_attr_meta_keywords':{meta_name:'keywords'}});

    // Krux('scrape',{'page_attr_domain':{url_domain: '2'}});

})();
</script></span><span class="kxtag kxinvisible" data-id="20651"><script type="text/javascript">Krux('social.init');</script></span><span class="kxtag kxinvisible" data-id="24671"><script>
(function() {

    Krux('scrape', {
        'page_attr_Citation_Title_Springer': {
            'meta_name': 'citation_title'
        },
        'page_attr_Citation_DOI_Springer': {
            'meta_name': 'citation_doi'
        },
        'page_attr_Product_Name_Springer': {
            'meta_name': 'citation_journal_title'
        },
        'page_attr_Article_Type_Springer': {
            'meta_name': 'citation_article_type'
        },
        'page_attr_Citation_Author_Springer': {
            'dom_multi': 'meta[name="citation_author"]:@content'
        }
    });

    if (typeof dataLayer !== 'undefined' && dataLayer[0]) {
        Krux('set', 'page_attr_Article_Keywords_Springer', dataLayer[0].Keywords);
        Krux('set', 'page_attr_Business_Partner_ID_Springer', dataLayer[0].Bpids);
        Krux('set', 'page_attr_Business_Partner_Names_Springer', dataLayer[0].Bpnames);
        Krux('set', 'page_attr__Subject_Codes_Srpinger', dataLayer[0].SubjectCodes);
    }

})();
</script></span></div></body><iframe id="kx-proxy-KDqyaFZ_" src="https://cdn.krxd.net/partnerjs/xdi/proxy.3d2100fd7107262ecb55ce6847f01fa5.html#%21kxcid=KDqyaFZ_&amp;kxt=https%3A%2F%2Flink.springer.com&amp;kxcl=cdn&amp;kxp=" style="display: none; visibility: hidden; height: 0; width: 0;"></iframe></html>